{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import matplotlib.pyplot as plt\n",
    "import operator\n",
    "import os\n",
    "import sys\n",
    "import typing\n",
    "import csv\n",
    "from collections import defaultdict\n",
    "from scipy import stats\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "\n",
    "from Bio import SeqIO\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.append(str(Path(os.path.dirname(os.getcwd())).parent))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from generate_summary_file import generate_summary\n",
    "from analysis.input_testing_data.generate_input_testing_data_for_modules import generate_testing_data_for_ecoli_and_bacillus\n",
    "from modules.main import run_modules\n",
    "from modules.main import run_input_processing\n",
    "from modules.main import EvaluationModule\n",
    "from modules.run_summary import RunSummary\n",
    "from modules import shared_functions_and_vars\n",
    "from modules import models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_user_input(optimization_method: str,\n",
    "                   is_ecoli_optimized: bool,\n",
    "                   orf_sequence: typing.Optional[str] = None,\n",
    "                   orf_sequence_file: typing.Optional[str] = None,\n",
    "                   output_path: typing.Optional[str] = None,\n",
    "                   optimization_cub_index: str = \"CAI\",\n",
    "                   tuning_param: float = 0.5,\n",
    "                   should_use_mrna_levels = True,\n",
    "                   should_use_protein_abundance = False):\n",
    "    default_user_inp_raw = generate_testing_data_for_ecoli_and_bacillus(\n",
    "        orf_optimization_method=optimization_method,\n",
    "        orf_optimization_cub_index=optimization_cub_index,\n",
    "        clusters_count=1,\n",
    "        tuning_param=tuning_param,\n",
    "        is_ecoli_optimized=is_ecoli_optimized,\n",
    "        sequence=orf_sequence,\n",
    "        sequence_file_path=orf_sequence_file,\n",
    "        output_path=os.path.join(\"results\", output_path),\n",
    "        should_use_mrna_levels=should_use_mrna_levels,\n",
    "        should_use_protein_abundance=should_use_protein_abundance,\n",
    "    )\n",
    "    return run_input_processing(default_user_inp_raw) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def evaluate_score_for_sequence(sequence: str,\n",
    "                                user_input: models.ModuleInput,\n",
    "                                optimization_method: str,\n",
    "                                is_ecoli_optimized: bool,\n",
    "                                orf_sequence: typing.Optional[str] = None,\n",
    "                                orf_sequence_file: typing.Optional[str] = None,\n",
    "                                output_path: typing.Optional[str] = None,\n",
    "                                optimization_cub_index: str = \"CAI\",\n",
    "                                tuning_param: float = 0.5,\n",
    "                                should_use_mrna_levels = True,\n",
    "                                should_use_protein_abundance = False):\n",
    "    run_summary = RunSummary()\n",
    "    return EvaluationModule.run_module(final_sequence=sequence, \n",
    "                                       user_input=user_input,\n",
    "                                       optimization_cub_index=user_input.optimization_cub_index,\n",
    "                                       run_summary=run_summary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_single_method_for_orf_sequence(optimization_method: str,\n",
    "                                       is_ecoli_optimized: bool,\n",
    "                                       orf_sequence: typing.Optional[str] = None,\n",
    "                                       orf_sequence_file: typing.Optional[str] = None,\n",
    "                                       output_path: typing.Optional[str] = None,\n",
    "                                       optimization_cub_index: str = \"CAI\",\n",
    "                                       tuning_param: float = 0.5,\n",
    "                                       should_use_mrna_levels = True,\n",
    "                                       should_use_protein_abundance = False,\n",
    "                                       initiation_optimization_method = \"original\"):\n",
    "    default_user_inp_raw = generate_testing_data_for_ecoli_and_bacillus(\n",
    "        orf_optimization_method=optimization_method,\n",
    "        orf_optimization_cub_index=optimization_cub_index,\n",
    "        clusters_count=1,\n",
    "        tuning_param=tuning_param,\n",
    "        is_ecoli_optimized=is_ecoli_optimized,\n",
    "        sequence=orf_sequence,\n",
    "        sequence_file_path=orf_sequence_file,\n",
    "        output_path=os.path.join(\"results\", output_path),\n",
    "        should_use_mrna_levels=should_use_mrna_levels,\n",
    "        should_use_protein_abundance=should_use_protein_abundance,\n",
    "        initiation_optimization_method=initiation_optimization_method,\n",
    "    )\n",
    "    return run_modules(default_user_inp_raw)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_all_methods(orf_sequence = None, orf_sequence_file = None, output_path = None, tuning_param: int = 0.5, should_use_mrna_levels: bool = True, should_use_protein_abundance: bool = False, initiation_optimization_method: str = \"original\"):\n",
    "    for optimization_method in [\n",
    "        \"single_codon_ratio\",\n",
    "        \"single_codon_diff\",\n",
    "        # \"single_codon_weakest_link\",\n",
    "        # \"zscore_single_aa_ratio\",\n",
    "        \"zscore_bulk_aa_ratio\",\n",
    "        # \"zscore_single_aa_diff\",\n",
    "        \"zscore_bulk_aa_diff\",\n",
    "        # \"zscore_single_aa_weakest_link\",\n",
    "        # \"zscore_bulk_aa_weakest_link\",\n",
    "    ]:\n",
    "        for optimization_cub_index in [\n",
    "            \"CAI\", \n",
    "            \"tAI\",\n",
    "        ]:\n",
    "            for direction in [True, False]:\n",
    "                run_single_method_for_orf_sequence(optimization_method=optimization_method,\n",
    "                                                   optimization_cub_index=optimization_cub_index,\n",
    "                                                   is_ecoli_optimized=direction,\n",
    "                                                   orf_sequence=orf_sequence,\n",
    "                                                   orf_sequence_file=orf_sequence_file,\n",
    "                                                   output_path=output_path,\n",
    "                                                   tuning_param=tuning_param,\n",
    "                                                   should_use_mrna_levels=should_use_mrna_levels,\n",
    "                                                   should_use_protein_abundance=should_use_protein_abundance,\n",
    "                                                   initiation_optimization_method=initiation_optimization_method)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_sequences_fasta_file(root_dir) -> None:\n",
    "    filename = \"run_summary.json\"\n",
    "\n",
    "    sequences = []\n",
    "    sequences_names = []\n",
    "    for root, dirs, files in os.walk(root_dir):\n",
    "        for file in files:\n",
    "            if file == filename:\n",
    "                directory_name = Path(root).name\n",
    "                file_path = os.path.join(root, file)\n",
    "                with open(file_path, \"r\") as summary_file:\n",
    "                    results_json = json.load(summary_file)\n",
    "\n",
    "                seq = results_json[\"final_evaluation\"][\"final_sequence\"]\n",
    "                sequences.append(seq)\n",
    "                sequences_names.append(directory_name[:-5])\n",
    "    \n",
    "    file_name = os.path.join(root_dir, \"mcherry_variants\")\n",
    "    shared_functions_and_vars.write_fasta(file_name, sequences, sequences_names)\n",
    "    \n",
    "    fields = ['variant', 'sequence']\n",
    "    variant_dict = [{\"variant\": sequences_names[i], \"sequence\": sequences[i]} for i in range(len(sequences))]\n",
    "    with open(file_name+\".csv\", 'w') as csvfile:\n",
    "        writer = csv.DictWriter(csvfile, fieldnames=fields)\n",
    "        writer.writeheader()\n",
    "        writer.writerows(variant_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_codon_frequencies(referece_cds: typing.Sequence[str]):\n",
    "    codons_counter = defaultdict(int)\n",
    "    \n",
    "    for cds in referece_cds:\n",
    "        if len(cds) % 3 != 0:\n",
    "            continue\n",
    "        for i in range(0, len(cds), 3):\n",
    "            codon = cds[i:i+3]\n",
    "            codons_counter[codon] += 1\n",
    "    \n",
    "    # Calculate relative frequncies\n",
    "    for amino_acid, codons in shared_functions_and_vars.synonymous_codons.items():\n",
    "        total_amino_acid_codons = 0\n",
    "        for amino_acid_codon in codons:\n",
    "            total_amino_acid_codons += codons_counter[amino_acid_codon]\n",
    "        \n",
    "        # In case a certain amino acid is missing from the reference cds collection, no need to normalize\n",
    "        if total_amino_acid_codons == 0:\n",
    "            continue\n",
    "        \n",
    "        for amino_acid_codon in codons:\n",
    "            codons_counter[amino_acid_codon] /= total_amino_acid_codons\n",
    "    \n",
    "    codons = list(codons_counter.keys())\n",
    "    df_dict = {\"codon\": codons, \"amino_acid\": [shared_functions_and_vars.nt_to_aa[codon] for codon in codons], \"frequency\": [codons_counter[codon] for codon in codons]}\n",
    "\n",
    "    return pd.DataFrame(df_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_codon_frequencies_lamda(referece_cds: typing.Sequence[str]):\n",
    "    codons_counter = defaultdict(int)\n",
    "    \n",
    "    for cds in referece_cds:\n",
    "        if len(cds) % 3 != 0:\n",
    "            continue\n",
    "        for i in range(0, len(cds), 3):\n",
    "            codon = cds[i:i+3]\n",
    "            codons_counter[codon] += 1\n",
    "    \n",
    "    # Calculate relative frequncies\n",
    "    for amino_acid, codons in shared_functions_and_vars.synonymous_codons.items():\n",
    "        amino_acid_codon_to_count = {codon: codons_counter[codon] for codon in codons}\n",
    "        most_frequent_codon = max(amino_acid_codon_to_count, key=amino_acid_codon_to_count.get)\n",
    "        \n",
    "        if amino_acid_codon_to_count[most_frequent_codon] == 0:\n",
    "            # In case a certain amino acid is missing from the reference cds collection, no need to normalize\n",
    "            continue\n",
    "            \n",
    "        for amino_acid_codon in codons:\n",
    "            codons_counter[amino_acid_codon] = 1 if amino_acid_codon == most_frequent_codon else 0 \n",
    "    \n",
    "    codons = list(codons_counter.keys())\n",
    "    df_dict = {\"codon\": codons, \"amino_acid\": [shared_functions_and_vars.nt_to_aa[codon] for codon in codons], \"frequency\": [codons_counter[codon] for codon in codons]}\n",
    "    \n",
    "    return pd.DataFrame(df_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_frequency_map(frequency_df):\n",
    "    sorted_df = frequency_df.sort_values(\"amino_acid\")\n",
    "    \n",
    "    plt.rcParams[\"figure.figsize\"] = [10, 5]\n",
    "    plt.rcParams[\"figure.autolayout\"] = True\n",
    "\n",
    "    sorted_df.plot(\"codon\", \"frequency\", kind=\"bar\", color = \"b\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_distance_between_frequencies(first_codons_freq, second_codons_freq):\n",
    "    return np.linalg.norm(first_codons_freq-second_codons_freq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_variants_df_dict(variants_root_path, ecoli_codon_frequencies, bacillus_codon_frequencies, ecoli_codon_frequencies_lamda, bacillus_codon_frequencies_lamda):\n",
    "    with open(variants_root_path, \"r\") as mcherry_variants_fasta_file:\n",
    "        mcherry_variants_dict = SeqIO.to_dict(SeqIO.parse(mcherry_variants_fasta_file, \"fasta\"))\n",
    "\n",
    "    unified_variants_dict = defaultdict(dict)\n",
    "\n",
    "    for variant_name, variant_seq in mcherry_variants_dict.items():  \n",
    "        stripped_name = variant_name.replace(\"_False\", \"\").replace(\"_True\", \"\").replace(\"_ecoli_opt\", \"\")\n",
    "       \n",
    "        mcherry_variant_codon_frequencies = calculate_codon_frequencies((str(variant_seq.seq),)).sort_values(\"codon\")\n",
    "        mcherry_variant_codon_frequencies_lamda = calculate_codon_frequencies_lamda((str(variant_seq.seq),)).sort_values(\"codon\")\n",
    "\n",
    "        series1 = mcherry_variant_codon_frequencies[\"frequency\"].reset_index(drop=True)\n",
    "        series2 =  ecoli_codon_frequencies[\"frequency\"].reset_index(drop=True)\n",
    "        series3 =  bacillus_codon_frequencies[\"frequency\"].reset_index(drop=True)\n",
    "        \n",
    "        ecoli_dist = calculate_distance_between_frequencies(series1, series2)\n",
    "        bacillus_dist = calculate_distance_between_frequencies(series1, series3)\n",
    "        \n",
    "        series1_lamda = mcherry_variant_codon_frequencies_lamda[\"frequency\"].reset_index(drop=True)\n",
    "        series2_lamda =  ecoli_codon_frequencies_lamda[\"frequency\"].reset_index(drop=True)\n",
    "        series3_lamda =  bacillus_codon_frequencies_lamda[\"frequency\"].reset_index(drop=True)\n",
    "        \n",
    "        ecoli_dist_lamda = calculate_distance_between_frequencies(series1_lamda, series2_lamda)\n",
    "        bacillus_dist_lamda = calculate_distance_between_frequencies(series1_lamda, series3_lamda)\n",
    "\n",
    "        if variant_name.endswith(\"True\"):\n",
    "            unified_variants_dict[stripped_name][\"ecoli_dist_for_optimized_ecoli\"] = ecoli_dist\n",
    "            unified_variants_dict[stripped_name][\"bacillus_dist_for_optimized_ecoli\"] = bacillus_dist\n",
    "            unified_variants_dict[stripped_name][\"ecoli_dist_lamda_for_optimized_ecoli\"] = ecoli_dist_lamda\n",
    "            unified_variants_dict[stripped_name][\"bacillus_dist_lamda_for_optimized_ecoli\"] = bacillus_dist_lamda\n",
    "        else:\n",
    "            unified_variants_dict[stripped_name][\"ecoli_dist_for_optimized_bacillus\"] = ecoli_dist\n",
    "            unified_variants_dict[stripped_name][\"bacillus_dist_for_optimized_bacillus\"] = bacillus_dist\n",
    "            unified_variants_dict[stripped_name][\"ecoli_dist_lamda_for_optimized_bacillus\"] = ecoli_dist_lamda\n",
    "            unified_variants_dict[stripped_name][\"bacillus_dist_lamda_for_optimized_bacillus\"] = bacillus_dist_lamda\n",
    "\n",
    "    df_dict = defaultdict(list)\n",
    "    for variant_name in unified_variants_dict.keys():\n",
    "        df_dict[\"variant\"].append(variant_name)\n",
    "        df_dict[\"ecoli_dist_for_original_mcherry\"].append(original_ecoli_dist)\n",
    "        df_dict[\"ecoli_dist_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"ecoli_dist_for_optimized_ecoli\"])\n",
    "        df_dict[\"ecoli_dist_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"ecoli_dist_for_optimized_bacillus\"])\n",
    "        df_dict[\"ecoli_dist_lamda_for_original_mcherry\"].append(original_ecoli_dist_lamda)\n",
    "        df_dict[\"ecoli_dist_lamda_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"ecoli_dist_lamda_for_optimized_ecoli\"])\n",
    "        df_dict[\"ecoli_dist_lamda_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"ecoli_dist_lamda_for_optimized_bacillus\"])\n",
    "\n",
    "        df_dict[\"bacillus_dist_for_original_mcherry\"].append(original_bacillus_dist)\n",
    "        df_dict[\"bacillus_dist_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"bacillus_dist_for_optimized_ecoli\"])\n",
    "        df_dict[\"bacillus_dist_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"bacillus_dist_for_optimized_bacillus\"])\n",
    "        df_dict[\"bacillus_dist_lamda_for_original_mcherry\"].append(original_bacillus_dist_lamda)\n",
    "        df_dict[\"bacillus_dist_lamda_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"bacillus_dist_lamda_for_optimized_ecoli\"])\n",
    "        df_dict[\"bacillus_dist_lamda_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"bacillus_dist_lamda_for_optimized_bacillus\"])\n",
    "    \n",
    "    return df_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def get_variants_df_dict_highly_expressed(variants_root_path):\n",
    "#     with open(variants_root_path, \"r\") as mcherry_variants_fasta_file:\n",
    "#         mcherry_variants_dict = SeqIO.to_dict(SeqIO.parse(mcherry_variants_fasta_file, \"fasta\"))\n",
    "\n",
    "#     unified_variants_dict = defaultdict(dict)\n",
    "\n",
    "#     for variant_name, variant_seq in mcherry_variants_dict.items():\n",
    "#         stripped_name = variant_name.replace(\"_False\", \"\").replace(\"_True\", \"\").replace(\"_ecoli_opt\", \"\")\n",
    "       \n",
    "#         mcherry_variant_codon_frequencies = calculate_codon_frequencies((str(variant_seq.seq),)).sort_values(\"codon\")\n",
    "#         mcherry_variant_codon_frequencies_lamda = calculate_codon_frequencies_lamda((str(variant_seq.seq),)).sort_values(\"codon\")\n",
    "\n",
    "#         series1 = mcherry_variant_codon_frequencies[\"frequency\"].reset_index(drop=True)\n",
    "#         series2 =  ecoli_highly_expressed_genes_codon_frequencies[\"frequency\"].reset_index(drop=True)\n",
    "#         series3 =  bacillus_highly_expressed_genes_codon_frequencies[\"frequency\"].reset_index(drop=True)\n",
    "\n",
    "#         ecoli_dist = calculate_distance_between_frequencies(series1, series2)\n",
    "#         bacillus_dist = calculate_distance_between_frequencies(series1, series3)\n",
    "        \n",
    "#         series1_lamda = mcherry_variant_codon_frequencies_lamda[\"frequency\"].reset_index(drop=True)\n",
    "#         series2_lamda =  ecoli_highly_expressed_genes_codon_frequencies_lamda[\"frequency\"].reset_index(drop=True)\n",
    "#         series3_lamda =  bacillus_highly_expressed_genes_codon_frequencies_lamda[\"frequency\"].reset_index(drop=True)\n",
    "\n",
    "#         ecoli_dist_lamda = calculate_distance_between_frequencies(series1_lamda, series2_lamda)\n",
    "#         bacillus_dist_lamda = calculate_distance_between_frequencies(series1_lamda, series3_lamda)\n",
    "\n",
    "#         if variant_name.endswith(\"True\"):\n",
    "#             unified_variants_dict[stripped_name][\"ecoli_dist_for_optimized_ecoli\"] = ecoli_dist\n",
    "#             unified_variants_dict[stripped_name][\"bacillus_dist_for_optimized_ecoli\"] = bacillus_dist\n",
    "#             unified_variants_dict[stripped_name][\"ecoli_dist_lamda_for_optimized_ecoli\"] = ecoli_dist_lamda\n",
    "#             unified_variants_dict[stripped_name][\"bacillus_dist_lamda_for_optimized_ecoli\"] = bacillus_dist_lamda\n",
    "#         else:\n",
    "#             unified_variants_dict[stripped_name][\"ecoli_dist_for_optimized_bacillus\"] = ecoli_dist\n",
    "#             unified_variants_dict[stripped_name][\"bacillus_dist_for_optimized_bacillus\"] = bacillus_dist\n",
    "#             unified_variants_dict[stripped_name][\"ecoli_dist_lamda_for_optimized_bacillus\"] = ecoli_dist_lamda\n",
    "#             unified_variants_dict[stripped_name][\"bacillus_dist_lamda_for_optimized_bacillus\"] = bacillus_dist_lamda\n",
    "\n",
    "#     df_dict = defaultdict(list)\n",
    "#     for variant_name in unified_variants_dict.keys():\n",
    "#         df_dict[\"variant\"].append(variant_name)\n",
    "#         df_dict[\"ecoli_dist_for_original_mcherry\"].append(original_ecoli_dist)\n",
    "#         df_dict[\"ecoli_dist_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"ecoli_dist_for_optimized_ecoli\"])\n",
    "#         df_dict[\"ecoli_dist_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"ecoli_dist_for_optimized_bacillus\"])\n",
    "#         df_dict[\"ecoli_dist_lamda_for_original_mcherry\"].append(original_ecoli_dist_lamda)\n",
    "#         df_dict[\"ecoli_dist_lamda_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"ecoli_dist_lamda_for_optimized_ecoli\"])\n",
    "#         df_dict[\"ecoli_dist_lamda_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"ecoli_dist_lamda_for_optimized_bacillus\"])\n",
    "\n",
    "#         df_dict[\"bacillus_dist_for_original_mcherry\"].append(original_bacillus_dist)\n",
    "#         df_dict[\"bacillus_dist_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"bacillus_dist_for_optimized_ecoli\"])\n",
    "#         df_dict[\"bacillus_dist_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"bacillus_dist_for_optimized_bacillus\"])\n",
    "#         df_dict[\"bacillus_dist_lamda_for_original_mcherry\"].append(original_bacillus_dist_lamda)\n",
    "#         df_dict[\"bacillus_dist_lamda_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"bacillus_dist_lamda_for_optimized_ecoli\"])\n",
    "#         df_dict[\"bacillus_dist_lamda_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"bacillus_dist_lamda_for_optimized_bacillus\"])\n",
    "    \n",
    "#     return df_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_codon_frequencies_df_dict(variants_root_path):\n",
    "    with open(variants_root_path, \"r\") as mcherry_variants_fasta_file:\n",
    "        mcherry_variants_dict = SeqIO.to_dict(SeqIO.parse(mcherry_variants_fasta_file, \"fasta\"))\n",
    "\n",
    "    unified_variants_dict = defaultdict(dict)\n",
    "\n",
    "    for variant_name, variant_seq in mcherry_variants_dict.items():\n",
    "        if \"weakest_link\" in variant_name:\n",
    "            continue\n",
    "        \n",
    "        stripped_name = variant_name.replace(\"_False\", \"\").replace(\"_True\", \"\").replace(\"_ecoli_opt\", \"\")\n",
    "       \n",
    "        mcherry_variant_codon_frequencies = calculate_codon_frequencies((str(variant_seq.seq),)).sort_values(\"codon\")\n",
    "        \n",
    "        return mcherry_variant_codon_frequencies \n",
    "    \n",
    "        series1 = mcherry_variant_codon_frequencies[\"frequency\"].reset_index(drop=True)\n",
    "        \n",
    "        if variant_name.endswith(\"True\"):\n",
    "            unified_variants_dict[stripped_name][\"codon_frequencies_for_optimized_ecoli\"] = mcherry_variant_codon_frequencies\n",
    "        else:\n",
    "            unified_variants_dict[stripped_name][\"codon_frequencies_for_optimized_bacillus\"] = mcherry_variant_codon_frequencies\n",
    "\n",
    "    df_dict = defaultdict(list)\n",
    "    for variant_name in unified_variants_dict.keys():\n",
    "        df_dict[\"variant\"].append(variant_name)\n",
    "        df_dict[\"codon_frequencies_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"codon_frequencies_for_optimized_ecoli\"])\n",
    "        df_dict[\"codon_frequencies_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"codon_frequencies_for_optimized_bacillus\"])\n",
    "    \n",
    "    return df_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_evaluation_summary(summary: typing.Dict[str, typing.Any], evaluation_method: str = \"average_distance_score\") -> typing.Dict[str, typing.Any]:\n",
    "    final_evaluation = summary[\"final_evaluation\"]\n",
    "    for evaluation_summary in summary[\"evaluation\"]:\n",
    "        if evaluation_summary[evaluation_method] == final_evaluation[evaluation_method]:\n",
    "            return evaluation_summary\n",
    "    raise RuntimeError(F\"Did not find a final evaluation for: {summary}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cub_score_df_dict(root_dir, evaluation_method: str = \"average_distance_score\"):\n",
    "    filename = \"run_summary.json\"\n",
    "    \n",
    "    unified_variants_dict = defaultdict(dict)\n",
    "    \n",
    "    for root, dirs, files in os.walk(root_dir):\n",
    "        for file in files:\n",
    "            if file == filename:\n",
    "                directory_name = Path(root).name\n",
    "                file_path = os.path.join(root, file)\n",
    "                with open(file_path, \"r\") as summary_file:\n",
    "                    results_json = json.load(summary_file)\n",
    "                \n",
    "                is_cai = directory_name.startswith(\"CAI\")\n",
    "                evaluation_summary = get_evaluation_summary(results_json)\n",
    "                organisms = evaluation_summary[\"organisms\"]\n",
    "                for organism in organisms:\n",
    "                    if organism[\"name\"] == \"Bacillus subtilis\":\n",
    "                        if is_cai:\n",
    "                            bacillus_initial_score = organism[\"cai_initial_score\"]\n",
    "                            bacillus_final_score = organism[\"cai_final_score\"]\n",
    "                        else:\n",
    "                            bacillus_initial_score = organism[\"tai_initial_score\"]\n",
    "                            bacillus_final_score = organism[\"tai_final_score\"]\n",
    "                    else:\n",
    "                        if is_cai:\n",
    "                            ecoli_initial_score = organism[\"cai_initial_score\"]\n",
    "                            ecoli_final_score = organism[\"cai_final_score\"]\n",
    "                        else:\n",
    "                            ecoli_initial_score = organism[\"tai_initial_score\"]\n",
    "                            ecoli_final_score = organism[\"tai_final_score\"]\n",
    "                            \n",
    "                \n",
    "                variant_name = directory_name[:-5]\n",
    "                stripped_name = variant_name.replace(\"_False\", \"\").replace(\"_True\", \"\").replace(\"_ecoli_opt\", \"\")\n",
    "                \n",
    "                if \"weakest_link\" in variant_name:\n",
    "                    continue\n",
    "                \n",
    "                if variant_name.endswith(\"True\"):\n",
    "                    suffix = \"_for_optimized_ecoli\"\n",
    "                else:\n",
    "                    suffix = \"_for_optimized_bacillus\"\n",
    "                \n",
    "                unified_variants_dict[stripped_name][\"bacillus_initial_score\" + suffix] = bacillus_initial_score\n",
    "                unified_variants_dict[stripped_name][\"bacillus_final_score\" + suffix] = bacillus_final_score\n",
    "                unified_variants_dict[stripped_name][\"ecoli_initial_score\" + suffix] = ecoli_initial_score\n",
    "                unified_variants_dict[stripped_name][\"ecoli_final_score\" + suffix] = ecoli_final_score\n",
    "                unified_variants_dict[stripped_name][\"average_distance_evaluation_score\" + suffix] = evaluation_summary[\"average_distance_score\"]\n",
    "                unified_variants_dict[stripped_name][\"weakest_link_evaluation_score\" + suffix] = evaluation_summary[\"weakest_link_score\"]\n",
    "                unified_variants_dict[stripped_name][\"ratio_evaluation_score\" + suffix] = evaluation_summary[\"ratio_score\"]\n",
    "                \n",
    "    \n",
    "    df_dict = defaultdict(list)\n",
    "    for variant_name in unified_variants_dict.keys():\n",
    "        df_dict[\"variant\"].append(variant_name)\n",
    "        df_dict[\"ecoli_initial_score_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"ecoli_initial_score_for_optimized_ecoli\"])\n",
    "        df_dict[\"ecoli_final_score_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"ecoli_final_score_for_optimized_ecoli\"])\n",
    "        df_dict[\"ecoli_initial_score_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"ecoli_initial_score_for_optimized_bacillus\"])\n",
    "        df_dict[\"ecoli_final_score_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"ecoli_final_score_for_optimized_bacillus\"])\n",
    "        df_dict[\"bacillus_initial_score_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"bacillus_initial_score_for_optimized_ecoli\"])\n",
    "        df_dict[\"bacillus_final_score_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"bacillus_final_score_for_optimized_ecoli\"])\n",
    "        df_dict[\"bacillus_initial_score_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"bacillus_initial_score_for_optimized_bacillus\"])\n",
    "        df_dict[\"bacillus_final_score_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"bacillus_final_score_for_optimized_bacillus\"])\n",
    "        df_dict[\"average_distance_evaluation_score_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"average_distance_evaluation_score_for_optimized_ecoli\"])\n",
    "        df_dict[\"average_distance_evaluation_score_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"average_distance_evaluation_score_for_optimized_bacillus\"])\n",
    "        df_dict[\"weakest_link_evaluation_score_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"weakest_link_evaluation_score_for_optimized_ecoli\"])\n",
    "        df_dict[\"weakest_link_evaluation_score_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"weakest_link_evaluation_score_for_optimized_bacillus\"])\n",
    "        df_dict[\"ratio_evaluation_score_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"ratio_evaluation_score_for_optimized_ecoli\"])\n",
    "        df_dict[\"ratio_evaluation_score_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"ratio_evaluation_score_for_optimized_bacillus\"])\n",
    "    \n",
    "    return df_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cub_weights(organism_summary_file_path: str):\n",
    "    with open(organism_summary_file_path, \"r\") as organism_summary_file:\n",
    "        organism_summary = json.load(organism_summary_file)\n",
    "    \n",
    "    cai_profile = organism_summary[\"cai_weights\"]\n",
    "    tai_profile = organism_summary[\"tai_weights\"]\n",
    "    \n",
    "    return cai_profile, tai_profile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cub_scores(organism_summary_file_path: str):\n",
    "    with open(organism_summary_file_path, \"r\") as organism_summary_file:\n",
    "        organism_summary = json.load(organism_summary_file)\n",
    "    \n",
    "    cai_scores = organism_summary[\"cai_scores\"]\n",
    "    tai_scores = organism_summary[\"tai_scores\"]\n",
    "    \n",
    "    return cai_scores, tai_scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_mrna_estimated_expression(endogenous_genes_fasta_file_path, mrna_levels_file_path):\n",
    "    expression_df = pd.read_csv(mrna_levels_file_path)\n",
    "    gene_name_to_mrna_level = {}\n",
    "    for _, pair in enumerate(zip(expression_df.gene.to_list(), expression_df.mRNA_level.to_list())):\n",
    "        measured_gene_name, expression_level = pair\n",
    "        try:\n",
    "            gene_name_to_mrna_level[measured_gene_name.lower()] = float(expression_level)\n",
    "        except:\n",
    "            continue\n",
    "            \n",
    "    with open(endogenous_genes_fasta_file_path, \"r\") as endogenous_genes_fasta_file:\n",
    "        endogenous_genes_dict = SeqIO.to_dict(SeqIO.parse(endogenous_genes_fasta_file, \"fasta\"), lambda x: x.description)\n",
    "        \n",
    "    estimated_expression = {description: {\"cds\": cds, \"mrna_level\": gene_name_to_mrna_level[description.split(\"|\")[0].lower()]}  for description, cds in endogenous_genes_dict.items() if description.split(\"|\")[0].lower() in gene_name_to_mrna_level}\n",
    "    return estimated_expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_pa_estimated_expression(endogenous_genes_fasta_file_path, pa_file_path):\n",
    "    expression_df = pd.read_json(pa_file_path)\n",
    "    gene_name_to_pa = {}\n",
    "    for _, pair in enumerate(zip(expression_df.name.to_list(), expression_df.abundance.to_list())):\n",
    "        measured_gene_name, expression_level = pair\n",
    "        try:\n",
    "            gene_name_to_pa[measured_gene_name.lower()] = float(expression_level)\n",
    "        except:\n",
    "            continue\n",
    "            \n",
    "    with open(endogenous_genes_fasta_file_path, \"r\") as endogenous_genes_fasta_file:\n",
    "        endogenous_genes_dict = SeqIO.to_dict(SeqIO.parse(endogenous_genes_fasta_file, \"fasta\"), lambda x: x.description)\n",
    "        \n",
    "    estimated_expression = {description: {\"cds\": cds, \"pa\": gene_name_to_pa[description.split(\"|\")[0].lower()]}  for description, cds in endogenous_genes_dict.items() if description.split(\"|\")[0].lower() in gene_name_to_pa}\n",
    "    return estimated_expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_estimated_expression(endogenous_genes_fasta_file_path, expression_levels_file_path, expression_type: str):\n",
    "    if expression_type == \"mrna_levels\":\n",
    "        return get_mrna_estimated_expression(endogenous_genes_fasta_file_path, expression_levels_file_path)\n",
    "    elif expression_type == \"protein_abundance\":\n",
    "        return get_pa_estimated_expression(endogenous_genes_fasta_file_path, expression_levels_file_path)\n",
    "    raise ValueError(\"Missing expression type support\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_highly_expressed_genes(endogenous_genes_fasta_file_path, expression_levels_file_path, expression_type):\n",
    "    estimated_expression = get_estimated_expression(endogenous_genes_fasta_file_path=endogenous_genes_fasta_file_path, expression_levels_file_path=expression_levels_file_path, expression_type=expression_type)\n",
    "    estimated_expression_threshold = 0.15\n",
    "    \n",
    "    expression_key = \"mrna_level\" if expression_type == \"mrna_levels\" else \"pa\"\n",
    "    sorted_estimated_expression = dict(\n",
    "        sorted(estimated_expression.items(), key=lambda x: x[1][expression_key], reverse=True)\n",
    "    )\n",
    "    highly_expressed_genes_count = round(len(sorted_estimated_expression) * estimated_expression_threshold)\n",
    "    \n",
    "    highly_expressed_names = list(sorted_estimated_expression.keys())[:highly_expressed_genes_count]\n",
    "    highly_expressed_genes = {key: value for key, value in estimated_expression.items() if key in highly_expressed_names}\n",
    "    return highly_expressed_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "base_path = os.path.join(Path(os.getcwd()).parent.resolve(), \"example_data\")\n",
    "macherry_file_path = os.path.join(base_path, \"mCherry_original.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(macherry_file_path, \"r\") as mcherry_fasta_file:\n",
    "    mcherry_dict = SeqIO.to_dict(SeqIO.parse(mcherry_fasta_file, \"fasta\"))\n",
    "\n",
    "original_mcherry_seq = list(mcherry_dict.values())[0].seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generating mcherry variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tuning parameter = 0.5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Using protein abundance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "##########################\n",
      "# User Input #\n",
      "##########################\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is:\n",
      "ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "------------------------------------------\n",
      "Parsing information for Escherichia coli:\n",
      "------------------------------------------\n",
      "Organism is defined as wanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 90 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3893\n",
      "Calculate CAI weights from a reference set of 584 highly expressed genes from estimated expression dictionary.\n",
      "cai_std=0.09377868043841883, cai_avg=0.6151304741138522\n",
      "------------------------------------------\n",
      "Parsing information for Bacillus subtilis:\n",
      "------------------------------------------\n",
      "Organism is defined as unwanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 98 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3544\n",
      "Calculate CAI weights from a reference set of 532 highly expressed genes from estimated expression dictionary.\n",
      "cai_std=0.042802002292947894, cai_avg=0.7334925635409068\n",
      "-----------------------------\n",
      "Normalized Prioritization Weights\n",
      "-----------------------------\n",
      "Escherichia coli : 1.0\n",
      "Bacillus subtilis : 1.0\n",
      "Total input processing time: 29.023258686065674\n",
      "\n",
      "##########################\n",
      "# Initiation #\n",
      "##########################\n",
      "Running initiation optimization InitiationOptimizationMethod.external_module on 15 codons at the start of the ORF\n",
      "Taking optimized sequence from configuration value config.INITIATION.EXTERNAL_INITIATION_ORF\n",
      "Optimized sequence is: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "Optimization method used is: ORFOptimizationMethod.single_codon_ratio\n",
      "Loss summary for: TGT\n",
      "loss for Escherichia coli: 1.6432327117286945\n",
      "loss for Bacillus subtilis: 1.6090122554987418\n",
      "Loss summary for: TGC\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGT': 1.0105780473545745, 'TGC': 1.0})\n",
      "Loss summary for: GAT\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: GAC\n",
      "loss for Escherichia coli: 1.32644912035017\n",
      "loss for Bacillus subtilis: 1.7487741357394064\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAT': 1.0, 'GAC': 0.8709203477283044})\n",
      "Loss summary for: TCT\n",
      "loss for Escherichia coli: 1.2358074271617692\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: TCG\n",
      "loss for Escherichia coli: 2.5325041749163573\n",
      "loss for Bacillus subtilis: 3.051635584744884\n",
      "Loss summary for: TCA\n",
      "loss for Escherichia coli: 2.940414537227555\n",
      "loss for Bacillus subtilis: 1.1546900193499807\n",
      "Loss summary for: TCC\n",
      "loss for Escherichia coli: 1.4589325892912588\n",
      "loss for Bacillus subtilis: 2.4779507004500814\n",
      "Loss summary for: AGC\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.2426615592968109\n",
      "Loss summary for: AGT\n",
      "loss for Escherichia coli: 2.911797534354834\n",
      "loss for Bacillus subtilis: 2.830032659527515\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TCT': 1.1116687578419073, 'TCG': 0.9109798056045083, 'TCA': 1.5957746973922708, 'TCC': 0.767310748640736, 'AGC': 0.8970642883686363, 'AGT': 1.0143430625248522})\n",
      "Loss summary for: CAA\n",
      "loss for Escherichia coli: 2.7969463044492056\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: CAG\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.3088053780514624\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAA': 1.6724073380756272, 'CAG': 0.8741027056874654})\n",
      "Loss summary for: ATG\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATG': 1.0})\n",
      "Loss summary for: AAC\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: AAT\n",
      "loss for Escherichia coli: 2.5897675631491137\n",
      "loss for Bacillus subtilis: 1.3943684266530345\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAC': 1.0, 'AAT': 1.3628298062012105})\n",
      "Loss summary for: CCT\n",
      "loss for Escherichia coli: 3.3057851239669427\n",
      "loss for Bacillus subtilis: 1.5776475136630435\n",
      "Loss summary for: CCG\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: CCA\n",
      "loss for Escherichia coli: 3.0314838894201217\n",
      "loss for Bacillus subtilis: 2.3382446297864634\n",
      "Loss summary for: CCC\n",
      "loss for Escherichia coli: 3.679421487603306\n",
      "loss for Bacillus subtilis: 3.7288941736028534\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CCT': 1.447545827476405, 'CCG': 1.0, 'CCA': 1.1386300978111386, 'CCC': 0.9933441558441559})\n",
      "Loss summary for: AAG\n",
      "loss for Escherichia coli: 3.0174622091692234\n",
      "loss for Bacillus subtilis: 2.731716735734297\n",
      "Loss summary for: AAA\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAG': 1.051000900574087, 'AAA': 1.0})\n",
      "Loss summary for: TAG\n",
      "loss for Escherichia coli: 3.853223593964335\n",
      "loss for Bacillus subtilis: 3.5819401803361797\n",
      "Loss summary for: TGA\n",
      "loss for Escherichia coli: 3.0911615190738604\n",
      "loss for Bacillus subtilis: 3.3509093705321793\n",
      "Loss summary for: TAA\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAG': 1.03717715193125, 'TGA': 0.9604604988453446, 'TAA': 1.0})\n",
      "Loss summary for: ACC\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 3.1600127668893183\n",
      "Loss summary for: ACA\n",
      "loss for Escherichia coli: 3.504371592388779\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: ACG\n",
      "loss for Escherichia coli: 2.7456517456510596\n",
      "loss for Bacillus subtilis: 2.128216248945088\n",
      "Loss summary for: ACT\n",
      "loss for Escherichia coli: 2.4857819856416143\n",
      "loss for Bacillus subtilis: 2.459228935403716\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ACC': 0.5625428140841212, 'ACA': 1.871996685998343, 'ACG': 1.1358339568986147, 'ACT': 1.0053841590051715})\n",
      "Loss summary for: TTT\n",
      "loss for Escherichia coli: 1.8071998373247593\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: TTC\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.289765409964395\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTT': 1.344321329639889, 'TTC': 0.8805309734513274})\n",
      "Loss summary for: GCA\n",
      "loss for Escherichia coli: 1.8758964212769875\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: GCC\n",
      "loss for Escherichia coli: 1.812421976933375\n",
      "loss for Bacillus subtilis: 2.129837116329259\n",
      "Loss summary for: GCG\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.101352459094135\n",
      "Loss summary for: GCT\n",
      "loss for Escherichia coli: 1.9423109359102864\n",
      "loss for Bacillus subtilis: 1.100806221057107\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GCA': 1.369633681418863, 'GCC': 0.9224789488619323, 'GCG': 0.9528769841269842, 'GCT': 1.3283237570851427})\n",
      "Loss summary for: GGT\n",
      "loss for Escherichia coli: 1.0427990176022806\n",
      "loss for Bacillus subtilis: 1.9495431577031723\n",
      "Loss summary for: GGG\n",
      "loss for Escherichia coli: 3.2817356454694036\n",
      "loss for Bacillus subtilis: 2.9297346559740687\n",
      "Loss summary for: GGA\n",
      "loss for Escherichia coli: 3.548016240581601\n",
      "loss for Bacillus subtilis: 1.388670309008689\n",
      "Loss summary for: GGC\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GGT': 0.7313645062044244, 'GGG': 1.0583703213877516, 'GGA': 1.5984285372331444, 'GGC': 1.0})\n",
      "Loss summary for: ATC\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.3089200005100767\n",
      "Loss summary for: ATA\n",
      "loss for Escherichia coli: 3.883668089603532\n",
      "loss for Bacillus subtilis: 3.6016578405776287\n",
      "Loss summary for: ATT\n",
      "loss for Escherichia coli: 1.599113350558424\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATC': 0.8740644321509926, 'ATA': 1.0384123083805088, 'ATT': 1.2645605365337098})\n",
      "Loss summary for: TTA\n",
      "loss for Escherichia coli: 3.631806659913921\n",
      "loss for Bacillus subtilis: 2.109220171036552\n",
      "Loss summary for: TTG\n",
      "loss for Escherichia coli: 3.5436923326098997\n",
      "loss for Bacillus subtilis: 2.600359572406518\n",
      "Loss summary for: CTC\n",
      "loss for Escherichia coli: 3.509954066352956\n",
      "loss for Bacillus subtilis: 2.8772310086632475\n",
      "Loss summary for: CTT\n",
      "loss for Escherichia coli: 3.5829592996359754\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: CTG\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.7802239801411086\n",
      "Loss summary for: CTA\n",
      "loss for Escherichia coli: 3.908995597888406\n",
      "loss for Bacillus subtilis: 3.46359234781819\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTA': 1.3122011147635726, 'TTG': 1.1673774549551195, 'CTC': 1.1044939732011871, 'CTT': 1.8928706505295008, 'CTG': 0.7494845360824742, 'CTA': 1.0623538715444918})\n",
      "Loss summary for: CAT\n",
      "loss for Escherichia coli: 1.6669947156956941\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: CAC\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.7428059544498995\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAT': 1.2911214953271029, 'CAC': 0.7574875207986689})\n",
      "Loss summary for: CGA\n",
      "loss for Escherichia coli: 3.868744772855011\n",
      "loss for Bacillus subtilis: 3.2787385877493485\n",
      "Loss summary for: CGC\n",
      "loss for Escherichia coli: 1.6579706335513271\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: CGG\n",
      "loss for Escherichia coli: 3.782199282654244\n",
      "loss for Bacillus subtilis: 2.96411287678867\n",
      "Loss summary for: CGT\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.019323224852071\n",
      "Loss summary for: AGG\n",
      "loss for Escherichia coli: 3.9610282186271286\n",
      "loss for Bacillus subtilis: 3.6400623268698062\n",
      "Loss summary for: AGA\n",
      "loss for Escherichia coli: 3.9388442858165034\n",
      "loss for Bacillus subtilis: 1.682377036584766\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CGA': 1.086254638087406, 'CGC': 1.2876220849113016, 'CGG': 1.1296003933480958, 'CGT': 0.9904761904761905, 'AGG': 1.0431567154433548, 'AGA': 1.5301103054141865})\n",
      "Loss summary for: TGG\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGG': 1.0})\n",
      "Loss summary for: GTA\n",
      "loss for Escherichia coli: 2.2154574764955095\n",
      "loss for Bacillus subtilis: 1.6324607736794863\n",
      "Loss summary for: GTC\n",
      "loss for Escherichia coli: 2.259279403605198\n",
      "loss for Bacillus subtilis: 1.7068049913928647\n",
      "Loss summary for: GTG\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.7691118224210975\n",
      "Loss summary for: GTT\n",
      "loss for Escherichia coli: 1.0374231441153496\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GTA': 1.1649581715973647, 'GTC': 1.1505169451198147, 'GTG': 0.7518346852066434, 'GTT': 1.018539711604486})\n",
      "Loss summary for: GAG\n",
      "loss for Escherichia coli: 2.6854688521243797\n",
      "loss for Bacillus subtilis: 2.5826598723651326\n",
      "Loss summary for: GAA\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAG': 1.0197094694060727, 'GAA': 1.0})\n",
      "Loss summary for: TAT\n",
      "loss for Escherichia coli: 1.3171976331360948\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: TAC\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.4803358808630471\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAT': 1.1476923076923078, 'TAC': 0.8219016779511333})\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4262385670731707]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.38120248233619]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4262385670731707]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4292446655321377]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.689467997530356]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5512262248110238]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4292446655321377]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.689467997530356]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4262385670731707]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.689467997530356]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4205725148436632]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.689467997530356]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.689467997530356]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.689467997530356]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.38120248233619]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.689467997530356]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4050486219579785]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4262385670731707]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.37314335931309695]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4205725148436632]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4262385670731707]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4292446655321377]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.37314335931309695]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4205725148436632]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.37314335931309695]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4262385670731707]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.689467997530356]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4262385670731707]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.689467997530356]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.38120248233619]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5512262248110238]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4262385670731707]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.689467997530356]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.37314335931309695]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4050486219579785]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2584847661069405]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.37314335931309695]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.27799163607995603]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.37314335931309695]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.689467997530356]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4262385670731707]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4205725148436632]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.38120248233619]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5512262248110238]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4262385670731707]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.37314335931309695]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.689467997530356]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.27799163607995603]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.689467997530356]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.38120248233619]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.689467997530356]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.37314335931309695]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.689467997530356]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4205725148436632]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.38120248233619]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.37314335931309695]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4050486219579785]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4292446655321377]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.37314335931309695]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.689467997530356]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.16532540959927366]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5512262248110238]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5512262248110238]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4205725148436632]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.37314335931309695]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4205725148436632]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4292446655321377]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5512262248110238]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.689467997530356]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.37314335931309695]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4205725148436632]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.689467997530356]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.689467997530356]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.38120248233619]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.689467997530356]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.38120248233619]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4292446655321377]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4050486219579785]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.37314335931309695]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.689467997530356]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.28926761860893596]\n",
      "\n",
      "##########################\n",
      "# EVALUATION #\n",
      "##########################\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTCATGCGTTTCAAAGTGCACATGGAAGGTTCCGTGAACGGTCACGAATTCGAAATCGAAGGTGAAGGTGAAGGTCGTCCCTACGAAGGTACCCAGACCGCCAAACTGAAAGTGACCAAAGGTGGCCCCCTGCCCTTCGCCTGGGACATCCTGTCCCCCCAGTTCATGTACGGTTCCAAAGCCTACGTGAAACACCCCGCCGACATCCCCGACTACCTGAAACTGTCCTTCCCCGAAGGTTTCAAATGGGAACGTGTGATGAACTTCGAAGACGGTGGCGTGGTTACCGTGACCCAGGACTCCAGCCTGCAGGACGGTGAATTCATCTACAAAGTGAAACTGCGTGGTACCAACTTCCCCTCCGACGGTCCCGTGATGCAGAAAAAGACCATGGGTTGGGAAGCCTCCAGCGAACGTATGTACCCCGAAGACGGTGCCCTGAAAGGTGAAATCAAACAGCGTCTGAAACTGAAAGACGGTGGCCACTACGACGCCGAAGTGAAAACCACTTACAAAGCCAAAAAGCCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTGGAACAGTACGAACGTGCCGAAGGTCGTCACTCCACCGGTGGCATGGACGAACTGTACAAATGA\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTGCACATGGAAGGTTCCGTGAACGGTCACGAATTCGAAATCGAAGGTGAAGGTGAAGGTCGTCCCTACGAAGGTACCCAGACCGCCAAACTGAAAGTGACCAAAGGTGGCCCCCTGCCCTTCGCCTGGGACATCCTGTCCCCCCAGTTCATGTACGGTTCCAAAGCCTACGTGAAACACCCCGCCGACATCCCCGACTACCTGAAACTGTCCTTCCCCGAAGGTTTCAAATGGGAACGTGTGATGAACTTCGAAGACGGTGGCGTGGTTACCGTGACCCAGGACTCCAGCCTGCAGGACGGTGAATTCATCTACAAAGTGAAACTGCGTGGTACCAACTTCCCCTCCGACGGTCCCGTGATGCAGAAAAAGACCATGGGTTGGGAAGCCTCCAGCGAACGTATGTACCCCGAAGACGGTGCCCTGAAAGGTGAAATCAAACAGCGTCTGAAACTGAAAGACGGTGGCCACTACGACGCCGAAGTGAAAACCACTTACAAAGCCAAAAAGCCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTGGAACAGTACGAACGTGCCGAAGGTCGTCACTCCACCGGTGGCATGGACGAACTGTACAAATGA', 'average_distance_score': 1.9074288080922197, 'weakest_link_score': 1.9074288080922197, 'ratio_score': 1.0833299803552052}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTGCACATGGAAGGTTCCGTGAACGGTCACGAATTCGAAATCGAAGGTGAAGGTGAAGGTCGTCCCTACGAAGGTACCCAGACCGCCAAACTGAAAGTGACCAAAGGTGGCCCCCTGCCCTTCGCCTGGGACATCCTGTCCCCCCAGTTCATGTACGGTTCCAAAGCCTACGTGAAACACCCCGCCGACATCCCCGACTACCTGAAACTGTCCTTCCCCGAAGGTTTCAAATGGGAACGTGTGATGAACTTCGAAGACGGTGGCGTGGTTACCGTGACCCAGGACTCCAGCCTGCAGGACGGTGAATTCATCTACAAAGTGAAACTGCGTGGTACCAACTTCCCCTCCGACGGTCCCGTGATGCAGAAAAAGACCATGGGTTGGGAAGCCTCCAGCGAACGTATGTACCCCGAAGACGGTGCCCTGAAAGGTGAAATCAAACAGCGTCTGAAACTGAAAGACGGTGGCCACTACGACGCCGAAGTGAAAACCACTTACAAAGCCAAAAAGCCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTGGAACAGTACGAACGTGCCGAAGGTCGTCACTCCACCGGTGGCATGGACGAACTGTACAAATGA', 'average_distance_score': 1.9074288080922197, 'weakest_link_score': 1.9074288080922197, 'ratio_score': 1.0833299803552052}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry-wanted-constraint-init-external-0.1\\CAI_single_codon_ratio_ecoli_opt_True_XSuW\n",
      "Final output: {'module_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'orf_optimization_method': 'single_codon_ratio', 'orf_optimization_cub_index': 'CAI', 'initiation_optimization_method': 'external', 'evaluation_score': 'average_distance', 'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.037037037037037035, 'TGA': 0.24183006535947713, 'GCA': 0.6303663185811371, 'GCC': 0.653737775567711, 'GCG': 1.0, 'GCT': 0.6063318415382065, 'TGC': 1.0, 'TGT': 0.7181136120042874, 'GAC': 0.848284271032921, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.36125998031280765, 'TTC': 1.0, 'TTT': 0.6556786703601108, 'GGA': 0.11638214051214708, 'GGC': 1.0, 'GGG': 0.18844386080105055, 'GGT': 0.9788246881155614, 'CAC': 1.0, 'CAT': 0.7088785046728973, 'ATA': 0.029297564419343447, 'ATC': 1.0, 'ATT': 0.7354394634662901, 'AAA': 1.0, 'AAG': 0.26291560102301786, 'CTA': 0.02288199697428139, 'CTC': 0.1265128593040847, 'CTG': 1.0, 'CTT': 0.10712934947049924, 'TTA': 0.09427004538577911, 'TTG': 0.11753025718608169, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.39072452229299365, 'CCA': 0.2588842975206611, 'CCC': 0.08181818181818182, 'CCG': 1.0, 'CCT': 0.1818181818181818, 'CAA': 0.32759266192437286, 'CAG': 1.0, 'AGA': 0.015347817420769383, 'AGG': 0.00976679290412597, 'CGA': 0.033087502491528795, 'CGC': 0.7123779150886984, 'CGG': 0.05521227825393661, 'CGT': 1.0, 'AGC': 1.0, 'AGT': 0.2936010037641154, 'TCA': 0.28523630280217477, 'TCC': 0.7921371810957758, 'TCG': 0.4086156419907988, 'TCT': 0.8883312421580928, 'ACA': 0.12800331400165701, 'ACC': 1.0, 'ACG': 0.34299917149958575, 'ACT': 0.4233637116818558, 'GTA': 0.5115587090867476, 'GTC': 0.49691004806591893, 'GTG': 1.0, 'GTT': 0.9814602883955139, 'TGG': 1.0, 'TAC': 1.0, 'TAT': 0.8523076923076923}, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'AAA': 0.7667390608953314, 'CGC': 0.40076028262096014, 'ATT': 0.5085852538544149, 'AGC': 0.27799163607995603, 'ACC': 0.4359429381095841, 'ATC': 0.4205725148436632, 'ACA': 0.13023262212317555, 'GGT': 0.33726003930064996, 'AAC': 0.5512262248110238, 'GCG': 0.3563679190097728, 'GGC': 0.4050486219579785, 'TGA': 0.28926761860893596, 'CGA': 0.06388523224887838, 'GTG': 0.3715940695510341, 'TTG': 0.12808970030428077, 'AAG': 0.23326093910466858, 'TTC': 0.4262385670731707, 'TCA': 0.12251736088908215, 'GCA': 0.21281773299119433, 'AAT': 0.4487737751889762, 'GAA': 0.689467997530356, 'CGT': 0.38120248233619, 'TTT': 0.5737614329268292, 'CTG': 0.4975546385450099, 'GTT': 0.2584847661069405, 'GCC': 0.26990242799092135, 'GAT': 0.626856640686903, 'AGG': 0.020162265266098916, 'CAG': 0.6530527088827319, 'GGG': 0.15057187484254547, 'GTC': 0.21619163129169194, 'CTC': 0.10435163186194824, 'TCT': 0.14544780223293646, 'CCC': 0.12361073915372255, 'CAC': 0.4292446655321377, 'GCT': 0.1609119200081115, 'TTA': 0.13009739763522432, 'CTT': 0.10341378016756284, 'ACG': 0.26849903016796667, 'GGA': 0.10711946389882603, 'CCG': 0.527785216925549, 'CAA': 0.3469472911172681, 'ACT': 0.16532540959927366, 'ATA': 0.07084223130192197, 'CAT': 0.5707553344678623, 'AGT': 0.15102374892893225, 'TGC': 0.55651396076928, 'GAG': 0.31053200246964396, 'TCG': 0.15406749964830613, 'GTA': 0.15372953305033354, 'TAC': 0.4315172054223149, 'TCC': 0.14895195222078703, 'GAC': 0.37314335931309695, 'TGT': 0.4434860392307201, 'TGG': 1.0, 'TAT': 0.5684827945776851, 'CCT': 0.15789033208355652, 'CCA': 0.19071371183717192, 'CGG': 0.09749348521075679, 'CTA': 0.03649285148597391, 'TAA': 0.6391064025794565, 'AGA': 0.036496252317115764, 'TAG': 0.07162597881160755}), 'cai_avg': 0.6151304741138522, 'cai_std': 0.09377868043841883, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.10739856801909307, 'TGA': 0.1694510739856802, 'GCA': 1.0, 'GCC': 0.5406038521603331, 'GCG': 0.950546590317543, 'GCT': 0.9508068714211347, 'TGC': 1.0, 'TGT': 0.7315315315315315, 'GAC': 0.6775877587758776, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.3929343907714492, 'TTC': 0.864321608040201, 'TTT': 1.0, 'GGA': 0.8215814372606443, 'GGC': 1.0, 'GGG': 0.28835323271006985, 'GGT': 0.6037395809867087, 'CAC': 0.6798462383305877, 'CAT': 1.0, 'ATA': 0.10219657483246464, 'ATC': 0.8559195830230827, 'ATT': 1.0, 'AAA': 1.0, 'AAG': 0.3472094095940959, 'CTA': 0.13892709766162312, 'CTC': 0.30375974323704724, 'CTG': 0.6657496561210454, 'CTT': 1.0, 'TTA': 0.5476845483723063, 'TTG': 0.3874369555249885, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.8191662154845695, 'CCA': 0.4708680142687277, 'CCC': 0.06896551724137931, 'CCG': 1.0, 'CCT': 0.7439556084026953, 'CAA': 1.0, 'CAG': 0.8559696778269108, 'AGA': 0.7029352226720648, 'AGG': 0.09210526315789473, 'CGA': 0.18927125506072875, 'CGC': 1.0, 'CGG': 0.27834008097165996, 'CGT': 0.9903846153846154, 'AGC': 0.8852526926263463, 'AGT': 0.3177299088649544, 'TCA': 0.9254349627174814, 'TCC': 0.4258492129246064, 'TCG': 0.2531068765534383, 'TCT': 1.0, 'ACA': 1.0, 'ACC': 0.22235752557233318, 'ACG': 0.5411592791037506, 'ACT': 0.43180711154408186, 'GTA': 0.7223221166195737, 'GTC': 0.6935525301823786, 'GTG': 0.6699203698946828, 'GTT': 1.0, 'TGG': 1.0, 'TAC': 0.7833094555873925, 'TAT': 1.0}, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'GAA': 0.6798186496752849, 'AAT': 0.5655317497600131, 'ATA': 0.1287010322849589, 'TTA': 0.198530453110291, 'GAC': 0.35960560391899904, 'CTG': 0.23973441522197322, 'TGG': 1.0, 'AAC': 0.4344682502399869, 'CAA': 0.5131548459124133, 'GCC': 0.20715414603308396, 'CTT': 0.2391679232236727, 'GCT': 0.24752923822568748, 'ATC': 0.366762837741479, 'AAA': 0.7016557540022625, 'AAG': 0.29834424599773746, 'TTG': 0.1592175745811708, 'AGC': 0.22534685577492342, 'CCG': 0.4334091260292722, 'AGT': 0.10611598754150686, 'TTT': 0.6847751982489505, 'GAG': 0.3201813503247151, 'ACT': 0.1614000537105004, 'TCA': 0.23653118484388272, 'ACC': 0.1588786441082565, 'CAC': 0.3265866087018494, 'GGC': 0.3390368073029434, 'GAT': 0.640394396081001, 'ACA': 0.4122430101751559, 'ACG': 0.2674782920060872, 'CCC': 0.0889644363010221, 'AGA': 0.26380392311639106, 'TCC': 0.12712038919920718, 'TAC': 0.34586274736720285, 'CAT': 0.6734133912981506, 'ATT': 0.5045361299735621, 'GCA': 0.2828363712991255, 'TAT': 0.6541372526327971, 'GGG': 0.16211598104043537, 'GTC': 0.25598186049286953, 'CCT': 0.28667299498885185, 'GTT': 0.28439644370189154, 'GCG': 0.26248024444210305, 'TCT': 0.2048572678833432, 'CTC': 0.11219040795754642, 'CCA': 0.19095344268085387, 'GGA': 0.3148341038094681, 'CGA': 0.09855152543710778, 'GTA': 0.19774449549495793, 'GTG': 0.26187720031028106, 'TTC': 0.31522480175104955, 'CGC': 0.2064569601768165, 'CAG': 0.4868451540875866, 'CGG': 0.1555630106168844, 'CGT': 0.18161187196589967, 'CTA': 0.05115922590534585, 'AGG': 0.09401270868690058, 'GGT': 0.18401310784715313, 'TAG': 0.14132712456344587, 'TGC': 0.5460141582025239, 'TCG': 0.10002831475713661, 'TAA': 0.630500582072177, 'TGA': 0.2281722933643772, 'TGT': 0.45398584179747614}), 'cai_avg': 0.7334925635409068, 'cai_std': 0.042802002292947894, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}]}, 'orf_debug': {'total_loss': {'C': {'TGC': 1.0, 'TGT': 1.0105780473545745}, 'D': {'GAC': 0.8709203477283044, 'GAT': 1.0}, 'S': {'TCC': 0.767310748640736, 'AGC': 0.8970642883686363, 'TCG': 0.9109798056045083, 'AGT': 1.0143430625248522, 'TCT': 1.1116687578419073, 'TCA': 1.5957746973922708}, 'Q': {'CAG': 0.8741027056874654, 'CAA': 1.6724073380756272}, 'M': {'ATG': 1.0}, 'N': {'AAC': 1.0, 'AAT': 1.3628298062012105}, 'P': {'CCC': 0.9933441558441559, 'CCG': 1.0, 'CCA': 1.1386300978111386, 'CCT': 1.447545827476405}, 'K': {'AAA': 1.0, 'AAG': 1.051000900574087}, '_': {'TGA': 0.9604604988453446, 'TAA': 1.0, 'TAG': 1.03717715193125}, 'T': {'ACC': 0.5625428140841212, 'ACT': 1.0053841590051715, 'ACG': 1.1358339568986147, 'ACA': 1.871996685998343}, 'F': {'TTC': 0.8805309734513274, 'TTT': 1.344321329639889}, 'A': {'GCC': 0.9224789488619323, 'GCG': 0.9528769841269842, 'GCT': 1.3283237570851427, 'GCA': 1.369633681418863}, 'G': {'GGT': 0.7313645062044244, 'GGC': 1.0, 'GGG': 1.0583703213877516, 'GGA': 1.5984285372331444}, 'I': {'ATC': 0.8740644321509926, 'ATA': 1.0384123083805088, 'ATT': 1.2645605365337098}, 'L': {'CTG': 0.7494845360824742, 'CTA': 1.0623538715444918, 'CTC': 1.1044939732011871, 'TTG': 1.1673774549551195, 'TTA': 1.3122011147635726, 'CTT': 1.8928706505295008}, 'H': {'CAC': 0.7574875207986689, 'CAT': 1.2911214953271029}, 'R': {'CGT': 0.9904761904761905, 'AGG': 1.0431567154433548, 'CGA': 1.086254638087406, 'CGG': 1.1296003933480958, 'CGC': 1.2876220849113016, 'AGA': 1.5301103054141865}, 'W': {'TGG': 1.0}, 'V': {'GTG': 0.7518346852066434, 'GTT': 1.018539711604486, 'GTC': 1.1505169451198147, 'GTA': 1.1649581715973647}, 'E': {'GAA': 1.0, 'GAG': 1.0197094694060727}, 'Y': {'TAC': 0.8219016779511333, 'TAT': 1.1476923076923078}}, 'optimized_loss': {'C': defaultdict(<class 'float'>, {'TGT': 1.6432327117286945, 'TGC': 1.0}), 'D': defaultdict(<class 'float'>, {'GAT': 1.0, 'GAC': 1.32644912035017}), 'S': defaultdict(<class 'float'>, {'TCT': 1.2358074271617692, 'TCG': 2.5325041749163573, 'TCA': 2.940414537227555, 'TCC': 1.4589325892912588, 'AGC': 1.0, 'AGT': 2.911797534354834}), 'Q': defaultdict(<class 'float'>, {'CAA': 2.7969463044492056, 'CAG': 1.0}), 'M': defaultdict(<class 'float'>, {'ATG': 1.0}), 'N': defaultdict(<class 'float'>, {'AAC': 1.0, 'AAT': 2.5897675631491137}), 'P': defaultdict(<class 'float'>, {'CCT': 3.3057851239669427, 'CCG': 1.0, 'CCA': 3.0314838894201217, 'CCC': 3.679421487603306}), 'K': defaultdict(<class 'float'>, {'AAG': 3.0174622091692234, 'AAA': 1.0}), '_': defaultdict(<class 'float'>, {'TAG': 3.853223593964335, 'TGA': 3.0911615190738604, 'TAA': 1.0}), 'T': defaultdict(<class 'float'>, {'ACC': 1.0, 'ACA': 3.504371592388779, 'ACG': 2.7456517456510596, 'ACT': 2.4857819856416143}), 'F': defaultdict(<class 'float'>, {'TTT': 1.8071998373247593, 'TTC': 1.0}), 'A': defaultdict(<class 'float'>, {'GCA': 1.8758964212769875, 'GCC': 1.812421976933375, 'GCG': 1.0, 'GCT': 1.9423109359102864}), 'G': defaultdict(<class 'float'>, {'GGT': 1.0427990176022806, 'GGG': 3.2817356454694036, 'GGA': 3.548016240581601, 'GGC': 1.0}), 'I': defaultdict(<class 'float'>, {'ATC': 1.0, 'ATA': 3.883668089603532, 'ATT': 1.599113350558424}), 'L': defaultdict(<class 'float'>, {'TTA': 3.631806659913921, 'TTG': 3.5436923326098997, 'CTC': 3.509954066352956, 'CTT': 3.5829592996359754, 'CTG': 1.0, 'CTA': 3.908995597888406}), 'H': defaultdict(<class 'float'>, {'CAT': 1.6669947156956941, 'CAC': 1.0}), 'R': defaultdict(<class 'float'>, {'CGA': 3.868744772855011, 'CGC': 1.6579706335513271, 'CGG': 3.782199282654244, 'CGT': 1.0, 'AGG': 3.9610282186271286, 'AGA': 3.9388442858165034}), 'W': defaultdict(<class 'float'>, {'TGG': 1.0}), 'V': defaultdict(<class 'float'>, {'GTA': 2.2154574764955095, 'GTC': 2.259279403605198, 'GTG': 1.0, 'GTT': 1.0374231441153496}), 'E': defaultdict(<class 'float'>, {'GAG': 2.6854688521243797, 'GAA': 1.0}), 'Y': defaultdict(<class 'float'>, {'TAT': 1.3171976331360948, 'TAC': 1.0})}, 'deoptimized_loss': {'C': defaultdict(<class 'float'>, {'TGT': 1.6090122554987418, 'TGC': 1.0}), 'D': defaultdict(<class 'float'>, {'GAT': 1.0, 'GAC': 1.7487741357394064}), 'S': defaultdict(<class 'float'>, {'TCT': 1.0, 'TCG': 3.051635584744884, 'TCA': 1.1546900193499807, 'TCC': 2.4779507004500814, 'AGC': 1.2426615592968109, 'AGT': 2.830032659527515}), 'Q': defaultdict(<class 'float'>, {'CAA': 1.0, 'CAG': 1.3088053780514624}), 'M': defaultdict(<class 'float'>, {'ATG': 1.0}), 'N': defaultdict(<class 'float'>, {'AAC': 1.0, 'AAT': 1.3943684266530345}), 'P': defaultdict(<class 'float'>, {'CCT': 1.5776475136630435, 'CCG': 1.0, 'CCA': 2.3382446297864634, 'CCC': 3.7288941736028534}), 'K': defaultdict(<class 'float'>, {'AAG': 2.731716735734297, 'AAA': 1.0}), '_': defaultdict(<class 'float'>, {'TAG': 3.5819401803361797, 'TGA': 3.3509093705321793, 'TAA': 1.0}), 'T': defaultdict(<class 'float'>, {'ACC': 3.1600127668893183, 'ACA': 1.0, 'ACG': 2.128216248945088, 'ACT': 2.459228935403716}), 'F': defaultdict(<class 'float'>, {'TTT': 1.0, 'TTC': 1.289765409964395}), 'A': defaultdict(<class 'float'>, {'GCA': 1.0, 'GCC': 2.129837116329259, 'GCG': 1.101352459094135, 'GCT': 1.100806221057107}), 'G': defaultdict(<class 'float'>, {'GGT': 1.9495431577031723, 'GGG': 2.9297346559740687, 'GGA': 1.388670309008689, 'GGC': 1.0}), 'I': defaultdict(<class 'float'>, {'ATC': 1.3089200005100767, 'ATA': 3.6016578405776287, 'ATT': 1.0}), 'L': defaultdict(<class 'float'>, {'TTA': 2.109220171036552, 'TTG': 2.600359572406518, 'CTC': 2.8772310086632475, 'CTT': 1.0, 'CTG': 1.7802239801411086, 'CTA': 3.46359234781819}), 'H': defaultdict(<class 'float'>, {'CAT': 1.0, 'CAC': 1.7428059544498995}), 'R': defaultdict(<class 'float'>, {'CGA': 3.2787385877493485, 'CGC': 1.0, 'CGG': 2.96411287678867, 'CGT': 1.019323224852071, 'AGG': 3.6400623268698062, 'AGA': 1.682377036584766}), 'W': defaultdict(<class 'float'>, {'TGG': 1.0}), 'V': defaultdict(<class 'float'>, {'GTA': 1.6324607736794863, 'GTC': 1.7068049913928647, 'GTG': 1.7691118224210975, 'GTT': 1.0}), 'E': defaultdict(<class 'float'>, {'GAG': 2.5826598723651326, 'GAA': 1.0}), 'Y': defaultdict(<class 'float'>, {'TAT': 1.0, 'TAC': 1.4803358808630471})}}, 'orf': {'orf_module_input_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'optimized_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTGCACATGGAAGGTTCCGTGAACGGTCACGAATTCGAAATCGAAGGTGAAGGTGAAGGTCGTCCCTACGAAGGTACCCAGACCGCCAAACTGAAAGTGACCAAAGGTGGCCCCCTGCCCTTCGCCTGGGACATCCTGTCCCCCCAGTTCATGTACGGTTCCAAAGCCTACGTGAAACACCCCGCCGACATCCCCGACTACCTGAAACTGTCCTTCCCCGAAGGTTTCAAATGGGAACGTGTGATGAACTTCGAAGACGGTGGCGTGGTTACCGTGACCCAGGACTCCAGCCTGCAGGACGGTGAATTCATCTACAAAGTGAAACTGCGTGGTACCAACTTCCCCTCCGACGGTCCCGTGATGCAGAAAAAGACCATGGGTTGGGAAGCCTCCAGCGAACGTATGTACCCCGAAGACGGTGCCCTGAAAGGTGAAATCAAACAGCGTCTGAAACTGAAAGACGGTGGCCACTACGACGCCGAAGTGAAAACCACTTACAAAGCCAAAAAGCCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTGGAACAGTACGAACGTGCCGAAGGTCGTCACTCCACCGGTGGCATGGACGAACTGTACAAATGA', 'aa_to_loss_mapping': {'C': {'TGC': 1.0, 'TGT': 1.0105780473545745}, 'D': {'GAC': 0.8709203477283044, 'GAT': 1.0}, 'S': {'TCC': 0.767310748640736, 'AGC': 0.8970642883686363, 'TCG': 0.9109798056045083, 'AGT': 1.0143430625248522, 'TCT': 1.1116687578419073, 'TCA': 1.5957746973922708}, 'Q': {'CAG': 0.8741027056874654, 'CAA': 1.6724073380756272}, 'M': {'ATG': 1.0}, 'N': {'AAC': 1.0, 'AAT': 1.3628298062012105}, 'P': {'CCC': 0.9933441558441559, 'CCG': 1.0, 'CCA': 1.1386300978111386, 'CCT': 1.447545827476405}, 'K': {'AAA': 1.0, 'AAG': 1.051000900574087}, '_': {'TGA': 0.9604604988453446, 'TAA': 1.0, 'TAG': 1.03717715193125}, 'T': {'ACC': 0.5625428140841212, 'ACT': 1.0053841590051715, 'ACG': 1.1358339568986147, 'ACA': 1.871996685998343}, 'F': {'TTC': 0.8805309734513274, 'TTT': 1.344321329639889}, 'A': {'GCC': 0.9224789488619323, 'GCG': 0.9528769841269842, 'GCT': 1.3283237570851427, 'GCA': 1.369633681418863}, 'G': {'GGT': 0.7313645062044244, 'GGC': 1.0, 'GGG': 1.0583703213877516, 'GGA': 1.5984285372331444}, 'I': {'ATC': 0.8740644321509926, 'ATA': 1.0384123083805088, 'ATT': 1.2645605365337098}, 'L': {'CTG': 0.7494845360824742, 'CTA': 1.0623538715444918, 'CTC': 1.1044939732011871, 'TTG': 1.1673774549551195, 'TTA': 1.3122011147635726, 'CTT': 1.8928706505295008}, 'H': {'CAC': 0.7574875207986689, 'CAT': 1.2911214953271029}, 'R': {'CGT': 0.9904761904761905, 'AGG': 1.0431567154433548, 'CGA': 1.086254638087406, 'CGG': 1.1296003933480958, 'CGC': 1.2876220849113016, 'AGA': 1.5301103054141865}, 'W': {'TGG': 1.0}, 'V': {'GTG': 0.7518346852066434, 'GTT': 1.018539711604486, 'GTC': 1.1505169451198147, 'GTA': 1.1649581715973647}, 'E': {'GAA': 1.0, 'GAG': 1.0197094694060727}, 'Y': {'TAC': 0.8219016779511333, 'TAT': 1.1476923076923078}}, 'run_time': 0.75}, 'evaluation': [{'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'cai_initial_score': 0.7915994634162647, 'cai_final_score': 0.8120707252714924, 'dist_score': 0.218293345134776}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'cai_initial_score': 0.7695788606254016, 'cai_final_score': 0.6156387084491994, 'dist_score': -3.5965642710496635}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTGCACATGGAAGGTTCCGTGAACGGTCACGAATTCGAAATCGAAGGTGAAGGTGAAGGTCGTCCCTACGAAGGTACCCAGACCGCCAAACTGAAAGTGACCAAAGGTGGCCCCCTGCCCTTCGCCTGGGACATCCTGTCCCCCCAGTTCATGTACGGTTCCAAAGCCTACGTGAAACACCCCGCCGACATCCCCGACTACCTGAAACTGTCCTTCCCCGAAGGTTTCAAATGGGAACGTGTGATGAACTTCGAAGACGGTGGCGTGGTTACCGTGACCCAGGACTCCAGCCTGCAGGACGGTGAATTCATCTACAAAGTGAAACTGCGTGGTACCAACTTCCCCTCCGACGGTCCCGTGATGCAGAAAAAGACCATGGGTTGGGAAGCCTCCAGCGAACGTATGTACCCCGAAGACGGTGCCCTGAAAGGTGAAATCAAACAGCGTCTGAAACTGAAAGACGGTGGCCACTACGACGCCGAAGTGAAAACCACTTACAAAGCCAAAAAGCCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTGGAACAGTACGAACGTGCCGAAGGTCGTCACTCCACCGGTGGCATGGACGAACTGTACAAATGA', 'average_distance_score': 1.9074288080922197, 'weakest_link_score': 1.9074288080922197, 'ratio_score': 1.0833299803552052, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTCATGCGTTTCAAAGTGCACATGGAAGGTTCCGTGAACGGTCACGAATTCGAAATCGAAGGTGAAGGTGAAGGTCGTCCCTACGAAGGTACCCAGACCGCCAAACTGAAAGTGACCAAAGGTGGCCCCCTGCCCTTCGCCTGGGACATCCTGTCCCCCCAGTTCATGTACGGTTCCAAAGCCTACGTGAAACACCCCGCCGACATCCCCGACTACCTGAAACTGTCCTTCCCCGAAGGTTTCAAATGGGAACGTGTGATGAACTTCGAAGACGGTGGCGTGGTTACCGTGACCCAGGACTCCAGCCTGCAGGACGGTGAATTCATCTACAAAGTGAAACTGCGTGGTACCAACTTCCCCTCCGACGGTCCCGTGATGCAGAAAAAGACCATGGGTTGGGAAGCCTCCAGCGAACGTATGTACCCCGAAGACGGTGCCCTGAAAGGTGAAATCAAACAGCGTCTGAAACTGAAAGACGGTGGCCACTACGACGCCGAAGTGAAAACCACTTACAAAGCCAAAAAGCCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTGGAACAGTACGAACGTGCCGAAGGTCGTCACTCCACCGGTGGCATGGACGAACTGTACAAATGA'}], 'final_evaluation': {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTGCACATGGAAGGTTCCGTGAACGGTCACGAATTCGAAATCGAAGGTGAAGGTGAAGGTCGTCCCTACGAAGGTACCCAGACCGCCAAACTGAAAGTGACCAAAGGTGGCCCCCTGCCCTTCGCCTGGGACATCCTGTCCCCCCAGTTCATGTACGGTTCCAAAGCCTACGTGAAACACCCCGCCGACATCCCCGACTACCTGAAACTGTCCTTCCCCGAAGGTTTCAAATGGGAACGTGTGATGAACTTCGAAGACGGTGGCGTGGTTACCGTGACCCAGGACTCCAGCCTGCAGGACGGTGAATTCATCTACAAAGTGAAACTGCGTGGTACCAACTTCCCCTCCGACGGTCCCGTGATGCAGAAAAAGACCATGGGTTGGGAAGCCTCCAGCGAACGTATGTACCCCGAAGACGGTGCCCTGAAAGGTGAAATCAAACAGCGTCTGAAACTGAAAGACGGTGGCCACTACGACGCCGAAGTGAAAACCACTTACAAAGCCAAAAAGCCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTGGAACAGTACGAACGTGCCGAAGGTCGTCACTCCACCGGTGGCATGGACGAACTGTACAAATGA', 'average_distance_score': 1.9074288080922197, 'weakest_link_score': 1.9074288080922197, 'ratio_score': 1.0833299803552052}, 'zip_output_file_path': 'results\\\\mcherry-wanted-constraint-init-external-0.1\\\\CAI_single_codon_ratio_ecoli_opt_True_XSuW\\\\communique_results.zip'}\n",
      "\n",
      "##########################\n",
      "# User Input #\n",
      "##########################\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is:\n",
      "ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "------------------------------------------\n",
      "Parsing information for Bacillus subtilis:\n",
      "------------------------------------------\n",
      "Organism is defined as wanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 98 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3544\n",
      "Calculate CAI weights from a reference set of 532 highly expressed genes from estimated expression dictionary.\n",
      "cai_std=0.042802002292947894, cai_avg=0.7334925635409068\n",
      "------------------------------------------\n",
      "Parsing information for Escherichia coli:\n",
      "------------------------------------------\n",
      "Organism is defined as unwanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 90 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3893\n",
      "Calculate CAI weights from a reference set of 584 highly expressed genes from estimated expression dictionary.\n",
      "cai_std=0.09377868043841883, cai_avg=0.6151304741138522\n",
      "-----------------------------\n",
      "Normalized Prioritization Weights\n",
      "-----------------------------\n",
      "Bacillus subtilis : 1.0\n",
      "Escherichia coli : 1.0\n",
      "Total input processing time: 27.842395782470703\n",
      "\n",
      "##########################\n",
      "# Initiation #\n",
      "##########################\n",
      "Running initiation optimization InitiationOptimizationMethod.external_module on 15 codons at the start of the ORF\n",
      "Taking optimized sequence from configuration value config.INITIATION.EXTERNAL_INITIATION_ORF\n",
      "Optimized sequence is: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "Optimization method used is: ORFOptimizationMethod.single_codon_ratio\n",
      "Loss summary for: TGT\n",
      "loss for Bacillus subtilis: 1.6090122554987418\n",
      "loss for Escherichia coli: 1.6432327117286945\n",
      "Loss summary for: TGC\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGT': 0.9895326764891983, 'TGC': 1.0})\n",
      "Loss summary for: GAT\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: GAC\n",
      "loss for Bacillus subtilis: 1.7487741357394064\n",
      "loss for Escherichia coli: 1.32644912035017\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAT': 1.0, 'GAC': 1.148210628685373})\n",
      "Loss summary for: TCT\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.2358074271617692\n",
      "Loss summary for: TCG\n",
      "loss for Bacillus subtilis: 3.051635584744884\n",
      "loss for Escherichia coli: 2.5325041749163573\n",
      "Loss summary for: TCA\n",
      "loss for Bacillus subtilis: 1.1546900193499807\n",
      "loss for Escherichia coli: 2.940414537227555\n",
      "Loss summary for: TCC\n",
      "loss for Bacillus subtilis: 2.4779507004500814\n",
      "loss for Escherichia coli: 1.4589325892912588\n",
      "Loss summary for: AGC\n",
      "loss for Bacillus subtilis: 1.2426615592968109\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: AGT\n",
      "loss for Bacillus subtilis: 2.830032659527515\n",
      "loss for Escherichia coli: 2.911797534354834\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TCT': 0.8995485327313768, 'TCG': 1.0977191742866568, 'TCA': 0.6266548790591467, 'TCC': 1.3032529542580562, 'AGC': 1.1147473073736536, 'AGT': 0.9858597519372291})\n",
      "Loss summary for: CAA\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 2.7969463044492056\n",
      "Loss summary for: CAG\n",
      "loss for Bacillus subtilis: 1.3088053780514624\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAA': 0.5979404522050593, 'CAG': 1.1440303221730892})\n",
      "Loss summary for: ATG\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATG': 1.0})\n",
      "Loss summary for: AAC\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: AAT\n",
      "loss for Bacillus subtilis: 1.3943684266530345\n",
      "loss for Escherichia coli: 2.5897675631491137\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAC': 1.0, 'AAT': 0.7337673387019819})\n",
      "Loss summary for: CCT\n",
      "loss for Bacillus subtilis: 1.5776475136630435\n",
      "loss for Escherichia coli: 3.3057851239669427\n",
      "Loss summary for: CCG\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: CCA\n",
      "loss for Bacillus subtilis: 2.3382446297864634\n",
      "loss for Escherichia coli: 3.0314838894201217\n",
      "Loss summary for: CCC\n",
      "loss for Bacillus subtilis: 3.7288941736028534\n",
      "loss for Escherichia coli: 3.679421487603306\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CCT': 0.6908244153785176, 'CCG': 1.0, 'CCA': 0.8782483459047534, 'CCC': 1.0067004412485698})\n",
      "Loss summary for: AAG\n",
      "loss for Bacillus subtilis: 2.731716735734297\n",
      "loss for Escherichia coli: 3.0174622091692234\n",
      "Loss summary for: AAA\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAG': 0.9514739706253069, 'AAA': 1.0})\n",
      "Loss summary for: TAG\n",
      "loss for Bacillus subtilis: 3.5819401803361797\n",
      "loss for Escherichia coli: 3.853223593964335\n",
      "Loss summary for: TGA\n",
      "loss for Bacillus subtilis: 3.3509093705321793\n",
      "loss for Escherichia coli: 3.0911615190738604\n",
      "Loss summary for: TAA\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAG': 0.9641554464808395, 'TGA': 1.041167233011862, 'TAA': 1.0})\n",
      "Loss summary for: ACC\n",
      "loss for Bacillus subtilis: 3.1600127668893183\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: ACA\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 3.504371592388779\n",
      "Loss summary for: ACG\n",
      "loss for Bacillus subtilis: 2.128216248945088\n",
      "loss for Escherichia coli: 2.7456517456510596\n",
      "Loss summary for: ACT\n",
      "loss for Bacillus subtilis: 2.459228935403716\n",
      "loss for Escherichia coli: 2.4857819856416143\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ACC': 1.7776424744276667, 'ACA': 0.5341889798628014, 'ACG': 0.8804103750608864, 'ACT': 0.9946446749166018})\n",
      "Loss summary for: TTT\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.8071998373247593\n",
      "Loss summary for: TTC\n",
      "loss for Bacillus subtilis: 1.289765409964395\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTT': 0.7438697712754998, 'TTC': 1.135678391959799})\n",
      "Loss summary for: GCA\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.8758964212769875\n",
      "Loss summary for: GCC\n",
      "loss for Bacillus subtilis: 2.129837116329259\n",
      "loss for Escherichia coli: 1.812421976933375\n",
      "Loss summary for: GCG\n",
      "loss for Bacillus subtilis: 1.101352459094135\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: GCT\n",
      "loss for Bacillus subtilis: 1.100806221057107\n",
      "loss for Escherichia coli: 1.9423109359102864\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GCA': 0.7301222316349993, 'GCC': 1.0840355774337245, 'GCG': 1.049453409682457, 'GCT': 0.7528285138815763})\n",
      "Loss summary for: GGT\n",
      "loss for Bacillus subtilis: 1.9495431577031723\n",
      "loss for Escherichia coli: 1.0427990176022806\n",
      "Loss summary for: GGG\n",
      "loss for Bacillus subtilis: 2.9297346559740687\n",
      "loss for Escherichia coli: 3.2817356454694036\n",
      "Loss summary for: GGA\n",
      "loss for Bacillus subtilis: 1.388670309008689\n",
      "loss for Escherichia coli: 3.548016240581601\n",
      "Loss summary for: GGC\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GGT': 1.3673072613131283, 'GGG': 0.9448488679168409, 'GGA': 0.6256144561401443, 'GGC': 1.0})\n",
      "Loss summary for: ATC\n",
      "loss for Bacillus subtilis: 1.3089200005100767\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: ATA\n",
      "loss for Bacillus subtilis: 3.6016578405776287\n",
      "loss for Escherichia coli: 3.883668089603532\n",
      "Loss summary for: ATT\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.599113350558424\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATC': 1.1440804169769172, 'ATA': 0.9630086160665641, 'ATT': 0.7907885554780182})\n",
      "Loss summary for: TTA\n",
      "loss for Bacillus subtilis: 2.109220171036552\n",
      "loss for Escherichia coli: 3.631806659913921\n",
      "Loss summary for: TTG\n",
      "loss for Bacillus subtilis: 2.600359572406518\n",
      "loss for Escherichia coli: 3.5436923326098997\n",
      "Loss summary for: CTC\n",
      "loss for Bacillus subtilis: 2.8772310086632475\n",
      "loss for Escherichia coli: 3.509954066352956\n",
      "Loss summary for: CTT\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 3.5829592996359754\n",
      "Loss summary for: CTG\n",
      "loss for Bacillus subtilis: 1.7802239801411086\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: CTA\n",
      "loss for Bacillus subtilis: 3.46359234781819\n",
      "loss for Escherichia coli: 3.908995597888406\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTA': 0.7620783039649957, 'TTG': 0.8566209633014076, 'CTC': 0.90539199331407, 'CTT': 0.5282981167890504, 'CTG': 1.3342503438789546, 'CTA': 0.9413059309005583})\n",
      "Loss summary for: CAT\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.6669947156956941\n",
      "Loss summary for: CAC\n",
      "loss for Bacillus subtilis: 1.7428059544498995\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAT': 0.7745204487875498, 'CAC': 1.3201537616694123})\n",
      "Loss summary for: CGA\n",
      "loss for Bacillus subtilis: 3.2787385877493485\n",
      "loss for Escherichia coli: 3.868744772855011\n",
      "Loss summary for: CGC\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.6579706335513271\n",
      "Loss summary for: CGG\n",
      "loss for Bacillus subtilis: 2.96411287678867\n",
      "loss for Escherichia coli: 3.782199282654244\n",
      "Loss summary for: CGT\n",
      "loss for Bacillus subtilis: 1.019323224852071\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: AGG\n",
      "loss for Bacillus subtilis: 3.6400623268698062\n",
      "loss for Escherichia coli: 3.9610282186271286\n",
      "Loss summary for: AGA\n",
      "loss for Bacillus subtilis: 1.682377036584766\n",
      "loss for Escherichia coli: 3.9388442858165034\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CGA': 0.920594458184062, 'CGC': 0.776625386996904, 'CGG': 0.885268813545678, 'CGT': 1.0096153846153846, 'AGG': 0.958628732572543, 'AGA': 0.6535476536963193})\n",
      "Loss summary for: TGG\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGG': 1.0})\n",
      "Loss summary for: GTA\n",
      "loss for Bacillus subtilis: 1.6324607736794863\n",
      "loss for Escherichia coli: 2.2154574764955095\n",
      "Loss summary for: GTC\n",
      "loss for Bacillus subtilis: 1.7068049913928647\n",
      "loss for Escherichia coli: 2.259279403605198\n",
      "Loss summary for: GTG\n",
      "loss for Bacillus subtilis: 1.7691118224210975\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: GTT\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.0374231441153496\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GTA': 0.8583999188819134, 'GTC': 0.8691745082432142, 'GTG': 1.3300796301053173, 'GTT': 0.9817977528089888})\n",
      "Loss summary for: GAG\n",
      "loss for Bacillus subtilis: 2.5826598723651326\n",
      "loss for Escherichia coli: 2.6854688521243797\n",
      "Loss summary for: GAA\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAG': 0.9806714853618527, 'GAA': 1.0})\n",
      "Loss summary for: TAT\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.3171976331360948\n",
      "Loss summary for: TAC\n",
      "loss for Bacillus subtilis: 1.4803358808630471\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAT': 0.8713136729222519, 'TAC': 1.2166905444126075})\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6847751982489505]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26380392311639106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6847751982489505]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6734133912981506]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23653118484388272]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5655317497600131]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6734133912981506]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6847751982489505]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5045361299735621]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26380392311639106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.28667299498885185]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2391679232236727]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.16211598104043537]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.28667299498885185]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2391679232236727]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.28667299498885185]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6847751982489505]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.640394396081001]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5045361299735621]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2391679232236727]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23653118484388272]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.28667299498885185]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6847751982489505]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23653118484388272]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6734133912981506]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.28667299498885185]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.640394396081001]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5045361299735621]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.28667299498885185]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.640394396081001]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2391679232236727]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2391679232236727]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23653118484388272]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6847751982489505]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.28667299498885185]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6847751982489505]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26380392311639106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5655317497600131]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6847751982489505]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.640394396081001]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.16211598104043537]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.25598186049286953]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.640394396081001]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23653118484388272]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23653118484388272]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2048572678833432]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2391679232236727]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.640394396081001]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6847751982489505]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5045361299735621]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2391679232236727]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26380392311639106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5655317497600131]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6847751982489505]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.28667299498885185]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23653118484388272]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.640394396081001]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.28667299498885185]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7016557540022625]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23653118484388272]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23653118484388272]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2048572678833432]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26380392311639106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.28667299498885185]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.640394396081001]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2391679232236727]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5045361299735621]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26380392311639106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2391679232236727]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2391679232236727]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.640394396081001]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.16211598104043537]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6734133912981506]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.640394396081001]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2674782920060872]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7016557540022625]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.28667299498885185]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2391679232236727]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.28667299498885185]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5655317497600131]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5655317497600131]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5045361299735621]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2391679232236727]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.640394396081001]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5045361299735621]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23653118484388272]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6734133912981506]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5655317497600131]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.640394396081001]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5045361299735621]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26380392311639106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26380392311639106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6734133912981506]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23653118484388272]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.16211598104043537]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.640394396081001]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2391679232236727]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14132712456344587]\n",
      "\n",
      "##########################\n",
      "# EVALUATION #\n",
      "##########################\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTTATGAGATTTAAGGTACATATGGAGGGATCAGTAAATGGACATGAGTTTGAGATTGAGGGAGAGGGAGAGGGAAGACCTTATGAGGGAACACAAACAGCAAAGCTTAAGGTAACAAAGGGAGGGCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAGGCATATGTAAAGCATCCTGCAGATATTCCTGATTATCTTAAGCTTTCATTTCCTGAGGGATTTAAGTGGGAGAGAGTAATGAATTTTGAGGATGGAGGGGTAGTCACAGTAACACAAGATTCATCTCTTCAAGATGGAGAGTTTATTTATAAGGTAAAGCTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTAATGCAAAAGAAAACAATGGGATGGGAGGCATCATCTGAGAGAATGTATCCTGAGGATGGAGCACTTAAGGGAGAGATTAAGCAAAGACTTAAGCTTAAGGATGGAGGGCATTATGATGCAGAGGTAAAGACAACGTATAAGGCAAAGAAACCTGTACAACTTCCTGGAGCATATAATGTAAATATTAAGCTTGATATTACATCACATAATGAGGATTATACAATTGTAGAGCAATATGAGAGAGCAGAGGGAAGACATTCAACAGGAGGGATGGATGAGCTTTATAAGTAG\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAGGTACATATGGAGGGATCAGTAAATGGACATGAGTTTGAGATTGAGGGAGAGGGAGAGGGAAGACCTTATGAGGGAACACAAACAGCAAAGCTTAAGGTAACAAAGGGAGGGCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAGGCATATGTAAAGCATCCTGCAGATATTCCTGATTATCTTAAGCTTTCATTTCCTGAGGGATTTAAGTGGGAGAGAGTAATGAATTTTGAGGATGGAGGGGTAGTCACAGTAACACAAGATTCATCTCTTCAAGATGGAGAGTTTATTTATAAGGTAAAGCTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTAATGCAAAAGAAAACAATGGGATGGGAGGCATCATCTGAGAGAATGTATCCTGAGGATGGAGCACTTAAGGGAGAGATTAAGCAAAGACTTAAGCTTAAGGATGGAGGGCATTATGATGCAGAGGTAAAGACAACGTATAAGGCAAAGAAACCTGTACAACTTCCTGGAGCATATAATGTAAATATTAAGCTTGATATTACATCACATAATGAGGATTATACAATTGTAGAGCAATATGAGAGAGCAGAGGGAAGACATTCAACAGGAGGGATGGATGAGCTTTATAAGTAG', 'average_distance_score': 2.2333536928868796, 'weakest_link_score': 2.2333536928868796, 'ratio_score': 1.1047359569220119}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAGGTACATATGGAGGGATCAGTAAATGGACATGAGTTTGAGATTGAGGGAGAGGGAGAGGGAAGACCTTATGAGGGAACACAAACAGCAAAGCTTAAGGTAACAAAGGGAGGGCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAGGCATATGTAAAGCATCCTGCAGATATTCCTGATTATCTTAAGCTTTCATTTCCTGAGGGATTTAAGTGGGAGAGAGTAATGAATTTTGAGGATGGAGGGGTAGTCACAGTAACACAAGATTCATCTCTTCAAGATGGAGAGTTTATTTATAAGGTAAAGCTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTAATGCAAAAGAAAACAATGGGATGGGAGGCATCATCTGAGAGAATGTATCCTGAGGATGGAGCACTTAAGGGAGAGATTAAGCAAAGACTTAAGCTTAAGGATGGAGGGCATTATGATGCAGAGGTAAAGACAACGTATAAGGCAAAGAAACCTGTACAACTTCCTGGAGCATATAATGTAAATATTAAGCTTGATATTACATCACATAATGAGGATTATACAATTGTAGAGCAATATGAGAGAGCAGAGGGAAGACATTCAACAGGAGGGATGGATGAGCTTTATAAGTAG', 'average_distance_score': 2.2333536928868796, 'weakest_link_score': 2.2333536928868796, 'ratio_score': 1.1047359569220119}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry-wanted-constraint-init-external-0.1\\CAI_single_codon_ratio_ecoli_opt_False_6rNS\n",
      "Final output: {'module_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'orf_optimization_method': 'single_codon_ratio', 'orf_optimization_cub_index': 'CAI', 'initiation_optimization_method': 'external', 'evaluation_score': 'average_distance', 'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.10739856801909307, 'TGA': 0.1694510739856802, 'GCA': 1.0, 'GCC': 0.5406038521603331, 'GCG': 0.950546590317543, 'GCT': 0.9508068714211347, 'TGC': 1.0, 'TGT': 0.7315315315315315, 'GAC': 0.6775877587758776, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.3929343907714492, 'TTC': 0.864321608040201, 'TTT': 1.0, 'GGA': 0.8215814372606443, 'GGC': 1.0, 'GGG': 0.28835323271006985, 'GGT': 0.6037395809867087, 'CAC': 0.6798462383305877, 'CAT': 1.0, 'ATA': 0.10219657483246464, 'ATC': 0.8559195830230827, 'ATT': 1.0, 'AAA': 1.0, 'AAG': 0.3472094095940959, 'CTA': 0.13892709766162312, 'CTC': 0.30375974323704724, 'CTG': 0.6657496561210454, 'CTT': 1.0, 'TTA': 0.5476845483723063, 'TTG': 0.3874369555249885, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.8191662154845695, 'CCA': 0.4708680142687277, 'CCC': 0.06896551724137931, 'CCG': 1.0, 'CCT': 0.7439556084026953, 'CAA': 1.0, 'CAG': 0.8559696778269108, 'AGA': 0.7029352226720648, 'AGG': 0.09210526315789473, 'CGA': 0.18927125506072875, 'CGC': 1.0, 'CGG': 0.27834008097165996, 'CGT': 0.9903846153846154, 'AGC': 0.8852526926263463, 'AGT': 0.3177299088649544, 'TCA': 0.9254349627174814, 'TCC': 0.4258492129246064, 'TCG': 0.2531068765534383, 'TCT': 1.0, 'ACA': 1.0, 'ACC': 0.22235752557233318, 'ACG': 0.5411592791037506, 'ACT': 0.43180711154408186, 'GTA': 0.7223221166195737, 'GTC': 0.6935525301823786, 'GTG': 0.6699203698946828, 'GTT': 1.0, 'TGG': 1.0, 'TAC': 0.7833094555873925, 'TAT': 1.0}, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'GAA': 0.6798186496752849, 'AAT': 0.5655317497600131, 'ATA': 0.1287010322849589, 'TTA': 0.198530453110291, 'GAC': 0.35960560391899904, 'CTG': 0.23973441522197322, 'TGG': 1.0, 'AAC': 0.4344682502399869, 'CAA': 0.5131548459124133, 'GCC': 0.20715414603308396, 'CTT': 0.2391679232236727, 'GCT': 0.24752923822568748, 'ATC': 0.366762837741479, 'AAA': 0.7016557540022625, 'AAG': 0.29834424599773746, 'TTG': 0.1592175745811708, 'AGC': 0.22534685577492342, 'CCG': 0.4334091260292722, 'AGT': 0.10611598754150686, 'TTT': 0.6847751982489505, 'GAG': 0.3201813503247151, 'ACT': 0.1614000537105004, 'TCA': 0.23653118484388272, 'ACC': 0.1588786441082565, 'CAC': 0.3265866087018494, 'GGC': 0.3390368073029434, 'GAT': 0.640394396081001, 'ACA': 0.4122430101751559, 'ACG': 0.2674782920060872, 'CCC': 0.0889644363010221, 'AGA': 0.26380392311639106, 'TCC': 0.12712038919920718, 'TAC': 0.34586274736720285, 'CAT': 0.6734133912981506, 'ATT': 0.5045361299735621, 'GCA': 0.2828363712991255, 'TAT': 0.6541372526327971, 'GGG': 0.16211598104043537, 'GTC': 0.25598186049286953, 'CCT': 0.28667299498885185, 'GTT': 0.28439644370189154, 'GCG': 0.26248024444210305, 'TCT': 0.2048572678833432, 'CTC': 0.11219040795754642, 'CCA': 0.19095344268085387, 'GGA': 0.3148341038094681, 'CGA': 0.09855152543710778, 'GTA': 0.19774449549495793, 'GTG': 0.26187720031028106, 'TTC': 0.31522480175104955, 'CGC': 0.2064569601768165, 'CAG': 0.4868451540875866, 'CGG': 0.1555630106168844, 'CGT': 0.18161187196589967, 'CTA': 0.05115922590534585, 'AGG': 0.09401270868690058, 'GGT': 0.18401310784715313, 'TAG': 0.14132712456344587, 'TGC': 0.5460141582025239, 'TCG': 0.10002831475713661, 'TAA': 0.630500582072177, 'TGA': 0.2281722933643772, 'TGT': 0.45398584179747614}), 'cai_avg': 0.7334925635409068, 'cai_std': 0.042802002292947894, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}, {'name': 'Escherichia coli', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.037037037037037035, 'TGA': 0.24183006535947713, 'GCA': 0.6303663185811371, 'GCC': 0.653737775567711, 'GCG': 1.0, 'GCT': 0.6063318415382065, 'TGC': 1.0, 'TGT': 0.7181136120042874, 'GAC': 0.848284271032921, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.36125998031280765, 'TTC': 1.0, 'TTT': 0.6556786703601108, 'GGA': 0.11638214051214708, 'GGC': 1.0, 'GGG': 0.18844386080105055, 'GGT': 0.9788246881155614, 'CAC': 1.0, 'CAT': 0.7088785046728973, 'ATA': 0.029297564419343447, 'ATC': 1.0, 'ATT': 0.7354394634662901, 'AAA': 1.0, 'AAG': 0.26291560102301786, 'CTA': 0.02288199697428139, 'CTC': 0.1265128593040847, 'CTG': 1.0, 'CTT': 0.10712934947049924, 'TTA': 0.09427004538577911, 'TTG': 0.11753025718608169, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.39072452229299365, 'CCA': 0.2588842975206611, 'CCC': 0.08181818181818182, 'CCG': 1.0, 'CCT': 0.1818181818181818, 'CAA': 0.32759266192437286, 'CAG': 1.0, 'AGA': 0.015347817420769383, 'AGG': 0.00976679290412597, 'CGA': 0.033087502491528795, 'CGC': 0.7123779150886984, 'CGG': 0.05521227825393661, 'CGT': 1.0, 'AGC': 1.0, 'AGT': 0.2936010037641154, 'TCA': 0.28523630280217477, 'TCC': 0.7921371810957758, 'TCG': 0.4086156419907988, 'TCT': 0.8883312421580928, 'ACA': 0.12800331400165701, 'ACC': 1.0, 'ACG': 0.34299917149958575, 'ACT': 0.4233637116818558, 'GTA': 0.5115587090867476, 'GTC': 0.49691004806591893, 'GTG': 1.0, 'GTT': 0.9814602883955139, 'TGG': 1.0, 'TAC': 1.0, 'TAT': 0.8523076923076923}, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'AAA': 0.7667390608953314, 'CGC': 0.40076028262096014, 'ATT': 0.5085852538544149, 'AGC': 0.27799163607995603, 'ACC': 0.4359429381095841, 'ATC': 0.4205725148436632, 'ACA': 0.13023262212317555, 'GGT': 0.33726003930064996, 'AAC': 0.5512262248110238, 'GCG': 0.3563679190097728, 'GGC': 0.4050486219579785, 'TGA': 0.28926761860893596, 'CGA': 0.06388523224887838, 'GTG': 0.3715940695510341, 'TTG': 0.12808970030428077, 'AAG': 0.23326093910466858, 'TTC': 0.4262385670731707, 'TCA': 0.12251736088908215, 'GCA': 0.21281773299119433, 'AAT': 0.4487737751889762, 'GAA': 0.689467997530356, 'CGT': 0.38120248233619, 'TTT': 0.5737614329268292, 'CTG': 0.4975546385450099, 'GTT': 0.2584847661069405, 'GCC': 0.26990242799092135, 'GAT': 0.626856640686903, 'AGG': 0.020162265266098916, 'CAG': 0.6530527088827319, 'GGG': 0.15057187484254547, 'GTC': 0.21619163129169194, 'CTC': 0.10435163186194824, 'TCT': 0.14544780223293646, 'CCC': 0.12361073915372255, 'CAC': 0.4292446655321377, 'GCT': 0.1609119200081115, 'TTA': 0.13009739763522432, 'CTT': 0.10341378016756284, 'ACG': 0.26849903016796667, 'GGA': 0.10711946389882603, 'CCG': 0.527785216925549, 'CAA': 0.3469472911172681, 'ACT': 0.16532540959927366, 'ATA': 0.07084223130192197, 'CAT': 0.5707553344678623, 'AGT': 0.15102374892893225, 'TGC': 0.55651396076928, 'GAG': 0.31053200246964396, 'TCG': 0.15406749964830613, 'GTA': 0.15372953305033354, 'TAC': 0.4315172054223149, 'TCC': 0.14895195222078703, 'GAC': 0.37314335931309695, 'TGT': 0.4434860392307201, 'TGG': 1.0, 'TAT': 0.5684827945776851, 'CCT': 0.15789033208355652, 'CCA': 0.19071371183717192, 'CGG': 0.09749348521075679, 'CTA': 0.03649285148597391, 'TAA': 0.6391064025794565, 'AGA': 0.036496252317115764, 'TAG': 0.07162597881160755}), 'cai_avg': 0.6151304741138522, 'cai_std': 0.09377868043841883, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}]}, 'orf_debug': {'total_loss': {'C': {'TGT': 0.9895326764891983, 'TGC': 1.0}, 'D': {'GAT': 1.0, 'GAC': 1.148210628685373}, 'S': {'TCA': 0.6266548790591467, 'TCT': 0.8995485327313768, 'AGT': 0.9858597519372291, 'TCG': 1.0977191742866568, 'AGC': 1.1147473073736536, 'TCC': 1.3032529542580562}, 'Q': {'CAA': 0.5979404522050593, 'CAG': 1.1440303221730892}, 'M': {'ATG': 1.0}, 'N': {'AAT': 0.7337673387019819, 'AAC': 1.0}, 'P': {'CCT': 0.6908244153785176, 'CCA': 0.8782483459047534, 'CCG': 1.0, 'CCC': 1.0067004412485698}, 'K': {'AAG': 0.9514739706253069, 'AAA': 1.0}, '_': {'TAG': 0.9641554464808395, 'TAA': 1.0, 'TGA': 1.041167233011862}, 'T': {'ACA': 0.5341889798628014, 'ACG': 0.8804103750608864, 'ACT': 0.9946446749166018, 'ACC': 1.7776424744276667}, 'F': {'TTT': 0.7438697712754998, 'TTC': 1.135678391959799}, 'A': {'GCA': 0.7301222316349993, 'GCT': 0.7528285138815763, 'GCG': 1.049453409682457, 'GCC': 1.0840355774337245}, 'G': {'GGA': 0.6256144561401443, 'GGG': 0.9448488679168409, 'GGC': 1.0, 'GGT': 1.3673072613131283}, 'I': {'ATT': 0.7907885554780182, 'ATA': 0.9630086160665641, 'ATC': 1.1440804169769172}, 'L': {'CTT': 0.5282981167890504, 'TTA': 0.7620783039649957, 'TTG': 0.8566209633014076, 'CTC': 0.90539199331407, 'CTA': 0.9413059309005583, 'CTG': 1.3342503438789546}, 'H': {'CAT': 0.7745204487875498, 'CAC': 1.3201537616694123}, 'R': {'AGA': 0.6535476536963193, 'CGC': 0.776625386996904, 'CGG': 0.885268813545678, 'CGA': 0.920594458184062, 'AGG': 0.958628732572543, 'CGT': 1.0096153846153846}, 'W': {'TGG': 1.0}, 'V': {'GTA': 0.8583999188819134, 'GTC': 0.8691745082432142, 'GTT': 0.9817977528089888, 'GTG': 1.3300796301053173}, 'E': {'GAG': 0.9806714853618527, 'GAA': 1.0}, 'Y': {'TAT': 0.8713136729222519, 'TAC': 1.2166905444126075}}, 'optimized_loss': {'C': defaultdict(<class 'float'>, {'TGT': 1.6090122554987418, 'TGC': 1.0}), 'D': defaultdict(<class 'float'>, {'GAT': 1.0, 'GAC': 1.7487741357394064}), 'S': defaultdict(<class 'float'>, {'TCT': 1.0, 'TCG': 3.051635584744884, 'TCA': 1.1546900193499807, 'TCC': 2.4779507004500814, 'AGC': 1.2426615592968109, 'AGT': 2.830032659527515}), 'Q': defaultdict(<class 'float'>, {'CAA': 1.0, 'CAG': 1.3088053780514624}), 'M': defaultdict(<class 'float'>, {'ATG': 1.0}), 'N': defaultdict(<class 'float'>, {'AAC': 1.0, 'AAT': 1.3943684266530345}), 'P': defaultdict(<class 'float'>, {'CCT': 1.5776475136630435, 'CCG': 1.0, 'CCA': 2.3382446297864634, 'CCC': 3.7288941736028534}), 'K': defaultdict(<class 'float'>, {'AAG': 2.731716735734297, 'AAA': 1.0}), '_': defaultdict(<class 'float'>, {'TAG': 3.5819401803361797, 'TGA': 3.3509093705321793, 'TAA': 1.0}), 'T': defaultdict(<class 'float'>, {'ACC': 3.1600127668893183, 'ACA': 1.0, 'ACG': 2.128216248945088, 'ACT': 2.459228935403716}), 'F': defaultdict(<class 'float'>, {'TTT': 1.0, 'TTC': 1.289765409964395}), 'A': defaultdict(<class 'float'>, {'GCA': 1.0, 'GCC': 2.129837116329259, 'GCG': 1.101352459094135, 'GCT': 1.100806221057107}), 'G': defaultdict(<class 'float'>, {'GGT': 1.9495431577031723, 'GGG': 2.9297346559740687, 'GGA': 1.388670309008689, 'GGC': 1.0}), 'I': defaultdict(<class 'float'>, {'ATC': 1.3089200005100767, 'ATA': 3.6016578405776287, 'ATT': 1.0}), 'L': defaultdict(<class 'float'>, {'TTA': 2.109220171036552, 'TTG': 2.600359572406518, 'CTC': 2.8772310086632475, 'CTT': 1.0, 'CTG': 1.7802239801411086, 'CTA': 3.46359234781819}), 'H': defaultdict(<class 'float'>, {'CAT': 1.0, 'CAC': 1.7428059544498995}), 'R': defaultdict(<class 'float'>, {'CGA': 3.2787385877493485, 'CGC': 1.0, 'CGG': 2.96411287678867, 'CGT': 1.019323224852071, 'AGG': 3.6400623268698062, 'AGA': 1.682377036584766}), 'W': defaultdict(<class 'float'>, {'TGG': 1.0}), 'V': defaultdict(<class 'float'>, {'GTA': 1.6324607736794863, 'GTC': 1.7068049913928647, 'GTG': 1.7691118224210975, 'GTT': 1.0}), 'E': defaultdict(<class 'float'>, {'GAG': 2.5826598723651326, 'GAA': 1.0}), 'Y': defaultdict(<class 'float'>, {'TAT': 1.0, 'TAC': 1.4803358808630471})}, 'deoptimized_loss': {'C': defaultdict(<class 'float'>, {'TGT': 1.6432327117286945, 'TGC': 1.0}), 'D': defaultdict(<class 'float'>, {'GAT': 1.0, 'GAC': 1.32644912035017}), 'S': defaultdict(<class 'float'>, {'TCT': 1.2358074271617692, 'TCG': 2.5325041749163573, 'TCA': 2.940414537227555, 'TCC': 1.4589325892912588, 'AGC': 1.0, 'AGT': 2.911797534354834}), 'Q': defaultdict(<class 'float'>, {'CAA': 2.7969463044492056, 'CAG': 1.0}), 'M': defaultdict(<class 'float'>, {'ATG': 1.0}), 'N': defaultdict(<class 'float'>, {'AAC': 1.0, 'AAT': 2.5897675631491137}), 'P': defaultdict(<class 'float'>, {'CCT': 3.3057851239669427, 'CCG': 1.0, 'CCA': 3.0314838894201217, 'CCC': 3.679421487603306}), 'K': defaultdict(<class 'float'>, {'AAG': 3.0174622091692234, 'AAA': 1.0}), '_': defaultdict(<class 'float'>, {'TAG': 3.853223593964335, 'TGA': 3.0911615190738604, 'TAA': 1.0}), 'T': defaultdict(<class 'float'>, {'ACC': 1.0, 'ACA': 3.504371592388779, 'ACG': 2.7456517456510596, 'ACT': 2.4857819856416143}), 'F': defaultdict(<class 'float'>, {'TTT': 1.8071998373247593, 'TTC': 1.0}), 'A': defaultdict(<class 'float'>, {'GCA': 1.8758964212769875, 'GCC': 1.812421976933375, 'GCG': 1.0, 'GCT': 1.9423109359102864}), 'G': defaultdict(<class 'float'>, {'GGT': 1.0427990176022806, 'GGG': 3.2817356454694036, 'GGA': 3.548016240581601, 'GGC': 1.0}), 'I': defaultdict(<class 'float'>, {'ATC': 1.0, 'ATA': 3.883668089603532, 'ATT': 1.599113350558424}), 'L': defaultdict(<class 'float'>, {'TTA': 3.631806659913921, 'TTG': 3.5436923326098997, 'CTC': 3.509954066352956, 'CTT': 3.5829592996359754, 'CTG': 1.0, 'CTA': 3.908995597888406}), 'H': defaultdict(<class 'float'>, {'CAT': 1.6669947156956941, 'CAC': 1.0}), 'R': defaultdict(<class 'float'>, {'CGA': 3.868744772855011, 'CGC': 1.6579706335513271, 'CGG': 3.782199282654244, 'CGT': 1.0, 'AGG': 3.9610282186271286, 'AGA': 3.9388442858165034}), 'W': defaultdict(<class 'float'>, {'TGG': 1.0}), 'V': defaultdict(<class 'float'>, {'GTA': 2.2154574764955095, 'GTC': 2.259279403605198, 'GTG': 1.0, 'GTT': 1.0374231441153496}), 'E': defaultdict(<class 'float'>, {'GAG': 2.6854688521243797, 'GAA': 1.0}), 'Y': defaultdict(<class 'float'>, {'TAT': 1.3171976331360948, 'TAC': 1.0})}}, 'orf': {'orf_module_input_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'optimized_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAGGTACATATGGAGGGATCAGTAAATGGACATGAGTTTGAGATTGAGGGAGAGGGAGAGGGAAGACCTTATGAGGGAACACAAACAGCAAAGCTTAAGGTAACAAAGGGAGGGCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAGGCATATGTAAAGCATCCTGCAGATATTCCTGATTATCTTAAGCTTTCATTTCCTGAGGGATTTAAGTGGGAGAGAGTAATGAATTTTGAGGATGGAGGGGTAGTCACAGTAACACAAGATTCATCTCTTCAAGATGGAGAGTTTATTTATAAGGTAAAGCTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTAATGCAAAAGAAAACAATGGGATGGGAGGCATCATCTGAGAGAATGTATCCTGAGGATGGAGCACTTAAGGGAGAGATTAAGCAAAGACTTAAGCTTAAGGATGGAGGGCATTATGATGCAGAGGTAAAGACAACGTATAAGGCAAAGAAACCTGTACAACTTCCTGGAGCATATAATGTAAATATTAAGCTTGATATTACATCACATAATGAGGATTATACAATTGTAGAGCAATATGAGAGAGCAGAGGGAAGACATTCAACAGGAGGGATGGATGAGCTTTATAAGTAG', 'aa_to_loss_mapping': {'C': {'TGT': 0.9895326764891983, 'TGC': 1.0}, 'D': {'GAT': 1.0, 'GAC': 1.148210628685373}, 'S': {'TCA': 0.6266548790591467, 'TCT': 0.8995485327313768, 'AGT': 0.9858597519372291, 'TCG': 1.0977191742866568, 'AGC': 1.1147473073736536, 'TCC': 1.3032529542580562}, 'Q': {'CAA': 0.5979404522050593, 'CAG': 1.1440303221730892}, 'M': {'ATG': 1.0}, 'N': {'AAT': 0.7337673387019819, 'AAC': 1.0}, 'P': {'CCT': 0.6908244153785176, 'CCA': 0.8782483459047534, 'CCG': 1.0, 'CCC': 1.0067004412485698}, 'K': {'AAG': 0.9514739706253069, 'AAA': 1.0}, '_': {'TAG': 0.9641554464808395, 'TAA': 1.0, 'TGA': 1.041167233011862}, 'T': {'ACA': 0.5341889798628014, 'ACG': 0.8804103750608864, 'ACT': 0.9946446749166018, 'ACC': 1.7776424744276667}, 'F': {'TTT': 0.7438697712754998, 'TTC': 1.135678391959799}, 'A': {'GCA': 0.7301222316349993, 'GCT': 0.7528285138815763, 'GCG': 1.049453409682457, 'GCC': 1.0840355774337245}, 'G': {'GGA': 0.6256144561401443, 'GGG': 0.9448488679168409, 'GGC': 1.0, 'GGT': 1.3673072613131283}, 'I': {'ATT': 0.7907885554780182, 'ATA': 0.9630086160665641, 'ATC': 1.1440804169769172}, 'L': {'CTT': 0.5282981167890504, 'TTA': 0.7620783039649957, 'TTG': 0.8566209633014076, 'CTC': 0.90539199331407, 'CTA': 0.9413059309005583, 'CTG': 1.3342503438789546}, 'H': {'CAT': 0.7745204487875498, 'CAC': 1.3201537616694123}, 'R': {'AGA': 0.6535476536963193, 'CGC': 0.776625386996904, 'CGG': 0.885268813545678, 'CGA': 0.920594458184062, 'AGG': 0.958628732572543, 'CGT': 1.0096153846153846}, 'W': {'TGG': 1.0}, 'V': {'GTA': 0.8583999188819134, 'GTC': 0.8691745082432142, 'GTT': 0.9817977528089888, 'GTG': 1.3300796301053173}, 'E': {'GAG': 0.9806714853618527, 'GAA': 1.0}, 'Y': {'TAT': 0.8713136729222519, 'TAC': 1.2166905444126075}}, 'run_time': 0.9059999999590218}, 'evaluation': [{'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'cai_initial_score': 0.7695788606254016, 'cai_final_score': 0.7335478217631582, 'dist_score': -0.8418073204995801}, {'name': 'Escherichia coli', 'is_wanted': False, 'cai_initial_score': 0.7915994634162647, 'cai_final_score': 0.29377395917401045, 'dist_score': -5.308514706273339}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAGGTACATATGGAGGGATCAGTAAATGGACATGAGTTTGAGATTGAGGGAGAGGGAGAGGGAAGACCTTATGAGGGAACACAAACAGCAAAGCTTAAGGTAACAAAGGGAGGGCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAGGCATATGTAAAGCATCCTGCAGATATTCCTGATTATCTTAAGCTTTCATTTCCTGAGGGATTTAAGTGGGAGAGAGTAATGAATTTTGAGGATGGAGGGGTAGTCACAGTAACACAAGATTCATCTCTTCAAGATGGAGAGTTTATTTATAAGGTAAAGCTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTAATGCAAAAGAAAACAATGGGATGGGAGGCATCATCTGAGAGAATGTATCCTGAGGATGGAGCACTTAAGGGAGAGATTAAGCAAAGACTTAAGCTTAAGGATGGAGGGCATTATGATGCAGAGGTAAAGACAACGTATAAGGCAAAGAAACCTGTACAACTTCCTGGAGCATATAATGTAAATATTAAGCTTGATATTACATCACATAATGAGGATTATACAATTGTAGAGCAATATGAGAGAGCAGAGGGAAGACATTCAACAGGAGGGATGGATGAGCTTTATAAGTAG', 'average_distance_score': 2.2333536928868796, 'weakest_link_score': 2.2333536928868796, 'ratio_score': 1.1047359569220119, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTTATGAGATTTAAGGTACATATGGAGGGATCAGTAAATGGACATGAGTTTGAGATTGAGGGAGAGGGAGAGGGAAGACCTTATGAGGGAACACAAACAGCAAAGCTTAAGGTAACAAAGGGAGGGCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAGGCATATGTAAAGCATCCTGCAGATATTCCTGATTATCTTAAGCTTTCATTTCCTGAGGGATTTAAGTGGGAGAGAGTAATGAATTTTGAGGATGGAGGGGTAGTCACAGTAACACAAGATTCATCTCTTCAAGATGGAGAGTTTATTTATAAGGTAAAGCTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTAATGCAAAAGAAAACAATGGGATGGGAGGCATCATCTGAGAGAATGTATCCTGAGGATGGAGCACTTAAGGGAGAGATTAAGCAAAGACTTAAGCTTAAGGATGGAGGGCATTATGATGCAGAGGTAAAGACAACGTATAAGGCAAAGAAACCTGTACAACTTCCTGGAGCATATAATGTAAATATTAAGCTTGATATTACATCACATAATGAGGATTATACAATTGTAGAGCAATATGAGAGAGCAGAGGGAAGACATTCAACAGGAGGGATGGATGAGCTTTATAAGTAG'}], 'final_evaluation': {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAGGTACATATGGAGGGATCAGTAAATGGACATGAGTTTGAGATTGAGGGAGAGGGAGAGGGAAGACCTTATGAGGGAACACAAACAGCAAAGCTTAAGGTAACAAAGGGAGGGCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAGGCATATGTAAAGCATCCTGCAGATATTCCTGATTATCTTAAGCTTTCATTTCCTGAGGGATTTAAGTGGGAGAGAGTAATGAATTTTGAGGATGGAGGGGTAGTCACAGTAACACAAGATTCATCTCTTCAAGATGGAGAGTTTATTTATAAGGTAAAGCTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTAATGCAAAAGAAAACAATGGGATGGGAGGCATCATCTGAGAGAATGTATCCTGAGGATGGAGCACTTAAGGGAGAGATTAAGCAAAGACTTAAGCTTAAGGATGGAGGGCATTATGATGCAGAGGTAAAGACAACGTATAAGGCAAAGAAACCTGTACAACTTCCTGGAGCATATAATGTAAATATTAAGCTTGATATTACATCACATAATGAGGATTATACAATTGTAGAGCAATATGAGAGAGCAGAGGGAAGACATTCAACAGGAGGGATGGATGAGCTTTATAAGTAG', 'average_distance_score': 2.2333536928868796, 'weakest_link_score': 2.2333536928868796, 'ratio_score': 1.1047359569220119}, 'zip_output_file_path': 'results\\\\mcherry-wanted-constraint-init-external-0.1\\\\CAI_single_codon_ratio_ecoli_opt_False_6rNS\\\\communique_results.zip'}\n",
      "\n",
      "##########################\n",
      "# User Input #\n",
      "##########################\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is:\n",
      "ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "------------------------------------------\n",
      "Parsing information for Escherichia coli:\n",
      "------------------------------------------\n",
      "Organism is defined as wanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Escherichia coli. Calculating tAI profile.\n",
      "tai_std=0.024913119712427708, tai_avg=0.24958213077316901\n",
      "------------------------------------------\n",
      "Parsing information for Bacillus subtilis:\n",
      "------------------------------------------\n",
      "Organism is defined as unwanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Bacillus subtilis. Calculating tAI profile.\n",
      "tai_std=0.028076995641981673, tai_avg=0.3271366648197227\n",
      "-----------------------------\n",
      "Normalized Prioritization Weights\n",
      "-----------------------------\n",
      "Escherichia coli : 1.0\n",
      "Bacillus subtilis : 1.0\n",
      "Total input processing time: 26.372096061706543\n",
      "\n",
      "##########################\n",
      "# Initiation #\n",
      "##########################\n",
      "Running initiation optimization InitiationOptimizationMethod.external_module on 15 codons at the start of the ORF\n",
      "Taking optimized sequence from configuration value config.INITIATION.EXTERNAL_INITIATION_ORF\n",
      "Optimized sequence is: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "Optimization method used is: ORFOptimizationMethod.single_codon_ratio\n",
      "Loss summary for: TGT\n",
      "loss for Escherichia coli: 1.1051265625\n",
      "loss for Bacillus subtilis: 1.141336111111111\n",
      "Loss summary for: TGC\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGT': 0.984009360374415, 'TGC': 1.0})\n",
      "Loss summary for: GAT\n",
      "loss for Escherichia coli: 1.3311390625000001\n",
      "loss for Bacillus subtilis: 1.6213777777777774\n",
      "Loss summary for: GAC\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAT': 0.906086387434555, 'GAC': 1.0})\n",
      "Loss summary for: TCT\n",
      "loss for Escherichia coli: 1.21550625\n",
      "loss for Bacillus subtilis: 1.5252249999999996\n",
      "Loss summary for: TCG\n",
      "loss for Escherichia coli: 1.177225\n",
      "loss for Bacillus subtilis: 1.5047111111111109\n",
      "Loss summary for: TCA\n",
      "loss for Escherichia coli: 1.2348765625000002\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: TCC\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.3611111111111114\n",
      "Loss summary for: AGC\n",
      "loss for Escherichia coli: 1.265625\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: AGT\n",
      "loss for Escherichia coli: 1.3835640625\n",
      "loss for Bacillus subtilis: 1.2920111111111112\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TCT': 0.8927125506072876, 'TCG': 0.8845108695652174, 'TCA': 1.11125, 'TCC': 0.8571428571428571, 'AGC': 1.125, 'AGT': 1.0348240469208212})\n",
      "Loss summary for: CAA\n",
      "loss for Escherichia coli: 1.10775625\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: CAG\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 2.1121777777777777\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAA': 1.0525, 'CAG': 0.6880733944954128})\n",
      "Loss summary for: ATG\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATG': 1.0})\n",
      "Loss summary for: AAC\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: AAT\n",
      "loss for Escherichia coli: 1.4520250000000001\n",
      "loss for Bacillus subtilis: 1.6213777777777774\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAC': 1.0, 'AAT': 0.9463350785340315})\n",
      "Loss summary for: CCT\n",
      "loss for Escherichia coli: 1.1908265625\n",
      "loss for Bacillus subtilis: 2.25\n",
      "Loss summary for: CCG\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.7956000000000003\n",
      "Loss summary for: CCA\n",
      "loss for Escherichia coli: 1.0531890625000002\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: CCC\n",
      "loss for Escherichia coli: 1.0816000000000001\n",
      "loss for Bacillus subtilis: 2.25\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CCT': 0.7275, 'CCG': 0.7462686567164178, 'CCA': 1.02625, 'CCC': 0.6933333333333334})\n",
      "Loss summary for: AAG\n",
      "loss for Escherichia coli: 2.2801\n",
      "loss for Bacillus subtilis: 2.1121777777777777\n",
      "Loss summary for: AAA\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAG': 1.0389908256880733, 'AAA': 1.0})\n",
      "Loss summary for: TAG\n",
      "loss for Escherichia coli: 1.5462535182753292\n",
      "loss for Bacillus subtilis: 1.7989356493995978\n",
      "Loss summary for: TGA\n",
      "loss for Escherichia coli: 1.5462535182753292\n",
      "loss for Bacillus subtilis: 1.7989356493995978\n",
      "Loss summary for: TAA\n",
      "loss for Escherichia coli: 1.5462535182753292\n",
      "loss for Bacillus subtilis: 1.7989356493995978\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAG': 0.9271127128057303, 'TGA': 0.9271127128057303, 'TAA': 0.9271127128057303})\n",
      "Loss summary for: ACC\n",
      "loss for Escherichia coli: 1.0816000000000001\n",
      "loss for Bacillus subtilis: 2.25\n",
      "Loss summary for: ACA\n",
      "loss for Escherichia coli: 1.2939062499999998\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: ACG\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 2.1121777777777777\n",
      "Loss summary for: ACT\n",
      "loss for Escherichia coli: 1.3053062500000001\n",
      "loss for Bacillus subtilis: 2.4596694444444447\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ACC': 0.6933333333333334, 'ACA': 1.1375, 'ACG': 0.6880733944954128, 'ACT': 0.7284803400637619})\n",
      "Loss summary for: TTT\n",
      "loss for Escherichia coli: 1.21550625\n",
      "loss for Bacillus subtilis: 1.4520250000000001\n",
      "Loss summary for: TTC\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTT': 0.9149377593360996, 'TTC': 1.0})\n",
      "Loss summary for: GCA\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: GCC\n",
      "loss for Escherichia coli: 1.265625\n",
      "loss for Bacillus subtilis: 2.777777777777778\n",
      "Loss summary for: GCG\n",
      "loss for Escherichia coli: 1.5750249999999997\n",
      "loss for Bacillus subtilis: 2.454444444444445\n",
      "Loss summary for: GCT\n",
      "loss for Escherichia coli: 1.50675625\n",
      "loss for Bacillus subtilis: 3.0102249999999997\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GCA': 1.0, 'GCC': 0.6749999999999999, 'GCG': 0.8010638297872339, 'GCT': 0.7074927953890491})\n",
      "Loss summary for: GGT\n",
      "loss for Escherichia coli: 1.4520250000000001\n",
      "loss for Bacillus subtilis: 1.6213777777777774\n",
      "Loss summary for: GGG\n",
      "loss for Escherichia coli: 1.782225\n",
      "loss for Bacillus subtilis: 2.2700444444444443\n",
      "Loss summary for: GGA\n",
      "loss for Escherichia coli: 1.8530015625000003\n",
      "loss for Bacillus subtilis: 1.3611111111111107\n",
      "Loss summary for: GGC\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GGT': 0.9463350785340315, 'GGG': 0.8860619469026548, 'GGA': 1.1667857142857145, 'GGC': 1.0})\n",
      "Loss summary for: ATC\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: ATA\n",
      "loss for Escherichia coli: 1.6002250000000002\n",
      "loss for Bacillus subtilis: 1.9321000000000004\n",
      "Loss summary for: ATT\n",
      "loss for Escherichia coli: 1.3311390625000001\n",
      "loss for Bacillus subtilis: 1.4520250000000001\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATC': 1.0, 'ATA': 0.9100719424460432, 'ATT': 0.9574688796680498})\n",
      "Loss summary for: TTA\n",
      "loss for Escherichia coli: 1.9635015625000003\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: TTG\n",
      "loss for Escherichia coli: 1.890625\n",
      "loss for Bacillus subtilis: 1.4200694444444444\n",
      "Loss summary for: CTC\n",
      "loss for Escherichia coli: 2.002225\n",
      "loss for Bacillus subtilis: 1.8270027777777775\n",
      "Loss summary for: CTT\n",
      "loss for Escherichia coli: 2.1498890625000002\n",
      "loss for Bacillus subtilis: 2.0164\n",
      "Loss summary for: CTG\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.5500250000000002\n",
      "Loss summary for: CTA\n",
      "loss for Escherichia coli: 1.8496000000000004\n",
      "loss for Bacillus subtilis: 1.3611111111111114\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTA': 1.40125, 'TTG': 1.153846153846154, 'CTC': 1.0468557336621456, 'CTT': 1.0325704225352113, 'CTG': 0.8032128514056224, 'CTA': 1.1657142857142857})\n",
      "Loss summary for: CAT\n",
      "loss for Escherichia coli: 1.1051265625\n",
      "loss for Bacillus subtilis: 1.2920111111111112\n",
      "Loss summary for: CAC\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAT': 0.9248533724340176, 'CAC': 1.0})\n",
      "Loss summary for: CGA\n",
      "loss for Escherichia coli: 2.20879044\n",
      "loss for Bacillus subtilis: 2.7167830044444448\n",
      "Loss summary for: CGC\n",
      "loss for Escherichia coli: 1.2996000000000003\n",
      "loss for Bacillus subtilis: 1.4081777777777775\n",
      "Loss summary for: CGG\n",
      "loss for Escherichia coli: 1.890625\n",
      "loss for Bacillus subtilis: 2.25\n",
      "Loss summary for: CGT\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: AGG\n",
      "loss for Escherichia coli: 1.782225\n",
      "loss for Bacillus subtilis: 2.0928444444444447\n",
      "Loss summary for: AGA\n",
      "loss for Escherichia coli: 1.8530015625000003\n",
      "loss for Bacillus subtilis: 2.195336111111111\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CGA': 0.9016744863290729, 'CGC': 0.960674157303371, 'CGG': 0.9166666666666666, 'CGT': 1.0, 'AGG': 0.922811059907834, 'AGA': 0.9187289088863893})\n",
      "Loss summary for: TGG\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGG': 1.0})\n",
      "Loss summary for: GTA\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: GTC\n",
      "loss for Escherichia coli: 1.890625\n",
      "loss for Bacillus subtilis: 2.25\n",
      "Loss summary for: GTG\n",
      "loss for Escherichia coli: 2.030625\n",
      "loss for Bacillus subtilis: 2.1121777777777777\n",
      "Loss summary for: GTT\n",
      "loss for Escherichia coli: 2.18300625\n",
      "loss for Bacillus subtilis: 2.4596694444444447\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GTA': 1.0, 'GTC': 0.9166666666666666, 'GTG': 0.9805045871559633, 'GTT': 0.9420828905419766})\n",
      "Loss summary for: GAG\n",
      "loss for Escherichia coli: 1.7956000000000003\n",
      "loss for Bacillus subtilis: 2.8224000000000005\n",
      "Loss summary for: GAA\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAG': 0.7976190476190476, 'GAA': 1.0})\n",
      "Loss summary for: TAT\n",
      "loss for Escherichia coli: 1.3311390625000001\n",
      "loss for Bacillus subtilis: 1.2920111111111112\n",
      "Loss summary for: TAC\n",
      "loss for Escherichia coli: 1.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAT': 1.0150293255131966, 'TAC': 1.0})\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5737614329268292]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.06388523224887838]\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.09749348521075679]\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.036496252317115764]\n",
      "Skipping codon AGA due to very low average frequency 0.036496252317115764 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.020162265266098916]\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.40076028262096014]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5737614329268292]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5707553344678623]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4487737751889762]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5707553344678623]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5737614329268292]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.07084223130192197]\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5085852538544149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.06388523224887838]\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.09749348521075679]\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.036496252317115764]\n",
      "Skipping codon AGA due to very low average frequency 0.036496252317115764 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.020162265266098916]\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.40076028262096014]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26849903016796667]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26849903016796667]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26849903016796667]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5737614329268292]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.626856640686903]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.07084223130192197]\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5085852538544149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5737614329268292]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5707553344678623]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.626856640686903]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.07084223130192197]\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5085852538544149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.626856640686903]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5737614329268292]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5737614329268292]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.06388523224887838]\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.09749348521075679]\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.036496252317115764]\n",
      "Skipping codon AGA due to very low average frequency 0.036496252317115764 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.020162265266098916]\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.40076028262096014]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4487737751889762]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5737614329268292]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.626856640686903]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2584847661069405]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26849903016796667]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26849903016796667]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.626856640686903]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15406749964830613]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.626856640686903]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5737614329268292]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.07084223130192197]\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5085852538544149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.06388523224887838]\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.09749348521075679]\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.036496252317115764]\n",
      "Skipping codon AGA due to very low average frequency 0.036496252317115764 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.020162265266098916]\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.40076028262096014]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26849903016796667]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4487737751889762]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5737614329268292]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.626856640686903]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26849903016796667]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15406749964830613]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.06388523224887838]\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.09749348521075679]\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.036496252317115764]\n",
      "Skipping codon AGA due to very low average frequency 0.036496252317115764 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.020162265266098916]\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.40076028262096014]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.626856640686903]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.07084223130192197]\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5085852538544149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.06388523224887838]\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.09749348521075679]\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.036496252317115764]\n",
      "Skipping codon AGA due to very low average frequency 0.036496252317115764 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.020162265266098916]\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.40076028262096014]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.626856640686903]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5707553344678623]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.626856640686903]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26849903016796667]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26849903016796667]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4487737751889762]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4487737751889762]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.07084223130192197]\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5085852538544149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.626856640686903]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.07084223130192197]\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5085852538544149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26849903016796667]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5707553344678623]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4487737751889762]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.626856640686903]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26849903016796667]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.07084223130192197]\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5085852538544149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.06388523224887838]\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.09749348521075679]\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.036496252317115764]\n",
      "Skipping codon AGA due to very low average frequency 0.036496252317115764 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.020162265266098916]\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.40076028262096014]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.06388523224887838]\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.09749348521075679]\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.036496252317115764]\n",
      "Skipping codon AGA due to very low average frequency 0.036496252317115764 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.020162265266098916]\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.40076028262096014]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5707553344678623]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26849903016796667]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.626856640686903]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.07162597881160755]\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.28926761860893596]\n",
      "\n",
      "##########################\n",
      "# EVALUATION #\n",
      "##########################\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTTATGCGCTTTAAAGTCCATATGGAGGGGTCCGTCAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCCTACGAGGGGACGCAGACGGCCAAACTGAAAGTCACGAAAGGGGGTCCCCTGCCCTTTGCCTGGGATATTCTGTCCCCCCAGTTTATGTACGGGTCCAAAGCCTACGTCAAACATCCCGCCGATATTCCCGATTACCTGAAACTGTCCTTTCCCGAGGGGTTTAAATGGGAGCGCGTCATGAATTTTGAGGATGGGGGTGTCGTTACGGTCACGCAGGATTCCTCGCTGCAGGATGGGGAGTTTATTTACAAAGTCAAACTGCGCGGGACGAATTTTCCCTCCGATGGGCCCGTCATGCAGAAAAAGACGATGGGGTGGGAGGCCTCCTCGGAGCGCATGTACCCCGAGGATGGGGCCCTGAAAGGGGAGATTAAACAGCGCCTGAAACTGAAAGATGGGGGTCATTACGATGCCGAGGTCAAAACGACCTACAAAGCCAAAAAGCCCGTCCAGCTGCCCGGGGCCTACAATGTCAATATTAAACTGGATATTACGTCCCATAATGAGGATTACACGATTGTCGAGCAGTACGAGCGCGCCGAGGGGCGCCATTCCACGGGGGGTATGGATGAGCTGTACAAATAA\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGCTTTAAAGTCCATATGGAGGGGTCCGTCAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCCTACGAGGGGACGCAGACGGCCAAACTGAAAGTCACGAAAGGGGGTCCCCTGCCCTTTGCCTGGGATATTCTGTCCCCCCAGTTTATGTACGGGTCCAAAGCCTACGTCAAACATCCCGCCGATATTCCCGATTACCTGAAACTGTCCTTTCCCGAGGGGTTTAAATGGGAGCGCGTCATGAATTTTGAGGATGGGGGTGTCGTTACGGTCACGCAGGATTCCTCGCTGCAGGATGGGGAGTTTATTTACAAAGTCAAACTGCGCGGGACGAATTTTCCCTCCGATGGGCCCGTCATGCAGAAAAAGACGATGGGGTGGGAGGCCTCCTCGGAGCGCATGTACCCCGAGGATGGGGCCCTGAAAGGGGAGATTAAACAGCGCCTGAAACTGAAAGATGGGGGTCATTACGATGCCGAGGTCAAAACGACCTACAAAGCCAAAAAGCCCGTCCAGCTGCCCGGGGCCTACAATGTCAATATTAAACTGGATATTACGTCCCATAATGAGGATTACACGATTGTCGAGCAGTACGAGCGCGCCGAGGGGCGCCATTCCACGGGGGGTATGGATGAGCTGTACAAATAA', 'average_distance_score': 0.47914818504126744, 'weakest_link_score': 0.47914818504126744, 'ratio_score': 1.023878204300951}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGCTTTAAAGTCCATATGGAGGGGTCCGTCAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCCTACGAGGGGACGCAGACGGCCAAACTGAAAGTCACGAAAGGGGGTCCCCTGCCCTTTGCCTGGGATATTCTGTCCCCCCAGTTTATGTACGGGTCCAAAGCCTACGTCAAACATCCCGCCGATATTCCCGATTACCTGAAACTGTCCTTTCCCGAGGGGTTTAAATGGGAGCGCGTCATGAATTTTGAGGATGGGGGTGTCGTTACGGTCACGCAGGATTCCTCGCTGCAGGATGGGGAGTTTATTTACAAAGTCAAACTGCGCGGGACGAATTTTCCCTCCGATGGGCCCGTCATGCAGAAAAAGACGATGGGGTGGGAGGCCTCCTCGGAGCGCATGTACCCCGAGGATGGGGCCCTGAAAGGGGAGATTAAACAGCGCCTGAAACTGAAAGATGGGGGTCATTACGATGCCGAGGTCAAAACGACCTACAAAGCCAAAAAGCCCGTCCAGCTGCCCGGGGCCTACAATGTCAATATTAAACTGGATATTACGTCCCATAATGAGGATTACACGATTGTCGAGCAGTACGAGCGCGCCGAGGGGCGCCATTCCACGGGGGGTATGGATGAGCTGTACAAATAA', 'average_distance_score': 0.47914818504126744, 'weakest_link_score': 0.47914818504126744, 'ratio_score': 1.023878204300951}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry-wanted-constraint-init-external-0.1\\tAI_single_codon_ratio_ecoli_opt_True_6ezh\n",
      "Final output: {'module_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'orf_optimization_method': 'single_codon_ratio', 'orf_optimization_cub_index': 'tAI', 'initiation_optimization_method': 'external', 'evaluation_score': 'average_distance', 'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': None, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'AAA': 0.7667390608953314, 'CGC': 0.40076028262096014, 'ATT': 0.5085852538544149, 'AGC': 0.27799163607995603, 'ACC': 0.4359429381095841, 'ATC': 0.4205725148436632, 'ACA': 0.13023262212317555, 'GGT': 0.33726003930064996, 'AAC': 0.5512262248110238, 'GCG': 0.3563679190097728, 'GGC': 0.4050486219579785, 'TGA': 0.28926761860893596, 'CGA': 0.06388523224887838, 'GTG': 0.3715940695510341, 'TTG': 0.12808970030428077, 'AAG': 0.23326093910466858, 'TTC': 0.4262385670731707, 'TCA': 0.12251736088908215, 'GCA': 0.21281773299119433, 'AAT': 0.4487737751889762, 'GAA': 0.689467997530356, 'CGT': 0.38120248233619, 'TTT': 0.5737614329268292, 'CTG': 0.4975546385450099, 'GTT': 0.2584847661069405, 'GCC': 0.26990242799092135, 'GAT': 0.626856640686903, 'AGG': 0.020162265266098916, 'CAG': 0.6530527088827319, 'GGG': 0.15057187484254547, 'GTC': 0.21619163129169194, 'CTC': 0.10435163186194824, 'TCT': 0.14544780223293646, 'CCC': 0.12361073915372255, 'CAC': 0.4292446655321377, 'GCT': 0.1609119200081115, 'TTA': 0.13009739763522432, 'CTT': 0.10341378016756284, 'ACG': 0.26849903016796667, 'GGA': 0.10711946389882603, 'CCG': 0.527785216925549, 'CAA': 0.3469472911172681, 'ACT': 0.16532540959927366, 'ATA': 0.07084223130192197, 'CAT': 0.5707553344678623, 'AGT': 0.15102374892893225, 'TGC': 0.55651396076928, 'GAG': 0.31053200246964396, 'TCG': 0.15406749964830613, 'GTA': 0.15372953305033354, 'TAC': 0.4315172054223149, 'TCC': 0.14895195222078703, 'GAC': 0.37314335931309695, 'TGT': 0.4434860392307201, 'TGG': 1.0, 'TAT': 0.5684827945776851, 'CCT': 0.15789033208355652, 'CCA': 0.19071371183717192, 'CGG': 0.09749348521075679, 'CTA': 0.03649285148597391, 'TAA': 0.6391064025794565, 'AGA': 0.036496252317115764, 'TAG': 0.07162597881160755}), 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.75, 'AAC': 0.5, 'AAG': 0.23999999999999996, 'AAT': 0.29500000000000004, 'ACA': 0.1525, 'ACC': 0.25, 'ACG': 0.29, 'ACT': 0.14750000000000002, 'AGA': 0.13874999999999998, 'AGC': 0.125, 'AGG': 0.16499999999999998, 'AGT': 0.07375000000000001, 'ATA': 0.10999999999999999, 'ATC': 0.375, 'ATG': 1.0, 'ATT': 0.22125000000000003, 'CAA': 0.27749999999999997, 'CAC': 0.125, 'CAG': 0.32999999999999996, 'CAT': 0.07375000000000001, 'CCA': 0.13874999999999998, 'CCC': 0.125, 'CCG': 0.16499999999999998, 'CCT': 0.07375000000000001, 'CGA': 0.013799999999999993, 'CGC': 0.36, 'CGG': 0.125, 'CGT': 0.5, 'CTA': 0.18, 'CTC': 0.125, 'CTG': 0.54, 'CTT': 0.07375000000000001, 'GAA': 0.5, 'GAC': 0.375, 'GAG': 0.15999999999999998, 'GAT': 0.22125000000000003, 'GCA': 0.375, 'GCC': 0.25, 'GCG': 0.11999999999999998, 'GCT': 0.14750000000000002, 'GGA': 0.13874999999999998, 'GGC': 0.5, 'GGG': 0.16499999999999998, 'GGT': 0.29500000000000004, 'GTA': 0.625, 'GTC': 0.25, 'GTG': 0.19999999999999996, 'GTT': 0.14750000000000002, 'TAC': 0.375, 'TAT': 0.22125000000000003, 'TCA': 0.13874999999999998, 'TCC': 0.25, 'TCG': 0.16499999999999998, 'TCT': 0.14750000000000002, 'TGC': 0.125, 'TGG': 0.125, 'TGT': 0.07375000000000001, 'TTA': 0.13874999999999998, 'TTC': 0.25, 'TTG': 0.16499999999999998, 'TTT': 0.14750000000000002}, 'tai_avg': 0.24958213077316901, 'tai_std': 0.024913119712427708}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': None, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'GAA': 0.6798186496752849, 'AAT': 0.5655317497600131, 'ATA': 0.1287010322849589, 'TTA': 0.198530453110291, 'GAC': 0.35960560391899904, 'CTG': 0.23973441522197322, 'TGG': 1.0, 'AAC': 0.4344682502399869, 'CAA': 0.5131548459124133, 'GCC': 0.20715414603308396, 'CTT': 0.2391679232236727, 'GCT': 0.24752923822568748, 'ATC': 0.366762837741479, 'AAA': 0.7016557540022625, 'AAG': 0.29834424599773746, 'TTG': 0.1592175745811708, 'AGC': 0.22534685577492342, 'CCG': 0.4334091260292722, 'AGT': 0.10611598754150686, 'TTT': 0.6847751982489505, 'GAG': 0.3201813503247151, 'ACT': 0.1614000537105004, 'TCA': 0.23653118484388272, 'ACC': 0.1588786441082565, 'CAC': 0.3265866087018494, 'GGC': 0.3390368073029434, 'GAT': 0.640394396081001, 'ACA': 0.4122430101751559, 'ACG': 0.2674782920060872, 'CCC': 0.0889644363010221, 'AGA': 0.26380392311639106, 'TCC': 0.12712038919920718, 'TAC': 0.34586274736720285, 'CAT': 0.6734133912981506, 'ATT': 0.5045361299735621, 'GCA': 0.2828363712991255, 'TAT': 0.6541372526327971, 'GGG': 0.16211598104043537, 'GTC': 0.25598186049286953, 'CCT': 0.28667299498885185, 'GTT': 0.28439644370189154, 'GCG': 0.26248024444210305, 'TCT': 0.2048572678833432, 'CTC': 0.11219040795754642, 'CCA': 0.19095344268085387, 'GGA': 0.3148341038094681, 'CGA': 0.09855152543710778, 'GTA': 0.19774449549495793, 'GTG': 0.26187720031028106, 'TTC': 0.31522480175104955, 'CGC': 0.2064569601768165, 'CAG': 0.4868451540875866, 'CGG': 0.1555630106168844, 'CGT': 0.18161187196589967, 'CTA': 0.05115922590534585, 'AGG': 0.09401270868690058, 'GGT': 0.18401310784715313, 'TAG': 0.14132712456344587, 'TGC': 0.5460141582025239, 'TCG': 0.10002831475713661, 'TAA': 0.630500582072177, 'TGA': 0.2281722933643772, 'TGT': 0.45398584179747614}), 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.6666666666666666, 'AAC': 0.6666666666666666, 'AAG': 0.2133333333333333, 'AAT': 0.39333333333333337, 'ACA': 0.6666666666666666, 'ACC': 0.16666666666666666, 'ACG': 0.2133333333333333, 'ACT': 0.09833333333333334, 'AGA': 0.18499999999999997, 'AGC': 0.3333333333333333, 'AGG': 0.21999999999999997, 'AGT': 0.19666666666666668, 'ATA': 0.10999999999999999, 'ATC': 0.5, 'ATG': 1.0, 'ATT': 0.29500000000000004, 'CAA': 0.6666666666666666, 'CAC': 0.3333333333333333, 'CAG': 0.2133333333333333, 'CAT': 0.19666666666666668, 'CCA': 0.5, 'CCG': 0.15999999999999998, 'CGA': 0.01839999999999999, 'CGC': 0.48, 'CGG': 0.16666666666666666, 'CGT': 0.6666666666666666, 'CTA': 0.3516666666666666, 'CTC': 0.16666666666666666, 'CTG': 0.2733333333333333, 'CTT': 0.09833333333333334, 'GAA': 1.0, 'GAC': 0.6666666666666666, 'GAG': 0.31999999999999995, 'GAT': 0.39333333333333337, 'GCA': 0.8333333333333334, 'GCC': 0.16666666666666666, 'GCG': 0.2666666666666666, 'GCT': 0.09833333333333334, 'GGA': 0.5, 'GGC': 0.6666666666666666, 'GGG': 0.15999999999999998, 'GGT': 0.39333333333333337, 'GTA': 0.6666666666666666, 'GTC': 0.16666666666666666, 'GTG': 0.2133333333333333, 'GTT': 0.09833333333333334, 'TAC': 0.3333333333333333, 'TAT': 0.19666666666666668, 'TCA': 0.3333333333333333, 'TCC': 0.16666666666666666, 'TCG': 0.10666666666666665, 'TCT': 0.09833333333333334, 'TGC': 0.16666666666666666, 'TGG': 0.16666666666666666, 'TGT': 0.09833333333333334, 'TTA': 0.5183333333333333, 'TTC': 0.5, 'TTG': 0.32666666666666666, 'TTT': 0.29500000000000004}, 'tai_avg': 0.3271366648197227, 'tai_std': 0.028076995641981673}]}, 'orf_debug': {'total_loss': {'C': {'TGT': 0.984009360374415, 'TGC': 1.0}, 'D': {'GAT': 0.906086387434555, 'GAC': 1.0}, 'S': {'TCC': 0.8571428571428571, 'TCG': 0.8845108695652174, 'TCT': 0.8927125506072876, 'AGT': 1.0348240469208212, 'TCA': 1.11125, 'AGC': 1.125}, 'Q': {'CAG': 0.6880733944954128, 'CAA': 1.0525}, 'M': {'ATG': 1.0}, 'N': {'AAT': 0.9463350785340315, 'AAC': 1.0}, 'P': {'CCC': 0.6933333333333334, 'CCT': 0.7275, 'CCG': 0.7462686567164178, 'CCA': 1.02625}, 'K': {'AAA': 1.0, 'AAG': 1.0389908256880733}, '_': {'TAG': 0.9271127128057303, 'TGA': 0.9271127128057303, 'TAA': 0.9271127128057303}, 'T': {'ACG': 0.6880733944954128, 'ACC': 0.6933333333333334, 'ACT': 0.7284803400637619, 'ACA': 1.1375}, 'F': {'TTT': 0.9149377593360996, 'TTC': 1.0}, 'A': {'GCC': 0.6749999999999999, 'GCT': 0.7074927953890491, 'GCG': 0.8010638297872339, 'GCA': 1.0}, 'G': {'GGG': 0.8860619469026548, 'GGT': 0.9463350785340315, 'GGC': 1.0, 'GGA': 1.1667857142857145}, 'I': {'ATA': 0.9100719424460432, 'ATT': 0.9574688796680498, 'ATC': 1.0}, 'L': {'CTG': 0.8032128514056224, 'CTT': 1.0325704225352113, 'CTC': 1.0468557336621456, 'TTG': 1.153846153846154, 'CTA': 1.1657142857142857, 'TTA': 1.40125}, 'H': {'CAT': 0.9248533724340176, 'CAC': 1.0}, 'R': {'CGA': 0.9016744863290729, 'CGG': 0.9166666666666666, 'AGA': 0.9187289088863893, 'AGG': 0.922811059907834, 'CGC': 0.960674157303371, 'CGT': 1.0}, 'W': {'TGG': 1.0}, 'V': {'GTC': 0.9166666666666666, 'GTT': 0.9420828905419766, 'GTG': 0.9805045871559633, 'GTA': 1.0}, 'E': {'GAG': 0.7976190476190476, 'GAA': 1.0}, 'Y': {'TAC': 1.0, 'TAT': 1.0150293255131966}}, 'optimized_loss': {'C': defaultdict(<class 'float'>, {'TGT': 1.1051265625, 'TGC': 1.0}), 'D': defaultdict(<class 'float'>, {'GAT': 1.3311390625000001, 'GAC': 1.0}), 'S': defaultdict(<class 'float'>, {'TCT': 1.21550625, 'TCG': 1.177225, 'TCA': 1.2348765625000002, 'TCC': 1.0, 'AGC': 1.265625, 'AGT': 1.3835640625}), 'Q': defaultdict(<class 'float'>, {'CAA': 1.10775625, 'CAG': 1.0}), 'M': defaultdict(<class 'float'>, {'ATG': 1.0}), 'N': defaultdict(<class 'float'>, {'AAC': 1.0, 'AAT': 1.4520250000000001}), 'P': defaultdict(<class 'float'>, {'CCT': 1.1908265625, 'CCG': 1.0, 'CCA': 1.0531890625000002, 'CCC': 1.0816000000000001}), 'K': defaultdict(<class 'float'>, {'AAG': 2.2801, 'AAA': 1.0}), '_': defaultdict(<class 'float'>, {'TAG': 1.5462535182753292, 'TGA': 1.5462535182753292, 'TAA': 1.5462535182753292}), 'T': defaultdict(<class 'float'>, {'ACC': 1.0816000000000001, 'ACA': 1.2939062499999998, 'ACG': 1.0, 'ACT': 1.3053062500000001}), 'F': defaultdict(<class 'float'>, {'TTT': 1.21550625, 'TTC': 1.0}), 'A': defaultdict(<class 'float'>, {'GCA': 1.0, 'GCC': 1.265625, 'GCG': 1.5750249999999997, 'GCT': 1.50675625}), 'G': defaultdict(<class 'float'>, {'GGT': 1.4520250000000001, 'GGG': 1.782225, 'GGA': 1.8530015625000003, 'GGC': 1.0}), 'I': defaultdict(<class 'float'>, {'ATC': 1.0, 'ATA': 1.6002250000000002, 'ATT': 1.3311390625000001}), 'L': defaultdict(<class 'float'>, {'TTA': 1.9635015625000003, 'TTG': 1.890625, 'CTC': 2.002225, 'CTT': 2.1498890625000002, 'CTG': 1.0, 'CTA': 1.8496000000000004}), 'H': defaultdict(<class 'float'>, {'CAT': 1.1051265625, 'CAC': 1.0}), 'R': defaultdict(<class 'float'>, {'CGA': 2.20879044, 'CGC': 1.2996000000000003, 'CGG': 1.890625, 'CGT': 1.0, 'AGG': 1.782225, 'AGA': 1.8530015625000003}), 'W': defaultdict(<class 'float'>, {'TGG': 1.0}), 'V': defaultdict(<class 'float'>, {'GTA': 1.0, 'GTC': 1.890625, 'GTG': 2.030625, 'GTT': 2.18300625}), 'E': defaultdict(<class 'float'>, {'GAG': 1.7956000000000003, 'GAA': 1.0}), 'Y': defaultdict(<class 'float'>, {'TAT': 1.3311390625000001, 'TAC': 1.0})}, 'deoptimized_loss': {'C': defaultdict(<class 'float'>, {'TGT': 1.141336111111111, 'TGC': 1.0}), 'D': defaultdict(<class 'float'>, {'GAT': 1.6213777777777774, 'GAC': 1.0}), 'S': defaultdict(<class 'float'>, {'TCT': 1.5252249999999996, 'TCG': 1.5047111111111109, 'TCA': 1.0, 'TCC': 1.3611111111111114, 'AGC': 1.0, 'AGT': 1.2920111111111112}), 'Q': defaultdict(<class 'float'>, {'CAA': 1.0, 'CAG': 2.1121777777777777}), 'M': defaultdict(<class 'float'>, {'ATG': 1.0}), 'N': defaultdict(<class 'float'>, {'AAC': 1.0, 'AAT': 1.6213777777777774}), 'P': defaultdict(<class 'float'>, {'CCT': 2.25, 'CCG': 1.7956000000000003, 'CCA': 1.0, 'CCC': 2.25}), 'K': defaultdict(<class 'float'>, {'AAG': 2.1121777777777777, 'AAA': 1.0}), '_': defaultdict(<class 'float'>, {'TAG': 1.7989356493995978, 'TGA': 1.7989356493995978, 'TAA': 1.7989356493995978}), 'T': defaultdict(<class 'float'>, {'ACC': 2.25, 'ACA': 1.0, 'ACG': 2.1121777777777777, 'ACT': 2.4596694444444447}), 'F': defaultdict(<class 'float'>, {'TTT': 1.4520250000000001, 'TTC': 1.0}), 'A': defaultdict(<class 'float'>, {'GCA': 1.0, 'GCC': 2.777777777777778, 'GCG': 2.454444444444445, 'GCT': 3.0102249999999997}), 'G': defaultdict(<class 'float'>, {'GGT': 1.6213777777777774, 'GGG': 2.2700444444444443, 'GGA': 1.3611111111111107, 'GGC': 1.0}), 'I': defaultdict(<class 'float'>, {'ATC': 1.0, 'ATA': 1.9321000000000004, 'ATT': 1.4520250000000001}), 'L': defaultdict(<class 'float'>, {'TTA': 1.0, 'TTG': 1.4200694444444444, 'CTC': 1.8270027777777775, 'CTT': 2.0164, 'CTG': 1.5500250000000002, 'CTA': 1.3611111111111114}), 'H': defaultdict(<class 'float'>, {'CAT': 1.2920111111111112, 'CAC': 1.0}), 'R': defaultdict(<class 'float'>, {'CGA': 2.7167830044444448, 'CGC': 1.4081777777777775, 'CGG': 2.25, 'CGT': 1.0, 'AGG': 2.0928444444444447, 'AGA': 2.195336111111111}), 'W': defaultdict(<class 'float'>, {'TGG': 1.0}), 'V': defaultdict(<class 'float'>, {'GTA': 1.0, 'GTC': 2.25, 'GTG': 2.1121777777777777, 'GTT': 2.4596694444444447}), 'E': defaultdict(<class 'float'>, {'GAG': 2.8224000000000005, 'GAA': 1.0}), 'Y': defaultdict(<class 'float'>, {'TAT': 1.2920111111111112, 'TAC': 1.0})}}, 'orf': {'orf_module_input_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'optimized_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGCTTTAAAGTCCATATGGAGGGGTCCGTCAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCCTACGAGGGGACGCAGACGGCCAAACTGAAAGTCACGAAAGGGGGTCCCCTGCCCTTTGCCTGGGATATTCTGTCCCCCCAGTTTATGTACGGGTCCAAAGCCTACGTCAAACATCCCGCCGATATTCCCGATTACCTGAAACTGTCCTTTCCCGAGGGGTTTAAATGGGAGCGCGTCATGAATTTTGAGGATGGGGGTGTCGTTACGGTCACGCAGGATTCCTCGCTGCAGGATGGGGAGTTTATTTACAAAGTCAAACTGCGCGGGACGAATTTTCCCTCCGATGGGCCCGTCATGCAGAAAAAGACGATGGGGTGGGAGGCCTCCTCGGAGCGCATGTACCCCGAGGATGGGGCCCTGAAAGGGGAGATTAAACAGCGCCTGAAACTGAAAGATGGGGGTCATTACGATGCCGAGGTCAAAACGACCTACAAAGCCAAAAAGCCCGTCCAGCTGCCCGGGGCCTACAATGTCAATATTAAACTGGATATTACGTCCCATAATGAGGATTACACGATTGTCGAGCAGTACGAGCGCGCCGAGGGGCGCCATTCCACGGGGGGTATGGATGAGCTGTACAAATAA', 'aa_to_loss_mapping': {'C': {'TGT': 0.984009360374415, 'TGC': 1.0}, 'D': {'GAT': 0.906086387434555, 'GAC': 1.0}, 'S': {'TCC': 0.8571428571428571, 'TCG': 0.8845108695652174, 'TCT': 0.8927125506072876, 'AGT': 1.0348240469208212, 'TCA': 1.11125, 'AGC': 1.125}, 'Q': {'CAG': 0.6880733944954128, 'CAA': 1.0525}, 'M': {'ATG': 1.0}, 'N': {'AAT': 0.9463350785340315, 'AAC': 1.0}, 'P': {'CCC': 0.6933333333333334, 'CCT': 0.7275, 'CCG': 0.7462686567164178, 'CCA': 1.02625}, 'K': {'AAA': 1.0, 'AAG': 1.0389908256880733}, '_': {'TAG': 0.9271127128057303, 'TGA': 0.9271127128057303, 'TAA': 0.9271127128057303}, 'T': {'ACG': 0.6880733944954128, 'ACC': 0.6933333333333334, 'ACT': 0.7284803400637619, 'ACA': 1.1375}, 'F': {'TTT': 0.9149377593360996, 'TTC': 1.0}, 'A': {'GCC': 0.6749999999999999, 'GCT': 0.7074927953890491, 'GCG': 0.8010638297872339, 'GCA': 1.0}, 'G': {'GGG': 0.8860619469026548, 'GGT': 0.9463350785340315, 'GGC': 1.0, 'GGA': 1.1667857142857145}, 'I': {'ATA': 0.9100719424460432, 'ATT': 0.9574688796680498, 'ATC': 1.0}, 'L': {'CTG': 0.8032128514056224, 'CTT': 1.0325704225352113, 'CTC': 1.0468557336621456, 'TTG': 1.153846153846154, 'CTA': 1.1657142857142857, 'TTA': 1.40125}, 'H': {'CAT': 0.9248533724340176, 'CAC': 1.0}, 'R': {'CGA': 0.9016744863290729, 'CGG': 0.9166666666666666, 'AGA': 0.9187289088863893, 'AGG': 0.922811059907834, 'CGC': 0.960674157303371, 'CGT': 1.0}, 'W': {'TGG': 1.0}, 'V': {'GTC': 0.9166666666666666, 'GTT': 0.9420828905419766, 'GTG': 0.9805045871559633, 'GTA': 1.0}, 'E': {'GAG': 0.7976190476190476, 'GAA': 1.0}, 'Y': {'TAC': 1.0, 'TAT': 1.0150293255131966}}, 'run_time': 0.8439999999245629}, 'evaluation': [{'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'tai_initial_score': 0.2816376469779175, 'tai_final_score': 0.25314983200053065, 'dist_score': -1.1434864563820935}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'tai_initial_score': 0.33696708001007003, 'tai_final_score': 0.27795533275103074, 'dist_score': -2.1017828264646283}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGCTTTAAAGTCCATATGGAGGGGTCCGTCAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCCTACGAGGGGACGCAGACGGCCAAACTGAAAGTCACGAAAGGGGGTCCCCTGCCCTTTGCCTGGGATATTCTGTCCCCCCAGTTTATGTACGGGTCCAAAGCCTACGTCAAACATCCCGCCGATATTCCCGATTACCTGAAACTGTCCTTTCCCGAGGGGTTTAAATGGGAGCGCGTCATGAATTTTGAGGATGGGGGTGTCGTTACGGTCACGCAGGATTCCTCGCTGCAGGATGGGGAGTTTATTTACAAAGTCAAACTGCGCGGGACGAATTTTCCCTCCGATGGGCCCGTCATGCAGAAAAAGACGATGGGGTGGGAGGCCTCCTCGGAGCGCATGTACCCCGAGGATGGGGCCCTGAAAGGGGAGATTAAACAGCGCCTGAAACTGAAAGATGGGGGTCATTACGATGCCGAGGTCAAAACGACCTACAAAGCCAAAAAGCCCGTCCAGCTGCCCGGGGCCTACAATGTCAATATTAAACTGGATATTACGTCCCATAATGAGGATTACACGATTGTCGAGCAGTACGAGCGCGCCGAGGGGCGCCATTCCACGGGGGGTATGGATGAGCTGTACAAATAA', 'average_distance_score': 0.47914818504126744, 'weakest_link_score': 0.47914818504126744, 'ratio_score': 1.023878204300951, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTTATGCGCTTTAAAGTCCATATGGAGGGGTCCGTCAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCCTACGAGGGGACGCAGACGGCCAAACTGAAAGTCACGAAAGGGGGTCCCCTGCCCTTTGCCTGGGATATTCTGTCCCCCCAGTTTATGTACGGGTCCAAAGCCTACGTCAAACATCCCGCCGATATTCCCGATTACCTGAAACTGTCCTTTCCCGAGGGGTTTAAATGGGAGCGCGTCATGAATTTTGAGGATGGGGGTGTCGTTACGGTCACGCAGGATTCCTCGCTGCAGGATGGGGAGTTTATTTACAAAGTCAAACTGCGCGGGACGAATTTTCCCTCCGATGGGCCCGTCATGCAGAAAAAGACGATGGGGTGGGAGGCCTCCTCGGAGCGCATGTACCCCGAGGATGGGGCCCTGAAAGGGGAGATTAAACAGCGCCTGAAACTGAAAGATGGGGGTCATTACGATGCCGAGGTCAAAACGACCTACAAAGCCAAAAAGCCCGTCCAGCTGCCCGGGGCCTACAATGTCAATATTAAACTGGATATTACGTCCCATAATGAGGATTACACGATTGTCGAGCAGTACGAGCGCGCCGAGGGGCGCCATTCCACGGGGGGTATGGATGAGCTGTACAAATAA'}], 'final_evaluation': {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGCTTTAAAGTCCATATGGAGGGGTCCGTCAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCCTACGAGGGGACGCAGACGGCCAAACTGAAAGTCACGAAAGGGGGTCCCCTGCCCTTTGCCTGGGATATTCTGTCCCCCCAGTTTATGTACGGGTCCAAAGCCTACGTCAAACATCCCGCCGATATTCCCGATTACCTGAAACTGTCCTTTCCCGAGGGGTTTAAATGGGAGCGCGTCATGAATTTTGAGGATGGGGGTGTCGTTACGGTCACGCAGGATTCCTCGCTGCAGGATGGGGAGTTTATTTACAAAGTCAAACTGCGCGGGACGAATTTTCCCTCCGATGGGCCCGTCATGCAGAAAAAGACGATGGGGTGGGAGGCCTCCTCGGAGCGCATGTACCCCGAGGATGGGGCCCTGAAAGGGGAGATTAAACAGCGCCTGAAACTGAAAGATGGGGGTCATTACGATGCCGAGGTCAAAACGACCTACAAAGCCAAAAAGCCCGTCCAGCTGCCCGGGGCCTACAATGTCAATATTAAACTGGATATTACGTCCCATAATGAGGATTACACGATTGTCGAGCAGTACGAGCGCGCCGAGGGGCGCCATTCCACGGGGGGTATGGATGAGCTGTACAAATAA', 'average_distance_score': 0.47914818504126744, 'weakest_link_score': 0.47914818504126744, 'ratio_score': 1.023878204300951}, 'zip_output_file_path': 'results\\\\mcherry-wanted-constraint-init-external-0.1\\\\tAI_single_codon_ratio_ecoli_opt_True_6ezh\\\\communique_results.zip'}\n",
      "\n",
      "##########################\n",
      "# User Input #\n",
      "##########################\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is:\n",
      "ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "------------------------------------------\n",
      "Parsing information for Bacillus subtilis:\n",
      "------------------------------------------\n",
      "Organism is defined as wanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Bacillus subtilis. Calculating tAI profile.\n",
      "tai_std=0.028076995641981673, tai_avg=0.3271366648197227\n",
      "------------------------------------------\n",
      "Parsing information for Escherichia coli:\n",
      "------------------------------------------\n",
      "Organism is defined as unwanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Escherichia coli. Calculating tAI profile.\n",
      "tai_std=0.024913119712427708, tai_avg=0.24958213077316901\n",
      "-----------------------------\n",
      "Normalized Prioritization Weights\n",
      "-----------------------------\n",
      "Bacillus subtilis : 1.0\n",
      "Escherichia coli : 1.0\n",
      "Total input processing time: 26.4543354511261\n",
      "\n",
      "##########################\n",
      "# Initiation #\n",
      "##########################\n",
      "Running initiation optimization InitiationOptimizationMethod.external_module on 15 codons at the start of the ORF\n",
      "Taking optimized sequence from configuration value config.INITIATION.EXTERNAL_INITIATION_ORF\n",
      "Optimized sequence is: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "Optimization method used is: ORFOptimizationMethod.single_codon_ratio\n",
      "Loss summary for: TGT\n",
      "loss for Bacillus subtilis: 1.141336111111111\n",
      "loss for Escherichia coli: 1.1051265625\n",
      "Loss summary for: TGC\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGT': 1.0162504954419342, 'TGC': 1.0})\n",
      "Loss summary for: GAT\n",
      "loss for Bacillus subtilis: 1.6213777777777774\n",
      "loss for Escherichia coli: 1.3311390625000001\n",
      "Loss summary for: GAC\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAT': 1.1036475261827372, 'GAC': 1.0})\n",
      "Loss summary for: TCT\n",
      "loss for Bacillus subtilis: 1.5252249999999996\n",
      "loss for Escherichia coli: 1.21550625\n",
      "Loss summary for: TCG\n",
      "loss for Bacillus subtilis: 1.5047111111111109\n",
      "loss for Escherichia coli: 1.177225\n",
      "Loss summary for: TCA\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.2348765625000002\n",
      "Loss summary for: TCC\n",
      "loss for Bacillus subtilis: 1.3611111111111114\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: AGC\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.265625\n",
      "Loss summary for: AGT\n",
      "loss for Bacillus subtilis: 1.2920111111111112\n",
      "loss for Escherichia coli: 1.3835640625\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TCT': 1.1201814058956914, 'TCG': 1.130568356374808, 'TCA': 0.8998875140607423, 'TCC': 1.1666666666666667, 'AGC': 0.8888888888888888, 'AGT': 0.9663478568898335})\n",
      "Loss summary for: CAA\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.10775625\n",
      "Loss summary for: CAG\n",
      "loss for Bacillus subtilis: 2.1121777777777777\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAA': 0.9501187648456058, 'CAG': 1.4533333333333334})\n",
      "Loss summary for: ATG\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATG': 1.0})\n",
      "Loss summary for: AAC\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: AAT\n",
      "loss for Bacillus subtilis: 1.6213777777777774\n",
      "loss for Escherichia coli: 1.4520250000000001\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAC': 1.0, 'AAT': 1.0567081604426})\n",
      "Loss summary for: CCT\n",
      "loss for Bacillus subtilis: 2.25\n",
      "loss for Escherichia coli: 1.1908265625\n",
      "Loss summary for: CCG\n",
      "loss for Bacillus subtilis: 1.7956000000000003\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: CCA\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.0531890625000002\n",
      "Loss summary for: CCC\n",
      "loss for Bacillus subtilis: 2.25\n",
      "loss for Escherichia coli: 1.0816000000000001\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CCT': 1.3745704467353952, 'CCG': 1.34, 'CCA': 0.9744214372716199, 'CCC': 1.4423076923076923})\n",
      "Loss summary for: AAG\n",
      "loss for Bacillus subtilis: 2.1121777777777777\n",
      "loss for Escherichia coli: 2.2801\n",
      "Loss summary for: AAA\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAG': 0.9624724061810155, 'AAA': 1.0})\n",
      "Loss summary for: TAG\n",
      "loss for Bacillus subtilis: 1.7989356493995978\n",
      "loss for Escherichia coli: 1.5462535182753292\n",
      "Loss summary for: TGA\n",
      "loss for Bacillus subtilis: 1.7989356493995978\n",
      "loss for Escherichia coli: 1.5462535182753292\n",
      "Loss summary for: TAA\n",
      "loss for Bacillus subtilis: 1.7989356493995978\n",
      "loss for Escherichia coli: 1.5462535182753292\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAG': 1.0786175037700543, 'TGA': 1.0786175037700543, 'TAA': 1.0786175037700543})\n",
      "Loss summary for: ACC\n",
      "loss for Bacillus subtilis: 2.25\n",
      "loss for Escherichia coli: 1.0816000000000001\n",
      "Loss summary for: ACA\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.2939062499999998\n",
      "Loss summary for: ACG\n",
      "loss for Bacillus subtilis: 2.1121777777777777\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: ACT\n",
      "loss for Bacillus subtilis: 2.4596694444444447\n",
      "loss for Escherichia coli: 1.3053062500000001\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ACC': 1.4423076923076923, 'ACA': 0.8791208791208791, 'ACG': 1.4533333333333334, 'ACT': 1.3727206418672502})\n",
      "Loss summary for: TTT\n",
      "loss for Bacillus subtilis: 1.4520250000000001\n",
      "loss for Escherichia coli: 1.21550625\n",
      "Loss summary for: TTC\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTT': 1.09297052154195, 'TTC': 1.0})\n",
      "Loss summary for: GCA\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: GCC\n",
      "loss for Bacillus subtilis: 2.777777777777778\n",
      "loss for Escherichia coli: 1.265625\n",
      "Loss summary for: GCG\n",
      "loss for Bacillus subtilis: 2.454444444444445\n",
      "loss for Escherichia coli: 1.5750249999999997\n",
      "Loss summary for: GCT\n",
      "loss for Bacillus subtilis: 3.0102249999999997\n",
      "loss for Escherichia coli: 1.50675625\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GCA': 1.0, 'GCC': 1.4814814814814816, 'GCG': 1.2483399734395753, 'GCT': 1.4134419551934825})\n",
      "Loss summary for: GGT\n",
      "loss for Bacillus subtilis: 1.6213777777777774\n",
      "loss for Escherichia coli: 1.4520250000000001\n",
      "Loss summary for: GGG\n",
      "loss for Bacillus subtilis: 2.2700444444444443\n",
      "loss for Escherichia coli: 1.782225\n",
      "Loss summary for: GGA\n",
      "loss for Bacillus subtilis: 1.3611111111111107\n",
      "loss for Escherichia coli: 1.8530015625000003\n",
      "Loss summary for: GGC\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GGT': 1.0567081604426, 'GGG': 1.128589263420724, 'GGA': 0.8570554025099478, 'GGC': 1.0})\n",
      "Loss summary for: ATC\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: ATA\n",
      "loss for Bacillus subtilis: 1.9321000000000004\n",
      "loss for Escherichia coli: 1.6002250000000002\n",
      "Loss summary for: ATT\n",
      "loss for Bacillus subtilis: 1.4520250000000001\n",
      "loss for Escherichia coli: 1.3311390625000001\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATC': 1.0, 'ATA': 1.098814229249012, 'ATT': 1.0444203683640303})\n",
      "Loss summary for: TTA\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.9635015625000003\n",
      "Loss summary for: TTG\n",
      "loss for Bacillus subtilis: 1.4200694444444444\n",
      "loss for Escherichia coli: 1.890625\n",
      "Loss summary for: CTC\n",
      "loss for Bacillus subtilis: 1.8270027777777775\n",
      "loss for Escherichia coli: 2.002225\n",
      "Loss summary for: CTT\n",
      "loss for Bacillus subtilis: 2.0164\n",
      "loss for Escherichia coli: 2.1498890625000002\n",
      "Loss summary for: CTG\n",
      "loss for Bacillus subtilis: 1.5500250000000002\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: CTA\n",
      "loss for Bacillus subtilis: 1.3611111111111114\n",
      "loss for Escherichia coli: 1.8496000000000004\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTA': 0.7136485280999108, 'TTG': 0.8666666666666667, 'CTC': 0.9552414605418138, 'CTT': 0.9684569479965899, 'CTG': 1.245, 'CTA': 0.8578431372549019})\n",
      "Loss summary for: CAT\n",
      "loss for Bacillus subtilis: 1.2920111111111112\n",
      "loss for Escherichia coli: 1.1051265625\n",
      "Loss summary for: CAC\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAT': 1.081252477209671, 'CAC': 1.0})\n",
      "Loss summary for: CGA\n",
      "loss for Bacillus subtilis: 2.7167830044444448\n",
      "loss for Escherichia coli: 2.20879044\n",
      "Loss summary for: CGC\n",
      "loss for Bacillus subtilis: 1.4081777777777775\n",
      "loss for Escherichia coli: 1.2996000000000003\n",
      "Loss summary for: CGG\n",
      "loss for Bacillus subtilis: 2.25\n",
      "loss for Escherichia coli: 1.890625\n",
      "Loss summary for: CGT\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: AGG\n",
      "loss for Bacillus subtilis: 2.0928444444444447\n",
      "loss for Escherichia coli: 1.782225\n",
      "Loss summary for: AGA\n",
      "loss for Bacillus subtilis: 2.195336111111111\n",
      "loss for Escherichia coli: 1.8530015625000003\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CGA': 1.1090476831292335, 'CGC': 1.0409356725146197, 'CGG': 1.0909090909090908, 'CGT': 1.0, 'AGG': 1.0836454431960052, 'AGA': 1.0884603611876338})\n",
      "Loss summary for: TGG\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGG': 1.0})\n",
      "Loss summary for: GTA\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: GTC\n",
      "loss for Bacillus subtilis: 2.25\n",
      "loss for Escherichia coli: 1.890625\n",
      "Loss summary for: GTG\n",
      "loss for Bacillus subtilis: 2.1121777777777777\n",
      "loss for Escherichia coli: 2.030625\n",
      "Loss summary for: GTT\n",
      "loss for Bacillus subtilis: 2.4596694444444447\n",
      "loss for Escherichia coli: 2.18300625\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GTA': 1.0, 'GTC': 1.0909090909090908, 'GTG': 1.0198830409356725, 'GTT': 1.0614777213761986})\n",
      "Loss summary for: GAG\n",
      "loss for Bacillus subtilis: 2.8224000000000005\n",
      "loss for Escherichia coli: 1.7956000000000003\n",
      "Loss summary for: GAA\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAG': 1.2537313432835822, 'GAA': 1.0})\n",
      "Loss summary for: TAT\n",
      "loss for Bacillus subtilis: 1.2920111111111112\n",
      "loss for Escherichia coli: 1.3311390625000001\n",
      "Loss summary for: TAC\n",
      "loss for Bacillus subtilis: 1.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAT': 0.9851932105453233, 'TAC': 1.0})\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31522480175104955]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.18161187196589967]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31522480175104955]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3265866087018494]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.22534685577492342]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4344682502399869]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3265866087018494]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31522480175104955]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.366762837741479]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.18161187196589967]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19095344268085387]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.198530453110291]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3390368073029434]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19095344268085387]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.198530453110291]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19095344268085387]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31522480175104955]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.35960560391899904]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.366762837741479]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.198530453110291]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.22534685577492342]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19095344268085387]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31522480175104955]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.22534685577492342]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3265866087018494]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19095344268085387]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.35960560391899904]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.366762837741479]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19095344268085387]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.35960560391899904]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.198530453110291]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.198530453110291]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.22534685577492342]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31522480175104955]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19095344268085387]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31522480175104955]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.18161187196589967]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4344682502399869]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31522480175104955]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.35960560391899904]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3390368073029434]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26187720031028106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.35960560391899904]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.22534685577492342]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.22534685577492342]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23653118484388272]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.198530453110291]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.35960560391899904]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31522480175104955]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.366762837741479]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.198530453110291]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.18161187196589967]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4344682502399869]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31522480175104955]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19095344268085387]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.22534685577492342]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.35960560391899904]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19095344268085387]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7016557540022625]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.22534685577492342]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.22534685577492342]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23653118484388272]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.18161187196589967]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19095344268085387]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.35960560391899904]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.198530453110291]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.366762837741479]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.18161187196589967]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.198530453110291]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.198530453110291]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.35960560391899904]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3390368073029434]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3265866087018494]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.35960560391899904]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.1614000537105004]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7016557540022625]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19095344268085387]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.198530453110291]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19095344268085387]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4344682502399869]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4344682502399869]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.366762837741479]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.198530453110291]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.35960560391899904]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.366762837741479]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.22534685577492342]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3265866087018494]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4344682502399869]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.35960560391899904]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.366762837741479]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.18161187196589967]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.18161187196589967]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3265866087018494]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.22534685577492342]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3390368073029434]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.35960560391899904]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.198530453110291]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14132712456344587]\n",
      "\n",
      "##########################\n",
      "# EVALUATION #\n",
      "##########################\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTCATGCGTTTCAAGGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATATGAAGGAACACAAACAGCAAAGTTAAAGGTAACAAAGGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTATGGAAGCAAGGCATATGTAAAGCACCCAGCAGACATCCCAGACTATTTAAAGTTAAGCTTCCCAGAAGGATTCAAGTGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTATAAGGTAAAGTTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAGAAAACAATGGGATGGGAAGCAAGCTCAGAACGTATGTATCCAGAAGACGGAGCATTAAAGGGAGAAATCAAGCAACGTTTAAAGTTAAAGGACGGAGGCCACTATGACGCAGAAGTAAAGACAACTTATAAGGCAAAGAAACCAGTACAATTACCAGGAGCATATAACGTAAACATCAAGTTAGACATCACAAGCCACAACGAAGACTATACAATCGTAGAACAATATGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATATAAGTAA\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATATGAAGGAACACAAACAGCAAAGTTAAAGGTAACAAAGGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTATGGAAGCAAGGCATATGTAAAGCACCCAGCAGACATCCCAGACTATTTAAAGTTAAGCTTCCCAGAAGGATTCAAGTGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTATAAGGTAAAGTTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAGAAAACAATGGGATGGGAAGCAAGCTCAGAACGTATGTATCCAGAAGACGGAGCATTAAAGGGAGAAATCAAGCAACGTTTAAAGTTAAAGGACGGAGGCCACTATGACGCAGAAGTAAAGACAACTTATAAGGCAAAGAAACCAGTACAATTACCAGGAGCATATAACGTAAACATCAAGTTAGACATCACAAGCCACAACGAAGACTATACAATCGTAGAACAATATGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATATAAGTAA', 'average_distance_score': 3.380712579917882, 'weakest_link_score': 3.380712579917882, 'ratio_score': 1.1494500671677315}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATATGAAGGAACACAAACAGCAAAGTTAAAGGTAACAAAGGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTATGGAAGCAAGGCATATGTAAAGCACCCAGCAGACATCCCAGACTATTTAAAGTTAAGCTTCCCAGAAGGATTCAAGTGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTATAAGGTAAAGTTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAGAAAACAATGGGATGGGAAGCAAGCTCAGAACGTATGTATCCAGAAGACGGAGCATTAAAGGGAGAAATCAAGCAACGTTTAAAGTTAAAGGACGGAGGCCACTATGACGCAGAAGTAAAGACAACTTATAAGGCAAAGAAACCAGTACAATTACCAGGAGCATATAACGTAAACATCAAGTTAGACATCACAAGCCACAACGAAGACTATACAATCGTAGAACAATATGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATATAAGTAA', 'average_distance_score': 3.380712579917882, 'weakest_link_score': 3.380712579917882, 'ratio_score': 1.1494500671677315}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry-wanted-constraint-init-external-0.1\\tAI_single_codon_ratio_ecoli_opt_False_2lJj\n",
      "Final output: {'module_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'orf_optimization_method': 'single_codon_ratio', 'orf_optimization_cub_index': 'tAI', 'initiation_optimization_method': 'external', 'evaluation_score': 'average_distance', 'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': None, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'GAA': 0.6798186496752849, 'AAT': 0.5655317497600131, 'ATA': 0.1287010322849589, 'TTA': 0.198530453110291, 'GAC': 0.35960560391899904, 'CTG': 0.23973441522197322, 'TGG': 1.0, 'AAC': 0.4344682502399869, 'CAA': 0.5131548459124133, 'GCC': 0.20715414603308396, 'CTT': 0.2391679232236727, 'GCT': 0.24752923822568748, 'ATC': 0.366762837741479, 'AAA': 0.7016557540022625, 'AAG': 0.29834424599773746, 'TTG': 0.1592175745811708, 'AGC': 0.22534685577492342, 'CCG': 0.4334091260292722, 'AGT': 0.10611598754150686, 'TTT': 0.6847751982489505, 'GAG': 0.3201813503247151, 'ACT': 0.1614000537105004, 'TCA': 0.23653118484388272, 'ACC': 0.1588786441082565, 'CAC': 0.3265866087018494, 'GGC': 0.3390368073029434, 'GAT': 0.640394396081001, 'ACA': 0.4122430101751559, 'ACG': 0.2674782920060872, 'CCC': 0.0889644363010221, 'AGA': 0.26380392311639106, 'TCC': 0.12712038919920718, 'TAC': 0.34586274736720285, 'CAT': 0.6734133912981506, 'ATT': 0.5045361299735621, 'GCA': 0.2828363712991255, 'TAT': 0.6541372526327971, 'GGG': 0.16211598104043537, 'GTC': 0.25598186049286953, 'CCT': 0.28667299498885185, 'GTT': 0.28439644370189154, 'GCG': 0.26248024444210305, 'TCT': 0.2048572678833432, 'CTC': 0.11219040795754642, 'CCA': 0.19095344268085387, 'GGA': 0.3148341038094681, 'CGA': 0.09855152543710778, 'GTA': 0.19774449549495793, 'GTG': 0.26187720031028106, 'TTC': 0.31522480175104955, 'CGC': 0.2064569601768165, 'CAG': 0.4868451540875866, 'CGG': 0.1555630106168844, 'CGT': 0.18161187196589967, 'CTA': 0.05115922590534585, 'AGG': 0.09401270868690058, 'GGT': 0.18401310784715313, 'TAG': 0.14132712456344587, 'TGC': 0.5460141582025239, 'TCG': 0.10002831475713661, 'TAA': 0.630500582072177, 'TGA': 0.2281722933643772, 'TGT': 0.45398584179747614}), 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.6666666666666666, 'AAC': 0.6666666666666666, 'AAG': 0.2133333333333333, 'AAT': 0.39333333333333337, 'ACA': 0.6666666666666666, 'ACC': 0.16666666666666666, 'ACG': 0.2133333333333333, 'ACT': 0.09833333333333334, 'AGA': 0.18499999999999997, 'AGC': 0.3333333333333333, 'AGG': 0.21999999999999997, 'AGT': 0.19666666666666668, 'ATA': 0.10999999999999999, 'ATC': 0.5, 'ATG': 1.0, 'ATT': 0.29500000000000004, 'CAA': 0.6666666666666666, 'CAC': 0.3333333333333333, 'CAG': 0.2133333333333333, 'CAT': 0.19666666666666668, 'CCA': 0.5, 'CCG': 0.15999999999999998, 'CGA': 0.01839999999999999, 'CGC': 0.48, 'CGG': 0.16666666666666666, 'CGT': 0.6666666666666666, 'CTA': 0.3516666666666666, 'CTC': 0.16666666666666666, 'CTG': 0.2733333333333333, 'CTT': 0.09833333333333334, 'GAA': 1.0, 'GAC': 0.6666666666666666, 'GAG': 0.31999999999999995, 'GAT': 0.39333333333333337, 'GCA': 0.8333333333333334, 'GCC': 0.16666666666666666, 'GCG': 0.2666666666666666, 'GCT': 0.09833333333333334, 'GGA': 0.5, 'GGC': 0.6666666666666666, 'GGG': 0.15999999999999998, 'GGT': 0.39333333333333337, 'GTA': 0.6666666666666666, 'GTC': 0.16666666666666666, 'GTG': 0.2133333333333333, 'GTT': 0.09833333333333334, 'TAC': 0.3333333333333333, 'TAT': 0.19666666666666668, 'TCA': 0.3333333333333333, 'TCC': 0.16666666666666666, 'TCG': 0.10666666666666665, 'TCT': 0.09833333333333334, 'TGC': 0.16666666666666666, 'TGG': 0.16666666666666666, 'TGT': 0.09833333333333334, 'TTA': 0.5183333333333333, 'TTC': 0.5, 'TTG': 0.32666666666666666, 'TTT': 0.29500000000000004}, 'tai_avg': 0.3271366648197227, 'tai_std': 0.028076995641981673}, {'name': 'Escherichia coli', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': None, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'AAA': 0.7667390608953314, 'CGC': 0.40076028262096014, 'ATT': 0.5085852538544149, 'AGC': 0.27799163607995603, 'ACC': 0.4359429381095841, 'ATC': 0.4205725148436632, 'ACA': 0.13023262212317555, 'GGT': 0.33726003930064996, 'AAC': 0.5512262248110238, 'GCG': 0.3563679190097728, 'GGC': 0.4050486219579785, 'TGA': 0.28926761860893596, 'CGA': 0.06388523224887838, 'GTG': 0.3715940695510341, 'TTG': 0.12808970030428077, 'AAG': 0.23326093910466858, 'TTC': 0.4262385670731707, 'TCA': 0.12251736088908215, 'GCA': 0.21281773299119433, 'AAT': 0.4487737751889762, 'GAA': 0.689467997530356, 'CGT': 0.38120248233619, 'TTT': 0.5737614329268292, 'CTG': 0.4975546385450099, 'GTT': 0.2584847661069405, 'GCC': 0.26990242799092135, 'GAT': 0.626856640686903, 'AGG': 0.020162265266098916, 'CAG': 0.6530527088827319, 'GGG': 0.15057187484254547, 'GTC': 0.21619163129169194, 'CTC': 0.10435163186194824, 'TCT': 0.14544780223293646, 'CCC': 0.12361073915372255, 'CAC': 0.4292446655321377, 'GCT': 0.1609119200081115, 'TTA': 0.13009739763522432, 'CTT': 0.10341378016756284, 'ACG': 0.26849903016796667, 'GGA': 0.10711946389882603, 'CCG': 0.527785216925549, 'CAA': 0.3469472911172681, 'ACT': 0.16532540959927366, 'ATA': 0.07084223130192197, 'CAT': 0.5707553344678623, 'AGT': 0.15102374892893225, 'TGC': 0.55651396076928, 'GAG': 0.31053200246964396, 'TCG': 0.15406749964830613, 'GTA': 0.15372953305033354, 'TAC': 0.4315172054223149, 'TCC': 0.14895195222078703, 'GAC': 0.37314335931309695, 'TGT': 0.4434860392307201, 'TGG': 1.0, 'TAT': 0.5684827945776851, 'CCT': 0.15789033208355652, 'CCA': 0.19071371183717192, 'CGG': 0.09749348521075679, 'CTA': 0.03649285148597391, 'TAA': 0.6391064025794565, 'AGA': 0.036496252317115764, 'TAG': 0.07162597881160755}), 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.75, 'AAC': 0.5, 'AAG': 0.23999999999999996, 'AAT': 0.29500000000000004, 'ACA': 0.1525, 'ACC': 0.25, 'ACG': 0.29, 'ACT': 0.14750000000000002, 'AGA': 0.13874999999999998, 'AGC': 0.125, 'AGG': 0.16499999999999998, 'AGT': 0.07375000000000001, 'ATA': 0.10999999999999999, 'ATC': 0.375, 'ATG': 1.0, 'ATT': 0.22125000000000003, 'CAA': 0.27749999999999997, 'CAC': 0.125, 'CAG': 0.32999999999999996, 'CAT': 0.07375000000000001, 'CCA': 0.13874999999999998, 'CCC': 0.125, 'CCG': 0.16499999999999998, 'CCT': 0.07375000000000001, 'CGA': 0.013799999999999993, 'CGC': 0.36, 'CGG': 0.125, 'CGT': 0.5, 'CTA': 0.18, 'CTC': 0.125, 'CTG': 0.54, 'CTT': 0.07375000000000001, 'GAA': 0.5, 'GAC': 0.375, 'GAG': 0.15999999999999998, 'GAT': 0.22125000000000003, 'GCA': 0.375, 'GCC': 0.25, 'GCG': 0.11999999999999998, 'GCT': 0.14750000000000002, 'GGA': 0.13874999999999998, 'GGC': 0.5, 'GGG': 0.16499999999999998, 'GGT': 0.29500000000000004, 'GTA': 0.625, 'GTC': 0.25, 'GTG': 0.19999999999999996, 'GTT': 0.14750000000000002, 'TAC': 0.375, 'TAT': 0.22125000000000003, 'TCA': 0.13874999999999998, 'TCC': 0.25, 'TCG': 0.16499999999999998, 'TCT': 0.14750000000000002, 'TGC': 0.125, 'TGG': 0.125, 'TGT': 0.07375000000000001, 'TTA': 0.13874999999999998, 'TTC': 0.25, 'TTG': 0.16499999999999998, 'TTT': 0.14750000000000002}, 'tai_avg': 0.24958213077316901, 'tai_std': 0.024913119712427708}]}, 'orf_debug': {'total_loss': {'C': {'TGC': 1.0, 'TGT': 1.0162504954419342}, 'D': {'GAC': 1.0, 'GAT': 1.1036475261827372}, 'S': {'AGC': 0.8888888888888888, 'TCA': 0.8998875140607423, 'AGT': 0.9663478568898335, 'TCT': 1.1201814058956914, 'TCG': 1.130568356374808, 'TCC': 1.1666666666666667}, 'Q': {'CAA': 0.9501187648456058, 'CAG': 1.4533333333333334}, 'M': {'ATG': 1.0}, 'N': {'AAC': 1.0, 'AAT': 1.0567081604426}, 'P': {'CCA': 0.9744214372716199, 'CCG': 1.34, 'CCT': 1.3745704467353952, 'CCC': 1.4423076923076923}, 'K': {'AAG': 0.9624724061810155, 'AAA': 1.0}, '_': {'TAG': 1.0786175037700543, 'TGA': 1.0786175037700543, 'TAA': 1.0786175037700543}, 'T': {'ACA': 0.8791208791208791, 'ACT': 1.3727206418672502, 'ACC': 1.4423076923076923, 'ACG': 1.4533333333333334}, 'F': {'TTC': 1.0, 'TTT': 1.09297052154195}, 'A': {'GCA': 1.0, 'GCG': 1.2483399734395753, 'GCT': 1.4134419551934825, 'GCC': 1.4814814814814816}, 'G': {'GGA': 0.8570554025099478, 'GGC': 1.0, 'GGT': 1.0567081604426, 'GGG': 1.128589263420724}, 'I': {'ATC': 1.0, 'ATT': 1.0444203683640303, 'ATA': 1.098814229249012}, 'L': {'TTA': 0.7136485280999108, 'CTA': 0.8578431372549019, 'TTG': 0.8666666666666667, 'CTC': 0.9552414605418138, 'CTT': 0.9684569479965899, 'CTG': 1.245}, 'H': {'CAC': 1.0, 'CAT': 1.081252477209671}, 'R': {'CGT': 1.0, 'CGC': 1.0409356725146197, 'AGG': 1.0836454431960052, 'AGA': 1.0884603611876338, 'CGG': 1.0909090909090908, 'CGA': 1.1090476831292335}, 'W': {'TGG': 1.0}, 'V': {'GTA': 1.0, 'GTG': 1.0198830409356725, 'GTT': 1.0614777213761986, 'GTC': 1.0909090909090908}, 'E': {'GAA': 1.0, 'GAG': 1.2537313432835822}, 'Y': {'TAT': 0.9851932105453233, 'TAC': 1.0}}, 'optimized_loss': {'C': defaultdict(<class 'float'>, {'TGT': 1.141336111111111, 'TGC': 1.0}), 'D': defaultdict(<class 'float'>, {'GAT': 1.6213777777777774, 'GAC': 1.0}), 'S': defaultdict(<class 'float'>, {'TCT': 1.5252249999999996, 'TCG': 1.5047111111111109, 'TCA': 1.0, 'TCC': 1.3611111111111114, 'AGC': 1.0, 'AGT': 1.2920111111111112}), 'Q': defaultdict(<class 'float'>, {'CAA': 1.0, 'CAG': 2.1121777777777777}), 'M': defaultdict(<class 'float'>, {'ATG': 1.0}), 'N': defaultdict(<class 'float'>, {'AAC': 1.0, 'AAT': 1.6213777777777774}), 'P': defaultdict(<class 'float'>, {'CCT': 2.25, 'CCG': 1.7956000000000003, 'CCA': 1.0, 'CCC': 2.25}), 'K': defaultdict(<class 'float'>, {'AAG': 2.1121777777777777, 'AAA': 1.0}), '_': defaultdict(<class 'float'>, {'TAG': 1.7989356493995978, 'TGA': 1.7989356493995978, 'TAA': 1.7989356493995978}), 'T': defaultdict(<class 'float'>, {'ACC': 2.25, 'ACA': 1.0, 'ACG': 2.1121777777777777, 'ACT': 2.4596694444444447}), 'F': defaultdict(<class 'float'>, {'TTT': 1.4520250000000001, 'TTC': 1.0}), 'A': defaultdict(<class 'float'>, {'GCA': 1.0, 'GCC': 2.777777777777778, 'GCG': 2.454444444444445, 'GCT': 3.0102249999999997}), 'G': defaultdict(<class 'float'>, {'GGT': 1.6213777777777774, 'GGG': 2.2700444444444443, 'GGA': 1.3611111111111107, 'GGC': 1.0}), 'I': defaultdict(<class 'float'>, {'ATC': 1.0, 'ATA': 1.9321000000000004, 'ATT': 1.4520250000000001}), 'L': defaultdict(<class 'float'>, {'TTA': 1.0, 'TTG': 1.4200694444444444, 'CTC': 1.8270027777777775, 'CTT': 2.0164, 'CTG': 1.5500250000000002, 'CTA': 1.3611111111111114}), 'H': defaultdict(<class 'float'>, {'CAT': 1.2920111111111112, 'CAC': 1.0}), 'R': defaultdict(<class 'float'>, {'CGA': 2.7167830044444448, 'CGC': 1.4081777777777775, 'CGG': 2.25, 'CGT': 1.0, 'AGG': 2.0928444444444447, 'AGA': 2.195336111111111}), 'W': defaultdict(<class 'float'>, {'TGG': 1.0}), 'V': defaultdict(<class 'float'>, {'GTA': 1.0, 'GTC': 2.25, 'GTG': 2.1121777777777777, 'GTT': 2.4596694444444447}), 'E': defaultdict(<class 'float'>, {'GAG': 2.8224000000000005, 'GAA': 1.0}), 'Y': defaultdict(<class 'float'>, {'TAT': 1.2920111111111112, 'TAC': 1.0})}, 'deoptimized_loss': {'C': defaultdict(<class 'float'>, {'TGT': 1.1051265625, 'TGC': 1.0}), 'D': defaultdict(<class 'float'>, {'GAT': 1.3311390625000001, 'GAC': 1.0}), 'S': defaultdict(<class 'float'>, {'TCT': 1.21550625, 'TCG': 1.177225, 'TCA': 1.2348765625000002, 'TCC': 1.0, 'AGC': 1.265625, 'AGT': 1.3835640625}), 'Q': defaultdict(<class 'float'>, {'CAA': 1.10775625, 'CAG': 1.0}), 'M': defaultdict(<class 'float'>, {'ATG': 1.0}), 'N': defaultdict(<class 'float'>, {'AAC': 1.0, 'AAT': 1.4520250000000001}), 'P': defaultdict(<class 'float'>, {'CCT': 1.1908265625, 'CCG': 1.0, 'CCA': 1.0531890625000002, 'CCC': 1.0816000000000001}), 'K': defaultdict(<class 'float'>, {'AAG': 2.2801, 'AAA': 1.0}), '_': defaultdict(<class 'float'>, {'TAG': 1.5462535182753292, 'TGA': 1.5462535182753292, 'TAA': 1.5462535182753292}), 'T': defaultdict(<class 'float'>, {'ACC': 1.0816000000000001, 'ACA': 1.2939062499999998, 'ACG': 1.0, 'ACT': 1.3053062500000001}), 'F': defaultdict(<class 'float'>, {'TTT': 1.21550625, 'TTC': 1.0}), 'A': defaultdict(<class 'float'>, {'GCA': 1.0, 'GCC': 1.265625, 'GCG': 1.5750249999999997, 'GCT': 1.50675625}), 'G': defaultdict(<class 'float'>, {'GGT': 1.4520250000000001, 'GGG': 1.782225, 'GGA': 1.8530015625000003, 'GGC': 1.0}), 'I': defaultdict(<class 'float'>, {'ATC': 1.0, 'ATA': 1.6002250000000002, 'ATT': 1.3311390625000001}), 'L': defaultdict(<class 'float'>, {'TTA': 1.9635015625000003, 'TTG': 1.890625, 'CTC': 2.002225, 'CTT': 2.1498890625000002, 'CTG': 1.0, 'CTA': 1.8496000000000004}), 'H': defaultdict(<class 'float'>, {'CAT': 1.1051265625, 'CAC': 1.0}), 'R': defaultdict(<class 'float'>, {'CGA': 2.20879044, 'CGC': 1.2996000000000003, 'CGG': 1.890625, 'CGT': 1.0, 'AGG': 1.782225, 'AGA': 1.8530015625000003}), 'W': defaultdict(<class 'float'>, {'TGG': 1.0}), 'V': defaultdict(<class 'float'>, {'GTA': 1.0, 'GTC': 1.890625, 'GTG': 2.030625, 'GTT': 2.18300625}), 'E': defaultdict(<class 'float'>, {'GAG': 1.7956000000000003, 'GAA': 1.0}), 'Y': defaultdict(<class 'float'>, {'TAT': 1.3311390625000001, 'TAC': 1.0})}}, 'orf': {'orf_module_input_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'optimized_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATATGAAGGAACACAAACAGCAAAGTTAAAGGTAACAAAGGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTATGGAAGCAAGGCATATGTAAAGCACCCAGCAGACATCCCAGACTATTTAAAGTTAAGCTTCCCAGAAGGATTCAAGTGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTATAAGGTAAAGTTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAGAAAACAATGGGATGGGAAGCAAGCTCAGAACGTATGTATCCAGAAGACGGAGCATTAAAGGGAGAAATCAAGCAACGTTTAAAGTTAAAGGACGGAGGCCACTATGACGCAGAAGTAAAGACAACTTATAAGGCAAAGAAACCAGTACAATTACCAGGAGCATATAACGTAAACATCAAGTTAGACATCACAAGCCACAACGAAGACTATACAATCGTAGAACAATATGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATATAAGTAA', 'aa_to_loss_mapping': {'C': {'TGC': 1.0, 'TGT': 1.0162504954419342}, 'D': {'GAC': 1.0, 'GAT': 1.1036475261827372}, 'S': {'AGC': 0.8888888888888888, 'TCA': 0.8998875140607423, 'AGT': 0.9663478568898335, 'TCT': 1.1201814058956914, 'TCG': 1.130568356374808, 'TCC': 1.1666666666666667}, 'Q': {'CAA': 0.9501187648456058, 'CAG': 1.4533333333333334}, 'M': {'ATG': 1.0}, 'N': {'AAC': 1.0, 'AAT': 1.0567081604426}, 'P': {'CCA': 0.9744214372716199, 'CCG': 1.34, 'CCT': 1.3745704467353952, 'CCC': 1.4423076923076923}, 'K': {'AAG': 0.9624724061810155, 'AAA': 1.0}, '_': {'TAG': 1.0786175037700543, 'TGA': 1.0786175037700543, 'TAA': 1.0786175037700543}, 'T': {'ACA': 0.8791208791208791, 'ACT': 1.3727206418672502, 'ACC': 1.4423076923076923, 'ACG': 1.4533333333333334}, 'F': {'TTC': 1.0, 'TTT': 1.09297052154195}, 'A': {'GCA': 1.0, 'GCG': 1.2483399734395753, 'GCT': 1.4134419551934825, 'GCC': 1.4814814814814816}, 'G': {'GGA': 0.8570554025099478, 'GGC': 1.0, 'GGT': 1.0567081604426, 'GGG': 1.128589263420724}, 'I': {'ATC': 1.0, 'ATT': 1.0444203683640303, 'ATA': 1.098814229249012}, 'L': {'TTA': 0.7136485280999108, 'CTA': 0.8578431372549019, 'TTG': 0.8666666666666667, 'CTC': 0.9552414605418138, 'CTT': 0.9684569479965899, 'CTG': 1.245}, 'H': {'CAC': 1.0, 'CAT': 1.081252477209671}, 'R': {'CGT': 1.0, 'CGC': 1.0409356725146197, 'AGG': 1.0836454431960052, 'AGA': 1.0884603611876338, 'CGG': 1.0909090909090908, 'CGA': 1.1090476831292335}, 'W': {'TGG': 1.0}, 'V': {'GTA': 1.0, 'GTG': 1.0198830409356725, 'GTT': 1.0614777213761986, 'GTC': 1.0909090909090908}, 'E': {'GAA': 1.0, 'GAG': 1.2537313432835822}, 'Y': {'TAT': 0.9851932105453233, 'TAC': 1.0}}, 'run_time': 0.7650000000139698}, 'evaluation': [{'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'tai_initial_score': 0.33696708001007003, 'tai_final_score': 0.5020103929506466, 'dist_score': 5.87823978908193}, {'name': 'Escherichia coli', 'is_wanted': False, 'tai_initial_score': 0.2816376469779175, 'tai_final_score': 0.25963474410806237, 'dist_score': -0.8831853707538343}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATATGAAGGAACACAAACAGCAAAGTTAAAGGTAACAAAGGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTATGGAAGCAAGGCATATGTAAAGCACCCAGCAGACATCCCAGACTATTTAAAGTTAAGCTTCCCAGAAGGATTCAAGTGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTATAAGGTAAAGTTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAGAAAACAATGGGATGGGAAGCAAGCTCAGAACGTATGTATCCAGAAGACGGAGCATTAAAGGGAGAAATCAAGCAACGTTTAAAGTTAAAGGACGGAGGCCACTATGACGCAGAAGTAAAGACAACTTATAAGGCAAAGAAACCAGTACAATTACCAGGAGCATATAACGTAAACATCAAGTTAGACATCACAAGCCACAACGAAGACTATACAATCGTAGAACAATATGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATATAAGTAA', 'average_distance_score': 3.380712579917882, 'weakest_link_score': 3.380712579917882, 'ratio_score': 1.1494500671677315, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTCATGCGTTTCAAGGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATATGAAGGAACACAAACAGCAAAGTTAAAGGTAACAAAGGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTATGGAAGCAAGGCATATGTAAAGCACCCAGCAGACATCCCAGACTATTTAAAGTTAAGCTTCCCAGAAGGATTCAAGTGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTATAAGGTAAAGTTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAGAAAACAATGGGATGGGAAGCAAGCTCAGAACGTATGTATCCAGAAGACGGAGCATTAAAGGGAGAAATCAAGCAACGTTTAAAGTTAAAGGACGGAGGCCACTATGACGCAGAAGTAAAGACAACTTATAAGGCAAAGAAACCAGTACAATTACCAGGAGCATATAACGTAAACATCAAGTTAGACATCACAAGCCACAACGAAGACTATACAATCGTAGAACAATATGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATATAAGTAA'}], 'final_evaluation': {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATATGAAGGAACACAAACAGCAAAGTTAAAGGTAACAAAGGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTATGGAAGCAAGGCATATGTAAAGCACCCAGCAGACATCCCAGACTATTTAAAGTTAAGCTTCCCAGAAGGATTCAAGTGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTATAAGGTAAAGTTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAGAAAACAATGGGATGGGAAGCAAGCTCAGAACGTATGTATCCAGAAGACGGAGCATTAAAGGGAGAAATCAAGCAACGTTTAAAGTTAAAGGACGGAGGCCACTATGACGCAGAAGTAAAGACAACTTATAAGGCAAAGAAACCAGTACAATTACCAGGAGCATATAACGTAAACATCAAGTTAGACATCACAAGCCACAACGAAGACTATACAATCGTAGAACAATATGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATATAAGTAA', 'average_distance_score': 3.380712579917882, 'weakest_link_score': 3.380712579917882, 'ratio_score': 1.1494500671677315}, 'zip_output_file_path': 'results\\\\mcherry-wanted-constraint-init-external-0.1\\\\tAI_single_codon_ratio_ecoli_opt_False_2lJj\\\\communique_results.zip'}\n",
      "\n",
      "##########################\n",
      "# User Input #\n",
      "##########################\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is:\n",
      "ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "------------------------------------------\n",
      "Parsing information for Escherichia coli:\n",
      "------------------------------------------\n",
      "Organism is defined as wanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 90 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3893\n",
      "Calculate CAI weights from a reference set of 584 highly expressed genes from estimated expression dictionary.\n",
      "cai_std=0.09377868043841883, cai_avg=0.6151304741138522\n",
      "------------------------------------------\n",
      "Parsing information for Bacillus subtilis:\n",
      "------------------------------------------\n",
      "Organism is defined as unwanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 98 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3544\n",
      "Calculate CAI weights from a reference set of 532 highly expressed genes from estimated expression dictionary.\n",
      "cai_std=0.042802002292947894, cai_avg=0.7334925635409068\n",
      "-----------------------------\n",
      "Normalized Prioritization Weights\n",
      "-----------------------------\n",
      "Escherichia coli : 1.0\n",
      "Bacillus subtilis : 1.0\n",
      "Total input processing time: 27.130823612213135\n",
      "\n",
      "##########################\n",
      "# Initiation #\n",
      "##########################\n",
      "Running initiation optimization InitiationOptimizationMethod.external_module on 15 codons at the start of the ORF\n",
      "Taking optimized sequence from configuration value config.INITIATION.EXTERNAL_INITIATION_ORF\n",
      "Optimized sequence is: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "Optimization method used is: ORFOptimizationMethod.single_codon_diff\n",
      "Loss summary for: TGT\n",
      "loss for Escherichia coli: 0.07945993573726945\n",
      "loss for Bacillus subtilis: 0.535138381624868\n",
      "Loss summary for: TGC\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGT': 0.3072991586810687, 'TGC': 0.5})\n",
      "Loss summary for: GAT\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: GAC\n",
      "loss for Escherichia coli: 0.023017662416012163\n",
      "loss for Bacillus subtilis: 0.45912517084291693\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAT': 0.5, 'GAC': 0.24107141662946455})\n",
      "Loss summary for: TCT\n",
      "loss for Escherichia coli: 0.012469911477954515\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: TCG\n",
      "loss for Escherichia coli: 0.34973545889795504\n",
      "loss for Bacillus subtilis: 0.06406309095863744\n",
      "Loss summary for: TCA\n",
      "loss for Escherichia coli: 0.5108871428319045\n",
      "loss for Bacillus subtilis: 0.8564298702199061\n",
      "Loss summary for: TCC\n",
      "loss for Escherichia coli: 0.043206951482810306\n",
      "loss for Bacillus subtilis: 0.18134755214850679\n",
      "Loss summary for: AGC\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.7836723298021964\n",
      "Loss summary for: AGT\n",
      "loss for Escherichia coli: 0.4989995418830652\n",
      "loss for Bacillus subtilis: 0.10095229498733223\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TCT': 0.5062349557389773, 'TCG': 0.20689927492829624, 'TCA': 0.6836585065259053, 'TCC': 0.11227725181565855, 'AGC': 0.3918361649010982, 'AGT': 0.2999759184351987})\n",
      "Loss summary for: CAA\n",
      "loss for Escherichia coli: 0.4521316282979507\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: CAG\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.7326840893591055\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAA': 0.7260658141489753, 'CAG': 0.3663420446795527})\n",
      "Loss summary for: ATG\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATG': 0.5})\n",
      "Loss summary for: AAC\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: AAT\n",
      "loss for Escherichia coli: 0.37121660773510085\n",
      "loss for Bacillus subtilis: 0.6710332885913121\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAC': 0.5, 'AAT': 0.5211249481632064})\n",
      "Loss summary for: CCT\n",
      "loss for Escherichia coli: 0.6694214876033059\n",
      "loss for Bacillus subtilis: 0.5534699472738245\n",
      "Loss summary for: CCG\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: CCA\n",
      "loss for Escherichia coli: 0.5492524844614439\n",
      "loss for Bacillus subtilis: 0.22171668686137477\n",
      "Loss summary for: CCC\n",
      "loss for Escherichia coli: 0.8430578512396695\n",
      "loss for Bacillus subtilis: 0.0047562425683709865\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CCT': 0.6114457174385652, 'CCG': 0.5, 'CCA': 0.38548458566140936, 'CCC': 0.42390704690402026})\n",
      "Loss summary for: AAG\n",
      "loss for Escherichia coli: 0.5432934112152591\n",
      "loss for Bacillus subtilis: 0.12055437411068066\n",
      "Loss summary for: AAA\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAG': 0.3319238926629699, 'AAA': 0.5})\n",
      "Loss summary for: TAG\n",
      "loss for Escherichia coli: 0.9272976680384089\n",
      "loss for Bacillus subtilis: 0.011534452412551762\n",
      "Loss summary for: TGA\n",
      "loss for Escherichia coli: 0.5748216497928147\n",
      "loss for Bacillus subtilis: 0.02871366647490047\n",
      "Loss summary for: TAA\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAG': 0.4694160602254803, 'TGA': 0.3017676581338576, 'TAA': 0.5})\n",
      "Loss summary for: ACC\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.0494428691786508\n",
      "Loss summary for: ACA\n",
      "loss for Escherichia coli: 0.7603782203920928\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: ACG\n",
      "loss for Escherichia coli: 0.4316500886502308\n",
      "loss for Bacillus subtilis: 0.29285336536009104\n",
      "Loss summary for: ACT\n",
      "loss for Escherichia coli: 0.3325094090053259\n",
      "loss for Bacillus subtilis: 0.18645738158004316\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ACC': 0.0247214345893254, 'ACA': 0.8801891101960464, 'ACG': 0.36225172700516095, 'ACT': 0.2594833952926845})\n",
      "Loss summary for: TTT\n",
      "loss for Escherichia coli: 0.11855717804498121\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: TTC\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.7470518421251988\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTT': 0.5592785890224906, 'TTC': 0.3735259210625994})\n",
      "Loss summary for: GCA\n",
      "loss for Escherichia coli: 0.13662905843926143\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: GCC\n",
      "loss for Escherichia coli: 0.1198975280687969\n",
      "loss for Bacillus subtilis: 0.29225252497059134\n",
      "Loss summary for: GCG\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.9035388203643069\n",
      "Loss summary for: GCT\n",
      "loss for Escherichia coli: 0.15497461898669973\n",
      "loss for Bacillus subtilis: 0.9040337067416463\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GCA': 0.5683145292196308, 'GCC': 0.20607502651969412, 'GCG': 0.45176941018215344, 'GCT': 0.529504162864173})\n",
      "Loss summary for: GGT\n",
      "loss for Escherichia coli: 0.0004483938334032446\n",
      "loss for Bacillus subtilis: 0.36450148165000656\n",
      "Loss summary for: GGG\n",
      "loss for Escherichia coli: 0.6586233670715046\n",
      "loss for Bacillus subtilis: 0.0831475868143477\n",
      "Loss summary for: GGA\n",
      "loss for Escherichia coli: 0.780780521605895\n",
      "loss for Bacillus subtilis: 0.674996058051266\n",
      "Loss summary for: GGC\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GGT': 0.1824749377417049, 'GGG': 0.3708854769429262, 'GGA': 0.7278882898285804, 'GGC': 0.5})\n",
      "Loss summary for: ATC\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.7325983326024077\n",
      "Loss summary for: ATA\n",
      "loss for Escherichia coli: 0.9422632184422188\n",
      "loss for Bacillus subtilis: 0.010444139907487545\n",
      "Loss summary for: ATT\n",
      "loss for Escherichia coli: 0.06999227749100446\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATC': 0.36629916630120385, 'ATA': 0.47635367917485316, 'ATT': 0.5349961387455022})\n",
      "Loss summary for: TTA\n",
      "loss for Escherichia coli: 0.8203467506854787\n",
      "loss for Bacillus subtilis: 0.2999583645257771\n",
      "Loss summary for: TTG\n",
      "loss for Escherichia coli: 0.7787528469820632\n",
      "loss for Bacillus subtilis: 0.15010739450647193\n",
      "Loss summary for: CTC\n",
      "loss for Escherichia coli: 0.7629797849611256\n",
      "loss for Bacillus subtilis: 0.09226998161143687\n",
      "Loss summary for: CTT\n",
      "loss for Escherichia coli: 0.7972179985769738\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: CTG\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.44322260462529023\n",
      "Loss summary for: CTA\n",
      "loss for Escherichia coli: 0.9547595918369682\n",
      "loss for Bacillus subtilis: 0.019300738464682166\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTA': 0.5601525576056279, 'TTG': 0.46443012074426754, 'CTC': 0.42762488328628123, 'CTT': 0.8986089992884869, 'CTG': 0.22161130231264511, 'CTA': 0.4870301651508252})\n",
      "Loss summary for: CAT\n",
      "loss for Escherichia coli: 0.0847517250414883\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: CAC\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.4621909077722502\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAT': 0.5423758625207441, 'CAC': 0.2310954538861251})\n",
      "Loss summary for: CGA\n",
      "loss for Escherichia coli: 0.9349197778380693\n",
      "loss for Bacillus subtilis: 0.03582360799226344\n",
      "Loss summary for: CGC\n",
      "loss for Escherichia coli: 0.08272646372872398\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: CGG\n",
      "loss for Escherichia coli: 0.892623839162117\n",
      "loss for Bacillus subtilis: 0.07747320067531023\n",
      "Loss summary for: CGT\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.9808616863905326\n",
      "Loss summary for: AGG\n",
      "loss for Escherichia coli: 0.9805618044353801\n",
      "loss for Bacillus subtilis: 0.00848337950138504\n",
      "Loss summary for: AGA\n",
      "loss for Escherichia coli: 0.9695399206580425\n",
      "loss for Bacillus subtilis: 0.4941179272730253\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CGA': 0.4853716929151664, 'CGC': 0.541363231864362, 'CGG': 0.4850485199187136, 'CGT': 0.4904308431952663, 'AGG': 0.49452259196838255, 'AGA': 0.7318289239655339})\n",
      "Loss summary for: TGG\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGG': 0.5})\n",
      "Loss summary for: GTA\n",
      "loss for Escherichia coli: 0.23857489466900444\n",
      "loss for Bacillus subtilis: 0.521749240157781\n",
      "Loss summary for: GTC\n",
      "loss for Escherichia coli: 0.25309949973703594\n",
      "loss for Bacillus subtilis: 0.4810151121223792\n",
      "Loss summary for: GTG\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.4487933019998287\n",
      "Loss summary for: GTT\n",
      "loss for Escherichia coli: 0.00034372090637751606\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GTA': 0.38016206741339276, 'GTC': 0.36705730592970753, 'GTG': 0.22439665099991435, 'GTT': 0.5001718604531887})\n",
      "Loss summary for: GAG\n",
      "loss for Escherichia coli: 0.40798881274999493\n",
      "loss for Bacillus subtilis: 0.15439743545092996\n",
      "Loss summary for: GAA\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAG': 0.28119312410046243, 'GAA': 0.5})\n",
      "Loss summary for: TAT\n",
      "loss for Escherichia coli: 0.021813017751479286\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss summary for: TAC\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.6135737032126173\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAT': 0.5109065088757396, 'TAC': 0.30678685160630864})\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4262385670731707]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.09749348521075679]\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.06388523224887838]\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.38120248233619]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4262385670731707]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4292446655321377]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5512262248110238]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4292446655321377]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4262385670731707]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4205725148436632]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.09749348521075679]\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.06388523224887838]\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.38120248233619]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19071371183717192]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19071371183717192]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19071371183717192]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4262385670731707]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.37314335931309695]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4205725148436632]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19071371183717192]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4262385670731707]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4292446655321377]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19071371183717192]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.37314335931309695]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4205725148436632]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19071371183717192]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.37314335931309695]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4262385670731707]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19071371183717192]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4262385670731707]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.09749348521075679]\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.06388523224887838]\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.38120248233619]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5512262248110238]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4262385670731707]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.37314335931309695]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.37314335931309695]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15406749964830613]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.37314335931309695]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4262385670731707]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4205725148436632]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.09749348521075679]\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.06388523224887838]\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.38120248233619]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5512262248110238]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4262385670731707]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19071371183717192]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.37314335931309695]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19071371183717192]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15406749964830613]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.09749348521075679]\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.06388523224887838]\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.38120248233619]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19071371183717192]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.37314335931309695]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4205725148436632]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.09749348521075679]\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.06388523224887838]\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.38120248233619]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.37314335931309695]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4292446655321377]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.37314335931309695]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.16532540959927366]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19071371183717192]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19071371183717192]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5512262248110238]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5512262248110238]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4205725148436632]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.37314335931309695]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4205725148436632]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4292446655321377]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5512262248110238]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.37314335931309695]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4205725148436632]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.09749348521075679]\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.06388523224887838]\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.38120248233619]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.09749348521075679]\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.06388523224887838]\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.38120248233619]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4292446655321377]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14895195222078703]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.37314335931309695]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4315172054223149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.28926761860893596]\n",
      "\n",
      "##########################\n",
      "# EVALUATION #\n",
      "##########################\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTCATGCGTTTCAAGGTGCACATGGAGGGTTCCGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCATACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCACTGCCATTCGCCTGGGACATCCTGTCCCCACAGTTCATGTACGGTTCCAAGGCCTACGTGAAGCACCCAGCCGACATCCCAGACTACCTGAAGCTGTCCTTCCCAGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTCACCGTGACCCAGGACTCCTCGCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCATCCGACGGTCCAGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCCTCGGAGCGTATGTACCCAGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCAGTGCAGCTGCCAGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCCCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCCACCGGTGGGATGGACGAGCTGTACAAGTGA\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGTTCCGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCATACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCACTGCCATTCGCCTGGGACATCCTGTCCCCACAGTTCATGTACGGTTCCAAGGCCTACGTGAAGCACCCAGCCGACATCCCAGACTACCTGAAGCTGTCCTTCCCAGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTCACCGTGACCCAGGACTCCTCGCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCATCCGACGGTCCAGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCCTCGGAGCGTATGTACCCAGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCAGTGCAGCTGCCAGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCCCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCCACCGGTGGGATGGACGAGCTGTACAAGTGA', 'average_distance_score': 1.9034506402682547, 'weakest_link_score': 1.9034506402682547, 'ratio_score': 1.0890597067107513}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGTTCCGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCATACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCACTGCCATTCGCCTGGGACATCCTGTCCCCACAGTTCATGTACGGTTCCAAGGCCTACGTGAAGCACCCAGCCGACATCCCAGACTACCTGAAGCTGTCCTTCCCAGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTCACCGTGACCCAGGACTCCTCGCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCATCCGACGGTCCAGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCCTCGGAGCGTATGTACCCAGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCAGTGCAGCTGCCAGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCCCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCCACCGGTGGGATGGACGAGCTGTACAAGTGA', 'average_distance_score': 1.9034506402682547, 'weakest_link_score': 1.9034506402682547, 'ratio_score': 1.0890597067107513}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry-wanted-constraint-init-external-0.1\\CAI_single_codon_diff_ecoli_opt_True_3AYt\n",
      "Final output: {'module_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'orf_optimization_method': 'single_codon_diff', 'orf_optimization_cub_index': 'CAI', 'initiation_optimization_method': 'external', 'evaluation_score': 'average_distance', 'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.037037037037037035, 'TGA': 0.24183006535947713, 'GCA': 0.6303663185811371, 'GCC': 0.653737775567711, 'GCG': 1.0, 'GCT': 0.6063318415382065, 'TGC': 1.0, 'TGT': 0.7181136120042874, 'GAC': 0.848284271032921, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.36125998031280765, 'TTC': 1.0, 'TTT': 0.6556786703601108, 'GGA': 0.11638214051214708, 'GGC': 1.0, 'GGG': 0.18844386080105055, 'GGT': 0.9788246881155614, 'CAC': 1.0, 'CAT': 0.7088785046728973, 'ATA': 0.029297564419343447, 'ATC': 1.0, 'ATT': 0.7354394634662901, 'AAA': 1.0, 'AAG': 0.26291560102301786, 'CTA': 0.02288199697428139, 'CTC': 0.1265128593040847, 'CTG': 1.0, 'CTT': 0.10712934947049924, 'TTA': 0.09427004538577911, 'TTG': 0.11753025718608169, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.39072452229299365, 'CCA': 0.2588842975206611, 'CCC': 0.08181818181818182, 'CCG': 1.0, 'CCT': 0.1818181818181818, 'CAA': 0.32759266192437286, 'CAG': 1.0, 'AGA': 0.015347817420769383, 'AGG': 0.00976679290412597, 'CGA': 0.033087502491528795, 'CGC': 0.7123779150886984, 'CGG': 0.05521227825393661, 'CGT': 1.0, 'AGC': 1.0, 'AGT': 0.2936010037641154, 'TCA': 0.28523630280217477, 'TCC': 0.7921371810957758, 'TCG': 0.4086156419907988, 'TCT': 0.8883312421580928, 'ACA': 0.12800331400165701, 'ACC': 1.0, 'ACG': 0.34299917149958575, 'ACT': 0.4233637116818558, 'GTA': 0.5115587090867476, 'GTC': 0.49691004806591893, 'GTG': 1.0, 'GTT': 0.9814602883955139, 'TGG': 1.0, 'TAC': 1.0, 'TAT': 0.8523076923076923}, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'AAA': 0.7667390608953314, 'CGC': 0.40076028262096014, 'ATT': 0.5085852538544149, 'AGC': 0.27799163607995603, 'ACC': 0.4359429381095841, 'ATC': 0.4205725148436632, 'ACA': 0.13023262212317555, 'GGT': 0.33726003930064996, 'AAC': 0.5512262248110238, 'GCG': 0.3563679190097728, 'GGC': 0.4050486219579785, 'TGA': 0.28926761860893596, 'CGA': 0.06388523224887838, 'GTG': 0.3715940695510341, 'TTG': 0.12808970030428077, 'AAG': 0.23326093910466858, 'TTC': 0.4262385670731707, 'TCA': 0.12251736088908215, 'GCA': 0.21281773299119433, 'AAT': 0.4487737751889762, 'GAA': 0.689467997530356, 'CGT': 0.38120248233619, 'TTT': 0.5737614329268292, 'CTG': 0.4975546385450099, 'GTT': 0.2584847661069405, 'GCC': 0.26990242799092135, 'GAT': 0.626856640686903, 'AGG': 0.020162265266098916, 'CAG': 0.6530527088827319, 'GGG': 0.15057187484254547, 'GTC': 0.21619163129169194, 'CTC': 0.10435163186194824, 'TCT': 0.14544780223293646, 'CCC': 0.12361073915372255, 'CAC': 0.4292446655321377, 'GCT': 0.1609119200081115, 'TTA': 0.13009739763522432, 'CTT': 0.10341378016756284, 'ACG': 0.26849903016796667, 'GGA': 0.10711946389882603, 'CCG': 0.527785216925549, 'CAA': 0.3469472911172681, 'ACT': 0.16532540959927366, 'ATA': 0.07084223130192197, 'CAT': 0.5707553344678623, 'AGT': 0.15102374892893225, 'TGC': 0.55651396076928, 'GAG': 0.31053200246964396, 'TCG': 0.15406749964830613, 'GTA': 0.15372953305033354, 'TAC': 0.4315172054223149, 'TCC': 0.14895195222078703, 'GAC': 0.37314335931309695, 'TGT': 0.4434860392307201, 'TGG': 1.0, 'TAT': 0.5684827945776851, 'CCT': 0.15789033208355652, 'CCA': 0.19071371183717192, 'CGG': 0.09749348521075679, 'CTA': 0.03649285148597391, 'TAA': 0.6391064025794565, 'AGA': 0.036496252317115764, 'TAG': 0.07162597881160755}), 'cai_avg': 0.6151304741138522, 'cai_std': 0.09377868043841883, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.10739856801909307, 'TGA': 0.1694510739856802, 'GCA': 1.0, 'GCC': 0.5406038521603331, 'GCG': 0.950546590317543, 'GCT': 0.9508068714211347, 'TGC': 1.0, 'TGT': 0.7315315315315315, 'GAC': 0.6775877587758776, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.3929343907714492, 'TTC': 0.864321608040201, 'TTT': 1.0, 'GGA': 0.8215814372606443, 'GGC': 1.0, 'GGG': 0.28835323271006985, 'GGT': 0.6037395809867087, 'CAC': 0.6798462383305877, 'CAT': 1.0, 'ATA': 0.10219657483246464, 'ATC': 0.8559195830230827, 'ATT': 1.0, 'AAA': 1.0, 'AAG': 0.3472094095940959, 'CTA': 0.13892709766162312, 'CTC': 0.30375974323704724, 'CTG': 0.6657496561210454, 'CTT': 1.0, 'TTA': 0.5476845483723063, 'TTG': 0.3874369555249885, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.8191662154845695, 'CCA': 0.4708680142687277, 'CCC': 0.06896551724137931, 'CCG': 1.0, 'CCT': 0.7439556084026953, 'CAA': 1.0, 'CAG': 0.8559696778269108, 'AGA': 0.7029352226720648, 'AGG': 0.09210526315789473, 'CGA': 0.18927125506072875, 'CGC': 1.0, 'CGG': 0.27834008097165996, 'CGT': 0.9903846153846154, 'AGC': 0.8852526926263463, 'AGT': 0.3177299088649544, 'TCA': 0.9254349627174814, 'TCC': 0.4258492129246064, 'TCG': 0.2531068765534383, 'TCT': 1.0, 'ACA': 1.0, 'ACC': 0.22235752557233318, 'ACG': 0.5411592791037506, 'ACT': 0.43180711154408186, 'GTA': 0.7223221166195737, 'GTC': 0.6935525301823786, 'GTG': 0.6699203698946828, 'GTT': 1.0, 'TGG': 1.0, 'TAC': 0.7833094555873925, 'TAT': 1.0}, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'GAA': 0.6798186496752849, 'AAT': 0.5655317497600131, 'ATA': 0.1287010322849589, 'TTA': 0.198530453110291, 'GAC': 0.35960560391899904, 'CTG': 0.23973441522197322, 'TGG': 1.0, 'AAC': 0.4344682502399869, 'CAA': 0.5131548459124133, 'GCC': 0.20715414603308396, 'CTT': 0.2391679232236727, 'GCT': 0.24752923822568748, 'ATC': 0.366762837741479, 'AAA': 0.7016557540022625, 'AAG': 0.29834424599773746, 'TTG': 0.1592175745811708, 'AGC': 0.22534685577492342, 'CCG': 0.4334091260292722, 'AGT': 0.10611598754150686, 'TTT': 0.6847751982489505, 'GAG': 0.3201813503247151, 'ACT': 0.1614000537105004, 'TCA': 0.23653118484388272, 'ACC': 0.1588786441082565, 'CAC': 0.3265866087018494, 'GGC': 0.3390368073029434, 'GAT': 0.640394396081001, 'ACA': 0.4122430101751559, 'ACG': 0.2674782920060872, 'CCC': 0.0889644363010221, 'AGA': 0.26380392311639106, 'TCC': 0.12712038919920718, 'TAC': 0.34586274736720285, 'CAT': 0.6734133912981506, 'ATT': 0.5045361299735621, 'GCA': 0.2828363712991255, 'TAT': 0.6541372526327971, 'GGG': 0.16211598104043537, 'GTC': 0.25598186049286953, 'CCT': 0.28667299498885185, 'GTT': 0.28439644370189154, 'GCG': 0.26248024444210305, 'TCT': 0.2048572678833432, 'CTC': 0.11219040795754642, 'CCA': 0.19095344268085387, 'GGA': 0.3148341038094681, 'CGA': 0.09855152543710778, 'GTA': 0.19774449549495793, 'GTG': 0.26187720031028106, 'TTC': 0.31522480175104955, 'CGC': 0.2064569601768165, 'CAG': 0.4868451540875866, 'CGG': 0.1555630106168844, 'CGT': 0.18161187196589967, 'CTA': 0.05115922590534585, 'AGG': 0.09401270868690058, 'GGT': 0.18401310784715313, 'TAG': 0.14132712456344587, 'TGC': 0.5460141582025239, 'TCG': 0.10002831475713661, 'TAA': 0.630500582072177, 'TGA': 0.2281722933643772, 'TGT': 0.45398584179747614}), 'cai_avg': 0.7334925635409068, 'cai_std': 0.042802002292947894, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}]}, 'orf_debug': {'total_loss': {'C': {'TGT': 0.3072991586810687, 'TGC': 0.5}, 'D': {'GAC': 0.24107141662946455, 'GAT': 0.5}, 'S': {'TCC': 0.11227725181565855, 'TCG': 0.20689927492829624, 'AGT': 0.2999759184351987, 'AGC': 0.3918361649010982, 'TCT': 0.5062349557389773, 'TCA': 0.6836585065259053}, 'Q': {'CAG': 0.3663420446795527, 'CAA': 0.7260658141489753}, 'M': {'ATG': 0.5}, 'N': {'AAC': 0.5, 'AAT': 0.5211249481632064}, 'P': {'CCA': 0.38548458566140936, 'CCC': 0.42390704690402026, 'CCG': 0.5, 'CCT': 0.6114457174385652}, 'K': {'AAG': 0.3319238926629699, 'AAA': 0.5}, '_': {'TGA': 0.3017676581338576, 'TAG': 0.4694160602254803, 'TAA': 0.5}, 'T': {'ACC': 0.0247214345893254, 'ACT': 0.2594833952926845, 'ACG': 0.36225172700516095, 'ACA': 0.8801891101960464}, 'F': {'TTC': 0.3735259210625994, 'TTT': 0.5592785890224906}, 'A': {'GCC': 0.20607502651969412, 'GCG': 0.45176941018215344, 'GCT': 0.529504162864173, 'GCA': 0.5683145292196308}, 'G': {'GGT': 0.1824749377417049, 'GGG': 0.3708854769429262, 'GGC': 0.5, 'GGA': 0.7278882898285804}, 'I': {'ATC': 0.36629916630120385, 'ATA': 0.47635367917485316, 'ATT': 0.5349961387455022}, 'L': {'CTG': 0.22161130231264511, 'CTC': 0.42762488328628123, 'TTG': 0.46443012074426754, 'CTA': 0.4870301651508252, 'TTA': 0.5601525576056279, 'CTT': 0.8986089992884869}, 'H': {'CAC': 0.2310954538861251, 'CAT': 0.5423758625207441}, 'R': {'CGG': 0.4850485199187136, 'CGA': 0.4853716929151664, 'CGT': 0.4904308431952663, 'AGG': 0.49452259196838255, 'CGC': 0.541363231864362, 'AGA': 0.7318289239655339}, 'W': {'TGG': 0.5}, 'V': {'GTG': 0.22439665099991435, 'GTC': 0.36705730592970753, 'GTA': 0.38016206741339276, 'GTT': 0.5001718604531887}, 'E': {'GAG': 0.28119312410046243, 'GAA': 0.5}, 'Y': {'TAC': 0.30678685160630864, 'TAT': 0.5109065088757396}}, 'optimized_loss': {'C': defaultdict(<class 'float'>, {'TGT': 0.07945993573726945, 'TGC': 0.0}), 'D': defaultdict(<class 'float'>, {'GAT': 0.0, 'GAC': 0.023017662416012163}), 'S': defaultdict(<class 'float'>, {'TCT': 0.012469911477954515, 'TCG': 0.34973545889795504, 'TCA': 0.5108871428319045, 'TCC': 0.043206951482810306, 'AGC': 0.0, 'AGT': 0.4989995418830652}), 'Q': defaultdict(<class 'float'>, {'CAA': 0.4521316282979507, 'CAG': 0.0}), 'M': defaultdict(<class 'float'>, {'ATG': 0.0}), 'N': defaultdict(<class 'float'>, {'AAC': 0.0, 'AAT': 0.37121660773510085}), 'P': defaultdict(<class 'float'>, {'CCT': 0.6694214876033059, 'CCG': 0.0, 'CCA': 0.5492524844614439, 'CCC': 0.8430578512396695}), 'K': defaultdict(<class 'float'>, {'AAG': 0.5432934112152591, 'AAA': 0.0}), '_': defaultdict(<class 'float'>, {'TAG': 0.9272976680384089, 'TGA': 0.5748216497928147, 'TAA': 0.0}), 'T': defaultdict(<class 'float'>, {'ACC': 0.0, 'ACA': 0.7603782203920928, 'ACG': 0.4316500886502308, 'ACT': 0.3325094090053259}), 'F': defaultdict(<class 'float'>, {'TTT': 0.11855717804498121, 'TTC': 0.0}), 'A': defaultdict(<class 'float'>, {'GCA': 0.13662905843926143, 'GCC': 0.1198975280687969, 'GCG': 0.0, 'GCT': 0.15497461898669973}), 'G': defaultdict(<class 'float'>, {'GGT': 0.0004483938334032446, 'GGG': 0.6586233670715046, 'GGA': 0.780780521605895, 'GGC': 0.0}), 'I': defaultdict(<class 'float'>, {'ATC': 0.0, 'ATA': 0.9422632184422188, 'ATT': 0.06999227749100446}), 'L': defaultdict(<class 'float'>, {'TTA': 0.8203467506854787, 'TTG': 0.7787528469820632, 'CTC': 0.7629797849611256, 'CTT': 0.7972179985769738, 'CTG': 0.0, 'CTA': 0.9547595918369682}), 'H': defaultdict(<class 'float'>, {'CAT': 0.0847517250414883, 'CAC': 0.0}), 'R': defaultdict(<class 'float'>, {'CGA': 0.9349197778380693, 'CGC': 0.08272646372872398, 'CGG': 0.892623839162117, 'CGT': 0.0, 'AGG': 0.9805618044353801, 'AGA': 0.9695399206580425}), 'W': defaultdict(<class 'float'>, {'TGG': 0.0}), 'V': defaultdict(<class 'float'>, {'GTA': 0.23857489466900444, 'GTC': 0.25309949973703594, 'GTG': 0.0, 'GTT': 0.00034372090637751606}), 'E': defaultdict(<class 'float'>, {'GAG': 0.40798881274999493, 'GAA': 0.0}), 'Y': defaultdict(<class 'float'>, {'TAT': 0.021813017751479286, 'TAC': 0.0})}, 'deoptimized_loss': {'C': defaultdict(<class 'float'>, {'TGT': 0.535138381624868, 'TGC': 1.0}), 'D': defaultdict(<class 'float'>, {'GAT': 1.0, 'GAC': 0.45912517084291693}), 'S': defaultdict(<class 'float'>, {'TCT': 1.0, 'TCG': 0.06406309095863744, 'TCA': 0.8564298702199061, 'TCC': 0.18134755214850679, 'AGC': 0.7836723298021964, 'AGT': 0.10095229498733223}), 'Q': defaultdict(<class 'float'>, {'CAA': 1.0, 'CAG': 0.7326840893591055}), 'M': defaultdict(<class 'float'>, {'ATG': 1.0}), 'N': defaultdict(<class 'float'>, {'AAC': 1.0, 'AAT': 0.6710332885913121}), 'P': defaultdict(<class 'float'>, {'CCT': 0.5534699472738245, 'CCG': 1.0, 'CCA': 0.22171668686137477, 'CCC': 0.0047562425683709865}), 'K': defaultdict(<class 'float'>, {'AAG': 0.12055437411068066, 'AAA': 1.0}), '_': defaultdict(<class 'float'>, {'TAG': 0.011534452412551762, 'TGA': 0.02871366647490047, 'TAA': 1.0}), 'T': defaultdict(<class 'float'>, {'ACC': 0.0494428691786508, 'ACA': 1.0, 'ACG': 0.29285336536009104, 'ACT': 0.18645738158004316}), 'F': defaultdict(<class 'float'>, {'TTT': 1.0, 'TTC': 0.7470518421251988}), 'A': defaultdict(<class 'float'>, {'GCA': 1.0, 'GCC': 0.29225252497059134, 'GCG': 0.9035388203643069, 'GCT': 0.9040337067416463}), 'G': defaultdict(<class 'float'>, {'GGT': 0.36450148165000656, 'GGG': 0.0831475868143477, 'GGA': 0.674996058051266, 'GGC': 1.0}), 'I': defaultdict(<class 'float'>, {'ATC': 0.7325983326024077, 'ATA': 0.010444139907487545, 'ATT': 1.0}), 'L': defaultdict(<class 'float'>, {'TTA': 0.2999583645257771, 'TTG': 0.15010739450647193, 'CTC': 0.09226998161143687, 'CTT': 1.0, 'CTG': 0.44322260462529023, 'CTA': 0.019300738464682166}), 'H': defaultdict(<class 'float'>, {'CAT': 1.0, 'CAC': 0.4621909077722502}), 'R': defaultdict(<class 'float'>, {'CGA': 0.03582360799226344, 'CGC': 1.0, 'CGG': 0.07747320067531023, 'CGT': 0.9808616863905326, 'AGG': 0.00848337950138504, 'AGA': 0.4941179272730253}), 'W': defaultdict(<class 'float'>, {'TGG': 1.0}), 'V': defaultdict(<class 'float'>, {'GTA': 0.521749240157781, 'GTC': 0.4810151121223792, 'GTG': 0.4487933019998287, 'GTT': 1.0}), 'E': defaultdict(<class 'float'>, {'GAG': 0.15439743545092996, 'GAA': 1.0}), 'Y': defaultdict(<class 'float'>, {'TAT': 1.0, 'TAC': 0.6135737032126173})}}, 'orf': {'orf_module_input_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'optimized_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGTTCCGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCATACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCACTGCCATTCGCCTGGGACATCCTGTCCCCACAGTTCATGTACGGTTCCAAGGCCTACGTGAAGCACCCAGCCGACATCCCAGACTACCTGAAGCTGTCCTTCCCAGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTCACCGTGACCCAGGACTCCTCGCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCATCCGACGGTCCAGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCCTCGGAGCGTATGTACCCAGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCAGTGCAGCTGCCAGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCCCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCCACCGGTGGGATGGACGAGCTGTACAAGTGA', 'aa_to_loss_mapping': {'C': {'TGT': 0.3072991586810687, 'TGC': 0.5}, 'D': {'GAC': 0.24107141662946455, 'GAT': 0.5}, 'S': {'TCC': 0.11227725181565855, 'TCG': 0.20689927492829624, 'AGT': 0.2999759184351987, 'AGC': 0.3918361649010982, 'TCT': 0.5062349557389773, 'TCA': 0.6836585065259053}, 'Q': {'CAG': 0.3663420446795527, 'CAA': 0.7260658141489753}, 'M': {'ATG': 0.5}, 'N': {'AAC': 0.5, 'AAT': 0.5211249481632064}, 'P': {'CCA': 0.38548458566140936, 'CCC': 0.42390704690402026, 'CCG': 0.5, 'CCT': 0.6114457174385652}, 'K': {'AAG': 0.3319238926629699, 'AAA': 0.5}, '_': {'TGA': 0.3017676581338576, 'TAG': 0.4694160602254803, 'TAA': 0.5}, 'T': {'ACC': 0.0247214345893254, 'ACT': 0.2594833952926845, 'ACG': 0.36225172700516095, 'ACA': 0.8801891101960464}, 'F': {'TTC': 0.3735259210625994, 'TTT': 0.5592785890224906}, 'A': {'GCC': 0.20607502651969412, 'GCG': 0.45176941018215344, 'GCT': 0.529504162864173, 'GCA': 0.5683145292196308}, 'G': {'GGT': 0.1824749377417049, 'GGG': 0.3708854769429262, 'GGC': 0.5, 'GGA': 0.7278882898285804}, 'I': {'ATC': 0.36629916630120385, 'ATA': 0.47635367917485316, 'ATT': 0.5349961387455022}, 'L': {'CTG': 0.22161130231264511, 'CTC': 0.42762488328628123, 'TTG': 0.46443012074426754, 'CTA': 0.4870301651508252, 'TTA': 0.5601525576056279, 'CTT': 0.8986089992884869}, 'H': {'CAC': 0.2310954538861251, 'CAT': 0.5423758625207441}, 'R': {'CGG': 0.4850485199187136, 'CGA': 0.4853716929151664, 'CGT': 0.4904308431952663, 'AGG': 0.49452259196838255, 'CGC': 0.541363231864362, 'AGA': 0.7318289239655339}, 'W': {'TGG': 0.5}, 'V': {'GTG': 0.22439665099991435, 'GTC': 0.36705730592970753, 'GTA': 0.38016206741339276, 'GTT': 0.5001718604531887}, 'E': {'GAG': 0.28119312410046243, 'GAA': 0.5}, 'Y': {'TAC': 0.30678685160630864, 'TAT': 0.5109065088757396}}, 'run_time': 0.75}, 'evaluation': [{'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'cai_initial_score': 0.7915994634162647, 'cai_final_score': 0.669446489947581, 'dist_score': -1.302566563078239}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'cai_initial_score': 0.7695788606254016, 'cai_final_score': 0.5508834062672596, 'dist_score': -5.109467843614748}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGTTCCGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCATACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCACTGCCATTCGCCTGGGACATCCTGTCCCCACAGTTCATGTACGGTTCCAAGGCCTACGTGAAGCACCCAGCCGACATCCCAGACTACCTGAAGCTGTCCTTCCCAGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTCACCGTGACCCAGGACTCCTCGCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCATCCGACGGTCCAGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCCTCGGAGCGTATGTACCCAGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCAGTGCAGCTGCCAGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCCCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCCACCGGTGGGATGGACGAGCTGTACAAGTGA', 'average_distance_score': 1.9034506402682547, 'weakest_link_score': 1.9034506402682547, 'ratio_score': 1.0890597067107513, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTCATGCGTTTCAAGGTGCACATGGAGGGTTCCGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCATACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCACTGCCATTCGCCTGGGACATCCTGTCCCCACAGTTCATGTACGGTTCCAAGGCCTACGTGAAGCACCCAGCCGACATCCCAGACTACCTGAAGCTGTCCTTCCCAGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTCACCGTGACCCAGGACTCCTCGCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCATCCGACGGTCCAGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCCTCGGAGCGTATGTACCCAGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCAGTGCAGCTGCCAGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCCCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCCACCGGTGGGATGGACGAGCTGTACAAGTGA'}], 'final_evaluation': {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGTTCCGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCATACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCACTGCCATTCGCCTGGGACATCCTGTCCCCACAGTTCATGTACGGTTCCAAGGCCTACGTGAAGCACCCAGCCGACATCCCAGACTACCTGAAGCTGTCCTTCCCAGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTCACCGTGACCCAGGACTCCTCGCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCATCCGACGGTCCAGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCCTCGGAGCGTATGTACCCAGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCAGTGCAGCTGCCAGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCCCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCCACCGGTGGGATGGACGAGCTGTACAAGTGA', 'average_distance_score': 1.9034506402682547, 'weakest_link_score': 1.9034506402682547, 'ratio_score': 1.0890597067107513}, 'zip_output_file_path': 'results\\\\mcherry-wanted-constraint-init-external-0.1\\\\CAI_single_codon_diff_ecoli_opt_True_3AYt\\\\communique_results.zip'}\n",
      "\n",
      "##########################\n",
      "# User Input #\n",
      "##########################\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is:\n",
      "ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "------------------------------------------\n",
      "Parsing information for Bacillus subtilis:\n",
      "------------------------------------------\n",
      "Organism is defined as wanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 98 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3544\n",
      "Calculate CAI weights from a reference set of 532 highly expressed genes from estimated expression dictionary.\n",
      "cai_std=0.042802002292947894, cai_avg=0.7334925635409068\n",
      "------------------------------------------\n",
      "Parsing information for Escherichia coli:\n",
      "------------------------------------------\n",
      "Organism is defined as unwanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 90 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3893\n",
      "Calculate CAI weights from a reference set of 584 highly expressed genes from estimated expression dictionary.\n",
      "cai_std=0.09377868043841883, cai_avg=0.6151304741138522\n",
      "-----------------------------\n",
      "Normalized Prioritization Weights\n",
      "-----------------------------\n",
      "Bacillus subtilis : 1.0\n",
      "Escherichia coli : 1.0\n",
      "Total input processing time: 27.31287980079651\n",
      "\n",
      "##########################\n",
      "# Initiation #\n",
      "##########################\n",
      "Running initiation optimization InitiationOptimizationMethod.external_module on 15 codons at the start of the ORF\n",
      "Taking optimized sequence from configuration value config.INITIATION.EXTERNAL_INITIATION_ORF\n",
      "Optimized sequence is: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "Optimization method used is: ORFOptimizationMethod.single_codon_diff\n",
      "Loss summary for: TGT\n",
      "loss for Bacillus subtilis: 0.07207531856180506\n",
      "loss for Escherichia coli: 0.5156871597458441\n",
      "Loss summary for: TGC\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGT': 0.2938812391538246, 'TGC': 0.5})\n",
      "Loss summary for: GAT\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: GAC\n",
      "loss for Bacillus subtilis: 0.10394965329116168\n",
      "loss for Escherichia coli: 0.7195862044818543\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAT': 0.5, 'GAC': 0.411767928886508})\n",
      "Loss summary for: TCT\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.7891323957941401\n",
      "Loss summary for: TCG\n",
      "loss for Bacillus subtilis: 0.5578493378517608\n",
      "loss for Escherichia coli: 0.16696674287955265\n",
      "Loss summary for: TCA\n",
      "loss for Bacillus subtilis: 0.005559944784943392\n",
      "loss for Escherichia coli: 0.08135974843625393\n",
      "Loss summary for: TCC\n",
      "loss for Bacillus subtilis: 0.32964912629929394\n",
      "loss for Escherichia coli: 0.6274813136743619\n",
      "Loss summary for: AGC\n",
      "loss for Bacillus subtilis: 0.013166944549503757\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: AGT\n",
      "loss for Bacillus subtilis: 0.46549247725742354\n",
      "loss for Escherichia coli: 0.08620154941129611\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TCT': 0.39456619789707004, 'TCG': 0.36240804036565677, 'TCA': 0.043459846610598665, 'TCC': 0.4785652199868279, 'AGC': 0.5065834722747519, 'AGT': 0.27584701333435985})\n",
      "Loss summary for: CAA\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.10731695214669645\n",
      "Loss summary for: CAG\n",
      "loss for Bacillus subtilis: 0.02074473370528387\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAA': 0.05365847607334823, 'CAG': 0.510372366852642})\n",
      "Loss summary for: ATG\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATG': 0.5})\n",
      "Loss summary for: AAC\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: AAT\n",
      "loss for Bacillus subtilis: 0.03270085762217316\n",
      "loss for Escherichia coli: 0.15266565232108809\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAC': 0.5, 'AAT': 0.09268325497163063})\n",
      "Loss summary for: CCT\n",
      "loss for Bacillus subtilis: 0.06555873046843394\n",
      "loss for Escherichia coli: 0.033057851239669415\n",
      "Loss summary for: CCG\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: CCA\n",
      "loss for Bacillus subtilis: 0.27998065832391933\n",
      "loss for Escherichia coli: 0.06702107950276619\n",
      "Loss summary for: CCC\n",
      "loss for Bacillus subtilis: 0.8668252080856124\n",
      "loss for Escherichia coli: 0.006694214876033058\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CCT': 0.04930829085405168, 'CCG': 0.5, 'CCA': 0.17350086891334277, 'CCC': 0.4367597114808227})\n",
      "Loss summary for: AAG\n",
      "loss for Bacillus subtilis: 0.4261355549224889\n",
      "loss for Escherichia coli: 0.06912461326129471\n",
      "Loss summary for: AAA\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAG': 0.2476300840918918, 'AAA': 0.5})\n",
      "Loss summary for: TAG\n",
      "loss for Bacillus subtilis: 0.7967373163743656\n",
      "loss for Escherichia coli: 0.0013717421124828531\n",
      "Loss summary for: TGA\n",
      "loss for Bacillus subtilis: 0.6898115185035399\n",
      "loss for Escherichia coli: 0.05848178051176898\n",
      "Loss summary for: TAA\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAG': 0.3990545292434242, 'TGA': 0.37414664950765447, 'TAA': 0.5})\n",
      "Loss summary for: ACC\n",
      "loss for Bacillus subtilis: 0.6047278180339846\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: ACA\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.0163848483954068\n",
      "Loss summary for: ACG\n",
      "loss for Bacillus subtilis: 0.21053480715258982\n",
      "loss for Escherichia coli: 0.11764843164940224\n",
      "Loss summary for: ACT\n",
      "loss for Bacillus subtilis: 0.3228431584918795\n",
      "loss for Escherichia coli: 0.17923683236903754\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ACC': 0.8023639090169923, 'ACA': 0.0081924241977034, 'ACG': 0.16409161940099604, 'ACT': 0.2510399954304585})\n",
      "Loss summary for: TTT\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.4299145187652029\n",
      "Loss summary for: TTC\n",
      "loss for Bacillus subtilis: 0.018408626044796858\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTT': 0.21495725938260146, 'TTC': 0.5092043130223984})\n",
      "Loss summary for: GCA\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.39736169560153567\n",
      "Loss summary for: GCC\n",
      "loss for Bacillus subtilis: 0.21104482064992505\n",
      "loss for Escherichia coli: 0.4273730792042188\n",
      "Loss summary for: GCG\n",
      "loss for Bacillus subtilis: 0.0024456397292209357\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: GCT\n",
      "loss for Bacillus subtilis: 0.0024199638993767703\n",
      "loss for Escherichia coli: 0.36763830206311277\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GCA': 0.19868084780076783, 'GCC': 0.31920894992707194, 'GCG': 0.5012228198646105, 'GCT': 0.18502913298124476})\n",
      "Loss summary for: GGT\n",
      "loss for Bacillus subtilis: 0.15702231967658922\n",
      "loss for Escherichia coli: 0.9580977700645261\n",
      "Loss summary for: GGG\n",
      "loss for Bacillus subtilis: 0.506441121394208\n",
      "loss for Escherichia coli: 0.035511088673605716\n",
      "Loss summary for: GGA\n",
      "loss for Bacillus subtilis: 0.031833183529977424\n",
      "loss for Escherichia coli: 0.013544802630189147\n",
      "Loss summary for: GGC\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GGT': 0.5575600448705577, 'GGG': 0.2709761050339069, 'GGA': 0.022688993080083285, 'GGC': 0.5})\n",
      "Loss summary for: ATC\n",
      "loss for Bacillus subtilis: 0.020759166556242364\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: ATA\n",
      "loss for Bacillus subtilis: 0.8060509902425582\n",
      "loss for Escherichia coli: 0.0008583472809055792\n",
      "Loss summary for: ATT\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.5408712044235846\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATC': 0.5103795832781212, 'ATA': 0.4034546687617319, 'ATT': 0.2704356022117923})\n",
      "Loss summary for: TTA\n",
      "loss for Bacillus subtilis: 0.20458926778116454\n",
      "loss for Escherichia coli: 0.008886841457036854\n",
      "Loss summary for: TTG\n",
      "loss for Bacillus subtilis: 0.3752334834564948\n",
      "loss for Escherichia coli: 0.013813361354226507\n",
      "Loss summary for: CTC\n",
      "loss for Bacillus subtilis: 0.48475049513734236\n",
      "loss for Escherichia coli: 0.016005503569295133\n",
      "Loss summary for: CTT\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.011476697517972355\n",
      "Loss summary for: CTG\n",
      "loss for Bacillus subtilis: 0.11172329238319939\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: CTA\n",
      "loss for Bacillus subtilis: 0.7414465431414359\n",
      "loss for Escherichia coli: 0.0005235857855310226\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTA': 0.1067380546191007, 'TTG': 0.19452342240536066, 'CTC': 0.2503779993533187, 'CTT': 0.005738348758986178, 'CTG': 0.5558616461915997, 'CTA': 0.37098506446348345})\n",
      "Loss summary for: CAT\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.5025087343872828\n",
      "Loss summary for: CAC\n",
      "loss for Bacillus subtilis: 0.10249843111107487\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAT': 0.2512543671936414, 'CAC': 0.5512492155555374})\n",
      "Loss summary for: CGA\n",
      "loss for Bacillus subtilis: 0.657281097870806\n",
      "loss for Escherichia coli: 0.0010947828211269241\n",
      "Loss summary for: CGC\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.5074822939061208\n",
      "Loss summary for: CGG\n",
      "loss for Bacillus subtilis: 0.5207930387319903\n",
      "loss for Escherichia coli: 0.0030483956699901217\n",
      "Loss summary for: CGT\n",
      "loss for Bacillus subtilis: 9.24556213017745e-05\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: AGG\n",
      "loss for Bacillus subtilis: 0.8242728531855956\n",
      "loss for Escherichia coli: 9.53902436320854e-05\n",
      "Loss summary for: AGA\n",
      "loss for Bacillus subtilis: 0.08824748192889574\n",
      "loss for Escherichia coli: 0.00023555549958127213\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CGA': 0.3291879403459665, 'CGC': 0.2537411469530604, 'CGG': 0.26192071720099025, 'CGT': 0.5000462278106509, 'AGG': 0.4121841217146138, 'AGA': 0.04424151871423851})\n",
      "Loss summary for: TGG\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGG': 0.5})\n",
      "Loss summary for: GTA\n",
      "loss for Bacillus subtilis: 0.07710500691863363\n",
      "loss for Escherichia coli: 0.26169231284249966\n",
      "Loss summary for: GTC\n",
      "loss for Bacillus subtilis: 0.09391005175762197\n",
      "loss for Escherichia coli: 0.24691959586887385\n",
      "Loss summary for: GTG\n",
      "loss for Bacillus subtilis: 0.108952562210463\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss summary for: GTT\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.9632642976974053\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GTA': 0.16939865988056665, 'GTC': 0.1704148238132479, 'GTG': 0.5544762811052315, 'GTT': 0.48163214884870265})\n",
      "Loss summary for: GAG\n",
      "loss for Bacillus subtilis: 0.3685286539080315\n",
      "loss for Escherichia coli: 0.13050877337561018\n",
      "Loss summary for: GAA\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAG': 0.24951871364182085, 'GAA': 0.5})\n",
      "Loss summary for: TAT\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.7264284023668639\n",
      "Loss summary for: TAC\n",
      "loss for Bacillus subtilis: 0.04695479203783222\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAT': 0.36321420118343195, 'TAC': 0.5234773960189161})\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6847751982489505]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26380392311639106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6847751982489505]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6734133912981506]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23653118484388272]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5655317497600131]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6734133912981506]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6847751982489505]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5045361299735621]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26380392311639106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.28667299498885185]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.24752923822568748]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2391679232236727]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.16211598104043537]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.28667299498885185]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2391679232236727]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.28667299498885185]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6847751982489505]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.24752923822568748]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.35960560391899904]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5045361299735621]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2391679232236727]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23653118484388272]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.28667299498885185]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6847751982489505]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23653118484388272]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.24752923822568748]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6734133912981506]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.28667299498885185]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.24752923822568748]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.35960560391899904]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5045361299735621]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.28667299498885185]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.35960560391899904]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2391679232236727]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2391679232236727]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23653118484388272]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6847751982489505]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.28667299498885185]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6847751982489505]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26380392311639106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5655317497600131]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6847751982489505]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.35960560391899904]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.16211598104043537]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.25598186049286953]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.35960560391899904]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23653118484388272]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23653118484388272]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.10611598754150686]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2391679232236727]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.35960560391899904]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6847751982489505]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5045361299735621]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2391679232236727]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26380392311639106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5655317497600131]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6847751982489505]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.28667299498885185]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23653118484388272]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.35960560391899904]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.28667299498885185]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7016557540022625]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.24752923822568748]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23653118484388272]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23653118484388272]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.10611598754150686]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26380392311639106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.28667299498885185]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.35960560391899904]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.24752923822568748]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2391679232236727]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5045361299735621]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26380392311639106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2391679232236727]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2391679232236727]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.35960560391899904]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.16211598104043537]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6734133912981506]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.35960560391899904]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.24752923822568748]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2674782920060872]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.24752923822568748]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7016557540022625]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.28667299498885185]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2391679232236727]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.28667299498885185]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.24752923822568748]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5655317497600131]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5655317497600131]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5045361299735621]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2391679232236727]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.35960560391899904]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5045361299735621]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23653118484388272]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6734133912981506]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5655317497600131]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.35960560391899904]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5045361299735621]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19774449549495793]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26380392311639106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.24752923822568748]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26380392311639106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6734133912981506]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23653118484388272]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.16211598104043537]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.35960560391899904]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3201813503247151]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2391679232236727]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2281722933643772]\n",
      "\n",
      "##########################\n",
      "# EVALUATION #\n",
      "##########################\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTTATGAGATTTAAGGTACATATGGAGGGATCAGTAAATGGACATGAGTTTGAGATTGAGGGAGAGGGAGAGGGAAGACCTTATGAGGGAACACAAACAGCTAAGCTTAAGGTAACAAAGGGAGGGCCTCTTCCTTTTGCTTGGGACATTCTTTCACCTCAATTTATGTATGGATCAAAGGCTTATGTAAAGCATCCTGCTGACATTCCTGACTATCTTAAGCTTTCATTTCCTGAGGGATTTAAGTGGGAGAGAGTAATGAATTTTGAGGACGGAGGGGTAGTCACAGTAACACAAGACTCAAGTCTTCAAGACGGAGAGTTTATTTATAAGGTAAAGCTTAGAGGAACAAATTTTCCTTCAGACGGACCTGTAATGCAAAAGAAAACAATGGGATGGGAGGCTTCAAGTGAGAGAATGTATCCTGAGGACGGAGCTCTTAAGGGAGAGATTAAGCAAAGACTTAAGCTTAAGGACGGAGGGCATTATGACGCTGAGGTAAAGACAACGTATAAGGCTAAGAAACCTGTACAACTTCCTGGAGCTTATAATGTAAATATTAAGCTTGACATTACATCACATAATGAGGACTATACAATTGTAGAGCAATATGAGAGAGCTGAGGGAAGACATTCAACAGGAGGGATGGACGAGCTTTATAAGTGA\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAGGTACATATGGAGGGATCAGTAAATGGACATGAGTTTGAGATTGAGGGAGAGGGAGAGGGAAGACCTTATGAGGGAACACAAACAGCTAAGCTTAAGGTAACAAAGGGAGGGCCTCTTCCTTTTGCTTGGGACATTCTTTCACCTCAATTTATGTATGGATCAAAGGCTTATGTAAAGCATCCTGCTGACATTCCTGACTATCTTAAGCTTTCATTTCCTGAGGGATTTAAGTGGGAGAGAGTAATGAATTTTGAGGACGGAGGGGTAGTCACAGTAACACAAGACTCAAGTCTTCAAGACGGAGAGTTTATTTATAAGGTAAAGCTTAGAGGAACAAATTTTCCTTCAGACGGACCTGTAATGCAAAAGAAAACAATGGGATGGGAGGCTTCAAGTGAGAGAATGTATCCTGAGGACGGAGCTCTTAAGGGAGAGATTAAGCAAAGACTTAAGCTTAAGGACGGAGGGCATTATGACGCTGAGGTAAAGACAACGTATAAGGCTAAGAAACCTGTACAACTTCCTGGAGCTTATAATGTAAATATTAAGCTTGACATTACATCACATAATGAGGACTATACAATTGTAGAGCAATATGAGAGAGCTGAGGGAAGACATTCAACAGGAGGGATGGACGAGCTTTATAAGTGA', 'average_distance_score': 1.9591117370140982, 'weakest_link_score': 1.9591117370140982, 'ratio_score': 1.0926009377970787}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAGGTACATATGGAGGGATCAGTAAATGGACATGAGTTTGAGATTGAGGGAGAGGGAGAGGGAAGACCTTATGAGGGAACACAAACAGCTAAGCTTAAGGTAACAAAGGGAGGGCCTCTTCCTTTTGCTTGGGACATTCTTTCACCTCAATTTATGTATGGATCAAAGGCTTATGTAAAGCATCCTGCTGACATTCCTGACTATCTTAAGCTTTCATTTCCTGAGGGATTTAAGTGGGAGAGAGTAATGAATTTTGAGGACGGAGGGGTAGTCACAGTAACACAAGACTCAAGTCTTCAAGACGGAGAGTTTATTTATAAGGTAAAGCTTAGAGGAACAAATTTTCCTTCAGACGGACCTGTAATGCAAAAGAAAACAATGGGATGGGAGGCTTCAAGTGAGAGAATGTATCCTGAGGACGGAGCTCTTAAGGGAGAGATTAAGCAAAGACTTAAGCTTAAGGACGGAGGGCATTATGACGCTGAGGTAAAGACAACGTATAAGGCTAAGAAACCTGTACAACTTCCTGGAGCTTATAATGTAAATATTAAGCTTGACATTACATCACATAATGAGGACTATACAATTGTAGAGCAATATGAGAGAGCTGAGGGAAGACATTCAACAGGAGGGATGGACGAGCTTTATAAGTGA', 'average_distance_score': 1.9591117370140982, 'weakest_link_score': 1.9591117370140982, 'ratio_score': 1.0926009377970787}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry-wanted-constraint-init-external-0.1\\CAI_single_codon_diff_ecoli_opt_False_sqYn\n",
      "Final output: {'module_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'orf_optimization_method': 'single_codon_diff', 'orf_optimization_cub_index': 'CAI', 'initiation_optimization_method': 'external', 'evaluation_score': 'average_distance', 'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.10739856801909307, 'TGA': 0.1694510739856802, 'GCA': 1.0, 'GCC': 0.5406038521603331, 'GCG': 0.950546590317543, 'GCT': 0.9508068714211347, 'TGC': 1.0, 'TGT': 0.7315315315315315, 'GAC': 0.6775877587758776, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.3929343907714492, 'TTC': 0.864321608040201, 'TTT': 1.0, 'GGA': 0.8215814372606443, 'GGC': 1.0, 'GGG': 0.28835323271006985, 'GGT': 0.6037395809867087, 'CAC': 0.6798462383305877, 'CAT': 1.0, 'ATA': 0.10219657483246464, 'ATC': 0.8559195830230827, 'ATT': 1.0, 'AAA': 1.0, 'AAG': 0.3472094095940959, 'CTA': 0.13892709766162312, 'CTC': 0.30375974323704724, 'CTG': 0.6657496561210454, 'CTT': 1.0, 'TTA': 0.5476845483723063, 'TTG': 0.3874369555249885, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.8191662154845695, 'CCA': 0.4708680142687277, 'CCC': 0.06896551724137931, 'CCG': 1.0, 'CCT': 0.7439556084026953, 'CAA': 1.0, 'CAG': 0.8559696778269108, 'AGA': 0.7029352226720648, 'AGG': 0.09210526315789473, 'CGA': 0.18927125506072875, 'CGC': 1.0, 'CGG': 0.27834008097165996, 'CGT': 0.9903846153846154, 'AGC': 0.8852526926263463, 'AGT': 0.3177299088649544, 'TCA': 0.9254349627174814, 'TCC': 0.4258492129246064, 'TCG': 0.2531068765534383, 'TCT': 1.0, 'ACA': 1.0, 'ACC': 0.22235752557233318, 'ACG': 0.5411592791037506, 'ACT': 0.43180711154408186, 'GTA': 0.7223221166195737, 'GTC': 0.6935525301823786, 'GTG': 0.6699203698946828, 'GTT': 1.0, 'TGG': 1.0, 'TAC': 0.7833094555873925, 'TAT': 1.0}, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'GAA': 0.6798186496752849, 'AAT': 0.5655317497600131, 'ATA': 0.1287010322849589, 'TTA': 0.198530453110291, 'GAC': 0.35960560391899904, 'CTG': 0.23973441522197322, 'TGG': 1.0, 'AAC': 0.4344682502399869, 'CAA': 0.5131548459124133, 'GCC': 0.20715414603308396, 'CTT': 0.2391679232236727, 'GCT': 0.24752923822568748, 'ATC': 0.366762837741479, 'AAA': 0.7016557540022625, 'AAG': 0.29834424599773746, 'TTG': 0.1592175745811708, 'AGC': 0.22534685577492342, 'CCG': 0.4334091260292722, 'AGT': 0.10611598754150686, 'TTT': 0.6847751982489505, 'GAG': 0.3201813503247151, 'ACT': 0.1614000537105004, 'TCA': 0.23653118484388272, 'ACC': 0.1588786441082565, 'CAC': 0.3265866087018494, 'GGC': 0.3390368073029434, 'GAT': 0.640394396081001, 'ACA': 0.4122430101751559, 'ACG': 0.2674782920060872, 'CCC': 0.0889644363010221, 'AGA': 0.26380392311639106, 'TCC': 0.12712038919920718, 'TAC': 0.34586274736720285, 'CAT': 0.6734133912981506, 'ATT': 0.5045361299735621, 'GCA': 0.2828363712991255, 'TAT': 0.6541372526327971, 'GGG': 0.16211598104043537, 'GTC': 0.25598186049286953, 'CCT': 0.28667299498885185, 'GTT': 0.28439644370189154, 'GCG': 0.26248024444210305, 'TCT': 0.2048572678833432, 'CTC': 0.11219040795754642, 'CCA': 0.19095344268085387, 'GGA': 0.3148341038094681, 'CGA': 0.09855152543710778, 'GTA': 0.19774449549495793, 'GTG': 0.26187720031028106, 'TTC': 0.31522480175104955, 'CGC': 0.2064569601768165, 'CAG': 0.4868451540875866, 'CGG': 0.1555630106168844, 'CGT': 0.18161187196589967, 'CTA': 0.05115922590534585, 'AGG': 0.09401270868690058, 'GGT': 0.18401310784715313, 'TAG': 0.14132712456344587, 'TGC': 0.5460141582025239, 'TCG': 0.10002831475713661, 'TAA': 0.630500582072177, 'TGA': 0.2281722933643772, 'TGT': 0.45398584179747614}), 'cai_avg': 0.7334925635409068, 'cai_std': 0.042802002292947894, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}, {'name': 'Escherichia coli', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.037037037037037035, 'TGA': 0.24183006535947713, 'GCA': 0.6303663185811371, 'GCC': 0.653737775567711, 'GCG': 1.0, 'GCT': 0.6063318415382065, 'TGC': 1.0, 'TGT': 0.7181136120042874, 'GAC': 0.848284271032921, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.36125998031280765, 'TTC': 1.0, 'TTT': 0.6556786703601108, 'GGA': 0.11638214051214708, 'GGC': 1.0, 'GGG': 0.18844386080105055, 'GGT': 0.9788246881155614, 'CAC': 1.0, 'CAT': 0.7088785046728973, 'ATA': 0.029297564419343447, 'ATC': 1.0, 'ATT': 0.7354394634662901, 'AAA': 1.0, 'AAG': 0.26291560102301786, 'CTA': 0.02288199697428139, 'CTC': 0.1265128593040847, 'CTG': 1.0, 'CTT': 0.10712934947049924, 'TTA': 0.09427004538577911, 'TTG': 0.11753025718608169, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.39072452229299365, 'CCA': 0.2588842975206611, 'CCC': 0.08181818181818182, 'CCG': 1.0, 'CCT': 0.1818181818181818, 'CAA': 0.32759266192437286, 'CAG': 1.0, 'AGA': 0.015347817420769383, 'AGG': 0.00976679290412597, 'CGA': 0.033087502491528795, 'CGC': 0.7123779150886984, 'CGG': 0.05521227825393661, 'CGT': 1.0, 'AGC': 1.0, 'AGT': 0.2936010037641154, 'TCA': 0.28523630280217477, 'TCC': 0.7921371810957758, 'TCG': 0.4086156419907988, 'TCT': 0.8883312421580928, 'ACA': 0.12800331400165701, 'ACC': 1.0, 'ACG': 0.34299917149958575, 'ACT': 0.4233637116818558, 'GTA': 0.5115587090867476, 'GTC': 0.49691004806591893, 'GTG': 1.0, 'GTT': 0.9814602883955139, 'TGG': 1.0, 'TAC': 1.0, 'TAT': 0.8523076923076923}, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'AAA': 0.7667390608953314, 'CGC': 0.40076028262096014, 'ATT': 0.5085852538544149, 'AGC': 0.27799163607995603, 'ACC': 0.4359429381095841, 'ATC': 0.4205725148436632, 'ACA': 0.13023262212317555, 'GGT': 0.33726003930064996, 'AAC': 0.5512262248110238, 'GCG': 0.3563679190097728, 'GGC': 0.4050486219579785, 'TGA': 0.28926761860893596, 'CGA': 0.06388523224887838, 'GTG': 0.3715940695510341, 'TTG': 0.12808970030428077, 'AAG': 0.23326093910466858, 'TTC': 0.4262385670731707, 'TCA': 0.12251736088908215, 'GCA': 0.21281773299119433, 'AAT': 0.4487737751889762, 'GAA': 0.689467997530356, 'CGT': 0.38120248233619, 'TTT': 0.5737614329268292, 'CTG': 0.4975546385450099, 'GTT': 0.2584847661069405, 'GCC': 0.26990242799092135, 'GAT': 0.626856640686903, 'AGG': 0.020162265266098916, 'CAG': 0.6530527088827319, 'GGG': 0.15057187484254547, 'GTC': 0.21619163129169194, 'CTC': 0.10435163186194824, 'TCT': 0.14544780223293646, 'CCC': 0.12361073915372255, 'CAC': 0.4292446655321377, 'GCT': 0.1609119200081115, 'TTA': 0.13009739763522432, 'CTT': 0.10341378016756284, 'ACG': 0.26849903016796667, 'GGA': 0.10711946389882603, 'CCG': 0.527785216925549, 'CAA': 0.3469472911172681, 'ACT': 0.16532540959927366, 'ATA': 0.07084223130192197, 'CAT': 0.5707553344678623, 'AGT': 0.15102374892893225, 'TGC': 0.55651396076928, 'GAG': 0.31053200246964396, 'TCG': 0.15406749964830613, 'GTA': 0.15372953305033354, 'TAC': 0.4315172054223149, 'TCC': 0.14895195222078703, 'GAC': 0.37314335931309695, 'TGT': 0.4434860392307201, 'TGG': 1.0, 'TAT': 0.5684827945776851, 'CCT': 0.15789033208355652, 'CCA': 0.19071371183717192, 'CGG': 0.09749348521075679, 'CTA': 0.03649285148597391, 'TAA': 0.6391064025794565, 'AGA': 0.036496252317115764, 'TAG': 0.07162597881160755}), 'cai_avg': 0.6151304741138522, 'cai_std': 0.09377868043841883, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}]}, 'orf_debug': {'total_loss': {'C': {'TGT': 0.2938812391538246, 'TGC': 0.5}, 'D': {'GAC': 0.411767928886508, 'GAT': 0.5}, 'S': {'TCA': 0.043459846610598665, 'AGT': 0.27584701333435985, 'TCG': 0.36240804036565677, 'TCT': 0.39456619789707004, 'TCC': 0.4785652199868279, 'AGC': 0.5065834722747519}, 'Q': {'CAA': 0.05365847607334823, 'CAG': 0.510372366852642}, 'M': {'ATG': 0.5}, 'N': {'AAT': 0.09268325497163063, 'AAC': 0.5}, 'P': {'CCT': 0.04930829085405168, 'CCA': 0.17350086891334277, 'CCC': 0.4367597114808227, 'CCG': 0.5}, 'K': {'AAG': 0.2476300840918918, 'AAA': 0.5}, '_': {'TGA': 0.37414664950765447, 'TAG': 0.3990545292434242, 'TAA': 0.5}, 'T': {'ACA': 0.0081924241977034, 'ACG': 0.16409161940099604, 'ACT': 0.2510399954304585, 'ACC': 0.8023639090169923}, 'F': {'TTT': 0.21495725938260146, 'TTC': 0.5092043130223984}, 'A': {'GCT': 0.18502913298124476, 'GCA': 0.19868084780076783, 'GCC': 0.31920894992707194, 'GCG': 0.5012228198646105}, 'G': {'GGA': 0.022688993080083285, 'GGG': 0.2709761050339069, 'GGC': 0.5, 'GGT': 0.5575600448705577}, 'I': {'ATT': 0.2704356022117923, 'ATA': 0.4034546687617319, 'ATC': 0.5103795832781212}, 'L': {'CTT': 0.005738348758986178, 'TTA': 0.1067380546191007, 'TTG': 0.19452342240536066, 'CTC': 0.2503779993533187, 'CTA': 0.37098506446348345, 'CTG': 0.5558616461915997}, 'H': {'CAT': 0.2512543671936414, 'CAC': 0.5512492155555374}, 'R': {'AGA': 0.04424151871423851, 'CGC': 0.2537411469530604, 'CGG': 0.26192071720099025, 'CGA': 0.3291879403459665, 'AGG': 0.4121841217146138, 'CGT': 0.5000462278106509}, 'W': {'TGG': 0.5}, 'V': {'GTA': 0.16939865988056665, 'GTC': 0.1704148238132479, 'GTT': 0.48163214884870265, 'GTG': 0.5544762811052315}, 'E': {'GAG': 0.24951871364182085, 'GAA': 0.5}, 'Y': {'TAT': 0.36321420118343195, 'TAC': 0.5234773960189161}}, 'optimized_loss': {'C': defaultdict(<class 'float'>, {'TGT': 0.07207531856180506, 'TGC': 0.0}), 'D': defaultdict(<class 'float'>, {'GAT': 0.0, 'GAC': 0.10394965329116168}), 'S': defaultdict(<class 'float'>, {'TCT': 0.0, 'TCG': 0.5578493378517608, 'TCA': 0.005559944784943392, 'TCC': 0.32964912629929394, 'AGC': 0.013166944549503757, 'AGT': 0.46549247725742354}), 'Q': defaultdict(<class 'float'>, {'CAA': 0.0, 'CAG': 0.02074473370528387}), 'M': defaultdict(<class 'float'>, {'ATG': 0.0}), 'N': defaultdict(<class 'float'>, {'AAC': 0.0, 'AAT': 0.03270085762217316}), 'P': defaultdict(<class 'float'>, {'CCT': 0.06555873046843394, 'CCG': 0.0, 'CCA': 0.27998065832391933, 'CCC': 0.8668252080856124}), 'K': defaultdict(<class 'float'>, {'AAG': 0.4261355549224889, 'AAA': 0.0}), '_': defaultdict(<class 'float'>, {'TAG': 0.7967373163743656, 'TGA': 0.6898115185035399, 'TAA': 0.0}), 'T': defaultdict(<class 'float'>, {'ACC': 0.6047278180339846, 'ACA': 0.0, 'ACG': 0.21053480715258982, 'ACT': 0.3228431584918795}), 'F': defaultdict(<class 'float'>, {'TTT': 0.0, 'TTC': 0.018408626044796858}), 'A': defaultdict(<class 'float'>, {'GCA': 0.0, 'GCC': 0.21104482064992505, 'GCG': 0.0024456397292209357, 'GCT': 0.0024199638993767703}), 'G': defaultdict(<class 'float'>, {'GGT': 0.15702231967658922, 'GGG': 0.506441121394208, 'GGA': 0.031833183529977424, 'GGC': 0.0}), 'I': defaultdict(<class 'float'>, {'ATC': 0.020759166556242364, 'ATA': 0.8060509902425582, 'ATT': 0.0}), 'L': defaultdict(<class 'float'>, {'TTA': 0.20458926778116454, 'TTG': 0.3752334834564948, 'CTC': 0.48475049513734236, 'CTT': 0.0, 'CTG': 0.11172329238319939, 'CTA': 0.7414465431414359}), 'H': defaultdict(<class 'float'>, {'CAT': 0.0, 'CAC': 0.10249843111107487}), 'R': defaultdict(<class 'float'>, {'CGA': 0.657281097870806, 'CGC': 0.0, 'CGG': 0.5207930387319903, 'CGT': 9.24556213017745e-05, 'AGG': 0.8242728531855956, 'AGA': 0.08824748192889574}), 'W': defaultdict(<class 'float'>, {'TGG': 0.0}), 'V': defaultdict(<class 'float'>, {'GTA': 0.07710500691863363, 'GTC': 0.09391005175762197, 'GTG': 0.108952562210463, 'GTT': 0.0}), 'E': defaultdict(<class 'float'>, {'GAG': 0.3685286539080315, 'GAA': 0.0}), 'Y': defaultdict(<class 'float'>, {'TAT': 0.0, 'TAC': 0.04695479203783222})}, 'deoptimized_loss': {'C': defaultdict(<class 'float'>, {'TGT': 0.5156871597458441, 'TGC': 1.0}), 'D': defaultdict(<class 'float'>, {'GAT': 1.0, 'GAC': 0.7195862044818543}), 'S': defaultdict(<class 'float'>, {'TCT': 0.7891323957941401, 'TCG': 0.16696674287955265, 'TCA': 0.08135974843625393, 'TCC': 0.6274813136743619, 'AGC': 1.0, 'AGT': 0.08620154941129611}), 'Q': defaultdict(<class 'float'>, {'CAA': 0.10731695214669645, 'CAG': 1.0}), 'M': defaultdict(<class 'float'>, {'ATG': 1.0}), 'N': defaultdict(<class 'float'>, {'AAC': 1.0, 'AAT': 0.15266565232108809}), 'P': defaultdict(<class 'float'>, {'CCT': 0.033057851239669415, 'CCG': 1.0, 'CCA': 0.06702107950276619, 'CCC': 0.006694214876033058}), 'K': defaultdict(<class 'float'>, {'AAG': 0.06912461326129471, 'AAA': 1.0}), '_': defaultdict(<class 'float'>, {'TAG': 0.0013717421124828531, 'TGA': 0.05848178051176898, 'TAA': 1.0}), 'T': defaultdict(<class 'float'>, {'ACC': 1.0, 'ACA': 0.0163848483954068, 'ACG': 0.11764843164940224, 'ACT': 0.17923683236903754}), 'F': defaultdict(<class 'float'>, {'TTT': 0.4299145187652029, 'TTC': 1.0}), 'A': defaultdict(<class 'float'>, {'GCA': 0.39736169560153567, 'GCC': 0.4273730792042188, 'GCG': 1.0, 'GCT': 0.36763830206311277}), 'G': defaultdict(<class 'float'>, {'GGT': 0.9580977700645261, 'GGG': 0.035511088673605716, 'GGA': 0.013544802630189147, 'GGC': 1.0}), 'I': defaultdict(<class 'float'>, {'ATC': 1.0, 'ATA': 0.0008583472809055792, 'ATT': 0.5408712044235846}), 'L': defaultdict(<class 'float'>, {'TTA': 0.008886841457036854, 'TTG': 0.013813361354226507, 'CTC': 0.016005503569295133, 'CTT': 0.011476697517972355, 'CTG': 1.0, 'CTA': 0.0005235857855310226}), 'H': defaultdict(<class 'float'>, {'CAT': 0.5025087343872828, 'CAC': 1.0}), 'R': defaultdict(<class 'float'>, {'CGA': 0.0010947828211269241, 'CGC': 0.5074822939061208, 'CGG': 0.0030483956699901217, 'CGT': 1.0, 'AGG': 9.53902436320854e-05, 'AGA': 0.00023555549958127213}), 'W': defaultdict(<class 'float'>, {'TGG': 1.0}), 'V': defaultdict(<class 'float'>, {'GTA': 0.26169231284249966, 'GTC': 0.24691959586887385, 'GTG': 1.0, 'GTT': 0.9632642976974053}), 'E': defaultdict(<class 'float'>, {'GAG': 0.13050877337561018, 'GAA': 1.0}), 'Y': defaultdict(<class 'float'>, {'TAT': 0.7264284023668639, 'TAC': 1.0})}}, 'orf': {'orf_module_input_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'optimized_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAGGTACATATGGAGGGATCAGTAAATGGACATGAGTTTGAGATTGAGGGAGAGGGAGAGGGAAGACCTTATGAGGGAACACAAACAGCTAAGCTTAAGGTAACAAAGGGAGGGCCTCTTCCTTTTGCTTGGGACATTCTTTCACCTCAATTTATGTATGGATCAAAGGCTTATGTAAAGCATCCTGCTGACATTCCTGACTATCTTAAGCTTTCATTTCCTGAGGGATTTAAGTGGGAGAGAGTAATGAATTTTGAGGACGGAGGGGTAGTCACAGTAACACAAGACTCAAGTCTTCAAGACGGAGAGTTTATTTATAAGGTAAAGCTTAGAGGAACAAATTTTCCTTCAGACGGACCTGTAATGCAAAAGAAAACAATGGGATGGGAGGCTTCAAGTGAGAGAATGTATCCTGAGGACGGAGCTCTTAAGGGAGAGATTAAGCAAAGACTTAAGCTTAAGGACGGAGGGCATTATGACGCTGAGGTAAAGACAACGTATAAGGCTAAGAAACCTGTACAACTTCCTGGAGCTTATAATGTAAATATTAAGCTTGACATTACATCACATAATGAGGACTATACAATTGTAGAGCAATATGAGAGAGCTGAGGGAAGACATTCAACAGGAGGGATGGACGAGCTTTATAAGTGA', 'aa_to_loss_mapping': {'C': {'TGT': 0.2938812391538246, 'TGC': 0.5}, 'D': {'GAC': 0.411767928886508, 'GAT': 0.5}, 'S': {'TCA': 0.043459846610598665, 'AGT': 0.27584701333435985, 'TCG': 0.36240804036565677, 'TCT': 0.39456619789707004, 'TCC': 0.4785652199868279, 'AGC': 0.5065834722747519}, 'Q': {'CAA': 0.05365847607334823, 'CAG': 0.510372366852642}, 'M': {'ATG': 0.5}, 'N': {'AAT': 0.09268325497163063, 'AAC': 0.5}, 'P': {'CCT': 0.04930829085405168, 'CCA': 0.17350086891334277, 'CCC': 0.4367597114808227, 'CCG': 0.5}, 'K': {'AAG': 0.2476300840918918, 'AAA': 0.5}, '_': {'TGA': 0.37414664950765447, 'TAG': 0.3990545292434242, 'TAA': 0.5}, 'T': {'ACA': 0.0081924241977034, 'ACG': 0.16409161940099604, 'ACT': 0.2510399954304585, 'ACC': 0.8023639090169923}, 'F': {'TTT': 0.21495725938260146, 'TTC': 0.5092043130223984}, 'A': {'GCT': 0.18502913298124476, 'GCA': 0.19868084780076783, 'GCC': 0.31920894992707194, 'GCG': 0.5012228198646105}, 'G': {'GGA': 0.022688993080083285, 'GGG': 0.2709761050339069, 'GGC': 0.5, 'GGT': 0.5575600448705577}, 'I': {'ATT': 0.2704356022117923, 'ATA': 0.4034546687617319, 'ATC': 0.5103795832781212}, 'L': {'CTT': 0.005738348758986178, 'TTA': 0.1067380546191007, 'TTG': 0.19452342240536066, 'CTC': 0.2503779993533187, 'CTA': 0.37098506446348345, 'CTG': 0.5558616461915997}, 'H': {'CAT': 0.2512543671936414, 'CAC': 0.5512492155555374}, 'R': {'AGA': 0.04424151871423851, 'CGC': 0.2537411469530604, 'CGG': 0.26192071720099025, 'CGA': 0.3291879403459665, 'AGG': 0.4121841217146138, 'CGT': 0.5000462278106509}, 'W': {'TGG': 0.5}, 'V': {'GTA': 0.16939865988056665, 'GTC': 0.1704148238132479, 'GTT': 0.48163214884870265, 'GTG': 0.5544762811052315}, 'E': {'GAG': 0.24951871364182085, 'GAA': 0.5}, 'Y': {'TAT': 0.36321420118343195, 'TAC': 0.5234773960189161}}, 'run_time': 0.7349999999860302}, 'evaluation': [{'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'cai_initial_score': 0.7695788606254016, 'cai_final_score': 0.7082628871665226, 'dist_score': -1.4325491840128575}, {'name': 'Escherichia coli', 'is_wanted': False, 'cai_initial_score': 0.7915994634162647, 'cai_final_score': 0.2898110642192039, 'dist_score': -5.350772658041054}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAGGTACATATGGAGGGATCAGTAAATGGACATGAGTTTGAGATTGAGGGAGAGGGAGAGGGAAGACCTTATGAGGGAACACAAACAGCTAAGCTTAAGGTAACAAAGGGAGGGCCTCTTCCTTTTGCTTGGGACATTCTTTCACCTCAATTTATGTATGGATCAAAGGCTTATGTAAAGCATCCTGCTGACATTCCTGACTATCTTAAGCTTTCATTTCCTGAGGGATTTAAGTGGGAGAGAGTAATGAATTTTGAGGACGGAGGGGTAGTCACAGTAACACAAGACTCAAGTCTTCAAGACGGAGAGTTTATTTATAAGGTAAAGCTTAGAGGAACAAATTTTCCTTCAGACGGACCTGTAATGCAAAAGAAAACAATGGGATGGGAGGCTTCAAGTGAGAGAATGTATCCTGAGGACGGAGCTCTTAAGGGAGAGATTAAGCAAAGACTTAAGCTTAAGGACGGAGGGCATTATGACGCTGAGGTAAAGACAACGTATAAGGCTAAGAAACCTGTACAACTTCCTGGAGCTTATAATGTAAATATTAAGCTTGACATTACATCACATAATGAGGACTATACAATTGTAGAGCAATATGAGAGAGCTGAGGGAAGACATTCAACAGGAGGGATGGACGAGCTTTATAAGTGA', 'average_distance_score': 1.9591117370140982, 'weakest_link_score': 1.9591117370140982, 'ratio_score': 1.0926009377970787, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTTATGAGATTTAAGGTACATATGGAGGGATCAGTAAATGGACATGAGTTTGAGATTGAGGGAGAGGGAGAGGGAAGACCTTATGAGGGAACACAAACAGCTAAGCTTAAGGTAACAAAGGGAGGGCCTCTTCCTTTTGCTTGGGACATTCTTTCACCTCAATTTATGTATGGATCAAAGGCTTATGTAAAGCATCCTGCTGACATTCCTGACTATCTTAAGCTTTCATTTCCTGAGGGATTTAAGTGGGAGAGAGTAATGAATTTTGAGGACGGAGGGGTAGTCACAGTAACACAAGACTCAAGTCTTCAAGACGGAGAGTTTATTTATAAGGTAAAGCTTAGAGGAACAAATTTTCCTTCAGACGGACCTGTAATGCAAAAGAAAACAATGGGATGGGAGGCTTCAAGTGAGAGAATGTATCCTGAGGACGGAGCTCTTAAGGGAGAGATTAAGCAAAGACTTAAGCTTAAGGACGGAGGGCATTATGACGCTGAGGTAAAGACAACGTATAAGGCTAAGAAACCTGTACAACTTCCTGGAGCTTATAATGTAAATATTAAGCTTGACATTACATCACATAATGAGGACTATACAATTGTAGAGCAATATGAGAGAGCTGAGGGAAGACATTCAACAGGAGGGATGGACGAGCTTTATAAGTGA'}], 'final_evaluation': {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAGGTACATATGGAGGGATCAGTAAATGGACATGAGTTTGAGATTGAGGGAGAGGGAGAGGGAAGACCTTATGAGGGAACACAAACAGCTAAGCTTAAGGTAACAAAGGGAGGGCCTCTTCCTTTTGCTTGGGACATTCTTTCACCTCAATTTATGTATGGATCAAAGGCTTATGTAAAGCATCCTGCTGACATTCCTGACTATCTTAAGCTTTCATTTCCTGAGGGATTTAAGTGGGAGAGAGTAATGAATTTTGAGGACGGAGGGGTAGTCACAGTAACACAAGACTCAAGTCTTCAAGACGGAGAGTTTATTTATAAGGTAAAGCTTAGAGGAACAAATTTTCCTTCAGACGGACCTGTAATGCAAAAGAAAACAATGGGATGGGAGGCTTCAAGTGAGAGAATGTATCCTGAGGACGGAGCTCTTAAGGGAGAGATTAAGCAAAGACTTAAGCTTAAGGACGGAGGGCATTATGACGCTGAGGTAAAGACAACGTATAAGGCTAAGAAACCTGTACAACTTCCTGGAGCTTATAATGTAAATATTAAGCTTGACATTACATCACATAATGAGGACTATACAATTGTAGAGCAATATGAGAGAGCTGAGGGAAGACATTCAACAGGAGGGATGGACGAGCTTTATAAGTGA', 'average_distance_score': 1.9591117370140982, 'weakest_link_score': 1.9591117370140982, 'ratio_score': 1.0926009377970787}, 'zip_output_file_path': 'results\\\\mcherry-wanted-constraint-init-external-0.1\\\\CAI_single_codon_diff_ecoli_opt_False_sqYn\\\\communique_results.zip'}\n",
      "\n",
      "##########################\n",
      "# User Input #\n",
      "##########################\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is:\n",
      "ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "------------------------------------------\n",
      "Parsing information for Escherichia coli:\n",
      "------------------------------------------\n",
      "Organism is defined as wanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Escherichia coli. Calculating tAI profile.\n",
      "tai_std=0.024913119712427708, tai_avg=0.24958213077316901\n",
      "------------------------------------------\n",
      "Parsing information for Bacillus subtilis:\n",
      "------------------------------------------\n",
      "Organism is defined as unwanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Bacillus subtilis. Calculating tAI profile.\n",
      "tai_std=0.028076995641981673, tai_avg=0.3271366648197227\n",
      "-----------------------------\n",
      "Normalized Prioritization Weights\n",
      "-----------------------------\n",
      "Escherichia coli : 1.0\n",
      "Bacillus subtilis : 1.0\n",
      "Total input processing time: 29.456879377365112\n",
      "\n",
      "##########################\n",
      "# Initiation #\n",
      "##########################\n",
      "Running initiation optimization InitiationOptimizationMethod.external_module on 15 codons at the start of the ORF\n",
      "Taking optimized sequence from configuration value config.INITIATION.EXTERNAL_INITIATION_ORF\n",
      "Optimized sequence is: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "Optimization method used is: ORFOptimizationMethod.single_codon_diff\n",
      "Loss summary for: TGT\n",
      "loss for Escherichia coli: 0.002626562499999999\n",
      "loss for Bacillus subtilis: 0.009669444444444446\n",
      "Loss summary for: TGC\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.027777777777777776\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGT': 0.006148003472222222, 'TGC': 0.013888888888888888})\n",
      "Loss summary for: GAT\n",
      "loss for Escherichia coli: 0.02363906249999999\n",
      "loss for Bacillus subtilis: 0.15471111111111113\n",
      "Loss summary for: GAC\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.4444444444444444\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAT': 0.08917508680555557, 'GAC': 0.2222222222222222})\n",
      "Loss summary for: TCT\n",
      "loss for Escherichia coli: 0.010506249999999996\n",
      "loss for Bacillus subtilis: 0.009669444444444446\n",
      "Loss summary for: TCG\n",
      "loss for Escherichia coli: 0.007225000000000003\n",
      "loss for Bacillus subtilis: 0.011377777777777773\n",
      "Loss summary for: TCA\n",
      "loss for Escherichia coli: 0.012376562500000004\n",
      "loss for Bacillus subtilis: 0.1111111111111111\n",
      "Loss summary for: TCC\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.027777777777777776\n",
      "Loss summary for: AGC\n",
      "loss for Escherichia coli: 0.015625\n",
      "loss for Bacillus subtilis: 0.1111111111111111\n",
      "Loss summary for: AGT\n",
      "loss for Escherichia coli: 0.031064062499999996\n",
      "loss for Bacillus subtilis: 0.03867777777777778\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TCT': 0.01008784722222222, 'TCG': 0.009301388888888889, 'TCA': 0.061743836805555556, 'TCC': 0.013888888888888888, 'AGC': 0.06336805555555555, 'AGT': 0.034870920138888886})\n",
      "Loss summary for: CAA\n",
      "loss for Escherichia coli: 0.002756249999999999\n",
      "loss for Bacillus subtilis: 0.4444444444444444\n",
      "Loss summary for: CAG\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.04551111111111109\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAA': 0.22360034722222222, 'CAG': 0.022755555555555546})\n",
      "Loss summary for: ATG\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATG': 0.5})\n",
      "Loss summary for: AAC\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.4444444444444444\n",
      "Loss summary for: AAT\n",
      "loss for Escherichia coli: 0.042024999999999986\n",
      "loss for Bacillus subtilis: 0.15471111111111113\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAC': 0.2222222222222222, 'AAT': 0.09836805555555556})\n",
      "Loss summary for: CCT\n",
      "loss for Escherichia coli: 0.008326562499999994\n",
      "loss for Bacillus subtilis: 0\n",
      "Loss summary for: CCG\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.02559999999999999\n",
      "Loss summary for: CCA\n",
      "loss for Escherichia coli: 0.0006890624999999998\n",
      "loss for Bacillus subtilis: 0.25\n",
      "Loss summary for: CCC\n",
      "loss for Escherichia coli: 0.0015999999999999983\n",
      "loss for Bacillus subtilis: 0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CCT': 0.004163281249999997, 'CCG': 0.012799999999999995, 'CCA': 0.12534453125, 'CCC': 0.0007999999999999992})\n",
      "Loss summary for: AAG\n",
      "loss for Escherichia coli: 0.2601\n",
      "loss for Bacillus subtilis: 0.04551111111111109\n",
      "Loss summary for: AAA\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.4444444444444444\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAG': 0.15280555555555556, 'AAA': 0.2222222222222222})\n",
      "Loss summary for: TAG\n",
      "loss for Escherichia coli: 0.059284665816312816\n",
      "loss for Bacillus subtilis: 0\n",
      "Loss summary for: TGA\n",
      "loss for Escherichia coli: 0.059284665816312816\n",
      "loss for Bacillus subtilis: 0\n",
      "Loss summary for: TAA\n",
      "loss for Escherichia coli: 0.059284665816312816\n",
      "loss for Bacillus subtilis: 0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAG': 0.029642332908156408, 'TGA': 0.029642332908156408, 'TAA': 0.029642332908156408})\n",
      "Loss summary for: ACC\n",
      "loss for Escherichia coli: 0.0015999999999999983\n",
      "loss for Bacillus subtilis: 0.027777777777777776\n",
      "Loss summary for: ACA\n",
      "loss for Escherichia coli: 0.018906249999999996\n",
      "loss for Bacillus subtilis: 0.4444444444444444\n",
      "Loss summary for: ACG\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.04551111111111109\n",
      "Loss summary for: ACT\n",
      "loss for Escherichia coli: 0.020306249999999988\n",
      "loss for Bacillus subtilis: 0.009669444444444446\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ACC': 0.014688888888888887, 'ACA': 0.23167534722222222, 'ACG': 0.022755555555555546, 'ACT': 0.014987847222222217})\n",
      "Loss summary for: TTT\n",
      "loss for Escherichia coli: 0.010506249999999996\n",
      "loss for Bacillus subtilis: 0.08702500000000002\n",
      "Loss summary for: TTC\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.25\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTT': 0.04876562500000001, 'TTC': 0.125})\n",
      "Loss summary for: GCA\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.6944444444444445\n",
      "Loss summary for: GCC\n",
      "loss for Escherichia coli: 0.015625\n",
      "loss for Bacillus subtilis: 0.027777777777777776\n",
      "Loss summary for: GCG\n",
      "loss for Escherichia coli: 0.065025\n",
      "loss for Bacillus subtilis: 0.07111111111111108\n",
      "Loss summary for: GCT\n",
      "loss for Escherichia coli: 0.05175624999999999\n",
      "loss for Bacillus subtilis: 0.009669444444444446\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GCA': 0.34722222222222227, 'GCC': 0.021701388888888888, 'GCG': 0.06806805555555553, 'GCT': 0.030712847222222218})\n",
      "Loss summary for: GGT\n",
      "loss for Escherichia coli: 0.042024999999999986\n",
      "loss for Bacillus subtilis: 0.15471111111111113\n",
      "Loss summary for: GGG\n",
      "loss for Escherichia coli: 0.11222500000000002\n",
      "loss for Bacillus subtilis: 0.02559999999999999\n",
      "Loss summary for: GGA\n",
      "loss for Escherichia coli: 0.1305015625\n",
      "loss for Bacillus subtilis: 0.25\n",
      "Loss summary for: GGC\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.4444444444444444\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GGT': 0.09836805555555556, 'GGG': 0.0689125, 'GGA': 0.19025078125, 'GGC': 0.2222222222222222})\n",
      "Loss summary for: ATC\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.25\n",
      "Loss summary for: ATA\n",
      "loss for Escherichia coli: 0.07022500000000001\n",
      "loss for Bacillus subtilis: 0.012099999999999998\n",
      "Loss summary for: ATT\n",
      "loss for Escherichia coli: 0.02363906249999999\n",
      "loss for Bacillus subtilis: 0.08702500000000002\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATC': 0.125, 'ATA': 0.041162500000000005, 'ATT': 0.055332031250000004})\n",
      "Loss summary for: TTA\n",
      "loss for Escherichia coli: 0.16100156250000003\n",
      "loss for Bacillus subtilis: 0.2686694444444444\n",
      "Loss summary for: TTG\n",
      "loss for Escherichia coli: 0.14062500000000006\n",
      "loss for Bacillus subtilis: 0.1067111111111111\n",
      "Loss summary for: CTC\n",
      "loss for Escherichia coli: 0.17222500000000002\n",
      "loss for Bacillus subtilis: 0.027777777777777776\n",
      "Loss summary for: CTT\n",
      "loss for Escherichia coli: 0.21738906250000006\n",
      "loss for Bacillus subtilis: 0.009669444444444446\n",
      "Loss summary for: CTG\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.0747111111111111\n",
      "Loss summary for: CTA\n",
      "loss for Escherichia coli: 0.12960000000000002\n",
      "loss for Bacillus subtilis: 0.12366944444444442\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTA': 0.21483550347222222, 'TTG': 0.12366805555555557, 'CTC': 0.1000013888888889, 'CTT': 0.11352925347222226, 'CTG': 0.03735555555555555, 'CTA': 0.12663472222222222})\n",
      "Loss summary for: CAT\n",
      "loss for Escherichia coli: 0.002626562499999999\n",
      "loss for Bacillus subtilis: 0.03867777777777778\n",
      "Loss summary for: CAC\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.1111111111111111\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAT': 0.02065217013888889, 'CAC': 0.05555555555555555})\n",
      "Loss summary for: CGA\n",
      "loss for Escherichia coli: 0.23639044\n",
      "loss for Bacillus subtilis: 0.00033855999999999963\n",
      "Loss summary for: CGC\n",
      "loss for Escherichia coli: 0.019600000000000003\n",
      "loss for Bacillus subtilis: 0.2304\n",
      "Loss summary for: CGG\n",
      "loss for Escherichia coli: 0.140625\n",
      "loss for Bacillus subtilis: 0.027777777777777776\n",
      "Loss summary for: CGT\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.4444444444444444\n",
      "Loss summary for: AGG\n",
      "loss for Escherichia coli: 0.11222500000000002\n",
      "loss for Bacillus subtilis: 0.04839999999999999\n",
      "Loss summary for: AGA\n",
      "loss for Escherichia coli: 0.1305015625\n",
      "loss for Bacillus subtilis: 0.03422499999999999\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CGA': 0.1183645, 'CGC': 0.125, 'CGG': 0.0842013888888889, 'CGT': 0.2222222222222222, 'AGG': 0.08031250000000001, 'AGA': 0.08236328125})\n",
      "Loss summary for: TGG\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.027777777777777776\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGG': 0.013888888888888888})\n",
      "Loss summary for: GTA\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.4444444444444444\n",
      "Loss summary for: GTC\n",
      "loss for Escherichia coli: 0.140625\n",
      "loss for Bacillus subtilis: 0.027777777777777776\n",
      "Loss summary for: GTG\n",
      "loss for Escherichia coli: 0.18062500000000004\n",
      "loss for Bacillus subtilis: 0.04551111111111109\n",
      "Loss summary for: GTT\n",
      "loss for Escherichia coli: 0.22800625\n",
      "loss for Bacillus subtilis: 0.009669444444444446\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GTA': 0.2222222222222222, 'GTC': 0.0842013888888889, 'GTG': 0.11306805555555556, 'GTT': 0.11883784722222222})\n",
      "Loss summary for: GAG\n",
      "loss for Escherichia coli: 0.11560000000000002\n",
      "loss for Bacillus subtilis: 0.10239999999999996\n",
      "Loss summary for: GAA\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAG': 0.10899999999999999, 'GAA': 0.5})\n",
      "Loss summary for: TAT\n",
      "loss for Escherichia coli: 0.02363906249999999\n",
      "loss for Bacillus subtilis: 0.03867777777777778\n",
      "Loss summary for: TAC\n",
      "loss for Escherichia coli: 0.0\n",
      "loss for Bacillus subtilis: 0.1111111111111111\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAT': 0.031158420138888886, 'TAC': 0.05555555555555555})\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5737614329268292]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.020162265266098916]\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.036496252317115764]\n",
      "Skipping codon AGA due to very low average frequency 0.036496252317115764 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.09749348521075679]\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.06388523224887838]\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.40076028262096014]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5737614329268292]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5707553344678623]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15406749964830613]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4487737751889762]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5707553344678623]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5737614329268292]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.07084223130192197]\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5085852538544149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.020162265266098916]\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.036496252317115764]\n",
      "Skipping codon AGA due to very low average frequency 0.036496252317115764 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.09749348521075679]\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.06388523224887838]\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.40076028262096014]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5684827945776851]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5737614329268292]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.626856640686903]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.07084223130192197]\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5085852538544149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15406749964830613]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5737614329268292]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5684827945776851]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15406749964830613]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5684827945776851]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5707553344678623]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.626856640686903]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.07084223130192197]\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5085852538544149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.626856640686903]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5684827945776851]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15406749964830613]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5737614329268292]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5737614329268292]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.020162265266098916]\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.036496252317115764]\n",
      "Skipping codon AGA due to very low average frequency 0.036496252317115764 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.09749348521075679]\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.06388523224887838]\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.40076028262096014]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4487737751889762]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5737614329268292]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.626856640686903]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3715940695510341]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.626856640686903]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15406749964830613]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15406749964830613]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14544780223293646]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.626856640686903]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5737614329268292]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.07084223130192197]\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5085852538544149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5684827945776851]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.020162265266098916]\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.036496252317115764]\n",
      "Skipping codon AGA due to very low average frequency 0.036496252317115764 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.09749348521075679]\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.06388523224887838]\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.40076028262096014]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4487737751889762]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5737614329268292]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15406749964830613]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.626856640686903]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15406749964830613]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15406749964830613]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14544780223293646]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.020162265266098916]\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.036496252317115764]\n",
      "Skipping codon AGA due to very low average frequency 0.036496252317115764 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.09749348521075679]\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.06388523224887838]\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.40076028262096014]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5684827945776851]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.626856640686903]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.07084223130192197]\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5085852538544149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.020162265266098916]\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.036496252317115764]\n",
      "Skipping codon AGA due to very low average frequency 0.036496252317115764 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.09749348521075679]\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.06388523224887838]\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.40076028262096014]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.626856640686903]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5707553344678623]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5684827945776851]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.626856640686903]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.16532540959927366]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5684827945776851]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7667390608953314]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.12361073915372255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5684827945776851]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4487737751889762]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4487737751889762]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.07084223130192197]\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5085852538544149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.626856640686903]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.07084223130192197]\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5085852538544149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15406749964830613]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5707553344678623]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4487737751889762]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.626856640686903]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5684827945776851]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.07084223130192197]\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5085852538544149]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.21619163129169194]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6530527088827319]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5684827945776851]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.020162265266098916]\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.036496252317115764]\n",
      "Skipping codon AGA due to very low average frequency 0.036496252317115764 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.09749348521075679]\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.06388523224887838]\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.40076028262096014]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26990242799092135]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.020162265266098916]\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.036496252317115764]\n",
      "Skipping codon AGA due to very low average frequency 0.036496252317115764 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.09749348521075679]\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.06388523224887838]\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.40076028262096014]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5707553344678623]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15406749964830613]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4359429381095841]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.15057187484254547]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.33726003930064996]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.626856640686903]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31053200246964396]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4975546385450099]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5684827945776851]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23326093910466858]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.07162597881160755]\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.28926761860893596]\n",
      "\n",
      "##########################\n",
      "# EVALUATION #\n",
      "##########################\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTTATGCGCTTTAAGGTCCATATGGAGGGGTCGGTCAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCCTATGAGGGGACCCAGACCGCCAAGCTGAAGGTCACCAAGGGGGGTCCCCTGCCCTTTGCCTGGGATATTCTGTCGCCCCAGTTTATGTATGGGTCGAAGGCCTATGTCAAGCATCCCGCCGATATTCCCGATTATCTGAAGCTGTCGTTTCCCGAGGGGTTTAAGTGGGAGCGCGTCATGAATTTTGAGGATGGGGGTGTCGTGACCGTCACCCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTCAAGCTGCGCGGGACCAATTTTCCCTCGGATGGGCCCGTCATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGTCTGAGCGCATGTATCCCGAGGATGGGGCCCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCCGAGGTCAAGACCACTTATAAGGCCAAGAAACCCGTCCAGCTGCCCGGGGCCTATAATGTCAATATTAAGCTGGATATTACCTCGCATAATGAGGATTATACCATTGTCGAGCAGTATGAGCGCGCCGAGGGGCGCCATTCGACCGGGGGTATGGATGAGCTGTATAAGTAA\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGCTTTAAGGTCCATATGGAGGGGTCGGTCAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCCTATGAGGGGACCCAGACCGCCAAGCTGAAGGTCACCAAGGGGGGTCCCCTGCCCTTTGCCTGGGATATTCTGTCGCCCCAGTTTATGTATGGGTCGAAGGCCTATGTCAAGCATCCCGCCGATATTCCCGATTATCTGAAGCTGTCGTTTCCCGAGGGGTTTAAGTGGGAGCGCGTCATGAATTTTGAGGATGGGGGTGTCGTGACCGTCACCCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTCAAGCTGCGCGGGACCAATTTTCCCTCGGATGGGCCCGTCATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGTCTGAGCGCATGTATCCCGAGGATGGGGCCCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCCGAGGTCAAGACCACTTATAAGGCCAAGAAACCCGTCCAGCTGCCCGGGGCCTATAATGTCAATATTAAGCTGGATATTACCTCGCATAATGAGGATTATACCATTGTCGAGCAGTATGAGCGCGCCGAGGGGCGCCATTCGACCGGGGGTATGGATGAGCTGTATAAGTAA', 'average_distance_score': 0.47700481472248213, 'weakest_link_score': 0.47700481472248213, 'ratio_score': 1.0256286534297445}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGCTTTAAGGTCCATATGGAGGGGTCGGTCAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCCTATGAGGGGACCCAGACCGCCAAGCTGAAGGTCACCAAGGGGGGTCCCCTGCCCTTTGCCTGGGATATTCTGTCGCCCCAGTTTATGTATGGGTCGAAGGCCTATGTCAAGCATCCCGCCGATATTCCCGATTATCTGAAGCTGTCGTTTCCCGAGGGGTTTAAGTGGGAGCGCGTCATGAATTTTGAGGATGGGGGTGTCGTGACCGTCACCCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTCAAGCTGCGCGGGACCAATTTTCCCTCGGATGGGCCCGTCATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGTCTGAGCGCATGTATCCCGAGGATGGGGCCCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCCGAGGTCAAGACCACTTATAAGGCCAAGAAACCCGTCCAGCTGCCCGGGGCCTATAATGTCAATATTAAGCTGGATATTACCTCGCATAATGAGGATTATACCATTGTCGAGCAGTATGAGCGCGCCGAGGGGCGCCATTCGACCGGGGGTATGGATGAGCTGTATAAGTAA', 'average_distance_score': 0.47700481472248213, 'weakest_link_score': 0.47700481472248213, 'ratio_score': 1.0256286534297445}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry-wanted-constraint-init-external-0.1\\tAI_single_codon_diff_ecoli_opt_True_XQ2U\n",
      "Final output: {'module_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'orf_optimization_method': 'single_codon_diff', 'orf_optimization_cub_index': 'tAI', 'initiation_optimization_method': 'external', 'evaluation_score': 'average_distance', 'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': None, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'AAA': 0.7667390608953314, 'CGC': 0.40076028262096014, 'ATT': 0.5085852538544149, 'AGC': 0.27799163607995603, 'ACC': 0.4359429381095841, 'ATC': 0.4205725148436632, 'ACA': 0.13023262212317555, 'GGT': 0.33726003930064996, 'AAC': 0.5512262248110238, 'GCG': 0.3563679190097728, 'GGC': 0.4050486219579785, 'TGA': 0.28926761860893596, 'CGA': 0.06388523224887838, 'GTG': 0.3715940695510341, 'TTG': 0.12808970030428077, 'AAG': 0.23326093910466858, 'TTC': 0.4262385670731707, 'TCA': 0.12251736088908215, 'GCA': 0.21281773299119433, 'AAT': 0.4487737751889762, 'GAA': 0.689467997530356, 'CGT': 0.38120248233619, 'TTT': 0.5737614329268292, 'CTG': 0.4975546385450099, 'GTT': 0.2584847661069405, 'GCC': 0.26990242799092135, 'GAT': 0.626856640686903, 'AGG': 0.020162265266098916, 'CAG': 0.6530527088827319, 'GGG': 0.15057187484254547, 'GTC': 0.21619163129169194, 'CTC': 0.10435163186194824, 'TCT': 0.14544780223293646, 'CCC': 0.12361073915372255, 'CAC': 0.4292446655321377, 'GCT': 0.1609119200081115, 'TTA': 0.13009739763522432, 'CTT': 0.10341378016756284, 'ACG': 0.26849903016796667, 'GGA': 0.10711946389882603, 'CCG': 0.527785216925549, 'CAA': 0.3469472911172681, 'ACT': 0.16532540959927366, 'ATA': 0.07084223130192197, 'CAT': 0.5707553344678623, 'AGT': 0.15102374892893225, 'TGC': 0.55651396076928, 'GAG': 0.31053200246964396, 'TCG': 0.15406749964830613, 'GTA': 0.15372953305033354, 'TAC': 0.4315172054223149, 'TCC': 0.14895195222078703, 'GAC': 0.37314335931309695, 'TGT': 0.4434860392307201, 'TGG': 1.0, 'TAT': 0.5684827945776851, 'CCT': 0.15789033208355652, 'CCA': 0.19071371183717192, 'CGG': 0.09749348521075679, 'CTA': 0.03649285148597391, 'TAA': 0.6391064025794565, 'AGA': 0.036496252317115764, 'TAG': 0.07162597881160755}), 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.75, 'AAC': 0.5, 'AAG': 0.23999999999999996, 'AAT': 0.29500000000000004, 'ACA': 0.1525, 'ACC': 0.25, 'ACG': 0.29, 'ACT': 0.14750000000000002, 'AGA': 0.13874999999999998, 'AGC': 0.125, 'AGG': 0.16499999999999998, 'AGT': 0.07375000000000001, 'ATA': 0.10999999999999999, 'ATC': 0.375, 'ATG': 1.0, 'ATT': 0.22125000000000003, 'CAA': 0.27749999999999997, 'CAC': 0.125, 'CAG': 0.32999999999999996, 'CAT': 0.07375000000000001, 'CCA': 0.13874999999999998, 'CCC': 0.125, 'CCG': 0.16499999999999998, 'CCT': 0.07375000000000001, 'CGA': 0.013799999999999993, 'CGC': 0.36, 'CGG': 0.125, 'CGT': 0.5, 'CTA': 0.18, 'CTC': 0.125, 'CTG': 0.54, 'CTT': 0.07375000000000001, 'GAA': 0.5, 'GAC': 0.375, 'GAG': 0.15999999999999998, 'GAT': 0.22125000000000003, 'GCA': 0.375, 'GCC': 0.25, 'GCG': 0.11999999999999998, 'GCT': 0.14750000000000002, 'GGA': 0.13874999999999998, 'GGC': 0.5, 'GGG': 0.16499999999999998, 'GGT': 0.29500000000000004, 'GTA': 0.625, 'GTC': 0.25, 'GTG': 0.19999999999999996, 'GTT': 0.14750000000000002, 'TAC': 0.375, 'TAT': 0.22125000000000003, 'TCA': 0.13874999999999998, 'TCC': 0.25, 'TCG': 0.16499999999999998, 'TCT': 0.14750000000000002, 'TGC': 0.125, 'TGG': 0.125, 'TGT': 0.07375000000000001, 'TTA': 0.13874999999999998, 'TTC': 0.25, 'TTG': 0.16499999999999998, 'TTT': 0.14750000000000002}, 'tai_avg': 0.24958213077316901, 'tai_std': 0.024913119712427708}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': None, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'GAA': 0.6798186496752849, 'AAT': 0.5655317497600131, 'ATA': 0.1287010322849589, 'TTA': 0.198530453110291, 'GAC': 0.35960560391899904, 'CTG': 0.23973441522197322, 'TGG': 1.0, 'AAC': 0.4344682502399869, 'CAA': 0.5131548459124133, 'GCC': 0.20715414603308396, 'CTT': 0.2391679232236727, 'GCT': 0.24752923822568748, 'ATC': 0.366762837741479, 'AAA': 0.7016557540022625, 'AAG': 0.29834424599773746, 'TTG': 0.1592175745811708, 'AGC': 0.22534685577492342, 'CCG': 0.4334091260292722, 'AGT': 0.10611598754150686, 'TTT': 0.6847751982489505, 'GAG': 0.3201813503247151, 'ACT': 0.1614000537105004, 'TCA': 0.23653118484388272, 'ACC': 0.1588786441082565, 'CAC': 0.3265866087018494, 'GGC': 0.3390368073029434, 'GAT': 0.640394396081001, 'ACA': 0.4122430101751559, 'ACG': 0.2674782920060872, 'CCC': 0.0889644363010221, 'AGA': 0.26380392311639106, 'TCC': 0.12712038919920718, 'TAC': 0.34586274736720285, 'CAT': 0.6734133912981506, 'ATT': 0.5045361299735621, 'GCA': 0.2828363712991255, 'TAT': 0.6541372526327971, 'GGG': 0.16211598104043537, 'GTC': 0.25598186049286953, 'CCT': 0.28667299498885185, 'GTT': 0.28439644370189154, 'GCG': 0.26248024444210305, 'TCT': 0.2048572678833432, 'CTC': 0.11219040795754642, 'CCA': 0.19095344268085387, 'GGA': 0.3148341038094681, 'CGA': 0.09855152543710778, 'GTA': 0.19774449549495793, 'GTG': 0.26187720031028106, 'TTC': 0.31522480175104955, 'CGC': 0.2064569601768165, 'CAG': 0.4868451540875866, 'CGG': 0.1555630106168844, 'CGT': 0.18161187196589967, 'CTA': 0.05115922590534585, 'AGG': 0.09401270868690058, 'GGT': 0.18401310784715313, 'TAG': 0.14132712456344587, 'TGC': 0.5460141582025239, 'TCG': 0.10002831475713661, 'TAA': 0.630500582072177, 'TGA': 0.2281722933643772, 'TGT': 0.45398584179747614}), 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.6666666666666666, 'AAC': 0.6666666666666666, 'AAG': 0.2133333333333333, 'AAT': 0.39333333333333337, 'ACA': 0.6666666666666666, 'ACC': 0.16666666666666666, 'ACG': 0.2133333333333333, 'ACT': 0.09833333333333334, 'AGA': 0.18499999999999997, 'AGC': 0.3333333333333333, 'AGG': 0.21999999999999997, 'AGT': 0.19666666666666668, 'ATA': 0.10999999999999999, 'ATC': 0.5, 'ATG': 1.0, 'ATT': 0.29500000000000004, 'CAA': 0.6666666666666666, 'CAC': 0.3333333333333333, 'CAG': 0.2133333333333333, 'CAT': 0.19666666666666668, 'CCA': 0.5, 'CCG': 0.15999999999999998, 'CGA': 0.01839999999999999, 'CGC': 0.48, 'CGG': 0.16666666666666666, 'CGT': 0.6666666666666666, 'CTA': 0.3516666666666666, 'CTC': 0.16666666666666666, 'CTG': 0.2733333333333333, 'CTT': 0.09833333333333334, 'GAA': 1.0, 'GAC': 0.6666666666666666, 'GAG': 0.31999999999999995, 'GAT': 0.39333333333333337, 'GCA': 0.8333333333333334, 'GCC': 0.16666666666666666, 'GCG': 0.2666666666666666, 'GCT': 0.09833333333333334, 'GGA': 0.5, 'GGC': 0.6666666666666666, 'GGG': 0.15999999999999998, 'GGT': 0.39333333333333337, 'GTA': 0.6666666666666666, 'GTC': 0.16666666666666666, 'GTG': 0.2133333333333333, 'GTT': 0.09833333333333334, 'TAC': 0.3333333333333333, 'TAT': 0.19666666666666668, 'TCA': 0.3333333333333333, 'TCC': 0.16666666666666666, 'TCG': 0.10666666666666665, 'TCT': 0.09833333333333334, 'TGC': 0.16666666666666666, 'TGG': 0.16666666666666666, 'TGT': 0.09833333333333334, 'TTA': 0.5183333333333333, 'TTC': 0.5, 'TTG': 0.32666666666666666, 'TTT': 0.29500000000000004}, 'tai_avg': 0.3271366648197227, 'tai_std': 0.028076995641981673}]}, 'orf_debug': {'total_loss': {'C': {'TGT': 0.006148003472222222, 'TGC': 0.013888888888888888}, 'D': {'GAT': 0.08917508680555557, 'GAC': 0.2222222222222222}, 'S': {'TCG': 0.009301388888888889, 'TCT': 0.01008784722222222, 'TCC': 0.013888888888888888, 'AGT': 0.034870920138888886, 'TCA': 0.061743836805555556, 'AGC': 0.06336805555555555}, 'Q': {'CAG': 0.022755555555555546, 'CAA': 0.22360034722222222}, 'M': {'ATG': 0.5}, 'N': {'AAT': 0.09836805555555556, 'AAC': 0.2222222222222222}, 'P': {'CCC': 0.0007999999999999992, 'CCT': 0.004163281249999997, 'CCG': 0.012799999999999995, 'CCA': 0.12534453125}, 'K': {'AAG': 0.15280555555555556, 'AAA': 0.2222222222222222}, '_': {'TAG': 0.029642332908156408, 'TGA': 0.029642332908156408, 'TAA': 0.029642332908156408}, 'T': {'ACC': 0.014688888888888887, 'ACT': 0.014987847222222217, 'ACG': 0.022755555555555546, 'ACA': 0.23167534722222222}, 'F': {'TTT': 0.04876562500000001, 'TTC': 0.125}, 'A': {'GCC': 0.021701388888888888, 'GCT': 0.030712847222222218, 'GCG': 0.06806805555555553, 'GCA': 0.34722222222222227}, 'G': {'GGG': 0.0689125, 'GGT': 0.09836805555555556, 'GGA': 0.19025078125, 'GGC': 0.2222222222222222}, 'I': {'ATA': 0.041162500000000005, 'ATT': 0.055332031250000004, 'ATC': 0.125}, 'L': {'CTG': 0.03735555555555555, 'CTC': 0.1000013888888889, 'CTT': 0.11352925347222226, 'TTG': 0.12366805555555557, 'CTA': 0.12663472222222222, 'TTA': 0.21483550347222222}, 'H': {'CAT': 0.02065217013888889, 'CAC': 0.05555555555555555}, 'R': {'AGG': 0.08031250000000001, 'AGA': 0.08236328125, 'CGG': 0.0842013888888889, 'CGA': 0.1183645, 'CGC': 0.125, 'CGT': 0.2222222222222222}, 'W': {'TGG': 0.013888888888888888}, 'V': {'GTC': 0.0842013888888889, 'GTG': 0.11306805555555556, 'GTT': 0.11883784722222222, 'GTA': 0.2222222222222222}, 'E': {'GAG': 0.10899999999999999, 'GAA': 0.5}, 'Y': {'TAT': 0.031158420138888886, 'TAC': 0.05555555555555555}}, 'optimized_loss': {'C': defaultdict(<class 'float'>, {'TGT': 0.002626562499999999, 'TGC': 0.0}), 'D': defaultdict(<class 'float'>, {'GAT': 0.02363906249999999, 'GAC': 0.0}), 'S': defaultdict(<class 'float'>, {'TCT': 0.010506249999999996, 'TCG': 0.007225000000000003, 'TCA': 0.012376562500000004, 'TCC': 0.0, 'AGC': 0.015625, 'AGT': 0.031064062499999996}), 'Q': defaultdict(<class 'float'>, {'CAA': 0.002756249999999999, 'CAG': 0.0}), 'M': defaultdict(<class 'float'>, {'ATG': 0.0}), 'N': defaultdict(<class 'float'>, {'AAC': 0.0, 'AAT': 0.042024999999999986}), 'P': defaultdict(<class 'float'>, {'CCT': 0.008326562499999994, 'CCG': 0.0, 'CCA': 0.0006890624999999998, 'CCC': 0.0015999999999999983}), 'K': defaultdict(<class 'float'>, {'AAG': 0.2601, 'AAA': 0.0}), '_': defaultdict(<class 'float'>, {'TAG': 0.059284665816312816, 'TGA': 0.059284665816312816, 'TAA': 0.059284665816312816}), 'T': defaultdict(<class 'float'>, {'ACC': 0.0015999999999999983, 'ACA': 0.018906249999999996, 'ACG': 0.0, 'ACT': 0.020306249999999988}), 'F': defaultdict(<class 'float'>, {'TTT': 0.010506249999999996, 'TTC': 0.0}), 'A': defaultdict(<class 'float'>, {'GCA': 0.0, 'GCC': 0.015625, 'GCG': 0.065025, 'GCT': 0.05175624999999999}), 'G': defaultdict(<class 'float'>, {'GGT': 0.042024999999999986, 'GGG': 0.11222500000000002, 'GGA': 0.1305015625, 'GGC': 0.0}), 'I': defaultdict(<class 'float'>, {'ATC': 0.0, 'ATA': 0.07022500000000001, 'ATT': 0.02363906249999999}), 'L': defaultdict(<class 'float'>, {'TTA': 0.16100156250000003, 'TTG': 0.14062500000000006, 'CTC': 0.17222500000000002, 'CTT': 0.21738906250000006, 'CTG': 0.0, 'CTA': 0.12960000000000002}), 'H': defaultdict(<class 'float'>, {'CAT': 0.002626562499999999, 'CAC': 0.0}), 'R': defaultdict(<class 'float'>, {'CGA': 0.23639044, 'CGC': 0.019600000000000003, 'CGG': 0.140625, 'CGT': 0.0, 'AGG': 0.11222500000000002, 'AGA': 0.1305015625}), 'W': defaultdict(<class 'float'>, {'TGG': 0.0}), 'V': defaultdict(<class 'float'>, {'GTA': 0.0, 'GTC': 0.140625, 'GTG': 0.18062500000000004, 'GTT': 0.22800625}), 'E': defaultdict(<class 'float'>, {'GAG': 0.11560000000000002, 'GAA': 0.0}), 'Y': defaultdict(<class 'float'>, {'TAT': 0.02363906249999999, 'TAC': 0.0})}, 'deoptimized_loss': {'C': defaultdict(<class 'float'>, {'TGT': 0.009669444444444446, 'TGC': 0.027777777777777776}), 'D': defaultdict(<class 'float'>, {'GAT': 0.15471111111111113, 'GAC': 0.4444444444444444}), 'S': defaultdict(<class 'float'>, {'TCT': 0.009669444444444446, 'TCG': 0.011377777777777773, 'TCA': 0.1111111111111111, 'TCC': 0.027777777777777776, 'AGC': 0.1111111111111111, 'AGT': 0.03867777777777778}), 'Q': defaultdict(<class 'float'>, {'CAA': 0.4444444444444444, 'CAG': 0.04551111111111109}), 'M': defaultdict(<class 'float'>, {'ATG': 1.0}), 'N': defaultdict(<class 'float'>, {'AAC': 0.4444444444444444, 'AAT': 0.15471111111111113}), 'P': defaultdict(<class 'float'>, {'CCT': 0, 'CCG': 0.02559999999999999, 'CCA': 0.25, 'CCC': 0}), 'K': defaultdict(<class 'float'>, {'AAG': 0.04551111111111109, 'AAA': 0.4444444444444444}), '_': defaultdict(<class 'float'>, {'TAG': 0, 'TGA': 0, 'TAA': 0}), 'T': defaultdict(<class 'float'>, {'ACC': 0.027777777777777776, 'ACA': 0.4444444444444444, 'ACG': 0.04551111111111109, 'ACT': 0.009669444444444446}), 'F': defaultdict(<class 'float'>, {'TTT': 0.08702500000000002, 'TTC': 0.25}), 'A': defaultdict(<class 'float'>, {'GCA': 0.6944444444444445, 'GCC': 0.027777777777777776, 'GCG': 0.07111111111111108, 'GCT': 0.009669444444444446}), 'G': defaultdict(<class 'float'>, {'GGT': 0.15471111111111113, 'GGG': 0.02559999999999999, 'GGA': 0.25, 'GGC': 0.4444444444444444}), 'I': defaultdict(<class 'float'>, {'ATC': 0.25, 'ATA': 0.012099999999999998, 'ATT': 0.08702500000000002}), 'L': defaultdict(<class 'float'>, {'TTA': 0.2686694444444444, 'TTG': 0.1067111111111111, 'CTC': 0.027777777777777776, 'CTT': 0.009669444444444446, 'CTG': 0.0747111111111111, 'CTA': 0.12366944444444442}), 'H': defaultdict(<class 'float'>, {'CAT': 0.03867777777777778, 'CAC': 0.1111111111111111}), 'R': defaultdict(<class 'float'>, {'CGA': 0.00033855999999999963, 'CGC': 0.2304, 'CGG': 0.027777777777777776, 'CGT': 0.4444444444444444, 'AGG': 0.04839999999999999, 'AGA': 0.03422499999999999}), 'W': defaultdict(<class 'float'>, {'TGG': 0.027777777777777776}), 'V': defaultdict(<class 'float'>, {'GTA': 0.4444444444444444, 'GTC': 0.027777777777777776, 'GTG': 0.04551111111111109, 'GTT': 0.009669444444444446}), 'E': defaultdict(<class 'float'>, {'GAG': 0.10239999999999996, 'GAA': 1.0}), 'Y': defaultdict(<class 'float'>, {'TAT': 0.03867777777777778, 'TAC': 0.1111111111111111})}}, 'orf': {'orf_module_input_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'optimized_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGCTTTAAGGTCCATATGGAGGGGTCGGTCAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCCTATGAGGGGACCCAGACCGCCAAGCTGAAGGTCACCAAGGGGGGTCCCCTGCCCTTTGCCTGGGATATTCTGTCGCCCCAGTTTATGTATGGGTCGAAGGCCTATGTCAAGCATCCCGCCGATATTCCCGATTATCTGAAGCTGTCGTTTCCCGAGGGGTTTAAGTGGGAGCGCGTCATGAATTTTGAGGATGGGGGTGTCGTGACCGTCACCCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTCAAGCTGCGCGGGACCAATTTTCCCTCGGATGGGCCCGTCATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGTCTGAGCGCATGTATCCCGAGGATGGGGCCCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCCGAGGTCAAGACCACTTATAAGGCCAAGAAACCCGTCCAGCTGCCCGGGGCCTATAATGTCAATATTAAGCTGGATATTACCTCGCATAATGAGGATTATACCATTGTCGAGCAGTATGAGCGCGCCGAGGGGCGCCATTCGACCGGGGGTATGGATGAGCTGTATAAGTAA', 'aa_to_loss_mapping': {'C': {'TGT': 0.006148003472222222, 'TGC': 0.013888888888888888}, 'D': {'GAT': 0.08917508680555557, 'GAC': 0.2222222222222222}, 'S': {'TCG': 0.009301388888888889, 'TCT': 0.01008784722222222, 'TCC': 0.013888888888888888, 'AGT': 0.034870920138888886, 'TCA': 0.061743836805555556, 'AGC': 0.06336805555555555}, 'Q': {'CAG': 0.022755555555555546, 'CAA': 0.22360034722222222}, 'M': {'ATG': 0.5}, 'N': {'AAT': 0.09836805555555556, 'AAC': 0.2222222222222222}, 'P': {'CCC': 0.0007999999999999992, 'CCT': 0.004163281249999997, 'CCG': 0.012799999999999995, 'CCA': 0.12534453125}, 'K': {'AAG': 0.15280555555555556, 'AAA': 0.2222222222222222}, '_': {'TAG': 0.029642332908156408, 'TGA': 0.029642332908156408, 'TAA': 0.029642332908156408}, 'T': {'ACC': 0.014688888888888887, 'ACT': 0.014987847222222217, 'ACG': 0.022755555555555546, 'ACA': 0.23167534722222222}, 'F': {'TTT': 0.04876562500000001, 'TTC': 0.125}, 'A': {'GCC': 0.021701388888888888, 'GCT': 0.030712847222222218, 'GCG': 0.06806805555555553, 'GCA': 0.34722222222222227}, 'G': {'GGG': 0.0689125, 'GGT': 0.09836805555555556, 'GGA': 0.19025078125, 'GGC': 0.2222222222222222}, 'I': {'ATA': 0.041162500000000005, 'ATT': 0.055332031250000004, 'ATC': 0.125}, 'L': {'CTG': 0.03735555555555555, 'CTC': 0.1000013888888889, 'CTT': 0.11352925347222226, 'TTG': 0.12366805555555557, 'CTA': 0.12663472222222222, 'TTA': 0.21483550347222222}, 'H': {'CAT': 0.02065217013888889, 'CAC': 0.05555555555555555}, 'R': {'AGG': 0.08031250000000001, 'AGA': 0.08236328125, 'CGG': 0.0842013888888889, 'CGA': 0.1183645, 'CGC': 0.125, 'CGT': 0.2222222222222222}, 'W': {'TGG': 0.013888888888888888}, 'V': {'GTC': 0.0842013888888889, 'GTG': 0.11306805555555556, 'GTT': 0.11883784722222222, 'GTA': 0.2222222222222222}, 'E': {'GAG': 0.10899999999999999, 'GAA': 0.5}, 'Y': {'TAT': 0.031158420138888886, 'TAC': 0.05555555555555555}}, 'run_time': 0.8440000000409782}, 'evaluation': [{'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'tai_initial_score': 0.2816376469779175, 'tai_final_score': 0.21684652426894746, 'dist_score': -2.6006828312493315}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'tai_initial_score': 0.33696708001007003, 'tai_final_score': 0.23716199528257118, 'dist_score': -3.554692460694296}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGCTTTAAGGTCCATATGGAGGGGTCGGTCAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCCTATGAGGGGACCCAGACCGCCAAGCTGAAGGTCACCAAGGGGGGTCCCCTGCCCTTTGCCTGGGATATTCTGTCGCCCCAGTTTATGTATGGGTCGAAGGCCTATGTCAAGCATCCCGCCGATATTCCCGATTATCTGAAGCTGTCGTTTCCCGAGGGGTTTAAGTGGGAGCGCGTCATGAATTTTGAGGATGGGGGTGTCGTGACCGTCACCCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTCAAGCTGCGCGGGACCAATTTTCCCTCGGATGGGCCCGTCATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGTCTGAGCGCATGTATCCCGAGGATGGGGCCCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCCGAGGTCAAGACCACTTATAAGGCCAAGAAACCCGTCCAGCTGCCCGGGGCCTATAATGTCAATATTAAGCTGGATATTACCTCGCATAATGAGGATTATACCATTGTCGAGCAGTATGAGCGCGCCGAGGGGCGCCATTCGACCGGGGGTATGGATGAGCTGTATAAGTAA', 'average_distance_score': 0.47700481472248213, 'weakest_link_score': 0.47700481472248213, 'ratio_score': 1.0256286534297445, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTTATGCGCTTTAAGGTCCATATGGAGGGGTCGGTCAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCCTATGAGGGGACCCAGACCGCCAAGCTGAAGGTCACCAAGGGGGGTCCCCTGCCCTTTGCCTGGGATATTCTGTCGCCCCAGTTTATGTATGGGTCGAAGGCCTATGTCAAGCATCCCGCCGATATTCCCGATTATCTGAAGCTGTCGTTTCCCGAGGGGTTTAAGTGGGAGCGCGTCATGAATTTTGAGGATGGGGGTGTCGTGACCGTCACCCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTCAAGCTGCGCGGGACCAATTTTCCCTCGGATGGGCCCGTCATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGTCTGAGCGCATGTATCCCGAGGATGGGGCCCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCCGAGGTCAAGACCACTTATAAGGCCAAGAAACCCGTCCAGCTGCCCGGGGCCTATAATGTCAATATTAAGCTGGATATTACCTCGCATAATGAGGATTATACCATTGTCGAGCAGTATGAGCGCGCCGAGGGGCGCCATTCGACCGGGGGTATGGATGAGCTGTATAAGTAA'}], 'final_evaluation': {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGCTTTAAGGTCCATATGGAGGGGTCGGTCAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCCTATGAGGGGACCCAGACCGCCAAGCTGAAGGTCACCAAGGGGGGTCCCCTGCCCTTTGCCTGGGATATTCTGTCGCCCCAGTTTATGTATGGGTCGAAGGCCTATGTCAAGCATCCCGCCGATATTCCCGATTATCTGAAGCTGTCGTTTCCCGAGGGGTTTAAGTGGGAGCGCGTCATGAATTTTGAGGATGGGGGTGTCGTGACCGTCACCCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTCAAGCTGCGCGGGACCAATTTTCCCTCGGATGGGCCCGTCATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGTCTGAGCGCATGTATCCCGAGGATGGGGCCCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCCGAGGTCAAGACCACTTATAAGGCCAAGAAACCCGTCCAGCTGCCCGGGGCCTATAATGTCAATATTAAGCTGGATATTACCTCGCATAATGAGGATTATACCATTGTCGAGCAGTATGAGCGCGCCGAGGGGCGCCATTCGACCGGGGGTATGGATGAGCTGTATAAGTAA', 'average_distance_score': 0.47700481472248213, 'weakest_link_score': 0.47700481472248213, 'ratio_score': 1.0256286534297445}, 'zip_output_file_path': 'results\\\\mcherry-wanted-constraint-init-external-0.1\\\\tAI_single_codon_diff_ecoli_opt_True_XQ2U\\\\communique_results.zip'}\n",
      "\n",
      "##########################\n",
      "# User Input #\n",
      "##########################\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is:\n",
      "ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "------------------------------------------\n",
      "Parsing information for Bacillus subtilis:\n",
      "------------------------------------------\n",
      "Organism is defined as wanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Bacillus subtilis. Calculating tAI profile.\n",
      "tai_std=0.028076995641981673, tai_avg=0.3271366648197227\n",
      "------------------------------------------\n",
      "Parsing information for Escherichia coli:\n",
      "------------------------------------------\n",
      "Organism is defined as unwanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Escherichia coli. Calculating tAI profile.\n",
      "tai_std=0.024913119712427708, tai_avg=0.24958213077316901\n",
      "-----------------------------\n",
      "Normalized Prioritization Weights\n",
      "-----------------------------\n",
      "Bacillus subtilis : 1.0\n",
      "Escherichia coli : 1.0\n",
      "Total input processing time: 28.604941368103027\n",
      "\n",
      "##########################\n",
      "# Initiation #\n",
      "##########################\n",
      "Running initiation optimization InitiationOptimizationMethod.external_module on 15 codons at the start of the ORF\n",
      "Taking optimized sequence from configuration value config.INITIATION.EXTERNAL_INITIATION_ORF\n",
      "Optimized sequence is: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "Optimization method used is: ORFOptimizationMethod.single_codon_diff\n",
      "Loss summary for: TGT\n",
      "loss for Bacillus subtilis: 0.004669444444444442\n",
      "loss for Escherichia coli: 0.005439062500000001\n",
      "Loss summary for: TGC\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.015625\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGT': 0.005054253472222221, 'TGC': 0.0078125})\n",
      "Loss summary for: GAT\n",
      "loss for Bacillus subtilis: 0.07471111111111108\n",
      "loss for Escherichia coli: 0.04895156250000001\n",
      "Loss summary for: GAC\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.140625\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAT': 0.06183133680555554, 'GAC': 0.0703125})\n",
      "Loss summary for: TCT\n",
      "loss for Bacillus subtilis: 0.055224999999999996\n",
      "loss for Escherichia coli: 0.021756250000000005\n",
      "Loss summary for: TCG\n",
      "loss for Bacillus subtilis: 0.051377777777777786\n",
      "loss for Escherichia coli: 0.027224999999999992\n",
      "Loss summary for: TCA\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.019251562499999996\n",
      "Loss summary for: TCC\n",
      "loss for Bacillus subtilis: 0.027777777777777776\n",
      "loss for Escherichia coli: 0.0625\n",
      "Loss summary for: AGC\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.015625\n",
      "Loss summary for: AGT\n",
      "loss for Bacillus subtilis: 0.01867777777777777\n",
      "loss for Escherichia coli: 0.005439062500000001\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TCT': 0.038490625, 'TCG': 0.039301388888888886, 'TCA': 0.009625781249999998, 'TCC': 0.04513888888888889, 'AGC': 0.0078125, 'AGT': 0.012058420138888885})\n",
      "Loss summary for: CAA\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.07700624999999998\n",
      "Loss summary for: CAG\n",
      "loss for Bacillus subtilis: 0.20551111111111114\n",
      "loss for Escherichia coli: 0.10889999999999997\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAA': 0.03850312499999999, 'CAG': 0.15720555555555554})\n",
      "Loss summary for: ATG\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 1.0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATG': 0.5})\n",
      "Loss summary for: AAC\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.25\n",
      "Loss summary for: AAT\n",
      "loss for Bacillus subtilis: 0.07471111111111108\n",
      "loss for Escherichia coli: 0.08702500000000002\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAC': 0.125, 'AAT': 0.08086805555555554})\n",
      "Loss summary for: CCT\n",
      "loss for Bacillus subtilis: 0.25\n",
      "loss for Escherichia coli: 0.005439062500000001\n",
      "Loss summary for: CCG\n",
      "loss for Bacillus subtilis: 0.11560000000000002\n",
      "loss for Escherichia coli: 0.027224999999999992\n",
      "Loss summary for: CCA\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.019251562499999996\n",
      "Loss summary for: CCC\n",
      "loss for Bacillus subtilis: 0.25\n",
      "loss for Escherichia coli: 0.015625\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CCT': 0.12771953125, 'CCG': 0.0714125, 'CCA': 0.009625781249999998, 'CCC': 0.1328125})\n",
      "Loss summary for: AAG\n",
      "loss for Bacillus subtilis: 0.20551111111111114\n",
      "loss for Escherichia coli: 0.057599999999999985\n",
      "Loss summary for: AAA\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.5625\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAG': 0.13155555555555556, 'AAA': 0.28125})\n",
      "Loss summary for: TAG\n",
      "loss for Bacillus subtilis: 0.11644751380637751\n",
      "loss for Escherichia coli: 0\n",
      "Loss summary for: TGA\n",
      "loss for Bacillus subtilis: 0.11644751380637751\n",
      "loss for Escherichia coli: 0\n",
      "Loss summary for: TAA\n",
      "loss for Bacillus subtilis: 0.11644751380637751\n",
      "loss for Escherichia coli: 0\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAG': 0.058223756903188754, 'TGA': 0.058223756903188754, 'TAA': 0.058223756903188754})\n",
      "Loss summary for: ACC\n",
      "loss for Bacillus subtilis: 0.25\n",
      "loss for Escherichia coli: 0.0625\n",
      "Loss summary for: ACA\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.02325625\n",
      "Loss summary for: ACG\n",
      "loss for Bacillus subtilis: 0.20551111111111114\n",
      "loss for Escherichia coli: 0.0841\n",
      "Loss summary for: ACT\n",
      "loss for Bacillus subtilis: 0.3230027777777777\n",
      "loss for Escherichia coli: 0.021756250000000005\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ACC': 0.15625, 'ACA': 0.011628125, 'ACG': 0.14480555555555558, 'ACT': 0.17237951388888884})\n",
      "Loss summary for: TTT\n",
      "loss for Bacillus subtilis: 0.042024999999999986\n",
      "loss for Escherichia coli: 0.021756250000000005\n",
      "Loss summary for: TTC\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.0625\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTT': 0.03189062499999999, 'TTC': 0.03125})\n",
      "Loss summary for: GCA\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.140625\n",
      "Loss summary for: GCC\n",
      "loss for Bacillus subtilis: 0.44444444444444453\n",
      "loss for Escherichia coli: 0.0625\n",
      "Loss summary for: GCG\n",
      "loss for Bacillus subtilis: 0.32111111111111124\n",
      "loss for Escherichia coli: 0.014399999999999996\n",
      "Loss summary for: GCT\n",
      "loss for Bacillus subtilis: 0.540225\n",
      "loss for Escherichia coli: 0.021756250000000005\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GCA': 0.0703125, 'GCC': 0.25347222222222227, 'GCG': 0.1677555555555556, 'GCT': 0.280990625})\n",
      "Loss summary for: GGT\n",
      "loss for Bacillus subtilis: 0.07471111111111108\n",
      "loss for Escherichia coli: 0.08702500000000002\n",
      "Loss summary for: GGG\n",
      "loss for Bacillus subtilis: 0.25671111111111106\n",
      "loss for Escherichia coli: 0.027224999999999992\n",
      "Loss summary for: GGA\n",
      "loss for Bacillus subtilis: 0.027777777777777766\n",
      "loss for Escherichia coli: 0.019251562499999996\n",
      "Loss summary for: GGC\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.25\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GGT': 0.08086805555555554, 'GGG': 0.14196805555555553, 'GGA': 0.02351467013888888, 'GGC': 0.125})\n",
      "Loss summary for: ATC\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.140625\n",
      "Loss summary for: ATA\n",
      "loss for Bacillus subtilis: 0.1521\n",
      "loss for Escherichia coli: 0.012099999999999998\n",
      "Loss summary for: ATT\n",
      "loss for Bacillus subtilis: 0.042024999999999986\n",
      "loss for Escherichia coli: 0.04895156250000001\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATC': 0.0703125, 'ATA': 0.0821, 'ATT': 0.04548828125})\n",
      "Loss summary for: TTA\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.019251562499999996\n",
      "Loss summary for: TTG\n",
      "loss for Bacillus subtilis: 0.03673611111111111\n",
      "loss for Escherichia coli: 0.027224999999999992\n",
      "Loss summary for: CTC\n",
      "loss for Bacillus subtilis: 0.12366944444444446\n",
      "loss for Escherichia coli: 0.015625\n",
      "Loss summary for: CTT\n",
      "loss for Bacillus subtilis: 0.17639999999999997\n",
      "loss for Escherichia coli: 0.005439062500000001\n",
      "Loss summary for: CTG\n",
      "loss for Bacillus subtilis: 0.060024999999999995\n",
      "loss for Escherichia coli: 0.2916\n",
      "Loss summary for: CTA\n",
      "loss for Bacillus subtilis: 0.027777777777777783\n",
      "loss for Escherichia coli: 0.0324\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTA': 0.009625781249999998, 'TTG': 0.031980555555555554, 'CTC': 0.06964722222222222, 'CTT': 0.09091953124999999, 'CTG': 0.1758125, 'CTA': 0.03008888888888889})\n",
      "Loss summary for: CAT\n",
      "loss for Bacillus subtilis: 0.01867777777777777\n",
      "loss for Escherichia coli: 0.005439062500000001\n",
      "Loss summary for: CAC\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.015625\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAT': 0.012058420138888885, 'CAC': 0.0078125})\n",
      "Loss summary for: CGA\n",
      "loss for Bacillus subtilis: 0.4202496711111111\n",
      "loss for Escherichia coli: 0.0001904399999999998\n",
      "Loss summary for: CGC\n",
      "loss for Bacillus subtilis: 0.034844444444444435\n",
      "loss for Escherichia coli: 0.1296\n",
      "Loss summary for: CGG\n",
      "loss for Bacillus subtilis: 0.25\n",
      "loss for Escherichia coli: 0.015625\n",
      "Loss summary for: CGT\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.25\n",
      "Loss summary for: AGG\n",
      "loss for Bacillus subtilis: 0.1995111111111111\n",
      "loss for Escherichia coli: 0.027224999999999992\n",
      "Loss summary for: AGA\n",
      "loss for Bacillus subtilis: 0.2320027777777778\n",
      "loss for Escherichia coli: 0.019251562499999996\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CGA': 0.21022005555555556, 'CGC': 0.08222222222222221, 'CGG': 0.1328125, 'CGT': 0.125, 'AGG': 0.11336805555555556, 'AGA': 0.1256271701388889})\n",
      "Loss summary for: TGG\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.015625\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGG': 0.0078125})\n",
      "Loss summary for: GTA\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.390625\n",
      "Loss summary for: GTC\n",
      "loss for Bacillus subtilis: 0.25\n",
      "loss for Escherichia coli: 0.0625\n",
      "Loss summary for: GTG\n",
      "loss for Bacillus subtilis: 0.20551111111111114\n",
      "loss for Escherichia coli: 0.03999999999999998\n",
      "Loss summary for: GTT\n",
      "loss for Bacillus subtilis: 0.3230027777777777\n",
      "loss for Escherichia coli: 0.021756250000000005\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GTA': 0.1953125, 'GTC': 0.15625, 'GTG': 0.12275555555555556, 'GTT': 0.17237951388888884})\n",
      "Loss summary for: GAG\n",
      "loss for Bacillus subtilis: 0.4624000000000001\n",
      "loss for Escherichia coli: 0.02559999999999999\n",
      "Loss summary for: GAA\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.25\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAG': 0.24400000000000005, 'GAA': 0.125})\n",
      "Loss summary for: TAT\n",
      "loss for Bacillus subtilis: 0.01867777777777777\n",
      "loss for Escherichia coli: 0.04895156250000001\n",
      "Loss summary for: TAC\n",
      "loss for Bacillus subtilis: 0.0\n",
      "loss for Escherichia coli: 0.140625\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAT': 0.03381467013888889, 'TAC': 0.0703125})\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31522480175104955]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2064569601768165]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31522480175104955]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26187720031028106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3265866087018494]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.22534685577492342]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26187720031028106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5655317497600131]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3265866087018494]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31522480175104955]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5045361299735621]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2064569601768165]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19095344268085387]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.198530453110291]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26187720031028106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.18401310784715313]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19095344268085387]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.198530453110291]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19095344268085387]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31522480175104955]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.640394396081001]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5045361299735621]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.198530453110291]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.22534685577492342]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19095344268085387]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31522480175104955]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.22534685577492342]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26187720031028106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3265866087018494]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19095344268085387]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.640394396081001]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5045361299735621]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19095344268085387]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.640394396081001]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.198530453110291]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.198530453110291]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.22534685577492342]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31522480175104955]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19095344268085387]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31522480175104955]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2064569601768165]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26187720031028106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5655317497600131]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31522480175104955]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.640394396081001]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.18401310784715313]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26187720031028106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26187720031028106]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.25598186049286953]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26187720031028106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.640394396081001]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.22534685577492342]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.22534685577492342]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23653118484388272]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.198530453110291]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.640394396081001]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31522480175104955]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5045361299735621]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26187720031028106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.198530453110291]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2064569601768165]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5655317497600131]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.31522480175104955]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19095344268085387]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.22534685577492342]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.640394396081001]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19095344268085387]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26187720031028106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7016557540022625]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.22534685577492342]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.22534685577492342]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.23653118484388272]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2064569601768165]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19095344268085387]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.640394396081001]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.198530453110291]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5045361299735621]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2064569601768165]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.198530453110291]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.198530453110291]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.640394396081001]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.18401310784715313]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3265866087018494]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.640394396081001]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26187720031028106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2674782920060872]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.7016557540022625]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19095344268085387]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26187720031028106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.198530453110291]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.19095344268085387]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5655317497600131]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26187720031028106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5655317497600131]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5045361299735621]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.198530453110291]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.640394396081001]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5045361299735621]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.22534685577492342]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3265866087018494]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5655317497600131]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.640394396081001]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5045361299735621]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.26187720031028106]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.5131548459124133]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2064569601768165]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2828363712991255]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.2064569601768165]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3265866087018494]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.22534685577492342]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.4122430101751559]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.3148341038094681]\n",
      "Choosing second-best option for amino-acid that is part of repeating sub-sequence.\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.18401310784715313]\n",
      "Candidate optimal codon frequencies in wanted organisms: [1.0]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.640394396081001]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6798186496752849]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.198530453110291]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.6541372526327971]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.29834424599773746]\n",
      "Candidate optimal codon frequencies in wanted organisms: [0.14132712456344587]\n",
      "\n",
      "##########################\n",
      "# EVALUATION #\n",
      "##########################\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTCATGCGCTTCAAGGTGCACATGGAAGGAAGCGTGAATGGACACGAATTCGAAATTGAAGGAGAAGGAGAAGGACGCCCATATGAAGGAACACAAACAGCAAAGTTAAAGGTGACAAAGGGAGGTCCATTACCATTCGCATGGGATATTTTAAGCCCACAATTCATGTATGGAAGCAAGGCATATGTGAAGCACCCAGCAGATATTCCAGATTATTTAAAGTTAAGCTTCCCAGAAGGATTCAAGTGGGAACGCGTGATGAATTTCGAAGATGGAGGTGTGGTCACAGTGACACAAGATAGCTCATTACAAGATGGAGAATTCATTTATAAGGTGAAGTTACGCGGAACAAATTTCCCAAGCGATGGACCAGTGATGCAAAAGAAAACAATGGGATGGGAAGCAAGCTCAGAACGCATGTATCCAGAAGATGGAGCATTAAAGGGAGAAATTAAGCAACGCTTAAAGTTAAAGGATGGAGGTCACTATGATGCAGAAGTGAAGACAACGTATAAGGCAAAGAAACCAGTGCAATTACCAGGAGCATATAATGTGAATATTAAGTTAGATATTACAAGCCACAATGAAGATTATACAATTGTGGAACAATATGAACGCGCAGAAGGACGCCACAGCACAGGAGGTATGGATGAATTATATAAGTAA\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAGGTGCACATGGAAGGAAGCGTGAATGGACACGAATTCGAAATTGAAGGAGAAGGAGAAGGACGCCCATATGAAGGAACACAAACAGCAAAGTTAAAGGTGACAAAGGGAGGTCCATTACCATTCGCATGGGATATTTTAAGCCCACAATTCATGTATGGAAGCAAGGCATATGTGAAGCACCCAGCAGATATTCCAGATTATTTAAAGTTAAGCTTCCCAGAAGGATTCAAGTGGGAACGCGTGATGAATTTCGAAGATGGAGGTGTGGTCACAGTGACACAAGATAGCTCATTACAAGATGGAGAATTCATTTATAAGGTGAAGTTACGCGGAACAAATTTCCCAAGCGATGGACCAGTGATGCAAAAGAAAACAATGGGATGGGAAGCAAGCTCAGAACGCATGTATCCAGAAGATGGAGCATTAAAGGGAGAAATTAAGCAACGCTTAAAGTTAAAGGATGGAGGTCACTATGATGCAGAAGTGAAGACAACGTATAAGGCAAAGAAACCAGTGCAATTACCAGGAGCATATAATGTGAATATTAAGTTAGATATTACAAGCCACAATGAAGATTATACAATTGTGGAACAATATGAACGCGCAGAAGGACGCCACAGCACAGGAGGTATGGATGAATTATATAAGTAA', 'average_distance_score': 2.8288931060000304, 'weakest_link_score': 2.8288931060000304, 'ratio_score': 1.1355593244176034}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAGGTGCACATGGAAGGAAGCGTGAATGGACACGAATTCGAAATTGAAGGAGAAGGAGAAGGACGCCCATATGAAGGAACACAAACAGCAAAGTTAAAGGTGACAAAGGGAGGTCCATTACCATTCGCATGGGATATTTTAAGCCCACAATTCATGTATGGAAGCAAGGCATATGTGAAGCACCCAGCAGATATTCCAGATTATTTAAAGTTAAGCTTCCCAGAAGGATTCAAGTGGGAACGCGTGATGAATTTCGAAGATGGAGGTGTGGTCACAGTGACACAAGATAGCTCATTACAAGATGGAGAATTCATTTATAAGGTGAAGTTACGCGGAACAAATTTCCCAAGCGATGGACCAGTGATGCAAAAGAAAACAATGGGATGGGAAGCAAGCTCAGAACGCATGTATCCAGAAGATGGAGCATTAAAGGGAGAAATTAAGCAACGCTTAAAGTTAAAGGATGGAGGTCACTATGATGCAGAAGTGAAGACAACGTATAAGGCAAAGAAACCAGTGCAATTACCAGGAGCATATAATGTGAATATTAAGTTAGATATTACAAGCCACAATGAAGATTATACAATTGTGGAACAATATGAACGCGCAGAAGGACGCCACAGCACAGGAGGTATGGATGAATTATATAAGTAA', 'average_distance_score': 2.8288931060000304, 'weakest_link_score': 2.8288931060000304, 'ratio_score': 1.1355593244176034}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry-wanted-constraint-init-external-0.1\\tAI_single_codon_diff_ecoli_opt_False_klLD\n",
      "Final output: {'module_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'orf_optimization_method': 'single_codon_diff', 'orf_optimization_cub_index': 'tAI', 'initiation_optimization_method': 'external', 'evaluation_score': 'average_distance', 'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': None, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'GAA': 0.6798186496752849, 'AAT': 0.5655317497600131, 'ATA': 0.1287010322849589, 'TTA': 0.198530453110291, 'GAC': 0.35960560391899904, 'CTG': 0.23973441522197322, 'TGG': 1.0, 'AAC': 0.4344682502399869, 'CAA': 0.5131548459124133, 'GCC': 0.20715414603308396, 'CTT': 0.2391679232236727, 'GCT': 0.24752923822568748, 'ATC': 0.366762837741479, 'AAA': 0.7016557540022625, 'AAG': 0.29834424599773746, 'TTG': 0.1592175745811708, 'AGC': 0.22534685577492342, 'CCG': 0.4334091260292722, 'AGT': 0.10611598754150686, 'TTT': 0.6847751982489505, 'GAG': 0.3201813503247151, 'ACT': 0.1614000537105004, 'TCA': 0.23653118484388272, 'ACC': 0.1588786441082565, 'CAC': 0.3265866087018494, 'GGC': 0.3390368073029434, 'GAT': 0.640394396081001, 'ACA': 0.4122430101751559, 'ACG': 0.2674782920060872, 'CCC': 0.0889644363010221, 'AGA': 0.26380392311639106, 'TCC': 0.12712038919920718, 'TAC': 0.34586274736720285, 'CAT': 0.6734133912981506, 'ATT': 0.5045361299735621, 'GCA': 0.2828363712991255, 'TAT': 0.6541372526327971, 'GGG': 0.16211598104043537, 'GTC': 0.25598186049286953, 'CCT': 0.28667299498885185, 'GTT': 0.28439644370189154, 'GCG': 0.26248024444210305, 'TCT': 0.2048572678833432, 'CTC': 0.11219040795754642, 'CCA': 0.19095344268085387, 'GGA': 0.3148341038094681, 'CGA': 0.09855152543710778, 'GTA': 0.19774449549495793, 'GTG': 0.26187720031028106, 'TTC': 0.31522480175104955, 'CGC': 0.2064569601768165, 'CAG': 0.4868451540875866, 'CGG': 0.1555630106168844, 'CGT': 0.18161187196589967, 'CTA': 0.05115922590534585, 'AGG': 0.09401270868690058, 'GGT': 0.18401310784715313, 'TAG': 0.14132712456344587, 'TGC': 0.5460141582025239, 'TCG': 0.10002831475713661, 'TAA': 0.630500582072177, 'TGA': 0.2281722933643772, 'TGT': 0.45398584179747614}), 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.6666666666666666, 'AAC': 0.6666666666666666, 'AAG': 0.2133333333333333, 'AAT': 0.39333333333333337, 'ACA': 0.6666666666666666, 'ACC': 0.16666666666666666, 'ACG': 0.2133333333333333, 'ACT': 0.09833333333333334, 'AGA': 0.18499999999999997, 'AGC': 0.3333333333333333, 'AGG': 0.21999999999999997, 'AGT': 0.19666666666666668, 'ATA': 0.10999999999999999, 'ATC': 0.5, 'ATG': 1.0, 'ATT': 0.29500000000000004, 'CAA': 0.6666666666666666, 'CAC': 0.3333333333333333, 'CAG': 0.2133333333333333, 'CAT': 0.19666666666666668, 'CCA': 0.5, 'CCG': 0.15999999999999998, 'CGA': 0.01839999999999999, 'CGC': 0.48, 'CGG': 0.16666666666666666, 'CGT': 0.6666666666666666, 'CTA': 0.3516666666666666, 'CTC': 0.16666666666666666, 'CTG': 0.2733333333333333, 'CTT': 0.09833333333333334, 'GAA': 1.0, 'GAC': 0.6666666666666666, 'GAG': 0.31999999999999995, 'GAT': 0.39333333333333337, 'GCA': 0.8333333333333334, 'GCC': 0.16666666666666666, 'GCG': 0.2666666666666666, 'GCT': 0.09833333333333334, 'GGA': 0.5, 'GGC': 0.6666666666666666, 'GGG': 0.15999999999999998, 'GGT': 0.39333333333333337, 'GTA': 0.6666666666666666, 'GTC': 0.16666666666666666, 'GTG': 0.2133333333333333, 'GTT': 0.09833333333333334, 'TAC': 0.3333333333333333, 'TAT': 0.19666666666666668, 'TCA': 0.3333333333333333, 'TCC': 0.16666666666666666, 'TCG': 0.10666666666666665, 'TCT': 0.09833333333333334, 'TGC': 0.16666666666666666, 'TGG': 0.16666666666666666, 'TGT': 0.09833333333333334, 'TTA': 0.5183333333333333, 'TTC': 0.5, 'TTG': 0.32666666666666666, 'TTT': 0.29500000000000004}, 'tai_avg': 0.3271366648197227, 'tai_std': 0.028076995641981673}, {'name': 'Escherichia coli', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': None, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'AAA': 0.7667390608953314, 'CGC': 0.40076028262096014, 'ATT': 0.5085852538544149, 'AGC': 0.27799163607995603, 'ACC': 0.4359429381095841, 'ATC': 0.4205725148436632, 'ACA': 0.13023262212317555, 'GGT': 0.33726003930064996, 'AAC': 0.5512262248110238, 'GCG': 0.3563679190097728, 'GGC': 0.4050486219579785, 'TGA': 0.28926761860893596, 'CGA': 0.06388523224887838, 'GTG': 0.3715940695510341, 'TTG': 0.12808970030428077, 'AAG': 0.23326093910466858, 'TTC': 0.4262385670731707, 'TCA': 0.12251736088908215, 'GCA': 0.21281773299119433, 'AAT': 0.4487737751889762, 'GAA': 0.689467997530356, 'CGT': 0.38120248233619, 'TTT': 0.5737614329268292, 'CTG': 0.4975546385450099, 'GTT': 0.2584847661069405, 'GCC': 0.26990242799092135, 'GAT': 0.626856640686903, 'AGG': 0.020162265266098916, 'CAG': 0.6530527088827319, 'GGG': 0.15057187484254547, 'GTC': 0.21619163129169194, 'CTC': 0.10435163186194824, 'TCT': 0.14544780223293646, 'CCC': 0.12361073915372255, 'CAC': 0.4292446655321377, 'GCT': 0.1609119200081115, 'TTA': 0.13009739763522432, 'CTT': 0.10341378016756284, 'ACG': 0.26849903016796667, 'GGA': 0.10711946389882603, 'CCG': 0.527785216925549, 'CAA': 0.3469472911172681, 'ACT': 0.16532540959927366, 'ATA': 0.07084223130192197, 'CAT': 0.5707553344678623, 'AGT': 0.15102374892893225, 'TGC': 0.55651396076928, 'GAG': 0.31053200246964396, 'TCG': 0.15406749964830613, 'GTA': 0.15372953305033354, 'TAC': 0.4315172054223149, 'TCC': 0.14895195222078703, 'GAC': 0.37314335931309695, 'TGT': 0.4434860392307201, 'TGG': 1.0, 'TAT': 0.5684827945776851, 'CCT': 0.15789033208355652, 'CCA': 0.19071371183717192, 'CGG': 0.09749348521075679, 'CTA': 0.03649285148597391, 'TAA': 0.6391064025794565, 'AGA': 0.036496252317115764, 'TAG': 0.07162597881160755}), 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.75, 'AAC': 0.5, 'AAG': 0.23999999999999996, 'AAT': 0.29500000000000004, 'ACA': 0.1525, 'ACC': 0.25, 'ACG': 0.29, 'ACT': 0.14750000000000002, 'AGA': 0.13874999999999998, 'AGC': 0.125, 'AGG': 0.16499999999999998, 'AGT': 0.07375000000000001, 'ATA': 0.10999999999999999, 'ATC': 0.375, 'ATG': 1.0, 'ATT': 0.22125000000000003, 'CAA': 0.27749999999999997, 'CAC': 0.125, 'CAG': 0.32999999999999996, 'CAT': 0.07375000000000001, 'CCA': 0.13874999999999998, 'CCC': 0.125, 'CCG': 0.16499999999999998, 'CCT': 0.07375000000000001, 'CGA': 0.013799999999999993, 'CGC': 0.36, 'CGG': 0.125, 'CGT': 0.5, 'CTA': 0.18, 'CTC': 0.125, 'CTG': 0.54, 'CTT': 0.07375000000000001, 'GAA': 0.5, 'GAC': 0.375, 'GAG': 0.15999999999999998, 'GAT': 0.22125000000000003, 'GCA': 0.375, 'GCC': 0.25, 'GCG': 0.11999999999999998, 'GCT': 0.14750000000000002, 'GGA': 0.13874999999999998, 'GGC': 0.5, 'GGG': 0.16499999999999998, 'GGT': 0.29500000000000004, 'GTA': 0.625, 'GTC': 0.25, 'GTG': 0.19999999999999996, 'GTT': 0.14750000000000002, 'TAC': 0.375, 'TAT': 0.22125000000000003, 'TCA': 0.13874999999999998, 'TCC': 0.25, 'TCG': 0.16499999999999998, 'TCT': 0.14750000000000002, 'TGC': 0.125, 'TGG': 0.125, 'TGT': 0.07375000000000001, 'TTA': 0.13874999999999998, 'TTC': 0.25, 'TTG': 0.16499999999999998, 'TTT': 0.14750000000000002}, 'tai_avg': 0.24958213077316901, 'tai_std': 0.024913119712427708}]}, 'orf_debug': {'total_loss': {'C': {'TGT': 0.005054253472222221, 'TGC': 0.0078125}, 'D': {'GAT': 0.06183133680555554, 'GAC': 0.0703125}, 'S': {'AGC': 0.0078125, 'TCA': 0.009625781249999998, 'AGT': 0.012058420138888885, 'TCT': 0.038490625, 'TCG': 0.039301388888888886, 'TCC': 0.04513888888888889}, 'Q': {'CAA': 0.03850312499999999, 'CAG': 0.15720555555555554}, 'M': {'ATG': 0.5}, 'N': {'AAT': 0.08086805555555554, 'AAC': 0.125}, 'P': {'CCA': 0.009625781249999998, 'CCG': 0.0714125, 'CCT': 0.12771953125, 'CCC': 0.1328125}, 'K': {'AAG': 0.13155555555555556, 'AAA': 0.28125}, '_': {'TAG': 0.058223756903188754, 'TGA': 0.058223756903188754, 'TAA': 0.058223756903188754}, 'T': {'ACA': 0.011628125, 'ACG': 0.14480555555555558, 'ACC': 0.15625, 'ACT': 0.17237951388888884}, 'F': {'TTC': 0.03125, 'TTT': 0.03189062499999999}, 'A': {'GCA': 0.0703125, 'GCG': 0.1677555555555556, 'GCC': 0.25347222222222227, 'GCT': 0.280990625}, 'G': {'GGA': 0.02351467013888888, 'GGT': 0.08086805555555554, 'GGC': 0.125, 'GGG': 0.14196805555555553}, 'I': {'ATT': 0.04548828125, 'ATC': 0.0703125, 'ATA': 0.0821}, 'L': {'TTA': 0.009625781249999998, 'CTA': 0.03008888888888889, 'TTG': 0.031980555555555554, 'CTC': 0.06964722222222222, 'CTT': 0.09091953124999999, 'CTG': 0.1758125}, 'H': {'CAC': 0.0078125, 'CAT': 0.012058420138888885}, 'R': {'CGC': 0.08222222222222221, 'AGG': 0.11336805555555556, 'CGT': 0.125, 'AGA': 0.1256271701388889, 'CGG': 0.1328125, 'CGA': 0.21022005555555556}, 'W': {'TGG': 0.0078125}, 'V': {'GTG': 0.12275555555555556, 'GTC': 0.15625, 'GTT': 0.17237951388888884, 'GTA': 0.1953125}, 'E': {'GAA': 0.125, 'GAG': 0.24400000000000005}, 'Y': {'TAT': 0.03381467013888889, 'TAC': 0.0703125}}, 'optimized_loss': {'C': defaultdict(<class 'float'>, {'TGT': 0.004669444444444442, 'TGC': 0.0}), 'D': defaultdict(<class 'float'>, {'GAT': 0.07471111111111108, 'GAC': 0.0}), 'S': defaultdict(<class 'float'>, {'TCT': 0.055224999999999996, 'TCG': 0.051377777777777786, 'TCA': 0.0, 'TCC': 0.027777777777777776, 'AGC': 0.0, 'AGT': 0.01867777777777777}), 'Q': defaultdict(<class 'float'>, {'CAA': 0.0, 'CAG': 0.20551111111111114}), 'M': defaultdict(<class 'float'>, {'ATG': 0.0}), 'N': defaultdict(<class 'float'>, {'AAC': 0.0, 'AAT': 0.07471111111111108}), 'P': defaultdict(<class 'float'>, {'CCT': 0.25, 'CCG': 0.11560000000000002, 'CCA': 0.0, 'CCC': 0.25}), 'K': defaultdict(<class 'float'>, {'AAG': 0.20551111111111114, 'AAA': 0.0}), '_': defaultdict(<class 'float'>, {'TAG': 0.11644751380637751, 'TGA': 0.11644751380637751, 'TAA': 0.11644751380637751}), 'T': defaultdict(<class 'float'>, {'ACC': 0.25, 'ACA': 0.0, 'ACG': 0.20551111111111114, 'ACT': 0.3230027777777777}), 'F': defaultdict(<class 'float'>, {'TTT': 0.042024999999999986, 'TTC': 0.0}), 'A': defaultdict(<class 'float'>, {'GCA': 0.0, 'GCC': 0.44444444444444453, 'GCG': 0.32111111111111124, 'GCT': 0.540225}), 'G': defaultdict(<class 'float'>, {'GGT': 0.07471111111111108, 'GGG': 0.25671111111111106, 'GGA': 0.027777777777777766, 'GGC': 0.0}), 'I': defaultdict(<class 'float'>, {'ATC': 0.0, 'ATA': 0.1521, 'ATT': 0.042024999999999986}), 'L': defaultdict(<class 'float'>, {'TTA': 0.0, 'TTG': 0.03673611111111111, 'CTC': 0.12366944444444446, 'CTT': 0.17639999999999997, 'CTG': 0.060024999999999995, 'CTA': 0.027777777777777783}), 'H': defaultdict(<class 'float'>, {'CAT': 0.01867777777777777, 'CAC': 0.0}), 'R': defaultdict(<class 'float'>, {'CGA': 0.4202496711111111, 'CGC': 0.034844444444444435, 'CGG': 0.25, 'CGT': 0.0, 'AGG': 0.1995111111111111, 'AGA': 0.2320027777777778}), 'W': defaultdict(<class 'float'>, {'TGG': 0.0}), 'V': defaultdict(<class 'float'>, {'GTA': 0.0, 'GTC': 0.25, 'GTG': 0.20551111111111114, 'GTT': 0.3230027777777777}), 'E': defaultdict(<class 'float'>, {'GAG': 0.4624000000000001, 'GAA': 0.0}), 'Y': defaultdict(<class 'float'>, {'TAT': 0.01867777777777777, 'TAC': 0.0})}, 'deoptimized_loss': {'C': defaultdict(<class 'float'>, {'TGT': 0.005439062500000001, 'TGC': 0.015625}), 'D': defaultdict(<class 'float'>, {'GAT': 0.04895156250000001, 'GAC': 0.140625}), 'S': defaultdict(<class 'float'>, {'TCT': 0.021756250000000005, 'TCG': 0.027224999999999992, 'TCA': 0.019251562499999996, 'TCC': 0.0625, 'AGC': 0.015625, 'AGT': 0.005439062500000001}), 'Q': defaultdict(<class 'float'>, {'CAA': 0.07700624999999998, 'CAG': 0.10889999999999997}), 'M': defaultdict(<class 'float'>, {'ATG': 1.0}), 'N': defaultdict(<class 'float'>, {'AAC': 0.25, 'AAT': 0.08702500000000002}), 'P': defaultdict(<class 'float'>, {'CCT': 0.005439062500000001, 'CCG': 0.027224999999999992, 'CCA': 0.019251562499999996, 'CCC': 0.015625}), 'K': defaultdict(<class 'float'>, {'AAG': 0.057599999999999985, 'AAA': 0.5625}), '_': defaultdict(<class 'float'>, {'TAG': 0, 'TGA': 0, 'TAA': 0}), 'T': defaultdict(<class 'float'>, {'ACC': 0.0625, 'ACA': 0.02325625, 'ACG': 0.0841, 'ACT': 0.021756250000000005}), 'F': defaultdict(<class 'float'>, {'TTT': 0.021756250000000005, 'TTC': 0.0625}), 'A': defaultdict(<class 'float'>, {'GCA': 0.140625, 'GCC': 0.0625, 'GCG': 0.014399999999999996, 'GCT': 0.021756250000000005}), 'G': defaultdict(<class 'float'>, {'GGT': 0.08702500000000002, 'GGG': 0.027224999999999992, 'GGA': 0.019251562499999996, 'GGC': 0.25}), 'I': defaultdict(<class 'float'>, {'ATC': 0.140625, 'ATA': 0.012099999999999998, 'ATT': 0.04895156250000001}), 'L': defaultdict(<class 'float'>, {'TTA': 0.019251562499999996, 'TTG': 0.027224999999999992, 'CTC': 0.015625, 'CTT': 0.005439062500000001, 'CTG': 0.2916, 'CTA': 0.0324}), 'H': defaultdict(<class 'float'>, {'CAT': 0.005439062500000001, 'CAC': 0.015625}), 'R': defaultdict(<class 'float'>, {'CGA': 0.0001904399999999998, 'CGC': 0.1296, 'CGG': 0.015625, 'CGT': 0.25, 'AGG': 0.027224999999999992, 'AGA': 0.019251562499999996}), 'W': defaultdict(<class 'float'>, {'TGG': 0.015625}), 'V': defaultdict(<class 'float'>, {'GTA': 0.390625, 'GTC': 0.0625, 'GTG': 0.03999999999999998, 'GTT': 0.021756250000000005}), 'E': defaultdict(<class 'float'>, {'GAG': 0.02559999999999999, 'GAA': 0.25}), 'Y': defaultdict(<class 'float'>, {'TAT': 0.04895156250000001, 'TAC': 0.140625})}}, 'orf': {'orf_module_input_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'optimized_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAGGTGCACATGGAAGGAAGCGTGAATGGACACGAATTCGAAATTGAAGGAGAAGGAGAAGGACGCCCATATGAAGGAACACAAACAGCAAAGTTAAAGGTGACAAAGGGAGGTCCATTACCATTCGCATGGGATATTTTAAGCCCACAATTCATGTATGGAAGCAAGGCATATGTGAAGCACCCAGCAGATATTCCAGATTATTTAAAGTTAAGCTTCCCAGAAGGATTCAAGTGGGAACGCGTGATGAATTTCGAAGATGGAGGTGTGGTCACAGTGACACAAGATAGCTCATTACAAGATGGAGAATTCATTTATAAGGTGAAGTTACGCGGAACAAATTTCCCAAGCGATGGACCAGTGATGCAAAAGAAAACAATGGGATGGGAAGCAAGCTCAGAACGCATGTATCCAGAAGATGGAGCATTAAAGGGAGAAATTAAGCAACGCTTAAAGTTAAAGGATGGAGGTCACTATGATGCAGAAGTGAAGACAACGTATAAGGCAAAGAAACCAGTGCAATTACCAGGAGCATATAATGTGAATATTAAGTTAGATATTACAAGCCACAATGAAGATTATACAATTGTGGAACAATATGAACGCGCAGAAGGACGCCACAGCACAGGAGGTATGGATGAATTATATAAGTAA', 'aa_to_loss_mapping': {'C': {'TGT': 0.005054253472222221, 'TGC': 0.0078125}, 'D': {'GAT': 0.06183133680555554, 'GAC': 0.0703125}, 'S': {'AGC': 0.0078125, 'TCA': 0.009625781249999998, 'AGT': 0.012058420138888885, 'TCT': 0.038490625, 'TCG': 0.039301388888888886, 'TCC': 0.04513888888888889}, 'Q': {'CAA': 0.03850312499999999, 'CAG': 0.15720555555555554}, 'M': {'ATG': 0.5}, 'N': {'AAT': 0.08086805555555554, 'AAC': 0.125}, 'P': {'CCA': 0.009625781249999998, 'CCG': 0.0714125, 'CCT': 0.12771953125, 'CCC': 0.1328125}, 'K': {'AAG': 0.13155555555555556, 'AAA': 0.28125}, '_': {'TAG': 0.058223756903188754, 'TGA': 0.058223756903188754, 'TAA': 0.058223756903188754}, 'T': {'ACA': 0.011628125, 'ACG': 0.14480555555555558, 'ACC': 0.15625, 'ACT': 0.17237951388888884}, 'F': {'TTC': 0.03125, 'TTT': 0.03189062499999999}, 'A': {'GCA': 0.0703125, 'GCG': 0.1677555555555556, 'GCC': 0.25347222222222227, 'GCT': 0.280990625}, 'G': {'GGA': 0.02351467013888888, 'GGT': 0.08086805555555554, 'GGC': 0.125, 'GGG': 0.14196805555555553}, 'I': {'ATT': 0.04548828125, 'ATC': 0.0703125, 'ATA': 0.0821}, 'L': {'TTA': 0.009625781249999998, 'CTA': 0.03008888888888889, 'TTG': 0.031980555555555554, 'CTC': 0.06964722222222222, 'CTT': 0.09091953124999999, 'CTG': 0.1758125}, 'H': {'CAC': 0.0078125, 'CAT': 0.012058420138888885}, 'R': {'CGC': 0.08222222222222221, 'AGG': 0.11336805555555556, 'CGT': 0.125, 'AGA': 0.1256271701388889, 'CGG': 0.1328125, 'CGA': 0.21022005555555556}, 'W': {'TGG': 0.0078125}, 'V': {'GTG': 0.12275555555555556, 'GTC': 0.15625, 'GTT': 0.17237951388888884, 'GTA': 0.1953125}, 'E': {'GAA': 0.125, 'GAG': 0.24400000000000005}, 'Y': {'TAT': 0.03381467013888889, 'TAC': 0.0703125}}, 'run_time': 0.7339999999385327}, 'evaluation': [{'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'tai_initial_score': 0.33696708001007003, 'tai_final_score': 0.42878031715627346, 'dist_score': 3.270052049618912}, {'name': 'Escherichia coli', 'is_wanted': False, 'tai_initial_score': 0.2816376469779175, 'tai_final_score': 0.22215173994906262, 'dist_score': -2.387734162381149}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAGGTGCACATGGAAGGAAGCGTGAATGGACACGAATTCGAAATTGAAGGAGAAGGAGAAGGACGCCCATATGAAGGAACACAAACAGCAAAGTTAAAGGTGACAAAGGGAGGTCCATTACCATTCGCATGGGATATTTTAAGCCCACAATTCATGTATGGAAGCAAGGCATATGTGAAGCACCCAGCAGATATTCCAGATTATTTAAAGTTAAGCTTCCCAGAAGGATTCAAGTGGGAACGCGTGATGAATTTCGAAGATGGAGGTGTGGTCACAGTGACACAAGATAGCTCATTACAAGATGGAGAATTCATTTATAAGGTGAAGTTACGCGGAACAAATTTCCCAAGCGATGGACCAGTGATGCAAAAGAAAACAATGGGATGGGAAGCAAGCTCAGAACGCATGTATCCAGAAGATGGAGCATTAAAGGGAGAAATTAAGCAACGCTTAAAGTTAAAGGATGGAGGTCACTATGATGCAGAAGTGAAGACAACGTATAAGGCAAAGAAACCAGTGCAATTACCAGGAGCATATAATGTGAATATTAAGTTAGATATTACAAGCCACAATGAAGATTATACAATTGTGGAACAATATGAACGCGCAGAAGGACGCCACAGCACAGGAGGTATGGATGAATTATATAAGTAA', 'average_distance_score': 2.8288931060000304, 'weakest_link_score': 2.8288931060000304, 'ratio_score': 1.1355593244176034, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTCATGCGCTTCAAGGTGCACATGGAAGGAAGCGTGAATGGACACGAATTCGAAATTGAAGGAGAAGGAGAAGGACGCCCATATGAAGGAACACAAACAGCAAAGTTAAAGGTGACAAAGGGAGGTCCATTACCATTCGCATGGGATATTTTAAGCCCACAATTCATGTATGGAAGCAAGGCATATGTGAAGCACCCAGCAGATATTCCAGATTATTTAAAGTTAAGCTTCCCAGAAGGATTCAAGTGGGAACGCGTGATGAATTTCGAAGATGGAGGTGTGGTCACAGTGACACAAGATAGCTCATTACAAGATGGAGAATTCATTTATAAGGTGAAGTTACGCGGAACAAATTTCCCAAGCGATGGACCAGTGATGCAAAAGAAAACAATGGGATGGGAAGCAAGCTCAGAACGCATGTATCCAGAAGATGGAGCATTAAAGGGAGAAATTAAGCAACGCTTAAAGTTAAAGGATGGAGGTCACTATGATGCAGAAGTGAAGACAACGTATAAGGCAAAGAAACCAGTGCAATTACCAGGAGCATATAATGTGAATATTAAGTTAGATATTACAAGCCACAATGAAGATTATACAATTGTGGAACAATATGAACGCGCAGAAGGACGCCACAGCACAGGAGGTATGGATGAATTATATAAGTAA'}], 'final_evaluation': {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAGGTGCACATGGAAGGAAGCGTGAATGGACACGAATTCGAAATTGAAGGAGAAGGAGAAGGACGCCCATATGAAGGAACACAAACAGCAAAGTTAAAGGTGACAAAGGGAGGTCCATTACCATTCGCATGGGATATTTTAAGCCCACAATTCATGTATGGAAGCAAGGCATATGTGAAGCACCCAGCAGATATTCCAGATTATTTAAAGTTAAGCTTCCCAGAAGGATTCAAGTGGGAACGCGTGATGAATTTCGAAGATGGAGGTGTGGTCACAGTGACACAAGATAGCTCATTACAAGATGGAGAATTCATTTATAAGGTGAAGTTACGCGGAACAAATTTCCCAAGCGATGGACCAGTGATGCAAAAGAAAACAATGGGATGGGAAGCAAGCTCAGAACGCATGTATCCAGAAGATGGAGCATTAAAGGGAGAAATTAAGCAACGCTTAAAGTTAAAGGATGGAGGTCACTATGATGCAGAAGTGAAGACAACGTATAAGGCAAAGAAACCAGTGCAATTACCAGGAGCATATAATGTGAATATTAAGTTAGATATTACAAGCCACAATGAAGATTATACAATTGTGGAACAATATGAACGCGCAGAAGGACGCCACAGCACAGGAGGTATGGATGAATTATATAAGTAA', 'average_distance_score': 2.8288931060000304, 'weakest_link_score': 2.8288931060000304, 'ratio_score': 1.1355593244176034}, 'zip_output_file_path': 'results\\\\mcherry-wanted-constraint-init-external-0.1\\\\tAI_single_codon_diff_ecoli_opt_False_klLD\\\\communique_results.zip'}\n",
      "\n",
      "##########################\n",
      "# User Input #\n",
      "##########################\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is:\n",
      "ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "------------------------------------------\n",
      "Parsing information for Escherichia coli:\n",
      "------------------------------------------\n",
      "Organism is defined as wanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 90 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3893\n",
      "Calculate CAI weights from a reference set of 584 highly expressed genes from estimated expression dictionary.\n",
      "cai_std=0.09377868043841883, cai_avg=0.6151304741138522\n",
      "------------------------------------------\n",
      "Parsing information for Bacillus subtilis:\n",
      "------------------------------------------\n",
      "Organism is defined as unwanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 98 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3544\n",
      "Calculate CAI weights from a reference set of 532 highly expressed genes from estimated expression dictionary.\n",
      "cai_std=0.042802002292947894, cai_avg=0.7334925635409068\n",
      "-----------------------------\n",
      "Normalized Prioritization Weights\n",
      "-----------------------------\n",
      "Escherichia coli : 1.0\n",
      "Bacillus subtilis : 1.0\n",
      "Total input processing time: 29.90017008781433\n",
      "\n",
      "##########################\n",
      "# Initiation #\n",
      "##########################\n",
      "Running initiation optimization InitiationOptimizationMethod.external_module on 15 codons at the start of the ORF\n",
      "Taking optimized sequence from configuration value config.INITIATION.EXTERNAL_INITIATION_ORF\n",
      "Optimized sequence is: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "Optimization method used is: ORFOptimizationMethod.zscore_bulk_aa_ratio\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CTA due to very low average frequency 0.03649285148597391 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CTA due to very low average frequency 0.03649285148597391 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CTA due to very low average frequency 0.03649285148597391 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGAGGTTCAAAGTGCATATGGAGGGTTCTGTTAACGGCCATGAATTCGAGATTGAAGGAGAGGGCGAGGGCAGACCATATGAAGGCACACAGACCGCTAAACTCAAAGTGACAAAGGGAGGACCTTTACCATTCGCTTGGGATATTCTTTCGCCCCAGTTTATGTATGGCAGTAAGGCATACGTGAAGCACCCTGCTGATATTCCGGATTACTTAAAGCTTTCGTTCCCCGAAGGTTTCAAATGGGAGCGAGTAATGAATTTCGAAGATGGGGGAGTTGTGACAGTTACCCAGGACTCGAGCCTACAGGACGGGGAATTTATCTATAAAGTCAAGCTCCGTGGTACAAATTTCCCCTCGGACGGGCCTGTGATGCAGAAGAAAACGATGGGATGGGAAGCCTCGAGTGAGCGAATGTACCCGGAAGACGGAGCCCTCAAAGGGGAAATTAAACAAAGGCTGAAGTTGAAGGATGGCGGGCACTATGACGCAGAAGTCAAGACGACCTATAAAGCGAAGAAACCTGTGCAACTACCTGGTGCATACAACGTGAACATCAAGCTGGATATTACCTCTCACAATGAGGATTACACCATAGTCGAGCAATACGAGAGGGCAGAAGGGCGACATAGTACCGGCGGGATGGACGAGCTGTACAAATAG\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CTA due to very low average frequency 0.03649285148597391 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CTA due to very low average frequency 0.03649285148597391 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGAGATTTAAAGTACATATGGAAGGTTCCGTCAACGGCCATGAATTTGAAATTGAAGGGGAAGGCGAGGGCCGGCCCTACGAAGGGACACAAACCGCTAAGTTAAAAGTGACTAAGGGGGGACCGCTACCTTTTGCATGGGACATTTTGTCGCCGCAGTTTATGTATGGTAGTAAGGCTTACGTCAAACATCCTGCGGATATCCCGGATTACCTTAAACTCTCTTTCCCTGAAGGGTTTAAATGGGAGCGAGTGATGAACTTCGAAGATGGGGGAGTGGTTACCGTCACGCAAGATTCTTCGCTCCAGGACGGTGAGTTTATATACAAAGTGAAACTTCGGGGTACCAATTTCCCTTCTGATGGGCCGGTTATGCAGAAAAAGACGATGGGCTGGGAAGCATCCTCAGAGAGGATGTACCCGGAAGACGGGGCGCTAAAAGGTGAGATAAAACAAAGGTTAAAGCTTAAGGACGGGGGGCACTATGACGCTGAAGTGAAGACCACTTACAAAGCGAAGAAGCCTGTACAGTTACCCGGCGCTTATAATGTAAATATCAAGCTGGACATTACCTCACATAATGAAGATTACACGATAGTCGAACAGTACGAAAGGGCGGAAGGTCGACACTCCACGGGGGGGATGGACGAACTATACAAGTAG\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CTA due to very low average frequency 0.03649285148597391 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CTA due to very low average frequency 0.03649285148597391 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGTTTCAAGGTTCACATGGAAGGTTCAGTGAATGGCCATGAGTTTGAAATCGAGGGGGAGGGCGAGGGGCGTCCTTACGAAGGCACGCAGACCGCGAAGTTGAAGGTTACTAAAGGGGGCCCCCTACCATTTGCATGGGACATATTGTCACCGCAATTTATGTACGGTTCTAAGGCGTATGTGAAGCATCCAGCGGACATCCCAGACTATCTGAAGCTCTCGTTCCCAGAAGGTTTTAAGTGGGAGAGGGTTATGAACTTTGAAGATGGGGGCGTTGTAACTGTAACACAGGACAGCAGTCTACAAGACGGGGAGTTTATTTACAAGGTCAAACTCAGGGGAACCAACTTTCCTTCTGACGGCCCTGTCATGCAGAAGAAGACAATGGGGTGGGAAGCATCTAGCGAACGCATGTACCCTGAGGACGGGGCTTTAAAAGGAGAGATTAAGCAAAGGCTCAAGTTAAAAGACGGAGGGCATTATGATGCTGAGGTGAAGACTACGTATAAAGCGAAGAAGCCCGTTCAGTTGCCAGGTGCGTACAACGTAAACATTAAGTTAGATATCACGTCCCATAACGAGGATTACACCATAGTCGAGCAGTACGAGCGTGCAGAAGGGCGTCATTCTACAGGTGGAATGGACGAACTCTATAAATAG\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CTA due to very low average frequency 0.03649285148597391 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CTA due to very low average frequency 0.03649285148597391 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "\n",
      "##########################\n",
      "# EVALUATION #\n",
      "##########################\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTCATGCGTTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATCGAGGGGGAGGGGGAGGGGCGTCCCTACGAGGGGACCCAGACCGCCAAGCTCAAGGTGACCAAGGGGGGTCCCCTCCCCTTCGCCTGGGACATCCTCTCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTCAAGCTCTCGTTCCCCGAGGGGTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGGGGTGTGGTCACCGTGACCCAGGACTCGTCCCTCCAGGACGGGGAGTTCATCTACAAGGTGAAGCTCCGTGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGGGCCCTCAAGGGGGAGATCAAGCAGCGTCTCAAGCTCAAGGACGGGGGTCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTCCCCGGGGCCTACAACGTGAACATCAAGCTCGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGGCGTCACTCGACCGGGGGTATGGACGAGCTCTACAAGTGA\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTCATGCGTTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATCGAGGGGGAGGGGGAGGGGCGTCCCTACGAGGGGACCCAGACCGCCAAGCTCAAGGTGACCAAGGGGGGTCCCCTCCCCTTCGCCTGGGACATCCTCTCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTCAAGCTCTCGTTCCCCGAGGGGTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGGGGTGTGGTCACCGTGACCCAGGACTCGAGTCTCCAGGACGGGGAGTTCATCTACAAGGTGAAGCTCCGTGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGAGTGAGCGTATGTACCCCGAGGACGGGGCCCTCAAGGGGGAGATCAAGCAGCGTCTCAAGCTCAAGGACGGGGGTCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTCCCCGGGGCCTACAACGTGAACATCAAGCTCGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGGCGTCACTCGACCGGGGGTATGGACGAGCTCTACAAGTGA\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTCATGCGTTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATCGAGGGGGAGGGGGAGGGGCGTCCCTACGAGGGGACCCAGACCGCCAAGCTCAAGGTGACCAAGGGGGGTCCCCTCCCCTTCGCCTGGGACATCCTCTCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTCAAGCTCTCGTTCCCCGAGGGGTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGGGGTGTGGTCACCGTGACCCAGGACTCGAGTCTCCAGGACGGGGAGTTCATCTACAAGGTGAAGCTCCGTGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGAGTGAGCGTATGTACCCCGAGGACGGGGCCCTCAAGGGGGAGATCAAGCAGCGTCTCAAGCTCAAGGACGGGGGTCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTCCCCGGGGCCTACAACGTGAACATCAAGCTCGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGGCGTCACTCGACCGGGGGTATGGACGAGCTCTACAAGTGA\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTCATGCGTTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATCGAGGGGGAGGGGGAGGGGCGTCCCTACGAGGGGACCCAGACCGCCAAGCTCAAGGTGACCAAGGGGGGTCCCCTCCCCTTCGCCTGGGACATCCTCTCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTCAAGCTCTCGTTCCCCGAGGGGTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGGGGTGTGGTCACCGTGACCCAGGACTCGAGTCTCCAGGACGGGGAGTTCATCTACAAGGTGAAGCTCCGTGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGAGTGAGCGTATGTACCCCGAGGACGGGGCCCTCAAGGGGGAGATCAAGCAGCGTCTCAAGCTCAAGGACGGGGGTCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTCCCCGGGGCCTACAACGTGAACATCAAGCTCGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGGCGTCACTCGACCGGGGGTATGGACGAGCTCTACAAGTGA\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATCGAGGGGGAGGGGGAGGGGCGTCCCTACGAGGGGACCCAGACCGCCAAGCTCAAGGTGACCAAGGGGGGTCCCCTCCCCTTCGCCTGGGACATCCTCTCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTCAAGCTCTCGTTCCCCGAGGGGTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGGGGTGTGGTCACCGTGACCCAGGACTCGTCCCTCCAGGACGGGGAGTTCATCTACAAGGTGAAGCTCCGTGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGGGCCCTCAAGGGGGAGATCAAGCAGCGTCTCAAGCTCAAGGACGGGGGTCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTCCCCGGGGCCTACAACGTGAACATCAAGCTCGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGGCGTCACTCGACCGGGGGTATGGACGAGCTCTACAAGTGA', 'average_distance_score': 2.2020095789486795, 'weakest_link_score': 2.2020095789486795, 'ratio_score': 1.1168280941156241}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATCGAGGGGGAGGGGGAGGGGCGTCCCTACGAGGGGACCCAGACCGCCAAGCTCAAGGTGACCAAGGGGGGTCCCCTCCCCTTCGCCTGGGACATCCTCTCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTCAAGCTCTCGTTCCCCGAGGGGTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGGGGTGTGGTCACCGTGACCCAGGACTCGAGTCTCCAGGACGGGGAGTTCATCTACAAGGTGAAGCTCCGTGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGAGTGAGCGTATGTACCCCGAGGACGGGGCCCTCAAGGGGGAGATCAAGCAGCGTCTCAAGCTCAAGGACGGGGGTCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTCCCCGGGGCCTACAACGTGAACATCAAGCTCGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGGCGTCACTCGACCGGGGGTATGGACGAGCTCTACAAGTGA', 'average_distance_score': 2.192378225606475, 'weakest_link_score': 2.192378225606475, 'ratio_score': 1.1165193947422447}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATCGAGGGGGAGGGGGAGGGGCGTCCCTACGAGGGGACCCAGACCGCCAAGCTCAAGGTGACCAAGGGGGGTCCCCTCCCCTTCGCCTGGGACATCCTCTCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTCAAGCTCTCGTTCCCCGAGGGGTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGGGGTGTGGTCACCGTGACCCAGGACTCGAGTCTCCAGGACGGGGAGTTCATCTACAAGGTGAAGCTCCGTGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGAGTGAGCGTATGTACCCCGAGGACGGGGCCCTCAAGGGGGAGATCAAGCAGCGTCTCAAGCTCAAGGACGGGGGTCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTCCCCGGGGCCTACAACGTGAACATCAAGCTCGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGGCGTCACTCGACCGGGGGTATGGACGAGCTCTACAAGTGA', 'average_distance_score': 2.192378225606475, 'weakest_link_score': 2.192378225606475, 'ratio_score': 1.1165193947422447}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATCGAGGGGGAGGGGGAGGGGCGTCCCTACGAGGGGACCCAGACCGCCAAGCTCAAGGTGACCAAGGGGGGTCCCCTCCCCTTCGCCTGGGACATCCTCTCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTCAAGCTCTCGTTCCCCGAGGGGTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGGGGTGTGGTCACCGTGACCCAGGACTCGAGTCTCCAGGACGGGGAGTTCATCTACAAGGTGAAGCTCCGTGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGAGTGAGCGTATGTACCCCGAGGACGGGGCCCTCAAGGGGGAGATCAAGCAGCGTCTCAAGCTCAAGGACGGGGGTCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTCCCCGGGGCCTACAACGTGAACATCAAGCTCGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGGCGTCACTCGACCGGGGGTATGGACGAGCTCTACAAGTGA', 'average_distance_score': 2.192378225606475, 'weakest_link_score': 2.192378225606475, 'ratio_score': 1.1165193947422447}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATCGAGGGGGAGGGGGAGGGGCGTCCCTACGAGGGGACCCAGACCGCCAAGCTCAAGGTGACCAAGGGGGGTCCCCTCCCCTTCGCCTGGGACATCCTCTCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTCAAGCTCTCGTTCCCCGAGGGGTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGGGGTGTGGTCACCGTGACCCAGGACTCGTCCCTCCAGGACGGGGAGTTCATCTACAAGGTGAAGCTCCGTGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGGGCCCTCAAGGGGGAGATCAAGCAGCGTCTCAAGCTCAAGGACGGGGGTCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTCCCCGGGGCCTACAACGTGAACATCAAGCTCGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGGCGTCACTCGACCGGGGGTATGGACGAGCTCTACAAGTGA', 'average_distance_score': 2.2020095789486795, 'weakest_link_score': 2.2020095789486795, 'ratio_score': 1.1168280941156241}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry-wanted-constraint-init-external-0.1\\CAI_zscore_bulk_aa_ratio_ecoli_opt_True_ZXaI\n",
      "Final output: {'module_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'orf_optimization_method': 'zscore_bulk_aa_ratio', 'orf_optimization_cub_index': 'CAI', 'initiation_optimization_method': 'external', 'evaluation_score': 'average_distance', 'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.037037037037037035, 'TGA': 0.24183006535947713, 'GCA': 0.6303663185811371, 'GCC': 0.653737775567711, 'GCG': 1.0, 'GCT': 0.6063318415382065, 'TGC': 1.0, 'TGT': 0.7181136120042874, 'GAC': 0.848284271032921, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.36125998031280765, 'TTC': 1.0, 'TTT': 0.6556786703601108, 'GGA': 0.11638214051214708, 'GGC': 1.0, 'GGG': 0.18844386080105055, 'GGT': 0.9788246881155614, 'CAC': 1.0, 'CAT': 0.7088785046728973, 'ATA': 0.029297564419343447, 'ATC': 1.0, 'ATT': 0.7354394634662901, 'AAA': 1.0, 'AAG': 0.26291560102301786, 'CTA': 0.02288199697428139, 'CTC': 0.1265128593040847, 'CTG': 1.0, 'CTT': 0.10712934947049924, 'TTA': 0.09427004538577911, 'TTG': 0.11753025718608169, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.39072452229299365, 'CCA': 0.2588842975206611, 'CCC': 0.08181818181818182, 'CCG': 1.0, 'CCT': 0.1818181818181818, 'CAA': 0.32759266192437286, 'CAG': 1.0, 'AGA': 0.015347817420769383, 'AGG': 0.00976679290412597, 'CGA': 0.033087502491528795, 'CGC': 0.7123779150886984, 'CGG': 0.05521227825393661, 'CGT': 1.0, 'AGC': 1.0, 'AGT': 0.2936010037641154, 'TCA': 0.28523630280217477, 'TCC': 0.7921371810957758, 'TCG': 0.4086156419907988, 'TCT': 0.8883312421580928, 'ACA': 0.12800331400165701, 'ACC': 1.0, 'ACG': 0.34299917149958575, 'ACT': 0.4233637116818558, 'GTA': 0.5115587090867476, 'GTC': 0.49691004806591893, 'GTG': 1.0, 'GTT': 0.9814602883955139, 'TGG': 1.0, 'TAC': 1.0, 'TAT': 0.8523076923076923}, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'AAA': 0.7667390608953314, 'CGC': 0.40076028262096014, 'ATT': 0.5085852538544149, 'AGC': 0.27799163607995603, 'ACC': 0.4359429381095841, 'ATC': 0.4205725148436632, 'ACA': 0.13023262212317555, 'GGT': 0.33726003930064996, 'AAC': 0.5512262248110238, 'GCG': 0.3563679190097728, 'GGC': 0.4050486219579785, 'TGA': 0.28926761860893596, 'CGA': 0.06388523224887838, 'GTG': 0.3715940695510341, 'TTG': 0.12808970030428077, 'AAG': 0.23326093910466858, 'TTC': 0.4262385670731707, 'TCA': 0.12251736088908215, 'GCA': 0.21281773299119433, 'AAT': 0.4487737751889762, 'GAA': 0.689467997530356, 'CGT': 0.38120248233619, 'TTT': 0.5737614329268292, 'CTG': 0.4975546385450099, 'GTT': 0.2584847661069405, 'GCC': 0.26990242799092135, 'GAT': 0.626856640686903, 'AGG': 0.020162265266098916, 'CAG': 0.6530527088827319, 'GGG': 0.15057187484254547, 'GTC': 0.21619163129169194, 'CTC': 0.10435163186194824, 'TCT': 0.14544780223293646, 'CCC': 0.12361073915372255, 'CAC': 0.4292446655321377, 'GCT': 0.1609119200081115, 'TTA': 0.13009739763522432, 'CTT': 0.10341378016756284, 'ACG': 0.26849903016796667, 'GGA': 0.10711946389882603, 'CCG': 0.527785216925549, 'CAA': 0.3469472911172681, 'ACT': 0.16532540959927366, 'ATA': 0.07084223130192197, 'CAT': 0.5707553344678623, 'AGT': 0.15102374892893225, 'TGC': 0.55651396076928, 'GAG': 0.31053200246964396, 'TCG': 0.15406749964830613, 'GTA': 0.15372953305033354, 'TAC': 0.4315172054223149, 'TCC': 0.14895195222078703, 'GAC': 0.37314335931309695, 'TGT': 0.4434860392307201, 'TGG': 1.0, 'TAT': 0.5684827945776851, 'CCT': 0.15789033208355652, 'CCA': 0.19071371183717192, 'CGG': 0.09749348521075679, 'CTA': 0.03649285148597391, 'TAA': 0.6391064025794565, 'AGA': 0.036496252317115764, 'TAG': 0.07162597881160755}), 'cai_avg': 0.6151304741138522, 'cai_std': 0.09377868043841883, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.10739856801909307, 'TGA': 0.1694510739856802, 'GCA': 1.0, 'GCC': 0.5406038521603331, 'GCG': 0.950546590317543, 'GCT': 0.9508068714211347, 'TGC': 1.0, 'TGT': 0.7315315315315315, 'GAC': 0.6775877587758776, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.3929343907714492, 'TTC': 0.864321608040201, 'TTT': 1.0, 'GGA': 0.8215814372606443, 'GGC': 1.0, 'GGG': 0.28835323271006985, 'GGT': 0.6037395809867087, 'CAC': 0.6798462383305877, 'CAT': 1.0, 'ATA': 0.10219657483246464, 'ATC': 0.8559195830230827, 'ATT': 1.0, 'AAA': 1.0, 'AAG': 0.3472094095940959, 'CTA': 0.13892709766162312, 'CTC': 0.30375974323704724, 'CTG': 0.6657496561210454, 'CTT': 1.0, 'TTA': 0.5476845483723063, 'TTG': 0.3874369555249885, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.8191662154845695, 'CCA': 0.4708680142687277, 'CCC': 0.06896551724137931, 'CCG': 1.0, 'CCT': 0.7439556084026953, 'CAA': 1.0, 'CAG': 0.8559696778269108, 'AGA': 0.7029352226720648, 'AGG': 0.09210526315789473, 'CGA': 0.18927125506072875, 'CGC': 1.0, 'CGG': 0.27834008097165996, 'CGT': 0.9903846153846154, 'AGC': 0.8852526926263463, 'AGT': 0.3177299088649544, 'TCA': 0.9254349627174814, 'TCC': 0.4258492129246064, 'TCG': 0.2531068765534383, 'TCT': 1.0, 'ACA': 1.0, 'ACC': 0.22235752557233318, 'ACG': 0.5411592791037506, 'ACT': 0.43180711154408186, 'GTA': 0.7223221166195737, 'GTC': 0.6935525301823786, 'GTG': 0.6699203698946828, 'GTT': 1.0, 'TGG': 1.0, 'TAC': 0.7833094555873925, 'TAT': 1.0}, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'GAA': 0.6798186496752849, 'AAT': 0.5655317497600131, 'ATA': 0.1287010322849589, 'TTA': 0.198530453110291, 'GAC': 0.35960560391899904, 'CTG': 0.23973441522197322, 'TGG': 1.0, 'AAC': 0.4344682502399869, 'CAA': 0.5131548459124133, 'GCC': 0.20715414603308396, 'CTT': 0.2391679232236727, 'GCT': 0.24752923822568748, 'ATC': 0.366762837741479, 'AAA': 0.7016557540022625, 'AAG': 0.29834424599773746, 'TTG': 0.1592175745811708, 'AGC': 0.22534685577492342, 'CCG': 0.4334091260292722, 'AGT': 0.10611598754150686, 'TTT': 0.6847751982489505, 'GAG': 0.3201813503247151, 'ACT': 0.1614000537105004, 'TCA': 0.23653118484388272, 'ACC': 0.1588786441082565, 'CAC': 0.3265866087018494, 'GGC': 0.3390368073029434, 'GAT': 0.640394396081001, 'ACA': 0.4122430101751559, 'ACG': 0.2674782920060872, 'CCC': 0.0889644363010221, 'AGA': 0.26380392311639106, 'TCC': 0.12712038919920718, 'TAC': 0.34586274736720285, 'CAT': 0.6734133912981506, 'ATT': 0.5045361299735621, 'GCA': 0.2828363712991255, 'TAT': 0.6541372526327971, 'GGG': 0.16211598104043537, 'GTC': 0.25598186049286953, 'CCT': 0.28667299498885185, 'GTT': 0.28439644370189154, 'GCG': 0.26248024444210305, 'TCT': 0.2048572678833432, 'CTC': 0.11219040795754642, 'CCA': 0.19095344268085387, 'GGA': 0.3148341038094681, 'CGA': 0.09855152543710778, 'GTA': 0.19774449549495793, 'GTG': 0.26187720031028106, 'TTC': 0.31522480175104955, 'CGC': 0.2064569601768165, 'CAG': 0.4868451540875866, 'CGG': 0.1555630106168844, 'CGT': 0.18161187196589967, 'CTA': 0.05115922590534585, 'AGG': 0.09401270868690058, 'GGT': 0.18401310784715313, 'TAG': 0.14132712456344587, 'TGC': 0.5460141582025239, 'TCG': 0.10002831475713661, 'TAA': 0.630500582072177, 'TGA': 0.2281722933643772, 'TGT': 0.45398584179747614}), 'cai_avg': 0.7334925635409068, 'cai_std': 0.042802002292947894, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}]}, 'orf': [{'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATCGAGGGGGAGGGGGAGGGGCGTCCCTACGAGGGGACCCAGACCGCCAAGCTCAAGGTGACCAAGGGGGGTCCCCTCCCCTTCGCCTGGGACATCCTCTCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTCAAGCTCTCGTTCCCCGAGGGGTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGGGGTGTGGTCACCGTGACCCAGGACTCGTCCCTCCAGGACGGGGAGTTCATCTACAAGGTGAAGCTCCGTGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGGGCCCTCAAGGGGGAGATCAAGCAGCGTCTCAAGCTCAAGGACGGGGGTCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTCCCCGGGGCCTACAACGTGAACATCAAGCTCGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGGCGTCACTCGACCGGGGGTATGGACGAGCTCTACAAGTGA', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATC', 'L': 'CTC', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'initial_sequence_optimization_score': 1.0808103487561067, 'final_sequence_optimization_score': 1.2934609924222038, 'run_time': 18.984000000054948, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 1.2828958818486456}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATC', 'L': 'CTC', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 1.2934609924222038}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATC', 'L': 'CTC', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 1.2938097063938578}]}, {'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGAGGTTCAAAGTGCATATGGAGGGTTCTGTTAACGGCCATGAATTCGAGATTGAAGGAGAGGGCGAGGGCAGACCATATGAAGGCACACAGACCGCTAAACTCAAAGTGACAAAGGGAGGACCTTTACCATTCGCTTGGGATATTCTTTCGCCCCAGTTTATGTATGGCAGTAAGGCATACGTGAAGCACCCTGCTGATATTCCGGATTACTTAAAGCTTTCGTTCCCCGAAGGTTTCAAATGGGAGCGAGTAATGAATTTCGAAGATGGGGGAGTTGTGACAGTTACCCAGGACTCGAGCCTACAGGACGGGGAATTTATCTATAAAGTCAAGCTCCGTGGTACAAATTTCCCCTCGGACGGGCCTGTGATGCAGAAGAAAACGATGGGATGGGAAGCCTCGAGTGAGCGAATGTACCCGGAAGACGGAGCCCTCAAAGGGGAAATTAAACAAAGGCTGAAGTTGAAGGATGGCGGGCACTATGACGCAGAAGTCAAGACGACCTATAAAGCGAAGAAACCTGTGCAACTACCTGGTGCATACAACGTGAACATCAAGCTGGATATTACCTCTCACAATGAGGATTACACCATAGTCGAGCAATACGAGAGGGCAGAAGGGCGACATAGTACCGGCGGGATGGACGAGCTGTACAAATAG', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATCGAGGGGGAGGGGGAGGGGCGTCCCTACGAGGGGACCCAGACCGCCAAGCTCAAGGTGACCAAGGGGGGTCCCCTCCCCTTCGCCTGGGACATCCTCTCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTCAAGCTCTCGTTCCCCGAGGGGTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGGGGTGTGGTCACCGTGACCCAGGACTCGAGTCTCCAGGACGGGGAGTTCATCTACAAGGTGAAGCTCCGTGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGAGTGAGCGTATGTACCCCGAGGACGGGGCCCTCAAGGGGGAGATCAAGCAGCGTCTCAAGCTCAAGGACGGGGGTCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTCCCCGGGGCCTACAACGTGAACATCAAGCTCGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGGCGTCACTCGACCGGGGGTATGGACGAGCTCTACAAGTGA', 'iterations_count': 2, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATC', 'L': 'CTC', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'initial_sequence_optimization_score': 1.1314976377976684, 'final_sequence_optimization_score': 1.3825349540414895, 'run_time': 12.686999999918044, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATC', 'L': 'CTC', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 1.3825349540414895}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATC', 'L': 'CTC', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 1.2939386204577115}]}, {'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGAGATTTAAAGTACATATGGAAGGTTCCGTCAACGGCCATGAATTTGAAATTGAAGGGGAAGGCGAGGGCCGGCCCTACGAAGGGACACAAACCGCTAAGTTAAAAGTGACTAAGGGGGGACCGCTACCTTTTGCATGGGACATTTTGTCGCCGCAGTTTATGTATGGTAGTAAGGCTTACGTCAAACATCCTGCGGATATCCCGGATTACCTTAAACTCTCTTTCCCTGAAGGGTTTAAATGGGAGCGAGTGATGAACTTCGAAGATGGGGGAGTGGTTACCGTCACGCAAGATTCTTCGCTCCAGGACGGTGAGTTTATATACAAAGTGAAACTTCGGGGTACCAATTTCCCTTCTGATGGGCCGGTTATGCAGAAAAAGACGATGGGCTGGGAAGCATCCTCAGAGAGGATGTACCCGGAAGACGGGGCGCTAAAAGGTGAGATAAAACAAAGGTTAAAGCTTAAGGACGGGGGGCACTATGACGCTGAAGTGAAGACCACTTACAAAGCGAAGAAGCCTGTACAGTTACCCGGCGCTTATAATGTAAATATCAAGCTGGACATTACCTCACATAATGAAGATTACACGATAGTCGAACAGTACGAAAGGGCGGAAGGTCGACACTCCACGGGGGGGATGGACGAACTATACAAGTAG', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATCGAGGGGGAGGGGGAGGGGCGTCCCTACGAGGGGACCCAGACCGCCAAGCTCAAGGTGACCAAGGGGGGTCCCCTCCCCTTCGCCTGGGACATCCTCTCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTCAAGCTCTCGTTCCCCGAGGGGTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGGGGTGTGGTCACCGTGACCCAGGACTCGAGTCTCCAGGACGGGGAGTTCATCTACAAGGTGAAGCTCCGTGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGAGTGAGCGTATGTACCCCGAGGACGGGGCCCTCAAGGGGGAGATCAAGCAGCGTCTCAAGCTCAAGGACGGGGGTCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTCCCCGGGGCCTACAACGTGAACATCAAGCTCGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGGCGTCACTCGACCGGGGGTATGGACGAGCTCTACAAGTGA', 'iterations_count': 2, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATC', 'L': 'CTC', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'initial_sequence_optimization_score': 1.110788217276731, 'final_sequence_optimization_score': 1.3815095260575794, 'run_time': 13.030999999959022, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATC', 'L': 'CTC', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 1.3815095260575794}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATC', 'L': 'CTC', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 1.2939386204577115}]}, {'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGTTTCAAGGTTCACATGGAAGGTTCAGTGAATGGCCATGAGTTTGAAATCGAGGGGGAGGGCGAGGGGCGTCCTTACGAAGGCACGCAGACCGCGAAGTTGAAGGTTACTAAAGGGGGCCCCCTACCATTTGCATGGGACATATTGTCACCGCAATTTATGTACGGTTCTAAGGCGTATGTGAAGCATCCAGCGGACATCCCAGACTATCTGAAGCTCTCGTTCCCAGAAGGTTTTAAGTGGGAGAGGGTTATGAACTTTGAAGATGGGGGCGTTGTAACTGTAACACAGGACAGCAGTCTACAAGACGGGGAGTTTATTTACAAGGTCAAACTCAGGGGAACCAACTTTCCTTCTGACGGCCCTGTCATGCAGAAGAAGACAATGGGGTGGGAAGCATCTAGCGAACGCATGTACCCTGAGGACGGGGCTTTAAAAGGAGAGATTAAGCAAAGGCTCAAGTTAAAAGACGGAGGGCATTATGATGCTGAGGTGAAGACTACGTATAAAGCGAAGAAGCCCGTTCAGTTGCCAGGTGCGTACAACGTAAACATTAAGTTAGATATCACGTCCCATAACGAGGATTACACCATAGTCGAGCAGTACGAGCGTGCAGAAGGGCGTCATTCTACAGGTGGAATGGACGAACTCTATAAATAG', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATCGAGGGGGAGGGGGAGGGGCGTCCCTACGAGGGGACCCAGACCGCCAAGCTCAAGGTGACCAAGGGGGGTCCCCTCCCCTTCGCCTGGGACATCCTCTCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTCAAGCTCTCGTTCCCCGAGGGGTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGGGGTGTGGTCACCGTGACCCAGGACTCGAGTCTCCAGGACGGGGAGTTCATCTACAAGGTGAAGCTCCGTGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGAGTGAGCGTATGTACCCCGAGGACGGGGCCCTCAAGGGGGAGATCAAGCAGCGTCTCAAGCTCAAGGACGGGGGTCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTCCCCGGGGCCTACAACGTGAACATCAAGCTCGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGGCGTCACTCGACCGGGGGTATGGACGAGCTCTACAAGTGA', 'iterations_count': 2, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATC', 'L': 'CTC', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'initial_sequence_optimization_score': 1.118136054719662, 'final_sequence_optimization_score': 1.3819710154889226, 'run_time': 13.202999999979511, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATC', 'L': 'CTC', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 1.3819710154889226}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATC', 'L': 'CTC', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 1.2939386204577115}]}], 'evaluation': [{'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'cai_initial_score': 0.7915994634162647, 'cai_final_score': 0.47659720075395334, 'dist_score': -3.358996535136388}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'cai_initial_score': 0.7695788606254016, 'cai_final_score': 0.4373062451319807, 'dist_score': -7.763015693033747}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATCGAGGGGGAGGGGGAGGGGCGTCCCTACGAGGGGACCCAGACCGCCAAGCTCAAGGTGACCAAGGGGGGTCCCCTCCCCTTCGCCTGGGACATCCTCTCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTCAAGCTCTCGTTCCCCGAGGGGTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGGGGTGTGGTCACCGTGACCCAGGACTCGTCCCTCCAGGACGGGGAGTTCATCTACAAGGTGAAGCTCCGTGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGGGCCCTCAAGGGGGAGATCAAGCAGCGTCTCAAGCTCAAGGACGGGGGTCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTCCCCGGGGCCTACAACGTGAACATCAAGCTCGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGGCGTCACTCGACCGGGGGTATGGACGAGCTCTACAAGTGA', 'average_distance_score': 2.2020095789486795, 'weakest_link_score': 2.2020095789486795, 'ratio_score': 1.1168280941156241, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTCATGCGTTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATCGAGGGGGAGGGGGAGGGGCGTCCCTACGAGGGGACCCAGACCGCCAAGCTCAAGGTGACCAAGGGGGGTCCCCTCCCCTTCGCCTGGGACATCCTCTCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTCAAGCTCTCGTTCCCCGAGGGGTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGGGGTGTGGTCACCGTGACCCAGGACTCGTCCCTCCAGGACGGGGAGTTCATCTACAAGGTGAAGCTCCGTGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGGGCCCTCAAGGGGGAGATCAAGCAGCGTCTCAAGCTCAAGGACGGGGGTCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTCCCCGGGGCCTACAACGTGAACATCAAGCTCGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGGCGTCACTCGACCGGGGGTATGGACGAGCTCTACAAGTGA'}, {'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'cai_initial_score': 0.7915994634162647, 'cai_final_score': 0.4721345395344538, 'dist_score': -3.406583696724037}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'cai_initial_score': 0.7695788606254016, 'cai_final_score': 0.4360939017482657, 'dist_score': -7.791340147936987}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATCGAGGGGGAGGGGGAGGGGCGTCCCTACGAGGGGACCCAGACCGCCAAGCTCAAGGTGACCAAGGGGGGTCCCCTCCCCTTCGCCTGGGACATCCTCTCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTCAAGCTCTCGTTCCCCGAGGGGTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGGGGTGTGGTCACCGTGACCCAGGACTCGAGTCTCCAGGACGGGGAGTTCATCTACAAGGTGAAGCTCCGTGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGAGTGAGCGTATGTACCCCGAGGACGGGGCCCTCAAGGGGGAGATCAAGCAGCGTCTCAAGCTCAAGGACGGGGGTCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTCCCCGGGGCCTACAACGTGAACATCAAGCTCGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGGCGTCACTCGACCGGGGGTATGGACGAGCTCTACAAGTGA', 'average_distance_score': 2.192378225606475, 'weakest_link_score': 2.192378225606475, 'ratio_score': 1.1165193947422447, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTCATGCGTTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATCGAGGGGGAGGGGGAGGGGCGTCCCTACGAGGGGACCCAGACCGCCAAGCTCAAGGTGACCAAGGGGGGTCCCCTCCCCTTCGCCTGGGACATCCTCTCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTCAAGCTCTCGTTCCCCGAGGGGTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGGGGTGTGGTCACCGTGACCCAGGACTCGAGTCTCCAGGACGGGGAGTTCATCTACAAGGTGAAGCTCCGTGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGAGTGAGCGTATGTACCCCGAGGACGGGGCCCTCAAGGGGGAGATCAAGCAGCGTCTCAAGCTCAAGGACGGGGGTCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTCCCCGGGGCCTACAACGTGAACATCAAGCTCGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGGCGTCACTCGACCGGGGGTATGGACGAGCTCTACAAGTGA'}, {'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'cai_initial_score': 0.7915994634162647, 'cai_final_score': 0.4721345395344538, 'dist_score': -3.406583696724037}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'cai_initial_score': 0.7695788606254016, 'cai_final_score': 0.4360939017482657, 'dist_score': -7.791340147936987}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATCGAGGGGGAGGGGGAGGGGCGTCCCTACGAGGGGACCCAGACCGCCAAGCTCAAGGTGACCAAGGGGGGTCCCCTCCCCTTCGCCTGGGACATCCTCTCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTCAAGCTCTCGTTCCCCGAGGGGTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGGGGTGTGGTCACCGTGACCCAGGACTCGAGTCTCCAGGACGGGGAGTTCATCTACAAGGTGAAGCTCCGTGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGAGTGAGCGTATGTACCCCGAGGACGGGGCCCTCAAGGGGGAGATCAAGCAGCGTCTCAAGCTCAAGGACGGGGGTCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTCCCCGGGGCCTACAACGTGAACATCAAGCTCGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGGCGTCACTCGACCGGGGGTATGGACGAGCTCTACAAGTGA', 'average_distance_score': 2.192378225606475, 'weakest_link_score': 2.192378225606475, 'ratio_score': 1.1165193947422447, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTCATGCGTTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATCGAGGGGGAGGGGGAGGGGCGTCCCTACGAGGGGACCCAGACCGCCAAGCTCAAGGTGACCAAGGGGGGTCCCCTCCCCTTCGCCTGGGACATCCTCTCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTCAAGCTCTCGTTCCCCGAGGGGTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGGGGTGTGGTCACCGTGACCCAGGACTCGAGTCTCCAGGACGGGGAGTTCATCTACAAGGTGAAGCTCCGTGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGAGTGAGCGTATGTACCCCGAGGACGGGGCCCTCAAGGGGGAGATCAAGCAGCGTCTCAAGCTCAAGGACGGGGGTCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTCCCCGGGGCCTACAACGTGAACATCAAGCTCGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGGCGTCACTCGACCGGGGGTATGGACGAGCTCTACAAGTGA'}, {'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'cai_initial_score': 0.7915994634162647, 'cai_final_score': 0.4721345395344538, 'dist_score': -3.406583696724037}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'cai_initial_score': 0.7695788606254016, 'cai_final_score': 0.4360939017482657, 'dist_score': -7.791340147936987}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATCGAGGGGGAGGGGGAGGGGCGTCCCTACGAGGGGACCCAGACCGCCAAGCTCAAGGTGACCAAGGGGGGTCCCCTCCCCTTCGCCTGGGACATCCTCTCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTCAAGCTCTCGTTCCCCGAGGGGTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGGGGTGTGGTCACCGTGACCCAGGACTCGAGTCTCCAGGACGGGGAGTTCATCTACAAGGTGAAGCTCCGTGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGAGTGAGCGTATGTACCCCGAGGACGGGGCCCTCAAGGGGGAGATCAAGCAGCGTCTCAAGCTCAAGGACGGGGGTCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTCCCCGGGGCCTACAACGTGAACATCAAGCTCGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGGCGTCACTCGACCGGGGGTATGGACGAGCTCTACAAGTGA', 'average_distance_score': 2.192378225606475, 'weakest_link_score': 2.192378225606475, 'ratio_score': 1.1165193947422447, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTCATGCGTTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATCGAGGGGGAGGGGGAGGGGCGTCCCTACGAGGGGACCCAGACCGCCAAGCTCAAGGTGACCAAGGGGGGTCCCCTCCCCTTCGCCTGGGACATCCTCTCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTCAAGCTCTCGTTCCCCGAGGGGTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGGGGTGTGGTCACCGTGACCCAGGACTCGAGTCTCCAGGACGGGGAGTTCATCTACAAGGTGAAGCTCCGTGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGAGTGAGCGTATGTACCCCGAGGACGGGGCCCTCAAGGGGGAGATCAAGCAGCGTCTCAAGCTCAAGGACGGGGGTCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTCCCCGGGGCCTACAACGTGAACATCAAGCTCGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGGCGTCACTCGACCGGGGGTATGGACGAGCTCTACAAGTGA'}], 'final_evaluation': {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATCGAGGGGGAGGGGGAGGGGCGTCCCTACGAGGGGACCCAGACCGCCAAGCTCAAGGTGACCAAGGGGGGTCCCCTCCCCTTCGCCTGGGACATCCTCTCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTCAAGCTCTCGTTCCCCGAGGGGTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGGGGTGTGGTCACCGTGACCCAGGACTCGTCCCTCCAGGACGGGGAGTTCATCTACAAGGTGAAGCTCCGTGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAAACCATGGGGTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGGGCCCTCAAGGGGGAGATCAAGCAGCGTCTCAAGCTCAAGGACGGGGGTCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTCCCCGGGGCCTACAACGTGAACATCAAGCTCGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGGCGTCACTCGACCGGGGGTATGGACGAGCTCTACAAGTGA', 'average_distance_score': 2.2020095789486795, 'weakest_link_score': 2.2020095789486795, 'ratio_score': 1.1168280941156241}, 'zip_output_file_path': 'results\\\\mcherry-wanted-constraint-init-external-0.1\\\\CAI_zscore_bulk_aa_ratio_ecoli_opt_True_ZXaI\\\\communique_results.zip'}\n",
      "\n",
      "##########################\n",
      "# User Input #\n",
      "##########################\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is:\n",
      "ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "------------------------------------------\n",
      "Parsing information for Bacillus subtilis:\n",
      "------------------------------------------\n",
      "Organism is defined as wanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 98 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3544\n",
      "Calculate CAI weights from a reference set of 532 highly expressed genes from estimated expression dictionary.\n",
      "cai_std=0.042802002292947894, cai_avg=0.7334925635409068\n",
      "------------------------------------------\n",
      "Parsing information for Escherichia coli:\n",
      "------------------------------------------\n",
      "Organism is defined as unwanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 90 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3893\n",
      "Calculate CAI weights from a reference set of 584 highly expressed genes from estimated expression dictionary.\n",
      "cai_std=0.09377868043841883, cai_avg=0.6151304741138522\n",
      "-----------------------------\n",
      "Normalized Prioritization Weights\n",
      "-----------------------------\n",
      "Bacillus subtilis : 1.0\n",
      "Escherichia coli : 1.0\n",
      "Total input processing time: 26.873214960098267\n",
      "\n",
      "##########################\n",
      "# Initiation #\n",
      "##########################\n",
      "Running initiation optimization InitiationOptimizationMethod.external_module on 15 codons at the start of the ORF\n",
      "Taking optimized sequence from configuration value config.INITIATION.EXTERNAL_INITIATION_ORF\n",
      "Optimized sequence is: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "Optimization method used is: ORFOptimizationMethod.zscore_bulk_aa_ratio\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGCTTCAAAGTCCACATGGAAGGGTCCGTTAACGGTCACGAGTTTGAAATTGAGGGTGAGGGTGAGGGAAGACCCTATGAAGGTACCCAGACGGCAAAGCTTAAGGTAACTAAGGGCGGCCCGCTCCCATTCGCTTGGGATATTCTCTCACCTCAGTTCATGTACGGATCAAAAGCGTATGTAAAACACCCTGCCGACATTCCAGATTATTTAAAGCTGTCCTTCCCTGAGGGGTTCAAATGGGAAAGAGTGATGAACTTCGAAGACGGTGGCGTGGTGACCGTCACCCAGGACTCGTCGCTACAGGACGGTGAGTTTATCTATAAAGTGAAGCTTCGTGGTACTAATTTTCCCTCTGATGGACCAGTAATGCAAAAAAAAACTATGGGATGGGAAGCAAGTTCAGAAAGGATGTACCCAGAAGATGGGGCCCTGAAAGGAGAAATCAAGCAACGTCTGAAACTGAAGGATGGAGGGCACTATGACGCGGAAGTAAAAACAACCTACAAGGCGAAGAAGCCGGTTCAACTGCCCGGAGCTTACAATGTGAATATCAAGCTCGACATCACTTCCCATAACGAAGACTATACTATAGTAGAGCAGTATGAGCGCGCGGAAGGTCGTCATTCCACGGGGGGCATGGATGAGCTTTACAAATAA\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTCCACATGGAGGGTTCAGTGAACGGTCATGAATTTGAAATAGAAGGCGAAGGTGAGGGAAGGCCATACGAGGGGACGCAAACCGCAAAACTAAAGGTTACTAAGGGCGGCCCGCTGCCTTTCGCGTGGGACATTTTGTCTCCGCAGTTCATGTATGGCAGCAAGGCATACGTCAAGCATCCAGCTGATATTCCTGATTATTTGAAACTGAGCTTTCCCGAAGGCTTCAAGTGGGAGCGCGTAATGAATTTTGAGGATGGGGGGGTCGTTACCGTTACGCAGGACTCCAGTTTGCAGGACGGCGAATTTATATACAAAGTGAAACTACGAGGGACAAACTTCCCGAGTGATGGGCCCGTGATGCAAAAGAAGACCATGGGTTGGGAGGCTTCTAGTGAGCGAATGTATCCAGAGGACGGAGCTCTGAAGGGGGAAATAAAGCAAAGGCTAAAGTTAAAAGACGGTGGACACTATGACGCGGAAGTTAAGACAACGTATAAAGCTAAAAAACCGGTACAGTTGCCGGGCGCGTACAATGTTAACATAAAGCTAGATATCACGAGTCACAACGAGGACTATACTATAGTGGAGCAATACGAGCGCGCTGAGGGTAGGCATAGTACTGGTGGTATGGATGAACTATATAAGTAA\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTCAAGGTGCATATGGAGGGTTCTGTGAATGGCCATGAATTCGAGATAGAAGGTGAAGGGGAGGGCCGGCCTTATGAAGGCACCCAAACGGCAAAGCTGAAAGTGACTAAGGGCGGACCTTTGCCGTTCGCATGGGACATACTTAGCCCCCAGTTCATGTATGGCAGTAAAGCATATGTGAAGCACCCCGCTGATATACCCGACTACTTAAAACTTTCCTTTCCCGAAGGTTTCAAGTGGGAACGTGTAATGAACTTTGAAGACGGAGGCGTAGTGACCGTTACTCAAGACAGTTCCTTACAGGATGGTGAATTTATCTACAAAGTTAAGTTGAGGGGAACTAACTTCCCTAGCGATGGGCCTGTAATGCAAAAAAAAACTATGGGATGGGAAGCATCAAGCGAGCGCATGTACCCTGAAGACGGAGCCTTAAAGGGGGAAATCAAACAACGTTTAAAGTTGAAGGATGGAGGCCACTACGATGCTGAAGTGAAGACAACTTATAAAGCCAAGAAACCGGTTCAACTGCCCGGGGCTTATAACGTTAATATCAAACTAGATATAACCTCACACAATGAAGACTATACAATTGTGGAACAATATGAACGCGCCGAAGGACGCCACAGTACCGGTGGGATGGACGAATTGTATAAATGA\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "\n",
      "##########################\n",
      "# EVALUATION #\n",
      "##########################\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA', 'average_distance_score': 3.977517411960954, 'weakest_link_score': 3.977517411960954, 'ratio_score': 1.1714276918403908}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA', 'average_distance_score': 3.977517411960954, 'weakest_link_score': 3.977517411960954, 'ratio_score': 1.1714276918403908}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA', 'average_distance_score': 3.977517411960954, 'weakest_link_score': 3.977517411960954, 'ratio_score': 1.1714276918403908}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA', 'average_distance_score': 3.977517411960954, 'weakest_link_score': 3.977517411960954, 'ratio_score': 1.1714276918403908}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA', 'average_distance_score': 3.977517411960954, 'weakest_link_score': 3.977517411960954, 'ratio_score': 1.1714276918403908}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry-wanted-constraint-init-external-0.1\\CAI_zscore_bulk_aa_ratio_ecoli_opt_False_6X6e\n",
      "Final output: {'module_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'orf_optimization_method': 'zscore_bulk_aa_ratio', 'orf_optimization_cub_index': 'CAI', 'initiation_optimization_method': 'external', 'evaluation_score': 'average_distance', 'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.10739856801909307, 'TGA': 0.1694510739856802, 'GCA': 1.0, 'GCC': 0.5406038521603331, 'GCG': 0.950546590317543, 'GCT': 0.9508068714211347, 'TGC': 1.0, 'TGT': 0.7315315315315315, 'GAC': 0.6775877587758776, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.3929343907714492, 'TTC': 0.864321608040201, 'TTT': 1.0, 'GGA': 0.8215814372606443, 'GGC': 1.0, 'GGG': 0.28835323271006985, 'GGT': 0.6037395809867087, 'CAC': 0.6798462383305877, 'CAT': 1.0, 'ATA': 0.10219657483246464, 'ATC': 0.8559195830230827, 'ATT': 1.0, 'AAA': 1.0, 'AAG': 0.3472094095940959, 'CTA': 0.13892709766162312, 'CTC': 0.30375974323704724, 'CTG': 0.6657496561210454, 'CTT': 1.0, 'TTA': 0.5476845483723063, 'TTG': 0.3874369555249885, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.8191662154845695, 'CCA': 0.4708680142687277, 'CCC': 0.06896551724137931, 'CCG': 1.0, 'CCT': 0.7439556084026953, 'CAA': 1.0, 'CAG': 0.8559696778269108, 'AGA': 0.7029352226720648, 'AGG': 0.09210526315789473, 'CGA': 0.18927125506072875, 'CGC': 1.0, 'CGG': 0.27834008097165996, 'CGT': 0.9903846153846154, 'AGC': 0.8852526926263463, 'AGT': 0.3177299088649544, 'TCA': 0.9254349627174814, 'TCC': 0.4258492129246064, 'TCG': 0.2531068765534383, 'TCT': 1.0, 'ACA': 1.0, 'ACC': 0.22235752557233318, 'ACG': 0.5411592791037506, 'ACT': 0.43180711154408186, 'GTA': 0.7223221166195737, 'GTC': 0.6935525301823786, 'GTG': 0.6699203698946828, 'GTT': 1.0, 'TGG': 1.0, 'TAC': 0.7833094555873925, 'TAT': 1.0}, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'GAA': 0.6798186496752849, 'AAT': 0.5655317497600131, 'ATA': 0.1287010322849589, 'TTA': 0.198530453110291, 'GAC': 0.35960560391899904, 'CTG': 0.23973441522197322, 'TGG': 1.0, 'AAC': 0.4344682502399869, 'CAA': 0.5131548459124133, 'GCC': 0.20715414603308396, 'CTT': 0.2391679232236727, 'GCT': 0.24752923822568748, 'ATC': 0.366762837741479, 'AAA': 0.7016557540022625, 'AAG': 0.29834424599773746, 'TTG': 0.1592175745811708, 'AGC': 0.22534685577492342, 'CCG': 0.4334091260292722, 'AGT': 0.10611598754150686, 'TTT': 0.6847751982489505, 'GAG': 0.3201813503247151, 'ACT': 0.1614000537105004, 'TCA': 0.23653118484388272, 'ACC': 0.1588786441082565, 'CAC': 0.3265866087018494, 'GGC': 0.3390368073029434, 'GAT': 0.640394396081001, 'ACA': 0.4122430101751559, 'ACG': 0.2674782920060872, 'CCC': 0.0889644363010221, 'AGA': 0.26380392311639106, 'TCC': 0.12712038919920718, 'TAC': 0.34586274736720285, 'CAT': 0.6734133912981506, 'ATT': 0.5045361299735621, 'GCA': 0.2828363712991255, 'TAT': 0.6541372526327971, 'GGG': 0.16211598104043537, 'GTC': 0.25598186049286953, 'CCT': 0.28667299498885185, 'GTT': 0.28439644370189154, 'GCG': 0.26248024444210305, 'TCT': 0.2048572678833432, 'CTC': 0.11219040795754642, 'CCA': 0.19095344268085387, 'GGA': 0.3148341038094681, 'CGA': 0.09855152543710778, 'GTA': 0.19774449549495793, 'GTG': 0.26187720031028106, 'TTC': 0.31522480175104955, 'CGC': 0.2064569601768165, 'CAG': 0.4868451540875866, 'CGG': 0.1555630106168844, 'CGT': 0.18161187196589967, 'CTA': 0.05115922590534585, 'AGG': 0.09401270868690058, 'GGT': 0.18401310784715313, 'TAG': 0.14132712456344587, 'TGC': 0.5460141582025239, 'TCG': 0.10002831475713661, 'TAA': 0.630500582072177, 'TGA': 0.2281722933643772, 'TGT': 0.45398584179747614}), 'cai_avg': 0.7334925635409068, 'cai_std': 0.042802002292947894, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}, {'name': 'Escherichia coli', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.037037037037037035, 'TGA': 0.24183006535947713, 'GCA': 0.6303663185811371, 'GCC': 0.653737775567711, 'GCG': 1.0, 'GCT': 0.6063318415382065, 'TGC': 1.0, 'TGT': 0.7181136120042874, 'GAC': 0.848284271032921, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.36125998031280765, 'TTC': 1.0, 'TTT': 0.6556786703601108, 'GGA': 0.11638214051214708, 'GGC': 1.0, 'GGG': 0.18844386080105055, 'GGT': 0.9788246881155614, 'CAC': 1.0, 'CAT': 0.7088785046728973, 'ATA': 0.029297564419343447, 'ATC': 1.0, 'ATT': 0.7354394634662901, 'AAA': 1.0, 'AAG': 0.26291560102301786, 'CTA': 0.02288199697428139, 'CTC': 0.1265128593040847, 'CTG': 1.0, 'CTT': 0.10712934947049924, 'TTA': 0.09427004538577911, 'TTG': 0.11753025718608169, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.39072452229299365, 'CCA': 0.2588842975206611, 'CCC': 0.08181818181818182, 'CCG': 1.0, 'CCT': 0.1818181818181818, 'CAA': 0.32759266192437286, 'CAG': 1.0, 'AGA': 0.015347817420769383, 'AGG': 0.00976679290412597, 'CGA': 0.033087502491528795, 'CGC': 0.7123779150886984, 'CGG': 0.05521227825393661, 'CGT': 1.0, 'AGC': 1.0, 'AGT': 0.2936010037641154, 'TCA': 0.28523630280217477, 'TCC': 0.7921371810957758, 'TCG': 0.4086156419907988, 'TCT': 0.8883312421580928, 'ACA': 0.12800331400165701, 'ACC': 1.0, 'ACG': 0.34299917149958575, 'ACT': 0.4233637116818558, 'GTA': 0.5115587090867476, 'GTC': 0.49691004806591893, 'GTG': 1.0, 'GTT': 0.9814602883955139, 'TGG': 1.0, 'TAC': 1.0, 'TAT': 0.8523076923076923}, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'AAA': 0.7667390608953314, 'CGC': 0.40076028262096014, 'ATT': 0.5085852538544149, 'AGC': 0.27799163607995603, 'ACC': 0.4359429381095841, 'ATC': 0.4205725148436632, 'ACA': 0.13023262212317555, 'GGT': 0.33726003930064996, 'AAC': 0.5512262248110238, 'GCG': 0.3563679190097728, 'GGC': 0.4050486219579785, 'TGA': 0.28926761860893596, 'CGA': 0.06388523224887838, 'GTG': 0.3715940695510341, 'TTG': 0.12808970030428077, 'AAG': 0.23326093910466858, 'TTC': 0.4262385670731707, 'TCA': 0.12251736088908215, 'GCA': 0.21281773299119433, 'AAT': 0.4487737751889762, 'GAA': 0.689467997530356, 'CGT': 0.38120248233619, 'TTT': 0.5737614329268292, 'CTG': 0.4975546385450099, 'GTT': 0.2584847661069405, 'GCC': 0.26990242799092135, 'GAT': 0.626856640686903, 'AGG': 0.020162265266098916, 'CAG': 0.6530527088827319, 'GGG': 0.15057187484254547, 'GTC': 0.21619163129169194, 'CTC': 0.10435163186194824, 'TCT': 0.14544780223293646, 'CCC': 0.12361073915372255, 'CAC': 0.4292446655321377, 'GCT': 0.1609119200081115, 'TTA': 0.13009739763522432, 'CTT': 0.10341378016756284, 'ACG': 0.26849903016796667, 'GGA': 0.10711946389882603, 'CCG': 0.527785216925549, 'CAA': 0.3469472911172681, 'ACT': 0.16532540959927366, 'ATA': 0.07084223130192197, 'CAT': 0.5707553344678623, 'AGT': 0.15102374892893225, 'TGC': 0.55651396076928, 'GAG': 0.31053200246964396, 'TCG': 0.15406749964830613, 'GTA': 0.15372953305033354, 'TAC': 0.4315172054223149, 'TCC': 0.14895195222078703, 'GAC': 0.37314335931309695, 'TGT': 0.4434860392307201, 'TGG': 1.0, 'TAT': 0.5684827945776851, 'CCT': 0.15789033208355652, 'CCA': 0.19071371183717192, 'CGG': 0.09749348521075679, 'CTA': 0.03649285148597391, 'TAA': 0.6391064025794565, 'AGA': 0.036496252317115764, 'TAG': 0.07162597881160755}), 'cai_avg': 0.6151304741138522, 'cai_std': 0.09377868043841883, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}]}, 'orf': [{'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA', 'iterations_count': 2, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'initial_sequence_optimization_score': 0.9252317033704289, 'final_sequence_optimization_score': 1.4142135623730951, 'run_time': 13.045999999972992, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 1.4142135623730951}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 1.3527453404443204}]}, {'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGCTTCAAAGTCCACATGGAAGGGTCCGTTAACGGTCACGAGTTTGAAATTGAGGGTGAGGGTGAGGGAAGACCCTATGAAGGTACCCAGACGGCAAAGCTTAAGGTAACTAAGGGCGGCCCGCTCCCATTCGCTTGGGATATTCTCTCACCTCAGTTCATGTACGGATCAAAAGCGTATGTAAAACACCCTGCCGACATTCCAGATTATTTAAAGCTGTCCTTCCCTGAGGGGTTCAAATGGGAAAGAGTGATGAACTTCGAAGACGGTGGCGTGGTGACCGTCACCCAGGACTCGTCGCTACAGGACGGTGAGTTTATCTATAAAGTGAAGCTTCGTGGTACTAATTTTCCCTCTGATGGACCAGTAATGCAAAAAAAAACTATGGGATGGGAAGCAAGTTCAGAAAGGATGTACCCAGAAGATGGGGCCCTGAAAGGAGAAATCAAGCAACGTCTGAAACTGAAGGATGGAGGGCACTATGACGCGGAAGTAAAAACAACCTACAAGGCGAAGAAGCCGGTTCAACTGCCCGGAGCTTACAATGTGAATATCAAGCTCGACATCACTTCCCATAACGAAGACTATACTATAGTAGAGCAGTATGAGCGCGCGGAAGGTCGTCATTCCACGGGGGGCATGGATGAGCTTTACAAATAA', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA', 'iterations_count': 2, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'initial_sequence_optimization_score': 0.9086026342396966, 'final_sequence_optimization_score': 1.333527178288054, 'run_time': 12.688000000081956, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 1.333527178288054}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 1.3527453404443204}]}, {'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTCCACATGGAGGGTTCAGTGAACGGTCATGAATTTGAAATAGAAGGCGAAGGTGAGGGAAGGCCATACGAGGGGACGCAAACCGCAAAACTAAAGGTTACTAAGGGCGGCCCGCTGCCTTTCGCGTGGGACATTTTGTCTCCGCAGTTCATGTATGGCAGCAAGGCATACGTCAAGCATCCAGCTGATATTCCTGATTATTTGAAACTGAGCTTTCCCGAAGGCTTCAAGTGGGAGCGCGTAATGAATTTTGAGGATGGGGGGGTCGTTACCGTTACGCAGGACTCCAGTTTGCAGGACGGCGAATTTATATACAAAGTGAAACTACGAGGGACAAACTTCCCGAGTGATGGGCCCGTGATGCAAAAGAAGACCATGGGTTGGGAGGCTTCTAGTGAGCGAATGTATCCAGAGGACGGAGCTCTGAAGGGGGAAATAAAGCAAAGGCTAAAGTTAAAAGACGGTGGACACTATGACGCGGAAGTTAAGACAACGTATAAAGCTAAAAAACCGGTACAGTTGCCGGGCGCGTACAATGTTAACATAAAGCTAGATATCACGAGTCACAACGAGGACTATACTATAGTGGAGCAATACGAGCGCGCTGAGGGTAGGCATAGTACTGGTGGTATGGATGAACTATATAAGTAA', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA', 'iterations_count': 2, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'initial_sequence_optimization_score': 0.8825942123631091, 'final_sequence_optimization_score': 1.259860777026206, 'run_time': 12.515999999945052, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 1.259860777026206}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 1.3527453404443204}]}, {'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTCAAGGTGCATATGGAGGGTTCTGTGAATGGCCATGAATTCGAGATAGAAGGTGAAGGGGAGGGCCGGCCTTATGAAGGCACCCAAACGGCAAAGCTGAAAGTGACTAAGGGCGGACCTTTGCCGTTCGCATGGGACATACTTAGCCCCCAGTTCATGTATGGCAGTAAAGCATATGTGAAGCACCCCGCTGATATACCCGACTACTTAAAACTTTCCTTTCCCGAAGGTTTCAAGTGGGAACGTGTAATGAACTTTGAAGACGGAGGCGTAGTGACCGTTACTCAAGACAGTTCCTTACAGGATGGTGAATTTATCTACAAAGTTAAGTTGAGGGGAACTAACTTCCCTAGCGATGGGCCTGTAATGCAAAAAAAAACTATGGGATGGGAAGCATCAAGCGAGCGCATGTACCCTGAAGACGGAGCCTTAAAGGGGGAAATCAAACAACGTTTAAAGTTGAAGGATGGAGGCCACTACGATGCTGAAGTGAAGACAACTTATAAAGCCAAGAAACCGGTTCAACTGCCCGGGGCTTATAACGTTAATATCAAACTAGATATAACCTCACACAATGAAGACTATACAATTGTGGAACAATATGAACGCGCCGAAGGACGCCACAGTACCGGTGGGATGGACGAATTGTATAAATGA', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA', 'iterations_count': 2, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'initial_sequence_optimization_score': 0.9217255331034807, 'final_sequence_optimization_score': 1.3205178428056774, 'run_time': 12.71799999999348, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 1.3205178428056774}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 1.3527453404443204}]}], 'evaluation': [{'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'cai_initial_score': 0.7695788606254016, 'cai_final_score': 0.9302799596537703, 'dist_score': 3.754522929289362}, {'name': 'Escherichia coli', 'is_wanted': False, 'cai_initial_score': 0.7915994634162647, 'cai_final_score': 0.39768100077174195, 'dist_score': -4.200511894632546}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA', 'average_distance_score': 3.977517411960954, 'weakest_link_score': 3.977517411960954, 'ratio_score': 1.1714276918403908, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA'}, {'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'cai_initial_score': 0.7695788606254016, 'cai_final_score': 0.9302799596537703, 'dist_score': 3.754522929289362}, {'name': 'Escherichia coli', 'is_wanted': False, 'cai_initial_score': 0.7915994634162647, 'cai_final_score': 0.39768100077174195, 'dist_score': -4.200511894632546}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA', 'average_distance_score': 3.977517411960954, 'weakest_link_score': 3.977517411960954, 'ratio_score': 1.1714276918403908, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA'}, {'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'cai_initial_score': 0.7695788606254016, 'cai_final_score': 0.9302799596537703, 'dist_score': 3.754522929289362}, {'name': 'Escherichia coli', 'is_wanted': False, 'cai_initial_score': 0.7915994634162647, 'cai_final_score': 0.39768100077174195, 'dist_score': -4.200511894632546}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA', 'average_distance_score': 3.977517411960954, 'weakest_link_score': 3.977517411960954, 'ratio_score': 1.1714276918403908, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA'}, {'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'cai_initial_score': 0.7695788606254016, 'cai_final_score': 0.9302799596537703, 'dist_score': 3.754522929289362}, {'name': 'Escherichia coli', 'is_wanted': False, 'cai_initial_score': 0.7915994634162647, 'cai_final_score': 0.39768100077174195, 'dist_score': -4.200511894632546}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA', 'average_distance_score': 3.977517411960954, 'weakest_link_score': 3.977517411960954, 'ratio_score': 1.1714276918403908, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA'}], 'final_evaluation': {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA', 'average_distance_score': 3.977517411960954, 'weakest_link_score': 3.977517411960954, 'ratio_score': 1.1714276918403908}, 'zip_output_file_path': 'results\\\\mcherry-wanted-constraint-init-external-0.1\\\\CAI_zscore_bulk_aa_ratio_ecoli_opt_False_6X6e\\\\communique_results.zip'}\n",
      "\n",
      "##########################\n",
      "# User Input #\n",
      "##########################\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is:\n",
      "ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "------------------------------------------\n",
      "Parsing information for Escherichia coli:\n",
      "------------------------------------------\n",
      "Organism is defined as wanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Escherichia coli. Calculating tAI profile.\n",
      "tai_std=0.024913119712427708, tai_avg=0.24958213077316901\n",
      "------------------------------------------\n",
      "Parsing information for Bacillus subtilis:\n",
      "------------------------------------------\n",
      "Organism is defined as unwanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Bacillus subtilis. Calculating tAI profile.\n",
      "tai_std=0.028076995641981673, tai_avg=0.3271366648197227\n",
      "-----------------------------\n",
      "Normalized Prioritization Weights\n",
      "-----------------------------\n",
      "Escherichia coli : 1.0\n",
      "Bacillus subtilis : 1.0\n",
      "Total input processing time: 26.78488254547119\n",
      "\n",
      "##########################\n",
      "# Initiation #\n",
      "##########################\n",
      "Running initiation optimization InitiationOptimizationMethod.external_module on 15 codons at the start of the ORF\n",
      "Taking optimized sequence from configuration value config.INITIATION.EXTERNAL_INITIATION_ORF\n",
      "Optimized sequence is: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "Optimization method used is: ORFOptimizationMethod.zscore_bulk_aa_ratio\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Skipping codon AGA due to very low average frequency 0.036496252317115764 in wanted hosts.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Skipping codon AGA due to very low average frequency 0.036496252317115764 in wanted hosts.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAGGTTCATATGGAAGGGTCAGTCAACGGTCATGAATTTGAGATCGAAGGCGAGGGAGAGGGCCGACCATATGAAGGAACGCAAACTGCTAAACTGAAAGTTACCAAAGGAGGTCCCCTACCGTTCGCCTGGGACATTCTTTCTCCGCAATTTATGTATGGGTCTAAGGCATACGTAAAACATCCAGCAGATATACCGGACTACTTGAAGCTTTCATTTCCCGAAGGTTTTAAGTGGGAAAGAGTAATGAATTTTGAGGATGGGGGCGTGGTCACGGTGACCCAGGACTCCTCTCTCCAAGATGGAGAGTTTATTTACAAGGTGAAGTTGCGCGGAACGAACTTTCCGTCTGACGGCCCGGTTATGCAGAAGAAGACCATGGGTTGGGAAGCGTCGAGTGAGAGAATGTATCCAGAAGACGGGGCATTGAAGGGAGAGATTAAGCAGAGGCTGAAACTAAAAGACGGAGGCCACTACGACGCGGAGGTTAAGACTACGTACAAAGCGAAAAAACCAGTTCAACTACCTGGGGCGTATAACGTGAACATTAAGTTGGACATAACTTCCCATAATGAAGACTATACTATAGTGGAGCAATATGAAAGGGCAGAGGGTAGGCATTCTACTGGGGGGATGGATGAGTTGTACAAGTAA\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Skipping codon AGA due to very low average frequency 0.036496252317115764 in wanted hosts.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Skipping codon AGA due to very low average frequency 0.036496252317115764 in wanted hosts.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGATTCAAAGTGCACATGGAGGGATCAGTTAATGGGCATGAATTCGAAATAGAAGGGGAAGGAGAGGGCCGCCCATACGAGGGTACCCAAACTGCTAAGCTCAAGGTGACCAAAGGGGGGCCTTTGCCGTTTGCTTGGGATATACTCTCTCCCCAATTTATGTACGGGTCAAAGGCGTACGTAAAACATCCAGCGGACATTCCAGATTACCTAAAGCTCAGCTTTCCCGAGGGGTTTAAATGGGAGCGGGTCATGAATTTTGAAGATGGCGGGGTGGTAACTGTTACTCAGGACAGTTCACTGCAGGACGGAGAATTCATCTACAAGGTGAAGCTTCGGGGAACCAATTTTCCATCAGATGGTCCTGTGATGCAAAAAAAGACTATGGGATGGGAGGCCTCTAGTGAAAGAATGTACCCGGAAGACGGTGCTCTGAAAGGCGAAATTAAACAACGGTTGAAGCTAAAGGACGGTGGGCATTATGACGCGGAGGTCAAAACTACTTATAAAGCAAAGAAGCCGGTACAGCTTCCAGGTGCCTACAATGTAAACATAAAATTGGATATCACGTCCCATAATGAGGATTACACAATAGTTGAACAATACGAGCGCGCAGAAGGGCGTCACTCGACGGGAGGAATGGATGAGCTGTATAAGTGA\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Skipping codon AGA due to very low average frequency 0.036496252317115764 in wanted hosts.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Skipping codon AGA due to very low average frequency 0.036496252317115764 in wanted hosts.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGATTTAAGGTACATATGGAGGGTAGTGTTAATGGCCATGAATTCGAGATCGAGGGCGAGGGCGAAGGTAGGCCGTATGAGGGTACGCAGACTGCCAAATTGAAAGTAACAAAGGGGGGTCCTCTCCCCTTCGCATGGGACATCTTAAGTCCACAATTTATGTACGGGTCAAAAGCCTACGTAAAGCATCCCGCGGATATACCTGACTACCTCAAGTTGTCCTTTCCAGAAGGTTTCAAGTGGGAGCGGGTCATGAACTTTGAAGACGGCGGCGTTGTAACCGTCACCCAAGACTCTTCGCTTCAGGATGGAGAGTTTATATACAAGGTAAAACTGCGAGGCACCAATTTTCCTAGTGACGGCCCCGTTATGCAGAAAAAGACCATGGGGTGGGAGGCAAGCTCGGAACGGATGTATCCCGAAGACGGCGCTCTAAAGGGAGAAATCAAGCAGCGATTAAAACTTAAAGACGGTGGGCACTACGACGCCGAAGTAAAAACCACTTACAAAGCCAAGAAACCAGTGCAGCTGCCCGGCGCATATAATGTGAATATAAAATTAGACATAACGTCTCACAATGAAGATTATACAATAGTAGAGCAATACGAGCGTGCCGAAGGTCGACATTCTACCGGTGGCATGGATGAATTATACAAATAG\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Skipping codon AGA due to very low average frequency 0.036496252317115764 in wanted hosts.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Skipping codon AGA due to very low average frequency 0.036496252317115764 in wanted hosts.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "\n",
      "##########################\n",
      "# EVALUATION #\n",
      "##########################\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTTATGCGCTTTAAGGTTCATATGGAGGGGTCGGTTAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGTCCGCTGCCGTTTGCTTGGGATATTCTGTCGCCGCAGTTTATGTATGGGTCGAAGGCTTATGTTAAGCATCCGGCTGATATTCCGGATTATCTGAAGCTGTCGTTTCCGGAGGGGTTTAAGTGGGAGCGCGTTATGAATTTTGAGGATGGGGGTGTTGTCACTGTTACTCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTTAAGCTGCGCGGGACTAATTTTCCGTCGGATGGGCCGGTTATGCAGAAGAAAACTATGGGGTGGGAGGCTTCGTCTGAGCGCATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCTGAGGTTAAGACTACCTATAAGGCTAAGAAACCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATTAAGCTGGATATTACTTCGCATAATGAGGATTATACTATTGTTGAGCAGTATGAGCGCGCTGAGGGGCGCCATTCGACTGGGGGTATGGATGAGCTGTATAAGTAA\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTTATGCGCTTTAAGGTTCATATGGAGGGGTCGGTTAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGTCCGCTGCCGTTTGCTTGGGATATTCTGTCGCCGCAGTTTATGTATGGGTCGAAGGCTTATGTTAAGCATCCGGCTGATATTCCGGATTATCTGAAGCTGTCGTTTCCGGAGGGGTTTAAGTGGGAGCGCGTTATGAATTTTGAGGATGGGGGTGTTGTCACTGTTACTCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTTAAGCTGCGCGGGACTAATTTTCCGTCGGATGGGCCGGTTATGCAGAAGAAAACTATGGGGTGGGAGGCTTCGTCTGAGCGCATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCTGAGGTTAAGACTACCTATAAGGCTAAGAAACCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATTAAGCTGGATATTACTTCGCATAATGAGGATTATACTATTGTTGAGCAGTATGAGCGCGCTGAGGGGCGCCATTCGACTGGGGGTATGGATGAGCTGTATAAGTAA\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTTATGCGCTTTAAGGTTCATATGGAGGGGTCGGTTAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGTCCGCTGCCGTTTGCTTGGGATATTCTGTCGCCGCAGTTTATGTATGGGTCGAAGGCTTATGTTAAGCATCCGGCTGATATTCCGGATTATCTGAAGCTGTCGTTTCCGGAGGGGTTTAAGTGGGAGCGCGTTATGAATTTTGAGGATGGGGGTGTTGTCACTGTTACTCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTTAAGCTGCGCGGGACTAATTTTCCGTCGGATGGGCCGGTTATGCAGAAGAAAACTATGGGGTGGGAGGCTTCGTCTGAGCGCATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCTGAGGTTAAGACTACCTATAAGGCTAAGAAACCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATTAAGCTGGATATTACTTCGCATAATGAGGATTATACTATTGTTGAGCAGTATGAGCGCGCTGAGGGGCGCCATTCGACTGGGGGTATGGATGAGCTGTATAAGTGA\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTTATGCGCTTTAAGGTTCATATGGAGGGGTCGGTTAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGTCCGCTGCCGTTTGCTTGGGATATTCTGTCGCCGCAGTTTATGTATGGGTCGAAGGCTTATGTTAAGCATCCGGCTGATATTCCGGATTATCTGAAGCTGTCGTTTCCGGAGGGGTTTAAGTGGGAGCGCGTTATGAATTTTGAGGATGGGGGTGTTGTCACTGTTACTCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTTAAGCTGCGCGGGACTAATTTTCCGTCGGATGGGCCGGTTATGCAGAAGAAAACTATGGGGTGGGAGGCTTCGTCTGAGCGCATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCTGAGGTTAAGACTACCTATAAGGCTAAGAAACCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATTAAGCTGGATATTACTTCGCATAATGAGGATTATACTATTGTTGAGCAGTATGAGCGCGCTGAGGGGCGCCATTCGACTGGGGGTATGGATGAGCTGTATAAGTAG\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGCTTTAAGGTTCATATGGAGGGGTCGGTTAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGTCCGCTGCCGTTTGCTTGGGATATTCTGTCGCCGCAGTTTATGTATGGGTCGAAGGCTTATGTTAAGCATCCGGCTGATATTCCGGATTATCTGAAGCTGTCGTTTCCGGAGGGGTTTAAGTGGGAGCGCGTTATGAATTTTGAGGATGGGGGTGTTGTCACTGTTACTCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTTAAGCTGCGCGGGACTAATTTTCCGTCGGATGGGCCGGTTATGCAGAAGAAAACTATGGGGTGGGAGGCTTCGTCTGAGCGCATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCTGAGGTTAAGACTACCTATAAGGCTAAGAAACCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATTAAGCTGGATATTACTTCGCATAATGAGGATTATACTATTGTTGAGCAGTATGAGCGCGCTGAGGGGCGCCATTCGACTGGGGGTATGGATGAGCTGTATAAGTAA', 'average_distance_score': 0.7032174309334969, 'weakest_link_score': 0.7032174309334969, 'ratio_score': 1.0396944557740175}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGCTTTAAGGTTCATATGGAGGGGTCGGTTAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGTCCGCTGCCGTTTGCTTGGGATATTCTGTCGCCGCAGTTTATGTATGGGTCGAAGGCTTATGTTAAGCATCCGGCTGATATTCCGGATTATCTGAAGCTGTCGTTTCCGGAGGGGTTTAAGTGGGAGCGCGTTATGAATTTTGAGGATGGGGGTGTTGTCACTGTTACTCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTTAAGCTGCGCGGGACTAATTTTCCGTCGGATGGGCCGGTTATGCAGAAGAAAACTATGGGGTGGGAGGCTTCGTCTGAGCGCATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCTGAGGTTAAGACTACCTATAAGGCTAAGAAACCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATTAAGCTGGATATTACTTCGCATAATGAGGATTATACTATTGTTGAGCAGTATGAGCGCGCTGAGGGGCGCCATTCGACTGGGGGTATGGATGAGCTGTATAAGTAA', 'average_distance_score': 0.7032174309334969, 'weakest_link_score': 0.7032174309334969, 'ratio_score': 1.0396944557740175}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGCTTTAAGGTTCATATGGAGGGGTCGGTTAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGTCCGCTGCCGTTTGCTTGGGATATTCTGTCGCCGCAGTTTATGTATGGGTCGAAGGCTTATGTTAAGCATCCGGCTGATATTCCGGATTATCTGAAGCTGTCGTTTCCGGAGGGGTTTAAGTGGGAGCGCGTTATGAATTTTGAGGATGGGGGTGTTGTCACTGTTACTCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTTAAGCTGCGCGGGACTAATTTTCCGTCGGATGGGCCGGTTATGCAGAAGAAAACTATGGGGTGGGAGGCTTCGTCTGAGCGCATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCTGAGGTTAAGACTACCTATAAGGCTAAGAAACCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATTAAGCTGGATATTACTTCGCATAATGAGGATTATACTATTGTTGAGCAGTATGAGCGCGCTGAGGGGCGCCATTCGACTGGGGGTATGGATGAGCTGTATAAGTGA', 'average_distance_score': 0.7032174309334969, 'weakest_link_score': 0.7032174309334969, 'ratio_score': 1.0396944557740175}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGCTTTAAGGTTCATATGGAGGGGTCGGTTAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGTCCGCTGCCGTTTGCTTGGGATATTCTGTCGCCGCAGTTTATGTATGGGTCGAAGGCTTATGTTAAGCATCCGGCTGATATTCCGGATTATCTGAAGCTGTCGTTTCCGGAGGGGTTTAAGTGGGAGCGCGTTATGAATTTTGAGGATGGGGGTGTTGTCACTGTTACTCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTTAAGCTGCGCGGGACTAATTTTCCGTCGGATGGGCCGGTTATGCAGAAGAAAACTATGGGGTGGGAGGCTTCGTCTGAGCGCATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCTGAGGTTAAGACTACCTATAAGGCTAAGAAACCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATTAAGCTGGATATTACTTCGCATAATGAGGATTATACTATTGTTGAGCAGTATGAGCGCGCTGAGGGGCGCCATTCGACTGGGGGTATGGATGAGCTGTATAAGTAG', 'average_distance_score': 0.7032174309334969, 'weakest_link_score': 0.7032174309334969, 'ratio_score': 1.0396944557740175}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGCTTTAAGGTTCATATGGAGGGGTCGGTTAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGTCCGCTGCCGTTTGCTTGGGATATTCTGTCGCCGCAGTTTATGTATGGGTCGAAGGCTTATGTTAAGCATCCGGCTGATATTCCGGATTATCTGAAGCTGTCGTTTCCGGAGGGGTTTAAGTGGGAGCGCGTTATGAATTTTGAGGATGGGGGTGTTGTCACTGTTACTCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTTAAGCTGCGCGGGACTAATTTTCCGTCGGATGGGCCGGTTATGCAGAAGAAAACTATGGGGTGGGAGGCTTCGTCTGAGCGCATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCTGAGGTTAAGACTACCTATAAGGCTAAGAAACCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATTAAGCTGGATATTACTTCGCATAATGAGGATTATACTATTGTTGAGCAGTATGAGCGCGCTGAGGGGCGCCATTCGACTGGGGGTATGGATGAGCTGTATAAGTAA', 'average_distance_score': 0.7032174309334969, 'weakest_link_score': 0.7032174309334969, 'ratio_score': 1.0396944557740175}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry-wanted-constraint-init-external-0.1\\tAI_zscore_bulk_aa_ratio_ecoli_opt_True_BbAg\n",
      "Final output: {'module_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'orf_optimization_method': 'zscore_bulk_aa_ratio', 'orf_optimization_cub_index': 'tAI', 'initiation_optimization_method': 'external', 'evaluation_score': 'average_distance', 'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': None, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'AAA': 0.7667390608953314, 'CGC': 0.40076028262096014, 'ATT': 0.5085852538544149, 'AGC': 0.27799163607995603, 'ACC': 0.4359429381095841, 'ATC': 0.4205725148436632, 'ACA': 0.13023262212317555, 'GGT': 0.33726003930064996, 'AAC': 0.5512262248110238, 'GCG': 0.3563679190097728, 'GGC': 0.4050486219579785, 'TGA': 0.28926761860893596, 'CGA': 0.06388523224887838, 'GTG': 0.3715940695510341, 'TTG': 0.12808970030428077, 'AAG': 0.23326093910466858, 'TTC': 0.4262385670731707, 'TCA': 0.12251736088908215, 'GCA': 0.21281773299119433, 'AAT': 0.4487737751889762, 'GAA': 0.689467997530356, 'CGT': 0.38120248233619, 'TTT': 0.5737614329268292, 'CTG': 0.4975546385450099, 'GTT': 0.2584847661069405, 'GCC': 0.26990242799092135, 'GAT': 0.626856640686903, 'AGG': 0.020162265266098916, 'CAG': 0.6530527088827319, 'GGG': 0.15057187484254547, 'GTC': 0.21619163129169194, 'CTC': 0.10435163186194824, 'TCT': 0.14544780223293646, 'CCC': 0.12361073915372255, 'CAC': 0.4292446655321377, 'GCT': 0.1609119200081115, 'TTA': 0.13009739763522432, 'CTT': 0.10341378016756284, 'ACG': 0.26849903016796667, 'GGA': 0.10711946389882603, 'CCG': 0.527785216925549, 'CAA': 0.3469472911172681, 'ACT': 0.16532540959927366, 'ATA': 0.07084223130192197, 'CAT': 0.5707553344678623, 'AGT': 0.15102374892893225, 'TGC': 0.55651396076928, 'GAG': 0.31053200246964396, 'TCG': 0.15406749964830613, 'GTA': 0.15372953305033354, 'TAC': 0.4315172054223149, 'TCC': 0.14895195222078703, 'GAC': 0.37314335931309695, 'TGT': 0.4434860392307201, 'TGG': 1.0, 'TAT': 0.5684827945776851, 'CCT': 0.15789033208355652, 'CCA': 0.19071371183717192, 'CGG': 0.09749348521075679, 'CTA': 0.03649285148597391, 'TAA': 0.6391064025794565, 'AGA': 0.036496252317115764, 'TAG': 0.07162597881160755}), 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.75, 'AAC': 0.5, 'AAG': 0.23999999999999996, 'AAT': 0.29500000000000004, 'ACA': 0.1525, 'ACC': 0.25, 'ACG': 0.29, 'ACT': 0.14750000000000002, 'AGA': 0.13874999999999998, 'AGC': 0.125, 'AGG': 0.16499999999999998, 'AGT': 0.07375000000000001, 'ATA': 0.10999999999999999, 'ATC': 0.375, 'ATG': 1.0, 'ATT': 0.22125000000000003, 'CAA': 0.27749999999999997, 'CAC': 0.125, 'CAG': 0.32999999999999996, 'CAT': 0.07375000000000001, 'CCA': 0.13874999999999998, 'CCC': 0.125, 'CCG': 0.16499999999999998, 'CCT': 0.07375000000000001, 'CGA': 0.013799999999999993, 'CGC': 0.36, 'CGG': 0.125, 'CGT': 0.5, 'CTA': 0.18, 'CTC': 0.125, 'CTG': 0.54, 'CTT': 0.07375000000000001, 'GAA': 0.5, 'GAC': 0.375, 'GAG': 0.15999999999999998, 'GAT': 0.22125000000000003, 'GCA': 0.375, 'GCC': 0.25, 'GCG': 0.11999999999999998, 'GCT': 0.14750000000000002, 'GGA': 0.13874999999999998, 'GGC': 0.5, 'GGG': 0.16499999999999998, 'GGT': 0.29500000000000004, 'GTA': 0.625, 'GTC': 0.25, 'GTG': 0.19999999999999996, 'GTT': 0.14750000000000002, 'TAC': 0.375, 'TAT': 0.22125000000000003, 'TCA': 0.13874999999999998, 'TCC': 0.25, 'TCG': 0.16499999999999998, 'TCT': 0.14750000000000002, 'TGC': 0.125, 'TGG': 0.125, 'TGT': 0.07375000000000001, 'TTA': 0.13874999999999998, 'TTC': 0.25, 'TTG': 0.16499999999999998, 'TTT': 0.14750000000000002}, 'tai_avg': 0.24958213077316901, 'tai_std': 0.024913119712427708}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': None, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'GAA': 0.6798186496752849, 'AAT': 0.5655317497600131, 'ATA': 0.1287010322849589, 'TTA': 0.198530453110291, 'GAC': 0.35960560391899904, 'CTG': 0.23973441522197322, 'TGG': 1.0, 'AAC': 0.4344682502399869, 'CAA': 0.5131548459124133, 'GCC': 0.20715414603308396, 'CTT': 0.2391679232236727, 'GCT': 0.24752923822568748, 'ATC': 0.366762837741479, 'AAA': 0.7016557540022625, 'AAG': 0.29834424599773746, 'TTG': 0.1592175745811708, 'AGC': 0.22534685577492342, 'CCG': 0.4334091260292722, 'AGT': 0.10611598754150686, 'TTT': 0.6847751982489505, 'GAG': 0.3201813503247151, 'ACT': 0.1614000537105004, 'TCA': 0.23653118484388272, 'ACC': 0.1588786441082565, 'CAC': 0.3265866087018494, 'GGC': 0.3390368073029434, 'GAT': 0.640394396081001, 'ACA': 0.4122430101751559, 'ACG': 0.2674782920060872, 'CCC': 0.0889644363010221, 'AGA': 0.26380392311639106, 'TCC': 0.12712038919920718, 'TAC': 0.34586274736720285, 'CAT': 0.6734133912981506, 'ATT': 0.5045361299735621, 'GCA': 0.2828363712991255, 'TAT': 0.6541372526327971, 'GGG': 0.16211598104043537, 'GTC': 0.25598186049286953, 'CCT': 0.28667299498885185, 'GTT': 0.28439644370189154, 'GCG': 0.26248024444210305, 'TCT': 0.2048572678833432, 'CTC': 0.11219040795754642, 'CCA': 0.19095344268085387, 'GGA': 0.3148341038094681, 'CGA': 0.09855152543710778, 'GTA': 0.19774449549495793, 'GTG': 0.26187720031028106, 'TTC': 0.31522480175104955, 'CGC': 0.2064569601768165, 'CAG': 0.4868451540875866, 'CGG': 0.1555630106168844, 'CGT': 0.18161187196589967, 'CTA': 0.05115922590534585, 'AGG': 0.09401270868690058, 'GGT': 0.18401310784715313, 'TAG': 0.14132712456344587, 'TGC': 0.5460141582025239, 'TCG': 0.10002831475713661, 'TAA': 0.630500582072177, 'TGA': 0.2281722933643772, 'TGT': 0.45398584179747614}), 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.6666666666666666, 'AAC': 0.6666666666666666, 'AAG': 0.2133333333333333, 'AAT': 0.39333333333333337, 'ACA': 0.6666666666666666, 'ACC': 0.16666666666666666, 'ACG': 0.2133333333333333, 'ACT': 0.09833333333333334, 'AGA': 0.18499999999999997, 'AGC': 0.3333333333333333, 'AGG': 0.21999999999999997, 'AGT': 0.19666666666666668, 'ATA': 0.10999999999999999, 'ATC': 0.5, 'ATG': 1.0, 'ATT': 0.29500000000000004, 'CAA': 0.6666666666666666, 'CAC': 0.3333333333333333, 'CAG': 0.2133333333333333, 'CAT': 0.19666666666666668, 'CCA': 0.5, 'CCG': 0.15999999999999998, 'CGA': 0.01839999999999999, 'CGC': 0.48, 'CGG': 0.16666666666666666, 'CGT': 0.6666666666666666, 'CTA': 0.3516666666666666, 'CTC': 0.16666666666666666, 'CTG': 0.2733333333333333, 'CTT': 0.09833333333333334, 'GAA': 1.0, 'GAC': 0.6666666666666666, 'GAG': 0.31999999999999995, 'GAT': 0.39333333333333337, 'GCA': 0.8333333333333334, 'GCC': 0.16666666666666666, 'GCG': 0.2666666666666666, 'GCT': 0.09833333333333334, 'GGA': 0.5, 'GGC': 0.6666666666666666, 'GGG': 0.15999999999999998, 'GGT': 0.39333333333333337, 'GTA': 0.6666666666666666, 'GTC': 0.16666666666666666, 'GTG': 0.2133333333333333, 'GTT': 0.09833333333333334, 'TAC': 0.3333333333333333, 'TAT': 0.19666666666666668, 'TCA': 0.3333333333333333, 'TCC': 0.16666666666666666, 'TCG': 0.10666666666666665, 'TCT': 0.09833333333333334, 'TGC': 0.16666666666666666, 'TGG': 0.16666666666666666, 'TGT': 0.09833333333333334, 'TTA': 0.5183333333333333, 'TTC': 0.5, 'TTG': 0.32666666666666666, 'TTT': 0.29500000000000004}, 'tai_avg': 0.3271366648197227, 'tai_std': 0.028076995641981673}]}, 'orf': [{'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGCTTTAAGGTTCATATGGAGGGGTCGGTTAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGTCCGCTGCCGTTTGCTTGGGATATTCTGTCGCCGCAGTTTATGTATGGGTCGAAGGCTTATGTTAAGCATCCGGCTGATATTCCGGATTATCTGAAGCTGTCGTTTCCGGAGGGGTTTAAGTGGGAGCGCGTTATGAATTTTGAGGATGGGGGTGTTGTCACTGTTACTCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTTAAGCTGCGCGGGACTAATTTTCCGTCGGATGGGCCGGTTATGCAGAAGAAAACTATGGGGTGGGAGGCTTCGTCTGAGCGCATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCTGAGGTTAAGACTACCTATAAGGCTAAGAAACCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATTAAGCTGGATATTACTTCGCATAATGAGGATTATACTATTGTTGAGCAGTATGAGCGCGCTGAGGGGCGCCATTCGACTGGGGGTATGGATGAGCTGTATAAGTAA', 'iterations_count': 2, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATT', 'L': 'CTG', 'H': 'CAT', 'R': 'CGC', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'initial_sequence_optimization_score': 1.093869981244511, 'final_sequence_optimization_score': 1.1713347016501405, 'run_time': 12.640999999945052, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATT', 'L': 'CTG', 'H': 'CAT', 'R': 'CGC', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'sequence_score': 1.1713347016501405}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATT', 'L': 'CTG', 'H': 'CAT', 'R': 'CGC', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'sequence_score': 1.2092521312220013}]}, {'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAGGTTCATATGGAAGGGTCAGTCAACGGTCATGAATTTGAGATCGAAGGCGAGGGAGAGGGCCGACCATATGAAGGAACGCAAACTGCTAAACTGAAAGTTACCAAAGGAGGTCCCCTACCGTTCGCCTGGGACATTCTTTCTCCGCAATTTATGTATGGGTCTAAGGCATACGTAAAACATCCAGCAGATATACCGGACTACTTGAAGCTTTCATTTCCCGAAGGTTTTAAGTGGGAAAGAGTAATGAATTTTGAGGATGGGGGCGTGGTCACGGTGACCCAGGACTCCTCTCTCCAAGATGGAGAGTTTATTTACAAGGTGAAGTTGCGCGGAACGAACTTTCCGTCTGACGGCCCGGTTATGCAGAAGAAGACCATGGGTTGGGAAGCGTCGAGTGAGAGAATGTATCCAGAAGACGGGGCATTGAAGGGAGAGATTAAGCAGAGGCTGAAACTAAAAGACGGAGGCCACTACGACGCGGAGGTTAAGACTACGTACAAAGCGAAAAAACCAGTTCAACTACCTGGGGCGTATAACGTGAACATTAAGTTGGACATAACTTCCCATAATGAAGACTATACTATAGTGGAGCAATATGAAAGGGCAGAGGGTAGGCATTCTACTGGGGGGATGGATGAGTTGTACAAGTAA', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGCTTTAAGGTTCATATGGAGGGGTCGGTTAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGTCCGCTGCCGTTTGCTTGGGATATTCTGTCGCCGCAGTTTATGTATGGGTCGAAGGCTTATGTTAAGCATCCGGCTGATATTCCGGATTATCTGAAGCTGTCGTTTCCGGAGGGGTTTAAGTGGGAGCGCGTTATGAATTTTGAGGATGGGGGTGTTGTCACTGTTACTCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTTAAGCTGCGCGGGACTAATTTTCCGTCGGATGGGCCGGTTATGCAGAAGAAAACTATGGGGTGGGAGGCTTCGTCTGAGCGCATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCTGAGGTTAAGACTACCTATAAGGCTAAGAAACCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATTAAGCTGGATATTACTTCGCATAATGAGGATTATACTATTGTTGAGCAGTATGAGCGCGCTGAGGGGCGCCATTCGACTGGGGGTATGGATGAGCTGTATAAGTAA', 'iterations_count': 2, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATT', 'L': 'CTG', 'H': 'CAT', 'R': 'CGC', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'initial_sequence_optimization_score': 0.9767809010220166, 'final_sequence_optimization_score': 1.2422181432770667, 'run_time': 13.734000000054948, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATT', 'L': 'CTG', 'H': 'CAT', 'R': 'CGC', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'sequence_score': 1.2422181432770667}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATT', 'L': 'CTG', 'H': 'CAT', 'R': 'CGC', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'sequence_score': 1.2092521312220013}]}, {'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGATTCAAAGTGCACATGGAGGGATCAGTTAATGGGCATGAATTCGAAATAGAAGGGGAAGGAGAGGGCCGCCCATACGAGGGTACCCAAACTGCTAAGCTCAAGGTGACCAAAGGGGGGCCTTTGCCGTTTGCTTGGGATATACTCTCTCCCCAATTTATGTACGGGTCAAAGGCGTACGTAAAACATCCAGCGGACATTCCAGATTACCTAAAGCTCAGCTTTCCCGAGGGGTTTAAATGGGAGCGGGTCATGAATTTTGAAGATGGCGGGGTGGTAACTGTTACTCAGGACAGTTCACTGCAGGACGGAGAATTCATCTACAAGGTGAAGCTTCGGGGAACCAATTTTCCATCAGATGGTCCTGTGATGCAAAAAAAGACTATGGGATGGGAGGCCTCTAGTGAAAGAATGTACCCGGAAGACGGTGCTCTGAAAGGCGAAATTAAACAACGGTTGAAGCTAAAGGACGGTGGGCATTATGACGCGGAGGTCAAAACTACTTATAAAGCAAAGAAGCCGGTACAGCTTCCAGGTGCCTACAATGTAAACATAAAATTGGATATCACGTCCCATAATGAGGATTACACAATAGTTGAACAATACGAGCGCGCAGAAGGGCGTCACTCGACGGGAGGAATGGATGAGCTGTATAAGTGA', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGCTTTAAGGTTCATATGGAGGGGTCGGTTAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGTCCGCTGCCGTTTGCTTGGGATATTCTGTCGCCGCAGTTTATGTATGGGTCGAAGGCTTATGTTAAGCATCCGGCTGATATTCCGGATTATCTGAAGCTGTCGTTTCCGGAGGGGTTTAAGTGGGAGCGCGTTATGAATTTTGAGGATGGGGGTGTTGTCACTGTTACTCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTTAAGCTGCGCGGGACTAATTTTCCGTCGGATGGGCCGGTTATGCAGAAGAAAACTATGGGGTGGGAGGCTTCGTCTGAGCGCATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCTGAGGTTAAGACTACCTATAAGGCTAAGAAACCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATTAAGCTGGATATTACTTCGCATAATGAGGATTATACTATTGTTGAGCAGTATGAGCGCGCTGAGGGGCGCCATTCGACTGGGGGTATGGATGAGCTGTATAAGTGA', 'iterations_count': 2, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATT', 'L': 'CTG', 'H': 'CAT', 'R': 'CGC', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'initial_sequence_optimization_score': 0.946335884794965, 'final_sequence_optimization_score': 1.2364108162724072, 'run_time': 12.98499999998603, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATT', 'L': 'CTG', 'H': 'CAT', 'R': 'CGC', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'sequence_score': 1.2364108162724072}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATT', 'L': 'CTG', 'H': 'CAT', 'R': 'CGC', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'sequence_score': 1.2092521312220013}]}, {'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGATTTAAGGTACATATGGAGGGTAGTGTTAATGGCCATGAATTCGAGATCGAGGGCGAGGGCGAAGGTAGGCCGTATGAGGGTACGCAGACTGCCAAATTGAAAGTAACAAAGGGGGGTCCTCTCCCCTTCGCATGGGACATCTTAAGTCCACAATTTATGTACGGGTCAAAAGCCTACGTAAAGCATCCCGCGGATATACCTGACTACCTCAAGTTGTCCTTTCCAGAAGGTTTCAAGTGGGAGCGGGTCATGAACTTTGAAGACGGCGGCGTTGTAACCGTCACCCAAGACTCTTCGCTTCAGGATGGAGAGTTTATATACAAGGTAAAACTGCGAGGCACCAATTTTCCTAGTGACGGCCCCGTTATGCAGAAAAAGACCATGGGGTGGGAGGCAAGCTCGGAACGGATGTATCCCGAAGACGGCGCTCTAAAGGGAGAAATCAAGCAGCGATTAAAACTTAAAGACGGTGGGCACTACGACGCCGAAGTAAAAACCACTTACAAAGCCAAGAAACCAGTGCAGCTGCCCGGCGCATATAATGTGAATATAAAATTAGACATAACGTCTCACAATGAAGATTATACAATAGTAGAGCAATACGAGCGTGCCGAAGGTCGACATTCTACCGGTGGCATGGATGAATTATACAAATAG', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGCTTTAAGGTTCATATGGAGGGGTCGGTTAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGTCCGCTGCCGTTTGCTTGGGATATTCTGTCGCCGCAGTTTATGTATGGGTCGAAGGCTTATGTTAAGCATCCGGCTGATATTCCGGATTATCTGAAGCTGTCGTTTCCGGAGGGGTTTAAGTGGGAGCGCGTTATGAATTTTGAGGATGGGGGTGTTGTCACTGTTACTCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTTAAGCTGCGCGGGACTAATTTTCCGTCGGATGGGCCGGTTATGCAGAAGAAAACTATGGGGTGGGAGGCTTCGTCTGAGCGCATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCTGAGGTTAAGACTACCTATAAGGCTAAGAAACCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATTAAGCTGGATATTACTTCGCATAATGAGGATTATACTATTGTTGAGCAGTATGAGCGCGCTGAGGGGCGCCATTCGACTGGGGGTATGGATGAGCTGTATAAGTAG', 'iterations_count': 2, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATT', 'L': 'CTG', 'H': 'CAT', 'R': 'CGC', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'initial_sequence_optimization_score': 0.9505786830378422, 'final_sequence_optimization_score': 1.215059265661401, 'run_time': 13.280999999959022, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATT', 'L': 'CTG', 'H': 'CAT', 'R': 'CGC', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'sequence_score': 1.215059265661401}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATT', 'L': 'CTG', 'H': 'CAT', 'R': 'CGC', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'sequence_score': 1.2092521312220013}]}], 'evaluation': [{'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'tai_initial_score': 0.2816376469779175, 'tai_final_score': 0.20306296762633808, 'dist_score': -3.1539478097712133}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'tai_initial_score': 0.33696708001007003, 'tai_final_score': 0.20892523561271537, 'dist_score': -4.560382671638207}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGCTTTAAGGTTCATATGGAGGGGTCGGTTAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGTCCGCTGCCGTTTGCTTGGGATATTCTGTCGCCGCAGTTTATGTATGGGTCGAAGGCTTATGTTAAGCATCCGGCTGATATTCCGGATTATCTGAAGCTGTCGTTTCCGGAGGGGTTTAAGTGGGAGCGCGTTATGAATTTTGAGGATGGGGGTGTTGTCACTGTTACTCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTTAAGCTGCGCGGGACTAATTTTCCGTCGGATGGGCCGGTTATGCAGAAGAAAACTATGGGGTGGGAGGCTTCGTCTGAGCGCATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCTGAGGTTAAGACTACCTATAAGGCTAAGAAACCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATTAAGCTGGATATTACTTCGCATAATGAGGATTATACTATTGTTGAGCAGTATGAGCGCGCTGAGGGGCGCCATTCGACTGGGGGTATGGATGAGCTGTATAAGTAA', 'average_distance_score': 0.7032174309334969, 'weakest_link_score': 0.7032174309334969, 'ratio_score': 1.0396944557740175, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTTATGCGCTTTAAGGTTCATATGGAGGGGTCGGTTAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGTCCGCTGCCGTTTGCTTGGGATATTCTGTCGCCGCAGTTTATGTATGGGTCGAAGGCTTATGTTAAGCATCCGGCTGATATTCCGGATTATCTGAAGCTGTCGTTTCCGGAGGGGTTTAAGTGGGAGCGCGTTATGAATTTTGAGGATGGGGGTGTTGTCACTGTTACTCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTTAAGCTGCGCGGGACTAATTTTCCGTCGGATGGGCCGGTTATGCAGAAGAAAACTATGGGGTGGGAGGCTTCGTCTGAGCGCATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCTGAGGTTAAGACTACCTATAAGGCTAAGAAACCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATTAAGCTGGATATTACTTCGCATAATGAGGATTATACTATTGTTGAGCAGTATGAGCGCGCTGAGGGGCGCCATTCGACTGGGGGTATGGATGAGCTGTATAAGTAA'}, {'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'tai_initial_score': 0.2816376469779175, 'tai_final_score': 0.20306296762633808, 'dist_score': -3.1539478097712133}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'tai_initial_score': 0.33696708001007003, 'tai_final_score': 0.20892523561271537, 'dist_score': -4.560382671638207}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGCTTTAAGGTTCATATGGAGGGGTCGGTTAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGTCCGCTGCCGTTTGCTTGGGATATTCTGTCGCCGCAGTTTATGTATGGGTCGAAGGCTTATGTTAAGCATCCGGCTGATATTCCGGATTATCTGAAGCTGTCGTTTCCGGAGGGGTTTAAGTGGGAGCGCGTTATGAATTTTGAGGATGGGGGTGTTGTCACTGTTACTCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTTAAGCTGCGCGGGACTAATTTTCCGTCGGATGGGCCGGTTATGCAGAAGAAAACTATGGGGTGGGAGGCTTCGTCTGAGCGCATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCTGAGGTTAAGACTACCTATAAGGCTAAGAAACCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATTAAGCTGGATATTACTTCGCATAATGAGGATTATACTATTGTTGAGCAGTATGAGCGCGCTGAGGGGCGCCATTCGACTGGGGGTATGGATGAGCTGTATAAGTAA', 'average_distance_score': 0.7032174309334969, 'weakest_link_score': 0.7032174309334969, 'ratio_score': 1.0396944557740175, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTTATGCGCTTTAAGGTTCATATGGAGGGGTCGGTTAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGTCCGCTGCCGTTTGCTTGGGATATTCTGTCGCCGCAGTTTATGTATGGGTCGAAGGCTTATGTTAAGCATCCGGCTGATATTCCGGATTATCTGAAGCTGTCGTTTCCGGAGGGGTTTAAGTGGGAGCGCGTTATGAATTTTGAGGATGGGGGTGTTGTCACTGTTACTCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTTAAGCTGCGCGGGACTAATTTTCCGTCGGATGGGCCGGTTATGCAGAAGAAAACTATGGGGTGGGAGGCTTCGTCTGAGCGCATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCTGAGGTTAAGACTACCTATAAGGCTAAGAAACCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATTAAGCTGGATATTACTTCGCATAATGAGGATTATACTATTGTTGAGCAGTATGAGCGCGCTGAGGGGCGCCATTCGACTGGGGGTATGGATGAGCTGTATAAGTAA'}, {'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'tai_initial_score': 0.2816376469779175, 'tai_final_score': 0.20306296762633808, 'dist_score': -3.1539478097712133}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'tai_initial_score': 0.33696708001007003, 'tai_final_score': 0.20892523561271537, 'dist_score': -4.560382671638207}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGCTTTAAGGTTCATATGGAGGGGTCGGTTAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGTCCGCTGCCGTTTGCTTGGGATATTCTGTCGCCGCAGTTTATGTATGGGTCGAAGGCTTATGTTAAGCATCCGGCTGATATTCCGGATTATCTGAAGCTGTCGTTTCCGGAGGGGTTTAAGTGGGAGCGCGTTATGAATTTTGAGGATGGGGGTGTTGTCACTGTTACTCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTTAAGCTGCGCGGGACTAATTTTCCGTCGGATGGGCCGGTTATGCAGAAGAAAACTATGGGGTGGGAGGCTTCGTCTGAGCGCATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCTGAGGTTAAGACTACCTATAAGGCTAAGAAACCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATTAAGCTGGATATTACTTCGCATAATGAGGATTATACTATTGTTGAGCAGTATGAGCGCGCTGAGGGGCGCCATTCGACTGGGGGTATGGATGAGCTGTATAAGTGA', 'average_distance_score': 0.7032174309334969, 'weakest_link_score': 0.7032174309334969, 'ratio_score': 1.0396944557740175, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTTATGCGCTTTAAGGTTCATATGGAGGGGTCGGTTAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGTCCGCTGCCGTTTGCTTGGGATATTCTGTCGCCGCAGTTTATGTATGGGTCGAAGGCTTATGTTAAGCATCCGGCTGATATTCCGGATTATCTGAAGCTGTCGTTTCCGGAGGGGTTTAAGTGGGAGCGCGTTATGAATTTTGAGGATGGGGGTGTTGTCACTGTTACTCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTTAAGCTGCGCGGGACTAATTTTCCGTCGGATGGGCCGGTTATGCAGAAGAAAACTATGGGGTGGGAGGCTTCGTCTGAGCGCATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCTGAGGTTAAGACTACCTATAAGGCTAAGAAACCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATTAAGCTGGATATTACTTCGCATAATGAGGATTATACTATTGTTGAGCAGTATGAGCGCGCTGAGGGGCGCCATTCGACTGGGGGTATGGATGAGCTGTATAAGTGA'}, {'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'tai_initial_score': 0.2816376469779175, 'tai_final_score': 0.20306296762633808, 'dist_score': -3.1539478097712133}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'tai_initial_score': 0.33696708001007003, 'tai_final_score': 0.20892523561271537, 'dist_score': -4.560382671638207}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGCTTTAAGGTTCATATGGAGGGGTCGGTTAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGTCCGCTGCCGTTTGCTTGGGATATTCTGTCGCCGCAGTTTATGTATGGGTCGAAGGCTTATGTTAAGCATCCGGCTGATATTCCGGATTATCTGAAGCTGTCGTTTCCGGAGGGGTTTAAGTGGGAGCGCGTTATGAATTTTGAGGATGGGGGTGTTGTCACTGTTACTCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTTAAGCTGCGCGGGACTAATTTTCCGTCGGATGGGCCGGTTATGCAGAAGAAAACTATGGGGTGGGAGGCTTCGTCTGAGCGCATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCTGAGGTTAAGACTACCTATAAGGCTAAGAAACCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATTAAGCTGGATATTACTTCGCATAATGAGGATTATACTATTGTTGAGCAGTATGAGCGCGCTGAGGGGCGCCATTCGACTGGGGGTATGGATGAGCTGTATAAGTAG', 'average_distance_score': 0.7032174309334969, 'weakest_link_score': 0.7032174309334969, 'ratio_score': 1.0396944557740175, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTTATGCGCTTTAAGGTTCATATGGAGGGGTCGGTTAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGTCCGCTGCCGTTTGCTTGGGATATTCTGTCGCCGCAGTTTATGTATGGGTCGAAGGCTTATGTTAAGCATCCGGCTGATATTCCGGATTATCTGAAGCTGTCGTTTCCGGAGGGGTTTAAGTGGGAGCGCGTTATGAATTTTGAGGATGGGGGTGTTGTCACTGTTACTCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTTAAGCTGCGCGGGACTAATTTTCCGTCGGATGGGCCGGTTATGCAGAAGAAAACTATGGGGTGGGAGGCTTCGTCTGAGCGCATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCTGAGGTTAAGACTACCTATAAGGCTAAGAAACCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATTAAGCTGGATATTACTTCGCATAATGAGGATTATACTATTGTTGAGCAGTATGAGCGCGCTGAGGGGCGCCATTCGACTGGGGGTATGGATGAGCTGTATAAGTAG'}], 'final_evaluation': {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGCTTTAAGGTTCATATGGAGGGGTCGGTTAATGGGCATGAGTTTGAGATTGAGGGGGAGGGGGAGGGGCGCCCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGTCCGCTGCCGTTTGCTTGGGATATTCTGTCGCCGCAGTTTATGTATGGGTCGAAGGCTTATGTTAAGCATCCGGCTGATATTCCGGATTATCTGAAGCTGTCGTTTCCGGAGGGGTTTAAGTGGGAGCGCGTTATGAATTTTGAGGATGGGGGTGTTGTCACTGTTACTCAGGATTCGTCTCTGCAGGATGGGGAGTTTATTTATAAGGTTAAGCTGCGCGGGACTAATTTTCCGTCGGATGGGCCGGTTATGCAGAAGAAAACTATGGGGTGGGAGGCTTCGTCTGAGCGCATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATTAAGCAGCGCCTGAAGCTGAAGGATGGGGGTCATTATGATGCTGAGGTTAAGACTACCTATAAGGCTAAGAAACCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATTAAGCTGGATATTACTTCGCATAATGAGGATTATACTATTGTTGAGCAGTATGAGCGCGCTGAGGGGCGCCATTCGACTGGGGGTATGGATGAGCTGTATAAGTAA', 'average_distance_score': 0.7032174309334969, 'weakest_link_score': 0.7032174309334969, 'ratio_score': 1.0396944557740175}, 'zip_output_file_path': 'results\\\\mcherry-wanted-constraint-init-external-0.1\\\\tAI_zscore_bulk_aa_ratio_ecoli_opt_True_BbAg\\\\communique_results.zip'}\n",
      "\n",
      "##########################\n",
      "# User Input #\n",
      "##########################\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is:\n",
      "ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "------------------------------------------\n",
      "Parsing information for Bacillus subtilis:\n",
      "------------------------------------------\n",
      "Organism is defined as wanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Bacillus subtilis. Calculating tAI profile.\n",
      "tai_std=0.028076995641981673, tai_avg=0.3271366648197227\n",
      "------------------------------------------\n",
      "Parsing information for Escherichia coli:\n",
      "------------------------------------------\n",
      "Organism is defined as unwanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Escherichia coli. Calculating tAI profile.\n",
      "tai_std=0.024913119712427708, tai_avg=0.24958213077316901\n",
      "-----------------------------\n",
      "Normalized Prioritization Weights\n",
      "-----------------------------\n",
      "Bacillus subtilis : 1.0\n",
      "Escherichia coli : 1.0\n",
      "Total input processing time: 28.133147478103638\n",
      "\n",
      "##########################\n",
      "# Initiation #\n",
      "##########################\n",
      "Running initiation optimization InitiationOptimizationMethod.external_module on 15 codons at the start of the ORF\n",
      "Taking optimized sequence from configuration value config.INITIATION.EXTERNAL_INITIATION_ORF\n",
      "Optimized sequence is: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "Optimization method used is: ORFOptimizationMethod.zscore_bulk_aa_ratio\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGGTTTAAAGTTCATATGGAAGGCAGCGTTAATGGGCATGAGTTCGAAATAGAGGGTGAGGGCGAGGGTCGTCCGTATGAAGGCACTCAGACGGCCAAACTCAAAGTTACGAAAGGGGGGCCCTTACCGTTTGCATGGGATATTCTTTCACCGCAGTTTATGTACGGTTCAAAGGCTTACGTTAAGCATCCTGCGGATATCCCTGACTACCTTAAGTTAAGCTTCCCGGAGGGATTCAAATGGGAGCGCGTGATGAATTTTGAAGATGGGGGGGTGGTTACAGTTACTCAGGACAGTTCTCTCCAGGACGGAGAGTTTATATATAAGGTAAAACTCCGGGGCACTAACTTCCCCAGTGACGGCCCTGTCATGCAGAAAAAAACAATGGGATGGGAGGCTTCCTCAGAGAGGATGTATCCCGAGGACGGTGCACTAAAGGGAGAGATCAAACAGCGTCTTAAACTCAAGGACGGGGGTCACTATGATGCGGAAGTTAAGACGACGTATAAAGCAAAAAAGCCCGTGCAGTTACCTGGTGCCTACAACGTGAACATTAAACTCGACATTACGAGCCATAATGAGGATTACACAATTGTGGAGCAGTACGAGCGCGCTGAGGGACGTCATTCAACGGGTGGGATGGACGAACTCTACAAATAA\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTTAAAGTTCACATGGAAGGCTCTGTTAACGGGCACGAATTCGAAATCGAAGGAGAGGGTGAGGGACGCCCGTATGAGGGAACCCAAACTGCAAAATTGAAAGTGACGAAAGGGGGCCCATTACCTTTCGCCTGGGACATTCTCTCTCCTCAGTTCATGTACGGCTCAAAAGCGTATGTTAAGCACCCAGCGGATATACCCGATTATTTAAAACTTTCGTTTCCGGAAGGATTTAAATGGGAGCGGGTCATGAATTTCGAAGATGGTGGGGTGGTTACCGTTACTCAGGATTCCTCCCTGCAAGATGGGGAATTTATTTACAAGGTCAAATTGCGTGGGACAAATTTTCCCTCGGACGGTCCAGTCATGCAGAAGAAAACGATGGGGTGGGAGGCATCGTCTGAACGTATGTACCCCGAGGACGGCGCACTGAAAGGAGAGATAAAGCAACGATTAAAGTTGAAAGATGGTGGGCACTATGATGCGGAGGTAAAGACTACGTATAAAGCAAAAAAGCCCGTGCAGCTACCGGGCGCCTATAACGTAAACATCAAGCTGGACATTACGAGCCATAACGAGGACTATACAATTGTCGAACAGTATGAACGTGCGGAAGGCCGGCACTCGACAGGTGGAATGGACGAACTCTACAAGTAA\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGGTTCAAAGTGCATATGGAAGGGTCTGTGAATGGGCATGAGTTTGAAATTGAGGGTGAGGGAGAGGGACGACCCTATGAAGGTACTCAAACCGCTAAATTAAAAGTCACAAAAGGAGGGCCACTACCATTCGCTTGGGATATTCTCTCTCCACAGTTTATGTATGGGTCAAAGGCCTACGTGAAACACCCCGCAGACATACCTGATTACCTAAAACTGAGTTTCCCTGAGGGTTTCAAATGGGAGAGGGTAATGAACTTTGAAGATGGTGGGGTAGTCACTGTTACTCAAGACTCATCGCTACAGGACGGAGAATTTATCTACAAAGTGAAGTTAAGAGGAACCAATTTCCCTTCTGACGGTCCGGTAATGCAGAAAAAGACGATGGGCTGGGAGGCATCATCGGAAAGGATGTATCCAGAAGATGGGGCGCTAAAAGGTGAAATTAAGCAACGTTTGAAATTAAAAGATGGTGGCCATTACGACGCCGAAGTGAAGACCACTTACAAAGCTAAAAAACCTGTCCAATTGCCAGGAGCGTACAATGTTAATATTAAGCTGGATATCACTTCTCATAACGAGGATTATACCATTGTTGAGCAATATGAGCGTGCGGAAGGCCGACACAGTACCGGCGGTATGGATGAATTATATAAATGA\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "\n",
      "##########################\n",
      "# EVALUATION #\n",
      "##########################\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCTTACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCTTTACCTTTCGCATGGGACATCTTAAGCCCTCAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCTGCAGACATCCCTGACTACTTAAAATTAAGCTTCCCTGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCAGTTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCTAGCGACGGACCTGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCAGTGAACGTATGTACCCTGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCTGTACAATTACCTGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCTTACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCTTTACCTTTCGCATGGGACATCTTAAGCCCTCAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCTGCAGACATCCCTGACTACTTAAAATTAAGCTTCCCTGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCAGTTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCTAGCGACGGACCTGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCAGTGAACGTATGTACCCTGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCTGTACAATTACCTGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCTTACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCTTTACCTTTCGCATGGGACATCTTAAGCCCTCAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCTGCAGACATCCCTGACTACTTAAAATTAAGCTTCCCTGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCAGTTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCTAGCGACGGACCTGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCAGTGAACGTATGTACCCTGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCTGTACAATTACCTGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCTTACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCTTTACCTTTCGCATGGGACATCTTAAGCCCTCAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCTGCAGACATCCCTGACTACTTAAAATTAAGCTTCCCTGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCAGTTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCTAGCGACGGACCTGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCAGTGAACGTATGTACCCTGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCTGTACAATTACCTGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATGA\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCTTACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCTTTACCTTTCGCATGGGACATCTTAAGCCCTCAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCTGCAGACATCCCTGACTACTTAAAATTAAGCTTCCCTGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCAGTTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCTAGCGACGGACCTGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCAGTGAACGTATGTACCCTGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCTGTACAATTACCTGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA', 'average_distance_score': 3.879985526875157, 'weakest_link_score': 3.879985526875157, 'ratio_score': 1.162982509615646}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCTTACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCTTTACCTTTCGCATGGGACATCTTAAGCCCTCAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCTGCAGACATCCCTGACTACTTAAAATTAAGCTTCCCTGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCAGTTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCTAGCGACGGACCTGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCAGTGAACGTATGTACCCTGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCTGTACAATTACCTGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA', 'average_distance_score': 3.879985526875157, 'weakest_link_score': 3.879985526875157, 'ratio_score': 1.162982509615646}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCTTACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCTTTACCTTTCGCATGGGACATCTTAAGCCCTCAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCTGCAGACATCCCTGACTACTTAAAATTAAGCTTCCCTGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCAGTTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCTAGCGACGGACCTGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCAGTGAACGTATGTACCCTGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCTGTACAATTACCTGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA', 'average_distance_score': 3.879985526875157, 'weakest_link_score': 3.879985526875157, 'ratio_score': 1.162982509615646}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCTTACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCTTTACCTTTCGCATGGGACATCTTAAGCCCTCAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCTGCAGACATCCCTGACTACTTAAAATTAAGCTTCCCTGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCAGTTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCTAGCGACGGACCTGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCAGTGAACGTATGTACCCTGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCTGTACAATTACCTGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATGA', 'average_distance_score': 3.879985526875157, 'weakest_link_score': 3.879985526875157, 'ratio_score': 1.162982509615646}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCTTACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCTTTACCTTTCGCATGGGACATCTTAAGCCCTCAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCTGCAGACATCCCTGACTACTTAAAATTAAGCTTCCCTGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCAGTTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCTAGCGACGGACCTGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCAGTGAACGTATGTACCCTGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCTGTACAATTACCTGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA', 'average_distance_score': 3.879985526875157, 'weakest_link_score': 3.879985526875157, 'ratio_score': 1.162982509615646}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry-wanted-constraint-init-external-0.1\\tAI_zscore_bulk_aa_ratio_ecoli_opt_False_cMCN\n",
      "Final output: {'module_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'orf_optimization_method': 'zscore_bulk_aa_ratio', 'orf_optimization_cub_index': 'tAI', 'initiation_optimization_method': 'external', 'evaluation_score': 'average_distance', 'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': None, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'GAA': 0.6798186496752849, 'AAT': 0.5655317497600131, 'ATA': 0.1287010322849589, 'TTA': 0.198530453110291, 'GAC': 0.35960560391899904, 'CTG': 0.23973441522197322, 'TGG': 1.0, 'AAC': 0.4344682502399869, 'CAA': 0.5131548459124133, 'GCC': 0.20715414603308396, 'CTT': 0.2391679232236727, 'GCT': 0.24752923822568748, 'ATC': 0.366762837741479, 'AAA': 0.7016557540022625, 'AAG': 0.29834424599773746, 'TTG': 0.1592175745811708, 'AGC': 0.22534685577492342, 'CCG': 0.4334091260292722, 'AGT': 0.10611598754150686, 'TTT': 0.6847751982489505, 'GAG': 0.3201813503247151, 'ACT': 0.1614000537105004, 'TCA': 0.23653118484388272, 'ACC': 0.1588786441082565, 'CAC': 0.3265866087018494, 'GGC': 0.3390368073029434, 'GAT': 0.640394396081001, 'ACA': 0.4122430101751559, 'ACG': 0.2674782920060872, 'CCC': 0.0889644363010221, 'AGA': 0.26380392311639106, 'TCC': 0.12712038919920718, 'TAC': 0.34586274736720285, 'CAT': 0.6734133912981506, 'ATT': 0.5045361299735621, 'GCA': 0.2828363712991255, 'TAT': 0.6541372526327971, 'GGG': 0.16211598104043537, 'GTC': 0.25598186049286953, 'CCT': 0.28667299498885185, 'GTT': 0.28439644370189154, 'GCG': 0.26248024444210305, 'TCT': 0.2048572678833432, 'CTC': 0.11219040795754642, 'CCA': 0.19095344268085387, 'GGA': 0.3148341038094681, 'CGA': 0.09855152543710778, 'GTA': 0.19774449549495793, 'GTG': 0.26187720031028106, 'TTC': 0.31522480175104955, 'CGC': 0.2064569601768165, 'CAG': 0.4868451540875866, 'CGG': 0.1555630106168844, 'CGT': 0.18161187196589967, 'CTA': 0.05115922590534585, 'AGG': 0.09401270868690058, 'GGT': 0.18401310784715313, 'TAG': 0.14132712456344587, 'TGC': 0.5460141582025239, 'TCG': 0.10002831475713661, 'TAA': 0.630500582072177, 'TGA': 0.2281722933643772, 'TGT': 0.45398584179747614}), 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.6666666666666666, 'AAC': 0.6666666666666666, 'AAG': 0.2133333333333333, 'AAT': 0.39333333333333337, 'ACA': 0.6666666666666666, 'ACC': 0.16666666666666666, 'ACG': 0.2133333333333333, 'ACT': 0.09833333333333334, 'AGA': 0.18499999999999997, 'AGC': 0.3333333333333333, 'AGG': 0.21999999999999997, 'AGT': 0.19666666666666668, 'ATA': 0.10999999999999999, 'ATC': 0.5, 'ATG': 1.0, 'ATT': 0.29500000000000004, 'CAA': 0.6666666666666666, 'CAC': 0.3333333333333333, 'CAG': 0.2133333333333333, 'CAT': 0.19666666666666668, 'CCA': 0.5, 'CCG': 0.15999999999999998, 'CGA': 0.01839999999999999, 'CGC': 0.48, 'CGG': 0.16666666666666666, 'CGT': 0.6666666666666666, 'CTA': 0.3516666666666666, 'CTC': 0.16666666666666666, 'CTG': 0.2733333333333333, 'CTT': 0.09833333333333334, 'GAA': 1.0, 'GAC': 0.6666666666666666, 'GAG': 0.31999999999999995, 'GAT': 0.39333333333333337, 'GCA': 0.8333333333333334, 'GCC': 0.16666666666666666, 'GCG': 0.2666666666666666, 'GCT': 0.09833333333333334, 'GGA': 0.5, 'GGC': 0.6666666666666666, 'GGG': 0.15999999999999998, 'GGT': 0.39333333333333337, 'GTA': 0.6666666666666666, 'GTC': 0.16666666666666666, 'GTG': 0.2133333333333333, 'GTT': 0.09833333333333334, 'TAC': 0.3333333333333333, 'TAT': 0.19666666666666668, 'TCA': 0.3333333333333333, 'TCC': 0.16666666666666666, 'TCG': 0.10666666666666665, 'TCT': 0.09833333333333334, 'TGC': 0.16666666666666666, 'TGG': 0.16666666666666666, 'TGT': 0.09833333333333334, 'TTA': 0.5183333333333333, 'TTC': 0.5, 'TTG': 0.32666666666666666, 'TTT': 0.29500000000000004}, 'tai_avg': 0.3271366648197227, 'tai_std': 0.028076995641981673}, {'name': 'Escherichia coli', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': None, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'AAA': 0.7667390608953314, 'CGC': 0.40076028262096014, 'ATT': 0.5085852538544149, 'AGC': 0.27799163607995603, 'ACC': 0.4359429381095841, 'ATC': 0.4205725148436632, 'ACA': 0.13023262212317555, 'GGT': 0.33726003930064996, 'AAC': 0.5512262248110238, 'GCG': 0.3563679190097728, 'GGC': 0.4050486219579785, 'TGA': 0.28926761860893596, 'CGA': 0.06388523224887838, 'GTG': 0.3715940695510341, 'TTG': 0.12808970030428077, 'AAG': 0.23326093910466858, 'TTC': 0.4262385670731707, 'TCA': 0.12251736088908215, 'GCA': 0.21281773299119433, 'AAT': 0.4487737751889762, 'GAA': 0.689467997530356, 'CGT': 0.38120248233619, 'TTT': 0.5737614329268292, 'CTG': 0.4975546385450099, 'GTT': 0.2584847661069405, 'GCC': 0.26990242799092135, 'GAT': 0.626856640686903, 'AGG': 0.020162265266098916, 'CAG': 0.6530527088827319, 'GGG': 0.15057187484254547, 'GTC': 0.21619163129169194, 'CTC': 0.10435163186194824, 'TCT': 0.14544780223293646, 'CCC': 0.12361073915372255, 'CAC': 0.4292446655321377, 'GCT': 0.1609119200081115, 'TTA': 0.13009739763522432, 'CTT': 0.10341378016756284, 'ACG': 0.26849903016796667, 'GGA': 0.10711946389882603, 'CCG': 0.527785216925549, 'CAA': 0.3469472911172681, 'ACT': 0.16532540959927366, 'ATA': 0.07084223130192197, 'CAT': 0.5707553344678623, 'AGT': 0.15102374892893225, 'TGC': 0.55651396076928, 'GAG': 0.31053200246964396, 'TCG': 0.15406749964830613, 'GTA': 0.15372953305033354, 'TAC': 0.4315172054223149, 'TCC': 0.14895195222078703, 'GAC': 0.37314335931309695, 'TGT': 0.4434860392307201, 'TGG': 1.0, 'TAT': 0.5684827945776851, 'CCT': 0.15789033208355652, 'CCA': 0.19071371183717192, 'CGG': 0.09749348521075679, 'CTA': 0.03649285148597391, 'TAA': 0.6391064025794565, 'AGA': 0.036496252317115764, 'TAG': 0.07162597881160755}), 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.75, 'AAC': 0.5, 'AAG': 0.23999999999999996, 'AAT': 0.29500000000000004, 'ACA': 0.1525, 'ACC': 0.25, 'ACG': 0.29, 'ACT': 0.14750000000000002, 'AGA': 0.13874999999999998, 'AGC': 0.125, 'AGG': 0.16499999999999998, 'AGT': 0.07375000000000001, 'ATA': 0.10999999999999999, 'ATC': 0.375, 'ATG': 1.0, 'ATT': 0.22125000000000003, 'CAA': 0.27749999999999997, 'CAC': 0.125, 'CAG': 0.32999999999999996, 'CAT': 0.07375000000000001, 'CCA': 0.13874999999999998, 'CCC': 0.125, 'CCG': 0.16499999999999998, 'CCT': 0.07375000000000001, 'CGA': 0.013799999999999993, 'CGC': 0.36, 'CGG': 0.125, 'CGT': 0.5, 'CTA': 0.18, 'CTC': 0.125, 'CTG': 0.54, 'CTT': 0.07375000000000001, 'GAA': 0.5, 'GAC': 0.375, 'GAG': 0.15999999999999998, 'GAT': 0.22125000000000003, 'GCA': 0.375, 'GCC': 0.25, 'GCG': 0.11999999999999998, 'GCT': 0.14750000000000002, 'GGA': 0.13874999999999998, 'GGC': 0.5, 'GGG': 0.16499999999999998, 'GGT': 0.29500000000000004, 'GTA': 0.625, 'GTC': 0.25, 'GTG': 0.19999999999999996, 'GTT': 0.14750000000000002, 'TAC': 0.375, 'TAT': 0.22125000000000003, 'TCA': 0.13874999999999998, 'TCC': 0.25, 'TCG': 0.16499999999999998, 'TCT': 0.14750000000000002, 'TGC': 0.125, 'TGG': 0.125, 'TGT': 0.07375000000000001, 'TTA': 0.13874999999999998, 'TTC': 0.25, 'TTG': 0.16499999999999998, 'TTT': 0.14750000000000002}, 'tai_avg': 0.24958213077316901, 'tai_std': 0.024913119712427708}]}, 'orf': [{'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCTTACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCTTTACCTTTCGCATGGGACATCTTAAGCCCTCAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCTGCAGACATCCCTGACTACTTAAAATTAAGCTTCCCTGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCAGTTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCTAGCGACGGACCTGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCAGTGAACGTATGTACCCTGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCTGTACAATTACCTGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA', 'iterations_count': 2, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'initial_sequence_optimization_score': 0.9141854307605061, 'final_sequence_optimization_score': 1.2546222234313034, 'run_time': 12.844000000040978, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 1.2546222234313034}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 1.290124356708438}]}, {'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGGTTTAAAGTTCATATGGAAGGCAGCGTTAATGGGCATGAGTTCGAAATAGAGGGTGAGGGCGAGGGTCGTCCGTATGAAGGCACTCAGACGGCCAAACTCAAAGTTACGAAAGGGGGGCCCTTACCGTTTGCATGGGATATTCTTTCACCGCAGTTTATGTACGGTTCAAAGGCTTACGTTAAGCATCCTGCGGATATCCCTGACTACCTTAAGTTAAGCTTCCCGGAGGGATTCAAATGGGAGCGCGTGATGAATTTTGAAGATGGGGGGGTGGTTACAGTTACTCAGGACAGTTCTCTCCAGGACGGAGAGTTTATATATAAGGTAAAACTCCGGGGCACTAACTTCCCCAGTGACGGCCCTGTCATGCAGAAAAAAACAATGGGATGGGAGGCTTCCTCAGAGAGGATGTATCCCGAGGACGGTGCACTAAAGGGAGAGATCAAACAGCGTCTTAAACTCAAGGACGGGGGTCACTATGATGCGGAAGTTAAGACGACGTATAAAGCAAAAAAGCCCGTGCAGTTACCTGGTGCCTACAACGTGAACATTAAACTCGACATTACGAGCCATAATGAGGATTACACAATTGTGGAGCAGTACGAGCGCGCTGAGGGACGTCATTCAACGGGTGGGATGGACGAACTCTACAAATAA', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCTTACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCTTTACCTTTCGCATGGGACATCTTAAGCCCTCAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCTGCAGACATCCCTGACTACTTAAAATTAAGCTTCCCTGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCAGTTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCTAGCGACGGACCTGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCAGTGAACGTATGTACCCTGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCTGTACAATTACCTGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA', 'iterations_count': 2, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'initial_sequence_optimization_score': 1.0563009009680548, 'final_sequence_optimization_score': 1.2252298508960495, 'run_time': 12.827999999979511, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 1.2252298508960495}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 1.290124356708438}]}, {'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTTAAAGTTCACATGGAAGGCTCTGTTAACGGGCACGAATTCGAAATCGAAGGAGAGGGTGAGGGACGCCCGTATGAGGGAACCCAAACTGCAAAATTGAAAGTGACGAAAGGGGGCCCATTACCTTTCGCCTGGGACATTCTCTCTCCTCAGTTCATGTACGGCTCAAAAGCGTATGTTAAGCACCCAGCGGATATACCCGATTATTTAAAACTTTCGTTTCCGGAAGGATTTAAATGGGAGCGGGTCATGAATTTCGAAGATGGTGGGGTGGTTACCGTTACTCAGGATTCCTCCCTGCAAGATGGGGAATTTATTTACAAGGTCAAATTGCGTGGGACAAATTTTCCCTCGGACGGTCCAGTCATGCAGAAGAAAACGATGGGGTGGGAGGCATCGTCTGAACGTATGTACCCCGAGGACGGCGCACTGAAAGGAGAGATAAAGCAACGATTAAAGTTGAAAGATGGTGGGCACTATGATGCGGAGGTAAAGACTACGTATAAAGCAAAAAAGCCCGTGCAGCTACCGGGCGCCTATAACGTAAACATCAAGCTGGACATTACGAGCCATAACGAGGACTATACAATTGTCGAACAGTATGAACGTGCGGAAGGCCGGCACTCGACAGGTGGAATGGACGAACTCTACAAGTAA', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCTTACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCTTTACCTTTCGCATGGGACATCTTAAGCCCTCAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCTGCAGACATCCCTGACTACTTAAAATTAAGCTTCCCTGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCAGTTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCTAGCGACGGACCTGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCAGTGAACGTATGTACCCTGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCTGTACAATTACCTGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'initial_sequence_optimization_score': 1.0599009617014645, 'final_sequence_optimization_score': 1.2903495874781674, 'run_time': 19.719000000040978, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGT', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 1.2548374457561324}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 1.2903495874781674}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 1.290124356708438}]}, {'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGGTTCAAAGTGCATATGGAAGGGTCTGTGAATGGGCATGAGTTTGAAATTGAGGGTGAGGGAGAGGGACGACCCTATGAAGGTACTCAAACCGCTAAATTAAAAGTCACAAAAGGAGGGCCACTACCATTCGCTTGGGATATTCTCTCTCCACAGTTTATGTATGGGTCAAAGGCCTACGTGAAACACCCCGCAGACATACCTGATTACCTAAAACTGAGTTTCCCTGAGGGTTTCAAATGGGAGAGGGTAATGAACTTTGAAGATGGTGGGGTAGTCACTGTTACTCAAGACTCATCGCTACAGGACGGAGAATTTATCTACAAAGTGAAGTTAAGAGGAACCAATTTCCCTTCTGACGGTCCGGTAATGCAGAAAAAGACGATGGGCTGGGAGGCATCATCGGAAAGGATGTATCCAGAAGATGGGGCGCTAAAAGGTGAAATTAAGCAACGTTTGAAATTAAAAGATGGTGGCCATTACGACGCCGAAGTGAAGACCACTTACAAAGCTAAAAAACCTGTCCAATTGCCAGGAGCGTACAATGTTAATATTAAGCTGGATATCACTTCTCATAACGAGGATTATACCATTGTTGAGCAATATGAGCGTGCGGAAGGCCGACACAGTACCGGCGGTATGGATGAATTATATAAATGA', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCTTACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCTTTACCTTTCGCATGGGACATCTTAAGCCCTCAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCTGCAGACATCCCTGACTACTTAAAATTAAGCTTCCCTGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCAGTTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCTAGCGACGGACCTGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCAGTGAACGTATGTACCCTGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCTGTACAATTACCTGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATGA', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'initial_sequence_optimization_score': 1.1059214927073737, 'final_sequence_optimization_score': 1.2708768774437023, 'run_time': 19.45299999997951, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGT', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCG', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 1.2628673657733738}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 1.2708768774437023}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 1.290124356708438}]}], 'evaluation': [{'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'tai_initial_score': 0.33696708001007003, 'tai_final_score': 0.557116807081934, 'dist_score': 7.840928918430598}, {'name': 'Escherichia coli', 'is_wanted': False, 'tai_initial_score': 0.2816376469779175, 'tai_final_score': 0.28365455995235994, 'dist_score': 0.0809578646802846}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCTTACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCTTTACCTTTCGCATGGGACATCTTAAGCCCTCAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCTGCAGACATCCCTGACTACTTAAAATTAAGCTTCCCTGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCAGTTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCTAGCGACGGACCTGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCAGTGAACGTATGTACCCTGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCTGTACAATTACCTGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA', 'average_distance_score': 3.879985526875157, 'weakest_link_score': 3.879985526875157, 'ratio_score': 1.162982509615646, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCTTACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCTTTACCTTTCGCATGGGACATCTTAAGCCCTCAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCTGCAGACATCCCTGACTACTTAAAATTAAGCTTCCCTGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCAGTTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCTAGCGACGGACCTGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCAGTGAACGTATGTACCCTGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCTGTACAATTACCTGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA'}, {'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'tai_initial_score': 0.33696708001007003, 'tai_final_score': 0.557116807081934, 'dist_score': 7.840928918430598}, {'name': 'Escherichia coli', 'is_wanted': False, 'tai_initial_score': 0.2816376469779175, 'tai_final_score': 0.28365455995235994, 'dist_score': 0.0809578646802846}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCTTACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCTTTACCTTTCGCATGGGACATCTTAAGCCCTCAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCTGCAGACATCCCTGACTACTTAAAATTAAGCTTCCCTGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCAGTTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCTAGCGACGGACCTGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCAGTGAACGTATGTACCCTGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCTGTACAATTACCTGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA', 'average_distance_score': 3.879985526875157, 'weakest_link_score': 3.879985526875157, 'ratio_score': 1.162982509615646, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCTTACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCTTTACCTTTCGCATGGGACATCTTAAGCCCTCAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCTGCAGACATCCCTGACTACTTAAAATTAAGCTTCCCTGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCAGTTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCTAGCGACGGACCTGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCAGTGAACGTATGTACCCTGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCTGTACAATTACCTGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA'}, {'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'tai_initial_score': 0.33696708001007003, 'tai_final_score': 0.557116807081934, 'dist_score': 7.840928918430598}, {'name': 'Escherichia coli', 'is_wanted': False, 'tai_initial_score': 0.2816376469779175, 'tai_final_score': 0.28365455995235994, 'dist_score': 0.0809578646802846}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCTTACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCTTTACCTTTCGCATGGGACATCTTAAGCCCTCAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCTGCAGACATCCCTGACTACTTAAAATTAAGCTTCCCTGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCAGTTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCTAGCGACGGACCTGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCAGTGAACGTATGTACCCTGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCTGTACAATTACCTGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA', 'average_distance_score': 3.879985526875157, 'weakest_link_score': 3.879985526875157, 'ratio_score': 1.162982509615646, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCTTACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCTTTACCTTTCGCATGGGACATCTTAAGCCCTCAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCTGCAGACATCCCTGACTACTTAAAATTAAGCTTCCCTGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCAGTTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCTAGCGACGGACCTGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCAGTGAACGTATGTACCCTGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCTGTACAATTACCTGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA'}, {'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'tai_initial_score': 0.33696708001007003, 'tai_final_score': 0.557116807081934, 'dist_score': 7.840928918430598}, {'name': 'Escherichia coli', 'is_wanted': False, 'tai_initial_score': 0.2816376469779175, 'tai_final_score': 0.28365455995235994, 'dist_score': 0.0809578646802846}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCTTACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCTTTACCTTTCGCATGGGACATCTTAAGCCCTCAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCTGCAGACATCCCTGACTACTTAAAATTAAGCTTCCCTGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCAGTTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCTAGCGACGGACCTGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCAGTGAACGTATGTACCCTGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCTGTACAATTACCTGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATGA', 'average_distance_score': 3.879985526875157, 'weakest_link_score': 3.879985526875157, 'ratio_score': 1.162982509615646, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCTTACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCTTTACCTTTCGCATGGGACATCTTAAGCCCTCAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCTGCAGACATCCCTGACTACTTAAAATTAAGCTTCCCTGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCAGTTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCTAGCGACGGACCTGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCAGTGAACGTATGTACCCTGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCTGTACAATTACCTGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATGA'}], 'final_evaluation': {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCTTACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCTTTACCTTTCGCATGGGACATCTTAAGCCCTCAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCTGCAGACATCCCTGACTACTTAAAATTAAGCTTCCCTGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCAGTTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCTAGCGACGGACCTGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCAGTGAACGTATGTACCCTGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCTGTACAATTACCTGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA', 'average_distance_score': 3.879985526875157, 'weakest_link_score': 3.879985526875157, 'ratio_score': 1.162982509615646}, 'zip_output_file_path': 'results\\\\mcherry-wanted-constraint-init-external-0.1\\\\tAI_zscore_bulk_aa_ratio_ecoli_opt_False_cMCN\\\\communique_results.zip'}\n",
      "\n",
      "##########################\n",
      "# User Input #\n",
      "##########################\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is:\n",
      "ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "------------------------------------------\n",
      "Parsing information for Escherichia coli:\n",
      "------------------------------------------\n",
      "Organism is defined as wanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 90 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3893\n",
      "Calculate CAI weights from a reference set of 584 highly expressed genes from estimated expression dictionary.\n",
      "cai_std=0.09377868043841883, cai_avg=0.6151304741138522\n",
      "------------------------------------------\n",
      "Parsing information for Bacillus subtilis:\n",
      "------------------------------------------\n",
      "Organism is defined as unwanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 98 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3544\n",
      "Calculate CAI weights from a reference set of 532 highly expressed genes from estimated expression dictionary.\n",
      "cai_std=0.042802002292947894, cai_avg=0.7334925635409068\n",
      "-----------------------------\n",
      "Normalized Prioritization Weights\n",
      "-----------------------------\n",
      "Escherichia coli : 1.0\n",
      "Bacillus subtilis : 1.0\n",
      "Total input processing time: 27.647403717041016\n",
      "\n",
      "##########################\n",
      "# Initiation #\n",
      "##########################\n",
      "Running initiation optimization InitiationOptimizationMethod.external_module on 15 codons at the start of the ORF\n",
      "Taking optimized sequence from configuration value config.INITIATION.EXTERNAL_INITIATION_ORF\n",
      "Optimized sequence is: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "Optimization method used is: ORFOptimizationMethod.zscore_bulk_aa_diff\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CTA due to very low average frequency 0.03649285148597391 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGGTTTAAAGTCCATATGGAGGGTTCTGTAAATGGGCATGAATTCGAGATAGAAGGCGAGGGCGAGGGCCGCCCTTACGAAGGCACACAAACTGCTAAACTCAAAGTCACCAAAGGGGGGCCACTCCCTTTCGCATGGGATATATTATCCCCGCAGTTTATGTATGGTAGTAAAGCCTACGTGAAGCACCCAGCGGATATCCCAGACTACCTAAAATTGAGTTTCCCTGAAGGTTTCAAATGGGAGCGTGTTATGAATTTCGAGGACGGTGGTGTCGTAACAGTAACGCAAGATTCAAGTTTACAAGATGGCGAATTCATTTATAAAGTTAAACTAAGAGGAACAAACTTTCCCTCTGACGGACCGGTTATGCAGAAGAAGACCATGGGTTGGGAGGCAAGTTCCGAGCGTATGTACCCTGAAGACGGCGCACTAAAGGGCGAAATCAAGCAGCGGTTGAAGCTCAAAGACGGGGGCCACTACGATGCGGAGGTCAAAACTACGTATAAGGCAAAGAAACCTGTGCAATTACCTGGTGCCTACAATGTCAATATTAAGCTTGATATCACGAGTCATAACGAGGACTATACCATCGTCGAACAGTATGAACGTGCAGAGGGGCGACATTCTACTGGGGGGATGGACGAGTTATACAAATGA\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CTA due to very low average frequency 0.03649285148597391 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGATTCAAGGTCCACATGGAGGGGTCGGTGAATGGACATGAATTCGAGATAGAAGGCGAAGGAGAAGGTCGGCCCTACGAGGGCACTCAGACAGCAAAATTAAAGGTCACAAAGGGGGGCCCTTTACCGTTTGCGTGGGATATTCTGTCCCCGCAATTCATGTACGGTAGCAAAGCGTATGTTAAGCATCCTGCGGACATCCCAGACTACCTAAAGCTGTCATTTCCAGAGGGCTTTAAATGGGAACGAGTTATGAATTTCGAGGACGGGGGCGTTGTGACAGTGACGCAAGACAGCAGTCTTCAAGATGGGGAATTTATTTATAAAGTTAAACTCAGAGGCACTAACTTTCCCTCGGACGGGCCAGTTATGCAGAAAAAGACAATGGGTTGGGAGGCCAGCTCCGAGCGCATGTATCCTGAGGACGGTGCACTTAAAGGAGAAATCAAACAAAGGCTCAAGCTTAAGGACGGTGGGCATTACGATGCAGAGGTCAAAACAACGTATAAAGCAAAAAAGCCTGTCCAGCTTCCAGGAGCATACAACGTGAATATAAAATTAGATATTACATCTCATAATGAGGACTATACTATCGTGGAGCAATATGAAAGAGCTGAAGGAAGGCATAGTACTGGCGGTATGGATGAGTTATATAAATGA\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CTA due to very low average frequency 0.03649285148597391 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAGGTGCATATGGAGGGATCCGTGAACGGACACGAGTTTGAAATCGAGGGGGAAGGCGAAGGACGACCATACGAAGGTACGCAGACAGCGAAATTGAAGGTGACGAAGGGAGGACCCCTACCCTTTGCTTGGGACATCTTATCGCCGCAATTTATGTATGGATCGAAGGCTTATGTAAAGCATCCAGCTGACATCCCGGACTACCTGAAACTCTCTTTTCCAGAGGGATTTAAGTGGGAAAGGGTTATGAACTTCGAAGATGGAGGCGTCGTCACAGTTACCCAGGATTCCTCGCTACAGGACGGTGAGTTTATCTATAAAGTAAAACTTCGAGGTACTAACTTCCCGAGTGATGGCCCGGTTATGCAAAAGAAGACCATGGGGTGGGAAGCCTCTTCAGAAAGGATGTACCCAGAGGATGGTGCTTTGAAGGGGGAAATAAAGCAGCGATTAAAACTAAAAGACGGCGGCCATTATGATGCCGAAGTAAAGACCACGTACAAGGCCAAGAAGCCCGTACAGTTACCTGGTGCGTACAACGTCAATATTAAGCTCGACATAACATCCCATAATGAGGACTACACCATCGTGGAGCAATACGAGCGTGCTGAGGGGCGGCATTCGACTGGAGGGATGGACGAACTATACAAATAG\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CTA due to very low average frequency 0.03649285148597391 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "\n",
      "##########################\n",
      "# EVALUATION #\n",
      "##########################\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCCCTGCCCTTCGCCTGGGACATCCTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTTACCGTGACCCAGGACTCGTCCCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGGATGGACGAGCTGTACAAGTGA\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCCCTGCCCTTCGCCTGGGACATCCTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTTACCGTGACCCAGGACTCGTCCCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGGATGGACGAGCTGTACAAGTGA\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCCCTGCCCTTCGCCTGGGACATCCTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTTACCGTGACCCAGGACTCGTCCCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGGATGGACGAGCTGTACAAGTGA\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCCCTGCCCTTCGCCTGGGACATCCTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTTACCGTGACCCAGGACTCGTCCCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGGATGGACGAGCTGTACAAGTGA\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCCCTGCCCTTCGCCTGGGACATCCTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTTACCGTGACCCAGGACTCGTCCCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGGATGGACGAGCTGTACAAGTGA', 'average_distance_score': 2.369961316026482, 'weakest_link_score': 2.369961316026482, 'ratio_score': 1.1180804115914358}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCCCTGCCCTTCGCCTGGGACATCCTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTTACCGTGACCCAGGACTCGTCCCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGGATGGACGAGCTGTACAAGTGA', 'average_distance_score': 2.369961316026482, 'weakest_link_score': 2.369961316026482, 'ratio_score': 1.1180804115914358}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCCCTGCCCTTCGCCTGGGACATCCTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTTACCGTGACCCAGGACTCGTCCCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGGATGGACGAGCTGTACAAGTGA', 'average_distance_score': 2.369961316026482, 'weakest_link_score': 2.369961316026482, 'ratio_score': 1.1180804115914358}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCCCTGCCCTTCGCCTGGGACATCCTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTTACCGTGACCCAGGACTCGTCCCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGGATGGACGAGCTGTACAAGTGA', 'average_distance_score': 2.369961316026482, 'weakest_link_score': 2.369961316026482, 'ratio_score': 1.1180804115914358}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCCCTGCCCTTCGCCTGGGACATCCTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTTACCGTGACCCAGGACTCGTCCCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGGATGGACGAGCTGTACAAGTGA', 'average_distance_score': 2.369961316026482, 'weakest_link_score': 2.369961316026482, 'ratio_score': 1.1180804115914358}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry-wanted-constraint-init-external-0.1\\CAI_zscore_bulk_aa_diff_ecoli_opt_True_YgOq\n",
      "Final output: {'module_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'orf_optimization_method': 'zscore_bulk_aa_diff', 'orf_optimization_cub_index': 'CAI', 'initiation_optimization_method': 'external', 'evaluation_score': 'average_distance', 'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.037037037037037035, 'TGA': 0.24183006535947713, 'GCA': 0.6303663185811371, 'GCC': 0.653737775567711, 'GCG': 1.0, 'GCT': 0.6063318415382065, 'TGC': 1.0, 'TGT': 0.7181136120042874, 'GAC': 0.848284271032921, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.36125998031280765, 'TTC': 1.0, 'TTT': 0.6556786703601108, 'GGA': 0.11638214051214708, 'GGC': 1.0, 'GGG': 0.18844386080105055, 'GGT': 0.9788246881155614, 'CAC': 1.0, 'CAT': 0.7088785046728973, 'ATA': 0.029297564419343447, 'ATC': 1.0, 'ATT': 0.7354394634662901, 'AAA': 1.0, 'AAG': 0.26291560102301786, 'CTA': 0.02288199697428139, 'CTC': 0.1265128593040847, 'CTG': 1.0, 'CTT': 0.10712934947049924, 'TTA': 0.09427004538577911, 'TTG': 0.11753025718608169, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.39072452229299365, 'CCA': 0.2588842975206611, 'CCC': 0.08181818181818182, 'CCG': 1.0, 'CCT': 0.1818181818181818, 'CAA': 0.32759266192437286, 'CAG': 1.0, 'AGA': 0.015347817420769383, 'AGG': 0.00976679290412597, 'CGA': 0.033087502491528795, 'CGC': 0.7123779150886984, 'CGG': 0.05521227825393661, 'CGT': 1.0, 'AGC': 1.0, 'AGT': 0.2936010037641154, 'TCA': 0.28523630280217477, 'TCC': 0.7921371810957758, 'TCG': 0.4086156419907988, 'TCT': 0.8883312421580928, 'ACA': 0.12800331400165701, 'ACC': 1.0, 'ACG': 0.34299917149958575, 'ACT': 0.4233637116818558, 'GTA': 0.5115587090867476, 'GTC': 0.49691004806591893, 'GTG': 1.0, 'GTT': 0.9814602883955139, 'TGG': 1.0, 'TAC': 1.0, 'TAT': 0.8523076923076923}, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'AAA': 0.7667390608953314, 'CGC': 0.40076028262096014, 'ATT': 0.5085852538544149, 'AGC': 0.27799163607995603, 'ACC': 0.4359429381095841, 'ATC': 0.4205725148436632, 'ACA': 0.13023262212317555, 'GGT': 0.33726003930064996, 'AAC': 0.5512262248110238, 'GCG': 0.3563679190097728, 'GGC': 0.4050486219579785, 'TGA': 0.28926761860893596, 'CGA': 0.06388523224887838, 'GTG': 0.3715940695510341, 'TTG': 0.12808970030428077, 'AAG': 0.23326093910466858, 'TTC': 0.4262385670731707, 'TCA': 0.12251736088908215, 'GCA': 0.21281773299119433, 'AAT': 0.4487737751889762, 'GAA': 0.689467997530356, 'CGT': 0.38120248233619, 'TTT': 0.5737614329268292, 'CTG': 0.4975546385450099, 'GTT': 0.2584847661069405, 'GCC': 0.26990242799092135, 'GAT': 0.626856640686903, 'AGG': 0.020162265266098916, 'CAG': 0.6530527088827319, 'GGG': 0.15057187484254547, 'GTC': 0.21619163129169194, 'CTC': 0.10435163186194824, 'TCT': 0.14544780223293646, 'CCC': 0.12361073915372255, 'CAC': 0.4292446655321377, 'GCT': 0.1609119200081115, 'TTA': 0.13009739763522432, 'CTT': 0.10341378016756284, 'ACG': 0.26849903016796667, 'GGA': 0.10711946389882603, 'CCG': 0.527785216925549, 'CAA': 0.3469472911172681, 'ACT': 0.16532540959927366, 'ATA': 0.07084223130192197, 'CAT': 0.5707553344678623, 'AGT': 0.15102374892893225, 'TGC': 0.55651396076928, 'GAG': 0.31053200246964396, 'TCG': 0.15406749964830613, 'GTA': 0.15372953305033354, 'TAC': 0.4315172054223149, 'TCC': 0.14895195222078703, 'GAC': 0.37314335931309695, 'TGT': 0.4434860392307201, 'TGG': 1.0, 'TAT': 0.5684827945776851, 'CCT': 0.15789033208355652, 'CCA': 0.19071371183717192, 'CGG': 0.09749348521075679, 'CTA': 0.03649285148597391, 'TAA': 0.6391064025794565, 'AGA': 0.036496252317115764, 'TAG': 0.07162597881160755}), 'cai_avg': 0.6151304741138522, 'cai_std': 0.09377868043841883, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.10739856801909307, 'TGA': 0.1694510739856802, 'GCA': 1.0, 'GCC': 0.5406038521603331, 'GCG': 0.950546590317543, 'GCT': 0.9508068714211347, 'TGC': 1.0, 'TGT': 0.7315315315315315, 'GAC': 0.6775877587758776, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.3929343907714492, 'TTC': 0.864321608040201, 'TTT': 1.0, 'GGA': 0.8215814372606443, 'GGC': 1.0, 'GGG': 0.28835323271006985, 'GGT': 0.6037395809867087, 'CAC': 0.6798462383305877, 'CAT': 1.0, 'ATA': 0.10219657483246464, 'ATC': 0.8559195830230827, 'ATT': 1.0, 'AAA': 1.0, 'AAG': 0.3472094095940959, 'CTA': 0.13892709766162312, 'CTC': 0.30375974323704724, 'CTG': 0.6657496561210454, 'CTT': 1.0, 'TTA': 0.5476845483723063, 'TTG': 0.3874369555249885, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.8191662154845695, 'CCA': 0.4708680142687277, 'CCC': 0.06896551724137931, 'CCG': 1.0, 'CCT': 0.7439556084026953, 'CAA': 1.0, 'CAG': 0.8559696778269108, 'AGA': 0.7029352226720648, 'AGG': 0.09210526315789473, 'CGA': 0.18927125506072875, 'CGC': 1.0, 'CGG': 0.27834008097165996, 'CGT': 0.9903846153846154, 'AGC': 0.8852526926263463, 'AGT': 0.3177299088649544, 'TCA': 0.9254349627174814, 'TCC': 0.4258492129246064, 'TCG': 0.2531068765534383, 'TCT': 1.0, 'ACA': 1.0, 'ACC': 0.22235752557233318, 'ACG': 0.5411592791037506, 'ACT': 0.43180711154408186, 'GTA': 0.7223221166195737, 'GTC': 0.6935525301823786, 'GTG': 0.6699203698946828, 'GTT': 1.0, 'TGG': 1.0, 'TAC': 0.7833094555873925, 'TAT': 1.0}, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'GAA': 0.6798186496752849, 'AAT': 0.5655317497600131, 'ATA': 0.1287010322849589, 'TTA': 0.198530453110291, 'GAC': 0.35960560391899904, 'CTG': 0.23973441522197322, 'TGG': 1.0, 'AAC': 0.4344682502399869, 'CAA': 0.5131548459124133, 'GCC': 0.20715414603308396, 'CTT': 0.2391679232236727, 'GCT': 0.24752923822568748, 'ATC': 0.366762837741479, 'AAA': 0.7016557540022625, 'AAG': 0.29834424599773746, 'TTG': 0.1592175745811708, 'AGC': 0.22534685577492342, 'CCG': 0.4334091260292722, 'AGT': 0.10611598754150686, 'TTT': 0.6847751982489505, 'GAG': 0.3201813503247151, 'ACT': 0.1614000537105004, 'TCA': 0.23653118484388272, 'ACC': 0.1588786441082565, 'CAC': 0.3265866087018494, 'GGC': 0.3390368073029434, 'GAT': 0.640394396081001, 'ACA': 0.4122430101751559, 'ACG': 0.2674782920060872, 'CCC': 0.0889644363010221, 'AGA': 0.26380392311639106, 'TCC': 0.12712038919920718, 'TAC': 0.34586274736720285, 'CAT': 0.6734133912981506, 'ATT': 0.5045361299735621, 'GCA': 0.2828363712991255, 'TAT': 0.6541372526327971, 'GGG': 0.16211598104043537, 'GTC': 0.25598186049286953, 'CCT': 0.28667299498885185, 'GTT': 0.28439644370189154, 'GCG': 0.26248024444210305, 'TCT': 0.2048572678833432, 'CTC': 0.11219040795754642, 'CCA': 0.19095344268085387, 'GGA': 0.3148341038094681, 'CGA': 0.09855152543710778, 'GTA': 0.19774449549495793, 'GTG': 0.26187720031028106, 'TTC': 0.31522480175104955, 'CGC': 0.2064569601768165, 'CAG': 0.4868451540875866, 'CGG': 0.1555630106168844, 'CGT': 0.18161187196589967, 'CTA': 0.05115922590534585, 'AGG': 0.09401270868690058, 'GGT': 0.18401310784715313, 'TAG': 0.14132712456344587, 'TGC': 0.5460141582025239, 'TCG': 0.10002831475713661, 'TAA': 0.630500582072177, 'TGA': 0.2281722933643772, 'TGT': 0.45398584179747614}), 'cai_avg': 0.7334925635409068, 'cai_std': 0.042802002292947894, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}]}, 'orf': [{'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCCCTGCCCTTCGCCTGGGACATCCTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTTACCGTGACCCAGGACTCGTCCCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGGATGGACGAGCTGTACAAGTGA', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'initial_sequence_optimization_score': 0.5193309344738618, 'final_sequence_optimization_score': 2.8892922505003438, 'run_time': 19.25, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 2.88856858683797}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 2.8892922505003438}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 2.8892922505003438}]}, {'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGGTTTAAAGTCCATATGGAGGGTTCTGTAAATGGGCATGAATTCGAGATAGAAGGCGAGGGCGAGGGCCGCCCTTACGAAGGCACACAAACTGCTAAACTCAAAGTCACCAAAGGGGGGCCACTCCCTTTCGCATGGGATATATTATCCCCGCAGTTTATGTATGGTAGTAAAGCCTACGTGAAGCACCCAGCGGATATCCCAGACTACCTAAAATTGAGTTTCCCTGAAGGTTTCAAATGGGAGCGTGTTATGAATTTCGAGGACGGTGGTGTCGTAACAGTAACGCAAGATTCAAGTTTACAAGATGGCGAATTCATTTATAAAGTTAAACTAAGAGGAACAAACTTTCCCTCTGACGGACCGGTTATGCAGAAGAAGACCATGGGTTGGGAGGCAAGTTCCGAGCGTATGTACCCTGAAGACGGCGCACTAAAGGGCGAAATCAAGCAGCGGTTGAAGCTCAAAGACGGGGGCCACTACGATGCGGAGGTCAAAACTACGTATAAGGCAAAGAAACCTGTGCAATTACCTGGTGCCTACAATGTCAATATTAAGCTTGATATCACGAGTCATAACGAGGACTATACCATCGTCGAACAGTATGAACGTGCAGAGGGGCGACATTCTACTGGGGGGATGGACGAGTTATACAAATGA', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCCCTGCCCTTCGCCTGGGACATCCTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTTACCGTGACCCAGGACTCGTCCCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGGATGGACGAGCTGTACAAGTGA', 'iterations_count': 4, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'initial_sequence_optimization_score': 0.3665427929108921, 'final_sequence_optimization_score': 2.8892922505003438, 'run_time': 27.781000000075437, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATC', 'L': 'CTC', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 2.721340513422541}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 2.88856858683797}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 2.8892922505003438}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 2.8892922505003438}]}, {'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGATTCAAGGTCCACATGGAGGGGTCGGTGAATGGACATGAATTCGAGATAGAAGGCGAAGGAGAAGGTCGGCCCTACGAGGGCACTCAGACAGCAAAATTAAAGGTCACAAAGGGGGGCCCTTTACCGTTTGCGTGGGATATTCTGTCCCCGCAATTCATGTACGGTAGCAAAGCGTATGTTAAGCATCCTGCGGACATCCCAGACTACCTAAAGCTGTCATTTCCAGAGGGCTTTAAATGGGAACGAGTTATGAATTTCGAGGACGGGGGCGTTGTGACAGTGACGCAAGACAGCAGTCTTCAAGATGGGGAATTTATTTATAAAGTTAAACTCAGAGGCACTAACTTTCCCTCGGACGGGCCAGTTATGCAGAAAAAGACAATGGGTTGGGAGGCCAGCTCCGAGCGCATGTATCCTGAGGACGGTGCACTTAAAGGAGAAATCAAACAAAGGCTCAAGCTTAAGGACGGTGGGCATTACGATGCAGAGGTCAAAACAACGTATAAAGCAAAAAAGCCTGTCCAGCTTCCAGGAGCATACAACGTGAATATAAAATTAGATATTACATCTCATAATGAGGACTATACTATCGTGGAGCAATATGAAAGAGCTGAAGGAAGGCATAGTACTGGCGGTATGGATGAGTTATATAAATGA', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCCCTGCCCTTCGCCTGGGACATCCTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTTACCGTGACCCAGGACTCGTCCCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGGATGGACGAGCTGTACAAGTGA', 'iterations_count': 4, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'initial_sequence_optimization_score': -0.007499705337304929, 'final_sequence_optimization_score': 2.8892922505003438, 'run_time': 26.67200000002049, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATC', 'L': 'CTC', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 2.721340513422541}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 2.88856858683797}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 2.8892922505003438}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 2.8892922505003438}]}, {'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAGGTGCATATGGAGGGATCCGTGAACGGACACGAGTTTGAAATCGAGGGGGAAGGCGAAGGACGACCATACGAAGGTACGCAGACAGCGAAATTGAAGGTGACGAAGGGAGGACCCCTACCCTTTGCTTGGGACATCTTATCGCCGCAATTTATGTATGGATCGAAGGCTTATGTAAAGCATCCAGCTGACATCCCGGACTACCTGAAACTCTCTTTTCCAGAGGGATTTAAGTGGGAAAGGGTTATGAACTTCGAAGATGGAGGCGTCGTCACAGTTACCCAGGATTCCTCGCTACAGGACGGTGAGTTTATCTATAAAGTAAAACTTCGAGGTACTAACTTCCCGAGTGATGGCCCGGTTATGCAAAAGAAGACCATGGGGTGGGAAGCCTCTTCAGAAAGGATGTACCCAGAGGATGGTGCTTTGAAGGGGGAAATAAAGCAGCGATTAAAACTAAAAGACGGCGGCCATTATGATGCCGAAGTAAAGACCACGTACAAGGCCAAGAAGCCCGTACAGTTACCTGGTGCGTACAACGTCAATATTAAGCTCGACATAACATCCCATAATGAGGACTACACCATCGTGGAGCAATACGAGCGTGCTGAGGGGCGGCATTCGACTGGAGGGATGGACGAACTATACAAATAG', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCCCTGCCCTTCGCCTGGGACATCCTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTTACCGTGACCCAGGACTCGTCCCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGGATGGACGAGCTGTACAAGTGA', 'iterations_count': 4, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'initial_sequence_optimization_score': 0.6833833774406468, 'final_sequence_optimization_score': 2.8892922505003438, 'run_time': 27.515999999945052, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATC', 'L': 'CTC', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 2.721340513422541}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 2.88856858683797}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 2.8892922505003438}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 2.8892922505003438}]}], 'evaluation': [{'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'cai_initial_score': 0.7915994634162647, 'cai_final_score': 0.6153593309815049, 'dist_score': -1.879319815664186}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'cai_initial_score': 0.7695788606254016, 'cai_final_score': 0.48626203020063413, 'dist_score': -6.619242447717149}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCCCTGCCCTTCGCCTGGGACATCCTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTTACCGTGACCCAGGACTCGTCCCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGGATGGACGAGCTGTACAAGTGA', 'average_distance_score': 2.369961316026482, 'weakest_link_score': 2.369961316026482, 'ratio_score': 1.1180804115914358, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCCCTGCCCTTCGCCTGGGACATCCTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTTACCGTGACCCAGGACTCGTCCCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGGATGGACGAGCTGTACAAGTGA'}, {'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'cai_initial_score': 0.7915994634162647, 'cai_final_score': 0.6153593309815049, 'dist_score': -1.879319815664186}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'cai_initial_score': 0.7695788606254016, 'cai_final_score': 0.48626203020063413, 'dist_score': -6.619242447717149}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCCCTGCCCTTCGCCTGGGACATCCTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTTACCGTGACCCAGGACTCGTCCCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGGATGGACGAGCTGTACAAGTGA', 'average_distance_score': 2.369961316026482, 'weakest_link_score': 2.369961316026482, 'ratio_score': 1.1180804115914358, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCCCTGCCCTTCGCCTGGGACATCCTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTTACCGTGACCCAGGACTCGTCCCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGGATGGACGAGCTGTACAAGTGA'}, {'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'cai_initial_score': 0.7915994634162647, 'cai_final_score': 0.6153593309815049, 'dist_score': -1.879319815664186}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'cai_initial_score': 0.7695788606254016, 'cai_final_score': 0.48626203020063413, 'dist_score': -6.619242447717149}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCCCTGCCCTTCGCCTGGGACATCCTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTTACCGTGACCCAGGACTCGTCCCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGGATGGACGAGCTGTACAAGTGA', 'average_distance_score': 2.369961316026482, 'weakest_link_score': 2.369961316026482, 'ratio_score': 1.1180804115914358, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCCCTGCCCTTCGCCTGGGACATCCTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTTACCGTGACCCAGGACTCGTCCCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGGATGGACGAGCTGTACAAGTGA'}, {'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'cai_initial_score': 0.7915994634162647, 'cai_final_score': 0.6153593309815049, 'dist_score': -1.879319815664186}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'cai_initial_score': 0.7695788606254016, 'cai_final_score': 0.48626203020063413, 'dist_score': -6.619242447717149}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCCCTGCCCTTCGCCTGGGACATCCTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTTACCGTGACCCAGGACTCGTCCCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGGATGGACGAGCTGTACAAGTGA', 'average_distance_score': 2.369961316026482, 'weakest_link_score': 2.369961316026482, 'ratio_score': 1.1180804115914358, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCCCTGCCCTTCGCCTGGGACATCCTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTTACCGTGACCCAGGACTCGTCCCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGGATGGACGAGCTGTACAAGTGA'}], 'final_evaluation': {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGGCCCCTGCCCTTCGCCTGGGACATCCTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGGGTGGTTACCGTGACCCAGGACTCGTCCCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAAACCATGGGTTGGGAGGCCTCGTCCGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGTCTGAAGCTGAAGGACGGTGGGCACTACGACGCCGAGGTGAAGACCACTTACAAGGCCAAGAAACCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCGCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGGATGGACGAGCTGTACAAGTGA', 'average_distance_score': 2.369961316026482, 'weakest_link_score': 2.369961316026482, 'ratio_score': 1.1180804115914358}, 'zip_output_file_path': 'results\\\\mcherry-wanted-constraint-init-external-0.1\\\\CAI_zscore_bulk_aa_diff_ecoli_opt_True_YgOq\\\\communique_results.zip'}\n",
      "\n",
      "##########################\n",
      "# User Input #\n",
      "##########################\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is:\n",
      "ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "------------------------------------------\n",
      "Parsing information for Bacillus subtilis:\n",
      "------------------------------------------\n",
      "Organism is defined as wanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 98 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3544\n",
      "Calculate CAI weights from a reference set of 532 highly expressed genes from estimated expression dictionary.\n",
      "cai_std=0.042802002292947894, cai_avg=0.7334925635409068\n",
      "------------------------------------------\n",
      "Parsing information for Escherichia coli:\n",
      "------------------------------------------\n",
      "Organism is defined as unwanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 90 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3893\n",
      "Calculate CAI weights from a reference set of 584 highly expressed genes from estimated expression dictionary.\n",
      "cai_std=0.09377868043841883, cai_avg=0.6151304741138522\n",
      "-----------------------------\n",
      "Normalized Prioritization Weights\n",
      "-----------------------------\n",
      "Bacillus subtilis : 1.0\n",
      "Escherichia coli : 1.0\n",
      "Total input processing time: 27.772412300109863\n",
      "\n",
      "##########################\n",
      "# Initiation #\n",
      "##########################\n",
      "Running initiation optimization InitiationOptimizationMethod.external_module on 15 codons at the start of the ORF\n",
      "Taking optimized sequence from configuration value config.INITIATION.EXTERNAL_INITIATION_ORF\n",
      "Optimized sequence is: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "Optimization method used is: ORFOptimizationMethod.zscore_bulk_aa_diff\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGATTCAAAGTGCACATGGAGGGATCAGTGAACGGCCATGAGTTCGAAATTGAGGGGGAGGGGGAAGGCAGGCCATATGAGGGAACCCAGACCGCAAAACTTAAGGTCACTAAAGGTGGCCCATTACCCTTCGCATGGGATATACTTTCTCCTCAATTCATGTATGGATCCAAGGCATACGTGAAGCACCCAGCAGACATTCCTGATTATCTCAAACTAAGTTTTCCTGAGGGTTTCAAGTGGGAGAGGGTAATGAATTTTGAGGATGGGGGTGTAGTAACCGTGACCCAGGACTCAAGCCTGCAGGATGGTGAGTTTATATATAAAGTAAAGCTGCGCGGAACCAATTTCCCGTCCGATGGCCCTGTGATGCAAAAAAAAACTATGGGGTGGGAAGCGTCCTCAGAACGCATGTACCCGGAGGATGGAGCACTGAAGGGGGAAATAAAGCAGCGTCTAAAGCTCAAGGACGGAGGTCACTATGATGCCGAAGTTAAGACCACTTATAAGGCTAAGAAGCCGGTGCAGCTGCCAGGGGCCTACAACGTGAATATTAAACTGGATATAACCTCTCATAACGAGGACTATACGATCGTAGAACAATACGAGCGAGCGGAGGGACGCCATTCAACAGGGGGAATGGACGAATTATATAAATAG\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGGTTTAAAGTACATATGGAAGGGTCAGTGAATGGTCACGAATTCGAAATCGAAGGCGAGGGGGAGGGTAGACCATACGAGGGTACCCAGACGGCTAAGCTCAAAGTAACAAAGGGGGGGCCACTACCCTTCGCGTGGGACATTTTGTCGCCACAGTTCATGTACGGGAGCAAAGCCTATGTTAAACACCCAGCAGACATTCCCGATTATCTGAAGCTGTCTTTTCCGGAAGGATTCAAATGGGAACGGGTTATGAACTTTGAAGATGGCGGCGTGGTTACTGTTACGCAAGATTCCTCCCTGCAGGATGGCGAGTTCATCTATAAGGTGAAACTCAGAGGAACGAACTTCCCAAGTGATGGACCCGTGATGCAAAAGAAGACGATGGGGTGGGAGGCTTCTAGTGAAAGAATGTATCCAGAGGACGGAGCGCTCAAGGGTGAGATTAAGCAGCGGTTAAAGCTCAAGGATGGAGGGCATTACGACGCTGAAGTGAAAACAACATATAAGGCCAAAAAACCGGTCCAATTACCCGGAGCTTACAATGTTAATATTAAATTGGACATCACGTCACATAATGAGGATTATACGATAGTGGAACAATACGAACGTGCTGAAGGCCGACATTCGACCGGCGGGATGGATGAGCTTTACAAATGA\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGAGGTTCAAGGTTCACATGGAAGGTTCTGTAAATGGGCATGAATTTGAAATCGAGGGCGAAGGTGAGGGACGGCCGTACGAAGGGACGCAGACAGCAAAGTTGAAGGTCACCAAAGGTGGACCACTTCCATTTGCATGGGACATCCTTTCGCCCCAATTTATGTATGGTAGCAAAGCCTACGTAAAACACCCGGCGGACATACCAGATTATCTAAAGCTTAGCTTTCCCGAAGGATTTAAGTGGGAGAGGGTGATGAATTTCGAGGATGGGGGTGTTGTCACGGTGACTCAGGACTCGTCATTGCAAGATGGGGAATTCATATATAAGGTAAAACTAAGAGGAACGAATTTCCCCAGCGATGGACCCGTTATGCAGAAGAAAACGATGGGGTGGGAGGCATCTAGCGAGAGAATGTATCCAGAGGATGGTGCCCTAAAAGGGGAGATTAAGCAAAGACTCAAACTCAAAGACGGCGGCCACTACGATGCAGAGGTGAAGACCACCTACAAGGCTAAAAAACCGGTTCAACTACCCGGAGCATACAACGTGAATATAAAACTTGACATAACGTCTCATAATGAAGACTACACCATTGTTGAACAGTACGAAAGGGCAGAGGGCCGGCATAGTACCGGTGGGATGGATGAGCTATACAAATGA\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "\n",
      "##########################\n",
      "# EVALUATION #\n",
      "##########################\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA', 'average_distance_score': 3.9819077948851493, 'weakest_link_score': 3.9819077948851493, 'ratio_score': 1.1707526264894996}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA', 'average_distance_score': 3.9819077948851493, 'weakest_link_score': 3.9819077948851493, 'ratio_score': 1.1707526264894996}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA', 'average_distance_score': 3.9819077948851493, 'weakest_link_score': 3.9819077948851493, 'ratio_score': 1.1707526264894996}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA', 'average_distance_score': 3.9819077948851493, 'weakest_link_score': 3.9819077948851493, 'ratio_score': 1.1707526264894996}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA', 'average_distance_score': 3.9819077948851493, 'weakest_link_score': 3.9819077948851493, 'ratio_score': 1.1707526264894996}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry-wanted-constraint-init-external-0.1\\CAI_zscore_bulk_aa_diff_ecoli_opt_False_WFw0\n",
      "Final output: {'module_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'orf_optimization_method': 'zscore_bulk_aa_diff', 'orf_optimization_cub_index': 'CAI', 'initiation_optimization_method': 'external', 'evaluation_score': 'average_distance', 'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.10739856801909307, 'TGA': 0.1694510739856802, 'GCA': 1.0, 'GCC': 0.5406038521603331, 'GCG': 0.950546590317543, 'GCT': 0.9508068714211347, 'TGC': 1.0, 'TGT': 0.7315315315315315, 'GAC': 0.6775877587758776, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.3929343907714492, 'TTC': 0.864321608040201, 'TTT': 1.0, 'GGA': 0.8215814372606443, 'GGC': 1.0, 'GGG': 0.28835323271006985, 'GGT': 0.6037395809867087, 'CAC': 0.6798462383305877, 'CAT': 1.0, 'ATA': 0.10219657483246464, 'ATC': 0.8559195830230827, 'ATT': 1.0, 'AAA': 1.0, 'AAG': 0.3472094095940959, 'CTA': 0.13892709766162312, 'CTC': 0.30375974323704724, 'CTG': 0.6657496561210454, 'CTT': 1.0, 'TTA': 0.5476845483723063, 'TTG': 0.3874369555249885, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.8191662154845695, 'CCA': 0.4708680142687277, 'CCC': 0.06896551724137931, 'CCG': 1.0, 'CCT': 0.7439556084026953, 'CAA': 1.0, 'CAG': 0.8559696778269108, 'AGA': 0.7029352226720648, 'AGG': 0.09210526315789473, 'CGA': 0.18927125506072875, 'CGC': 1.0, 'CGG': 0.27834008097165996, 'CGT': 0.9903846153846154, 'AGC': 0.8852526926263463, 'AGT': 0.3177299088649544, 'TCA': 0.9254349627174814, 'TCC': 0.4258492129246064, 'TCG': 0.2531068765534383, 'TCT': 1.0, 'ACA': 1.0, 'ACC': 0.22235752557233318, 'ACG': 0.5411592791037506, 'ACT': 0.43180711154408186, 'GTA': 0.7223221166195737, 'GTC': 0.6935525301823786, 'GTG': 0.6699203698946828, 'GTT': 1.0, 'TGG': 1.0, 'TAC': 0.7833094555873925, 'TAT': 1.0}, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'GAA': 0.6798186496752849, 'AAT': 0.5655317497600131, 'ATA': 0.1287010322849589, 'TTA': 0.198530453110291, 'GAC': 0.35960560391899904, 'CTG': 0.23973441522197322, 'TGG': 1.0, 'AAC': 0.4344682502399869, 'CAA': 0.5131548459124133, 'GCC': 0.20715414603308396, 'CTT': 0.2391679232236727, 'GCT': 0.24752923822568748, 'ATC': 0.366762837741479, 'AAA': 0.7016557540022625, 'AAG': 0.29834424599773746, 'TTG': 0.1592175745811708, 'AGC': 0.22534685577492342, 'CCG': 0.4334091260292722, 'AGT': 0.10611598754150686, 'TTT': 0.6847751982489505, 'GAG': 0.3201813503247151, 'ACT': 0.1614000537105004, 'TCA': 0.23653118484388272, 'ACC': 0.1588786441082565, 'CAC': 0.3265866087018494, 'GGC': 0.3390368073029434, 'GAT': 0.640394396081001, 'ACA': 0.4122430101751559, 'ACG': 0.2674782920060872, 'CCC': 0.0889644363010221, 'AGA': 0.26380392311639106, 'TCC': 0.12712038919920718, 'TAC': 0.34586274736720285, 'CAT': 0.6734133912981506, 'ATT': 0.5045361299735621, 'GCA': 0.2828363712991255, 'TAT': 0.6541372526327971, 'GGG': 0.16211598104043537, 'GTC': 0.25598186049286953, 'CCT': 0.28667299498885185, 'GTT': 0.28439644370189154, 'GCG': 0.26248024444210305, 'TCT': 0.2048572678833432, 'CTC': 0.11219040795754642, 'CCA': 0.19095344268085387, 'GGA': 0.3148341038094681, 'CGA': 0.09855152543710778, 'GTA': 0.19774449549495793, 'GTG': 0.26187720031028106, 'TTC': 0.31522480175104955, 'CGC': 0.2064569601768165, 'CAG': 0.4868451540875866, 'CGG': 0.1555630106168844, 'CGT': 0.18161187196589967, 'CTA': 0.05115922590534585, 'AGG': 0.09401270868690058, 'GGT': 0.18401310784715313, 'TAG': 0.14132712456344587, 'TGC': 0.5460141582025239, 'TCG': 0.10002831475713661, 'TAA': 0.630500582072177, 'TGA': 0.2281722933643772, 'TGT': 0.45398584179747614}), 'cai_avg': 0.7334925635409068, 'cai_std': 0.042802002292947894, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}, {'name': 'Escherichia coli', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.037037037037037035, 'TGA': 0.24183006535947713, 'GCA': 0.6303663185811371, 'GCC': 0.653737775567711, 'GCG': 1.0, 'GCT': 0.6063318415382065, 'TGC': 1.0, 'TGT': 0.7181136120042874, 'GAC': 0.848284271032921, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.36125998031280765, 'TTC': 1.0, 'TTT': 0.6556786703601108, 'GGA': 0.11638214051214708, 'GGC': 1.0, 'GGG': 0.18844386080105055, 'GGT': 0.9788246881155614, 'CAC': 1.0, 'CAT': 0.7088785046728973, 'ATA': 0.029297564419343447, 'ATC': 1.0, 'ATT': 0.7354394634662901, 'AAA': 1.0, 'AAG': 0.26291560102301786, 'CTA': 0.02288199697428139, 'CTC': 0.1265128593040847, 'CTG': 1.0, 'CTT': 0.10712934947049924, 'TTA': 0.09427004538577911, 'TTG': 0.11753025718608169, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.39072452229299365, 'CCA': 0.2588842975206611, 'CCC': 0.08181818181818182, 'CCG': 1.0, 'CCT': 0.1818181818181818, 'CAA': 0.32759266192437286, 'CAG': 1.0, 'AGA': 0.015347817420769383, 'AGG': 0.00976679290412597, 'CGA': 0.033087502491528795, 'CGC': 0.7123779150886984, 'CGG': 0.05521227825393661, 'CGT': 1.0, 'AGC': 1.0, 'AGT': 0.2936010037641154, 'TCA': 0.28523630280217477, 'TCC': 0.7921371810957758, 'TCG': 0.4086156419907988, 'TCT': 0.8883312421580928, 'ACA': 0.12800331400165701, 'ACC': 1.0, 'ACG': 0.34299917149958575, 'ACT': 0.4233637116818558, 'GTA': 0.5115587090867476, 'GTC': 0.49691004806591893, 'GTG': 1.0, 'GTT': 0.9814602883955139, 'TGG': 1.0, 'TAC': 1.0, 'TAT': 0.8523076923076923}, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'AAA': 0.7667390608953314, 'CGC': 0.40076028262096014, 'ATT': 0.5085852538544149, 'AGC': 0.27799163607995603, 'ACC': 0.4359429381095841, 'ATC': 0.4205725148436632, 'ACA': 0.13023262212317555, 'GGT': 0.33726003930064996, 'AAC': 0.5512262248110238, 'GCG': 0.3563679190097728, 'GGC': 0.4050486219579785, 'TGA': 0.28926761860893596, 'CGA': 0.06388523224887838, 'GTG': 0.3715940695510341, 'TTG': 0.12808970030428077, 'AAG': 0.23326093910466858, 'TTC': 0.4262385670731707, 'TCA': 0.12251736088908215, 'GCA': 0.21281773299119433, 'AAT': 0.4487737751889762, 'GAA': 0.689467997530356, 'CGT': 0.38120248233619, 'TTT': 0.5737614329268292, 'CTG': 0.4975546385450099, 'GTT': 0.2584847661069405, 'GCC': 0.26990242799092135, 'GAT': 0.626856640686903, 'AGG': 0.020162265266098916, 'CAG': 0.6530527088827319, 'GGG': 0.15057187484254547, 'GTC': 0.21619163129169194, 'CTC': 0.10435163186194824, 'TCT': 0.14544780223293646, 'CCC': 0.12361073915372255, 'CAC': 0.4292446655321377, 'GCT': 0.1609119200081115, 'TTA': 0.13009739763522432, 'CTT': 0.10341378016756284, 'ACG': 0.26849903016796667, 'GGA': 0.10711946389882603, 'CCG': 0.527785216925549, 'CAA': 0.3469472911172681, 'ACT': 0.16532540959927366, 'ATA': 0.07084223130192197, 'CAT': 0.5707553344678623, 'AGT': 0.15102374892893225, 'TGC': 0.55651396076928, 'GAG': 0.31053200246964396, 'TCG': 0.15406749964830613, 'GTA': 0.15372953305033354, 'TAC': 0.4315172054223149, 'TCC': 0.14895195222078703, 'GAC': 0.37314335931309695, 'TGT': 0.4434860392307201, 'TGG': 1.0, 'TAT': 0.5684827945776851, 'CCT': 0.15789033208355652, 'CCA': 0.19071371183717192, 'CGG': 0.09749348521075679, 'CTA': 0.03649285148597391, 'TAA': 0.6391064025794565, 'AGA': 0.036496252317115764, 'TAG': 0.07162597881160755}), 'cai_avg': 0.6151304741138522, 'cai_std': 0.09377868043841883, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}]}, 'orf': [{'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'initial_sequence_optimization_score': -0.5193309344738618, 'final_sequence_optimization_score': 3.4625768604112874, 'run_time': 20.43700000003446, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 3.4581864774870925}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 3.4625768604112874}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 3.4625768604112874}]}, {'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGATTCAAAGTGCACATGGAGGGATCAGTGAACGGCCATGAGTTCGAAATTGAGGGGGAGGGGGAAGGCAGGCCATATGAGGGAACCCAGACCGCAAAACTTAAGGTCACTAAAGGTGGCCCATTACCCTTCGCATGGGATATACTTTCTCCTCAATTCATGTATGGATCCAAGGCATACGTGAAGCACCCAGCAGACATTCCTGATTATCTCAAACTAAGTTTTCCTGAGGGTTTCAAGTGGGAGAGGGTAATGAATTTTGAGGATGGGGGTGTAGTAACCGTGACCCAGGACTCAAGCCTGCAGGATGGTGAGTTTATATATAAAGTAAAGCTGCGCGGAACCAATTTCCCGTCCGATGGCCCTGTGATGCAAAAAAAAACTATGGGGTGGGAAGCGTCCTCAGAACGCATGTACCCGGAGGATGGAGCACTGAAGGGGGAAATAAAGCAGCGTCTAAAGCTCAAGGACGGAGGTCACTATGATGCCGAAGTTAAGACCACTTATAAGGCTAAGAAGCCGGTGCAGCTGCCAGGGGCCTACAACGTGAATATTAAACTGGATATAACCTCTCATAACGAGGACTATACGATCGTAGAACAATACGAGCGAGCGGAGGGACGCCATTCAACAGGGGGAATGGACGAATTATATAAATAG', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'initial_sequence_optimization_score': -0.6600145429325787, 'final_sequence_optimization_score': 3.4625768604112874, 'run_time': 20.219000000040978, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 3.4581864774870925}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 3.4625768604112874}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 3.4625768604112874}]}, {'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGGTTTAAAGTACATATGGAAGGGTCAGTGAATGGTCACGAATTCGAAATCGAAGGCGAGGGGGAGGGTAGACCATACGAGGGTACCCAGACGGCTAAGCTCAAAGTAACAAAGGGGGGGCCACTACCCTTCGCGTGGGACATTTTGTCGCCACAGTTCATGTACGGGAGCAAAGCCTATGTTAAACACCCAGCAGACATTCCCGATTATCTGAAGCTGTCTTTTCCGGAAGGATTCAAATGGGAACGGGTTATGAACTTTGAAGATGGCGGCGTGGTTACTGTTACGCAAGATTCCTCCCTGCAGGATGGCGAGTTCATCTATAAGGTGAAACTCAGAGGAACGAACTTCCCAAGTGATGGACCCGTGATGCAAAAGAAGACGATGGGGTGGGAGGCTTCTAGTGAAAGAATGTATCCAGAGGACGGAGCGCTCAAGGGTGAGATTAAGCAGCGGTTAAAGCTCAAGGATGGAGGGCATTACGACGCTGAAGTGAAAACAACATATAAGGCCAAAAAACCGGTCCAATTACCCGGAGCTTACAATGTTAATATTAAATTGGACATCACGTCACATAATGAGGATTATACGATAGTGGAACAATACGAACGTGCTGAAGGCCGACATTCGACCGGCGGGATGGATGAGCTTTACAAATGA', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'initial_sequence_optimization_score': -0.3682868028922508, 'final_sequence_optimization_score': 3.4625768604112874, 'run_time': 20.140999999945052, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 3.4581864774870925}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 3.4625768604112874}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 3.4625768604112874}]}, {'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGAGGTTCAAGGTTCACATGGAAGGTTCTGTAAATGGGCATGAATTTGAAATCGAGGGCGAAGGTGAGGGACGGCCGTACGAAGGGACGCAGACAGCAAAGTTGAAGGTCACCAAAGGTGGACCACTTCCATTTGCATGGGACATCCTTTCGCCCCAATTTATGTATGGTAGCAAAGCCTACGTAAAACACCCGGCGGACATACCAGATTATCTAAAGCTTAGCTTTCCCGAAGGATTTAAGTGGGAGAGGGTGATGAATTTCGAGGATGGGGGTGTTGTCACGGTGACTCAGGACTCGTCATTGCAAGATGGGGAATTCATATATAAGGTAAAACTAAGAGGAACGAATTTCCCCAGCGATGGACCCGTTATGCAGAAGAAAACGATGGGGTGGGAGGCATCTAGCGAGAGAATGTATCCAGAGGATGGTGCCCTAAAAGGGGAGATTAAGCAAAGACTCAAACTCAAAGACGGCGGCCACTACGATGCAGAGGTGAAGACCACCTACAAGGCTAAAAAACCGGTTCAACTACCCGGAGCATACAACGTGAATATAAAACTTGACATAACGTCTCATAATGAAGACTACACCATTGTTGAACAGTACGAAAGGGCAGAGGGCCGGCATAGTACCGGTGGGATGGATGAGCTATACAAATGA', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'initial_sequence_optimization_score': -0.7240910504855531, 'final_sequence_optimization_score': 3.4625768604112874, 'run_time': 20.18799999996554, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 3.4581864774870925}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 3.4625768604112874}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 3.4625768604112874}]}], 'evaluation': [{'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'cai_initial_score': 0.7695788606254016, 'cai_final_score': 0.9355715782388977, 'dist_score': 3.878153093806204}, {'name': 'Escherichia coli', 'is_wanted': False, 'cai_initial_score': 0.7915994634162647, 'cai_final_score': 0.4084514258280152, 'dist_score': -4.085662495964095}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA', 'average_distance_score': 3.9819077948851493, 'weakest_link_score': 3.9819077948851493, 'ratio_score': 1.1707526264894996, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA'}, {'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'cai_initial_score': 0.7695788606254016, 'cai_final_score': 0.9355715782388977, 'dist_score': 3.878153093806204}, {'name': 'Escherichia coli', 'is_wanted': False, 'cai_initial_score': 0.7915994634162647, 'cai_final_score': 0.4084514258280152, 'dist_score': -4.085662495964095}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA', 'average_distance_score': 3.9819077948851493, 'weakest_link_score': 3.9819077948851493, 'ratio_score': 1.1707526264894996, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA'}, {'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'cai_initial_score': 0.7695788606254016, 'cai_final_score': 0.9355715782388977, 'dist_score': 3.878153093806204}, {'name': 'Escherichia coli', 'is_wanted': False, 'cai_initial_score': 0.7915994634162647, 'cai_final_score': 0.4084514258280152, 'dist_score': -4.085662495964095}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA', 'average_distance_score': 3.9819077948851493, 'weakest_link_score': 3.9819077948851493, 'ratio_score': 1.1707526264894996, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA'}, {'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'cai_initial_score': 0.7695788606254016, 'cai_final_score': 0.9355715782388977, 'dist_score': 3.878153093806204}, {'name': 'Escherichia coli', 'is_wanted': False, 'cai_initial_score': 0.7915994634162647, 'cai_final_score': 0.4084514258280152, 'dist_score': -4.085662495964095}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA', 'average_distance_score': 3.9819077948851493, 'weakest_link_score': 3.9819077948851493, 'ratio_score': 1.1707526264894996, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA'}], 'final_evaluation': {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGCCCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGCGTTGTAACAGTTACACAAGATTCATCTCTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAGACAATGGGATGGGAAGCATCATCTGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGCCATTATGATGCAGAAGTTAAAACAACGTATAAAGCAAAAAAGCCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGCATGGATGAACTTTATAAATAA', 'average_distance_score': 3.9819077948851493, 'weakest_link_score': 3.9819077948851493, 'ratio_score': 1.1707526264894996}, 'zip_output_file_path': 'results\\\\mcherry-wanted-constraint-init-external-0.1\\\\CAI_zscore_bulk_aa_diff_ecoli_opt_False_WFw0\\\\communique_results.zip'}\n",
      "\n",
      "##########################\n",
      "# User Input #\n",
      "##########################\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is:\n",
      "ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "------------------------------------------\n",
      "Parsing information for Escherichia coli:\n",
      "------------------------------------------\n",
      "Organism is defined as wanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Escherichia coli. Calculating tAI profile.\n",
      "tai_std=0.024913119712427708, tai_avg=0.24958213077316901\n",
      "------------------------------------------\n",
      "Parsing information for Bacillus subtilis:\n",
      "------------------------------------------\n",
      "Organism is defined as unwanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Bacillus subtilis. Calculating tAI profile.\n",
      "tai_std=0.028076995641981673, tai_avg=0.3271366648197227\n",
      "-----------------------------\n",
      "Normalized Prioritization Weights\n",
      "-----------------------------\n",
      "Escherichia coli : 1.0\n",
      "Bacillus subtilis : 1.0\n",
      "Total input processing time: 28.01015567779541\n",
      "\n",
      "##########################\n",
      "# Initiation #\n",
      "##########################\n",
      "Running initiation optimization InitiationOptimizationMethod.external_module on 15 codons at the start of the ORF\n",
      "Taking optimized sequence from configuration value config.INITIATION.EXTERNAL_INITIATION_ORF\n",
      "Optimized sequence is: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "Optimization method used is: ORFOptimizationMethod.zscore_bulk_aa_diff\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Skipping codon AGA due to very low average frequency 0.036496252317115764 in wanted hosts.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGAGGTTCAAGGTGCATATGGAAGGTTCGGTAAACGGGCATGAATTCGAAATCGAGGGGGAAGGGGAAGGGCGACCATATGAAGGTACACAGACGGCCAAACTTAAAGTTACTAAGGGAGGGCCTCTGCCGTTCGCGTGGGACATTCTATCGCCCCAATTTATGTATGGGTCAAAGGCTTATGTGAAGCATCCGGCCGACATTCCTGATTACCTGAAGTTAAGTTTCCCAGAAGGTTTTAAGTGGGAGCGTGTGATGAACTTTGAAGATGGAGGAGTGGTCACGGTCACTCAGGATAGCAGCTTACAGGATGGGGAGTTTATTTATAAAGTGAAGTTGCGCGGGACAAATTTCCCTAGCGACGGCCCCGTAATGCAAAAAAAAACAATGGGTTGGGAAGCATCCTCGGAGAGAATGTACCCCGAAGATGGTGCACTTAAAGGAGAGATAAAACAAAGGCTCAAATTAAAAGACGGAGGTCATTATGATGCAGAGGTTAAGACGACATACAAAGCGAAGAAACCAGTCCAATTACCCGGAGCCTACAACGTTAATATCAAACTCGATATCACCTCACACAATGAGGATTATACAATTGTTGAGCAATATGAACGAGCAGAAGGTCGTCATTCAACCGGTGGTATGGACGAGCTATACAAATAG\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Skipping codon AGA due to very low average frequency 0.036496252317115764 in wanted hosts.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTACATATGGAAGGAAGTGTGAACGGACACGAATTCGAGATCGAAGGGGAAGGAGAGGGACGCCCCTATGAGGGAACCCAAACTGCCAAATTGAAGGTCACTAAGGGAGGTCCGCTCCCTTTTGCTTGGGATATACTCAGTCCACAATTTATGTATGGGTCTAAAGCTTATGTTAAGCATCCCGCTGATATACCTGACTACCTGAAATTGTCCTTCCCCGAGGGTTTTAAATGGGAACGAGTCATGAACTTCGAGGATGGGGGGGTAGTGACTGTGACCCAAGACAGCTCGTTACAGGACGGGGAATTCATATATAAGGTGAAATTGCGTGGGACTAACTTTCCTTCCGACGGACCCGTCATGCAAAAAAAGACAATGGGGTGGGAAGCATCATCCGAAAGAATGTACCCGGAGGATGGAGCACTCAAGGGCGAAATCAAACAAAGACTCAAACTCAAGGACGGTGGACATTACGATGCCGAAGTTAAAACCACGTATAAGGCAAAGAAACCAGTCCAATTGCCTGGTGCCTATAATGTAAACATCAAACTCGACATTACCTCCCATAACGAAGATTATACAATAGTAGAGCAATACGAGCGTGCTGAGGGCCGGCATTCAACGGGAGGTATGGACGAGCTTTACAAATAG\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Skipping codon AGA due to very low average frequency 0.036496252317115764 in wanted hosts.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAGGTCCATATGGAGGGAAGTGTCAACGGACATGAGTTTGAGATCGAAGGAGAAGGCGAAGGTCGGCCGTACGAAGGCACCCAGACGGCGAAGCTCAAAGTCACAAAGGGTGGCCCTCTTCCCTTCGCCTGGGACATCTTATCACCACAGTTCATGTACGGTTCCAAAGCGTACGTGAAACACCCCGCGGATATTCCTGATTATTTAAAGTTAAGCTTTCCTGAGGGTTTCAAATGGGAGCGTGTTATGAACTTCGAGGACGGCGGTGTAGTAACAGTCACTCAAGATAGTAGTCTACAAGACGGTGAATTCATTTATAAAGTTAAGCTAAGAGGTACTAATTTCCCCTCGGACGGTCCCGTAATGCAGAAAAAGACCATGGGTTGGGAGGCTAGCAGTGAGCGTATGTACCCAGAAGATGGCGCACTAAAGGGCGAAATTAAGCAAAGACTCAAACTAAAAGATGGTGGACACTATGACGCGGAGGTGAAAACCACCTATAAGGCAAAAAAGCCGGTGCAACTGCCGGGCGCGTATAATGTCAACATAAAGTTGGACATTACCAGCCATAACGAGGATTATACAATTGTTGAACAGTATGAGAGAGCTGAAGGGCGCCATAGTACCGGTGGTATGGACGAGTTATATAAATAA\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CGA due to very low average frequency 0.06388523224887838 in wanted hosts.\n",
      "Skipping codon CGG due to very low average frequency 0.09749348521075679 in wanted hosts.\n",
      "Skipping codon AGA due to very low average frequency 0.036496252317115764 in wanted hosts.\n",
      "Skipping codon AGG due to very low average frequency 0.020162265266098916 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon TAG due to very low average frequency 0.07162597881160755 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon ATA due to very low average frequency 0.07084223130192197 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "\n",
      "##########################\n",
      "# EVALUATION #\n",
      "##########################\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATCGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGCCCGCTGCCGTTCGCCTGGGACATCCTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGCGTCGTTACCGTCACCCAGGACTCCTCGCTGCAGGACGGGGAATTCATCTACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAGACCATGGGGTGGGAAGCCTCCTCGGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATCAAACAGCGTCTGAAACTGAAAGACGGGGGCCACTACGACGCCGAAGTCAAAACCACTTACAAAGCCAAAAAGCCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGCATGGACGAACTGTACAAATAA\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATCGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGCCCGCTGCCGTTCGCCTGGGACATCCTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGCGTCGTTACCGTCACCCAGGACTCCTCGCTGCAGGACGGGGAATTCATCTACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAGACCATGGGGTGGGAAGCCTCCTCGGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATCAAACAGCGTCTGAAACTGAAAGACGGGGGCCACTACGACGCCGAAGTCAAAACCACTTACAAAGCCAAAAAGCCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGCATGGACGAACTGTACAAATAG\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATCGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGCCCGCTGCCGTTCGCCTGGGACATCCTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGCGTCGTTACCGTCACCCAGGACTCCTCGCTGCAGGACGGGGAATTCATCTACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAGACCATGGGGTGGGAAGCCTCCTCGGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATCAAACAGCGTCTGAAACTGAAAGACGGGGGCCACTACGACGCCGAAGTCAAAACCACTTACAAAGCCAAAAAGCCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGCATGGACGAACTGTACAAATAG\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATCGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGCCCGCTGCCGTTCGCCTGGGACATCCTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGCGTCGTTACCGTCACCCAGGACTCCTCGCTGCAGGACGGGGAATTCATCTACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAGACCATGGGGTGGGAAGCCTCCTCGGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATCAAACAGCGTCTGAAACTGAAAGACGGGGGCCACTACGACGCCGAAGTCAAAACCACTTACAAAGCCAAAAAGCCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGCATGGACGAACTGTACAAATAA\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATCGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGCCCGCTGCCGTTCGCCTGGGACATCCTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGCGTCGTTACCGTCACCCAGGACTCCTCGCTGCAGGACGGGGAATTCATCTACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAGACCATGGGGTGGGAAGCCTCCTCGGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATCAAACAGCGTCTGAAACTGAAAGACGGGGGCCACTACGACGCCGAAGTCAAAACCACTTACAAAGCCAAAAAGCCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGCATGGACGAACTGTACAAATAA', 'average_distance_score': 0.9097083250513909, 'weakest_link_score': 0.9097083250513909, 'ratio_score': 1.0408188642521476}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATCGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGCCCGCTGCCGTTCGCCTGGGACATCCTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGCGTCGTTACCGTCACCCAGGACTCCTCGCTGCAGGACGGGGAATTCATCTACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAGACCATGGGGTGGGAAGCCTCCTCGGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATCAAACAGCGTCTGAAACTGAAAGACGGGGGCCACTACGACGCCGAAGTCAAAACCACTTACAAAGCCAAAAAGCCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGCATGGACGAACTGTACAAATAG', 'average_distance_score': 0.9097083250513909, 'weakest_link_score': 0.9097083250513909, 'ratio_score': 1.0408188642521476}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATCGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGCCCGCTGCCGTTCGCCTGGGACATCCTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGCGTCGTTACCGTCACCCAGGACTCCTCGCTGCAGGACGGGGAATTCATCTACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAGACCATGGGGTGGGAAGCCTCCTCGGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATCAAACAGCGTCTGAAACTGAAAGACGGGGGCCACTACGACGCCGAAGTCAAAACCACTTACAAAGCCAAAAAGCCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGCATGGACGAACTGTACAAATAG', 'average_distance_score': 0.9097083250513909, 'weakest_link_score': 0.9097083250513909, 'ratio_score': 1.0408188642521476}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATCGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGCCCGCTGCCGTTCGCCTGGGACATCCTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGCGTCGTTACCGTCACCCAGGACTCCTCGCTGCAGGACGGGGAATTCATCTACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAGACCATGGGGTGGGAAGCCTCCTCGGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATCAAACAGCGTCTGAAACTGAAAGACGGGGGCCACTACGACGCCGAAGTCAAAACCACTTACAAAGCCAAAAAGCCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGCATGGACGAACTGTACAAATAA', 'average_distance_score': 0.9097083250513909, 'weakest_link_score': 0.9097083250513909, 'ratio_score': 1.0408188642521476}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATCGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGCCCGCTGCCGTTCGCCTGGGACATCCTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGCGTCGTTACCGTCACCCAGGACTCCTCGCTGCAGGACGGGGAATTCATCTACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAGACCATGGGGTGGGAAGCCTCCTCGGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATCAAACAGCGTCTGAAACTGAAAGACGGGGGCCACTACGACGCCGAAGTCAAAACCACTTACAAAGCCAAAAAGCCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGCATGGACGAACTGTACAAATAA', 'average_distance_score': 0.9097083250513909, 'weakest_link_score': 0.9097083250513909, 'ratio_score': 1.0408188642521476}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry-wanted-constraint-init-external-0.1\\tAI_zscore_bulk_aa_diff_ecoli_opt_True_AYoX\n",
      "Final output: {'module_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'orf_optimization_method': 'zscore_bulk_aa_diff', 'orf_optimization_cub_index': 'tAI', 'initiation_optimization_method': 'external', 'evaluation_score': 'average_distance', 'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': None, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'AAA': 0.7667390608953314, 'CGC': 0.40076028262096014, 'ATT': 0.5085852538544149, 'AGC': 0.27799163607995603, 'ACC': 0.4359429381095841, 'ATC': 0.4205725148436632, 'ACA': 0.13023262212317555, 'GGT': 0.33726003930064996, 'AAC': 0.5512262248110238, 'GCG': 0.3563679190097728, 'GGC': 0.4050486219579785, 'TGA': 0.28926761860893596, 'CGA': 0.06388523224887838, 'GTG': 0.3715940695510341, 'TTG': 0.12808970030428077, 'AAG': 0.23326093910466858, 'TTC': 0.4262385670731707, 'TCA': 0.12251736088908215, 'GCA': 0.21281773299119433, 'AAT': 0.4487737751889762, 'GAA': 0.689467997530356, 'CGT': 0.38120248233619, 'TTT': 0.5737614329268292, 'CTG': 0.4975546385450099, 'GTT': 0.2584847661069405, 'GCC': 0.26990242799092135, 'GAT': 0.626856640686903, 'AGG': 0.020162265266098916, 'CAG': 0.6530527088827319, 'GGG': 0.15057187484254547, 'GTC': 0.21619163129169194, 'CTC': 0.10435163186194824, 'TCT': 0.14544780223293646, 'CCC': 0.12361073915372255, 'CAC': 0.4292446655321377, 'GCT': 0.1609119200081115, 'TTA': 0.13009739763522432, 'CTT': 0.10341378016756284, 'ACG': 0.26849903016796667, 'GGA': 0.10711946389882603, 'CCG': 0.527785216925549, 'CAA': 0.3469472911172681, 'ACT': 0.16532540959927366, 'ATA': 0.07084223130192197, 'CAT': 0.5707553344678623, 'AGT': 0.15102374892893225, 'TGC': 0.55651396076928, 'GAG': 0.31053200246964396, 'TCG': 0.15406749964830613, 'GTA': 0.15372953305033354, 'TAC': 0.4315172054223149, 'TCC': 0.14895195222078703, 'GAC': 0.37314335931309695, 'TGT': 0.4434860392307201, 'TGG': 1.0, 'TAT': 0.5684827945776851, 'CCT': 0.15789033208355652, 'CCA': 0.19071371183717192, 'CGG': 0.09749348521075679, 'CTA': 0.03649285148597391, 'TAA': 0.6391064025794565, 'AGA': 0.036496252317115764, 'TAG': 0.07162597881160755}), 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.75, 'AAC': 0.5, 'AAG': 0.23999999999999996, 'AAT': 0.29500000000000004, 'ACA': 0.1525, 'ACC': 0.25, 'ACG': 0.29, 'ACT': 0.14750000000000002, 'AGA': 0.13874999999999998, 'AGC': 0.125, 'AGG': 0.16499999999999998, 'AGT': 0.07375000000000001, 'ATA': 0.10999999999999999, 'ATC': 0.375, 'ATG': 1.0, 'ATT': 0.22125000000000003, 'CAA': 0.27749999999999997, 'CAC': 0.125, 'CAG': 0.32999999999999996, 'CAT': 0.07375000000000001, 'CCA': 0.13874999999999998, 'CCC': 0.125, 'CCG': 0.16499999999999998, 'CCT': 0.07375000000000001, 'CGA': 0.013799999999999993, 'CGC': 0.36, 'CGG': 0.125, 'CGT': 0.5, 'CTA': 0.18, 'CTC': 0.125, 'CTG': 0.54, 'CTT': 0.07375000000000001, 'GAA': 0.5, 'GAC': 0.375, 'GAG': 0.15999999999999998, 'GAT': 0.22125000000000003, 'GCA': 0.375, 'GCC': 0.25, 'GCG': 0.11999999999999998, 'GCT': 0.14750000000000002, 'GGA': 0.13874999999999998, 'GGC': 0.5, 'GGG': 0.16499999999999998, 'GGT': 0.29500000000000004, 'GTA': 0.625, 'GTC': 0.25, 'GTG': 0.19999999999999996, 'GTT': 0.14750000000000002, 'TAC': 0.375, 'TAT': 0.22125000000000003, 'TCA': 0.13874999999999998, 'TCC': 0.25, 'TCG': 0.16499999999999998, 'TCT': 0.14750000000000002, 'TGC': 0.125, 'TGG': 0.125, 'TGT': 0.07375000000000001, 'TTA': 0.13874999999999998, 'TTC': 0.25, 'TTG': 0.16499999999999998, 'TTT': 0.14750000000000002}, 'tai_avg': 0.24958213077316901, 'tai_std': 0.024913119712427708}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': None, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'GAA': 0.6798186496752849, 'AAT': 0.5655317497600131, 'ATA': 0.1287010322849589, 'TTA': 0.198530453110291, 'GAC': 0.35960560391899904, 'CTG': 0.23973441522197322, 'TGG': 1.0, 'AAC': 0.4344682502399869, 'CAA': 0.5131548459124133, 'GCC': 0.20715414603308396, 'CTT': 0.2391679232236727, 'GCT': 0.24752923822568748, 'ATC': 0.366762837741479, 'AAA': 0.7016557540022625, 'AAG': 0.29834424599773746, 'TTG': 0.1592175745811708, 'AGC': 0.22534685577492342, 'CCG': 0.4334091260292722, 'AGT': 0.10611598754150686, 'TTT': 0.6847751982489505, 'GAG': 0.3201813503247151, 'ACT': 0.1614000537105004, 'TCA': 0.23653118484388272, 'ACC': 0.1588786441082565, 'CAC': 0.3265866087018494, 'GGC': 0.3390368073029434, 'GAT': 0.640394396081001, 'ACA': 0.4122430101751559, 'ACG': 0.2674782920060872, 'CCC': 0.0889644363010221, 'AGA': 0.26380392311639106, 'TCC': 0.12712038919920718, 'TAC': 0.34586274736720285, 'CAT': 0.6734133912981506, 'ATT': 0.5045361299735621, 'GCA': 0.2828363712991255, 'TAT': 0.6541372526327971, 'GGG': 0.16211598104043537, 'GTC': 0.25598186049286953, 'CCT': 0.28667299498885185, 'GTT': 0.28439644370189154, 'GCG': 0.26248024444210305, 'TCT': 0.2048572678833432, 'CTC': 0.11219040795754642, 'CCA': 0.19095344268085387, 'GGA': 0.3148341038094681, 'CGA': 0.09855152543710778, 'GTA': 0.19774449549495793, 'GTG': 0.26187720031028106, 'TTC': 0.31522480175104955, 'CGC': 0.2064569601768165, 'CAG': 0.4868451540875866, 'CGG': 0.1555630106168844, 'CGT': 0.18161187196589967, 'CTA': 0.05115922590534585, 'AGG': 0.09401270868690058, 'GGT': 0.18401310784715313, 'TAG': 0.14132712456344587, 'TGC': 0.5460141582025239, 'TCG': 0.10002831475713661, 'TAA': 0.630500582072177, 'TGA': 0.2281722933643772, 'TGT': 0.45398584179747614}), 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.6666666666666666, 'AAC': 0.6666666666666666, 'AAG': 0.2133333333333333, 'AAT': 0.39333333333333337, 'ACA': 0.6666666666666666, 'ACC': 0.16666666666666666, 'ACG': 0.2133333333333333, 'ACT': 0.09833333333333334, 'AGA': 0.18499999999999997, 'AGC': 0.3333333333333333, 'AGG': 0.21999999999999997, 'AGT': 0.19666666666666668, 'ATA': 0.10999999999999999, 'ATC': 0.5, 'ATG': 1.0, 'ATT': 0.29500000000000004, 'CAA': 0.6666666666666666, 'CAC': 0.3333333333333333, 'CAG': 0.2133333333333333, 'CAT': 0.19666666666666668, 'CCA': 0.5, 'CCG': 0.15999999999999998, 'CGA': 0.01839999999999999, 'CGC': 0.48, 'CGG': 0.16666666666666666, 'CGT': 0.6666666666666666, 'CTA': 0.3516666666666666, 'CTC': 0.16666666666666666, 'CTG': 0.2733333333333333, 'CTT': 0.09833333333333334, 'GAA': 1.0, 'GAC': 0.6666666666666666, 'GAG': 0.31999999999999995, 'GAT': 0.39333333333333337, 'GCA': 0.8333333333333334, 'GCC': 0.16666666666666666, 'GCG': 0.2666666666666666, 'GCT': 0.09833333333333334, 'GGA': 0.5, 'GGC': 0.6666666666666666, 'GGG': 0.15999999999999998, 'GGT': 0.39333333333333337, 'GTA': 0.6666666666666666, 'GTC': 0.16666666666666666, 'GTG': 0.2133333333333333, 'GTT': 0.09833333333333334, 'TAC': 0.3333333333333333, 'TAT': 0.19666666666666668, 'TCA': 0.3333333333333333, 'TCC': 0.16666666666666666, 'TCG': 0.10666666666666665, 'TCT': 0.09833333333333334, 'TGC': 0.16666666666666666, 'TGG': 0.16666666666666666, 'TGT': 0.09833333333333334, 'TTA': 0.5183333333333333, 'TTC': 0.5, 'TTG': 0.32666666666666666, 'TTT': 0.29500000000000004}, 'tai_avg': 0.3271366648197227, 'tai_std': 0.028076995641981673}]}, 'orf': [{'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATCGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGCCCGCTGCCGTTCGCCTGGGACATCCTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGCGTCGTTACCGTCACCCAGGACTCCTCGCTGCAGGACGGGGAATTCATCTACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAGACCATGGGGTGGGAAGCCTCCTCGGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATCAAACAGCGTCTGAAACTGAAAGACGGGGGCCACTACGACGCCGAAGTCAAAACCACTTACAAAGCCAAAAAGCCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGCATGGACGAACTGTACAAATAA', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCC', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCG', 'K': 'AAA', '_': 'TAA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTC', 'E': 'GAA', 'Y': 'TAC'}, 'initial_sequence_optimization_score': 0.46828435194706775, 'final_sequence_optimization_score': 1.3779926769984585, 'run_time': 19.875, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATT', 'L': 'CTG', 'H': 'CAT', 'R': 'CGC', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'sequence_score': 1.1715017828805643}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCC', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCG', 'K': 'AAA', '_': 'TAA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTC', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 1.3779926769984585}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCC', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCG', 'K': 'AAA', '_': 'TAA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTC', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 1.3779926769984585}]}, {'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGAGGTTCAAGGTGCATATGGAAGGTTCGGTAAACGGGCATGAATTCGAAATCGAGGGGGAAGGGGAAGGGCGACCATATGAAGGTACACAGACGGCCAAACTTAAAGTTACTAAGGGAGGGCCTCTGCCGTTCGCGTGGGACATTCTATCGCCCCAATTTATGTATGGGTCAAAGGCTTATGTGAAGCATCCGGCCGACATTCCTGATTACCTGAAGTTAAGTTTCCCAGAAGGTTTTAAGTGGGAGCGTGTGATGAACTTTGAAGATGGAGGAGTGGTCACGGTCACTCAGGATAGCAGCTTACAGGATGGGGAGTTTATTTATAAAGTGAAGTTGCGCGGGACAAATTTCCCTAGCGACGGCCCCGTAATGCAAAAAAAAACAATGGGTTGGGAAGCATCCTCGGAGAGAATGTACCCCGAAGATGGTGCACTTAAAGGAGAGATAAAACAAAGGCTCAAATTAAAAGACGGAGGTCATTATGATGCAGAGGTTAAGACGACATACAAAGCGAAGAAACCAGTCCAATTACCCGGAGCCTACAACGTTAATATCAAACTCGATATCACCTCACACAATGAGGATTATACAATTGTTGAGCAATATGAACGAGCAGAAGGTCGTCATTCAACCGGTGGTATGGACGAGCTATACAAATAG', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATCGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGCCCGCTGCCGTTCGCCTGGGACATCCTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGCGTCGTTACCGTCACCCAGGACTCCTCGCTGCAGGACGGGGAATTCATCTACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAGACCATGGGGTGGGAAGCCTCCTCGGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATCAAACAGCGTCTGAAACTGAAAGACGGGGGCCACTACGACGCCGAAGTCAAAACCACTTACAAAGCCAAAAAGCCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGCATGGACGAACTGTACAAATAG', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCC', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCG', 'K': 'AAA', '_': 'TAA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTC', 'E': 'GAA', 'Y': 'TAC'}, 'initial_sequence_optimization_score': -0.5470880593030707, 'final_sequence_optimization_score': 1.3779926769984585, 'run_time': 20.18799999996554, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATT', 'L': 'CTG', 'H': 'CAT', 'R': 'CGC', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'sequence_score': 1.1715017828805643}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCC', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCG', 'K': 'AAA', '_': 'TAA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTC', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 1.3779926769984585}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCC', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCG', 'K': 'AAA', '_': 'TAA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTC', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 1.3779926769984585}]}, {'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTACATATGGAAGGAAGTGTGAACGGACACGAATTCGAGATCGAAGGGGAAGGAGAGGGACGCCCCTATGAGGGAACCCAAACTGCCAAATTGAAGGTCACTAAGGGAGGTCCGCTCCCTTTTGCTTGGGATATACTCAGTCCACAATTTATGTATGGGTCTAAAGCTTATGTTAAGCATCCCGCTGATATACCTGACTACCTGAAATTGTCCTTCCCCGAGGGTTTTAAATGGGAACGAGTCATGAACTTCGAGGATGGGGGGGTAGTGACTGTGACCCAAGACAGCTCGTTACAGGACGGGGAATTCATATATAAGGTGAAATTGCGTGGGACTAACTTTCCTTCCGACGGACCCGTCATGCAAAAAAAGACAATGGGGTGGGAAGCATCATCCGAAAGAATGTACCCGGAGGATGGAGCACTCAAGGGCGAAATCAAACAAAGACTCAAACTCAAGGACGGTGGACATTACGATGCCGAAGTTAAAACCACGTATAAGGCAAAGAAACCAGTCCAATTGCCTGGTGCCTATAATGTAAACATCAAACTCGACATTACCTCCCATAACGAAGATTATACAATAGTAGAGCAATACGAGCGTGCTGAGGGCCGGCATTCAACGGGAGGTATGGACGAGCTTTACAAATAG', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATCGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGCCCGCTGCCGTTCGCCTGGGACATCCTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGCGTCGTTACCGTCACCCAGGACTCCTCGCTGCAGGACGGGGAATTCATCTACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAGACCATGGGGTGGGAAGCCTCCTCGGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATCAAACAGCGTCTGAAACTGAAAGACGGGGGCCACTACGACGCCGAAGTCAAAACCACTTACAAAGCCAAAAAGCCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGCATGGACGAACTGTACAAATAG', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCC', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCG', 'K': 'AAA', '_': 'TAA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTC', 'E': 'GAA', 'Y': 'TAC'}, 'initial_sequence_optimization_score': -0.3662091447147084, 'final_sequence_optimization_score': 1.3779926769984585, 'run_time': 20.31299999996554, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATT', 'L': 'CTG', 'H': 'CAT', 'R': 'CGC', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'sequence_score': 1.1715017828805643}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCC', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCG', 'K': 'AAA', '_': 'TAA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTC', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 1.3779926769984585}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCC', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCG', 'K': 'AAA', '_': 'TAA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTC', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 1.3779926769984585}]}, {'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGAGATTTAAGGTCCATATGGAGGGAAGTGTCAACGGACATGAGTTTGAGATCGAAGGAGAAGGCGAAGGTCGGCCGTACGAAGGCACCCAGACGGCGAAGCTCAAAGTCACAAAGGGTGGCCCTCTTCCCTTCGCCTGGGACATCTTATCACCACAGTTCATGTACGGTTCCAAAGCGTACGTGAAACACCCCGCGGATATTCCTGATTATTTAAAGTTAAGCTTTCCTGAGGGTTTCAAATGGGAGCGTGTTATGAACTTCGAGGACGGCGGTGTAGTAACAGTCACTCAAGATAGTAGTCTACAAGACGGTGAATTCATTTATAAAGTTAAGCTAAGAGGTACTAATTTCCCCTCGGACGGTCCCGTAATGCAGAAAAAGACCATGGGTTGGGAGGCTAGCAGTGAGCGTATGTACCCAGAAGATGGCGCACTAAAGGGCGAAATTAAGCAAAGACTCAAACTAAAAGATGGTGGACACTATGACGCGGAGGTGAAAACCACCTATAAGGCAAAAAAGCCGGTGCAACTGCCGGGCGCGTATAATGTCAACATAAAGTTGGACATTACCAGCCATAACGAGGATTATACAATTGTTGAACAGTATGAGAGAGCTGAAGGGCGCCATAGTACCGGTGGTATGGACGAGTTATATAAATAA', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATCGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGCCCGCTGCCGTTCGCCTGGGACATCCTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGCGTCGTTACCGTCACCCAGGACTCCTCGCTGCAGGACGGGGAATTCATCTACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAGACCATGGGGTGGGAAGCCTCCTCGGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATCAAACAGCGTCTGAAACTGAAAGACGGGGGCCACTACGACGCCGAAGTCAAAACCACTTACAAAGCCAAAAAGCCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGCATGGACGAACTGTACAAATAA', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCC', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCG', 'K': 'AAA', '_': 'TAA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTC', 'E': 'GAA', 'Y': 'TAC'}, 'initial_sequence_optimization_score': -0.5531460755317997, 'final_sequence_optimization_score': 1.3779926769984585, 'run_time': 19.76500000001397, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATT', 'L': 'CTG', 'H': 'CAT', 'R': 'CGC', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'sequence_score': 1.1715017828805643}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCC', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCG', 'K': 'AAA', '_': 'TAA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTC', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 1.3779926769984585}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCC', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCG', 'K': 'AAA', '_': 'TAA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTC', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 1.3779926769984585}]}], 'evaluation': [{'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'tai_initial_score': 0.2816376469779175, 'tai_final_score': 0.32470626161202354, 'dist_score': 1.7287523654703758}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'tai_initial_score': 0.33696708001007003, 'tai_final_score': 0.3344214992855626, 'dist_score': -0.09066428463240588}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATCGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGCCCGCTGCCGTTCGCCTGGGACATCCTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGCGTCGTTACCGTCACCCAGGACTCCTCGCTGCAGGACGGGGAATTCATCTACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAGACCATGGGGTGGGAAGCCTCCTCGGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATCAAACAGCGTCTGAAACTGAAAGACGGGGGCCACTACGACGCCGAAGTCAAAACCACTTACAAAGCCAAAAAGCCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGCATGGACGAACTGTACAAATAA', 'average_distance_score': 0.9097083250513909, 'weakest_link_score': 0.9097083250513909, 'ratio_score': 1.0408188642521476, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATCGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGCCCGCTGCCGTTCGCCTGGGACATCCTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGCGTCGTTACCGTCACCCAGGACTCCTCGCTGCAGGACGGGGAATTCATCTACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAGACCATGGGGTGGGAAGCCTCCTCGGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATCAAACAGCGTCTGAAACTGAAAGACGGGGGCCACTACGACGCCGAAGTCAAAACCACTTACAAAGCCAAAAAGCCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGCATGGACGAACTGTACAAATAA'}, {'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'tai_initial_score': 0.2816376469779175, 'tai_final_score': 0.32470626161202354, 'dist_score': 1.7287523654703758}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'tai_initial_score': 0.33696708001007003, 'tai_final_score': 0.3344214992855626, 'dist_score': -0.09066428463240588}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATCGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGCCCGCTGCCGTTCGCCTGGGACATCCTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGCGTCGTTACCGTCACCCAGGACTCCTCGCTGCAGGACGGGGAATTCATCTACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAGACCATGGGGTGGGAAGCCTCCTCGGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATCAAACAGCGTCTGAAACTGAAAGACGGGGGCCACTACGACGCCGAAGTCAAAACCACTTACAAAGCCAAAAAGCCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGCATGGACGAACTGTACAAATAG', 'average_distance_score': 0.9097083250513909, 'weakest_link_score': 0.9097083250513909, 'ratio_score': 1.0408188642521476, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATCGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGCCCGCTGCCGTTCGCCTGGGACATCCTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGCGTCGTTACCGTCACCCAGGACTCCTCGCTGCAGGACGGGGAATTCATCTACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAGACCATGGGGTGGGAAGCCTCCTCGGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATCAAACAGCGTCTGAAACTGAAAGACGGGGGCCACTACGACGCCGAAGTCAAAACCACTTACAAAGCCAAAAAGCCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGCATGGACGAACTGTACAAATAG'}, {'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'tai_initial_score': 0.2816376469779175, 'tai_final_score': 0.32470626161202354, 'dist_score': 1.7287523654703758}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'tai_initial_score': 0.33696708001007003, 'tai_final_score': 0.3344214992855626, 'dist_score': -0.09066428463240588}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATCGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGCCCGCTGCCGTTCGCCTGGGACATCCTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGCGTCGTTACCGTCACCCAGGACTCCTCGCTGCAGGACGGGGAATTCATCTACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAGACCATGGGGTGGGAAGCCTCCTCGGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATCAAACAGCGTCTGAAACTGAAAGACGGGGGCCACTACGACGCCGAAGTCAAAACCACTTACAAAGCCAAAAAGCCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGCATGGACGAACTGTACAAATAG', 'average_distance_score': 0.9097083250513909, 'weakest_link_score': 0.9097083250513909, 'ratio_score': 1.0408188642521476, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATCGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGCCCGCTGCCGTTCGCCTGGGACATCCTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGCGTCGTTACCGTCACCCAGGACTCCTCGCTGCAGGACGGGGAATTCATCTACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAGACCATGGGGTGGGAAGCCTCCTCGGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATCAAACAGCGTCTGAAACTGAAAGACGGGGGCCACTACGACGCCGAAGTCAAAACCACTTACAAAGCCAAAAAGCCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGCATGGACGAACTGTACAAATAG'}, {'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'tai_initial_score': 0.2816376469779175, 'tai_final_score': 0.32470626161202354, 'dist_score': 1.7287523654703758}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'tai_initial_score': 0.33696708001007003, 'tai_final_score': 0.3344214992855626, 'dist_score': -0.09066428463240588}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATCGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGCCCGCTGCCGTTCGCCTGGGACATCCTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGCGTCGTTACCGTCACCCAGGACTCCTCGCTGCAGGACGGGGAATTCATCTACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAGACCATGGGGTGGGAAGCCTCCTCGGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATCAAACAGCGTCTGAAACTGAAAGACGGGGGCCACTACGACGCCGAAGTCAAAACCACTTACAAAGCCAAAAAGCCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGCATGGACGAACTGTACAAATAA', 'average_distance_score': 0.9097083250513909, 'weakest_link_score': 0.9097083250513909, 'ratio_score': 1.0408188642521476, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATCGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGCCCGCTGCCGTTCGCCTGGGACATCCTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGCGTCGTTACCGTCACCCAGGACTCCTCGCTGCAGGACGGGGAATTCATCTACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAGACCATGGGGTGGGAAGCCTCCTCGGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATCAAACAGCGTCTGAAACTGAAAGACGGGGGCCACTACGACGCCGAAGTCAAAACCACTTACAAAGCCAAAAAGCCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGCATGGACGAACTGTACAAATAA'}], 'final_evaluation': {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATCGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGCCCGCTGCCGTTCGCCTGGGACATCCTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGCGTCGTTACCGTCACCCAGGACTCCTCGCTGCAGGACGGGGAATTCATCTACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAGACCATGGGGTGGGAAGCCTCCTCGGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATCAAACAGCGTCTGAAACTGAAAGACGGGGGCCACTACGACGCCGAAGTCAAAACCACTTACAAAGCCAAAAAGCCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGCATGGACGAACTGTACAAATAA', 'average_distance_score': 0.9097083250513909, 'weakest_link_score': 0.9097083250513909, 'ratio_score': 1.0408188642521476}, 'zip_output_file_path': 'results\\\\mcherry-wanted-constraint-init-external-0.1\\\\tAI_zscore_bulk_aa_diff_ecoli_opt_True_AYoX\\\\communique_results.zip'}\n",
      "\n",
      "##########################\n",
      "# User Input #\n",
      "##########################\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is:\n",
      "ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "------------------------------------------\n",
      "Parsing information for Bacillus subtilis:\n",
      "------------------------------------------\n",
      "Organism is defined as wanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Bacillus subtilis. Calculating tAI profile.\n",
      "tai_std=0.028076995641981673, tai_avg=0.3271366648197227\n",
      "------------------------------------------\n",
      "Parsing information for Escherichia coli:\n",
      "------------------------------------------\n",
      "Organism is defined as unwanted\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Escherichia coli. Calculating tAI profile.\n",
      "tai_std=0.024913119712427708, tai_avg=0.24958213077316901\n",
      "-----------------------------\n",
      "Normalized Prioritization Weights\n",
      "-----------------------------\n",
      "Bacillus subtilis : 1.0\n",
      "Escherichia coli : 1.0\n",
      "Total input processing time: 27.914368629455566\n",
      "\n",
      "##########################\n",
      "# Initiation #\n",
      "##########################\n",
      "Running initiation optimization InitiationOptimizationMethod.external_module on 15 codons at the start of the ORF\n",
      "Taking optimized sequence from configuration value config.INITIATION.EXTERNAL_INITIATION_ORF\n",
      "Optimized sequence is: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "Optimization method used is: ORFOptimizationMethod.zscore_bulk_aa_diff\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CTA due to very low average frequency 0.05115922590534585 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CTA due to very low average frequency 0.05115922590534585 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTTAAAGTACACATGGAAGGCTCCGTGAATGGGCACGAGTTCGAAATTGAGGGGGAGGGGGAAGGGCGTCCTTACGAAGGAACACAGACAGCTAAGCTCAAAGTGACAAAAGGAGGCCCTTTACCGTTCGCATGGGATATTCTCAGTCCCCAATTCATGTATGGCTCAAAGGCCTATGTCAAACATCCGGCCGATATCCCTGATTATCTTAAATTGAGTTTCCCCGAGGGTTTTAAATGGGAGCGTGTCATGAACTTTGAGGATGGCGGAGTTGTTACTGTCACGCAAGATAGCTCACTACAAGATGGCGAGTTCATATATAAGGTTAAGTTACGTGGCACCAATTTCCCGTCCGATGGGCCCGTGATGCAAAAGAAGACCATGGGGTGGGAGGCTAGTAGTGAAAGGATGTACCCTGAGGATGGAGCGTTGAAAGGTGAGATTAAACAACGTTTAAAGCTGAAAGACGGTGGCCATTATGATGCTGAGGTAAAAACGACCTACAAGGCCAAGAAACCGGTGCAACTTCCTGGCGCCTACAACGTGAATATCAAACTGGACATTACTTCCCACAACGAGGATTATACCATCGTCGAGCAATATGAGCGTGCTGAGGGAAGGCACTCCACAGGCGGGATGGACGAGCTTTATAAGTGA\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CTA due to very low average frequency 0.05115922590534585 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CTA due to very low average frequency 0.05115922590534585 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CTA due to very low average frequency 0.05115922590534585 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGGTTTAAGGTCCATATGGAGGGTTCGGTAAACGGACATGAGTTTGAAATCGAAGGCGAAGGTGAAGGAAGGCCGTACGAGGGAACGCAGACTGCAAAGCTAAAGGTTACAAAAGGCGGCCCTTTACCTTTCGCCTGGGATATATTGTCTCCCCAATTCATGTACGGTTCCAAAGCCTACGTAAAACATCCAGCCGATATCCCGGATTACCTAAAGTTAAGCTTCCCTGAGGGATTCAAGTGGGAGAGGGTAATGAACTTTGAGGACGGGGGTGTGGTCACAGTGACCCAGGATAGTAGTTTACAAGACGGTGAATTTATTTACAAGGTAAAACTACGAGGGACCAACTTCCCGTCGGACGGTCCTGTGATGCAGAAAAAGACGATGGGTTGGGAGGCTAGTTCTGAAAGGATGTACCCTGAAGATGGGGCTCTCAAAGGCGAAATTAAACAGAGACTGAAACTCAAGGATGGGGGACATTATGACGCTGAGGTTAAGACCACCTACAAGGCAAAGAAGCCGGTTCAGCTACCGGGTGCCTACAACGTCAATATAAAGCTGGACATTACATCCCACAATGAAGATTATACTATAGTAGAACAATACGAAAGAGCCGAGGGGCGGCATAGCACTGGCGGGATGGACGAACTTTACAAATAA\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CTA due to very low average frequency 0.05115922590534585 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CTA due to very low average frequency 0.05115922590534585 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CTA due to very low average frequency 0.05115922590534585 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Running ORF optimization for initial sequence: ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTTCACATGGAGGGCAGTGTTAATGGGCACGAATTCGAGATAGAAGGCGAAGGAGAAGGACGGCCATACGAGGGCACGCAAACCGCTAAACTAAAGGTCACAAAGGGAGGACCGCTTCCATTCGCATGGGATATTCTTAGCCCTCAATTTATGTATGGGAGCAAAGCATACGTGAAGCACCCAGCGGACATTCCTGACTACTTAAAACTAAGCTTTCCTGAGGGCTTCAAATGGGAACGCGTAATGAACTTTGAGGATGGTGGGGTAGTTACCGTCACTCAAGATTCGTCTCTGCAGGACGGTGAGTTTATTTATAAAGTTAAGCTCCGCGGAACAAATTTTCCTAGCGACGGGCCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCTTCATCGGAACGCATGTATCCAGAGGATGGAGCTCTGAAAGGAGAAATTAAACAGAGGCTTAAACTTAAGGACGGTGGCCATTACGACGCTGAAGTCAAAACCACCTATAAAGCGAAAAAACCCGTTCAATTACCTGGTGCATATAATGTCAATATTAAATTGGATATCACCAGCCACAACGAAGATTACACAATCGTGGAGCAGTATGAACGCGCCGAAGGGCGTCATTCAACGGGCGGGATGGATGAGCTTTACAAATAA\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CTA due to very low average frequency 0.05115922590534585 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CTA due to very low average frequency 0.05115922590534585 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Skipping codon CTA due to very low average frequency 0.05115922590534585 in wanted hosts.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "Found optimal codon. Now looking for second-best option.\n",
      "\n",
      "##########################\n",
      "# EVALUATION #\n",
      "##########################\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCTCAGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCTCAGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATGA\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCTCAGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA\n",
      "Calculating evaluation score based on initial sequence: TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA andfinal sequence: TTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCTCAGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCTCAGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA', 'average_distance_score': 3.9049289997955983, 'weakest_link_score': 3.9049289997955983, 'ratio_score': 1.1606489296917233}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCTCAGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATGA', 'average_distance_score': 3.9049289997955983, 'weakest_link_score': 3.9049289997955983, 'ratio_score': 1.1606489296917233}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCTCAGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA', 'average_distance_score': 3.9049289997955983, 'weakest_link_score': 3.9049289997955983, 'ratio_score': 1.1606489296917233}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCTCAGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA', 'average_distance_score': 3.9049289997955983, 'weakest_link_score': 3.9049289997955983, 'ratio_score': 1.1606489296917233}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCTCAGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA', 'average_distance_score': 3.9049289997955983, 'weakest_link_score': 3.9049289997955983, 'ratio_score': 1.1606489296917233}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry-wanted-constraint-init-external-0.1\\tAI_zscore_bulk_aa_diff_ecoli_opt_False_j5s4\n",
      "Final output: {'module_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'orf_optimization_method': 'zscore_bulk_aa_diff', 'orf_optimization_cub_index': 'tAI', 'initiation_optimization_method': 'external', 'evaluation_score': 'average_distance', 'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': None, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'GAA': 0.6798186496752849, 'AAT': 0.5655317497600131, 'ATA': 0.1287010322849589, 'TTA': 0.198530453110291, 'GAC': 0.35960560391899904, 'CTG': 0.23973441522197322, 'TGG': 1.0, 'AAC': 0.4344682502399869, 'CAA': 0.5131548459124133, 'GCC': 0.20715414603308396, 'CTT': 0.2391679232236727, 'GCT': 0.24752923822568748, 'ATC': 0.366762837741479, 'AAA': 0.7016557540022625, 'AAG': 0.29834424599773746, 'TTG': 0.1592175745811708, 'AGC': 0.22534685577492342, 'CCG': 0.4334091260292722, 'AGT': 0.10611598754150686, 'TTT': 0.6847751982489505, 'GAG': 0.3201813503247151, 'ACT': 0.1614000537105004, 'TCA': 0.23653118484388272, 'ACC': 0.1588786441082565, 'CAC': 0.3265866087018494, 'GGC': 0.3390368073029434, 'GAT': 0.640394396081001, 'ACA': 0.4122430101751559, 'ACG': 0.2674782920060872, 'CCC': 0.0889644363010221, 'AGA': 0.26380392311639106, 'TCC': 0.12712038919920718, 'TAC': 0.34586274736720285, 'CAT': 0.6734133912981506, 'ATT': 0.5045361299735621, 'GCA': 0.2828363712991255, 'TAT': 0.6541372526327971, 'GGG': 0.16211598104043537, 'GTC': 0.25598186049286953, 'CCT': 0.28667299498885185, 'GTT': 0.28439644370189154, 'GCG': 0.26248024444210305, 'TCT': 0.2048572678833432, 'CTC': 0.11219040795754642, 'CCA': 0.19095344268085387, 'GGA': 0.3148341038094681, 'CGA': 0.09855152543710778, 'GTA': 0.19774449549495793, 'GTG': 0.26187720031028106, 'TTC': 0.31522480175104955, 'CGC': 0.2064569601768165, 'CAG': 0.4868451540875866, 'CGG': 0.1555630106168844, 'CGT': 0.18161187196589967, 'CTA': 0.05115922590534585, 'AGG': 0.09401270868690058, 'GGT': 0.18401310784715313, 'TAG': 0.14132712456344587, 'TGC': 0.5460141582025239, 'TCG': 0.10002831475713661, 'TAA': 0.630500582072177, 'TGA': 0.2281722933643772, 'TGT': 0.45398584179747614}), 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.6666666666666666, 'AAC': 0.6666666666666666, 'AAG': 0.2133333333333333, 'AAT': 0.39333333333333337, 'ACA': 0.6666666666666666, 'ACC': 0.16666666666666666, 'ACG': 0.2133333333333333, 'ACT': 0.09833333333333334, 'AGA': 0.18499999999999997, 'AGC': 0.3333333333333333, 'AGG': 0.21999999999999997, 'AGT': 0.19666666666666668, 'ATA': 0.10999999999999999, 'ATC': 0.5, 'ATG': 1.0, 'ATT': 0.29500000000000004, 'CAA': 0.6666666666666666, 'CAC': 0.3333333333333333, 'CAG': 0.2133333333333333, 'CAT': 0.19666666666666668, 'CCA': 0.5, 'CCG': 0.15999999999999998, 'CGA': 0.01839999999999999, 'CGC': 0.48, 'CGG': 0.16666666666666666, 'CGT': 0.6666666666666666, 'CTA': 0.3516666666666666, 'CTC': 0.16666666666666666, 'CTG': 0.2733333333333333, 'CTT': 0.09833333333333334, 'GAA': 1.0, 'GAC': 0.6666666666666666, 'GAG': 0.31999999999999995, 'GAT': 0.39333333333333337, 'GCA': 0.8333333333333334, 'GCC': 0.16666666666666666, 'GCG': 0.2666666666666666, 'GCT': 0.09833333333333334, 'GGA': 0.5, 'GGC': 0.6666666666666666, 'GGG': 0.15999999999999998, 'GGT': 0.39333333333333337, 'GTA': 0.6666666666666666, 'GTC': 0.16666666666666666, 'GTG': 0.2133333333333333, 'GTT': 0.09833333333333334, 'TAC': 0.3333333333333333, 'TAT': 0.19666666666666668, 'TCA': 0.3333333333333333, 'TCC': 0.16666666666666666, 'TCG': 0.10666666666666665, 'TCT': 0.09833333333333334, 'TGC': 0.16666666666666666, 'TGG': 0.16666666666666666, 'TGT': 0.09833333333333334, 'TTA': 0.5183333333333333, 'TTC': 0.5, 'TTG': 0.32666666666666666, 'TTT': 0.29500000000000004}, 'tai_avg': 0.3271366648197227, 'tai_std': 0.028076995641981673}, {'name': 'Escherichia coli', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': None, 'codon_frequencies': defaultdict(<class 'int'>, {'ATG': 1.0, 'AAA': 0.7667390608953314, 'CGC': 0.40076028262096014, 'ATT': 0.5085852538544149, 'AGC': 0.27799163607995603, 'ACC': 0.4359429381095841, 'ATC': 0.4205725148436632, 'ACA': 0.13023262212317555, 'GGT': 0.33726003930064996, 'AAC': 0.5512262248110238, 'GCG': 0.3563679190097728, 'GGC': 0.4050486219579785, 'TGA': 0.28926761860893596, 'CGA': 0.06388523224887838, 'GTG': 0.3715940695510341, 'TTG': 0.12808970030428077, 'AAG': 0.23326093910466858, 'TTC': 0.4262385670731707, 'TCA': 0.12251736088908215, 'GCA': 0.21281773299119433, 'AAT': 0.4487737751889762, 'GAA': 0.689467997530356, 'CGT': 0.38120248233619, 'TTT': 0.5737614329268292, 'CTG': 0.4975546385450099, 'GTT': 0.2584847661069405, 'GCC': 0.26990242799092135, 'GAT': 0.626856640686903, 'AGG': 0.020162265266098916, 'CAG': 0.6530527088827319, 'GGG': 0.15057187484254547, 'GTC': 0.21619163129169194, 'CTC': 0.10435163186194824, 'TCT': 0.14544780223293646, 'CCC': 0.12361073915372255, 'CAC': 0.4292446655321377, 'GCT': 0.1609119200081115, 'TTA': 0.13009739763522432, 'CTT': 0.10341378016756284, 'ACG': 0.26849903016796667, 'GGA': 0.10711946389882603, 'CCG': 0.527785216925549, 'CAA': 0.3469472911172681, 'ACT': 0.16532540959927366, 'ATA': 0.07084223130192197, 'CAT': 0.5707553344678623, 'AGT': 0.15102374892893225, 'TGC': 0.55651396076928, 'GAG': 0.31053200246964396, 'TCG': 0.15406749964830613, 'GTA': 0.15372953305033354, 'TAC': 0.4315172054223149, 'TCC': 0.14895195222078703, 'GAC': 0.37314335931309695, 'TGT': 0.4434860392307201, 'TGG': 1.0, 'TAT': 0.5684827945776851, 'CCT': 0.15789033208355652, 'CCA': 0.19071371183717192, 'CGG': 0.09749348521075679, 'CTA': 0.03649285148597391, 'TAA': 0.6391064025794565, 'AGA': 0.036496252317115764, 'TAG': 0.07162597881160755}), 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.75, 'AAC': 0.5, 'AAG': 0.23999999999999996, 'AAT': 0.29500000000000004, 'ACA': 0.1525, 'ACC': 0.25, 'ACG': 0.29, 'ACT': 0.14750000000000002, 'AGA': 0.13874999999999998, 'AGC': 0.125, 'AGG': 0.16499999999999998, 'AGT': 0.07375000000000001, 'ATA': 0.10999999999999999, 'ATC': 0.375, 'ATG': 1.0, 'ATT': 0.22125000000000003, 'CAA': 0.27749999999999997, 'CAC': 0.125, 'CAG': 0.32999999999999996, 'CAT': 0.07375000000000001, 'CCA': 0.13874999999999998, 'CCC': 0.125, 'CCG': 0.16499999999999998, 'CCT': 0.07375000000000001, 'CGA': 0.013799999999999993, 'CGC': 0.36, 'CGG': 0.125, 'CGT': 0.5, 'CTA': 0.18, 'CTC': 0.125, 'CTG': 0.54, 'CTT': 0.07375000000000001, 'GAA': 0.5, 'GAC': 0.375, 'GAG': 0.15999999999999998, 'GAT': 0.22125000000000003, 'GCA': 0.375, 'GCC': 0.25, 'GCG': 0.11999999999999998, 'GCT': 0.14750000000000002, 'GGA': 0.13874999999999998, 'GGC': 0.5, 'GGG': 0.16499999999999998, 'GGT': 0.29500000000000004, 'GTA': 0.625, 'GTC': 0.25, 'GTG': 0.19999999999999996, 'GTT': 0.14750000000000002, 'TAC': 0.375, 'TAT': 0.22125000000000003, 'TCA': 0.13874999999999998, 'TCC': 0.25, 'TCG': 0.16499999999999998, 'TCT': 0.14750000000000002, 'TGC': 0.125, 'TGG': 0.125, 'TGT': 0.07375000000000001, 'TTA': 0.13874999999999998, 'TTC': 0.25, 'TTG': 0.16499999999999998, 'TTT': 0.14750000000000002}, 'tai_avg': 0.24958213077316901, 'tai_std': 0.024913119712427708}]}, 'orf': [{'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCTCAGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCA', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'initial_sequence_optimization_score': -0.46828435194706775, 'final_sequence_optimization_score': 3.4366446478485297, 'run_time': 21.35999999998603, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 3.4117011749280888}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCA', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 3.4366446478485297}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCA', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 3.4366446478485297}]}, {'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGCTTTAAAGTACACATGGAAGGCTCCGTGAATGGGCACGAGTTCGAAATTGAGGGGGAGGGGGAAGGGCGTCCTTACGAAGGAACACAGACAGCTAAGCTCAAAGTGACAAAAGGAGGCCCTTTACCGTTCGCATGGGATATTCTCAGTCCCCAATTCATGTATGGCTCAAAGGCCTATGTCAAACATCCGGCCGATATCCCTGATTATCTTAAATTGAGTTTCCCCGAGGGTTTTAAATGGGAGCGTGTCATGAACTTTGAGGATGGCGGAGTTGTTACTGTCACGCAAGATAGCTCACTACAAGATGGCGAGTTCATATATAAGGTTAAGTTACGTGGCACCAATTTCCCGTCCGATGGGCCCGTGATGCAAAAGAAGACCATGGGGTGGGAGGCTAGTAGTGAAAGGATGTACCCTGAGGATGGAGCGTTGAAAGGTGAGATTAAACAACGTTTAAAGCTGAAAGACGGTGGCCATTATGATGCTGAGGTAAAAACGACCTACAAGGCCAAGAAACCGGTGCAACTTCCTGGCGCCTACAACGTGAATATCAAACTGGACATTACTTCCCACAACGAGGATTATACCATCGTCGAGCAATATGAGCGTGCTGAGGGAAGGCACTCCACAGGCGGGATGGACGAGCTTTATAAGTGA', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCTCAGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATGA', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCA', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'initial_sequence_optimization_score': 0.3508372096937051, 'final_sequence_optimization_score': 3.4366446478485297, 'run_time': 20.030999999959022, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 3.4272388564599634}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCA', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 3.4366446478485297}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCA', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 3.4366446478485297}]}, {'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTTATGCGGTTTAAGGTCCATATGGAGGGTTCGGTAAACGGACATGAGTTTGAAATCGAAGGCGAAGGTGAAGGAAGGCCGTACGAGGGAACGCAGACTGCAAAGCTAAAGGTTACAAAAGGCGGCCCTTTACCTTTCGCCTGGGATATATTGTCTCCCCAATTCATGTACGGTTCCAAAGCCTACGTAAAACATCCAGCCGATATCCCGGATTACCTAAAGTTAAGCTTCCCTGAGGGATTCAAGTGGGAGAGGGTAATGAACTTTGAGGACGGGGGTGTGGTCACAGTGACCCAGGATAGTAGTTTACAAGACGGTGAATTTATTTACAAGGTAAAACTACGAGGGACCAACTTCCCGTCGGACGGTCCTGTGATGCAGAAAAAGACGATGGGTTGGGAGGCTAGTTCTGAAAGGATGTACCCTGAAGATGGGGCTCTCAAAGGCGAAATTAAACAGAGACTGAAACTCAAGGATGGGGGACATTATGACGCTGAGGTTAAGACCACCTACAAGGCAAAGAAGCCGGTTCAGCTACCGGGTGCCTACAACGTCAATATAAAGCTGGACATTACATCCCACAATGAAGATTATACTATAGTAGAACAATACGAAAGAGCCGAGGGGCGGCATAGCACTGGCGGGATGGACGAACTTTACAAATAA', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCTCAGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCA', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'initial_sequence_optimization_score': 0.10312242758445866, 'final_sequence_optimization_score': 3.4366446478485297, 'run_time': 23.859000000054948, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 3.4272388564599634}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCA', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 3.4366446478485297}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCA', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 3.4366446478485297}]}, {'initial_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAGGTTCACATGGAGGGCAGTGTTAATGGGCACGAATTCGAGATAGAAGGCGAAGGAGAAGGACGGCCATACGAGGGCACGCAAACCGCTAAACTAAAGGTCACAAAGGGAGGACCGCTTCCATTCGCATGGGATATTCTTAGCCCTCAATTTATGTATGGGAGCAAAGCATACGTGAAGCACCCAGCGGACATTCCTGACTACTTAAAACTAAGCTTTCCTGAGGGCTTCAAATGGGAACGCGTAATGAACTTTGAGGATGGTGGGGTAGTTACCGTCACTCAAGATTCGTCTCTGCAGGACGGTGAGTTTATTTATAAAGTTAAGCTCCGCGGAACAAATTTTCCTAGCGACGGGCCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCTTCATCGGAACGCATGTATCCAGAGGATGGAGCTCTGAAAGGAGAAATTAAACAGAGGCTTAAACTTAAGGACGGTGGCCATTACGACGCTGAAGTCAAAACCACCTATAAAGCGAAAAAACCCGTTCAATTACCTGGTGCATATAATGTCAATATTAAATTGGATATCACCAGCCACAACGAAGATTACACAATCGTGGAGCAGTATGAACGCGCCGAAGGGCGTCATTCAACGGGCGGGATGGATGAGCTTTACAAATAA', 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCTCAGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCA', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'initial_sequence_optimization_score': 0.6111542405623543, 'final_sequence_optimization_score': 3.4366446478485297, 'run_time': 20.719000000040978, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 3.4272388564599634}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCA', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 3.4366446478485297}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCA', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 3.4366446478485297}]}], 'evaluation': [{'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'tai_initial_score': 0.33696708001007003, 'tai_final_score': 0.5722076091551204, 'dist_score': 8.378408151095439}, {'name': 'Escherichia coli', 'is_wanted': False, 'tai_initial_score': 0.2816376469779175, 'tai_final_score': 0.2958020049648616, 'dist_score': 0.5685501515042423}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCTCAGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA', 'average_distance_score': 3.9049289997955983, 'weakest_link_score': 3.9049289997955983, 'ratio_score': 1.1606489296917233, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCTCAGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA'}, {'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'tai_initial_score': 0.33696708001007003, 'tai_final_score': 0.5722076091551204, 'dist_score': 8.378408151095439}, {'name': 'Escherichia coli', 'is_wanted': False, 'tai_initial_score': 0.2816376469779175, 'tai_final_score': 0.2958020049648616, 'dist_score': 0.5685501515042423}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCTCAGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATGA', 'average_distance_score': 3.9049289997955983, 'weakest_link_score': 3.9049289997955983, 'ratio_score': 1.1606489296917233, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCTCAGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATGA'}, {'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'tai_initial_score': 0.33696708001007003, 'tai_final_score': 0.5722076091551204, 'dist_score': 8.378408151095439}, {'name': 'Escherichia coli', 'is_wanted': False, 'tai_initial_score': 0.2816376469779175, 'tai_final_score': 0.2958020049648616, 'dist_score': 0.5685501515042423}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCTCAGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA', 'average_distance_score': 3.9049289997955983, 'weakest_link_score': 3.9049289997955983, 'ratio_score': 1.1606489296917233, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCTCAGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA'}, {'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'tai_initial_score': 0.33696708001007003, 'tai_final_score': 0.5722076091551204, 'dist_score': 8.378408151095439}, {'name': 'Escherichia coli', 'is_wanted': False, 'tai_initial_score': 0.2816376469779175, 'tai_final_score': 0.2958020049648616, 'dist_score': 0.5685501515042423}], 'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCTCAGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA', 'average_distance_score': 3.9049289997955983, 'weakest_link_score': 3.9049289997955983, 'ratio_score': 1.1606489296917233, 'init_evaluation_sequence': 'TTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_evaluation_sequence': 'TTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCTCAGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA'}], 'final_evaluation': {'final_sequence': 'ATGGTTTCAAAAGGGGAGGAGGACAATATGGCAATTATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGCCCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGCGTAGTGACAGTAACACAAGACAGCTCATTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAGACAATGGGATGGGAAGCAAGCTCAGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGCCACTACGACGCAGAAGTAAAAACAACGTACAAAGCAAAAAAGCCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGCATGGACGAATTATACAAATAA', 'average_distance_score': 3.9049289997955983, 'weakest_link_score': 3.9049289997955983, 'ratio_score': 1.1606489296917233}, 'zip_output_file_path': 'results\\\\mcherry-wanted-constraint-init-external-0.1\\\\tAI_zscore_bulk_aa_diff_ecoli_opt_False_j5s4\\\\communique_results.zip'}\n"
     ]
    }
   ],
   "source": [
    "initiation_optimization_method=\"external\"\n",
    "mcherry_output_path = f\"mcherry-wanted-constraint-init-{initiation_optimization_method}-0.1\"\n",
    "run_all_methods(orf_sequence_file=macherry_file_path, output_path=mcherry_output_path, should_use_protein_abundance=True, should_use_mrna_levels=False, initiation_optimization_method=initiation_optimization_method)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "mcherry_root_dir = fr\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\orf_model_analysis\\results\\{mcherry_output_path}\"\n",
    "mcherry_variants_file = os.path.join(mcherry_root_dir, \"mcherry_variants.fasta\")\n",
    "generate_sequences_fasta_file(mcherry_root_dir)\n",
    "generate_summary(results_directory=mcherry_root_dir)\n",
    "\n",
    "reference_mcherry_protein_seq  = \"MVSKGEEDNMAIIKEFMRFKVHMEGSVNGHEFEIEGEGEGRPYEGTQTAKLKVTKGGPLPFAWDILSPQFMYGSKAYVKHPADIPDYLKLSFPEGFKWERVMNFEDGGVVTVTQDSSLQDGEFIYKVKLRGTNFPSDGPVMQKKTMGWEASSERMYPEDGALKGEIKQRLKLKDGGHYDAEVKTTYKAKKPVQLPGAYNVNIKLDITSHNEDYTIVEQYERAEGRHSTGGMDELYK*\"\n",
    "formatted_reference_mcherry_protein_seq = reference_mcherry_protein_seq.replace(\"*\", \"_\")\n",
    "\n",
    "with open(os.path.join(mcherry_root_dir, \"mcherry_variants.fasta\"), \"r\") as fasta_file:\n",
    "     mcherry_dict = SeqIO.to_dict(SeqIO.parse(fasta_file, \"fasta\"))\n",
    "     for seq_key, seq in mcherry_dict.items():\n",
    "         translated_variant_mcherry_seq = shared_functions_and_vars.translate(str(seq.seq))\n",
    "     \n",
    "         assert translated_variant_mcherry_seq == formatted_reference_mcherry_protein_seq\n",
    "         \n",
    "         seqeunce_str = str(seq.seq)\n",
    "         seqeunce_str_split = [seqeunce_str[i:i+3].upper() for i in range(0, len(seqeunce_str), 3)]\n",
    "         for i in range(15, len(seqeunce_str_split)):\n",
    "            prev_codon = seqeunce_str_split[i-1]\n",
    "            codon = seqeunce_str_split[i]\n",
    "            if prev_codon == codon:\n",
    "                print(f\"codon {codon} is the same as it previous, location {i} in {seq_key}\")\n",
    "     \n",
    "     variant_seqs = [str(seq.seq) for seq in mcherry_dict.values()] \n",
    "     assert len(variant_seqs) == len(set(variant_seqs)) == 16\n",
    "    \n",
    "     variant_seqs_init = [seq[:15*3] for seq in variant_seqs]\n",
    "     assert len(set(variant_seqs_init)) == 1\n",
    "    \n",
    "     "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tuning parameter = 1 (optimizing only wanted host and neglecting unwanted host)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Using protein abundance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run_all_methods(orf_sequence_file=macherry_file_path, output_path=\"mcherry_greater_alpha\", tuning_param=1, should_use_protein_abundance=True, should_use_mrna_levels=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mcherry_greater_alpha_root_dir = r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\orf_model_analysis\\results\\mcherry_greater_alpha\"\n",
    "mcherry_greater_alpha_variants_file = os.path.join(mcherry_greater_alpha_root_dir, \"mcherry_variants.fasta\")\n",
    "generate_sequences_fasta_file(mcherry_greater_alpha_root_dir)\n",
    "generate_summary(results_directory=mcherry_greater_alpha_root_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tuning parameter = 0 (de-optimizing uwanted host and neglecting wanted host)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Using protein abundance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run_all_methods(orf_sequence_file=macherry_file_path, output_path=\"mcherry_lower_alpha\", tuning_param=0, should_use_protein_abundance=True, should_use_mrna_levels=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mcherry_lower_alpha_root_dir = r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\orf_model_analysis\\results\\mcherry_lower_alpha\"\n",
    "mcherry_lower_alpha_variants_file = os.path.join(mcherry_lower_alpha_root_dir, \"mcherry_variants.fasta\")\n",
    "generate_sequences_fasta_file(mcherry_lower_alpha_root_dir)\n",
    "generate_summary(results_directory=mcherry_lower_alpha_root_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Using ribosomal genes as highly expressed genes (instead of mRNA levels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run_all_methods(orf_sequence_file=macherry_file_path, output_path=\"ribosomal\", should_use_mrna_levels=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# mcherry_ribosomal_root_dir = r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\orf_model_analysis\\results\\mrna\\ribosomal\"\n",
    "# mcherry_ribosomal_variants_file = os.path.join(mcherry_ribosomal_root_dir, \"mcherry_variants.fasta\")\n",
    "# generate_sequences_fasta_file(mcherry_ribosomal_root_dir)\n",
    "# generate_summary(results_directory=mcherry_ribosomal_root_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CUB and codon frequencies of Bacillus and E.coli"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bacillus_endogenous_genes_fasta_file_path = r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\Bacillus-subtilis.fasta\"\n",
    "ecoli_endogenous_genes_fasta_file_path =  r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\Escherichia-coli.fasta\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(bacillus_endogenous_genes_fasta_file_path, \"r\") as bacillus_endogenous_genes_fasta_file:\n",
    "    bacillus_endogenous_genes_dict = SeqIO.to_dict(SeqIO.parse(bacillus_endogenous_genes_fasta_file, \"fasta\"), lambda record: record.description)\n",
    "\n",
    "bacillus_codon_frequencies = calculate_codon_frequencies([str(gene.seq) for gene in bacillus_endogenous_genes_dict.values()])\n",
    "bacillus_codon_frequencies = bacillus_codon_frequencies.sort_values(\"codon\")\n",
    "\n",
    "bacillus_codon_frequencies_lamda = calculate_codon_frequencies_lamda([str(gene.seq) for gene in bacillus_endogenous_genes_dict.values()])\n",
    "bacillus_codon_frequencies_lamda = bacillus_codon_frequencies_lamda.sort_values(\"codon\")\n",
    "\n",
    "with open(ecoli_endogenous_genes_fasta_file_path, \"r\") as ecoli_endogenous_genes_fasta_file:\n",
    "    ecoli_endogenous_genes_dict = SeqIO.to_dict(SeqIO.parse(ecoli_endogenous_genes_fasta_file, \"fasta\"), lambda x: x.description)\n",
    "\n",
    "ecoli_codon_frequencies = calculate_codon_frequencies([str(gene.seq) for gene in ecoli_endogenous_genes_dict.values()])\n",
    "ecoli_codon_frequencies = ecoli_codon_frequencies.sort_values(\"codon\")\n",
    "\n",
    "ecoli_codon_frequencies_lamda = calculate_codon_frequencies_lamda([str(gene.seq) for gene in ecoli_endogenous_genes_dict.values()])\n",
    "ecoli_codon_frequencies_lamda = ecoli_codon_frequencies_lamda.sort_values(\"codon\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bacillus_codon_frequencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ecoli_codon_frequencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ecoli_highly_expressed_genes = get_highly_expressed_genes(endogenous_genes_fasta_file_path=ecoli_endogenous_genes_fasta_file_path,\n",
    "                                                          expression_levels_file_path=r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\PA_ecoli.json\",\n",
    "                                                          expression_type=\"protein_abundance\")\n",
    "# ecoli_highly_expressed_genes = get_highly_expressed_genes(endogenous_genes_fasta_file_path=ecoli_endogenous_genes_fasta_file_path,\n",
    "#                                                           expression_levels_file_path=r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\ecoli_mrna_level.csv\",\n",
    "#                                                           expression_type=\"mrna_levels\")\n",
    "ecoli_highly_expressed_genes_codon_frequencies = calculate_codon_frequencies([str(gene[\"cds\"].seq) for gene in ecoli_highly_expressed_genes.values()])\n",
    "ecoli_highly_expressed_genes_codon_frequencies = ecoli_highly_expressed_genes_codon_frequencies.sort_values(\"codon\")\n",
    "\n",
    "ecoli_highly_expressed_genes_codon_frequencies_lamda = calculate_codon_frequencies_lamda([str(gene[\"cds\"].seq) for gene in ecoli_highly_expressed_genes.values()])\n",
    "ecoli_highly_expressed_genes_codon_frequencies_lamda = ecoli_highly_expressed_genes_codon_frequencies_lamda.sort_values(\"codon\")\n",
    "\n",
    "# bacillus_highly_expressed_genes = get_highly_expressed_genes(endogenous_genes_fasta_file_path=bacillus_endogenous_genes_fasta_file_path,\n",
    "#                                                           expression_levels_file_path=r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\bacillus_mrna_level.csv\",\n",
    "#                                                             expression_type=\"mrna_levels\")\n",
    "bacillus_highly_expressed_genes = get_highly_expressed_genes(endogenous_genes_fasta_file_path=bacillus_endogenous_genes_fasta_file_path,\n",
    "                                                             expression_levels_file_path=r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\PA_bacillus.json\",\n",
    "                                                            expression_type=\"protein_abundance\")\n",
    "bacillus_highly_expressed_genes_codon_frequencies = calculate_codon_frequencies([str(gene[\"cds\"].seq) for gene in bacillus_highly_expressed_genes.values()])\n",
    "bacillus_highly_expressed_genes_codon_frequencies = bacillus_highly_expressed_genes_codon_frequencies.sort_values(\"codon\")\n",
    "\n",
    "bacillus_highly_expressed_genes_codon_frequencies_lamda = calculate_codon_frequencies_lamda([str(gene[\"cds\"].seq) for gene in bacillus_highly_expressed_genes.values()])\n",
    "bacillus_highly_expressed_genes_codon_frequencies_lamda = bacillus_highly_expressed_genes_codon_frequencies_lamda.sort_values(\"codon\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run only once - for generating both tai and cai weights dict:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from modules.user_IO.user_input import UserInputModule\n",
    "from modules.run_summary import RunSummary\n",
    "\n",
    "from generate_summary_file import generate_summary\n",
    "from analysis.input_testing_data.generate_input_testing_data_for_modules import generate_testing_data_for_ecoli_and_bacillus\n",
    "from modules.main import run_modules\n",
    "from modules import shared_functions_and_vars\n",
    "\n",
    "default_user_inp_raw = generate_testing_data_for_ecoli_and_bacillus(\n",
    "    orf_optimization_method=\"single_codon_diff\",\n",
    "    orf_optimization_cub_index=\"max_CAI_tAI\",\n",
    "    clusters_count=1,\n",
    "    tuning_param=0.5,\n",
    "    is_ecoli_optimized=True,\n",
    "    sequence=original_mcherry_seq,\n",
    "    output_path=\"results\",\n",
    "    should_use_mrna_levels=False,\n",
    "    should_use_protein_abundance=True,\n",
    ")\n",
    "\n",
    "UserInputModule.run_module(default_user_inp_raw, RunSummary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ecoli_cai_weights, ecoli_tai_weights = get_cub_weights(r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\Escherichia-coli_True_parsed.json\")\n",
    "bacillus_cai_weights, bacillus_tai_weights = get_cub_weights(r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\Bacillus-subtilis_False_parsed.json\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "codons = ecoli_cai_weights.keys()\n",
    "weights_df_dict = {\n",
    "    \"codon\": codons, \n",
    "    \"amino_acid\": [shared_functions_and_vars.nt_to_aa[k] for k in codons],\n",
    "    \"ecoli_cai_weights\": [ecoli_cai_weights[k] for k in codons],\n",
    "    \"ecoli_tai_weights\": [ecoli_tai_weights.get(k, 0) for k in codons],\n",
    "    \"bacillus_cai_weights\": [bacillus_cai_weights[k] for k in codons],\n",
    "    \"bacillus_tai_weights\": [bacillus_tai_weights.get(k, 0) for k in codons],\n",
    "}\n",
    "weights_dataframe = pd.DataFrame(weights_df_dict)\n",
    "weights_dataframe.sort_values(\"amino_acid\")\n",
    "weights_dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(nrows=1,ncols=2, figsize=(20,5))\n",
    "scatter = weights_dataframe.plot.scatter(x=\"ecoli_cai_weights\", y=\"bacillus_cai_weights\", title=\"E.coli CAI weights vs. Bacillus CAI weights\", ax=axes[0])\n",
    "r, p = stats.spearmanr(weights_dataframe[\"ecoli_cai_weights\"], weights_dataframe[\"bacillus_cai_weights\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = weights_dataframe.plot.scatter(x=\"ecoli_tai_weights\", y=\"bacillus_tai_weights\", title=\"E.coli tAI weights vs. Bacillus tAI weights\", ax=axes[1])\n",
    "r, p = stats.spearmanr(weights_dataframe[\"ecoli_tai_weights\"], weights_dataframe[\"bacillus_tai_weights\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_weights_codon_frequencies_df = pd.merge(weights_dataframe, ecoli_codon_frequencies, on='codon', how='inner')\n",
    "merged_weights_codon_frequencies_df = pd.merge(merged_weights_codon_frequencies_df, bacillus_codon_frequencies, on='codon', how='inner', suffixes=(\"_ecoli\", \"_bacillus\"))\n",
    "\n",
    "fig, axes = plt.subplots(nrows=2,ncols=2, figsize=(20,15))\n",
    "scatter = merged_weights_codon_frequencies_df.plot.scatter(x=\"ecoli_cai_weights\", y=\"frequency_ecoli\", title=\"E.coli CAI weights vs. E.coli codon frequencies\", ax=axes[0,0])\n",
    "r, p = stats.spearmanr(merged_weights_codon_frequencies_df[\"ecoli_cai_weights\"], merged_weights_codon_frequencies_df[\"frequency_ecoli\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_weights_codon_frequencies_df.plot.scatter(x=\"bacillus_cai_weights\", y=\"frequency_bacillus\", title=\"Bacillus CAI weights vs. Bacillus codon frequencies\", ax=axes[0,1])\n",
    "r, p = stats.spearmanr(merged_weights_codon_frequencies_df[\"ecoli_cai_weights\"], merged_weights_codon_frequencies_df[\"frequency_bacillus\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_weights_codon_frequencies_df.plot.scatter(x=\"ecoli_tai_weights\", y=\"frequency_ecoli\", title=\"E.coli tAI weights vs. E.coli codon frequencies\", ax=axes[1,0])\n",
    "r, p = stats.spearmanr(merged_weights_codon_frequencies_df[\"ecoli_tai_weights\"], merged_weights_codon_frequencies_df[\"frequency_ecoli\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_weights_codon_frequencies_df.plot.scatter(x=\"bacillus_tai_weights\", y=\"frequency_bacillus\", title=\"Bacillus tAI weights vs. Bacillus codon frequencies\", ax=axes[1,1])\n",
    "r, p = stats.spearmanr(merged_weights_codon_frequencies_df[\"ecoli_tai_weights\"], merged_weights_codon_frequencies_df[\"frequency_bacillus\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_weights_codon_frequencies_df = pd.merge(weights_dataframe, ecoli_highly_expressed_genes_codon_frequencies, on='codon', how='inner')\n",
    "merged_weights_codon_frequencies_df = pd.merge(merged_weights_codon_frequencies_df, bacillus_highly_expressed_genes_codon_frequencies, on='codon', how='inner', suffixes=(\"_ecoli\", \"_bacillus\"))\n",
    "\n",
    "fig, axes = plt.subplots(nrows=2,ncols=2, figsize=(20,15))\n",
    "scatter = merged_weights_codon_frequencies_df.plot.scatter(x=\"ecoli_cai_weights\", y=\"frequency_ecoli\", title=\"E.coli CAI weights vs. E.coli codon frequencies in highly expressed genes\", ax=axes[0,0])\n",
    "r, p = stats.spearmanr(merged_weights_codon_frequencies_df[\"ecoli_cai_weights\"], merged_weights_codon_frequencies_df[\"frequency_ecoli\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_weights_codon_frequencies_df.plot.scatter(x=\"bacillus_cai_weights\", y=\"frequency_bacillus\", title=\"Bacillus CAI weights vs. Bacillus codon frequencies in highly expressed genes\", ax=axes[0,1])\n",
    "r, p = stats.spearmanr(merged_weights_codon_frequencies_df[\"ecoli_cai_weights\"], merged_weights_codon_frequencies_df[\"frequency_bacillus\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_weights_codon_frequencies_df.plot.scatter(x=\"ecoli_tai_weights\", y=\"frequency_ecoli\", title=\"E.coli tAI weights vs. E.coli codon frequencies in highly expressed genes\", ax=axes[1,0])\n",
    "r, p = stats.spearmanr(merged_weights_codon_frequencies_df[\"ecoli_tai_weights\"], merged_weights_codon_frequencies_df[\"frequency_ecoli\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_weights_codon_frequencies_df.plot.scatter(x=\"bacillus_tai_weights\", y=\"frequency_bacillus\", title=\"Bacillus tAI weights vs. Bacillus codon frequencies in highly expressed genes\", ax=axes[1,1])\n",
    "r, p = stats.spearmanr(merged_weights_codon_frequencies_df[\"ecoli_tai_weights\"], merged_weights_codon_frequencies_df[\"frequency_bacillus\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ecoli_cai_scores, ecoli_tai_scores = get_cub_scores(r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\Escherichia-coli_True_parsed.json\")\n",
    "bacillus_cai_scores, bacillus_tai_scores = get_cub_scores(r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\Bacillus-subtilis_False_parsed.json\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_scores(estimated_expression_dict, cai_scores_dict, tai_scores_dict, expression_type, selected_genes=None):\n",
    "    all_cds = cai_scores_dict.keys()\n",
    "    selected_genes = selected_genes or all_cds\n",
    "    filtered_cds = []\n",
    "    expression_levels = []\n",
    "    \n",
    "    expression_levels_key = \"mrna_level\" if expression_type == \"mrna_levels\" else \"pa\"\n",
    "    for k in all_cds:\n",
    "        if not any(k.startswith(x) for x in selected_genes):\n",
    "            continue\n",
    "        short_name = k.split(\" \")[0]\n",
    "        # print(f\"k: {k}, sort_name: {short_name}\")\n",
    "        for gene_name, expression in estimated_expression_dict.items():\n",
    "            if gene_name.startswith(short_name):\n",
    "                expression_levels.append(expression[expression_levels_key])\n",
    "                filtered_cds.append(k)\n",
    "                break\n",
    "\n",
    "        \n",
    "    scores_expression_levels_df_dict = {\n",
    "        \"gene\": filtered_cds, \n",
    "        \"cai_scores\": [cai_scores_dict[k] for k in filtered_cds],\n",
    "        \"tai_scores\": [tai_scores_dict[k] for k in filtered_cds],\n",
    "        expression_levels_key: expression_levels,\n",
    "    }\n",
    "    return pd.DataFrame(scores_expression_levels_df_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_type = \"mrna_levels\"\n",
    "ecoli_estimated_expression = get_estimated_expression(endogenous_genes_fasta_file_path=ecoli_endogenous_genes_fasta_file_path,\n",
    "                                                      expression_levels_file_path=r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\ecoli_mrna_level.csv\",\n",
    "                                                      expression_type=expression_type)\n",
    "ecoli_scores_dataframe =  get_scores(estimated_expression_dict=ecoli_estimated_expression, cai_scores_dict=ecoli_cai_scores, tai_scores_dict=ecoli_tai_scores, expression_type=expression_type)\n",
    "print(f\"Ecoli mrna levels stats:\\n records:{len(ecoli_scores_dataframe['mrna_level'])}\\n min: {ecoli_scores_dataframe['mrna_level'].min()} \\n max: {ecoli_scores_dataframe['mrna_level'].max()}\\n mean: {ecoli_scores_dataframe['mrna_level'].mean()}\\n std: {ecoli_scores_dataframe['mrna_level'].std()}\")  \n",
    "ecoli_scores_dataframe[\"mrna_level\"].sort_values(ascending=False).reset_index(drop=True).plot()\n",
    "\n",
    "bacillus_estimated_expression = get_estimated_expression(endogenous_genes_fasta_file_path=bacillus_endogenous_genes_fasta_file_path,\n",
    "                                                         expression_levels_file_path=r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\bacillus_mrna_level.csv\",\n",
    "                                                          expression_type=expression_type)\n",
    "bacillus_scores_dataframe =  get_scores(estimated_expression_dict=bacillus_estimated_expression, cai_scores_dict=bacillus_cai_scores, tai_scores_dict=bacillus_tai_scores, expression_type=expression_type)\n",
    "print(f\"Bacillus mrna levels stats:\\n records:{len(bacillus_scores_dataframe['mrna_level'])}\\n min: {bacillus_scores_dataframe['mrna_level'].min()} \\n max: {bacillus_scores_dataframe['mrna_level'].max()}\\n mean: {bacillus_scores_dataframe['mrna_level'].mean()}\\n std: {bacillus_scores_dataframe['mrna_level'].std()}\")  \n",
    "bacillus_scores_dataframe[\"mrna_level\"].sort_values(ascending=False).reset_index(drop=True).plot()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(nrows=2,ncols=2, figsize=(20,15))\n",
    "\n",
    "scatter = ecoli_scores_dataframe.plot.scatter(x=\"cai_scores\", y=\"mrna_level\", title=\"E.coli mRNA levels vs. E.coli CAI scores\", ax=axes[0,0])\n",
    "r, p = stats.spearmanr(ecoli_scores_dataframe[\"cai_scores\"], ecoli_scores_dataframe[\"mrna_level\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = ecoli_scores_dataframe.plot.scatter(x=\"tai_scores\", y=\"mrna_level\", title=\"E.coli mRNA levels vs. E.coli tAI scores\", ax=axes[0,1])\n",
    "r, p = stats.spearmanr(ecoli_scores_dataframe[\"tai_scores\"], ecoli_scores_dataframe[\"mrna_level\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = bacillus_scores_dataframe.plot.scatter(x=\"cai_scores\", y=\"mrna_level\", title=\"Baciilus mRNA levels vs. Baciilus CAI scores\", ax=axes[1,0])\n",
    "r, p = stats.spearmanr(bacillus_scores_dataframe[\"cai_scores\"], bacillus_scores_dataframe[\"mrna_level\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = bacillus_scores_dataframe.plot.scatter(x=\"tai_scores\", y=\"mrna_level\", title=\"Baciilus mRNA levels vs. Baciilus tAI scores\", ax=axes[1,1])\n",
    "r, p = stats.spearmanr(bacillus_scores_dataframe[\"tai_scores\"], bacillus_scores_dataframe[\"mrna_level\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_type = \"mrna_levels\"\n",
    "ecoli_estimated_expression = get_estimated_expression(endogenous_genes_fasta_file_path=ecoli_endogenous_genes_fasta_file_path,\n",
    "                                                      expression_levels_file_path=r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\ecoli_mrna_level.csv\",\n",
    "                                                      expression_type=expression_type)\n",
    "ecoli_highly_expressed_scores_dataframe =  get_scores(estimated_expression_dict=ecoli_estimated_expression, cai_scores_dict=ecoli_cai_scores, tai_scores_dict=ecoli_tai_scores, selected_genes=ecoli_highly_expressed_genes.keys(), expression_type=expression_type)\n",
    "print(f\"Ecoli mrna levels stats:\\n records:{len(ecoli_highly_expressed_scores_dataframe['mrna_level'])}\\n min: {ecoli_highly_expressed_scores_dataframe['mrna_level'].min()} \\n max: {ecoli_highly_expressed_scores_dataframe['mrna_level'].max()}\\n mean: {ecoli_highly_expressed_scores_dataframe['mrna_level'].mean()}\\n std: {ecoli_highly_expressed_scores_dataframe['mrna_level'].std()}\")  \n",
    "ecoli_highly_expressed_scores_dataframe[\"mrna_level\"].sort_values(ascending=False).reset_index(drop=True).plot()\n",
    "bacillus_estimated_expression = get_estimated_expression(endogenous_genes_fasta_file_path=bacillus_endogenous_genes_fasta_file_path,\n",
    "                                                      expression_levels_file_path=r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\bacillus_mrna_level.csv\",\n",
    "                                                        expression_type=expression_type)\n",
    "bacillus_highly_expressed_scores_dataframe =  get_scores(estimated_expression_dict=bacillus_estimated_expression, cai_scores_dict=bacillus_cai_scores, tai_scores_dict=bacillus_tai_scores, selected_genes=bacillus_highly_expressed_genes.keys(), expression_type=expression_type)\n",
    "print(f\"Bacillus mrna levels stats:\\n records:{len(bacillus_highly_expressed_scores_dataframe['mrna_level'])}\\n min: {bacillus_highly_expressed_scores_dataframe['mrna_level'].min()} \\n max: {bacillus_highly_expressed_scores_dataframe['mrna_level'].max()}\\n mean: {bacillus_highly_expressed_scores_dataframe['mrna_level'].mean()}\\n std: {bacillus_highly_expressed_scores_dataframe['mrna_level'].std()}\")  \n",
    "bacillus_highly_expressed_scores_dataframe[\"mrna_level\"].sort_values(ascending=False).reset_index(drop=True).plot()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(nrows=2,ncols=2, figsize=(20,15))\n",
    "\n",
    "scatter = ecoli_highly_expressed_scores_dataframe.plot.scatter(x=\"cai_scores\", y=\"mrna_level\", title=\"E.coli mRNA levels (Highly expressed) vs. E.coli CAI scores\", ax=axes[0,0])\n",
    "r, p = stats.spearmanr(ecoli_highly_expressed_scores_dataframe[\"cai_scores\"], ecoli_highly_expressed_scores_dataframe[\"mrna_level\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = ecoli_highly_expressed_scores_dataframe.plot.scatter(x=\"tai_scores\", y=\"mrna_level\", title=\"E.coli mRNA levels (Highly expressed) vs. E.coli tAI scores\", ax=axes[0,1])\n",
    "r, p = stats.spearmanr(ecoli_highly_expressed_scores_dataframe[\"tai_scores\"], ecoli_highly_expressed_scores_dataframe[\"mrna_level\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = bacillus_highly_expressed_scores_dataframe.plot.scatter(x=\"cai_scores\", y=\"mrna_level\", title=\"Baciilus mRNA levels (Highly expressed) vs. Baciilus CAI scores\", ax=axes[1,0])\n",
    "r, p = stats.spearmanr(bacillus_highly_expressed_scores_dataframe[\"cai_scores\"], bacillus_highly_expressed_scores_dataframe[\"mrna_level\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = bacillus_highly_expressed_scores_dataframe.plot.scatter(x=\"tai_scores\", y=\"mrna_level\", title=\"Baciilus mRNA levels (Highly expressed) vs. Baciilus tAI scores\", ax=axes[1,1])\n",
    "r, p = stats.spearmanr(bacillus_highly_expressed_scores_dataframe[\"tai_scores\"], bacillus_highly_expressed_scores_dataframe[\"mrna_level\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_type = \"protein_abundance\"\n",
    "ecoli_estimated_expression_pa = get_estimated_expression(endogenous_genes_fasta_file_path=ecoli_endogenous_genes_fasta_file_path,\n",
    "                                                         expression_levels_file_path=r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\PA_ecoli.json\",\n",
    "                                                         expression_type=expression_type)\n",
    "ecoli_scores_dataframe_pa =  get_scores(estimated_expression_dict=ecoli_estimated_expression_pa, cai_scores_dict=ecoli_cai_scores, tai_scores_dict=ecoli_tai_scores, expression_type=expression_type)\n",
    "print(f\"Ecoli protein abundance stats:\\n records:{len(ecoli_scores_dataframe_pa['pa'])}\\n min: {ecoli_scores_dataframe_pa['pa'].min()} \\n max: {ecoli_scores_dataframe_pa['pa'].max()}\\n mean: {ecoli_scores_dataframe_pa['pa'].mean()}\\n std: {ecoli_scores_dataframe_pa['pa'].std()}\")  \n",
    "ecoli_scores_dataframe_pa[\"pa\"].sort_values(ascending=False).reset_index(drop=True).plot()\n",
    "\n",
    "bacillus_estimated_expression_pa = get_estimated_expression(endogenous_genes_fasta_file_path=bacillus_endogenous_genes_fasta_file_path,\n",
    "                                                         expression_levels_file_path=r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\PA_bacillus.json\",\n",
    "                                                          expression_type=expression_type)\n",
    "bacillus_scores_dataframe_pa =  get_scores(estimated_expression_dict=bacillus_estimated_expression_pa, cai_scores_dict=bacillus_cai_scores, tai_scores_dict=bacillus_tai_scores, expression_type=expression_type)\n",
    "print(f\"Bacillus protein abundance stats:\\n records:{len(bacillus_scores_dataframe_pa['pa'])}\\n min: {bacillus_scores_dataframe_pa['pa'].min()} \\n max: {bacillus_scores_dataframe_pa['pa'].max()}\\n mean: {bacillus_scores_dataframe_pa['pa'].mean()}\\n std: {bacillus_scores_dataframe_pa['pa'].std()}\")  \n",
    "bacillus_scores_dataframe_pa[\"pa\"].sort_values(ascending=False).reset_index(drop=True).plot()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(nrows=2,ncols=2, figsize=(20,15))\n",
    "\n",
    "ecoli_scores_dataframe_pa[\"pa_logarithmic_base10\"] = np.log10(ecoli_scores_dataframe_pa['pa'])\n",
    "bacillus_scores_dataframe_pa[\"pa_logarithmic_base10\"] = np.log10(bacillus_scores_dataframe_pa['pa'])\n",
    "                                                              \n",
    "\n",
    "scatter = ecoli_scores_dataframe_pa.plot.scatter(x=\"cai_scores\", y=\"pa_logarithmic_base10\", title=\"E.coli protein abundance vs. E.coli CAI scores\", ax=axes[0,0])\n",
    "r, p = stats.spearmanr(ecoli_scores_dataframe_pa[\"cai_scores\"], ecoli_scores_dataframe_pa[\"pa\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = ecoli_scores_dataframe_pa.plot.scatter(x=\"tai_scores\", y=\"pa_logarithmic_base10\", title=\"E.coli protein abundance vs. E.coli tAI scores\", ax=axes[0,1])\n",
    "r, p = stats.spearmanr(ecoli_scores_dataframe_pa[\"tai_scores\"], ecoli_scores_dataframe_pa[\"pa\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = bacillus_scores_dataframe_pa.plot.scatter(x=\"cai_scores\", y=\"pa_logarithmic_base10\", title=\"Baciilus protein abundance vs. Baciilus CAI scores\", ax=axes[1,0])\n",
    "r, p = stats.spearmanr(bacillus_scores_dataframe_pa[\"cai_scores\"], bacillus_scores_dataframe_pa[\"pa\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = bacillus_scores_dataframe_pa.plot.scatter(x=\"tai_scores\", y=\"pa_logarithmic_base10\", title=\"Baciilus protein abundance vs. Baciilus tAI scores\", ax=axes[1,1])\n",
    "r, p = stats.spearmanr(bacillus_scores_dataframe_pa[\"tai_scores\"], bacillus_scores_dataframe_pa[\"pa\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_type = \"protein_abundance\"\n",
    "ecoli_estimated_expression_pa = get_estimated_expression(endogenous_genes_fasta_file_path=ecoli_endogenous_genes_fasta_file_path,\n",
    "                                                         expression_levels_file_path=r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\PA_ecoli.json\",\n",
    "                                                         expression_type=expression_type)\n",
    "ecoli_scores_dataframe_pa_highly_expressed =  get_scores(estimated_expression_dict=ecoli_estimated_expression_pa, cai_scores_dict=ecoli_cai_scores, tai_scores_dict=ecoli_tai_scores, expression_type=expression_type, selected_genes=ecoli_highly_expressed_genes.keys())\n",
    "print(f\"Ecoli protein abundance stats:\\n records:{len(ecoli_scores_dataframe_pa_highly_expressed['pa'])}\\n min: {ecoli_scores_dataframe_pa_highly_expressed['pa'].min()} \\n max: {ecoli_scores_dataframe_pa_highly_expressed['pa'].max()}\\n mean: {ecoli_scores_dataframe_pa_highly_expressed['pa'].mean()}\\n std: {ecoli_scores_dataframe_pa_highly_expressed['pa'].std()}\")  \n",
    "ecoli_scores_dataframe_pa_highly_expressed[\"pa\"].sort_values(ascending=False).reset_index(drop=True).plot()\n",
    "\n",
    "bacillus_estimated_expression_pa = get_estimated_expression(endogenous_genes_fasta_file_path=bacillus_endogenous_genes_fasta_file_path,\n",
    "                                                         expression_levels_file_path=r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\PA_bacillus.json\",\n",
    "                                                          expression_type=expression_type)\n",
    "bacillus_scores_dataframe_pa_highly_expressed =  get_scores(estimated_expression_dict=bacillus_estimated_expression_pa, cai_scores_dict=bacillus_cai_scores, tai_scores_dict=bacillus_tai_scores, expression_type=expression_type, selected_genes=bacillus_highly_expressed_genes.keys())\n",
    "print(f\"Bacillus protein abundance stats:\\n records:{len(bacillus_scores_dataframe_pa_highly_expressed['pa'])}\\n min: {bacillus_scores_dataframe_pa_highly_expressed['pa'].min()} \\n max: {bacillus_scores_dataframe_pa_highly_expressed['pa'].max()}\\n mean: {bacillus_scores_dataframe_pa_highly_expressed['pa'].mean()}\\n std: {bacillus_scores_dataframe_pa_highly_expressed['pa'].std()}\")  \n",
    "bacillus_scores_dataframe_pa_highly_expressed[\"pa\"].sort_values(ascending=False).reset_index(drop=True).plot()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(nrows=2,ncols=2, figsize=(20,15))\n",
    "\n",
    "ecoli_scores_dataframe_pa_highly_expressed[\"pa_logarithmic_base10\"] = np.log10(ecoli_scores_dataframe_pa_highly_expressed['pa'])\n",
    "bacillus_scores_dataframe_pa_highly_expressed[\"pa_logarithmic_base10\"] = np.log10(bacillus_scores_dataframe_pa_highly_expressed['pa'])\n",
    "\n",
    "scatter = ecoli_scores_dataframe_pa_highly_expressed.plot.scatter(x=\"cai_scores\", y=\"pa_logarithmic_base10\", title=\"E.coli protein abundance vs. E.coli CAI scores - Highly expressed\", ax=axes[0,0])\n",
    "r, p = stats.spearmanr(ecoli_scores_dataframe_pa_highly_expressed[\"cai_scores\"], ecoli_scores_dataframe_pa_highly_expressed[\"pa\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = ecoli_scores_dataframe_pa_highly_expressed.plot.scatter(x=\"tai_scores\", y=\"pa_logarithmic_base10\", title=\"E.coli protein abundance vs. E.coli tAI scores - Highly expressed\", ax=axes[0,1])\n",
    "r, p = stats.spearmanr(ecoli_scores_dataframe_pa_highly_expressed[\"tai_scores\"], ecoli_scores_dataframe_pa_highly_expressed[\"pa\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = bacillus_scores_dataframe_pa_highly_expressed.plot.scatter(x=\"cai_scores\", y=\"pa_logarithmic_base10\", title=\"Baciilus protein abundance vs. Baciilus CAI scores - Highly expressed \", ax=axes[1,0])\n",
    "r, p = stats.spearmanr(bacillus_scores_dataframe_pa_highly_expressed[\"cai_scores\"], bacillus_scores_dataframe_pa_highly_expressed[\"pa\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = bacillus_scores_dataframe_pa.plot.scatter(x=\"tai_scores\", y=\"pa_logarithmic_base10\", title=\"Baciilus protein abundance vs. Baciilus tAI scores - Highly expressed\", ax=axes[1,1])\n",
    "r, p = stats.spearmanr(bacillus_scores_dataframe_pa_highly_expressed[\"tai_scores\"], bacillus_scores_dataframe_pa_highly_expressed[\"pa\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Verification of variants translation to correct amino acid sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Validating original mcherry sequence:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "aa E is the same as it previous, location 6\n",
      "aa I is the same as it previous, location 12\n",
      "aa G is the same as it previous, location 56\n",
      "aa G is the same as it previous, location 107\n",
      "aa V is the same as it previous, location 109\n",
      "aa S is the same as it previous, location 116\n",
      "aa K is the same as it previous, location 143\n",
      "aa S is the same as it previous, location 151\n",
      "aa G is the same as it previous, location 175\n",
      "aa T is the same as it previous, location 184\n",
      "aa K is the same as it previous, location 189\n",
      "aa G is the same as it previous, location 229\n"
     ]
    }
   ],
   "source": [
    "# Reference aa sequence is taken from: https://www.addgene.org/browse/sequence_vdb/6496/\n",
    "reference_mcherry_protein_seq  = \"MVSKGEEDNMAIIKEFMRFKVHMEGSVNGHEFEIEGEGEGRPYEGTQTAKLKVTKGGPLPFAWDILSPQFMYGSKAYVKHPADIPDYLKLSFPEGFKWERVMNFEDGGVVTVTQDSSLQDGEFIYKVKLRGTNFPSDGPVMQKKTMGWEASSERMYPEDGALKGEIKQRLKLKDGGHYDAEVKTTYKAKKPVQLPGAYNVNIKLDITSHNEDYTIVEQYERAEGRHSTGGMDELYK*\"\n",
    "formatted_reference_mcherry_protein_seq = reference_mcherry_protein_seq.replace(\"*\", \"_\")\n",
    "\n",
    "for i in range(1, len(formatted_reference_mcherry_protein_seq)):\n",
    "    prev_aa = formatted_reference_mcherry_protein_seq[i-1]\n",
    "    aa = formatted_reference_mcherry_protein_seq[i]\n",
    "    if prev_aa == aa:\n",
    "        print(f\"aa {aa} is the same as it previous, location {i}\")\n",
    "\n",
    "translated_original_mcherry_seq = shared_functions_and_vars.translate(original_mcherry_seq)\n",
    "     \n",
    "assert translated_original_mcherry_seq == formatted_reference_mcherry_protein_seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Validating generated mcherry variants:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(mcherry_variants_file, \"r\") as mcherry_variants_fasta_file:\n",
    "    mcherry_variants_dict = SeqIO.to_dict(SeqIO.parse(mcherry_variants_fasta_file, \"fasta\"))\n",
    "\n",
    "for variant_name, variant_seq in mcherry_variants_dict.items():\n",
    "    translated_seq = shared_functions_and_vars.translate(variant_seq.seq)\n",
    "    if translated_seq != formatted_reference_mcherry_protein_seq:\n",
    "        raise ValueError(f\"Variant {variant_name} is not coding to correct sequence. \\nReference sequence: {formatted_reference_mcherry_protein_seq} \\nVariant sequence: {translated_seq}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Verify codon frequencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bacillus_codon_frequencies\n",
    "ecoli_codon_frequencies\n",
    "\n",
    "# import codonbias as cb\n",
    "# codon_counter = cb.scores.CodonCounter(genetic_code=1, ignore_stop=False)\n",
    "# codon_counter.count(ref_seq)\\\n",
    "#             .get_aa_table(normed=True, pseudocount=pseudocount)\\\n",
    "#             .groupby('aa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Skip if not taking the original sequence prefix\n",
    "# skipped_codons = 12\n",
    "# skipped_codons_in_nt = 12 * 3\n",
    "# for variant_name, variant_seq in mcherry_variants_dict.items():\n",
    "#     assert original_mcherry_seq[:skipped_codons_in_nt].lower() == str(variant_seq.seq)[:skipped_codons_in_nt].lower(), f\"variant {variant_name} has different initiation prefix\"\n",
    "#     assert original_mcherry_seq.lower() != str(variant_seq.seq).lower(), \"variant is identical to the original sequence\"\n",
    "#     # print(original_mcherry_seq[:skipped_codons_in_nt].lower())\n",
    "#     # print(str(variant_seq.seq)[:skipped_codons_in_nt].lower())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Measurements for assessing variants scores "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Eucledean distance of codon frequencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "original_mcherry_codon_frequencies = calculate_codon_frequencies((str(original_mcherry_seq),))\n",
    "original_mcherry_codon_frequencies = original_mcherry_codon_frequencies.sort_values(\"codon\")\n",
    "\n",
    "original_mcherry_codon_frequencies_lamda = calculate_codon_frequencies_lamda((str(original_mcherry_seq),))\n",
    "original_mcherry_codon_frequencies_lamda = original_mcherry_codon_frequencies_lamda.sort_values(\"codon\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "original_bacillus_dist = calculate_distance_between_frequencies(original_mcherry_codon_frequencies[\"frequency\"].reset_index(drop=True), bacillus_codon_frequencies[\"frequency\"].reset_index(drop=True))\n",
    "original_ecoli_dist = calculate_distance_between_frequencies(original_mcherry_codon_frequencies[\"frequency\"].reset_index(drop=True), ecoli_codon_frequencies[\"frequency\"].reset_index(drop=True))\n",
    "\n",
    "original_bacillus_dist_lamda = calculate_distance_between_frequencies(original_mcherry_codon_frequencies_lamda[\"frequency\"].reset_index(drop=True), bacillus_codon_frequencies_lamda[\"frequency\"].reset_index(drop=True))\n",
    "original_ecoli_dist_lamda = calculate_distance_between_frequencies(original_mcherry_codon_frequencies_lamda[\"frequency\"].reset_index(drop=True), ecoli_codon_frequencies_lamda[\"frequency\"].reset_index(drop=True))\n",
    "\n",
    "original_bacillus_dist_highly_expressed = calculate_distance_between_frequencies(original_mcherry_codon_frequencies[\"frequency\"].reset_index(drop=True), bacillus_highly_expressed_genes_codon_frequencies[\"frequency\"].reset_index(drop=True))\n",
    "original_ecoli_dist_highly_expressed = calculate_distance_between_frequencies(original_mcherry_codon_frequencies[\"frequency\"].reset_index(drop=True), ecoli_highly_expressed_genes_codon_frequencies[\"frequency\"].reset_index(drop=True))\n",
    "\n",
    "original_bacillus_dist_highly_expressed_lamda = calculate_distance_between_frequencies(original_mcherry_codon_frequencies_lamda[\"frequency\"].reset_index(drop=True), bacillus_highly_expressed_genes_codon_frequencies_lamda[\"frequency\"].reset_index(drop=True))\n",
    "original_ecoli_dist_highly_expressed_lamda = calculate_distance_between_frequencies(original_mcherry_codon_frequencies_lamda[\"frequency\"].reset_index(drop=True), ecoli_highly_expressed_genes_codon_frequencies_lamda[\"frequency\"].reset_index(drop=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_dict = get_variants_df_dict(mcherry_variants_file, ecoli_codon_frequencies, bacillus_codon_frequencies, ecoli_codon_frequencies_lamda, bacillus_codon_frequencies_lamda)\n",
    "\n",
    "ecoli_opt_df_dict = {key: value for key, value in df_dict.items() if not key.endswith(\"for_optimized_bacillus\")}\n",
    "ecoli_opt_variants_dataframe = pd.DataFrame(ecoli_opt_df_dict)\n",
    "\n",
    "bacillus_opt_df_dict = {key: value for key, value in df_dict.items() if not key.endswith(\"for_optimized_ecoli\")}\n",
    "bacillus_opt_variants_dataframe = pd.DataFrame(bacillus_opt_df_dict)\n",
    "\n",
    "df_dict_highly_expressed = get_variants_df_dict(mcherry_variants_file, ecoli_highly_expressed_genes_codon_frequencies, bacillus_highly_expressed_genes_codon_frequencies, ecoli_highly_expressed_genes_codon_frequencies_lamda, bacillus_highly_expressed_genes_codon_frequencies_lamda)\n",
    "\n",
    "ecoli_opt_df_dict_highly_expressed = {key: value for key, value in df_dict_highly_expressed.items() if not key.endswith(\"for_optimized_bacillus\")}\n",
    "ecoli_opt_variants_dataframe_highly_expressed = pd.DataFrame(ecoli_opt_df_dict_highly_expressed)\n",
    "\n",
    "bacillus_opt_df_dict_highly_expressed = {key: value for key, value in df_dict_highly_expressed.items() if not key.endswith(\"for_optimized_ecoli\")}\n",
    "bacillus_opt_variants_dataframe_highly_expressed = pd.DataFrame(bacillus_opt_df_dict_highly_expressed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bacillus_A = bacillus_codon_frequencies[bacillus_codon_frequencies[\"amino_acid\"] == \"A\"].sort_values(\"codon\")\n",
    "x = pd.DataFrame(get_codon_frequencies_df_dict(mcherry_variants_file))\n",
    "x_A = x[x[\"amino_acid\"] == \"A\"].sort_values(\"codon\")\n",
    "original_A = original_mcherry_codon_frequencies[original_mcherry_codon_frequencies[\"amino_acid\"] == \"A\"].sort_values(\"codon\")\n",
    "\n",
    "\n",
    "original_dist = calculate_distance_between_frequencies(original_A[\"frequency\"].reset_index(drop=True), bacillus_A[\"frequency\"].reset_index(drop=True))\n",
    "variant_dist = calculate_distance_between_frequencies(x_A[\"frequency\"].reset_index(drop=True), bacillus_A[\"frequency\"].reset_index(drop=True))\n",
    "bacillus_A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "original_A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x_A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"original seq dist: {original_dist}, variant dist: {variant_dist}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### CUB scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_dict = get_cub_score_df_dict(mcherry_root_dir)\n",
    "cub_ecoli_opt_df_dict = {key.replace(\"_for_optimized_ecoli\", \"\"): value for key, value in df_dict.items() if not (key.endswith(\"_for_optimized_bacillus\") or key.startswith(\"evaluation_score\"))}\n",
    "cub_ecoli_opt_variants_dataframe = pd.DataFrame(cub_ecoli_opt_df_dict)\n",
    "\n",
    "evaluation_ecoli_opt_df_dict = {key: value for key, value in df_dict.items() if key in [\"variant\", \"average_distance_evaluation_score_for_optimized_ecoli\", \"ratio_evaluation_score_for_optimized_ecoli\"]}\n",
    "evaluation_ecoli_opt_variants_dataframe = pd.DataFrame(evaluation_ecoli_opt_df_dict)\n",
    "\n",
    "cub_bacillus_opt_df_dict = {key.replace(\"_for_optimized_bacillus\", \"\"): value for key, value in df_dict.items() if not (key.endswith(\"for_optimized_ecoli\") or key.startswith(\"evaluation_score\"))}\n",
    "cub_bacillus_opt_variants_dataframe = pd.DataFrame(cub_bacillus_opt_df_dict)\n",
    "\n",
    "evaluation_bacillus_opt_df_dict = {key: value for key, value in df_dict.items() if key in [\"variant\", \"average_distance_evaluation_score_for_optimized_bacillus\", \"ratio_evaluation_score_for_optimized_bacillus\"]}\n",
    "evaluation_bacillus_opt_variants_dataframe = pd.DataFrame(evaluation_bacillus_opt_df_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "evaluation_ecoli_opt_variants_dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### tuning parameter = 1 - Eucledean distance of codon frequencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "greater_alpha_df_dict = get_variants_df_dict(mcherry_greater_alpha_variants_file, ecoli_codon_frequencies, bacillus_codon_frequencies, ecoli_codon_frequencies_lamda, bacillus_codon_frequencies_lamda)\n",
    "    \n",
    "greater_alpha_ecoli_opt_df_dict = {key: value for key, value in greater_alpha_df_dict.items() if not key.endswith(\"for_optimized_bacillus\")}\n",
    "greater_alpha_ecoli_opt_variants_dataframe = pd.DataFrame(greater_alpha_ecoli_opt_df_dict)\n",
    "\n",
    "greater_alpha_bacillus_opt_df_dict = {key: value for key, value in greater_alpha_df_dict.items() if not key.endswith(\"for_optimized_ecoli\")}\n",
    "greater_alpha_bacillus_opt_variants_dataframe = pd.DataFrame(greater_alpha_bacillus_opt_df_dict)\n",
    "\n",
    "greater_alpha_df_dict_highly_expressed = get_variants_df_dict(mcherry_greater_alpha_variants_file, ecoli_highly_expressed_genes_codon_frequencies, bacillus_highly_expressed_genes_codon_frequencies, ecoli_highly_expressed_genes_codon_frequencies_lamda, bacillus_highly_expressed_genes_codon_frequencies_lamda)\n",
    "\n",
    "greater_alpha_ecoli_opt_df_dict_highly_expressed = {key: value for key, value in greater_alpha_df_dict_highly_expressed.items() if not key.endswith(\"for_optimized_bacillus\")}\n",
    "greater_alpha_ecoli_opt_variants_dataframe_highly_expressed = pd.DataFrame(greater_alpha_ecoli_opt_df_dict_highly_expressed)\n",
    "\n",
    "greater_alpha_bacillus_opt_df_dict_highly_expressed = {key: value for key, value in greater_alpha_df_dict_highly_expressed.items() if not key.endswith(\"for_optimized_ecoli\")}\n",
    "greater_alpha_bacillus_opt_variants_dataframe_highly_expressed = pd.DataFrame(greater_alpha_bacillus_opt_df_dict_highly_expressed)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### tuning parameter = 1 - CUB scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "greater_alpha_df_dict = get_cub_score_df_dict(mcherry_greater_alpha_root_dir)\n",
    "cub_ecoli_opt_greater_alpha_df_dict = {key.replace(\"_for_optimized_ecoli\", \"\"): value for key, value in greater_alpha_df_dict.items() if not (key.endswith(\"_for_optimized_bacillus\") or key.startswith(\"evaluation_score\"))}\n",
    "cub_ecoli_opt_greater_alpha_variants_dataframe = pd.DataFrame(cub_ecoli_opt_greater_alpha_df_dict)\n",
    "\n",
    "evaluation_ecoli_opt_greate_alpha_df_dict = {key: value for key, value in greater_alpha_df_dict.items() if key in [\"variant\", \"average_distance_evaluation_score_for_optimized_ecoli\", \"ratio_evaluation_score_for_optimized_ecoli\"]}\n",
    "evaluation_ecoli_opt_greater_alpha_variants_dataframe = pd.DataFrame(evaluation_ecoli_opt_greate_alpha_df_dict)\n",
    "\n",
    "cub_bacillus_opt_greater_alpha_df_dict = {key.replace(\"_for_optimized_bacillus\", \"\"): value for key, value in greater_alpha_df_dict.items() if not (key.endswith(\"for_optimized_ecoli\") or key.startswith(\"evaluation_score\"))}\n",
    "cub_bacillus_opt_greater_alpha_variants_dataframe = pd.DataFrame(cub_bacillus_opt_greater_alpha_df_dict)\n",
    "\n",
    "evaluation_bacillus_opt_greater_alpha_df_dict = {key: value for key, value in greater_alpha_df_dict.items() if key in [\"variant\", \"evaluation_score_for_optimized_bacillus\", \"average_distance_evaluation_score_for_optimized_bacillus\", \"ratio_evaluation_score_for_optimized_bacillus\"]}\n",
    "evaluation_bacillus_opt_greater_alpha_variants_dataframe = pd.DataFrame(evaluation_bacillus_opt_greater_alpha_df_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### tuning parameter = 0 - Eucledean distance of codon frequencies & CUB scores "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lower_alpha_df_dict = get_variants_df_dict(mcherry_lower_alpha_variants_file,  ecoli_codon_frequencies, bacillus_codon_frequencies, ecoli_codon_frequencies_lamda, bacillus_codon_frequencies_lamda)\n",
    "    \n",
    "lower_alpha_ecoli_opt_df_dict = {key: value for key, value in lower_alpha_df_dict.items() if not key.endswith(\"for_optimized_bacillus\")}\n",
    "lower_alpha_ecoli_opt_variants_dataframe = pd.DataFrame(lower_alpha_ecoli_opt_df_dict)\n",
    "\n",
    "lower_alpha_bacillus_opt_df_dict = {key: value for key, value in lower_alpha_df_dict.items() if not key.endswith(\"for_optimized_ecoli\")}\n",
    "lower_alpha_bacillus_opt_variants_dataframe = pd.DataFrame(lower_alpha_bacillus_opt_df_dict)\n",
    "\n",
    "lower_alpha_df_dict_highly_expressed = get_variants_df_dict(mcherry_lower_alpha_variants_file, ecoli_highly_expressed_genes_codon_frequencies, bacillus_highly_expressed_genes_codon_frequencies, ecoli_highly_expressed_genes_codon_frequencies_lamda, bacillus_highly_expressed_genes_codon_frequencies_lamda)\n",
    "\n",
    "lower_alpha_ecoli_opt_df_dict_highly_expressed = {key: value for key, value in lower_alpha_df_dict_highly_expressed.items() if not key.endswith(\"for_optimized_bacillus\")}\n",
    "lower_alpha_ecoli_opt_variants_dataframe_highly_expressed = pd.DataFrame(lower_alpha_ecoli_opt_df_dict_highly_expressed)\n",
    "\n",
    "lower_alpha_bacillus_opt_df_dict_highly_expressed = {key: value for key, value in lower_alpha_df_dict_highly_expressed.items() if not key.endswith(\"for_optimized_ecoli\")}\n",
    "lower_alpha_bacillus_opt_variants_dataframe_highly_expressed = pd.DataFrame(lower_alpha_bacillus_opt_df_dict_highly_expressed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lower_alpha_df_dict = get_cub_score_df_dict(mcherry_lower_alpha_root_dir)\n",
    "cub_ecoli_opt_lower_alpha_df_dict = {key.replace(\"_for_optimized_ecoli\", \"\"): value for key, value in lower_alpha_df_dict.items() if not (key.endswith(\"_for_optimized_bacillus\") or key.startswith(\"evaluation_score\"))}\n",
    "cub_ecoli_opt_lower_alpha_variants_dataframe = pd.DataFrame(cub_ecoli_opt_lower_alpha_df_dict)\n",
    "\n",
    "evaluation_ecoli_opt_lower_alpha_df_dict = {key: value for key, value in lower_alpha_df_dict.items() if key in [\"variant\", \"average_distance_evaluation_score_for_optimized_ecoli\", \"ratio_evaluation_score_for_optimized_ecoli\"]}\n",
    "evaluation_ecoli_opt_lower_alpha_variants_dataframe = pd.DataFrame(evaluation_ecoli_opt_lower_alpha_df_dict)\n",
    "\n",
    "cub_bacillus_opt_lower_alpha_df_dict = {key.replace(\"_for_optimized_bacillus\", \"\"): value for key, value in lower_alpha_df_dict.items() if not (key.endswith(\"for_optimized_ecoli\") or key.startswith(\"evaluation_score\"))}\n",
    "cub_bacillus_opt_lower_alpha_variants_dataframe = pd.DataFrame(cub_bacillus_opt_lower_alpha_df_dict)\n",
    "\n",
    "evaluation_bacillus_opt_lower_alpha_df_dict = {key: value for key, value in lower_alpha_df_dict.items() if key in [\"variant\", \"average_distance_evaluation_score_for_optimized_bacillus\", \"ratio_evaluation_score_for_optimized_bacillus\"]}\n",
    "evaluation_bacillus_opt_lower_alpha_variants_dataframe = pd.DataFrame(evaluation_bacillus_opt_lower_alpha_df_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Comparative analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(nrows=1,ncols=2, figsize=(27,10))\n",
    "\n",
    "fig.suptitle(\"Variants Evaluation Scores\")\n",
    "evaluation_ecoli_opt_variants_dataframe.plot(\"variant\", kind=\"barh\", ax=axes[0], title=\"E.coli optimized\")\n",
    "evaluation_bacillus_opt_variants_dataframe.plot(\"variant\", kind=\"barh\", ax=axes[1], title=\"Bacillus optimized\")\n",
    "# evaluation_ecoli_opt_variants_dataframe.plot(\"variant\", kind=\"barh\", ax=axes[0], title=\"E.coli optimized\")\n",
    "# evaluation_bacillus_opt_variants_dataframe.plot(\"variant\", kind=\"barh\", ax=axes[1], title=\"Bacillus optimized\", color=\"#dc4d01\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparing variant scores with random sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "random_sequences_num = 1000\n",
    "random_sequences = [shared_functions_and_vars.synonymous_codon_permutation(original_mcherry_seq) for _ in range(random_sequences_num)]\n",
    "\n",
    "random_evaluation_scores = {}\n",
    "dummy_optimization_method = \"single_codon_ratio\"\n",
    "dummy_output_path = \"results/mcherry_debug\"\n",
    "for optimization_cub_index in [\"CAI\", \"tAI\"]:\n",
    "    for direction in [True, False]:\n",
    "        user_input = get_user_input(\n",
    "            optimization_method=dummy_optimization_method,\n",
    "            optimization_cub_index=optimization_cub_index,\n",
    "            is_ecoli_optimized=direction,\n",
    "            orf_sequence=str(original_mcherry_seq),\n",
    "            output_path=dummy_output_path,\n",
    "            should_use_protein_abundance=True,\n",
    "        )\n",
    "        key = f\"{optimization_cub_index}_ecoli_opt_{direction}\"\n",
    "        random_evaluation_scores[key] = []\n",
    "        for random_sequence in random_sequences:\n",
    "            evaluation_score = evaluate_score_for_sequence(\n",
    "                sequence=random_sequence,\n",
    "                user_input=user_input,\n",
    "                optimization_method=dummy_optimization_method,\n",
    "                optimization_cub_index=optimization_cub_index,\n",
    "                is_ecoli_optimized=direction,\n",
    "                orf_sequence=original_mcherry_seq,\n",
    "                output_path=dummy_output_path,\n",
    "            )\n",
    "            random_evaluation_scores[key].append(evaluation_score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_random_scores_df(is_ecoli_opt: bool):\n",
    "    random_scores = defaultdict(dict)\n",
    "    variants = evaluation_ecoli_opt_variants_dataframe[\"variant\"]\n",
    "    for key, values in random_evaluation_scores.items():\n",
    "        if str(is_ecoli_opt) not in key:\n",
    "            continue\n",
    "        average_distance_scores = [x.average_distance_score for x  in values]\n",
    "        average_distance_average = np.average(average_distance_scores)\n",
    "        average_distance_std =  np.std(average_distance_scores)\n",
    "        ratio_scores = [x.ratio_score for x  in values]\n",
    "        ratio_average = np.average(ratio_scores)\n",
    "        ratio_std =  np.std(ratio_scores)\n",
    "        for variant in variants:\n",
    "            if variant.startswith(key.split(\"_\")[0]):\n",
    "                random_scores[variant][\"average_distance_average\"] = average_distance_average\n",
    "                random_scores[variant][\"average_distance_std\"] = average_distance_std\n",
    "                random_scores[variant][\"ratio_average\"] = ratio_average\n",
    "                random_scores[variant][\"ratio_std\"] = ratio_std\n",
    "    \n",
    "    random_scores_df_dict = {\n",
    "        \"variant\": variants, \n",
    "        \"random_average_distance_average\": [random_scores[variant][\"average_distance_average\"] for variant in variants],\n",
    "        \"random_average_distance_std\": [random_scores[variant][\"average_distance_std\"] for variant in variants],\n",
    "        \"random_ratio_average\": [random_scores[variant][\"ratio_average\"] for variant in variants],\n",
    "        \"random_ratio_std\": [random_scores[variant][\"ratio_std\"] for variant in variants],\n",
    "    }\n",
    "\n",
    "    return pd.DataFrame(random_scores_df_dict)   \n",
    "\n",
    "ecoli_opt_random_scores_dataframe = get_random_scores_df(is_ecoli_opt=True)\n",
    "new_evaluation_ecoli_opt_variants_dataframe = pd.merge(evaluation_ecoli_opt_variants_dataframe, ecoli_opt_random_scores_dataframe, on='variant', how='inner')\n",
    "new_evaluation_ecoli_opt_variants_dataframe[\"average_distance_zscore\"] = (new_evaluation_ecoli_opt_variants_dataframe[\"average_distance_evaluation_score_for_optimized_ecoli\"] - new_evaluation_ecoli_opt_variants_dataframe[\"random_average_distance_average\"]) / new_evaluation_ecoli_opt_variants_dataframe[\"random_average_distance_std\"]\n",
    "new_evaluation_ecoli_opt_variants_dataframe[\"ratio_zscore\"] = (new_evaluation_ecoli_opt_variants_dataframe[\"ratio_evaluation_score_for_optimized_ecoli\"] - new_evaluation_ecoli_opt_variants_dataframe[\"random_ratio_average\"]) / new_evaluation_ecoli_opt_variants_dataframe[\"random_ratio_std\"]\n",
    "\n",
    "\n",
    "bacillus_opt_random_scores_dataframe = get_random_scores_df(is_ecoli_opt=False)\n",
    "new_evaluation_bacillus_opt_variants_dataframe = pd.merge(evaluation_bacillus_opt_variants_dataframe, bacillus_opt_random_scores_dataframe, on='variant', how='inner')\n",
    "new_evaluation_bacillus_opt_variants_dataframe[\"average_distance_zscore\"] = (new_evaluation_bacillus_opt_variants_dataframe[\"average_distance_evaluation_score_for_optimized_bacillus\"] - new_evaluation_bacillus_opt_variants_dataframe[\"random_average_distance_average\"]) / new_evaluation_bacillus_opt_variants_dataframe[\"random_average_distance_std\"]\n",
    "new_evaluation_bacillus_opt_variants_dataframe[\"ratio_zscore\"] = (new_evaluation_bacillus_opt_variants_dataframe[\"ratio_evaluation_score_for_optimized_bacillus\"] - new_evaluation_bacillus_opt_variants_dataframe[\"random_ratio_average\"]) / new_evaluation_bacillus_opt_variants_dataframe[\"random_ratio_std\"]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_evaluation_ecoli_opt_variants_dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_evaluation_bacillus_opt_variants_dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for key, values in random_evaluation_scores.items():\n",
    "    variants = evaluation_ecoli_opt_variants_dataframe[\"variant\"]\n",
    "    for score_type in [\"average_distance\", \"ratio\"]:\n",
    "        scores = [getattr(x, f\"{score_type}_score\") for x  in values]\n",
    "        for variant in variants:\n",
    "            if variant.startswith(key.split(\"_\")[0]):\n",
    "                for score in scores:\n",
    "                    if \"True\" in key:\n",
    "                        evaluation = evaluation_ecoli_opt_variants_dataframe[evaluation_ecoli_opt_variants_dataframe[\"variant\"] == variant][f\"{score_type}_evaluation_score_for_optimized_ecoli\"]\n",
    "                    else:\n",
    "                        evaluation = evaluation_bacillus_opt_variants_dataframe[evaluation_bacillus_opt_variants_dataframe[\"variant\"] == variant][f\"{score_type}_evaluation_score_for_optimized_bacillus\"]\n",
    "                    evaluation = evaluation.reset_index(drop=True)\n",
    "                    score_diff = evaluation - score\n",
    "                    # print(score_diff)\n",
    "                    if score_diff[0] <= 0 :\n",
    "                        print(f\"Found better random score: {score} for key: {key} and variant: {variant}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(nrows=1,ncols=2, figsize=(30,10))\n",
    "\n",
    "fig.suptitle(\"Variants Evaluation Z-Scores\")\n",
    "new_evaluation_ecoli_opt_variants_dataframe.plot(\"variant\", \"average_distance_zscore\", kind=\"barh\", ax=axes[0], title=\"E.coli optimized\")\n",
    "new_evaluation_bacillus_opt_variants_dataframe.plot(\"variant\", \"average_distance_zscore\", kind=\"barh\", ax=axes[1], title=\"Bacillus optimized\", color=\"#dc4d01\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(nrows=1,ncols=2, figsize=(30,10))\n",
    "\n",
    "fig.suptitle(\"Variants Evaluation Z-Scores\")\n",
    "new_evaluation_ecoli_opt_variants_dataframe.plot(\"variant\", \"ratio_zscore\", kind=\"barh\", ax=axes[0], title=\"E.coli optimized\")\n",
    "new_evaluation_bacillus_opt_variants_dataframe.plot(\"variant\", \"ratio_zscore\", kind=\"barh\", ax=axes[1], title=\"Bacillus optimized\", color=\"#dc4d01\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df_ecoli = pd.merge(evaluation_ecoli_opt_variants_dataframe, ecoli_opt_variants_dataframe, on='variant', how='inner')\n",
    "merged_df_ecoli['ecoli_freq_dist_diff'] =  merged_df_ecoli['ecoli_dist_for_optimized_ecoli'] - merged_df_ecoli['ecoli_dist_for_original_mcherry']\n",
    "merged_df_ecoli['bacillus_freq_dist_diff'] = merged_df_ecoli['bacillus_dist_for_optimized_ecoli'] - merged_df_ecoli['bacillus_dist_for_original_mcherry']\n",
    "\n",
    "merged_df_bacillus = pd.merge(evaluation_bacillus_opt_variants_dataframe, bacillus_opt_variants_dataframe, on='variant', how='inner')\n",
    "merged_df_bacillus['ecoli_freq_dist_diff'] =  merged_df_bacillus['ecoli_dist_for_optimized_bacillus'] - merged_df_bacillus['ecoli_dist_for_original_mcherry']\n",
    "merged_df_bacillus['bacillus_freq_dist_diff'] = merged_df_bacillus['bacillus_dist_for_optimized_bacillus'] - merged_df_bacillus['bacillus_dist_for_original_mcherry'] \n",
    "\n",
    "\n",
    "fig, axes = plt.subplots(nrows=2,ncols=2, figsize=(20,15))\n",
    "fig.suptitle(\"Ecaluation Score vs. Codon Frequencies distance\")\n",
    "\n",
    "scatter = merged_df_ecoli.plot.scatter(x=\"average_distance_evaluation_score_for_optimized_ecoli\", y=\"ecoli_freq_dist_diff\", ax=axes[0,0], title=\"E.coli optimized - Evaluation score vs.\\nE.coli codon frequencies diff (modified seq dist-original seq dist)\")\n",
    "r, p = stats.spearmanr(merged_df_ecoli[\"average_distance_evaluation_score_for_optimized_ecoli\"], merged_df_ecoli[\"ecoli_freq_dist_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_ecoli.plot.scatter(x=\"average_distance_evaluation_score_for_optimized_ecoli\", y=\"bacillus_freq_dist_diff\", ax=axes[0,1], title=\"E.coli optimized - Evaluation score vs.\\nBacillus codon frequencies diff (modified seq dist-original seq dist)\")\n",
    "r, p = stats.spearmanr(merged_df_ecoli[\"average_distance_evaluation_score_for_optimized_ecoli\"], merged_df_ecoli[\"bacillus_freq_dist_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_bacillus.plot.scatter(x=\"average_distance_evaluation_score_for_optimized_bacillus\", y=\"ecoli_freq_dist_diff\", ax=axes[1,0], title=\"Bacillus optimized - Evaluation score vs.\\nE.coli codon frequencies diff (modified seq dist-original seq dist)\")\n",
    "r, p = stats.spearmanr(merged_df_bacillus[\"average_distance_evaluation_score_for_optimized_bacillus\"], merged_df_bacillus[\"ecoli_freq_dist_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_bacillus.plot.scatter(x=\"average_distance_evaluation_score_for_optimized_bacillus\", y=\"bacillus_freq_dist_diff\", ax=axes[1,1], title=\"Bacillus optimized - Evaluation score vs.\\nBacillus codon frequencies diff (modified seq dist-original seq dist)\")\n",
    "r, p = stats.spearmanr(merged_df_bacillus[\"average_distance_evaluation_score_for_optimized_bacillus\"], merged_df_bacillus[\"bacillus_freq_dist_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df_ecoli = pd.merge(evaluation_ecoli_opt_greater_alpha_variants_dataframe, greater_alpha_ecoli_opt_variants_dataframe, on='variant', how='inner')\n",
    "merged_df_ecoli['ecoli_freq_dist_diff'] =  merged_df_ecoli['ecoli_dist_for_optimized_ecoli'] - merged_df_ecoli['ecoli_dist_for_original_mcherry']\n",
    "merged_df_ecoli['bacillus_freq_dist_diff'] = merged_df_ecoli['bacillus_dist_for_optimized_ecoli'] - merged_df_ecoli['bacillus_dist_for_original_mcherry']\n",
    "\n",
    "merged_df_bacillus = pd.merge(evaluation_bacillus_opt_greater_alpha_variants_dataframe, greater_alpha_bacillus_opt_variants_dataframe, on='variant', how='inner')\n",
    "merged_df_bacillus['ecoli_freq_dist_diff'] =  merged_df_bacillus['ecoli_dist_for_optimized_bacillus'] - merged_df_bacillus['ecoli_dist_for_original_mcherry']\n",
    "merged_df_bacillus['bacillus_freq_dist_diff'] = merged_df_bacillus['bacillus_dist_for_optimized_bacillus'] - merged_df_bacillus['bacillus_dist_for_original_mcherry'] \n",
    "\n",
    "\n",
    "fig, axes = plt.subplots(nrows=2,ncols=2, figsize=(20,13))\n",
    "fig.suptitle(\"Ecaluation Score vs. Codon Frequencies distance (alpha=1)\")\n",
    "\n",
    "scatter = merged_df_ecoli.plot.scatter(x=\"evaluation_score_for_optimized_ecoli\", y=\"ecoli_freq_dist_diff\", ax=axes[0,0], title=\"E.coli optimized - Evaluation score vs.\\nE.coli codon frequencies diff (modified seq dist-original seq dist)\")\n",
    "r, p = stats.pearsonr(merged_df_ecoli[\"evaluation_score_for_optimized_ecoli\"], merged_df_ecoli[\"ecoli_freq_dist_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_ecoli.plot.scatter(x=\"evaluation_score_for_optimized_ecoli\", y=\"bacillus_freq_dist_diff\", ax=axes[0,1], title=\"E.coli optimized - Evaluation score vs.\\nBacillus codon frequencies diff (modified seq dist-original seq dist)\")\n",
    "r, p = stats.pearsonr(merged_df_ecoli[\"evaluation_score_for_optimized_ecoli\"], merged_df_ecoli[\"bacillus_freq_dist_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_bacillus.plot.scatter(x=\"evaluation_score_for_optimized_bacillus\", y=\"ecoli_freq_dist_diff\", ax=axes[1,0], title=\"Bacillus optimized - Evaluation score vs.\\nE.coli codon frequencies diff (modified seq dist-original seq dist)\")\n",
    "r, p = stats.pearsonr(merged_df_bacillus[\"evaluation_score_for_optimized_bacillus\"], merged_df_bacillus[\"ecoli_freq_dist_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_bacillus.plot.scatter(x=\"evaluation_score_for_optimized_bacillus\", y=\"bacillus_freq_dist_diff\", ax=axes[1,1], title=\"Bacillus optimized - Evaluation score vs.\\nBacillus codon frequencies diff (modified seq dist-original seq dist)\")\n",
    "r, p = stats.pearsonr(merged_df_bacillus[\"evaluation_score_for_optimized_bacillus\"], merged_df_bacillus[\"bacillus_freq_dist_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df_ecoli = pd.merge(evaluation_ecoli_opt_lower_alpha_variants_dataframe, lower_alpha_ecoli_opt_variants_dataframe, on='variant', how='inner')\n",
    "merged_df_ecoli['ecoli_freq_dist_diff'] =  merged_df_ecoli['ecoli_dist_for_optimized_ecoli'] - merged_df_ecoli['ecoli_dist_for_original_mcherry']\n",
    "merged_df_ecoli['bacillus_freq_dist_diff'] = merged_df_ecoli['bacillus_dist_for_optimized_ecoli'] - merged_df_ecoli['bacillus_dist_for_original_mcherry']\n",
    "\n",
    "merged_df_bacillus = pd.merge(evaluation_bacillus_opt_lower_alpha_variants_dataframe, lower_alpha_bacillus_opt_variants_dataframe, on='variant', how='inner')\n",
    "merged_df_bacillus['ecoli_freq_dist_diff'] =  merged_df_bacillus['ecoli_dist_for_optimized_bacillus'] - merged_df_bacillus['ecoli_dist_for_original_mcherry']\n",
    "merged_df_bacillus['bacillus_freq_dist_diff'] = merged_df_bacillus['bacillus_dist_for_optimized_bacillus'] - merged_df_bacillus['bacillus_dist_for_original_mcherry'] \n",
    "\n",
    "\n",
    "fig, axes = plt.subplots(nrows=2,ncols=2, figsize=(20,13))\n",
    "fig.suptitle(\"Ecaluation Score vs. Codon Frequencies distance (alpha=0)\")\n",
    "\n",
    "scatter = merged_df_ecoli.plot.scatter(x=\"evaluation_score_for_optimized_ecoli\", y=\"ecoli_freq_dist_diff\", ax=axes[0,0], title=\"Bacillus deoptimized - Evaluation score vs.\\nE.coli codon frequencies diff (modified seq dist-original seq dist)\")\n",
    "r, p = stats.pearsonr(merged_df_ecoli[\"evaluation_score_for_optimized_ecoli\"], merged_df_ecoli[\"ecoli_freq_dist_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_ecoli.plot.scatter(x=\"evaluation_score_for_optimized_ecoli\", y=\"bacillus_freq_dist_diff\", ax=axes[0,1], title=\"Bacillus deoptimized - Evaluation score vs.\\nBacillus codon frequencies diff (modified seq dist-original seq dist)\")\n",
    "r, p = stats.pearsonr(merged_df_ecoli[\"evaluation_score_for_optimized_ecoli\"], merged_df_ecoli[\"bacillus_freq_dist_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_bacillus.plot.scatter(x=\"evaluation_score_for_optimized_bacillus\", y=\"ecoli_freq_dist_diff\", ax=axes[1,0], title=\"E.coli deoptimized - Evaluation score vs.\\nE.coli codon frequencies diff (modified seq dist-original seq dist)\")\n",
    "r, p = stats.pearsonr(merged_df_bacillus[\"evaluation_score_for_optimized_bacillus\"], merged_df_bacillus[\"ecoli_freq_dist_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_bacillus.plot.scatter(x=\"evaluation_score_for_optimized_bacillus\", y=\"bacillus_freq_dist_diff\", ax=axes[1,1], title=\"E.coli deoptimized - Evaluation score vs.\\nBacillus codon frequencies diff (modified seq dist-original seq dist)\")\n",
    "r, p = stats.pearsonr(merged_df_bacillus[\"evaluation_score_for_optimized_bacillus\"], merged_df_bacillus[\"bacillus_freq_dist_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df_ecoli = pd.merge(evaluation_ecoli_opt_variants_dataframe, cub_ecoli_opt_variants_dataframe, on='variant', how='inner')\n",
    "merged_df_ecoli['ecoli_score_diff'] = merged_df_ecoli['ecoli_final_score'] - merged_df_ecoli['ecoli_initial_score']\n",
    "merged_df_ecoli['bacillus_score_diff'] = merged_df_ecoli['bacillus_final_score'] - merged_df_ecoli['bacillus_initial_score']\n",
    "\n",
    "merged_df_bacillus = pd.merge(evaluation_bacillus_opt_variants_dataframe, cub_bacillus_opt_variants_dataframe, on='variant', how='inner')\n",
    "merged_df_bacillus['ecoli_score_diff'] = merged_df_bacillus['ecoli_final_score'] - merged_df_bacillus['ecoli_initial_score']\n",
    "merged_df_bacillus['bacillus_score_diff'] = merged_df_bacillus['bacillus_final_score'] - merged_df_bacillus['bacillus_initial_score']\n",
    "\n",
    "fig, axes = plt.subplots(nrows=2,ncols=2, figsize=(20,15))\n",
    "\n",
    "scatter = merged_df_ecoli.plot.scatter(x=\"evaluation_score_for_optimized_ecoli\", y=\"ecoli_score_diff\", ax=axes[0,0], title=\"E.coli optimized - Evaluation score vs.\\nE.coli CUB score diff (modified seq score-original seq score)\")\n",
    "r, p = stats.spearmanr(merged_df_ecoli[\"evaluation_score_for_optimized_ecoli\"], merged_df_ecoli[\"ecoli_score_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_ecoli.plot.scatter(x=\"evaluation_score_for_optimized_ecoli\", y=\"bacillus_score_diff\", ax=axes[0,1], title=\"E.coli optimized - Evaluation score vs.\\nBacillus CUB score diff (modified seq score-original seq score)\")\n",
    "r, p = stats.spearmanr(merged_df_ecoli[\"evaluation_score_for_optimized_ecoli\"], merged_df_ecoli[\"bacillus_score_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_bacillus.plot.scatter(x=\"evaluation_score_for_optimized_bacillus\", y=\"ecoli_score_diff\", ax=axes[1,0], title=\"Bacillus optimized - Evaluation score vs.\\nE.coli CUB score diff (modified seq score-original seq score)\")\n",
    "r, p = stats.spearmanr(merged_df_bacillus[\"evaluation_score_for_optimized_bacillus\"], merged_df_bacillus[\"ecoli_score_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_bacillus.plot.scatter(x=\"evaluation_score_for_optimized_bacillus\", y=\"bacillus_score_diff\", ax=axes[1,1], title=\"Bacillus optimized - Evaluation score vs.\\nBacillus CUB score diff (modified seq score-original seq score)\")\n",
    "r, p = stats.spearmanr(merged_df_bacillus[\"evaluation_score_for_optimized_bacillus\"], merged_df_bacillus[\"bacillus_score_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df_ecoli_greater_alpha = pd.merge(evaluation_ecoli_opt_greater_alpha_variants_dataframe, cub_ecoli_opt_greater_alpha_variants_dataframe, on='variant', how='inner')\n",
    "merged_df_ecoli_greater_alpha['ecoli_score_diff'] = merged_df_ecoli_greater_alpha['ecoli_final_score'] - merged_df_ecoli_greater_alpha['ecoli_initial_score']\n",
    "merged_df_ecoli_greater_alpha['bacillus_score_diff'] = merged_df_ecoli_greater_alpha['bacillus_final_score'] - merged_df_ecoli_greater_alpha['bacillus_initial_score']\n",
    "\n",
    "merged_df_bacillus_greater_alpha = pd.merge(evaluation_bacillus_opt_greater_alpha_variants_dataframe, cub_bacillus_opt_greater_alpha_variants_dataframe, on='variant', how='inner')\n",
    "merged_df_bacillus_greater_alpha['ecoli_score_diff'] = merged_df_bacillus_greater_alpha['ecoli_final_score'] - merged_df_bacillus_greater_alpha['ecoli_initial_score']\n",
    "merged_df_bacillus_greater_alpha['bacillus_score_diff'] = merged_df_bacillus_greater_alpha['bacillus_final_score'] - merged_df_bacillus_greater_alpha['bacillus_initial_score']\n",
    "\n",
    "fig, axes = plt.subplots(nrows=2,ncols=2, figsize=(20,15))\n",
    "\n",
    "scatter = merged_df_ecoli_greater_alpha.plot.scatter(x=\"evaluation_score_for_optimized_ecoli\", y=\"ecoli_score_diff\", ax=axes[0,0], title=\"E.coli optimized (alpha=1)- Evaluation score vs.\\nE.coli CUB score diff (modified seq score-original seq score\")\n",
    "r, p = stats.spearmanr(merged_df_ecoli_greater_alpha[\"evaluation_score_for_optimized_ecoli\"], merged_df_ecoli_greater_alpha[\"ecoli_score_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_ecoli_greater_alpha.plot.scatter(x=\"evaluation_score_for_optimized_ecoli\", y=\"bacillus_score_diff\", ax=axes[0,1], title=\"E.coli optimized (alpha=1)- Evaluation score vs.\\nBacillus CUB score diff (modified seq score-original seq score\")\n",
    "r, p = stats.spearmanr(merged_df_ecoli_greater_alpha[\"evaluation_score_for_optimized_ecoli\"], merged_df_ecoli_greater_alpha[\"bacillus_score_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_bacillus_greater_alpha.plot.scatter(x=\"evaluation_score_for_optimized_bacillus\", y=\"ecoli_score_diff\", ax=axes[1,0], title=\"Bacillus optimized (alpha=1)- Evaluation score vs.\\nE.coli CUB score diff (modified seq score-original seq score\")\n",
    "r, p = stats.spearmanr(merged_df_bacillus_greater_alpha[\"evaluation_score_for_optimized_bacillus\"], merged_df_bacillus_greater_alpha[\"ecoli_score_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_bacillus_greater_alpha.plot.scatter(x=\"evaluation_score_for_optimized_bacillus\", y=\"bacillus_score_diff\", ax=axes[1,1], title=\"Bacillus optimized (alpha=1)- Evaluation score vs.\\nBacillus CUB score diff (modified seq score-original seq score\")\n",
    "r, p = stats.spearmanr(merged_df_bacillus_greater_alpha[\"evaluation_score_for_optimized_bacillus\"], merged_df_bacillus_greater_alpha[\"bacillus_score_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df_ecoli_lower_alpha = pd.merge(evaluation_ecoli_opt_lower_alpha_variants_dataframe, cub_ecoli_opt_lower_alpha_variants_dataframe, on='variant', how='inner')\n",
    "merged_df_ecoli_lower_alpha['ecoli_score_diff'] = merged_df_ecoli_lower_alpha['ecoli_final_score'] - merged_df_ecoli_lower_alpha['ecoli_initial_score']\n",
    "merged_df_ecoli_lower_alpha['bacillus_score_diff'] = merged_df_ecoli_lower_alpha['bacillus_final_score'] - merged_df_ecoli_lower_alpha['bacillus_initial_score']\n",
    "\n",
    "merged_df_bacillus_lower_alpha = pd.merge(evaluation_bacillus_opt_lower_alpha_variants_dataframe, cub_bacillus_opt_lower_alpha_variants_dataframe, on='variant', how='inner')\n",
    "merged_df_bacillus_lower_alpha['ecoli_score_diff'] = merged_df_bacillus_lower_alpha['ecoli_final_score'] - merged_df_bacillus_lower_alpha['ecoli_initial_score']\n",
    "merged_df_bacillus_lower_alpha['bacillus_score_diff'] = merged_df_bacillus_lower_alpha['bacillus_final_score'] - merged_df_bacillus_lower_alpha['bacillus_initial_score']\n",
    "\n",
    "fig, axes = plt.subplots(nrows=2,ncols=2, figsize=(20,15))\n",
    "\n",
    "scatter = merged_df_ecoli_lower_alpha.plot.scatter(x=\"evaluation_score_for_optimized_ecoli\", y=\"ecoli_score_diff\", ax=axes[0,0], title=\"Bacillus deoptimized (alpha=0)- Evaluation score vs.\\nE.coli CUB score diff (modified seq score-original seq score\")\n",
    "r, p = stats.spearmanr(merged_df_ecoli_lower_alpha[\"evaluation_score_for_optimized_ecoli\"], merged_df_ecoli_lower_alpha[\"ecoli_score_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_ecoli_lower_alpha.plot.scatter(x=\"evaluation_score_for_optimized_ecoli\", y=\"bacillus_score_diff\", ax=axes[0,1], title=\"Bacillus deoptimized (alpha=0)- Evaluation score vs.\\nBacillus CUB score diff (modified seq score-original seq score\")\n",
    "r, p = stats.spearmanr(merged_df_ecoli_lower_alpha[\"evaluation_score_for_optimized_ecoli\"], merged_df_ecoli_lower_alpha[\"bacillus_score_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_bacillus_lower_alpha.plot.scatter(x=\"evaluation_score_for_optimized_bacillus\", y=\"ecoli_score_diff\", ax=axes[1,0], title=\"E.coli deoptimized (alpha=0)- Evaluation score vs.\\nE.coli CUB score diff (modified seq score-original seq score\")\n",
    "r, p = stats.spearmanr(merged_df_bacillus_lower_alpha[\"evaluation_score_for_optimized_bacillus\"], merged_df_bacillus_lower_alpha[\"ecoli_score_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_bacillus_lower_alpha.plot.scatter(x=\"evaluation_score_for_optimized_bacillus\", y=\"bacillus_score_diff\", ax=axes[1,1], title=\"E.coli deoptimized (alpha=0)- Evaluation score vs.\\nBacillus CUB score diff (modified seq score-original seq score\")\n",
    "r, p = stats.spearmanr(merged_df_bacillus_lower_alpha[\"evaluation_score_for_optimized_bacillus\"], merged_df_bacillus_lower_alpha[\"bacillus_score_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "my_colors = ['#3b5b92', '#3b5b92', '#dc4d01', '#dc4d01'] \n",
    "my_colors_extended = ['cornflowerblue', 'cornflowerblue', 'navy', 'navy', 'orange', 'orange', 'orangered', 'orangered'] \n",
    "for q in range(cub_ecoli_opt_greater_alpha_variants_dataframe.shape[0]):\n",
    "\n",
    "    fig, axes = plt.subplots(nrows=1,ncols=3, figsize=(20,5))\n",
    "    fig.suptitle(greater_alpha_ecoli_opt_variants_dataframe.iloc[q,0] + \" - codon frequency distance\")\n",
    "    lower_alpha_ecoli_opt_variants_dataframe.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[0], stacked=True, color=my_colors_extended)\n",
    "    axes[0].set_title(\"alpha = 0\")\n",
    "    ecoli_opt_variants_dataframe.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[1], stacked=True, color=my_colors_extended)\n",
    "    axes[1].set_title(\"alpha = 0.5\")\n",
    "    greater_alpha_ecoli_opt_variants_dataframe.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[2], stacked=True, color=my_colors_extended)\n",
    "    axes[2].set_title(\"alpha = 1\")\n",
    "    \n",
    "    fig, axes = plt.subplots(nrows=1,ncols=3, figsize=(20,5))\n",
    "    fig.suptitle(greater_alpha_ecoli_opt_variants_dataframe_highly_expressed.iloc[q,0] + \" - codon frequency distance (highly expressed)\")\n",
    "    lower_alpha_ecoli_opt_variants_dataframe_highly_expressed.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[0], stacked=True, color=my_colors_extended)\n",
    "    axes[0].set_title(\"alpha = 0\")\n",
    "    ecoli_opt_variants_dataframe_highly_expressed.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[1], stacked=True, color=my_colors_extended)\n",
    "    axes[1].set_title(\"alpha = 0.5\")\n",
    "    greater_alpha_ecoli_opt_variants_dataframe_highly_expressed.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[2], stacked=True, color=my_colors_extended)\n",
    "    axes[2].set_title(\"alpha = 1\")\n",
    "\n",
    "\n",
    "    fig, axes = plt.subplots(nrows=1,ncols=3, figsize=(20,5))\n",
    "    fig.suptitle(cub_ecoli_opt_greater_alpha_variants_dataframe.iloc[q,0] + \" - CUB scores\")\n",
    "    cub_ecoli_opt_lower_alpha_variants_dataframe.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[0], stacked=True, color=my_colors)\n",
    "    axes[0].set_title(\"alpha = 0\")\n",
    "    cub_ecoli_opt_variants_dataframe.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[1], stacked=True, color=my_colors)\n",
    "    axes[1].set_title(\"alpha = 0.5\")\n",
    "    cub_ecoli_opt_greater_alpha_variants_dataframe.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[2], stacked=True, color=my_colors)\n",
    "    axes[2].set_title(\"alpha = 1\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Optimizing for Bacillus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "my_colors = ['#3b5b92', '#3b5b92', '#dc4d01', '#dc4d01'] \n",
    "my_colors_extended = ['cornflowerblue', 'cornflowerblue', 'navy', 'navy', 'orange', 'orange', 'orangered', 'orangered'] \n",
    "for q in range(cub_bacillus_opt_greater_alpha_variants_dataframe.shape[0]):\n",
    "\n",
    "    fig, axes = plt.subplots(nrows=1,ncols=3, figsize=(20,5))\n",
    "    fig.suptitle(greater_alpha_bacillus_opt_variants_dataframe.iloc[q,0] + \" - codon frequency distance\")\n",
    "    lower_alpha_bacillus_opt_variants_dataframe.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[0], stacked=True, color=my_colors_extended)\n",
    "    axes[0].set_title(\"alpha = 0\")\n",
    "    bacillus_opt_variants_dataframe.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[1], stacked=True, color=my_colors_extended)\n",
    "    axes[1].set_title(\"alpha = 0.5\")\n",
    "    greater_alpha_bacillus_opt_variants_dataframe.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[2], stacked=True, color=my_colors_extended)\n",
    "    axes[2].set_title(\"alpha = 1\")\n",
    "    \n",
    "    fig, axes = plt.subplots(nrows=1,ncols=3, figsize=(20,5))\n",
    "    fig.suptitle(greater_alpha_bacillus_opt_variants_dataframe_highly_expressed.iloc[q,0] + \" - codon frequency distance (highly expressed)\")\n",
    "    lower_alpha_bacillus_opt_variants_dataframe_highly_expressed.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[0], stacked=True, color=my_colors_extended)\n",
    "    axes[0].set_title(\"alpha = 0\")\n",
    "    bacillus_opt_variants_dataframe_highly_expressed.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[1], stacked=True, color=my_colors_extended)\n",
    "    axes[1].set_title(\"alpha = 0.5\")\n",
    "    greater_alpha_bacillus_opt_variants_dataframe_highly_expressed.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[2], stacked=True, color=my_colors_extended)\n",
    "    axes[2].set_title(\"alpha = 1\")\n",
    "\n",
    "\n",
    "    fig, axes = plt.subplots(nrows=1,ncols=3, figsize=(20,5))\n",
    "    fig.suptitle(cub_bacillus_opt_greater_alpha_variants_dataframe.iloc[q,0] + \" - CUB scores\")\n",
    "    cub_bacillus_opt_lower_alpha_variants_dataframe.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[0], stacked=True, color=my_colors)\n",
    "    axes[0].set_title(\"alpha = 0\")\n",
    "    cub_bacillus_opt_variants_dataframe.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[1], stacked=True, color=my_colors)\n",
    "    axes[1].set_title(\"alpha = 0.5\")\n",
    "    cub_bacillus_opt_greater_alpha_variants_dataframe.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[2], stacked=True, color=my_colors)\n",
    "    axes[2].set_title(\"alpha = 1\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Debugging negative non normalized score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "root_dir = r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\orf_model_analysis\\results\\endogenous-remote-debug\\CAI_single_codon_diff_ecoli_opt_False_DppD\"\n",
    "run_summary_path = os.path.join(root_dir, \"run_summary.json\")\n",
    "\n",
    "with open(run_summary_path, \"r\") as json_file:\n",
    "    run_summary = json.load(json_file)\n",
    "\n",
    "stats_dict = defaultdict(list)\n",
    "for codon in run_summary[\"orf_sequence_stats\"].keys():\n",
    "    stats_dict[\"codon\"].append(codon)\n",
    "    stats_dict[\"aa\"].append(shared_functions_and_vars.nt_to_aa[codon])\n",
    "    stats_dict[\"counts\"].append(run_summary[\"orf_sequence_stats\"][codon])\n",
    "    stats_dict[\"optimized_loss\"].append(run_summary[\"orf_debug\"][\"optimized_loss\"][shared_functions_and_vars.nt_to_aa[codon]][codon])\n",
    "    stats_dict[\"deoptimized_loss\"].append(run_summary[\"orf_debug\"][\"deoptimized_loss\"][shared_functions_and_vars.nt_to_aa[codon]][codon])\n",
    "    stats_dict[\"total_loss\"].append(run_summary[\"orf_debug\"][\"total_loss\"][shared_functions_and_vars.nt_to_aa[codon]][codon])\n",
    "    stats_dict[\"optimal\"].append(run_summary[\"orf\"][\"aa_to_optimal_codon\"][shared_functions_and_vars.nt_to_aa[codon]] == codon)\n",
    "\n",
    "stats_df = pd.DataFrame(stats_dict)\n",
    "total = stats_df.groupby(stats_df[\"aa\"]).aggregate({\"counts\": \"sum\"})\n",
    "\n",
    "filtered_df = stats_df[stats_df.optimal]\n",
    "merged = filtered_df.merge(total, on='aa', how='inner', suffixes=(\"\", \"_aa\") )\n",
    "merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# scatter = merged_df_ecoli_lower_alpha.plot.scatter(x=\"evaluation_score_for_optimized_ecoli\", y=\"ecoli_score_diff\", ax=axes[0,0], title=\"Bacillus deoptimized (alpha=0)- Evaluation score vs.\\nE.coli CUB score diff (modified seq score-original seq score\")\n",
    "# TODO - think how to visualize the data in order to show the trend\n",
    "# stats_df.plot()\n",
    "\n",
    "colors = ['r', 'b', 'k', 'g', 'm']\n",
    "merged.plot.scatter(x=\"optimized_loss\", y=\"deoptimized_loss\", s=10*merged.counts_aa)\n",
    "stats_df.plot.scatter(x=\"optimized_loss\", y=\"deoptimized_loss\", s=10*stats_df.counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
