{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import matplotlib.pyplot as plt\n",
    "import operator\n",
    "import os\n",
    "import sys\n",
    "import typing\n",
    "from collections import defaultdict\n",
    "from scipy import stats\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "\n",
    "from Bio import SeqIO\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.append(str(Path(os.path.dirname(os.getcwd())).parent))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from generate_summary_file import generate_summary\n",
    "from input_testing_data.generate_input_testing_data_for_modules import generate_testing_data_for_ecoli_and_bacillus\n",
    "from modules.main import run_modules\n",
    "from modules import shared_functions_and_vars"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_single_method_for_orf_sequence(optimization_method: str,\n",
    "                                       is_ecoli_optimized: bool,\n",
    "                                       orf_sequence: typing.Optional[str] = None,\n",
    "                                       orf_sequence_file: typing.Optional[str] = None,\n",
    "                                       output_path: typing.Optional[str] = None,\n",
    "                                       optimization_cub_index: str = \"CAI\",\n",
    "                                       tuning_param: float = 0.5,\n",
    "                                       should_use_mrna_levels = True,\n",
    "                                       should_use_protein_abundance = False):\n",
    "    default_user_inp_raw = generate_testing_data_for_ecoli_and_bacillus(\n",
    "        optimization_method=optimization_method,\n",
    "        optimization_cub_index=optimization_cub_index,\n",
    "        clusters_count=1,\n",
    "        tuning_param=tuning_param,\n",
    "        is_ecoli_optimized=is_ecoli_optimized,\n",
    "        sequence=orf_sequence,\n",
    "        sequence_file_path=orf_sequence_file,\n",
    "        output_path=os.path.join(\"results\", output_path),\n",
    "        should_use_mrna_levels=should_use_mrna_levels,\n",
    "        should_use_protein_abundance=should_use_protein_abundance,\n",
    "    )\n",
    "    return run_modules(default_user_inp_raw)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_all_methods(orf_sequence = None, orf_sequence_file = None, output_path = None, tuning_param: int = 0.5, should_use_mrna_levels: bool = True, should_use_protein_abundance: bool = False):\n",
    "    for optimization_method in [\n",
    "        \"single_codon_ratio\",\n",
    "        \"single_codon_diff\",\n",
    "        # \"single_codon_weakest_link\",\n",
    "        # \"zscore_single_aa_ratio\",\n",
    "        \"zscore_bulk_aa_ratio\",\n",
    "        # \"zscore_single_aa_diff\",\n",
    "        \"zscore_bulk_aa_diff\",\n",
    "        # \"zscore_single_aa_weakest_link\",\n",
    "        # \"zscore_bulk_aa_weakest_link\",\n",
    "    ]:\n",
    "        for optimization_cub_index in [\"CAI\", \"tAI\"]:\n",
    "            for direction in [True, False]:\n",
    "                run_single_method_for_orf_sequence(optimization_method=optimization_method,\n",
    "                                                   optimization_cub_index=optimization_cub_index,\n",
    "                                                   is_ecoli_optimized=direction,\n",
    "                                                   orf_sequence=orf_sequence,\n",
    "                                                   orf_sequence_file=orf_sequence_file,\n",
    "                                                   output_path=output_path,\n",
    "                                                   tuning_param=tuning_param,\n",
    "                                                   should_use_mrna_levels=should_use_mrna_levels,\n",
    "                                                   should_use_protein_abundance=should_use_protein_abundance)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_sequences_fasta_file(root_dir) -> None:\n",
    "    filename = \"run_summary.json\"\n",
    "\n",
    "    sequences = []\n",
    "    sequences_names = []\n",
    "    for root, dirs, files in os.walk(root_dir):\n",
    "        for file in files:\n",
    "            if file == filename:\n",
    "                directory_name = Path(root).name\n",
    "                file_path = os.path.join(root, file)\n",
    "                with open(file_path, \"r\") as summary_file:\n",
    "                    results_json = json.load(summary_file)\n",
    "\n",
    "                seq = results_json[\"final_evaluation\"][\"final_sequence\"]\n",
    "                sequences.append(seq)\n",
    "                sequences_names.append(directory_name[:-5])\n",
    "\n",
    "    shared_functions_and_vars.write_fasta(os.path.join(root_dir, \"mcherry_variants\"), sequences, sequences_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_codon_frequencies(referece_cds: typing.Sequence[str]):\n",
    "    codons_counter = defaultdict(int)\n",
    "    \n",
    "    for cds in referece_cds:\n",
    "        if len(cds) % 3 != 0:\n",
    "            continue\n",
    "        for i in range(0, len(cds), 3):\n",
    "            codon = cds[i:i+3]\n",
    "            codons_counter[codon] += 1\n",
    "    \n",
    "    # Calculate relative frequncies\n",
    "    for amino_acid, codons in shared_functions_and_vars.synonymous_codons.items():\n",
    "        total_amino_acid_codons = 0\n",
    "        for amino_acid_codon in codons:\n",
    "            total_amino_acid_codons += codons_counter[amino_acid_codon]\n",
    "        \n",
    "        # In case a certain amino acid is missing from the reference cds collection, no need to normalize\n",
    "        if total_amino_acid_codons == 0:\n",
    "            continue\n",
    "        \n",
    "        for amino_acid_codon in codons:\n",
    "            codons_counter[amino_acid_codon] /= total_amino_acid_codons\n",
    "    \n",
    "    codons = list(codons_counter.keys())\n",
    "    df_dict = {\"codon\": codons, \"amino_acid\": [shared_functions_and_vars.nt_to_aa[codon] for codon in codons], \"frequency\": [codons_counter[codon] for codon in codons]}\n",
    "\n",
    "    return pd.DataFrame(df_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_codon_frequencies_lamda(referece_cds: typing.Sequence[str]):\n",
    "    codons_counter = defaultdict(int)\n",
    "    \n",
    "    for cds in referece_cds:\n",
    "        if len(cds) % 3 != 0:\n",
    "            continue\n",
    "        for i in range(0, len(cds), 3):\n",
    "            codon = cds[i:i+3]\n",
    "            codons_counter[codon] += 1\n",
    "    \n",
    "    # Calculate relative frequncies\n",
    "    for amino_acid, codons in shared_functions_and_vars.synonymous_codons.items():\n",
    "        amino_acid_codon_to_count = {codon: codons_counter[codon] for codon in codons}\n",
    "        most_frequent_codon = max(amino_acid_codon_to_count, key=amino_acid_codon_to_count.get)\n",
    "        \n",
    "        if amino_acid_codon_to_count[most_frequent_codon] == 0:\n",
    "            # In case a certain amino acid is missing from the reference cds collection, no need to normalize\n",
    "            continue\n",
    "            \n",
    "        for amino_acid_codon in codons:\n",
    "            codons_counter[amino_acid_codon] = 1 if amino_acid_codon == most_frequent_codon else 0 \n",
    "    \n",
    "    codons = list(codons_counter.keys())\n",
    "    df_dict = {\"codon\": codons, \"amino_acid\": [shared_functions_and_vars.nt_to_aa[codon] for codon in codons], \"frequency\": [codons_counter[codon] for codon in codons]}\n",
    "    \n",
    "    return pd.DataFrame(df_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_frequency_map(frequency_df):\n",
    "    sorted_df = frequency_df.sort_values(\"amino_acid\")\n",
    "    \n",
    "    plt.rcParams[\"figure.figsize\"] = [10, 5]\n",
    "    plt.rcParams[\"figure.autolayout\"] = True\n",
    "\n",
    "    sorted_df.plot(\"codon\", \"frequency\", kind=\"bar\", color = \"b\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_distance_between_frequencies(first_codons_freq, second_codons_freq):\n",
    "    return np.linalg.norm(first_codons_freq-second_codons_freq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_variants_df_dict(variants_root_path):\n",
    "    with open(variants_root_path, \"r\") as mcherry_variants_fasta_file:\n",
    "        mcherry_variants_dict = SeqIO.to_dict(SeqIO.parse(mcherry_variants_fasta_file, \"fasta\"))\n",
    "\n",
    "    unified_variants_dict = defaultdict(dict)\n",
    "\n",
    "    for variant_name, variant_seq in mcherry_variants_dict.items():\n",
    "        if \"weakest_link\" in variant_name:\n",
    "            continue\n",
    "        \n",
    "        stripped_name = variant_name.replace(\"_False\", \"\").replace(\"_True\", \"\").replace(\"_ecoli_opt\", \"\")\n",
    "       \n",
    "        mcherry_variant_codon_frequencies = calculate_codon_frequencies((str(variant_seq.seq),)).sort_values(\"codon\")\n",
    "        mcherry_variant_codon_frequencies_lamda = calculate_codon_frequencies_lamda((str(variant_seq.seq),)).sort_values(\"codon\")\n",
    "\n",
    "        series1 = mcherry_variant_codon_frequencies[\"frequency\"].reset_index(drop=True)\n",
    "        series2 =  ecoli_codon_frequencies[\"frequency\"].reset_index(drop=True)\n",
    "        series3 =  bacillus_codon_frequencies[\"frequency\"].reset_index(drop=True)\n",
    "        \n",
    "        ecoli_dist = calculate_distance_between_frequencies(series1, series2)\n",
    "        bacillus_dist = calculate_distance_between_frequencies(series1, series3)\n",
    "        \n",
    "        series1_lamda = mcherry_variant_codon_frequencies_lamda[\"frequency\"].reset_index(drop=True)\n",
    "        series2_lamda =  ecoli_codon_frequencies_lamda[\"frequency\"].reset_index(drop=True)\n",
    "        series3_lamda =  bacillus_codon_frequencies_lamda[\"frequency\"].reset_index(drop=True)\n",
    "        \n",
    "        ecoli_dist_lamda = calculate_distance_between_frequencies(series1_lamda, series2_lamda)\n",
    "        bacillus_dist_lamda = calculate_distance_between_frequencies(series1_lamda, series3_lamda)\n",
    "\n",
    "        if variant_name.endswith(\"True\"):\n",
    "            unified_variants_dict[stripped_name][\"ecoli_dist_for_optimized_ecoli\"] = ecoli_dist\n",
    "            unified_variants_dict[stripped_name][\"bacillus_dist_for_optimized_ecoli\"] = bacillus_dist\n",
    "            unified_variants_dict[stripped_name][\"ecoli_dist_lamda_for_optimized_ecoli\"] = ecoli_dist_lamda\n",
    "            unified_variants_dict[stripped_name][\"bacillus_dist_lamda_for_optimized_ecoli\"] = bacillus_dist_lamda\n",
    "        else:\n",
    "            unified_variants_dict[stripped_name][\"ecoli_dist_for_optimized_bacillus\"] = ecoli_dist\n",
    "            unified_variants_dict[stripped_name][\"bacillus_dist_for_optimized_bacillus\"] = bacillus_dist\n",
    "            unified_variants_dict[stripped_name][\"ecoli_dist_lamda_for_optimized_bacillus\"] = ecoli_dist_lamda\n",
    "            unified_variants_dict[stripped_name][\"bacillus_dist_lamda_for_optimized_bacillus\"] = bacillus_dist_lamda\n",
    "\n",
    "    df_dict = defaultdict(list)\n",
    "    for variant_name in unified_variants_dict.keys():\n",
    "        df_dict[\"variant\"].append(variant_name)\n",
    "        df_dict[\"ecoli_dist_for_original_mcherry\"].append(original_ecoli_dist)\n",
    "        df_dict[\"ecoli_dist_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"ecoli_dist_for_optimized_ecoli\"])\n",
    "        df_dict[\"ecoli_dist_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"ecoli_dist_for_optimized_bacillus\"])\n",
    "        df_dict[\"ecoli_dist_lamda_for_original_mcherry\"].append(original_ecoli_dist_lamda)\n",
    "        df_dict[\"ecoli_dist_lamda_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"ecoli_dist_lamda_for_optimized_ecoli\"])\n",
    "        df_dict[\"ecoli_dist_lamda_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"ecoli_dist_lamda_for_optimized_bacillus\"])\n",
    "\n",
    "        df_dict[\"bacillus_dist_for_original_mcherry\"].append(original_bacillus_dist)\n",
    "        df_dict[\"bacillus_dist_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"bacillus_dist_for_optimized_ecoli\"])\n",
    "        df_dict[\"bacillus_dist_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"bacillus_dist_for_optimized_bacillus\"])\n",
    "        df_dict[\"bacillus_dist_lamda_for_original_mcherry\"].append(original_bacillus_dist_lamda)\n",
    "        df_dict[\"bacillus_dist_lamda_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"bacillus_dist_lamda_for_optimized_ecoli\"])\n",
    "        df_dict[\"bacillus_dist_lamda_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"bacillus_dist_lamda_for_optimized_bacillus\"])\n",
    "    \n",
    "    return df_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_variants_df_dict_highly_expressed(variants_root_path):\n",
    "    with open(variants_root_path, \"r\") as mcherry_variants_fasta_file:\n",
    "        mcherry_variants_dict = SeqIO.to_dict(SeqIO.parse(mcherry_variants_fasta_file, \"fasta\"))\n",
    "\n",
    "    unified_variants_dict = defaultdict(dict)\n",
    "\n",
    "    for variant_name, variant_seq in mcherry_variants_dict.items():\n",
    "        if \"weakest_link\" in variant_name:\n",
    "            continue\n",
    "        \n",
    "        stripped_name = variant_name.replace(\"_False\", \"\").replace(\"_True\", \"\").replace(\"_ecoli_opt\", \"\")\n",
    "       \n",
    "        mcherry_variant_codon_frequencies = calculate_codon_frequencies((str(variant_seq.seq),)).sort_values(\"codon\")\n",
    "        mcherry_variant_codon_frequencies_lamda = calculate_codon_frequencies_lamda((str(variant_seq.seq),)).sort_values(\"codon\")\n",
    "\n",
    "        series1 = mcherry_variant_codon_frequencies[\"frequency\"].reset_index(drop=True)\n",
    "        series2 =  ecoli_highly_expressed_genes_codon_frequencies[\"frequency\"].reset_index(drop=True)\n",
    "        series3 =  bacillus_highly_expressed_genes_codon_frequencies[\"frequency\"].reset_index(drop=True)\n",
    "\n",
    "        ecoli_dist = calculate_distance_between_frequencies(series1, series2)\n",
    "        bacillus_dist = calculate_distance_between_frequencies(series1, series3)\n",
    "        \n",
    "        series1_lamda = mcherry_variant_codon_frequencies_lamda[\"frequency\"].reset_index(drop=True)\n",
    "        series2_lamda =  ecoli_highly_expressed_genes_codon_frequencies_lamda[\"frequency\"].reset_index(drop=True)\n",
    "        series3_lamda =  bacillus_highly_expressed_genes_codon_frequencies_lamda[\"frequency\"].reset_index(drop=True)\n",
    "\n",
    "        ecoli_dist_lamda = calculate_distance_between_frequencies(series1_lamda, series2_lamda)\n",
    "        bacillus_dist_lamda = calculate_distance_between_frequencies(series1_lamda, series3_lamda)\n",
    "\n",
    "        if variant_name.endswith(\"True\"):\n",
    "            unified_variants_dict[stripped_name][\"ecoli_dist_for_optimized_ecoli\"] = ecoli_dist\n",
    "            unified_variants_dict[stripped_name][\"bacillus_dist_for_optimized_ecoli\"] = bacillus_dist\n",
    "            unified_variants_dict[stripped_name][\"ecoli_dist_lamda_for_optimized_ecoli\"] = ecoli_dist_lamda\n",
    "            unified_variants_dict[stripped_name][\"bacillus_dist_lamda_for_optimized_ecoli\"] = bacillus_dist_lamda\n",
    "        else:\n",
    "            unified_variants_dict[stripped_name][\"ecoli_dist_for_optimized_bacillus\"] = ecoli_dist\n",
    "            unified_variants_dict[stripped_name][\"bacillus_dist_for_optimized_bacillus\"] = bacillus_dist\n",
    "            unified_variants_dict[stripped_name][\"ecoli_dist_lamda_for_optimized_bacillus\"] = ecoli_dist_lamda\n",
    "            unified_variants_dict[stripped_name][\"bacillus_dist_lamda_for_optimized_bacillus\"] = bacillus_dist_lamda\n",
    "\n",
    "    df_dict = defaultdict(list)\n",
    "    for variant_name in unified_variants_dict.keys():\n",
    "        df_dict[\"variant\"].append(variant_name)\n",
    "        df_dict[\"ecoli_dist_for_original_mcherry\"].append(original_ecoli_dist)\n",
    "        df_dict[\"ecoli_dist_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"ecoli_dist_for_optimized_ecoli\"])\n",
    "        df_dict[\"ecoli_dist_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"ecoli_dist_for_optimized_bacillus\"])\n",
    "        df_dict[\"ecoli_dist_lamda_for_original_mcherry\"].append(original_ecoli_dist_lamda)\n",
    "        df_dict[\"ecoli_dist_lamda_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"ecoli_dist_lamda_for_optimized_ecoli\"])\n",
    "        df_dict[\"ecoli_dist_lamda_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"ecoli_dist_lamda_for_optimized_bacillus\"])\n",
    "\n",
    "        df_dict[\"bacillus_dist_for_original_mcherry\"].append(original_bacillus_dist)\n",
    "        df_dict[\"bacillus_dist_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"bacillus_dist_for_optimized_ecoli\"])\n",
    "        df_dict[\"bacillus_dist_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"bacillus_dist_for_optimized_bacillus\"])\n",
    "        df_dict[\"bacillus_dist_lamda_for_original_mcherry\"].append(original_bacillus_dist_lamda)\n",
    "        df_dict[\"bacillus_dist_lamda_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"bacillus_dist_lamda_for_optimized_ecoli\"])\n",
    "        df_dict[\"bacillus_dist_lamda_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"bacillus_dist_lamda_for_optimized_bacillus\"])\n",
    "    \n",
    "    return df_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_codon_frequencies_df_dict(variants_root_path):\n",
    "    with open(variants_root_path, \"r\") as mcherry_variants_fasta_file:\n",
    "        mcherry_variants_dict = SeqIO.to_dict(SeqIO.parse(mcherry_variants_fasta_file, \"fasta\"))\n",
    "\n",
    "    unified_variants_dict = defaultdict(dict)\n",
    "\n",
    "    for variant_name, variant_seq in mcherry_variants_dict.items():\n",
    "        if \"weakest_link\" in variant_name:\n",
    "            continue\n",
    "        \n",
    "        stripped_name = variant_name.replace(\"_False\", \"\").replace(\"_True\", \"\").replace(\"_ecoli_opt\", \"\")\n",
    "       \n",
    "        mcherry_variant_codon_frequencies = calculate_codon_frequencies((str(variant_seq.seq),)).sort_values(\"codon\")\n",
    "        \n",
    "        return mcherry_variant_codon_frequencies \n",
    "    \n",
    "        series1 = mcherry_variant_codon_frequencies[\"frequency\"].reset_index(drop=True)\n",
    "        \n",
    "        if variant_name.endswith(\"True\"):\n",
    "            unified_variants_dict[stripped_name][\"codon_frequencies_for_optimized_ecoli\"] = mcherry_variant_codon_frequencies\n",
    "        else:\n",
    "            unified_variants_dict[stripped_name][\"codon_frequencies_for_optimized_bacillus\"] = mcherry_variant_codon_frequencies\n",
    "\n",
    "    df_dict = defaultdict(list)\n",
    "    for variant_name in unified_variants_dict.keys():\n",
    "        df_dict[\"variant\"].append(variant_name)\n",
    "        df_dict[\"codon_frequencies_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"codon_frequencies_for_optimized_ecoli\"])\n",
    "        df_dict[\"codon_frequencies_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"codon_frequencies_for_optimized_bacillus\"])\n",
    "    \n",
    "    return df_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cub_score_df_dict(root_dir) -> None:\n",
    "    filename = \"run_summary.json\"\n",
    "    \n",
    "    unified_variants_dict = defaultdict(dict)\n",
    "    \n",
    "    for root, dirs, files in os.walk(root_dir):\n",
    "        for file in files:\n",
    "            if file == filename:\n",
    "                directory_name = Path(root).name\n",
    "                file_path = os.path.join(root, file)\n",
    "                with open(file_path, \"r\") as summary_file:\n",
    "                    results_json = json.load(summary_file)\n",
    "                \n",
    "                is_cai = directory_name.startswith(\"CAI\")\n",
    "                organisms = results_json[\"evaluation\"][\"organisms\"]\n",
    "                for organism in organisms:\n",
    "                    if organism[\"name\"] == \"Bacillus subtilis\":\n",
    "                        if is_cai:\n",
    "                            bacillus_initial_score = organism[\"cai_initial_score\"]\n",
    "                            bacillus_final_score = organism[\"cai_final_score\"]\n",
    "                        else:\n",
    "                            bacillus_initial_score = organism[\"tai_initial_score\"]\n",
    "                            bacillus_final_score = organism[\"tai_final_score\"]\n",
    "                    else:\n",
    "                        if is_cai:\n",
    "                            ecoli_initial_score = organism[\"cai_initial_score\"]\n",
    "                            ecoli_final_score = organism[\"cai_final_score\"]\n",
    "                        else:\n",
    "                            ecoli_initial_score = organism[\"tai_initial_score\"]\n",
    "                            ecoli_final_score = organism[\"tai_final_score\"]\n",
    "                            \n",
    "                \n",
    "                variant_name = directory_name[:-5]\n",
    "                stripped_name = variant_name.replace(\"_False\", \"\").replace(\"_True\", \"\").replace(\"_ecoli_opt\", \"\")\n",
    "                \n",
    "                if \"weakest_link\" in variant_name:\n",
    "                    continue\n",
    "                \n",
    "                if variant_name.endswith(\"True\"):\n",
    "                    suffix = \"_for_optimized_ecoli\"\n",
    "                else:\n",
    "                    suffix = \"_for_optimized_bacillus\"\n",
    "                \n",
    "                unified_variants_dict[stripped_name][\"bacillus_initial_score\" + suffix] = bacillus_initial_score\n",
    "                unified_variants_dict[stripped_name][\"bacillus_final_score\" + suffix] = bacillus_final_score\n",
    "                unified_variants_dict[stripped_name][\"ecoli_initial_score\" + suffix] = ecoli_initial_score\n",
    "                unified_variants_dict[stripped_name][\"ecoli_final_score\" + suffix] = ecoli_final_score\n",
    "                unified_variants_dict[stripped_name][\"evaluation_score\" + suffix] = results_json[\"evaluation\"][\"average_distance_score\"]\n",
    "                \n",
    "    \n",
    "    df_dict = defaultdict(list)\n",
    "    for variant_name in unified_variants_dict.keys():\n",
    "        df_dict[\"variant\"].append(variant_name)\n",
    "        df_dict[\"ecoli_initial_score_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"ecoli_initial_score_for_optimized_ecoli\"])\n",
    "        df_dict[\"ecoli_final_score_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"ecoli_final_score_for_optimized_ecoli\"])\n",
    "        df_dict[\"ecoli_initial_score_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"ecoli_initial_score_for_optimized_bacillus\"])\n",
    "        df_dict[\"ecoli_final_score_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"ecoli_final_score_for_optimized_bacillus\"])\n",
    "        df_dict[\"bacillus_initial_score_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"bacillus_initial_score_for_optimized_ecoli\"])\n",
    "        df_dict[\"bacillus_final_score_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"bacillus_final_score_for_optimized_ecoli\"])\n",
    "        df_dict[\"bacillus_initial_score_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"bacillus_initial_score_for_optimized_bacillus\"])\n",
    "        df_dict[\"bacillus_final_score_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"bacillus_final_score_for_optimized_bacillus\"])\n",
    "        df_dict[\"evaluation_score_for_optimized_ecoli\"].append(unified_variants_dict[variant_name][\"evaluation_score_for_optimized_ecoli\"])\n",
    "        df_dict[\"evaluation_score_for_optimized_bacillus\"].append(unified_variants_dict[variant_name][\"evaluation_score_for_optimized_bacillus\"])\n",
    "    \n",
    "\n",
    "    return df_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cub_weights(organism_summary_file_path: str):\n",
    "    with open(organism_summary_file_path, \"r\") as organism_summary_file:\n",
    "        organism_summary = json.load(organism_summary_file)\n",
    "    \n",
    "    cai_profile = organism_summary[\"cai_weights\"]\n",
    "    tai_profile = organism_summary[\"tai_weights\"]\n",
    "    \n",
    "    return cai_profile, tai_profile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cub_scores(organism_summary_file_path: str):\n",
    "    with open(organism_summary_file_path, \"r\") as organism_summary_file:\n",
    "        organism_summary = json.load(organism_summary_file)\n",
    "    \n",
    "    cai_scores = organism_summary[\"cai_scores\"]\n",
    "    tai_scores = organism_summary[\"tai_scores\"]\n",
    "    \n",
    "    return cai_scores, tai_scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_mrna_estimated_expression(endogenous_genes_fasta_file_path, mrna_levels_file_path):\n",
    "    expression_df = pd.read_csv(mrna_levels_file_path)\n",
    "    gene_name_to_mrna_level = {}\n",
    "    for _, pair in enumerate(zip(expression_df.gene.to_list(), expression_df.mRNA_level.to_list())):\n",
    "        measured_gene_name, expression_level = pair\n",
    "        try:\n",
    "            gene_name_to_mrna_level[measured_gene_name.lower()] = float(expression_level)\n",
    "        except:\n",
    "            continue\n",
    "            \n",
    "    with open(endogenous_genes_fasta_file_path, \"r\") as endogenous_genes_fasta_file:\n",
    "        endogenous_genes_dict = SeqIO.to_dict(SeqIO.parse(endogenous_genes_fasta_file, \"fasta\"))\n",
    "        \n",
    "    estimated_expression = {description: {\"cds\": cds, \"mrna_level\": gene_name_to_mrna_level[description.split(\"|\")[0].lower()]}  for description, cds in endogenous_genes_dict.items() if description.split(\"|\")[0].lower() in gene_name_to_mrna_level}\n",
    "    return estimated_expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_pa_estimated_expression(endogenous_genes_fasta_file_path, pa_file_path):\n",
    "    expression_df = pd.read_json(pa_file_path)\n",
    "    gene_name_to_pa = {}\n",
    "    for _, pair in enumerate(zip(expression_df.name.to_list(), expression_df.abundance.to_list())):\n",
    "        measured_gene_name, expression_level = pair\n",
    "        try:\n",
    "            gene_name_to_pa[measured_gene_name.lower()] = float(expression_level)\n",
    "        except:\n",
    "            continue\n",
    "            \n",
    "    with open(endogenous_genes_fasta_file_path, \"r\") as endogenous_genes_fasta_file:\n",
    "        endogenous_genes_dict = SeqIO.to_dict(SeqIO.parse(endogenous_genes_fasta_file, \"fasta\"))\n",
    "        \n",
    "    estimated_expression = {description: {\"cds\": cds, \"pa\": gene_name_to_pa[description.split(\"|\")[0].lower()]}  for description, cds in endogenous_genes_dict.items() if description.split(\"|\")[0].lower() in gene_name_to_pa}\n",
    "    return estimated_expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_estimated_expression(endogenous_genes_fasta_file_path, expression_levels_file_path, expression_type: str):\n",
    "    if expression_type == \"mrna_levels\":\n",
    "        return get_mrna_estimated_expression(endogenous_genes_fasta_file_path, expression_levels_file_path)\n",
    "    elif expression_type == \"protein_abundance\":\n",
    "        return get_pa_estimated_expression(endogenous_genes_fasta_file_path, expression_levels_file_path)\n",
    "    raise ValueError(\"Missing expression type support\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_highly_expressed_genes(endogenous_genes_fasta_file_path, expression_levels_file_path, expression_type):\n",
    "    estimated_expression = get_estimated_expression(endogenous_genes_fasta_file_path=endogenous_genes_fasta_file_path, expression_levels_file_path=expression_levels_file_path, expression_type=expression_type)\n",
    "    estimated_expression_threshold = 0.15\n",
    "    \n",
    "    expression_key = \"mrna_level\" if expression_type == \"mrna_levels\" else \"pa\"\n",
    "    sorted_estimated_expression = dict(\n",
    "        sorted(estimated_expression.items(), key=lambda x: x[1][expression_key], reverse=True)\n",
    "    )\n",
    "    highly_expressed_genes_count = round(len(sorted_estimated_expression) * estimated_expression_threshold)\n",
    "    \n",
    "    highly_expressed_names = list(sorted_estimated_expression.keys())[:highly_expressed_genes_count]\n",
    "    highly_expressed_genes = {key: value for key, value in estimated_expression.items() if key in highly_expressed_names}\n",
    "    return highly_expressed_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "base_path = os.path.join(Path(os.getcwd()).parent.resolve(), \"example_data\")\n",
    "macherry_file_path = os.path.join(base_path, \"mCherry_original.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(macherry_file_path, \"r\") as mcherry_fasta_file:\n",
    "    mcherry_dict = SeqIO.to_dict(SeqIO.parse(mcherry_fasta_file, \"fasta\"))\n",
    "\n",
    "original_mcherry_seq = list(mcherry_dict.values())[0].seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generating mcherry variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tuning parameter = 0.5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run_all_methods(orf_sequence_file=macherry_file_path, output_path=\"mcherry\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Using protein abundance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "##########################\n",
      "# USER INPUT INFORMATION #\n",
      "##########################\n",
      "Organism is optimized\n",
      "Information about Escherichia coli:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 90 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3895\n",
      "Calculate CAI weights from a reference set of 584 highly expressed genes from estimated expression dictionary.\n",
      "name=Escherichia coli, cai_std=0.09192830548402788, cai_avg=0.6260311851192147\n",
      "Organism is de-optimized\n",
      "Information about Bacillus subtilis:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 98 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3550\n",
      "Calculate CAI weights from a reference set of 532 highly expressed genes from estimated expression dictionary.\n",
      "name=Bacillus subtilis, cai_std=0.04172052669683352, cai_avg=0.7354163127574911\n",
      "Escherichia coli has weight of 1.0\n",
      "Bacillus subtilis has weight of 1.0\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is: ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Total input processing time: 20.604766845703125\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "\n",
      "Optimization method used is: OptimizationMethod.single_codon_ratio\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGT': 1.0266699244446114, 'TGC': 1.0})\n",
      "Optimal codon for C is: TGC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAT': 1.0, 'GAC': 0.8704743117735279})\n",
      "Optimal codon for D is: GAC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TCT': 1.1098288897731794, 'TCG': 0.9122339163709154, 'TCA': 1.6124930847398409, 'TCC': 0.7659716433061695, 'AGC': 0.8955866523143166, 'AGT': 1.014780408686112})\n",
      "Optimal codon for S is: TCC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAA': 1.6488052136133238, 'CAG': 0.8511814982403219})\n",
      "Optimal codon for Q is: CAG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATG': 1.0})\n",
      "Optimal codon for M is: ATG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAC': 1.0, 'AAT': 1.3664053706204924})\n",
      "Optimal codon for N is: AAC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CCT': 1.4515318315786538, 'CCG': 1.0, 'CCA': 1.1487500141191709, 'CCC': 0.9928440508587798})\n",
      "Optimal codon for P is: CCC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAG': 1.0411554812226402, 'AAA': 1.0})\n",
      "Optimal codon for K is: AAA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAG': 1.03717715193125, 'TGA': 0.9604604988453446, 'TAA': 1.0})\n",
      "Optimal codon for _ is: TGA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ACC': 0.5622484746202778, 'ACA': 1.852982246159984, 'ACG': 1.1275996541820978, 'ACT': 0.995344828845795})\n",
      "Optimal codon for T is: ACC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTT': 1.3108766233766234, 'TTC': 0.850358185195012})\n",
      "Optimal codon for F is: TTC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GCA': 1.3445989650711514, 'GCC': 0.9046857885919708, 'GCG': 0.9256386861313867, 'GCT': 1.2909863109263566})\n",
      "Optimal codon for A is: GCC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GGT': 0.7327043728383656, 'GGG': 1.0599359180021022, 'GGA': 1.631314789300235, 'GGC': 1.0})\n",
      "Optimal codon for G is: GGT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATC': 0.8610985703536493, 'ATA': 1.0398187526044447, 'ATT': 1.2436006102729276})\n",
      "Optimal codon for I is: ATC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTA': 1.342905017544206, 'TTG': 1.18186580635295, 'CTC': 1.1010138419157345, 'CTT': 1.884274082883542, 'CTG': 0.7485467530310581, 'CTA': 1.0654890450933776})\n",
      "Optimal codon for L is: CTG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAT': 1.274040632054176, 'CAC': 0.7563227619429947})\n",
      "Optimal codon for H is: CAC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CGA': 1.0889573611066927, 'CGC': 1.2660281680045868, 'CGG': 1.1270420583103506, 'CGT': 1.0, 'AGG': 1.0447076760109593, 'AGA': 1.5477869687535553})\n",
      "Optimal codon for R is: CGT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGG': 1.0})\n",
      "Optimal codon for W is: TGG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GTA': 1.1599639749538844, 'GTC': 1.1315778472258138, 'GTG': 0.7523086143311546, 'GTT': 1.0})\n",
      "Optimal codon for V is: GTG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAG': 1.016179802561771, 'GAA': 1.0})\n",
      "Optimal codon for E is: GAA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAT': 1.142005218039508, 'TAC': 0.8188261979018996})\n",
      "Optimal codon for Y is: TAC\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTGTCCAAAGGTGAAGAAGACAACATGGCCATCATCAAAGAATTCATGCGTTTCAAAGTGCACATGGAAGGTTCCGTGAACGGTCACGAATTCGAAATCGAAGGTGAAGGTGAAGGTCGTCCCTACGAAGGTACCCAGACCGCCAAACTGAAAGTGACCAAAGGTGGTCCCCTGCCCTTCGCCTGGGACATCCTGTCCCCCCAGTTCATGTACGGTTCCAAAGCCTACGTGAAACACCCCGCCGACATCCCCGACTACCTGAAACTGTCCTTCCCCGAAGGTTTCAAATGGGAACGTGTGATGAACTTCGAAGACGGTGGTGTGGTGACCGTGACCCAGGACTCCTCCCTGCAGGACGGTGAATTCATCTACAAAGTGAAACTGCGTGGTACCAACTTCCCCTCCGACGGTCCCGTGATGCAGAAAAAAACCATGGGTTGGGAAGCCTCCTCCGAACGTATGTACCCCGAAGACGGTGCCCTGAAAGGTGAAATCAAACAGCGTCTGAAACTGAAAGACGGTGGTCACTACGACGCCGAAGTGAAAACCACCTACAAAGCCAAAAAACCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTGGAACAGTACGAACGTGCCGAAGGTCGTCACTCCACCGGTGGTATGGACGAACTGTACAAATGA', 'average_distance_score': 1.964045494451632, 'weakest_link_score': 1.964045494451632}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTGTCCAAAGGTGAAGAAGACAACATGGCCATCATCAAAGAATTCATGCGTTTCAAAGTGCACATGGAAGGTTCCGTGAACGGTCACGAATTCGAAATCGAAGGTGAAGGTGAAGGTCGTCCCTACGAAGGTACCCAGACCGCCAAACTGAAAGTGACCAAAGGTGGTCCCCTGCCCTTCGCCTGGGACATCCTGTCCCCCCAGTTCATGTACGGTTCCAAAGCCTACGTGAAACACCCCGCCGACATCCCCGACTACCTGAAACTGTCCTTCCCCGAAGGTTTCAAATGGGAACGTGTGATGAACTTCGAAGACGGTGGTGTGGTGACCGTGACCCAGGACTCCTCCCTGCAGGACGGTGAATTCATCTACAAAGTGAAACTGCGTGGTACCAACTTCCCCTCCGACGGTCCCGTGATGCAGAAAAAAACCATGGGTTGGGAAGCCTCCTCCGAACGTATGTACCCCGAAGACGGTGCCCTGAAAGGTGAAATCAAACAGCGTCTGAAACTGAAAGACGGTGGTCACTACGACGCCGAAGTGAAAACCACCTACAAAGCCAAAAAACCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTGGAACAGTACGAACGTGCCGAAGGTCGTCACTCCACCGGTGGTATGGACGAACTGTACAAATGA', 'average_distance_score': 1.964045494451632, 'weakest_link_score': 1.964045494451632}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry_debug\\CAI_single_codon_ratio_ecoli_opt_True_IwnZ\n",
      "Final output: {'user_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'optimization_method': 'single_codon_ratio', 'optimization_cub_index': 'CAI', 'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.037037037037037035, 'TGA': 0.24183006535947713, 'GCA': 0.6554010349288486, 'GCC': 0.6652652005174644, 'GCG': 1.0, 'GCT': 0.6266170763260025, 'TGC': 1.0, 'TGT': 0.7162162162162162, 'GAC': 0.8422464375523889, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.36087048383688997, 'TTC': 1.0, 'TTT': 0.6891233766233766, 'GGA': 0.12073112073112073, 'GGC': 1.0, 'GGG': 0.19111752445085778, 'GGT': 0.9866923200256533, 'CAC': 1.0, 'CAT': 0.7259593679458239, 'ATA': 0.03593829462620784, 'ATC': 1.0, 'ATT': 0.7563993897270725, 'AAA': 1.0, 'AAG': 0.2715144230769231, 'CTA': 0.0255237398225231, 'CTC': 0.1333821242338304, 'CTG': 1.0, 'CTT': 0.11572591711645777, 'TTA': 0.10557131095050773, 'TTG': 0.12322751806788033, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.4079954093343534, 'CCA': 0.2681349767629825, 'CCC': 0.08830066680137402, 'CCG': 1.0, 'CCT': 0.19337239846433624, 'CAA': 0.3511947863866763, 'CAG': 1.0, 'AGA': 0.018922018348623854, 'AGG': 0.011085626911314986, 'CGA': 0.03631498470948013, 'CGC': 0.7138761467889909, 'CGG': 0.05600152905198777, 'CGT': 1.0, 'AGC': 1.0, 'AGT': 0.3131715081575806, 'TCA': 0.3048149621965778, 'TCC': 0.7942697970553123, 'TCG': 0.40907282132908873, 'TCT': 0.8901711102268206, 'ACA': 0.14701775384001597, 'ACC': 1.0, 'ACG': 0.3457011769399561, 'ACT': 0.4374625972471574, 'GTA': 0.531166518254675, 'GTC': 0.5075690115761353, 'GTG': 0.9971059661620658, 'GTT': 1.0, 'TGG': 1.0, 'TAC': 1.0, 'TAT': 0.857994781960492}, 'cai_avg': 0.6260311851192147, 'cai_std': 0.09192830548402788, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.10739856801909307, 'TGA': 0.1694510739856802, 'GCA': 1.0, 'GCC': 0.5246426811237062, 'GCG': 0.9196648595367175, 'GCT': 0.9361754558896007, 'TGC': 1.0, 'TGT': 0.7495652173913042, 'GAC': 0.6699738903394257, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.3869690068323596, 'TTC': 0.82402496099844, 'TTT': 1.0, 'GGA': 0.8480035296712993, 'GGC': 1.0, 'GGG': 0.29340392675932053, 'GGT': 0.6170306640194132, 'CAC': 0.6778131634819533, 'CAT': 1.0, 'ATA': 0.11114994757077944, 'ATC': 0.8386927647675637, 'ATT': 1.0, 'AAA': 1.0, 'AAG': 0.3398391420911528, 'CTA': 0.14688262702462837, 'CTC': 0.3046372309740404, 'CTG': 0.6640781007321944, 'CTT': 1.0, 'TTA': 0.5893055247392945, 'TTG': 0.41202573774129136, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.8348958333333333, 'CCA': 0.4923917284432306, 'CCC': 0.07452204447912603, 'CCG': 1.0, 'CCT': 0.7553648068669527, 'CAA': 1.0, 'CAG': 0.8251624335499115, 'AGA': 0.72005772005772, 'AGG': 0.09620009620009619, 'CGA': 0.19672919672919673, 'CGC': 0.984126984126984, 'CGG': 0.2751322751322751, 'CGT': 1.0, 'AGC': 0.8834134615384617, 'AGT': 0.33774038461538464, 'TCA': 0.9487179487179488, 'TCC': 0.42588141025641024, 'TCG': 0.2560096153846154, 'TCT': 1.0, 'ACA': 1.0, 'ACC': 0.22142692218887092, 'ACG': 0.5329023320249365, 'ACT': 0.4301546986839067, 'GTA': 0.7337249143416544, 'GTC': 0.681106216348507, 'GTG': 0.6669114047968673, 'GTT': 1.0, 'TGG': 1.0, 'TAC': 0.7787396121883656, 'TAT': 1.0}, 'cai_avg': 0.7354163127574911, 'cai_std': 0.04172052669683352, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}]}, 'orf_debug': {'C': defaultdict(<class 'float'>, {'TGT': 1.0266699244446114, 'TGC': 1.0}), 'D': defaultdict(<class 'float'>, {'GAT': 1.0, 'GAC': 0.8704743117735279}), 'S': defaultdict(<class 'float'>, {'TCT': 1.1098288897731794, 'TCG': 0.9122339163709154, 'TCA': 1.6124930847398409, 'TCC': 0.7659716433061695, 'AGC': 0.8955866523143166, 'AGT': 1.014780408686112}), 'Q': defaultdict(<class 'float'>, {'CAA': 1.6488052136133238, 'CAG': 0.8511814982403219}), 'M': defaultdict(<class 'float'>, {'ATG': 1.0}), 'N': defaultdict(<class 'float'>, {'AAC': 1.0, 'AAT': 1.3664053706204924}), 'P': defaultdict(<class 'float'>, {'CCT': 1.4515318315786538, 'CCG': 1.0, 'CCA': 1.1487500141191709, 'CCC': 0.9928440508587798}), 'K': defaultdict(<class 'float'>, {'AAG': 1.0411554812226402, 'AAA': 1.0}), '_': defaultdict(<class 'float'>, {'TAG': 1.03717715193125, 'TGA': 0.9604604988453446, 'TAA': 1.0}), 'T': defaultdict(<class 'float'>, {'ACC': 0.5622484746202778, 'ACA': 1.852982246159984, 'ACG': 1.1275996541820978, 'ACT': 0.995344828845795}), 'F': defaultdict(<class 'float'>, {'TTT': 1.3108766233766234, 'TTC': 0.850358185195012}), 'A': defaultdict(<class 'float'>, {'GCA': 1.3445989650711514, 'GCC': 0.9046857885919708, 'GCG': 0.9256386861313867, 'GCT': 1.2909863109263566}), 'G': defaultdict(<class 'float'>, {'GGT': 0.7327043728383656, 'GGG': 1.0599359180021022, 'GGA': 1.631314789300235, 'GGC': 1.0}), 'I': defaultdict(<class 'float'>, {'ATC': 0.8610985703536493, 'ATA': 1.0398187526044447, 'ATT': 1.2436006102729276}), 'L': defaultdict(<class 'float'>, {'TTA': 1.342905017544206, 'TTG': 1.18186580635295, 'CTC': 1.1010138419157345, 'CTT': 1.884274082883542, 'CTG': 0.7485467530310581, 'CTA': 1.0654890450933776}), 'H': defaultdict(<class 'float'>, {'CAT': 1.274040632054176, 'CAC': 0.7563227619429947}), 'R': defaultdict(<class 'float'>, {'CGA': 1.0889573611066927, 'CGC': 1.2660281680045868, 'CGG': 1.1270420583103506, 'CGT': 1.0, 'AGG': 1.0447076760109593, 'AGA': 1.5477869687535553}), 'W': defaultdict(<class 'float'>, {'TGG': 1.0}), 'V': defaultdict(<class 'float'>, {'GTA': 1.1599639749538844, 'GTC': 1.1315778472258138, 'GTG': 0.7523086143311546, 'GTT': 1.0}), 'E': defaultdict(<class 'float'>, {'GAG': 1.016179802561771, 'GAA': 1.0}), 'Y': defaultdict(<class 'float'>, {'TAT': 1.142005218039508, 'TAC': 0.8188261979018996})}, 'orf': {'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCC', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAA', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAA', 'Y': 'TAC'}, 'run_time': 0.07900000002700835}, 'evaluation': [{'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'cai_initial_score': 0.7891479965436429, 'cai_final_score': 0.8343280182260585, 'dist_score': 0.4914701891276073}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'cai_initial_score': 0.7608140098814632, 'cai_final_score': 0.6174363800575295, 'dist_score': -3.4366207997756564}], 'final_sequence': 'ATGGTGTCCAAAGGTGAAGAAGACAACATGGCCATCATCAAAGAATTCATGCGTTTCAAAGTGCACATGGAAGGTTCCGTGAACGGTCACGAATTCGAAATCGAAGGTGAAGGTGAAGGTCGTCCCTACGAAGGTACCCAGACCGCCAAACTGAAAGTGACCAAAGGTGGTCCCCTGCCCTTCGCCTGGGACATCCTGTCCCCCCAGTTCATGTACGGTTCCAAAGCCTACGTGAAACACCCCGCCGACATCCCCGACTACCTGAAACTGTCCTTCCCCGAAGGTTTCAAATGGGAACGTGTGATGAACTTCGAAGACGGTGGTGTGGTGACCGTGACCCAGGACTCCTCCCTGCAGGACGGTGAATTCATCTACAAAGTGAAACTGCGTGGTACCAACTTCCCCTCCGACGGTCCCGTGATGCAGAAAAAAACCATGGGTTGGGAAGCCTCCTCCGAACGTATGTACCCCGAAGACGGTGCCCTGAAAGGTGAAATCAAACAGCGTCTGAAACTGAAAGACGGTGGTCACTACGACGCCGAAGTGAAAACCACCTACAAAGCCAAAAAACCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTGGAACAGTACGAACGTGCCGAAGGTCGTCACTCCACCGGTGGTATGGACGAACTGTACAAATGA', 'average_distance_score': 1.964045494451632, 'weakest_link_score': 1.964045494451632}], 'final_evaluation': {'final_sequence': 'ATGGTGTCCAAAGGTGAAGAAGACAACATGGCCATCATCAAAGAATTCATGCGTTTCAAAGTGCACATGGAAGGTTCCGTGAACGGTCACGAATTCGAAATCGAAGGTGAAGGTGAAGGTCGTCCCTACGAAGGTACCCAGACCGCCAAACTGAAAGTGACCAAAGGTGGTCCCCTGCCCTTCGCCTGGGACATCCTGTCCCCCCAGTTCATGTACGGTTCCAAAGCCTACGTGAAACACCCCGCCGACATCCCCGACTACCTGAAACTGTCCTTCCCCGAAGGTTTCAAATGGGAACGTGTGATGAACTTCGAAGACGGTGGTGTGGTGACCGTGACCCAGGACTCCTCCCTGCAGGACGGTGAATTCATCTACAAAGTGAAACTGCGTGGTACCAACTTCCCCTCCGACGGTCCCGTGATGCAGAAAAAAACCATGGGTTGGGAAGCCTCCTCCGAACGTATGTACCCCGAAGACGGTGCCCTGAAAGGTGAAATCAAACAGCGTCTGAAACTGAAAGACGGTGGTCACTACGACGCCGAAGTGAAAACCACCTACAAAGCCAAAAAACCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTGGAACAGTACGAACGTGCCGAAGGTCGTCACTCCACCGGTGGTATGGACGAACTGTACAAATGA', 'average_distance_score': 1.964045494451632, 'weakest_link_score': 1.964045494451632}, 'zip_output_file_path': 'results\\\\mcherry_debug\\\\CAI_single_codon_ratio_ecoli_opt_True_IwnZ\\\\communique_results.zip'}\n",
      "##########################\n",
      "# USER INPUT INFORMATION #\n",
      "##########################\n",
      "Organism is optimized\n",
      "Information about Bacillus subtilis:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 98 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3550\n",
      "Calculate CAI weights from a reference set of 532 highly expressed genes from estimated expression dictionary.\n",
      "name=Bacillus subtilis, cai_std=0.04172052669683352, cai_avg=0.7354163127574911\n",
      "Organism is de-optimized\n",
      "Information about Escherichia coli:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 90 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3895\n",
      "Calculate CAI weights from a reference set of 584 highly expressed genes from estimated expression dictionary.\n",
      "name=Escherichia coli, cai_std=0.09192830548402788, cai_avg=0.6260311851192147\n",
      "Bacillus subtilis has weight of 1.0\n",
      "Escherichia coli has weight of 1.0\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is: ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Total input processing time: 20.32265567779541\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "\n",
      "Optimization method used is: OptimizationMethod.single_codon_ratio\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGT': 0.9740228832951947, 'TGC': 1.0})\n",
      "Optimal codon for C is: TGT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAT': 1.0, 'GAC': 1.148798978297904})\n",
      "Optimal codon for D is: GAT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TCT': 0.9010397992111868, 'TCG': 1.0962100641666987, 'TCA': 0.6201576983267123, 'TCC': 1.305531358424304, 'AGC': 1.1165865384615383, 'AGT': 0.9854348698894877})\n",
      "Optimal codon for S is: TCA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAA': 0.6064997804128238, 'CAG': 1.1748375664500885})\n",
      "Optimal codon for Q is: CAA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATG': 1.0})\n",
      "Optimal codon for M is: ATG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAC': 1.0, 'AAT': 0.7318472405783171})\n",
      "Optimal codon for N is: AAT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CCT': 0.6889273650391959, 'CCG': 1.0, 'CCA': 0.8705114147630908, 'CCC': 1.0072075258294901})\n",
      "Optimal codon for P is: CCT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAG': 0.9604713398095827, 'AAA': 1.0})\n",
      "Optimal codon for K is: AAG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAG': 0.9641554464808395, 'TGA': 1.041167233011862, 'TAA': 1.0})\n",
      "Optimal codon for _ is: TAG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ACC': 1.778573077811129, 'ACA': 0.5396705781031327, 'ACG': 0.8868395766982989, 'ACT': 1.00467694312493})\n",
      "Optimal codon for T is: ACA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTT': 0.7628482972136224, 'TTC': 1.17597503900156})\n",
      "Optimal codon for F is: TTT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GCA': 0.7437161755862898, 'GCC': 1.1053561497372184, 'GCG': 1.0803351404632826, 'GCT': 0.774601551957931})\n",
      "Optimal codon for A is: GCA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GGT': 1.3648069222327404, 'GGG': 0.9434532626131995, 'GGA': 0.6130024729494162, 'GGC': 1.0})\n",
      "Optimal codon for G is: GGA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATC': 1.1613072352324363, 'ATA': 0.961706064153286, 'ATT': 0.8041166848418757})\n",
      "Optimal codon for I is: ATT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTA': 0.744654303123178, 'TTG': 0.8461197494881767, 'CTC': 0.9082537947570458, 'CTT': 0.530708355585765, 'CTG': 1.3359218992678055, 'CTA': 0.9385361629057029})\n",
      "Optimal codon for L is: CTT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAT': 0.7849043231750532, 'CAC': 1.3221868365180467})\n",
      "Optimal codon for H is: CAT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CGA': 0.9183096011984467, 'CGC': 0.789871841142461, 'CGG': 0.88727833413705, 'CGT': 1.0, 'AGG': 0.9572055637786945, 'AGA': 0.6460837442024129})\n",
      "Optimal codon for R is: AGA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGG': 1.0})\n",
      "Optimal codon for W is: TGG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GTA': 0.8620957388265063, 'GTC': 0.8837217894037153, 'GTG': 1.3292417246731345, 'GTT': 1.0})\n",
      "Optimal codon for V is: GTA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAG': 0.9840778152439342, 'GAA': 1.0})\n",
      "Optimal codon for E is: GAG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAT': 0.8756527415143603, 'TAC': 1.2212603878116344})\n",
      "Optimal codon for Y is: TAT\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTATCAAAGGGAGAGGAGGATAATATGGCAATTATTAAGGAGTTTATGAGATTTAAGGTACATATGGAGGGATCAGTAAATGGACATGAGTTTGAGATTGAGGGAGAGGGAGAGGGAAGACCTTATGAGGGAACACAAACAGCAAAGCTTAAGGTAACAAAGGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAGGCATATGTAAAGCATCCTGCAGATATTCCTGATTATCTTAAGCTTTCATTTCCTGAGGGATTTAAGTGGGAGAGAGTAATGAATTTTGAGGATGGAGGAGTAGTAACAGTAACACAAGATTCATCACTTCAAGATGGAGAGTTTATTTATAAGGTAAAGCTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTAATGCAAAAGAAGACAATGGGATGGGAGGCATCATCAGAGAGAATGTATCCTGAGGATGGAGCACTTAAGGGAGAGATTAAGCAAAGACTTAAGCTTAAGGATGGAGGACATTATGATGCAGAGGTAAAGACAACATATAAGGCAAAGAAGCCTGTACAACTTCCTGGAGCATATAATGTAAATATTAAGCTTGATATTACATCACATAATGAGGATTATACAATTGTAGAGCAATATGAGAGAGCAGAGGGAAGACATTCAACAGGAGGAATGGATGAGCTTTATAAGTAG', 'average_distance_score': 2.391561709043347, 'weakest_link_score': 2.391561709043347}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTATCAAAGGGAGAGGAGGATAATATGGCAATTATTAAGGAGTTTATGAGATTTAAGGTACATATGGAGGGATCAGTAAATGGACATGAGTTTGAGATTGAGGGAGAGGGAGAGGGAAGACCTTATGAGGGAACACAAACAGCAAAGCTTAAGGTAACAAAGGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAGGCATATGTAAAGCATCCTGCAGATATTCCTGATTATCTTAAGCTTTCATTTCCTGAGGGATTTAAGTGGGAGAGAGTAATGAATTTTGAGGATGGAGGAGTAGTAACAGTAACACAAGATTCATCACTTCAAGATGGAGAGTTTATTTATAAGGTAAAGCTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTAATGCAAAAGAAGACAATGGGATGGGAGGCATCATCAGAGAGAATGTATCCTGAGGATGGAGCACTTAAGGGAGAGATTAAGCAAAGACTTAAGCTTAAGGATGGAGGACATTATGATGCAGAGGTAAAGACAACATATAAGGCAAAGAAGCCTGTACAACTTCCTGGAGCATATAATGTAAATATTAAGCTTGATATTACATCACATAATGAGGATTATACAATTGTAGAGCAATATGAGAGAGCAGAGGGAAGACATTCAACAGGAGGAATGGATGAGCTTTATAAGTAG', 'average_distance_score': 2.391561709043347, 'weakest_link_score': 2.391561709043347}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry_debug\\CAI_single_codon_ratio_ecoli_opt_False_tjgP\n",
      "Final output: {'user_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'optimization_method': 'single_codon_ratio', 'optimization_cub_index': 'CAI', 'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.10739856801909307, 'TGA': 0.1694510739856802, 'GCA': 1.0, 'GCC': 0.5246426811237062, 'GCG': 0.9196648595367175, 'GCT': 0.9361754558896007, 'TGC': 1.0, 'TGT': 0.7495652173913042, 'GAC': 0.6699738903394257, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.3869690068323596, 'TTC': 0.82402496099844, 'TTT': 1.0, 'GGA': 0.8480035296712993, 'GGC': 1.0, 'GGG': 0.29340392675932053, 'GGT': 0.6170306640194132, 'CAC': 0.6778131634819533, 'CAT': 1.0, 'ATA': 0.11114994757077944, 'ATC': 0.8386927647675637, 'ATT': 1.0, 'AAA': 1.0, 'AAG': 0.3398391420911528, 'CTA': 0.14688262702462837, 'CTC': 0.3046372309740404, 'CTG': 0.6640781007321944, 'CTT': 1.0, 'TTA': 0.5893055247392945, 'TTG': 0.41202573774129136, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.8348958333333333, 'CCA': 0.4923917284432306, 'CCC': 0.07452204447912603, 'CCG': 1.0, 'CCT': 0.7553648068669527, 'CAA': 1.0, 'CAG': 0.8251624335499115, 'AGA': 0.72005772005772, 'AGG': 0.09620009620009619, 'CGA': 0.19672919672919673, 'CGC': 0.984126984126984, 'CGG': 0.2751322751322751, 'CGT': 1.0, 'AGC': 0.8834134615384617, 'AGT': 0.33774038461538464, 'TCA': 0.9487179487179488, 'TCC': 0.42588141025641024, 'TCG': 0.2560096153846154, 'TCT': 1.0, 'ACA': 1.0, 'ACC': 0.22142692218887092, 'ACG': 0.5329023320249365, 'ACT': 0.4301546986839067, 'GTA': 0.7337249143416544, 'GTC': 0.681106216348507, 'GTG': 0.6669114047968673, 'GTT': 1.0, 'TGG': 1.0, 'TAC': 0.7787396121883656, 'TAT': 1.0}, 'cai_avg': 0.7354163127574911, 'cai_std': 0.04172052669683352, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}, {'name': 'Escherichia coli', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.037037037037037035, 'TGA': 0.24183006535947713, 'GCA': 0.6554010349288486, 'GCC': 0.6652652005174644, 'GCG': 1.0, 'GCT': 0.6266170763260025, 'TGC': 1.0, 'TGT': 0.7162162162162162, 'GAC': 0.8422464375523889, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.36087048383688997, 'TTC': 1.0, 'TTT': 0.6891233766233766, 'GGA': 0.12073112073112073, 'GGC': 1.0, 'GGG': 0.19111752445085778, 'GGT': 0.9866923200256533, 'CAC': 1.0, 'CAT': 0.7259593679458239, 'ATA': 0.03593829462620784, 'ATC': 1.0, 'ATT': 0.7563993897270725, 'AAA': 1.0, 'AAG': 0.2715144230769231, 'CTA': 0.0255237398225231, 'CTC': 0.1333821242338304, 'CTG': 1.0, 'CTT': 0.11572591711645777, 'TTA': 0.10557131095050773, 'TTG': 0.12322751806788033, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.4079954093343534, 'CCA': 0.2681349767629825, 'CCC': 0.08830066680137402, 'CCG': 1.0, 'CCT': 0.19337239846433624, 'CAA': 0.3511947863866763, 'CAG': 1.0, 'AGA': 0.018922018348623854, 'AGG': 0.011085626911314986, 'CGA': 0.03631498470948013, 'CGC': 0.7138761467889909, 'CGG': 0.05600152905198777, 'CGT': 1.0, 'AGC': 1.0, 'AGT': 0.3131715081575806, 'TCA': 0.3048149621965778, 'TCC': 0.7942697970553123, 'TCG': 0.40907282132908873, 'TCT': 0.8901711102268206, 'ACA': 0.14701775384001597, 'ACC': 1.0, 'ACG': 0.3457011769399561, 'ACT': 0.4374625972471574, 'GTA': 0.531166518254675, 'GTC': 0.5075690115761353, 'GTG': 0.9971059661620658, 'GTT': 1.0, 'TGG': 1.0, 'TAC': 1.0, 'TAT': 0.857994781960492}, 'cai_avg': 0.6260311851192147, 'cai_std': 0.09192830548402788, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}]}, 'orf_debug': {'C': defaultdict(<class 'float'>, {'TGT': 0.9740228832951947, 'TGC': 1.0}), 'D': defaultdict(<class 'float'>, {'GAT': 1.0, 'GAC': 1.148798978297904}), 'S': defaultdict(<class 'float'>, {'TCT': 0.9010397992111868, 'TCG': 1.0962100641666987, 'TCA': 0.6201576983267123, 'TCC': 1.305531358424304, 'AGC': 1.1165865384615383, 'AGT': 0.9854348698894877}), 'Q': defaultdict(<class 'float'>, {'CAA': 0.6064997804128238, 'CAG': 1.1748375664500885}), 'M': defaultdict(<class 'float'>, {'ATG': 1.0}), 'N': defaultdict(<class 'float'>, {'AAC': 1.0, 'AAT': 0.7318472405783171}), 'P': defaultdict(<class 'float'>, {'CCT': 0.6889273650391959, 'CCG': 1.0, 'CCA': 0.8705114147630908, 'CCC': 1.0072075258294901}), 'K': defaultdict(<class 'float'>, {'AAG': 0.9604713398095827, 'AAA': 1.0}), '_': defaultdict(<class 'float'>, {'TAG': 0.9641554464808395, 'TGA': 1.041167233011862, 'TAA': 1.0}), 'T': defaultdict(<class 'float'>, {'ACC': 1.778573077811129, 'ACA': 0.5396705781031327, 'ACG': 0.8868395766982989, 'ACT': 1.00467694312493}), 'F': defaultdict(<class 'float'>, {'TTT': 0.7628482972136224, 'TTC': 1.17597503900156}), 'A': defaultdict(<class 'float'>, {'GCA': 0.7437161755862898, 'GCC': 1.1053561497372184, 'GCG': 1.0803351404632826, 'GCT': 0.774601551957931}), 'G': defaultdict(<class 'float'>, {'GGT': 1.3648069222327404, 'GGG': 0.9434532626131995, 'GGA': 0.6130024729494162, 'GGC': 1.0}), 'I': defaultdict(<class 'float'>, {'ATC': 1.1613072352324363, 'ATA': 0.961706064153286, 'ATT': 0.8041166848418757}), 'L': defaultdict(<class 'float'>, {'TTA': 0.744654303123178, 'TTG': 0.8461197494881767, 'CTC': 0.9082537947570458, 'CTT': 0.530708355585765, 'CTG': 1.3359218992678055, 'CTA': 0.9385361629057029}), 'H': defaultdict(<class 'float'>, {'CAT': 0.7849043231750532, 'CAC': 1.3221868365180467}), 'R': defaultdict(<class 'float'>, {'CGA': 0.9183096011984467, 'CGC': 0.789871841142461, 'CGG': 0.88727833413705, 'CGT': 1.0, 'AGG': 0.9572055637786945, 'AGA': 0.6460837442024129}), 'W': defaultdict(<class 'float'>, {'TGG': 1.0}), 'V': defaultdict(<class 'float'>, {'GTA': 0.8620957388265063, 'GTC': 0.8837217894037153, 'GTG': 1.3292417246731345, 'GTT': 1.0}), 'E': defaultdict(<class 'float'>, {'GAG': 0.9840778152439342, 'GAA': 1.0}), 'Y': defaultdict(<class 'float'>, {'TAT': 0.8756527415143603, 'TAC': 1.2212603878116344})}, 'orf': {'aa_to_optimal_codon': {'C': 'TGT', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAG', '_': 'TAG', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTA', 'E': 'GAG', 'Y': 'TAT'}, 'run_time': 0.06199999991804361}, 'evaluation': [{'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'cai_initial_score': 0.7608140098814632, 'cai_final_score': 0.7395795772447973, 'dist_score': -0.5089684699085438}, {'name': 'Escherichia coli', 'is_wanted': False, 'cai_initial_score': 0.7891479965436429, 'cai_final_score': 0.30265495681447085, 'dist_score': -5.292091887995237}], 'final_sequence': 'ATGGTATCAAAGGGAGAGGAGGATAATATGGCAATTATTAAGGAGTTTATGAGATTTAAGGTACATATGGAGGGATCAGTAAATGGACATGAGTTTGAGATTGAGGGAGAGGGAGAGGGAAGACCTTATGAGGGAACACAAACAGCAAAGCTTAAGGTAACAAAGGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAGGCATATGTAAAGCATCCTGCAGATATTCCTGATTATCTTAAGCTTTCATTTCCTGAGGGATTTAAGTGGGAGAGAGTAATGAATTTTGAGGATGGAGGAGTAGTAACAGTAACACAAGATTCATCACTTCAAGATGGAGAGTTTATTTATAAGGTAAAGCTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTAATGCAAAAGAAGACAATGGGATGGGAGGCATCATCAGAGAGAATGTATCCTGAGGATGGAGCACTTAAGGGAGAGATTAAGCAAAGACTTAAGCTTAAGGATGGAGGACATTATGATGCAGAGGTAAAGACAACATATAAGGCAAAGAAGCCTGTACAACTTCCTGGAGCATATAATGTAAATATTAAGCTTGATATTACATCACATAATGAGGATTATACAATTGTAGAGCAATATGAGAGAGCAGAGGGAAGACATTCAACAGGAGGAATGGATGAGCTTTATAAGTAG', 'average_distance_score': 2.391561709043347, 'weakest_link_score': 2.391561709043347}], 'final_evaluation': {'final_sequence': 'ATGGTATCAAAGGGAGAGGAGGATAATATGGCAATTATTAAGGAGTTTATGAGATTTAAGGTACATATGGAGGGATCAGTAAATGGACATGAGTTTGAGATTGAGGGAGAGGGAGAGGGAAGACCTTATGAGGGAACACAAACAGCAAAGCTTAAGGTAACAAAGGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAGGCATATGTAAAGCATCCTGCAGATATTCCTGATTATCTTAAGCTTTCATTTCCTGAGGGATTTAAGTGGGAGAGAGTAATGAATTTTGAGGATGGAGGAGTAGTAACAGTAACACAAGATTCATCACTTCAAGATGGAGAGTTTATTTATAAGGTAAAGCTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTAATGCAAAAGAAGACAATGGGATGGGAGGCATCATCAGAGAGAATGTATCCTGAGGATGGAGCACTTAAGGGAGAGATTAAGCAAAGACTTAAGCTTAAGGATGGAGGACATTATGATGCAGAGGTAAAGACAACATATAAGGCAAAGAAGCCTGTACAACTTCCTGGAGCATATAATGTAAATATTAAGCTTGATATTACATCACATAATGAGGATTATACAATTGTAGAGCAATATGAGAGAGCAGAGGGAAGACATTCAACAGGAGGAATGGATGAGCTTTATAAGTAG', 'average_distance_score': 2.391561709043347, 'weakest_link_score': 2.391561709043347}, 'zip_output_file_path': 'results\\\\mcherry_debug\\\\CAI_single_codon_ratio_ecoli_opt_False_tjgP\\\\communique_results.zip'}\n",
      "##########################\n",
      "# USER INPUT INFORMATION #\n",
      "##########################\n",
      "Organism is optimized\n",
      "Information about Escherichia coli:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Escherichia coli. Calculating tAI profile.\n",
      "name=Escherichia coli, tai_std=0.023481599705588187, tai_avg=0.2504950429983414\n",
      "Organism is de-optimized\n",
      "Information about Bacillus subtilis:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Bacillus subtilis. Calculating tAI profile.\n",
      "name=Bacillus subtilis, tai_std=0.027038419737753252, tai_avg=0.3292474875006557\n",
      "Escherichia coli has weight of 1.0\n",
      "Bacillus subtilis has weight of 1.0\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is: ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Total input processing time: 20.663880109786987\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "\n",
      "Optimization method used is: OptimizationMethod.single_codon_ratio\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGT': 0.984009360374415, 'TGC': 1.0})\n",
      "Optimal codon for C is: TGT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAT': 0.906086387434555, 'GAC': 1.0})\n",
      "Optimal codon for D is: GAT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TCT': 0.8927125506072876, 'TCG': 0.8845108695652174, 'TCA': 1.11125, 'TCC': 0.8571428571428571, 'AGC': 1.125, 'AGT': 1.0348240469208212})\n",
      "Optimal codon for S is: TCC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAA': 1.0525, 'CAG': 0.6880733944954128})\n",
      "Optimal codon for Q is: CAG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATG': 1.0})\n",
      "Optimal codon for M is: ATG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAC': 1.0, 'AAT': 0.9463350785340315})\n",
      "Optimal codon for N is: AAT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CCT': 0.7275, 'CCG': 0.7462686567164178, 'CCA': 1.02625, 'CCC': 0.6933333333333334})\n",
      "Optimal codon for P is: CCC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAG': 1.0389908256880733, 'AAA': 1.0})\n",
      "Optimal codon for K is: AAA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAG': 0.9271127128057303, 'TGA': 0.9271127128057303, 'TAA': 0.9271127128057303})\n",
      "Optimal codon for _ is: TAG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ACC': 0.6933333333333334, 'ACA': 1.1375, 'ACG': 0.6880733944954128, 'ACT': 0.7284803400637619})\n",
      "Optimal codon for T is: ACG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTT': 0.9149377593360996, 'TTC': 1.0})\n",
      "Optimal codon for F is: TTT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GCA': 1.0, 'GCC': 0.6749999999999999, 'GCG': 0.8010638297872339, 'GCT': 0.7074927953890491})\n",
      "Optimal codon for A is: GCC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GGT': 0.9463350785340315, 'GGG': 0.8860619469026548, 'GGA': 1.1667857142857145, 'GGC': 1.0})\n",
      "Optimal codon for G is: GGG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATC': 1.0, 'ATA': 0.9100719424460432, 'ATT': 0.9574688796680498})\n",
      "Optimal codon for I is: ATA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTA': 1.40125, 'TTG': 1.153846153846154, 'CTC': 1.0468557336621456, 'CTT': 1.0325704225352113, 'CTG': 0.8032128514056224, 'CTA': 1.1657142857142857})\n",
      "Optimal codon for L is: CTG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAT': 0.9248533724340176, 'CAC': 1.0})\n",
      "Optimal codon for H is: CAT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CGA': 0.9016744863290729, 'CGC': 0.960674157303371, 'CGG': 0.9166666666666666, 'CGT': 1.0, 'AGG': 0.922811059907834, 'AGA': 0.9187289088863893})\n",
      "Optimal codon for R is: CGA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGG': 1.0})\n",
      "Optimal codon for W is: TGG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GTA': 1.0, 'GTC': 0.9166666666666666, 'GTG': 0.9805045871559633, 'GTT': 0.9420828905419766})\n",
      "Optimal codon for V is: GTC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAG': 0.7976190476190476, 'GAA': 1.0})\n",
      "Optimal codon for E is: GAG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAT': 1.0150293255131966, 'TAC': 1.0})\n",
      "Optimal codon for Y is: TAC\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTCTCCAAAGGGGAGGAGGATAATATGGCCATAATAAAAGAGTTTATGCGATTTAAAGTCCATATGGAGGGGTCCGTCAATGGGCATGAGTTTGAGATAGAGGGGGAGGGGGAGGGGCGACCCTACGAGGGGACGCAGACGGCCAAACTGAAAGTCACGAAAGGGGGGCCCCTGCCCTTTGCCTGGGATATACTGTCCCCCCAGTTTATGTACGGGTCCAAAGCCTACGTCAAACATCCCGCCGATATACCCGATTACCTGAAACTGTCCTTTCCCGAGGGGTTTAAATGGGAGCGAGTCATGAATTTTGAGGATGGGGGGGTCGTCACGGTCACGCAGGATTCCTCCCTGCAGGATGGGGAGTTTATATACAAAGTCAAACTGCGAGGGACGAATTTTCCCTCCGATGGGCCCGTCATGCAGAAAAAAACGATGGGGTGGGAGGCCTCCTCCGAGCGAATGTACCCCGAGGATGGGGCCCTGAAAGGGGAGATAAAACAGCGACTGAAACTGAAAGATGGGGGGCATTACGATGCCGAGGTCAAAACGACGTACAAAGCCAAAAAACCCGTCCAGCTGCCCGGGGCCTACAATGTCAATATAAAACTGGATATAACGTCCCATAATGAGGATTACACGATAGTCGAGCAGTACGAGCGAGCCGAGGGGCGACATTCCACGGGGGGGATGGATGAGCTGTACAAATAA', 'average_distance_score': 0.5137405766337915, 'weakest_link_score': 0.5137405766337915}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTCTCCAAAGGGGAGGAGGATAATATGGCCATAATAAAAGAGTTTATGCGATTTAAAGTCCATATGGAGGGGTCCGTCAATGGGCATGAGTTTGAGATAGAGGGGGAGGGGGAGGGGCGACCCTACGAGGGGACGCAGACGGCCAAACTGAAAGTCACGAAAGGGGGGCCCCTGCCCTTTGCCTGGGATATACTGTCCCCCCAGTTTATGTACGGGTCCAAAGCCTACGTCAAACATCCCGCCGATATACCCGATTACCTGAAACTGTCCTTTCCCGAGGGGTTTAAATGGGAGCGAGTCATGAATTTTGAGGATGGGGGGGTCGTCACGGTCACGCAGGATTCCTCCCTGCAGGATGGGGAGTTTATATACAAAGTCAAACTGCGAGGGACGAATTTTCCCTCCGATGGGCCCGTCATGCAGAAAAAAACGATGGGGTGGGAGGCCTCCTCCGAGCGAATGTACCCCGAGGATGGGGCCCTGAAAGGGGAGATAAAACAGCGACTGAAACTGAAAGATGGGGGGCATTACGATGCCGAGGTCAAAACGACGTACAAAGCCAAAAAACCCGTCCAGCTGCCCGGGGCCTACAATGTCAATATAAAACTGGATATAACGTCCCATAATGAGGATTACACGATAGTCGAGCAGTACGAGCGAGCCGAGGGGCGACATTCCACGGGGGGGATGGATGAGCTGTACAAATAA', 'average_distance_score': 0.5137405766337915, 'weakest_link_score': 0.5137405766337915}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry_debug\\tAI_single_codon_ratio_ecoli_opt_True_7FgF\n",
      "Final output: {'user_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'optimization_method': 'single_codon_ratio', 'optimization_cub_index': 'tAI', 'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': None, 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.75, 'AAC': 0.5, 'AAG': 0.23999999999999996, 'AAT': 0.29500000000000004, 'ACA': 0.1525, 'ACC': 0.25, 'ACG': 0.29, 'ACT': 0.14750000000000002, 'AGA': 0.13874999999999998, 'AGC': 0.125, 'AGG': 0.16499999999999998, 'AGT': 0.07375000000000001, 'ATA': 0.10999999999999999, 'ATC': 0.375, 'ATG': 1.0, 'ATT': 0.22125000000000003, 'CAA': 0.27749999999999997, 'CAC': 0.125, 'CAG': 0.32999999999999996, 'CAT': 0.07375000000000001, 'CCA': 0.13874999999999998, 'CCC': 0.125, 'CCG': 0.16499999999999998, 'CCT': 0.07375000000000001, 'CGA': 0.013799999999999993, 'CGC': 0.36, 'CGG': 0.125, 'CGT': 0.5, 'CTA': 0.18, 'CTC': 0.125, 'CTG': 0.54, 'CTT': 0.07375000000000001, 'GAA': 0.5, 'GAC': 0.375, 'GAG': 0.15999999999999998, 'GAT': 0.22125000000000003, 'GCA': 0.375, 'GCC': 0.25, 'GCG': 0.11999999999999998, 'GCT': 0.14750000000000002, 'GGA': 0.13874999999999998, 'GGC': 0.5, 'GGG': 0.16499999999999998, 'GGT': 0.29500000000000004, 'GTA': 0.625, 'GTC': 0.25, 'GTG': 0.19999999999999996, 'GTT': 0.14750000000000002, 'TAC': 0.375, 'TAT': 0.22125000000000003, 'TCA': 0.13874999999999998, 'TCC': 0.25, 'TCG': 0.16499999999999998, 'TCT': 0.14750000000000002, 'TGC': 0.125, 'TGG': 0.125, 'TGT': 0.07375000000000001, 'TTA': 0.13874999999999998, 'TTC': 0.25, 'TTG': 0.16499999999999998, 'TTT': 0.14750000000000002}, 'tai_avg': 0.2504950429983414, 'tai_std': 0.023481599705588187}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': None, 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.6666666666666666, 'AAC': 0.6666666666666666, 'AAG': 0.2133333333333333, 'AAT': 0.39333333333333337, 'ACA': 0.6666666666666666, 'ACC': 0.16666666666666666, 'ACG': 0.2133333333333333, 'ACT': 0.09833333333333334, 'AGA': 0.18499999999999997, 'AGC': 0.3333333333333333, 'AGG': 0.21999999999999997, 'AGT': 0.19666666666666668, 'ATA': 0.10999999999999999, 'ATC': 0.5, 'ATG': 1.0, 'ATT': 0.29500000000000004, 'CAA': 0.6666666666666666, 'CAC': 0.3333333333333333, 'CAG': 0.2133333333333333, 'CAT': 0.19666666666666668, 'CCA': 0.5, 'CCG': 0.15999999999999998, 'CGA': 0.01839999999999999, 'CGC': 0.48, 'CGG': 0.16666666666666666, 'CGT': 0.6666666666666666, 'CTA': 0.3516666666666666, 'CTC': 0.16666666666666666, 'CTG': 0.2733333333333333, 'CTT': 0.09833333333333334, 'GAA': 1.0, 'GAC': 0.6666666666666666, 'GAG': 0.31999999999999995, 'GAT': 0.39333333333333337, 'GCA': 0.8333333333333334, 'GCC': 0.16666666666666666, 'GCG': 0.2666666666666666, 'GCT': 0.09833333333333334, 'GGA': 0.5, 'GGC': 0.6666666666666666, 'GGG': 0.15999999999999998, 'GGT': 0.39333333333333337, 'GTA': 0.6666666666666666, 'GTC': 0.16666666666666666, 'GTG': 0.2133333333333333, 'GTT': 0.09833333333333334, 'TAC': 0.3333333333333333, 'TAT': 0.19666666666666668, 'TCA': 0.3333333333333333, 'TCC': 0.16666666666666666, 'TCG': 0.10666666666666665, 'TCT': 0.09833333333333334, 'TGC': 0.16666666666666666, 'TGG': 0.16666666666666666, 'TGT': 0.09833333333333334, 'TTA': 0.5183333333333333, 'TTC': 0.5, 'TTG': 0.32666666666666666, 'TTT': 0.29500000000000004}, 'tai_avg': 0.3292474875006557, 'tai_std': 0.027038419737753252}]}, 'orf_debug': {'C': defaultdict(<class 'float'>, {'TGT': 0.984009360374415, 'TGC': 1.0}), 'D': defaultdict(<class 'float'>, {'GAT': 0.906086387434555, 'GAC': 1.0}), 'S': defaultdict(<class 'float'>, {'TCT': 0.8927125506072876, 'TCG': 0.8845108695652174, 'TCA': 1.11125, 'TCC': 0.8571428571428571, 'AGC': 1.125, 'AGT': 1.0348240469208212}), 'Q': defaultdict(<class 'float'>, {'CAA': 1.0525, 'CAG': 0.6880733944954128}), 'M': defaultdict(<class 'float'>, {'ATG': 1.0}), 'N': defaultdict(<class 'float'>, {'AAC': 1.0, 'AAT': 0.9463350785340315}), 'P': defaultdict(<class 'float'>, {'CCT': 0.7275, 'CCG': 0.7462686567164178, 'CCA': 1.02625, 'CCC': 0.6933333333333334}), 'K': defaultdict(<class 'float'>, {'AAG': 1.0389908256880733, 'AAA': 1.0}), '_': defaultdict(<class 'float'>, {'TAG': 0.9271127128057303, 'TGA': 0.9271127128057303, 'TAA': 0.9271127128057303}), 'T': defaultdict(<class 'float'>, {'ACC': 0.6933333333333334, 'ACA': 1.1375, 'ACG': 0.6880733944954128, 'ACT': 0.7284803400637619}), 'F': defaultdict(<class 'float'>, {'TTT': 0.9149377593360996, 'TTC': 1.0}), 'A': defaultdict(<class 'float'>, {'GCA': 1.0, 'GCC': 0.6749999999999999, 'GCG': 0.8010638297872339, 'GCT': 0.7074927953890491}), 'G': defaultdict(<class 'float'>, {'GGT': 0.9463350785340315, 'GGG': 0.8860619469026548, 'GGA': 1.1667857142857145, 'GGC': 1.0}), 'I': defaultdict(<class 'float'>, {'ATC': 1.0, 'ATA': 0.9100719424460432, 'ATT': 0.9574688796680498}), 'L': defaultdict(<class 'float'>, {'TTA': 1.40125, 'TTG': 1.153846153846154, 'CTC': 1.0468557336621456, 'CTT': 1.0325704225352113, 'CTG': 0.8032128514056224, 'CTA': 1.1657142857142857}), 'H': defaultdict(<class 'float'>, {'CAT': 0.9248533724340176, 'CAC': 1.0}), 'R': defaultdict(<class 'float'>, {'CGA': 0.9016744863290729, 'CGC': 0.960674157303371, 'CGG': 0.9166666666666666, 'CGT': 1.0, 'AGG': 0.922811059907834, 'AGA': 0.9187289088863893}), 'W': defaultdict(<class 'float'>, {'TGG': 1.0}), 'V': defaultdict(<class 'float'>, {'GTA': 1.0, 'GTC': 0.9166666666666666, 'GTG': 0.9805045871559633, 'GTT': 0.9420828905419766}), 'E': defaultdict(<class 'float'>, {'GAG': 0.7976190476190476, 'GAA': 1.0}), 'Y': defaultdict(<class 'float'>, {'TAT': 1.0150293255131966, 'TAC': 1.0})}, 'orf': {'aa_to_optimal_codon': {'C': 'TGT', 'D': 'GAT', 'S': 'TCC', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCC', 'K': 'AAA', '_': 'TAA', 'T': 'ACG', 'F': 'TTT', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAT', 'R': 'CGA', 'W': 'TGG', 'V': 'GTC', 'E': 'GAG', 'Y': 'TAC'}, 'run_time': 0.062000000034458935}, 'evaluation': [{'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'tai_initial_score': 0.2799603826276079, 'tai_final_score': 0.21990240129720437, 'dist_score': -2.557661406522949}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'tai_initial_score': 0.33499296920046817, 'tai_final_score': 0.23805637984916864, 'dist_score': -3.585142559790532}], 'final_sequence': 'ATGGTCTCCAAAGGGGAGGAGGATAATATGGCCATAATAAAAGAGTTTATGCGATTTAAAGTCCATATGGAGGGGTCCGTCAATGGGCATGAGTTTGAGATAGAGGGGGAGGGGGAGGGGCGACCCTACGAGGGGACGCAGACGGCCAAACTGAAAGTCACGAAAGGGGGGCCCCTGCCCTTTGCCTGGGATATACTGTCCCCCCAGTTTATGTACGGGTCCAAAGCCTACGTCAAACATCCCGCCGATATACCCGATTACCTGAAACTGTCCTTTCCCGAGGGGTTTAAATGGGAGCGAGTCATGAATTTTGAGGATGGGGGGGTCGTCACGGTCACGCAGGATTCCTCCCTGCAGGATGGGGAGTTTATATACAAAGTCAAACTGCGAGGGACGAATTTTCCCTCCGATGGGCCCGTCATGCAGAAAAAAACGATGGGGTGGGAGGCCTCCTCCGAGCGAATGTACCCCGAGGATGGGGCCCTGAAAGGGGAGATAAAACAGCGACTGAAACTGAAAGATGGGGGGCATTACGATGCCGAGGTCAAAACGACGTACAAAGCCAAAAAACCCGTCCAGCTGCCCGGGGCCTACAATGTCAATATAAAACTGGATATAACGTCCCATAATGAGGATTACACGATAGTCGAGCAGTACGAGCGAGCCGAGGGGCGACATTCCACGGGGGGGATGGATGAGCTGTACAAATAA', 'average_distance_score': 0.5137405766337915, 'weakest_link_score': 0.5137405766337915}], 'final_evaluation': {'final_sequence': 'ATGGTCTCCAAAGGGGAGGAGGATAATATGGCCATAATAAAAGAGTTTATGCGATTTAAAGTCCATATGGAGGGGTCCGTCAATGGGCATGAGTTTGAGATAGAGGGGGAGGGGGAGGGGCGACCCTACGAGGGGACGCAGACGGCCAAACTGAAAGTCACGAAAGGGGGGCCCCTGCCCTTTGCCTGGGATATACTGTCCCCCCAGTTTATGTACGGGTCCAAAGCCTACGTCAAACATCCCGCCGATATACCCGATTACCTGAAACTGTCCTTTCCCGAGGGGTTTAAATGGGAGCGAGTCATGAATTTTGAGGATGGGGGGGTCGTCACGGTCACGCAGGATTCCTCCCTGCAGGATGGGGAGTTTATATACAAAGTCAAACTGCGAGGGACGAATTTTCCCTCCGATGGGCCCGTCATGCAGAAAAAAACGATGGGGTGGGAGGCCTCCTCCGAGCGAATGTACCCCGAGGATGGGGCCCTGAAAGGGGAGATAAAACAGCGACTGAAACTGAAAGATGGGGGGCATTACGATGCCGAGGTCAAAACGACGTACAAAGCCAAAAAACCCGTCCAGCTGCCCGGGGCCTACAATGTCAATATAAAACTGGATATAACGTCCCATAATGAGGATTACACGATAGTCGAGCAGTACGAGCGAGCCGAGGGGCGACATTCCACGGGGGGGATGGATGAGCTGTACAAATAA', 'average_distance_score': 0.5137405766337915, 'weakest_link_score': 0.5137405766337915}, 'zip_output_file_path': 'results\\\\mcherry_debug\\\\tAI_single_codon_ratio_ecoli_opt_True_7FgF\\\\communique_results.zip'}\n",
      "##########################\n",
      "# USER INPUT INFORMATION #\n",
      "##########################\n",
      "Organism is optimized\n",
      "Information about Bacillus subtilis:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Bacillus subtilis. Calculating tAI profile.\n",
      "name=Bacillus subtilis, tai_std=0.027038419737753252, tai_avg=0.3292474875006557\n",
      "Organism is de-optimized\n",
      "Information about Escherichia coli:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Escherichia coli. Calculating tAI profile.\n",
      "name=Escherichia coli, tai_std=0.023481599705588187, tai_avg=0.2504950429983414\n",
      "Bacillus subtilis has weight of 1.0\n",
      "Escherichia coli has weight of 1.0\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is: ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Total input processing time: 21.189903020858765\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "\n",
      "Optimization method used is: OptimizationMethod.single_codon_ratio\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGT': 1.0162504954419342, 'TGC': 1.0})\n",
      "Optimal codon for C is: TGC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAT': 1.1036475261827372, 'GAC': 1.0})\n",
      "Optimal codon for D is: GAC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TCT': 1.1201814058956914, 'TCG': 1.130568356374808, 'TCA': 0.8998875140607423, 'TCC': 1.1666666666666667, 'AGC': 0.8888888888888888, 'AGT': 0.9663478568898335})\n",
      "Optimal codon for S is: AGC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAA': 0.9501187648456058, 'CAG': 1.4533333333333334})\n",
      "Optimal codon for Q is: CAA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATG': 1.0})\n",
      "Optimal codon for M is: ATG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAC': 1.0, 'AAT': 1.0567081604426})\n",
      "Optimal codon for N is: AAC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CCT': 1.3745704467353952, 'CCG': 1.34, 'CCA': 0.9744214372716199, 'CCC': 1.4423076923076923})\n",
      "Optimal codon for P is: CCA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAG': 0.9624724061810155, 'AAA': 1.0})\n",
      "Optimal codon for K is: AAG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAG': 1.0786175037700543, 'TGA': 1.0786175037700543, 'TAA': 1.0786175037700543})\n",
      "Optimal codon for _ is: TAG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ACC': 1.4423076923076923, 'ACA': 0.8791208791208791, 'ACG': 1.4533333333333334, 'ACT': 1.3727206418672502})\n",
      "Optimal codon for T is: ACA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTT': 1.09297052154195, 'TTC': 1.0})\n",
      "Optimal codon for F is: TTC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GCA': 1.0, 'GCC': 1.4814814814814816, 'GCG': 1.2483399734395753, 'GCT': 1.4134419551934825})\n",
      "Optimal codon for A is: GCA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GGT': 1.0567081604426, 'GGG': 1.128589263420724, 'GGA': 0.8570554025099478, 'GGC': 1.0})\n",
      "Optimal codon for G is: GGA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATC': 1.0, 'ATA': 1.098814229249012, 'ATT': 1.0444203683640303})\n",
      "Optimal codon for I is: ATC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTA': 0.7136485280999108, 'TTG': 0.8666666666666667, 'CTC': 0.9552414605418138, 'CTT': 0.9684569479965899, 'CTG': 1.245, 'CTA': 0.8578431372549019})\n",
      "Optimal codon for L is: TTA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAT': 1.081252477209671, 'CAC': 1.0})\n",
      "Optimal codon for H is: CAC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CGA': 1.1090476831292335, 'CGC': 1.0409356725146197, 'CGG': 1.0909090909090908, 'CGT': 1.0, 'AGG': 1.0836454431960052, 'AGA': 1.0884603611876338})\n",
      "Optimal codon for R is: CGT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGG': 1.0})\n",
      "Optimal codon for W is: TGG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GTA': 1.0, 'GTC': 1.0909090909090908, 'GTG': 1.0198830409356725, 'GTT': 1.0614777213761986})\n",
      "Optimal codon for V is: GTA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAG': 1.2537313432835822, 'GAA': 1.0})\n",
      "Optimal codon for E is: GAA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAT': 0.9851932105453233, 'TAC': 1.0})\n",
      "Optimal codon for Y is: TAT\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTAAGCAAGGGAGAAGAAGACAACATGGCAATCATCAAGGAATTCATGCGTTTCAAGGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATATGAAGGAACACAAACAGCAAAGTTAAAGGTAACAAAGGGAGGACCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTATGGAAGCAAGGCATATGTAAAGCACCCAGCAGACATCCCAGACTATTTAAAGTTAAGCTTCCCAGAAGGATTCAAGTGGGAACGTGTAATGAACTTCGAAGACGGAGGAGTAGTAACAGTAACACAAGACAGCAGCTTACAAGACGGAGAATTCATCTATAAGGTAAAGTTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAGAAGACAATGGGATGGGAAGCAAGCAGCGAACGTATGTATCCAGAAGACGGAGCATTAAAGGGAGAAATCAAGCAACGTTTAAAGTTAAAGGACGGAGGACACTATGACGCAGAAGTAAAGACAACATATAAGGCAAAGAAGCCAGTACAATTACCAGGAGCATATAACGTAAACATCAAGTTAGACATCACAAGCCACAACGAAGACTATACAATCGTAGAACAATATGAACGTGCAGAAGGACGTCACAGCACAGGAGGAATGGACGAATTATATAAGTAA', 'average_distance_score': 3.632588568818941, 'weakest_link_score': 3.632588568818941}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTAAGCAAGGGAGAAGAAGACAACATGGCAATCATCAAGGAATTCATGCGTTTCAAGGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATATGAAGGAACACAAACAGCAAAGTTAAAGGTAACAAAGGGAGGACCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTATGGAAGCAAGGCATATGTAAAGCACCCAGCAGACATCCCAGACTATTTAAAGTTAAGCTTCCCAGAAGGATTCAAGTGGGAACGTGTAATGAACTTCGAAGACGGAGGAGTAGTAACAGTAACACAAGACAGCAGCTTACAAGACGGAGAATTCATCTATAAGGTAAAGTTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAGAAGACAATGGGATGGGAAGCAAGCAGCGAACGTATGTATCCAGAAGACGGAGCATTAAAGGGAGAAATCAAGCAACGTTTAAAGTTAAAGGACGGAGGACACTATGACGCAGAAGTAAAGACAACATATAAGGCAAAGAAGCCAGTACAATTACCAGGAGCATATAACGTAAACATCAAGTTAGACATCACAAGCCACAACGAAGACTATACAATCGTAGAACAATATGAACGTGCAGAAGGACGTCACAGCACAGGAGGAATGGACGAATTATATAAGTAA', 'average_distance_score': 3.632588568818941, 'weakest_link_score': 3.632588568818941}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry_debug\\tAI_single_codon_ratio_ecoli_opt_False_EVbh\n",
      "Final output: {'user_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'optimization_method': 'single_codon_ratio', 'optimization_cub_index': 'tAI', 'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': None, 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.6666666666666666, 'AAC': 0.6666666666666666, 'AAG': 0.2133333333333333, 'AAT': 0.39333333333333337, 'ACA': 0.6666666666666666, 'ACC': 0.16666666666666666, 'ACG': 0.2133333333333333, 'ACT': 0.09833333333333334, 'AGA': 0.18499999999999997, 'AGC': 0.3333333333333333, 'AGG': 0.21999999999999997, 'AGT': 0.19666666666666668, 'ATA': 0.10999999999999999, 'ATC': 0.5, 'ATG': 1.0, 'ATT': 0.29500000000000004, 'CAA': 0.6666666666666666, 'CAC': 0.3333333333333333, 'CAG': 0.2133333333333333, 'CAT': 0.19666666666666668, 'CCA': 0.5, 'CCG': 0.15999999999999998, 'CGA': 0.01839999999999999, 'CGC': 0.48, 'CGG': 0.16666666666666666, 'CGT': 0.6666666666666666, 'CTA': 0.3516666666666666, 'CTC': 0.16666666666666666, 'CTG': 0.2733333333333333, 'CTT': 0.09833333333333334, 'GAA': 1.0, 'GAC': 0.6666666666666666, 'GAG': 0.31999999999999995, 'GAT': 0.39333333333333337, 'GCA': 0.8333333333333334, 'GCC': 0.16666666666666666, 'GCG': 0.2666666666666666, 'GCT': 0.09833333333333334, 'GGA': 0.5, 'GGC': 0.6666666666666666, 'GGG': 0.15999999999999998, 'GGT': 0.39333333333333337, 'GTA': 0.6666666666666666, 'GTC': 0.16666666666666666, 'GTG': 0.2133333333333333, 'GTT': 0.09833333333333334, 'TAC': 0.3333333333333333, 'TAT': 0.19666666666666668, 'TCA': 0.3333333333333333, 'TCC': 0.16666666666666666, 'TCG': 0.10666666666666665, 'TCT': 0.09833333333333334, 'TGC': 0.16666666666666666, 'TGG': 0.16666666666666666, 'TGT': 0.09833333333333334, 'TTA': 0.5183333333333333, 'TTC': 0.5, 'TTG': 0.32666666666666666, 'TTT': 0.29500000000000004}, 'tai_avg': 0.3292474875006557, 'tai_std': 0.027038419737753252}, {'name': 'Escherichia coli', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': None, 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.75, 'AAC': 0.5, 'AAG': 0.23999999999999996, 'AAT': 0.29500000000000004, 'ACA': 0.1525, 'ACC': 0.25, 'ACG': 0.29, 'ACT': 0.14750000000000002, 'AGA': 0.13874999999999998, 'AGC': 0.125, 'AGG': 0.16499999999999998, 'AGT': 0.07375000000000001, 'ATA': 0.10999999999999999, 'ATC': 0.375, 'ATG': 1.0, 'ATT': 0.22125000000000003, 'CAA': 0.27749999999999997, 'CAC': 0.125, 'CAG': 0.32999999999999996, 'CAT': 0.07375000000000001, 'CCA': 0.13874999999999998, 'CCC': 0.125, 'CCG': 0.16499999999999998, 'CCT': 0.07375000000000001, 'CGA': 0.013799999999999993, 'CGC': 0.36, 'CGG': 0.125, 'CGT': 0.5, 'CTA': 0.18, 'CTC': 0.125, 'CTG': 0.54, 'CTT': 0.07375000000000001, 'GAA': 0.5, 'GAC': 0.375, 'GAG': 0.15999999999999998, 'GAT': 0.22125000000000003, 'GCA': 0.375, 'GCC': 0.25, 'GCG': 0.11999999999999998, 'GCT': 0.14750000000000002, 'GGA': 0.13874999999999998, 'GGC': 0.5, 'GGG': 0.16499999999999998, 'GGT': 0.29500000000000004, 'GTA': 0.625, 'GTC': 0.25, 'GTG': 0.19999999999999996, 'GTT': 0.14750000000000002, 'TAC': 0.375, 'TAT': 0.22125000000000003, 'TCA': 0.13874999999999998, 'TCC': 0.25, 'TCG': 0.16499999999999998, 'TCT': 0.14750000000000002, 'TGC': 0.125, 'TGG': 0.125, 'TGT': 0.07375000000000001, 'TTA': 0.13874999999999998, 'TTC': 0.25, 'TTG': 0.16499999999999998, 'TTT': 0.14750000000000002}, 'tai_avg': 0.2504950429983414, 'tai_std': 0.023481599705588187}]}, 'orf_debug': {'C': defaultdict(<class 'float'>, {'TGT': 1.0162504954419342, 'TGC': 1.0}), 'D': defaultdict(<class 'float'>, {'GAT': 1.1036475261827372, 'GAC': 1.0}), 'S': defaultdict(<class 'float'>, {'TCT': 1.1201814058956914, 'TCG': 1.130568356374808, 'TCA': 0.8998875140607423, 'TCC': 1.1666666666666667, 'AGC': 0.8888888888888888, 'AGT': 0.9663478568898335}), 'Q': defaultdict(<class 'float'>, {'CAA': 0.9501187648456058, 'CAG': 1.4533333333333334}), 'M': defaultdict(<class 'float'>, {'ATG': 1.0}), 'N': defaultdict(<class 'float'>, {'AAC': 1.0, 'AAT': 1.0567081604426}), 'P': defaultdict(<class 'float'>, {'CCT': 1.3745704467353952, 'CCG': 1.34, 'CCA': 0.9744214372716199, 'CCC': 1.4423076923076923}), 'K': defaultdict(<class 'float'>, {'AAG': 0.9624724061810155, 'AAA': 1.0}), '_': defaultdict(<class 'float'>, {'TAG': 1.0786175037700543, 'TGA': 1.0786175037700543, 'TAA': 1.0786175037700543}), 'T': defaultdict(<class 'float'>, {'ACC': 1.4423076923076923, 'ACA': 0.8791208791208791, 'ACG': 1.4533333333333334, 'ACT': 1.3727206418672502}), 'F': defaultdict(<class 'float'>, {'TTT': 1.09297052154195, 'TTC': 1.0}), 'A': defaultdict(<class 'float'>, {'GCA': 1.0, 'GCC': 1.4814814814814816, 'GCG': 1.2483399734395753, 'GCT': 1.4134419551934825}), 'G': defaultdict(<class 'float'>, {'GGT': 1.0567081604426, 'GGG': 1.128589263420724, 'GGA': 0.8570554025099478, 'GGC': 1.0}), 'I': defaultdict(<class 'float'>, {'ATC': 1.0, 'ATA': 1.098814229249012, 'ATT': 1.0444203683640303}), 'L': defaultdict(<class 'float'>, {'TTA': 0.7136485280999108, 'TTG': 0.8666666666666667, 'CTC': 0.9552414605418138, 'CTT': 0.9684569479965899, 'CTG': 1.245, 'CTA': 0.8578431372549019}), 'H': defaultdict(<class 'float'>, {'CAT': 1.081252477209671, 'CAC': 1.0}), 'R': defaultdict(<class 'float'>, {'CGA': 1.1090476831292335, 'CGC': 1.0409356725146197, 'CGG': 1.0909090909090908, 'CGT': 1.0, 'AGG': 1.0836454431960052, 'AGA': 1.0884603611876338}), 'W': defaultdict(<class 'float'>, {'TGG': 1.0}), 'V': defaultdict(<class 'float'>, {'GTA': 1.0, 'GTC': 1.0909090909090908, 'GTG': 1.0198830409356725, 'GTT': 1.0614777213761986}), 'E': defaultdict(<class 'float'>, {'GAG': 1.2537313432835822, 'GAA': 1.0}), 'Y': defaultdict(<class 'float'>, {'TAT': 0.9851932105453233, 'TAC': 1.0})}, 'orf': {'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCA', 'K': 'AAG', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAT'}, 'run_time': 0.06299999996554106}, 'evaluation': [{'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'tai_initial_score': 0.33499296920046817, 'tai_final_score': 0.5040673316441382, 'dist_score': 6.2531155327688985}, {'name': 'Escherichia coli', 'is_wanted': False, 'tai_initial_score': 0.2799603826276079, 'tai_final_score': 0.2561955571446793, 'dist_score': -1.0120616048689834}], 'final_sequence': 'ATGGTAAGCAAGGGAGAAGAAGACAACATGGCAATCATCAAGGAATTCATGCGTTTCAAGGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATATGAAGGAACACAAACAGCAAAGTTAAAGGTAACAAAGGGAGGACCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTATGGAAGCAAGGCATATGTAAAGCACCCAGCAGACATCCCAGACTATTTAAAGTTAAGCTTCCCAGAAGGATTCAAGTGGGAACGTGTAATGAACTTCGAAGACGGAGGAGTAGTAACAGTAACACAAGACAGCAGCTTACAAGACGGAGAATTCATCTATAAGGTAAAGTTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAGAAGACAATGGGATGGGAAGCAAGCAGCGAACGTATGTATCCAGAAGACGGAGCATTAAAGGGAGAAATCAAGCAACGTTTAAAGTTAAAGGACGGAGGACACTATGACGCAGAAGTAAAGACAACATATAAGGCAAAGAAGCCAGTACAATTACCAGGAGCATATAACGTAAACATCAAGTTAGACATCACAAGCCACAACGAAGACTATACAATCGTAGAACAATATGAACGTGCAGAAGGACGTCACAGCACAGGAGGAATGGACGAATTATATAAGTAA', 'average_distance_score': 3.632588568818941, 'weakest_link_score': 3.632588568818941}], 'final_evaluation': {'final_sequence': 'ATGGTAAGCAAGGGAGAAGAAGACAACATGGCAATCATCAAGGAATTCATGCGTTTCAAGGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATATGAAGGAACACAAACAGCAAAGTTAAAGGTAACAAAGGGAGGACCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTATGGAAGCAAGGCATATGTAAAGCACCCAGCAGACATCCCAGACTATTTAAAGTTAAGCTTCCCAGAAGGATTCAAGTGGGAACGTGTAATGAACTTCGAAGACGGAGGAGTAGTAACAGTAACACAAGACAGCAGCTTACAAGACGGAGAATTCATCTATAAGGTAAAGTTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAGAAGACAATGGGATGGGAAGCAAGCAGCGAACGTATGTATCCAGAAGACGGAGCATTAAAGGGAGAAATCAAGCAACGTTTAAAGTTAAAGGACGGAGGACACTATGACGCAGAAGTAAAGACAACATATAAGGCAAAGAAGCCAGTACAATTACCAGGAGCATATAACGTAAACATCAAGTTAGACATCACAAGCCACAACGAAGACTATACAATCGTAGAACAATATGAACGTGCAGAAGGACGTCACAGCACAGGAGGAATGGACGAATTATATAAGTAA', 'average_distance_score': 3.632588568818941, 'weakest_link_score': 3.632588568818941}, 'zip_output_file_path': 'results\\\\mcherry_debug\\\\tAI_single_codon_ratio_ecoli_opt_False_EVbh\\\\communique_results.zip'}\n",
      "##########################\n",
      "# USER INPUT INFORMATION #\n",
      "##########################\n",
      "Organism is optimized\n",
      "Information about Escherichia coli:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 90 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3895\n",
      "Calculate CAI weights from a reference set of 584 highly expressed genes from estimated expression dictionary.\n",
      "name=Escherichia coli, cai_std=0.09192830548402788, cai_avg=0.6260311851192147\n",
      "Organism is de-optimized\n",
      "Information about Bacillus subtilis:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 98 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3550\n",
      "Calculate CAI weights from a reference set of 532 highly expressed genes from estimated expression dictionary.\n",
      "name=Bacillus subtilis, cai_std=0.04172052669683352, cai_avg=0.7354163127574911\n",
      "Escherichia coli has weight of 1.0\n",
      "Bacillus subtilis has weight of 1.0\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is: ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Total input processing time: 20.491962671279907\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "\n",
      "Optimization method used is: OptimizationMethod.single_codon_diff\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGT': 0.3211906255307573, 'TGC': 0.5})\n",
      "Optimal codon for C is: TGT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAT': 0.5, 'GAC': 0.23687560010072853})\n",
      "Optimal codon for D is: GAC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TCT': 0.5060311925144046, 'TCG': 0.2073679268306709, 'TCA': 0.6916739915026691, 'TCC': 0.1118499460028256, 'AGC': 0.39020967201368356, 'AGT': 0.2929009723033402})\n",
      "Optimal codon for S is: TCC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAA': 0.7104741026059154, 'CAG': 0.340446520871006})\n",
      "Optimal codon for Q is: CAG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATG': 0.5})\n",
      "Optimal codon for M is: ATG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAC': 0.5, 'AAT': 0.5237602439432804})\n",
      "Optimal codon for N is: AAC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CCT': 0.6106120395061632, 'CCG': 0.5, 'CCA': 0.38903801323851617, 'CCC': 0.41837460463408405})\n",
      "Optimal codon for P is: CCA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAG': 0.3230909391410995, 'AAA': 0.5})\n",
      "Optimal codon for K is: AAG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAG': 0.4694160602254803, 'TGA': 0.3017676581338576, 'TAA': 0.5})\n",
      "Optimal codon for _ is: TGA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ACC': 0.02451494093501815, 'ACA': 0.8637893561320659, 'ACG': 0.3560459226676872, 'ACT': 0.2507406971478782})\n",
      "Optimal codon for T is: ACC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTT': 0.5483221374810254, 'TTC': 0.3395085681742403})\n",
      "Optimal codon for F is: TTC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GCA': 0.5593742233640543, 'GCC': 0.19364866442064205, 'GCG': 0.42289172693334515, 'GCT': 0.507919645950722})\n",
      "Optimal codon for A is: GCC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GGT': 0.1904519673432688, 'GGG': 0.37018836174414865, 'GGA': 0.7461118741928667, 'GGC': 0.5})\n",
      "Optimal codon for G is: GGT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATC': 0.35170277683672996, 'ATA': 0.4708846413066058, 'ATT': 0.5296706286626713})\n",
      "Optimal codon for I is: ATC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTA': 0.5736418406415243, 'TTG': 0.46924759681733225, 'CTC': 0.42191519254649945, 'CTT': 0.8909703268297648, 'CTG': 0.2204998619360393, 'CTA': 0.4855892438855689})\n",
      "Optimal codon for L is: CTG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAT': 0.5375491340083262, 'CAC': 0.22971534229470658})\n",
      "Optimal codon for H is: CAC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CGA': 0.48369559277060226, 'CGC': 0.525186390131594, 'CGG': 0.4834154409858235, 'CGT': 0.5, 'AGG': 0.4936030479051473, 'AGA': 0.74049856314793})\n",
      "Optimal codon for R is: CGG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGG': 0.5})\n",
      "Optimal codon for W is: TGG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GTA': 0.37907854176555605, 'GTC': 0.3531969781543418, 'GTG': 0.22238959863999305, 'GTT': 0.5})\n",
      "Optimal codon for V is: GTG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAG': 0.279115775339857, 'GAA': 0.5})\n",
      "Optimal codon for E is: GAG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAT': 0.5100827409752241, 'TAC': 0.30321769179564306})\n",
      "Optimal codon for Y is: TAC\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTGTCCAAGGGTGAGGAGGACAACATGGCCATCATCAAGGAGTTCATGCGGTTCAAGGTGCACATGGAGGGTTCCGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGGCCATACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGTCCACTGCCATTCGCCTGGGACATCCTGTCCCCACAGTTCATGTACGGTTCCAAGGCCTACGTGAAGCACCCAGCCGACATCCCAGACTACCTGAAGCTGTCCTTCCCAGAGGGTTTCAAGTGGGAGCGGGTGATGAACTTCGAGGACGGTGGTGTGGTGACCGTGACCCAGGACTCCTCCCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGGGGTACCAACTTCCCATCCGACGGTCCAGTGATGCAGAAGAAGACCATGGGTTGGGAGGCCTCCTCCGAGCGGATGTACCCAGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGGCTGAAGCTGAAGGACGGTGGTCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCAGTGCAGCTGCCAGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCCCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGGGCCGAGGGTCGGCACTCCACCGGTGGTATGGACGAGCTGTACAAGTGA', 'average_distance_score': 1.9223416370411865, 'weakest_link_score': 1.9223416370411865}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTGTCCAAGGGTGAGGAGGACAACATGGCCATCATCAAGGAGTTCATGCGGTTCAAGGTGCACATGGAGGGTTCCGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGGCCATACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGTCCACTGCCATTCGCCTGGGACATCCTGTCCCCACAGTTCATGTACGGTTCCAAGGCCTACGTGAAGCACCCAGCCGACATCCCAGACTACCTGAAGCTGTCCTTCCCAGAGGGTTTCAAGTGGGAGCGGGTGATGAACTTCGAGGACGGTGGTGTGGTGACCGTGACCCAGGACTCCTCCCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGGGGTACCAACTTCCCATCCGACGGTCCAGTGATGCAGAAGAAGACCATGGGTTGGGAGGCCTCCTCCGAGCGGATGTACCCAGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGGCTGAAGCTGAAGGACGGTGGTCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCAGTGCAGCTGCCAGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCCCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGGGCCGAGGGTCGGCACTCCACCGGTGGTATGGACGAGCTGTACAAGTGA', 'average_distance_score': 1.9223416370411865, 'weakest_link_score': 1.9223416370411865}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry_debug\\CAI_single_codon_diff_ecoli_opt_True_dxlH\n",
      "Final output: {'user_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'optimization_method': 'single_codon_diff', 'optimization_cub_index': 'CAI', 'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.037037037037037035, 'TGA': 0.24183006535947713, 'GCA': 0.6554010349288486, 'GCC': 0.6652652005174644, 'GCG': 1.0, 'GCT': 0.6266170763260025, 'TGC': 1.0, 'TGT': 0.7162162162162162, 'GAC': 0.8422464375523889, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.36087048383688997, 'TTC': 1.0, 'TTT': 0.6891233766233766, 'GGA': 0.12073112073112073, 'GGC': 1.0, 'GGG': 0.19111752445085778, 'GGT': 0.9866923200256533, 'CAC': 1.0, 'CAT': 0.7259593679458239, 'ATA': 0.03593829462620784, 'ATC': 1.0, 'ATT': 0.7563993897270725, 'AAA': 1.0, 'AAG': 0.2715144230769231, 'CTA': 0.0255237398225231, 'CTC': 0.1333821242338304, 'CTG': 1.0, 'CTT': 0.11572591711645777, 'TTA': 0.10557131095050773, 'TTG': 0.12322751806788033, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.4079954093343534, 'CCA': 0.2681349767629825, 'CCC': 0.08830066680137402, 'CCG': 1.0, 'CCT': 0.19337239846433624, 'CAA': 0.3511947863866763, 'CAG': 1.0, 'AGA': 0.018922018348623854, 'AGG': 0.011085626911314986, 'CGA': 0.03631498470948013, 'CGC': 0.7138761467889909, 'CGG': 0.05600152905198777, 'CGT': 1.0, 'AGC': 1.0, 'AGT': 0.3131715081575806, 'TCA': 0.3048149621965778, 'TCC': 0.7942697970553123, 'TCG': 0.40907282132908873, 'TCT': 0.8901711102268206, 'ACA': 0.14701775384001597, 'ACC': 1.0, 'ACG': 0.3457011769399561, 'ACT': 0.4374625972471574, 'GTA': 0.531166518254675, 'GTC': 0.5075690115761353, 'GTG': 0.9971059661620658, 'GTT': 1.0, 'TGG': 1.0, 'TAC': 1.0, 'TAT': 0.857994781960492}, 'cai_avg': 0.6260311851192147, 'cai_std': 0.09192830548402788, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.10739856801909307, 'TGA': 0.1694510739856802, 'GCA': 1.0, 'GCC': 0.5246426811237062, 'GCG': 0.9196648595367175, 'GCT': 0.9361754558896007, 'TGC': 1.0, 'TGT': 0.7495652173913042, 'GAC': 0.6699738903394257, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.3869690068323596, 'TTC': 0.82402496099844, 'TTT': 1.0, 'GGA': 0.8480035296712993, 'GGC': 1.0, 'GGG': 0.29340392675932053, 'GGT': 0.6170306640194132, 'CAC': 0.6778131634819533, 'CAT': 1.0, 'ATA': 0.11114994757077944, 'ATC': 0.8386927647675637, 'ATT': 1.0, 'AAA': 1.0, 'AAG': 0.3398391420911528, 'CTA': 0.14688262702462837, 'CTC': 0.3046372309740404, 'CTG': 0.6640781007321944, 'CTT': 1.0, 'TTA': 0.5893055247392945, 'TTG': 0.41202573774129136, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.8348958333333333, 'CCA': 0.4923917284432306, 'CCC': 0.07452204447912603, 'CCG': 1.0, 'CCT': 0.7553648068669527, 'CAA': 1.0, 'CAG': 0.8251624335499115, 'AGA': 0.72005772005772, 'AGG': 0.09620009620009619, 'CGA': 0.19672919672919673, 'CGC': 0.984126984126984, 'CGG': 0.2751322751322751, 'CGT': 1.0, 'AGC': 0.8834134615384617, 'AGT': 0.33774038461538464, 'TCA': 0.9487179487179488, 'TCC': 0.42588141025641024, 'TCG': 0.2560096153846154, 'TCT': 1.0, 'ACA': 1.0, 'ACC': 0.22142692218887092, 'ACG': 0.5329023320249365, 'ACT': 0.4301546986839067, 'GTA': 0.7337249143416544, 'GTC': 0.681106216348507, 'GTG': 0.6669114047968673, 'GTT': 1.0, 'TGG': 1.0, 'TAC': 0.7787396121883656, 'TAT': 1.0}, 'cai_avg': 0.7354163127574911, 'cai_std': 0.04172052669683352, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}]}, 'orf_debug': {'C': defaultdict(<class 'float'>, {'TGT': 0.3211906255307573, 'TGC': 0.5}), 'D': defaultdict(<class 'float'>, {'GAT': 0.5, 'GAC': 0.23687560010072853}), 'S': defaultdict(<class 'float'>, {'TCT': 0.5060311925144046, 'TCG': 0.2073679268306709, 'TCA': 0.6916739915026691, 'TCC': 0.1118499460028256, 'AGC': 0.39020967201368356, 'AGT': 0.2929009723033402}), 'Q': defaultdict(<class 'float'>, {'CAA': 0.7104741026059154, 'CAG': 0.340446520871006}), 'M': defaultdict(<class 'float'>, {'ATG': 0.5}), 'N': defaultdict(<class 'float'>, {'AAC': 0.5, 'AAT': 0.5237602439432804}), 'P': defaultdict(<class 'float'>, {'CCT': 0.6106120395061632, 'CCG': 0.5, 'CCA': 0.38903801323851617, 'CCC': 0.41837460463408405}), 'K': defaultdict(<class 'float'>, {'AAG': 0.3230909391410995, 'AAA': 0.5}), '_': defaultdict(<class 'float'>, {'TAG': 0.4694160602254803, 'TGA': 0.3017676581338576, 'TAA': 0.5}), 'T': defaultdict(<class 'float'>, {'ACC': 0.02451494093501815, 'ACA': 0.8637893561320659, 'ACG': 0.3560459226676872, 'ACT': 0.2507406971478782}), 'F': defaultdict(<class 'float'>, {'TTT': 0.5483221374810254, 'TTC': 0.3395085681742403}), 'A': defaultdict(<class 'float'>, {'GCA': 0.5593742233640543, 'GCC': 0.19364866442064205, 'GCG': 0.42289172693334515, 'GCT': 0.507919645950722}), 'G': defaultdict(<class 'float'>, {'GGT': 0.1904519673432688, 'GGG': 0.37018836174414865, 'GGA': 0.7461118741928667, 'GGC': 0.5}), 'I': defaultdict(<class 'float'>, {'ATC': 0.35170277683672996, 'ATA': 0.4708846413066058, 'ATT': 0.5296706286626713}), 'L': defaultdict(<class 'float'>, {'TTA': 0.5736418406415243, 'TTG': 0.46924759681733225, 'CTC': 0.42191519254649945, 'CTT': 0.8909703268297648, 'CTG': 0.2204998619360393, 'CTA': 0.4855892438855689}), 'H': defaultdict(<class 'float'>, {'CAT': 0.5375491340083262, 'CAC': 0.22971534229470658}), 'R': defaultdict(<class 'float'>, {'CGA': 0.48369559277060226, 'CGC': 0.525186390131594, 'CGG': 0.4834154409858235, 'CGT': 0.5, 'AGG': 0.4936030479051473, 'AGA': 0.74049856314793}), 'W': defaultdict(<class 'float'>, {'TGG': 0.5}), 'V': defaultdict(<class 'float'>, {'GTA': 0.37907854176555605, 'GTC': 0.3531969781543418, 'GTG': 0.22238959863999305, 'GTT': 0.5}), 'E': defaultdict(<class 'float'>, {'GAG': 0.279115775339857, 'GAA': 0.5}), 'Y': defaultdict(<class 'float'>, {'TAT': 0.5100827409752241, 'TAC': 0.30321769179564306})}, 'orf': {'aa_to_optimal_codon': {'C': 'TGT', 'D': 'GAC', 'S': 'TCC', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCA', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATC', 'L': 'CTG', 'H': 'CAC', 'R': 'CGG', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'run_time': 0.0779999999795109}, 'evaluation': [{'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'cai_initial_score': 0.7891479965436429, 'cai_final_score': 0.6228454402648508, 'dist_score': -1.8090462497177913}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'cai_initial_score': 0.7608140098814632, 'cai_final_score': 0.5249374363470826, 'dist_score': -5.653729523800164}], 'final_sequence': 'ATGGTGTCCAAGGGTGAGGAGGACAACATGGCCATCATCAAGGAGTTCATGCGGTTCAAGGTGCACATGGAGGGTTCCGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGGCCATACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGTCCACTGCCATTCGCCTGGGACATCCTGTCCCCACAGTTCATGTACGGTTCCAAGGCCTACGTGAAGCACCCAGCCGACATCCCAGACTACCTGAAGCTGTCCTTCCCAGAGGGTTTCAAGTGGGAGCGGGTGATGAACTTCGAGGACGGTGGTGTGGTGACCGTGACCCAGGACTCCTCCCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGGGGTACCAACTTCCCATCCGACGGTCCAGTGATGCAGAAGAAGACCATGGGTTGGGAGGCCTCCTCCGAGCGGATGTACCCAGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGGCTGAAGCTGAAGGACGGTGGTCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCAGTGCAGCTGCCAGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCCCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGGGCCGAGGGTCGGCACTCCACCGGTGGTATGGACGAGCTGTACAAGTGA', 'average_distance_score': 1.9223416370411865, 'weakest_link_score': 1.9223416370411865}], 'final_evaluation': {'final_sequence': 'ATGGTGTCCAAGGGTGAGGAGGACAACATGGCCATCATCAAGGAGTTCATGCGGTTCAAGGTGCACATGGAGGGTTCCGTGAACGGTCACGAGTTCGAGATCGAGGGTGAGGGTGAGGGTCGGCCATACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGTCCACTGCCATTCGCCTGGGACATCCTGTCCCCACAGTTCATGTACGGTTCCAAGGCCTACGTGAAGCACCCAGCCGACATCCCAGACTACCTGAAGCTGTCCTTCCCAGAGGGTTTCAAGTGGGAGCGGGTGATGAACTTCGAGGACGGTGGTGTGGTGACCGTGACCCAGGACTCCTCCCTGCAGGACGGTGAGTTCATCTACAAGGTGAAGCTGCGGGGTACCAACTTCCCATCCGACGGTCCAGTGATGCAGAAGAAGACCATGGGTTGGGAGGCCTCCTCCGAGCGGATGTACCCAGAGGACGGTGCCCTGAAGGGTGAGATCAAGCAGCGGCTGAAGCTGAAGGACGGTGGTCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCAGTGCAGCTGCCAGGTGCCTACAACGTGAACATCAAGCTGGACATCACCTCCCACAACGAGGACTACACCATCGTGGAGCAGTACGAGCGGGCCGAGGGTCGGCACTCCACCGGTGGTATGGACGAGCTGTACAAGTGA', 'average_distance_score': 1.9223416370411865, 'weakest_link_score': 1.9223416370411865}, 'zip_output_file_path': 'results\\\\mcherry_debug\\\\CAI_single_codon_diff_ecoli_opt_True_dxlH\\\\communique_results.zip'}\n",
      "##########################\n",
      "# USER INPUT INFORMATION #\n",
      "##########################\n",
      "Organism is optimized\n",
      "Information about Bacillus subtilis:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 98 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3550\n",
      "Calculate CAI weights from a reference set of 532 highly expressed genes from estimated expression dictionary.\n",
      "name=Bacillus subtilis, cai_std=0.04172052669683352, cai_avg=0.7354163127574911\n",
      "Organism is de-optimized\n",
      "Information about Escherichia coli:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 90 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3895\n",
      "Calculate CAI weights from a reference set of 584 highly expressed genes from estimated expression dictionary.\n",
      "name=Escherichia coli, cai_std=0.09192830548402788, cai_avg=0.6260311851192147\n",
      "Bacillus subtilis has weight of 1.0\n",
      "Escherichia coli has weight of 1.0\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is: ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Total input processing time: 20.503247022628784\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "\n",
      "Optimization method used is: OptimizationMethod.single_codon_diff\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGT': 0.2878416243556693, 'TGC': 0.5})\n",
      "Optimal codon for C is: TGT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAT': 0.5, 'GAC': 0.4091481473136918})\n",
      "Optimal codon for D is: GAC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TCT': 0.3962023027412252, 'TCG': 0.36043113277514427, 'TCA': 0.047771004981298035, 'TCC': 0.4802383328017277, 'AGC': 0.5067962104752218, 'AGT': 0.2683320958455362})\n",
      "Optimal codon for S is: TCA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAA': 0.06166888899259159, 'CAG': 0.5152840873210945})\n",
      "Optimal codon for Q is: CAA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATG': 0.5})\n",
      "Optimal codon for M is: ATG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAC': 0.5, 'AAT': 0.09685981994430053})\n",
      "Optimal codon for N is: AAT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CCT': 0.04861963110354668, 'CCG': 0.5, 'CCA': 0.16478126155826806, 'CCC': 0.43215322695633207})\n",
      "Optimal codon for P is: CCT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAG': 0.2547662201268697, 'AAA': 0.5})\n",
      "Optimal codon for K is: AAG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAG': 0.3990545292434242, 'TGA': 0.37414664950765447, 'TAA': 0.5})\n",
      "Optimal codon for _ is: TGA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ACC': 0.8030880187461472, 'ACA': 0.010807109972081766, 'ACG': 0.16884476758270672, 'ACT': 0.2580485957111289})\n",
      "Optimal codon for T is: ACA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTT': 0.2374455141044021, 'TTC': 0.5154836071758003})\n",
      "Optimal codon for F is: TTT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GCA': 0.21477525829290292, 'GCC': 0.33427118381440035, 'GCG': 0.5032268673966277, 'GCT': 0.19836126638712376})\n",
      "Optimal codon for A is: GCT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GGT': 0.560113623349509, 'GGG': 0.26790195943568595, 'GGA': 0.018839465252688016, 'GGC': 0.5})\n",
      "Optimal codon for G is: GGA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATC': 0.5130100120691663, 'ATA': 0.39567298836203413, 'ATT': 0.2860700183897439})\n",
      "Optimal codon for I is: ATT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTA': 0.0899076268527375, 'TTG': 0.18044937714392123, 'CTC': 0.2506600858062895, 'CTT': 0.006696243946222627, 'CTG': 0.5564217612038449, 'CTA': 0.3642303566834636})\n",
      "Optimal codon for L is: CTT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAT': 0.2635085019541501, 'CAC': 0.5519021788127533})\n",
      "Optimal codon for H is: CAT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CGA': 0.32328138075088564, 'CGC': 0.254935552793601, 'CGG': 0.26428469490553613, 'CGT': 0.5, 'AGG': 0.4084885786163661, 'AGA': 0.039362861438833774})\n",
      "Optimal codon for R is: AGA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGG': 0.5})\n",
      "Optimal codon for W is: TGG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GTA': 0.1765201456785766, 'GTC': 0.1796597733819701, 'GTG': 0.5525841600051915, 'GTT': 0.5})\n",
      "Optimal codon for V is: GTA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAG': 0.2530172523443873, 'GAA': 0.5})\n",
      "Optimal codon for E is: GAG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAT': 0.36807752293571605, 'TAC': 0.5244780796072774})\n",
      "Optimal codon for Y is: TAT\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTATCAAAGGGAGAGGAGGACAATATGGCTATTATTAAGGAGTTTATGAGATTTAAGGTACATATGGAGGGATCAGTAAATGGACATGAGTTTGAGATTGAGGGAGAGGGAGAGGGAAGACCTTATGAGGGAACACAAACAGCTAAGCTTAAGGTAACAAAGGGAGGACCTCTTCCTTTTGCTTGGGACATTCTTTCACCTCAATTTATGTATGGATCAAAGGCTTATGTAAAGCATCCTGCTGACATTCCTGACTATCTTAAGCTTTCATTTCCTGAGGGATTTAAGTGGGAGAGAGTAATGAATTTTGAGGACGGAGGAGTAGTAACAGTAACACAAGACTCATCACTTCAAGACGGAGAGTTTATTTATAAGGTAAAGCTTAGAGGAACAAATTTTCCTTCAGACGGACCTGTAATGCAAAAGAAGACAATGGGATGGGAGGCTTCATCAGAGAGAATGTATCCTGAGGACGGAGCTCTTAAGGGAGAGATTAAGCAAAGACTTAAGCTTAAGGACGGAGGACATTATGACGCTGAGGTAAAGACAACATATAAGGCTAAGAAGCCTGTACAACTTCCTGGAGCTTATAATGTAAATATTAAGCTTGACATTACATCACATAATGAGGACTATACAATTGTAGAGCAATATGAGAGAGCTGAGGGAAGACATTCAACAGGAGGAATGGACGAGCTTTATAAGTGA', 'average_distance_score': 2.169537763253943, 'weakest_link_score': 2.169537763253943}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTATCAAAGGGAGAGGAGGACAATATGGCTATTATTAAGGAGTTTATGAGATTTAAGGTACATATGGAGGGATCAGTAAATGGACATGAGTTTGAGATTGAGGGAGAGGGAGAGGGAAGACCTTATGAGGGAACACAAACAGCTAAGCTTAAGGTAACAAAGGGAGGACCTCTTCCTTTTGCTTGGGACATTCTTTCACCTCAATTTATGTATGGATCAAAGGCTTATGTAAAGCATCCTGCTGACATTCCTGACTATCTTAAGCTTTCATTTCCTGAGGGATTTAAGTGGGAGAGAGTAATGAATTTTGAGGACGGAGGAGTAGTAACAGTAACACAAGACTCATCACTTCAAGACGGAGAGTTTATTTATAAGGTAAAGCTTAGAGGAACAAATTTTCCTTCAGACGGACCTGTAATGCAAAAGAAGACAATGGGATGGGAGGCTTCATCAGAGAGAATGTATCCTGAGGACGGAGCTCTTAAGGGAGAGATTAAGCAAAGACTTAAGCTTAAGGACGGAGGACATTATGACGCTGAGGTAAAGACAACATATAAGGCTAAGAAGCCTGTACAACTTCCTGGAGCTTATAATGTAAATATTAAGCTTGACATTACATCACATAATGAGGACTATACAATTGTAGAGCAATATGAGAGAGCTGAGGGAAGACATTCAACAGGAGGAATGGACGAGCTTTATAAGTGA', 'average_distance_score': 2.169537763253943, 'weakest_link_score': 2.169537763253943}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry_debug\\CAI_single_codon_diff_ecoli_opt_False_NkHQ\n",
      "Final output: {'user_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'optimization_method': 'single_codon_diff', 'optimization_cub_index': 'CAI', 'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.10739856801909307, 'TGA': 0.1694510739856802, 'GCA': 1.0, 'GCC': 0.5246426811237062, 'GCG': 0.9196648595367175, 'GCT': 0.9361754558896007, 'TGC': 1.0, 'TGT': 0.7495652173913042, 'GAC': 0.6699738903394257, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.3869690068323596, 'TTC': 0.82402496099844, 'TTT': 1.0, 'GGA': 0.8480035296712993, 'GGC': 1.0, 'GGG': 0.29340392675932053, 'GGT': 0.6170306640194132, 'CAC': 0.6778131634819533, 'CAT': 1.0, 'ATA': 0.11114994757077944, 'ATC': 0.8386927647675637, 'ATT': 1.0, 'AAA': 1.0, 'AAG': 0.3398391420911528, 'CTA': 0.14688262702462837, 'CTC': 0.3046372309740404, 'CTG': 0.6640781007321944, 'CTT': 1.0, 'TTA': 0.5893055247392945, 'TTG': 0.41202573774129136, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.8348958333333333, 'CCA': 0.4923917284432306, 'CCC': 0.07452204447912603, 'CCG': 1.0, 'CCT': 0.7553648068669527, 'CAA': 1.0, 'CAG': 0.8251624335499115, 'AGA': 0.72005772005772, 'AGG': 0.09620009620009619, 'CGA': 0.19672919672919673, 'CGC': 0.984126984126984, 'CGG': 0.2751322751322751, 'CGT': 1.0, 'AGC': 0.8834134615384617, 'AGT': 0.33774038461538464, 'TCA': 0.9487179487179488, 'TCC': 0.42588141025641024, 'TCG': 0.2560096153846154, 'TCT': 1.0, 'ACA': 1.0, 'ACC': 0.22142692218887092, 'ACG': 0.5329023320249365, 'ACT': 0.4301546986839067, 'GTA': 0.7337249143416544, 'GTC': 0.681106216348507, 'GTG': 0.6669114047968673, 'GTT': 1.0, 'TGG': 1.0, 'TAC': 0.7787396121883656, 'TAT': 1.0}, 'cai_avg': 0.7354163127574911, 'cai_std': 0.04172052669683352, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}, {'name': 'Escherichia coli', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.037037037037037035, 'TGA': 0.24183006535947713, 'GCA': 0.6554010349288486, 'GCC': 0.6652652005174644, 'GCG': 1.0, 'GCT': 0.6266170763260025, 'TGC': 1.0, 'TGT': 0.7162162162162162, 'GAC': 0.8422464375523889, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.36087048383688997, 'TTC': 1.0, 'TTT': 0.6891233766233766, 'GGA': 0.12073112073112073, 'GGC': 1.0, 'GGG': 0.19111752445085778, 'GGT': 0.9866923200256533, 'CAC': 1.0, 'CAT': 0.7259593679458239, 'ATA': 0.03593829462620784, 'ATC': 1.0, 'ATT': 0.7563993897270725, 'AAA': 1.0, 'AAG': 0.2715144230769231, 'CTA': 0.0255237398225231, 'CTC': 0.1333821242338304, 'CTG': 1.0, 'CTT': 0.11572591711645777, 'TTA': 0.10557131095050773, 'TTG': 0.12322751806788033, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.4079954093343534, 'CCA': 0.2681349767629825, 'CCC': 0.08830066680137402, 'CCG': 1.0, 'CCT': 0.19337239846433624, 'CAA': 0.3511947863866763, 'CAG': 1.0, 'AGA': 0.018922018348623854, 'AGG': 0.011085626911314986, 'CGA': 0.03631498470948013, 'CGC': 0.7138761467889909, 'CGG': 0.05600152905198777, 'CGT': 1.0, 'AGC': 1.0, 'AGT': 0.3131715081575806, 'TCA': 0.3048149621965778, 'TCC': 0.7942697970553123, 'TCG': 0.40907282132908873, 'TCT': 0.8901711102268206, 'ACA': 0.14701775384001597, 'ACC': 1.0, 'ACG': 0.3457011769399561, 'ACT': 0.4374625972471574, 'GTA': 0.531166518254675, 'GTC': 0.5075690115761353, 'GTG': 0.9971059661620658, 'GTT': 1.0, 'TGG': 1.0, 'TAC': 1.0, 'TAT': 0.857994781960492}, 'cai_avg': 0.6260311851192147, 'cai_std': 0.09192830548402788, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}]}, 'orf_debug': {'C': defaultdict(<class 'float'>, {'TGT': 0.2878416243556693, 'TGC': 0.5}), 'D': defaultdict(<class 'float'>, {'GAT': 0.5, 'GAC': 0.4091481473136918}), 'S': defaultdict(<class 'float'>, {'TCT': 0.3962023027412252, 'TCG': 0.36043113277514427, 'TCA': 0.047771004981298035, 'TCC': 0.4802383328017277, 'AGC': 0.5067962104752218, 'AGT': 0.2683320958455362}), 'Q': defaultdict(<class 'float'>, {'CAA': 0.06166888899259159, 'CAG': 0.5152840873210945}), 'M': defaultdict(<class 'float'>, {'ATG': 0.5}), 'N': defaultdict(<class 'float'>, {'AAC': 0.5, 'AAT': 0.09685981994430053}), 'P': defaultdict(<class 'float'>, {'CCT': 0.04861963110354668, 'CCG': 0.5, 'CCA': 0.16478126155826806, 'CCC': 0.43215322695633207}), 'K': defaultdict(<class 'float'>, {'AAG': 0.2547662201268697, 'AAA': 0.5}), '_': defaultdict(<class 'float'>, {'TAG': 0.3990545292434242, 'TGA': 0.37414664950765447, 'TAA': 0.5}), 'T': defaultdict(<class 'float'>, {'ACC': 0.8030880187461472, 'ACA': 0.010807109972081766, 'ACG': 0.16884476758270672, 'ACT': 0.2580485957111289}), 'F': defaultdict(<class 'float'>, {'TTT': 0.2374455141044021, 'TTC': 0.5154836071758003}), 'A': defaultdict(<class 'float'>, {'GCA': 0.21477525829290292, 'GCC': 0.33427118381440035, 'GCG': 0.5032268673966277, 'GCT': 0.19836126638712376}), 'G': defaultdict(<class 'float'>, {'GGT': 0.560113623349509, 'GGG': 0.26790195943568595, 'GGA': 0.018839465252688016, 'GGC': 0.5}), 'I': defaultdict(<class 'float'>, {'ATC': 0.5130100120691663, 'ATA': 0.39567298836203413, 'ATT': 0.2860700183897439}), 'L': defaultdict(<class 'float'>, {'TTA': 0.0899076268527375, 'TTG': 0.18044937714392123, 'CTC': 0.2506600858062895, 'CTT': 0.006696243946222627, 'CTG': 0.5564217612038449, 'CTA': 0.3642303566834636}), 'H': defaultdict(<class 'float'>, {'CAT': 0.2635085019541501, 'CAC': 0.5519021788127533}), 'R': defaultdict(<class 'float'>, {'CGA': 0.32328138075088564, 'CGC': 0.254935552793601, 'CGG': 0.26428469490553613, 'CGT': 0.5, 'AGG': 0.4084885786163661, 'AGA': 0.039362861438833774}), 'W': defaultdict(<class 'float'>, {'TGG': 0.5}), 'V': defaultdict(<class 'float'>, {'GTA': 0.1765201456785766, 'GTC': 0.1796597733819701, 'GTG': 0.5525841600051915, 'GTT': 0.5}), 'E': defaultdict(<class 'float'>, {'GAG': 0.2530172523443873, 'GAA': 0.5}), 'Y': defaultdict(<class 'float'>, {'TAT': 0.36807752293571605, 'TAC': 0.5244780796072774})}, 'orf': {'aa_to_optimal_codon': {'C': 'TGT', 'D': 'GAC', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAG', '_': 'TGA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCT', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTA', 'E': 'GAG', 'Y': 'TAT'}, 'run_time': 0.062000000034458935}, 'evaluation': [{'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'cai_initial_score': 0.7608140098814632, 'cai_final_score': 0.7204288515987123, 'dist_score': -0.9679925322183441}, {'name': 'Escherichia coli', 'is_wanted': False, 'cai_initial_score': 0.7891479965436429, 'cai_final_score': 0.30127822281653116, 'dist_score': -5.30706805872623}], 'final_sequence': 'ATGGTATCAAAGGGAGAGGAGGACAATATGGCTATTATTAAGGAGTTTATGAGATTTAAGGTACATATGGAGGGATCAGTAAATGGACATGAGTTTGAGATTGAGGGAGAGGGAGAGGGAAGACCTTATGAGGGAACACAAACAGCTAAGCTTAAGGTAACAAAGGGAGGACCTCTTCCTTTTGCTTGGGACATTCTTTCACCTCAATTTATGTATGGATCAAAGGCTTATGTAAAGCATCCTGCTGACATTCCTGACTATCTTAAGCTTTCATTTCCTGAGGGATTTAAGTGGGAGAGAGTAATGAATTTTGAGGACGGAGGAGTAGTAACAGTAACACAAGACTCATCACTTCAAGACGGAGAGTTTATTTATAAGGTAAAGCTTAGAGGAACAAATTTTCCTTCAGACGGACCTGTAATGCAAAAGAAGACAATGGGATGGGAGGCTTCATCAGAGAGAATGTATCCTGAGGACGGAGCTCTTAAGGGAGAGATTAAGCAAAGACTTAAGCTTAAGGACGGAGGACATTATGACGCTGAGGTAAAGACAACATATAAGGCTAAGAAGCCTGTACAACTTCCTGGAGCTTATAATGTAAATATTAAGCTTGACATTACATCACATAATGAGGACTATACAATTGTAGAGCAATATGAGAGAGCTGAGGGAAGACATTCAACAGGAGGAATGGACGAGCTTTATAAGTGA', 'average_distance_score': 2.169537763253943, 'weakest_link_score': 2.169537763253943}], 'final_evaluation': {'final_sequence': 'ATGGTATCAAAGGGAGAGGAGGACAATATGGCTATTATTAAGGAGTTTATGAGATTTAAGGTACATATGGAGGGATCAGTAAATGGACATGAGTTTGAGATTGAGGGAGAGGGAGAGGGAAGACCTTATGAGGGAACACAAACAGCTAAGCTTAAGGTAACAAAGGGAGGACCTCTTCCTTTTGCTTGGGACATTCTTTCACCTCAATTTATGTATGGATCAAAGGCTTATGTAAAGCATCCTGCTGACATTCCTGACTATCTTAAGCTTTCATTTCCTGAGGGATTTAAGTGGGAGAGAGTAATGAATTTTGAGGACGGAGGAGTAGTAACAGTAACACAAGACTCATCACTTCAAGACGGAGAGTTTATTTATAAGGTAAAGCTTAGAGGAACAAATTTTCCTTCAGACGGACCTGTAATGCAAAAGAAGACAATGGGATGGGAGGCTTCATCAGAGAGAATGTATCCTGAGGACGGAGCTCTTAAGGGAGAGATTAAGCAAAGACTTAAGCTTAAGGACGGAGGACATTATGACGCTGAGGTAAAGACAACATATAAGGCTAAGAAGCCTGTACAACTTCCTGGAGCTTATAATGTAAATATTAAGCTTGACATTACATCACATAATGAGGACTATACAATTGTAGAGCAATATGAGAGAGCTGAGGGAAGACATTCAACAGGAGGAATGGACGAGCTTTATAAGTGA', 'average_distance_score': 2.169537763253943, 'weakest_link_score': 2.169537763253943}, 'zip_output_file_path': 'results\\\\mcherry_debug\\\\CAI_single_codon_diff_ecoli_opt_False_NkHQ\\\\communique_results.zip'}\n",
      "##########################\n",
      "# USER INPUT INFORMATION #\n",
      "##########################\n",
      "Organism is optimized\n",
      "Information about Escherichia coli:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Escherichia coli. Calculating tAI profile.\n",
      "name=Escherichia coli, tai_std=0.023481599705588187, tai_avg=0.2504950429983414\n",
      "Organism is de-optimized\n",
      "Information about Bacillus subtilis:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Bacillus subtilis. Calculating tAI profile.\n",
      "name=Bacillus subtilis, tai_std=0.027038419737753252, tai_avg=0.3292474875006557\n",
      "Escherichia coli has weight of 1.0\n",
      "Bacillus subtilis has weight of 1.0\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is: ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Total input processing time: 23.590357780456543\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "\n",
      "Optimization method used is: OptimizationMethod.single_codon_diff\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGT': 0.006148003472222222, 'TGC': 0.013888888888888888})\n",
      "Optimal codon for C is: TGT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAT': 0.08917508680555557, 'GAC': 0.2222222222222222})\n",
      "Optimal codon for D is: GAT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TCT': 0.01008784722222222, 'TCG': 0.009301388888888889, 'TCA': 0.061743836805555556, 'TCC': 0.013888888888888888, 'AGC': 0.06336805555555555, 'AGT': 0.034870920138888886})\n",
      "Optimal codon for S is: TCG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAA': 0.22360034722222222, 'CAG': 0.022755555555555546})\n",
      "Optimal codon for Q is: CAG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATG': 0.5})\n",
      "Optimal codon for M is: ATG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAC': 0.2222222222222222, 'AAT': 0.09836805555555556})\n",
      "Optimal codon for N is: AAT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CCT': 0.004163281249999997, 'CCG': 0.012799999999999995, 'CCA': 0.12534453125, 'CCC': 0.0007999999999999992})\n",
      "Optimal codon for P is: CCC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAG': 0.15280555555555556, 'AAA': 0.2222222222222222})\n",
      "Optimal codon for K is: AAG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAG': 0.029642332908156408, 'TGA': 0.029642332908156408, 'TAA': 0.029642332908156408})\n",
      "Optimal codon for _ is: TAG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ACC': 0.014688888888888887, 'ACA': 0.23167534722222222, 'ACG': 0.022755555555555546, 'ACT': 0.014987847222222217})\n",
      "Optimal codon for T is: ACC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTT': 0.04876562500000001, 'TTC': 0.125})\n",
      "Optimal codon for F is: TTT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GCA': 0.34722222222222227, 'GCC': 0.021701388888888888, 'GCG': 0.06806805555555553, 'GCT': 0.030712847222222218})\n",
      "Optimal codon for A is: GCC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GGT': 0.09836805555555556, 'GGG': 0.0689125, 'GGA': 0.19025078125, 'GGC': 0.2222222222222222})\n",
      "Optimal codon for G is: GGG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATC': 0.125, 'ATA': 0.041162500000000005, 'ATT': 0.055332031250000004})\n",
      "Optimal codon for I is: ATA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTA': 0.21483550347222222, 'TTG': 0.12366805555555557, 'CTC': 0.1000013888888889, 'CTT': 0.11352925347222226, 'CTG': 0.03735555555555555, 'CTA': 0.12663472222222222})\n",
      "Optimal codon for L is: CTG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAT': 0.02065217013888889, 'CAC': 0.05555555555555555})\n",
      "Optimal codon for H is: CAT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CGA': 0.1183645, 'CGC': 0.125, 'CGG': 0.0842013888888889, 'CGT': 0.2222222222222222, 'AGG': 0.08031250000000001, 'AGA': 0.08236328125})\n",
      "Optimal codon for R is: AGG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGG': 0.013888888888888888})\n",
      "Optimal codon for W is: TGG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GTA': 0.2222222222222222, 'GTC': 0.0842013888888889, 'GTG': 0.11306805555555556, 'GTT': 0.11883784722222222})\n",
      "Optimal codon for V is: GTC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAG': 0.10899999999999999, 'GAA': 0.5})\n",
      "Optimal codon for E is: GAG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAT': 0.031158420138888886, 'TAC': 0.05555555555555555})\n",
      "Optimal codon for Y is: TAT\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTCTCGAAGGGGGAGGAGGATAATATGGCCATAATAAAGGAGTTTATGAGGTTTAAGGTCCATATGGAGGGGTCGGTCAATGGGCATGAGTTTGAGATAGAGGGGGAGGGGGAGGGGAGGCCCTATGAGGGGACCCAGACCGCCAAGCTGAAGGTCACCAAGGGGGGGCCCCTGCCCTTTGCCTGGGATATACTGTCGCCCCAGTTTATGTATGGGTCGAAGGCCTATGTCAAGCATCCCGCCGATATACCCGATTATCTGAAGCTGTCGTTTCCCGAGGGGTTTAAGTGGGAGAGGGTCATGAATTTTGAGGATGGGGGGGTCGTCACCGTCACCCAGGATTCGTCGCTGCAGGATGGGGAGTTTATATATAAGGTCAAGCTGAGGGGGACCAATTTTCCCTCGGATGGGCCCGTCATGCAGAAGAAGACCATGGGGTGGGAGGCCTCGTCGGAGAGGATGTATCCCGAGGATGGGGCCCTGAAGGGGGAGATAAAGCAGAGGCTGAAGCTGAAGGATGGGGGGCATTATGATGCCGAGGTCAAGACCACCTATAAGGCCAAGAAGCCCGTCCAGCTGCCCGGGGCCTATAATGTCAATATAAAGCTGGATATAACCTCGCATAATGAGGATTATACCATAGTCGAGCAGTATGAGAGGGCCGAGGGGAGGCATTCGACCGGGGGGATGGATGAGCTGTATAAGTAA', 'average_distance_score': 0.5168220227201525, 'weakest_link_score': 0.5168220227201525}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTCTCGAAGGGGGAGGAGGATAATATGGCCATAATAAAGGAGTTTATGAGGTTTAAGGTCCATATGGAGGGGTCGGTCAATGGGCATGAGTTTGAGATAGAGGGGGAGGGGGAGGGGAGGCCCTATGAGGGGACCCAGACCGCCAAGCTGAAGGTCACCAAGGGGGGGCCCCTGCCCTTTGCCTGGGATATACTGTCGCCCCAGTTTATGTATGGGTCGAAGGCCTATGTCAAGCATCCCGCCGATATACCCGATTATCTGAAGCTGTCGTTTCCCGAGGGGTTTAAGTGGGAGAGGGTCATGAATTTTGAGGATGGGGGGGTCGTCACCGTCACCCAGGATTCGTCGCTGCAGGATGGGGAGTTTATATATAAGGTCAAGCTGAGGGGGACCAATTTTCCCTCGGATGGGCCCGTCATGCAGAAGAAGACCATGGGGTGGGAGGCCTCGTCGGAGAGGATGTATCCCGAGGATGGGGCCCTGAAGGGGGAGATAAAGCAGAGGCTGAAGCTGAAGGATGGGGGGCATTATGATGCCGAGGTCAAGACCACCTATAAGGCCAAGAAGCCCGTCCAGCTGCCCGGGGCCTATAATGTCAATATAAAGCTGGATATAACCTCGCATAATGAGGATTATACCATAGTCGAGCAGTATGAGAGGGCCGAGGGGAGGCATTCGACCGGGGGGATGGATGAGCTGTATAAGTAA', 'average_distance_score': 0.5168220227201525, 'weakest_link_score': 0.5168220227201525}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry_debug\\tAI_single_codon_diff_ecoli_opt_True_2M1V\n",
      "Final output: {'user_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'optimization_method': 'single_codon_diff', 'optimization_cub_index': 'tAI', 'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': None, 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.75, 'AAC': 0.5, 'AAG': 0.23999999999999996, 'AAT': 0.29500000000000004, 'ACA': 0.1525, 'ACC': 0.25, 'ACG': 0.29, 'ACT': 0.14750000000000002, 'AGA': 0.13874999999999998, 'AGC': 0.125, 'AGG': 0.16499999999999998, 'AGT': 0.07375000000000001, 'ATA': 0.10999999999999999, 'ATC': 0.375, 'ATG': 1.0, 'ATT': 0.22125000000000003, 'CAA': 0.27749999999999997, 'CAC': 0.125, 'CAG': 0.32999999999999996, 'CAT': 0.07375000000000001, 'CCA': 0.13874999999999998, 'CCC': 0.125, 'CCG': 0.16499999999999998, 'CCT': 0.07375000000000001, 'CGA': 0.013799999999999993, 'CGC': 0.36, 'CGG': 0.125, 'CGT': 0.5, 'CTA': 0.18, 'CTC': 0.125, 'CTG': 0.54, 'CTT': 0.07375000000000001, 'GAA': 0.5, 'GAC': 0.375, 'GAG': 0.15999999999999998, 'GAT': 0.22125000000000003, 'GCA': 0.375, 'GCC': 0.25, 'GCG': 0.11999999999999998, 'GCT': 0.14750000000000002, 'GGA': 0.13874999999999998, 'GGC': 0.5, 'GGG': 0.16499999999999998, 'GGT': 0.29500000000000004, 'GTA': 0.625, 'GTC': 0.25, 'GTG': 0.19999999999999996, 'GTT': 0.14750000000000002, 'TAC': 0.375, 'TAT': 0.22125000000000003, 'TCA': 0.13874999999999998, 'TCC': 0.25, 'TCG': 0.16499999999999998, 'TCT': 0.14750000000000002, 'TGC': 0.125, 'TGG': 0.125, 'TGT': 0.07375000000000001, 'TTA': 0.13874999999999998, 'TTC': 0.25, 'TTG': 0.16499999999999998, 'TTT': 0.14750000000000002}, 'tai_avg': 0.2504950429983414, 'tai_std': 0.023481599705588187}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': None, 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.6666666666666666, 'AAC': 0.6666666666666666, 'AAG': 0.2133333333333333, 'AAT': 0.39333333333333337, 'ACA': 0.6666666666666666, 'ACC': 0.16666666666666666, 'ACG': 0.2133333333333333, 'ACT': 0.09833333333333334, 'AGA': 0.18499999999999997, 'AGC': 0.3333333333333333, 'AGG': 0.21999999999999997, 'AGT': 0.19666666666666668, 'ATA': 0.10999999999999999, 'ATC': 0.5, 'ATG': 1.0, 'ATT': 0.29500000000000004, 'CAA': 0.6666666666666666, 'CAC': 0.3333333333333333, 'CAG': 0.2133333333333333, 'CAT': 0.19666666666666668, 'CCA': 0.5, 'CCG': 0.15999999999999998, 'CGA': 0.01839999999999999, 'CGC': 0.48, 'CGG': 0.16666666666666666, 'CGT': 0.6666666666666666, 'CTA': 0.3516666666666666, 'CTC': 0.16666666666666666, 'CTG': 0.2733333333333333, 'CTT': 0.09833333333333334, 'GAA': 1.0, 'GAC': 0.6666666666666666, 'GAG': 0.31999999999999995, 'GAT': 0.39333333333333337, 'GCA': 0.8333333333333334, 'GCC': 0.16666666666666666, 'GCG': 0.2666666666666666, 'GCT': 0.09833333333333334, 'GGA': 0.5, 'GGC': 0.6666666666666666, 'GGG': 0.15999999999999998, 'GGT': 0.39333333333333337, 'GTA': 0.6666666666666666, 'GTC': 0.16666666666666666, 'GTG': 0.2133333333333333, 'GTT': 0.09833333333333334, 'TAC': 0.3333333333333333, 'TAT': 0.19666666666666668, 'TCA': 0.3333333333333333, 'TCC': 0.16666666666666666, 'TCG': 0.10666666666666665, 'TCT': 0.09833333333333334, 'TGC': 0.16666666666666666, 'TGG': 0.16666666666666666, 'TGT': 0.09833333333333334, 'TTA': 0.5183333333333333, 'TTC': 0.5, 'TTG': 0.32666666666666666, 'TTT': 0.29500000000000004}, 'tai_avg': 0.3292474875006557, 'tai_std': 0.027038419737753252}]}, 'orf_debug': {'C': defaultdict(<class 'float'>, {'TGT': 0.006148003472222222, 'TGC': 0.013888888888888888}), 'D': defaultdict(<class 'float'>, {'GAT': 0.08917508680555557, 'GAC': 0.2222222222222222}), 'S': defaultdict(<class 'float'>, {'TCT': 0.01008784722222222, 'TCG': 0.009301388888888889, 'TCA': 0.061743836805555556, 'TCC': 0.013888888888888888, 'AGC': 0.06336805555555555, 'AGT': 0.034870920138888886}), 'Q': defaultdict(<class 'float'>, {'CAA': 0.22360034722222222, 'CAG': 0.022755555555555546}), 'M': defaultdict(<class 'float'>, {'ATG': 0.5}), 'N': defaultdict(<class 'float'>, {'AAC': 0.2222222222222222, 'AAT': 0.09836805555555556}), 'P': defaultdict(<class 'float'>, {'CCT': 0.004163281249999997, 'CCG': 0.012799999999999995, 'CCA': 0.12534453125, 'CCC': 0.0007999999999999992}), 'K': defaultdict(<class 'float'>, {'AAG': 0.15280555555555556, 'AAA': 0.2222222222222222}), '_': defaultdict(<class 'float'>, {'TAG': 0.029642332908156408, 'TGA': 0.029642332908156408, 'TAA': 0.029642332908156408}), 'T': defaultdict(<class 'float'>, {'ACC': 0.014688888888888887, 'ACA': 0.23167534722222222, 'ACG': 0.022755555555555546, 'ACT': 0.014987847222222217}), 'F': defaultdict(<class 'float'>, {'TTT': 0.04876562500000001, 'TTC': 0.125}), 'A': defaultdict(<class 'float'>, {'GCA': 0.34722222222222227, 'GCC': 0.021701388888888888, 'GCG': 0.06806805555555553, 'GCT': 0.030712847222222218}), 'G': defaultdict(<class 'float'>, {'GGT': 0.09836805555555556, 'GGG': 0.0689125, 'GGA': 0.19025078125, 'GGC': 0.2222222222222222}), 'I': defaultdict(<class 'float'>, {'ATC': 0.125, 'ATA': 0.041162500000000005, 'ATT': 0.055332031250000004}), 'L': defaultdict(<class 'float'>, {'TTA': 0.21483550347222222, 'TTG': 0.12366805555555557, 'CTC': 0.1000013888888889, 'CTT': 0.11352925347222226, 'CTG': 0.03735555555555555, 'CTA': 0.12663472222222222}), 'H': defaultdict(<class 'float'>, {'CAT': 0.02065217013888889, 'CAC': 0.05555555555555555}), 'R': defaultdict(<class 'float'>, {'CGA': 0.1183645, 'CGC': 0.125, 'CGG': 0.0842013888888889, 'CGT': 0.2222222222222222, 'AGG': 0.08031250000000001, 'AGA': 0.08236328125}), 'W': defaultdict(<class 'float'>, {'TGG': 0.013888888888888888}), 'V': defaultdict(<class 'float'>, {'GTA': 0.2222222222222222, 'GTC': 0.0842013888888889, 'GTG': 0.11306805555555556, 'GTT': 0.11883784722222222}), 'E': defaultdict(<class 'float'>, {'GAG': 0.10899999999999999, 'GAA': 0.5}), 'Y': defaultdict(<class 'float'>, {'TAT': 0.031158420138888886, 'TAC': 0.05555555555555555})}, 'orf': {'aa_to_optimal_codon': {'C': 'TGT', 'D': 'GAT', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCC', 'K': 'AAG', '_': 'TAA', 'T': 'ACC', 'F': 'TTT', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAT', 'R': 'AGG', 'W': 'TGG', 'V': 'GTC', 'E': 'GAG', 'Y': 'TAT'}, 'run_time': 0.0470000000204891}, 'evaluation': [{'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'tai_initial_score': 0.2799603826276079, 'tai_final_score': 0.20058573010517175, 'dist_score': -3.380291526882066}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'tai_initial_score': 0.33499296920046817, 'tai_final_score': 0.21564712650061577, 'dist_score': -4.413935572322371}], 'final_sequence': 'ATGGTCTCGAAGGGGGAGGAGGATAATATGGCCATAATAAAGGAGTTTATGAGGTTTAAGGTCCATATGGAGGGGTCGGTCAATGGGCATGAGTTTGAGATAGAGGGGGAGGGGGAGGGGAGGCCCTATGAGGGGACCCAGACCGCCAAGCTGAAGGTCACCAAGGGGGGGCCCCTGCCCTTTGCCTGGGATATACTGTCGCCCCAGTTTATGTATGGGTCGAAGGCCTATGTCAAGCATCCCGCCGATATACCCGATTATCTGAAGCTGTCGTTTCCCGAGGGGTTTAAGTGGGAGAGGGTCATGAATTTTGAGGATGGGGGGGTCGTCACCGTCACCCAGGATTCGTCGCTGCAGGATGGGGAGTTTATATATAAGGTCAAGCTGAGGGGGACCAATTTTCCCTCGGATGGGCCCGTCATGCAGAAGAAGACCATGGGGTGGGAGGCCTCGTCGGAGAGGATGTATCCCGAGGATGGGGCCCTGAAGGGGGAGATAAAGCAGAGGCTGAAGCTGAAGGATGGGGGGCATTATGATGCCGAGGTCAAGACCACCTATAAGGCCAAGAAGCCCGTCCAGCTGCCCGGGGCCTATAATGTCAATATAAAGCTGGATATAACCTCGCATAATGAGGATTATACCATAGTCGAGCAGTATGAGAGGGCCGAGGGGAGGCATTCGACCGGGGGGATGGATGAGCTGTATAAGTAA', 'average_distance_score': 0.5168220227201525, 'weakest_link_score': 0.5168220227201525}], 'final_evaluation': {'final_sequence': 'ATGGTCTCGAAGGGGGAGGAGGATAATATGGCCATAATAAAGGAGTTTATGAGGTTTAAGGTCCATATGGAGGGGTCGGTCAATGGGCATGAGTTTGAGATAGAGGGGGAGGGGGAGGGGAGGCCCTATGAGGGGACCCAGACCGCCAAGCTGAAGGTCACCAAGGGGGGGCCCCTGCCCTTTGCCTGGGATATACTGTCGCCCCAGTTTATGTATGGGTCGAAGGCCTATGTCAAGCATCCCGCCGATATACCCGATTATCTGAAGCTGTCGTTTCCCGAGGGGTTTAAGTGGGAGAGGGTCATGAATTTTGAGGATGGGGGGGTCGTCACCGTCACCCAGGATTCGTCGCTGCAGGATGGGGAGTTTATATATAAGGTCAAGCTGAGGGGGACCAATTTTCCCTCGGATGGGCCCGTCATGCAGAAGAAGACCATGGGGTGGGAGGCCTCGTCGGAGAGGATGTATCCCGAGGATGGGGCCCTGAAGGGGGAGATAAAGCAGAGGCTGAAGCTGAAGGATGGGGGGCATTATGATGCCGAGGTCAAGACCACCTATAAGGCCAAGAAGCCCGTCCAGCTGCCCGGGGCCTATAATGTCAATATAAAGCTGGATATAACCTCGCATAATGAGGATTATACCATAGTCGAGCAGTATGAGAGGGCCGAGGGGAGGCATTCGACCGGGGGGATGGATGAGCTGTATAAGTAA', 'average_distance_score': 0.5168220227201525, 'weakest_link_score': 0.5168220227201525}, 'zip_output_file_path': 'results\\\\mcherry_debug\\\\tAI_single_codon_diff_ecoli_opt_True_2M1V\\\\communique_results.zip'}\n",
      "##########################\n",
      "# USER INPUT INFORMATION #\n",
      "##########################\n",
      "Organism is optimized\n",
      "Information about Bacillus subtilis:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Bacillus subtilis. Calculating tAI profile.\n",
      "name=Bacillus subtilis, tai_std=0.027038419737753252, tai_avg=0.3292474875006557\n",
      "Organism is de-optimized\n",
      "Information about Escherichia coli:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Escherichia coli. Calculating tAI profile.\n",
      "name=Escherichia coli, tai_std=0.023481599705588187, tai_avg=0.2504950429983414\n",
      "Bacillus subtilis has weight of 1.0\n",
      "Escherichia coli has weight of 1.0\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is: ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Total input processing time: 21.108784437179565\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "\n",
      "Optimization method used is: OptimizationMethod.single_codon_diff\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGT': 0.005054253472222221, 'TGC': 0.0078125})\n",
      "Optimal codon for C is: TGT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAT': 0.06183133680555554, 'GAC': 0.0703125})\n",
      "Optimal codon for D is: GAT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TCT': 0.038490625, 'TCG': 0.039301388888888886, 'TCA': 0.009625781249999998, 'TCC': 0.04513888888888889, 'AGC': 0.0078125, 'AGT': 0.012058420138888885})\n",
      "Optimal codon for S is: AGC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAA': 0.03850312499999999, 'CAG': 0.15720555555555554})\n",
      "Optimal codon for Q is: CAA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATG': 0.5})\n",
      "Optimal codon for M is: ATG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAC': 0.125, 'AAT': 0.08086805555555554})\n",
      "Optimal codon for N is: AAT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CCT': 0.12771953125, 'CCG': 0.0714125, 'CCA': 0.009625781249999998, 'CCC': 0.1328125})\n",
      "Optimal codon for P is: CCA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'AAG': 0.13155555555555556, 'AAA': 0.28125})\n",
      "Optimal codon for K is: AAG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAG': 0.058223756903188754, 'TGA': 0.058223756903188754, 'TAA': 0.058223756903188754})\n",
      "Optimal codon for _ is: TAG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ACC': 0.15625, 'ACA': 0.011628125, 'ACG': 0.14480555555555558, 'ACT': 0.17237951388888884})\n",
      "Optimal codon for T is: ACA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTT': 0.03189062499999999, 'TTC': 0.03125})\n",
      "Optimal codon for F is: TTC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GCA': 0.0703125, 'GCC': 0.25347222222222227, 'GCG': 0.1677555555555556, 'GCT': 0.280990625})\n",
      "Optimal codon for A is: GCA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GGT': 0.08086805555555554, 'GGG': 0.14196805555555553, 'GGA': 0.02351467013888888, 'GGC': 0.125})\n",
      "Optimal codon for G is: GGA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'ATC': 0.0703125, 'ATA': 0.0821, 'ATT': 0.04548828125})\n",
      "Optimal codon for I is: ATT\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TTA': 0.009625781249999998, 'TTG': 0.031980555555555554, 'CTC': 0.06964722222222222, 'CTT': 0.09091953124999999, 'CTG': 0.1758125, 'CTA': 0.03008888888888889})\n",
      "Optimal codon for L is: TTA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CAT': 0.012058420138888885, 'CAC': 0.0078125})\n",
      "Optimal codon for H is: CAC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'CGA': 0.21022005555555556, 'CGC': 0.08222222222222221, 'CGG': 0.1328125, 'CGT': 0.125, 'AGG': 0.11336805555555556, 'AGA': 0.1256271701388889})\n",
      "Optimal codon for R is: CGC\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TGG': 0.0078125})\n",
      "Optimal codon for W is: TGG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GTA': 0.1953125, 'GTC': 0.15625, 'GTG': 0.12275555555555556, 'GTT': 0.17237951388888884})\n",
      "Optimal codon for V is: GTG\n",
      "Loss dict is: defaultdict(<class 'float'>, {'GAG': 0.24400000000000005, 'GAA': 0.125})\n",
      "Optimal codon for E is: GAA\n",
      "Loss dict is: defaultdict(<class 'float'>, {'TAT': 0.03381467013888889, 'TAC': 0.0703125})\n",
      "Optimal codon for Y is: TAT\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTGAGCAAGGGAGAAGAAGATAATATGGCAATTATTAAGGAATTCATGCGCTTCAAGGTGCACATGGAAGGAAGCGTGAATGGACACGAATTCGAAATTGAAGGAGAAGGAGAAGGACGCCCATATGAAGGAACACAAACAGCAAAGTTAAAGGTGACAAAGGGAGGACCATTACCATTCGCATGGGATATTTTAAGCCCACAATTCATGTATGGAAGCAAGGCATATGTGAAGCACCCAGCAGATATTCCAGATTATTTAAAGTTAAGCTTCCCAGAAGGATTCAAGTGGGAACGCGTGATGAATTTCGAAGATGGAGGAGTGGTGACAGTGACACAAGATAGCAGCTTACAAGATGGAGAATTCATTTATAAGGTGAAGTTACGCGGAACAAATTTCCCAAGCGATGGACCAGTGATGCAAAAGAAGACAATGGGATGGGAAGCAAGCAGCGAACGCATGTATCCAGAAGATGGAGCATTAAAGGGAGAAATTAAGCAACGCTTAAAGTTAAAGGATGGAGGACACTATGATGCAGAAGTGAAGACAACATATAAGGCAAAGAAGCCAGTGCAATTACCAGGAGCATATAATGTGAATATTAAGTTAGATATTACAAGCCACAATGAAGATTATACAATTGTGGAACAATATGAACGCGCAGAAGGACGCCACAGCACAGGAGGAATGGATGAATTATATAAGTAA', 'average_distance_score': 3.0574937724559748, 'weakest_link_score': 3.0574937724559748}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTGAGCAAGGGAGAAGAAGATAATATGGCAATTATTAAGGAATTCATGCGCTTCAAGGTGCACATGGAAGGAAGCGTGAATGGACACGAATTCGAAATTGAAGGAGAAGGAGAAGGACGCCCATATGAAGGAACACAAACAGCAAAGTTAAAGGTGACAAAGGGAGGACCATTACCATTCGCATGGGATATTTTAAGCCCACAATTCATGTATGGAAGCAAGGCATATGTGAAGCACCCAGCAGATATTCCAGATTATTTAAAGTTAAGCTTCCCAGAAGGATTCAAGTGGGAACGCGTGATGAATTTCGAAGATGGAGGAGTGGTGACAGTGACACAAGATAGCAGCTTACAAGATGGAGAATTCATTTATAAGGTGAAGTTACGCGGAACAAATTTCCCAAGCGATGGACCAGTGATGCAAAAGAAGACAATGGGATGGGAAGCAAGCAGCGAACGCATGTATCCAGAAGATGGAGCATTAAAGGGAGAAATTAAGCAACGCTTAAAGTTAAAGGATGGAGGACACTATGATGCAGAAGTGAAGACAACATATAAGGCAAAGAAGCCAGTGCAATTACCAGGAGCATATAATGTGAATATTAAGTTAGATATTACAAGCCACAATGAAGATTATACAATTGTGGAACAATATGAACGCGCAGAAGGACGCCACAGCACAGGAGGAATGGATGAATTATATAAGTAA', 'average_distance_score': 3.0574937724559748, 'weakest_link_score': 3.0574937724559748}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry_debug\\tAI_single_codon_diff_ecoli_opt_False_eItl\n",
      "Final output: {'user_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'optimization_method': 'single_codon_diff', 'optimization_cub_index': 'tAI', 'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': None, 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.6666666666666666, 'AAC': 0.6666666666666666, 'AAG': 0.2133333333333333, 'AAT': 0.39333333333333337, 'ACA': 0.6666666666666666, 'ACC': 0.16666666666666666, 'ACG': 0.2133333333333333, 'ACT': 0.09833333333333334, 'AGA': 0.18499999999999997, 'AGC': 0.3333333333333333, 'AGG': 0.21999999999999997, 'AGT': 0.19666666666666668, 'ATA': 0.10999999999999999, 'ATC': 0.5, 'ATG': 1.0, 'ATT': 0.29500000000000004, 'CAA': 0.6666666666666666, 'CAC': 0.3333333333333333, 'CAG': 0.2133333333333333, 'CAT': 0.19666666666666668, 'CCA': 0.5, 'CCG': 0.15999999999999998, 'CGA': 0.01839999999999999, 'CGC': 0.48, 'CGG': 0.16666666666666666, 'CGT': 0.6666666666666666, 'CTA': 0.3516666666666666, 'CTC': 0.16666666666666666, 'CTG': 0.2733333333333333, 'CTT': 0.09833333333333334, 'GAA': 1.0, 'GAC': 0.6666666666666666, 'GAG': 0.31999999999999995, 'GAT': 0.39333333333333337, 'GCA': 0.8333333333333334, 'GCC': 0.16666666666666666, 'GCG': 0.2666666666666666, 'GCT': 0.09833333333333334, 'GGA': 0.5, 'GGC': 0.6666666666666666, 'GGG': 0.15999999999999998, 'GGT': 0.39333333333333337, 'GTA': 0.6666666666666666, 'GTC': 0.16666666666666666, 'GTG': 0.2133333333333333, 'GTT': 0.09833333333333334, 'TAC': 0.3333333333333333, 'TAT': 0.19666666666666668, 'TCA': 0.3333333333333333, 'TCC': 0.16666666666666666, 'TCG': 0.10666666666666665, 'TCT': 0.09833333333333334, 'TGC': 0.16666666666666666, 'TGG': 0.16666666666666666, 'TGT': 0.09833333333333334, 'TTA': 0.5183333333333333, 'TTC': 0.5, 'TTG': 0.32666666666666666, 'TTT': 0.29500000000000004}, 'tai_avg': 0.3292474875006557, 'tai_std': 0.027038419737753252}, {'name': 'Escherichia coli', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': None, 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.75, 'AAC': 0.5, 'AAG': 0.23999999999999996, 'AAT': 0.29500000000000004, 'ACA': 0.1525, 'ACC': 0.25, 'ACG': 0.29, 'ACT': 0.14750000000000002, 'AGA': 0.13874999999999998, 'AGC': 0.125, 'AGG': 0.16499999999999998, 'AGT': 0.07375000000000001, 'ATA': 0.10999999999999999, 'ATC': 0.375, 'ATG': 1.0, 'ATT': 0.22125000000000003, 'CAA': 0.27749999999999997, 'CAC': 0.125, 'CAG': 0.32999999999999996, 'CAT': 0.07375000000000001, 'CCA': 0.13874999999999998, 'CCC': 0.125, 'CCG': 0.16499999999999998, 'CCT': 0.07375000000000001, 'CGA': 0.013799999999999993, 'CGC': 0.36, 'CGG': 0.125, 'CGT': 0.5, 'CTA': 0.18, 'CTC': 0.125, 'CTG': 0.54, 'CTT': 0.07375000000000001, 'GAA': 0.5, 'GAC': 0.375, 'GAG': 0.15999999999999998, 'GAT': 0.22125000000000003, 'GCA': 0.375, 'GCC': 0.25, 'GCG': 0.11999999999999998, 'GCT': 0.14750000000000002, 'GGA': 0.13874999999999998, 'GGC': 0.5, 'GGG': 0.16499999999999998, 'GGT': 0.29500000000000004, 'GTA': 0.625, 'GTC': 0.25, 'GTG': 0.19999999999999996, 'GTT': 0.14750000000000002, 'TAC': 0.375, 'TAT': 0.22125000000000003, 'TCA': 0.13874999999999998, 'TCC': 0.25, 'TCG': 0.16499999999999998, 'TCT': 0.14750000000000002, 'TGC': 0.125, 'TGG': 0.125, 'TGT': 0.07375000000000001, 'TTA': 0.13874999999999998, 'TTC': 0.25, 'TTG': 0.16499999999999998, 'TTT': 0.14750000000000002}, 'tai_avg': 0.2504950429983414, 'tai_std': 0.023481599705588187}]}, 'orf_debug': {'C': defaultdict(<class 'float'>, {'TGT': 0.005054253472222221, 'TGC': 0.0078125}), 'D': defaultdict(<class 'float'>, {'GAT': 0.06183133680555554, 'GAC': 0.0703125}), 'S': defaultdict(<class 'float'>, {'TCT': 0.038490625, 'TCG': 0.039301388888888886, 'TCA': 0.009625781249999998, 'TCC': 0.04513888888888889, 'AGC': 0.0078125, 'AGT': 0.012058420138888885}), 'Q': defaultdict(<class 'float'>, {'CAA': 0.03850312499999999, 'CAG': 0.15720555555555554}), 'M': defaultdict(<class 'float'>, {'ATG': 0.5}), 'N': defaultdict(<class 'float'>, {'AAC': 0.125, 'AAT': 0.08086805555555554}), 'P': defaultdict(<class 'float'>, {'CCT': 0.12771953125, 'CCG': 0.0714125, 'CCA': 0.009625781249999998, 'CCC': 0.1328125}), 'K': defaultdict(<class 'float'>, {'AAG': 0.13155555555555556, 'AAA': 0.28125}), '_': defaultdict(<class 'float'>, {'TAG': 0.058223756903188754, 'TGA': 0.058223756903188754, 'TAA': 0.058223756903188754}), 'T': defaultdict(<class 'float'>, {'ACC': 0.15625, 'ACA': 0.011628125, 'ACG': 0.14480555555555558, 'ACT': 0.17237951388888884}), 'F': defaultdict(<class 'float'>, {'TTT': 0.03189062499999999, 'TTC': 0.03125}), 'A': defaultdict(<class 'float'>, {'GCA': 0.0703125, 'GCC': 0.25347222222222227, 'GCG': 0.1677555555555556, 'GCT': 0.280990625}), 'G': defaultdict(<class 'float'>, {'GGT': 0.08086805555555554, 'GGG': 0.14196805555555553, 'GGA': 0.02351467013888888, 'GGC': 0.125}), 'I': defaultdict(<class 'float'>, {'ATC': 0.0703125, 'ATA': 0.0821, 'ATT': 0.04548828125}), 'L': defaultdict(<class 'float'>, {'TTA': 0.009625781249999998, 'TTG': 0.031980555555555554, 'CTC': 0.06964722222222222, 'CTT': 0.09091953124999999, 'CTG': 0.1758125, 'CTA': 0.03008888888888889}), 'H': defaultdict(<class 'float'>, {'CAT': 0.012058420138888885, 'CAC': 0.0078125}), 'R': defaultdict(<class 'float'>, {'CGA': 0.21022005555555556, 'CGC': 0.08222222222222221, 'CGG': 0.1328125, 'CGT': 0.125, 'AGG': 0.11336805555555556, 'AGA': 0.1256271701388889}), 'W': defaultdict(<class 'float'>, {'TGG': 0.0078125}), 'V': defaultdict(<class 'float'>, {'GTA': 0.1953125, 'GTC': 0.15625, 'GTG': 0.12275555555555556, 'GTT': 0.17237951388888884}), 'E': defaultdict(<class 'float'>, {'GAG': 0.24400000000000005, 'GAA': 0.125}), 'Y': defaultdict(<class 'float'>, {'TAT': 0.03381467013888889, 'TAC': 0.0703125})}, 'orf': {'aa_to_optimal_codon': {'C': 'TGT', 'D': 'GAT', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCA', 'K': 'AAG', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'TTA', 'H': 'CAC', 'R': 'CGC', 'W': 'TGG', 'V': 'GTG', 'E': 'GAA', 'Y': 'TAT'}, 'run_time': 0.030999999959021807}, 'evaluation': [{'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'tai_initial_score': 0.33499296920046817, 'tai_final_score': 0.4290852254770808, 'dist_score': 3.479946579319993}, {'name': 'Escherichia coli', 'is_wanted': False, 'tai_initial_score': 0.2799603826276079, 'tai_final_score': 0.218085405465751, 'dist_score': -2.635040965591957}], 'final_sequence': 'ATGGTGAGCAAGGGAGAAGAAGATAATATGGCAATTATTAAGGAATTCATGCGCTTCAAGGTGCACATGGAAGGAAGCGTGAATGGACACGAATTCGAAATTGAAGGAGAAGGAGAAGGACGCCCATATGAAGGAACACAAACAGCAAAGTTAAAGGTGACAAAGGGAGGACCATTACCATTCGCATGGGATATTTTAAGCCCACAATTCATGTATGGAAGCAAGGCATATGTGAAGCACCCAGCAGATATTCCAGATTATTTAAAGTTAAGCTTCCCAGAAGGATTCAAGTGGGAACGCGTGATGAATTTCGAAGATGGAGGAGTGGTGACAGTGACACAAGATAGCAGCTTACAAGATGGAGAATTCATTTATAAGGTGAAGTTACGCGGAACAAATTTCCCAAGCGATGGACCAGTGATGCAAAAGAAGACAATGGGATGGGAAGCAAGCAGCGAACGCATGTATCCAGAAGATGGAGCATTAAAGGGAGAAATTAAGCAACGCTTAAAGTTAAAGGATGGAGGACACTATGATGCAGAAGTGAAGACAACATATAAGGCAAAGAAGCCAGTGCAATTACCAGGAGCATATAATGTGAATATTAAGTTAGATATTACAAGCCACAATGAAGATTATACAATTGTGGAACAATATGAACGCGCAGAAGGACGCCACAGCACAGGAGGAATGGATGAATTATATAAGTAA', 'average_distance_score': 3.0574937724559748, 'weakest_link_score': 3.0574937724559748}], 'final_evaluation': {'final_sequence': 'ATGGTGAGCAAGGGAGAAGAAGATAATATGGCAATTATTAAGGAATTCATGCGCTTCAAGGTGCACATGGAAGGAAGCGTGAATGGACACGAATTCGAAATTGAAGGAGAAGGAGAAGGACGCCCATATGAAGGAACACAAACAGCAAAGTTAAAGGTGACAAAGGGAGGACCATTACCATTCGCATGGGATATTTTAAGCCCACAATTCATGTATGGAAGCAAGGCATATGTGAAGCACCCAGCAGATATTCCAGATTATTTAAAGTTAAGCTTCCCAGAAGGATTCAAGTGGGAACGCGTGATGAATTTCGAAGATGGAGGAGTGGTGACAGTGACACAAGATAGCAGCTTACAAGATGGAGAATTCATTTATAAGGTGAAGTTACGCGGAACAAATTTCCCAAGCGATGGACCAGTGATGCAAAAGAAGACAATGGGATGGGAAGCAAGCAGCGAACGCATGTATCCAGAAGATGGAGCATTAAAGGGAGAAATTAAGCAACGCTTAAAGTTAAAGGATGGAGGACACTATGATGCAGAAGTGAAGACAACATATAAGGCAAAGAAGCCAGTGCAATTACCAGGAGCATATAATGTGAATATTAAGTTAGATATTACAAGCCACAATGAAGATTATACAATTGTGGAACAATATGAACGCGCAGAAGGACGCCACAGCACAGGAGGAATGGATGAATTATATAAGTAA', 'average_distance_score': 3.0574937724559748, 'weakest_link_score': 3.0574937724559748}, 'zip_output_file_path': 'results\\\\mcherry_debug\\\\tAI_single_codon_diff_ecoli_opt_False_eItl\\\\communique_results.zip'}\n",
      "##########################\n",
      "# USER INPUT INFORMATION #\n",
      "##########################\n",
      "Organism is optimized\n",
      "Information about Escherichia coli:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 90 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3895\n",
      "Calculate CAI weights from a reference set of 584 highly expressed genes from estimated expression dictionary.\n",
      "name=Escherichia coli, cai_std=0.09192830548402788, cai_avg=0.6260311851192147\n",
      "Organism is de-optimized\n",
      "Information about Bacillus subtilis:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 98 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3550\n",
      "Calculate CAI weights from a reference set of 532 highly expressed genes from estimated expression dictionary.\n",
      "name=Bacillus subtilis, cai_std=0.04172052669683352, cai_avg=0.7354163127574911\n",
      "Escherichia coli has weight of 1.0\n",
      "Bacillus subtilis has weight of 1.0\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is: ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Total input processing time: 20.60048007965088\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "\n",
      "Optimization method used is: OptimizationMethod.zscore_bulk_aa_ratio\n",
      "Running ORF optimization for initial sequence: ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Running ORF optimization for initial sequence: ATGGTGTCGAAAGGGGAAGAAGACAACATGGCCATCATCAAAGAGTTCATGAGATTTAAAGTCCATATGGAAGGATCTGTGAACGGACACGAGTTCGAGATTGAAGGTGAAGGGGAAGGTAGACCATACGAGGGTACGCAGACAGCCAAGCTCAAGGTAACGAAGGGAGGCCCTCTCCCGTTCGCCTGGGACATACTTAGTCCGCAATTTATGTACGGAAGTAAGGCCTACGTGAAGCATCCTGCAGATATCCCTGATTACTTGAAGTTGAGTTTCCCTGAAGGCTTTAAATGGGAAAGAGTCATGAACTTTGAGGATGGTGGCGTTGTTACAGTGACCCAAGACTCCTCTCTTCAGGATGGAGAGTTTATCTACAAGGTTAAACTCCGTGGAACGAACTTTCCCTCAGATGGGCCGGTGATGCAGAAAAAGACCATGGGGTGGGAGGCGTCGTCGGAGAGGATGTACCCTGAGGACGGTGCACTCAAAGGTGAGATAAAGCAGCGCCTGAAACTAAAAGACGGTGGCCACTATGACGCTGAGGTGAAAACCACATATAAGGCGAAAAAACCAGTGCAACTTCCTGGCGCGTACAACGTAAACATCAAGCTTGATATCACTAGCCACAACGAGGACTACACTATCGTAGAACAGTATGAACGCGCTGAAGGGCGACACTCCACGGGTGGTATGGACGAGCTGTATAAGTAA\n",
      "Running ORF optimization for initial sequence: ATGGTTTCGAAGGGTGAAGAGGATAACATGGCCATCATTAAAGAATTCATGAGGTTCAAAGTACATATGGAAGGCTCGGTGAATGGGCATGAATTTGAAATCGAGGGCGAAGGCGAGGGACGCCCGTATGAGGGGACACAGACCGCCAAACTAAAAGTCACTAAAGGCGGGCCGTTGCCCTTCGCTTGGGACATTTTGTCACCGCAATTTATGTACGGGTCAAAAGCTTATGTTAAGCATCCGGCTGATATACCCGACTACCTAAAGCTTTCCTTTCCCGAGGGTTTTAAGTGGGAGCGAGTGATGAATTTTGAAGACGGAGGTGTAGTCACCGTCACGCAAGACAGCTCTCTACAAGATGGGGAGTTTATCTACAAGGTCAAATTGCGTGGGACAAACTTCCCGAGTGATGGACCAGTTATGCAGAAAAAAACCATGGGTTGGGAGGCCTCTTCAGAACGTATGTATCCAGAGGACGGAGCTCTGAAGGGGGAAATAAAACAGCGGCTAAAGTTAAAGGACGGAGGCCACTATGACGCCGAAGTAAAGACTACTTATAAAGCAAAGAAGCCCGTACAGCTTCCGGGCGCTTATAACGTAAATATAAAATTGGACATAACGTCTCACAACGAAGATTACACTATTGTTGAGCAATACGAAAGAGCGGAGGGACGACATTCAACCGGAGGAATGGACGAACTCTATAAATGA\n",
      "Running ORF optimization for initial sequence: ATGGTTAGCAAAGGGGAAGAGGACAATATGGCGATCATCAAAGAGTTCATGAGGTTTAAGGTTCATATGGAGGGATCGGTGAATGGTCATGAGTTCGAGATTGAAGGCGAGGGAGAAGGGAGGCCCTACGAGGGAACTCAGACAGCGAAACTCAAGGTGACAAAAGGGGGGCCGCTCCCCTTCGCATGGGATATCCTCAGTCCTCAGTTCATGTACGGAAGTAAGGCCTATGTTAAACATCCAGCGGATATACCGGATTACTTGAAGTTATCGTTCCCGGAGGGCTTCAAGTGGGAGAGGGTAATGAACTTCGAAGATGGCGGGGTGGTTACCGTAACCCAAGATAGCTCACTCCAAGACGGGGAGTTCATTTATAAAGTGAAACTTCGCGGTACAAATTTCCCGAGCGACGGTCCAGTGATGCAGAAAAAGACCATGGGTTGGGAGGCCTCTTCCGAACGGATGTATCCCGAGGACGGCGCACTCAAGGGGGAAATTAAACAAAGACTGAAGTTAAAAGACGGTGGTCATTACGATGCTGAAGTCAAGACCACATACAAAGCTAAAAAGCCGGTTCAGCTGCCTGGCGCCTACAATGTCAATATCAAACTAGATATTACTTCCCATAACGAAGATTACACCATCGTCGAACAATACGAGAGAGCGGAAGGGAGACATTCAACTGGCGGGATGGATGAACTATACAAATAG\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTGTCGAAGGGGGAGGAGGACAACATGGCCATAATAAAGGAGTTCATGAGGTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATAGAGGGGGAGGGGGAGGGGAGGCCCTACGAGGGGACCCAGACCGCCAAGCTAAAGGTGACCAAGGGGGGGCCCCTACCCTTCGCCTGGGACATACTATCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATACCCGACTACCTAAAGCTATCGTTCCCCGAGGGGTTCAAGTGGGAGAGGGTGATGAACTTCGAGGACGGGGGGGTGGTGACCGTGACCCAGGACTCGTCGCTACAGGACGGGGAGTTCATATACAAGGTGAAGCTAAGGGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAGACCATGGGGTGGGAGGCCTCGTCGGAGAGGATGTACCCCGAGGACGGGGCCCTAAAGGGGGAGATAAAGCAGAGGCTAAAGCTAAAGGACGGGGGGCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTACCCGGGGCCTACAACGTGAACATAAAGCTAGACATAACCTCGCACAACGAGGACTACACCATAGTGGAGCAGTACGAGAGGGCCGAGGGGAGGCACTCGACCGGGGGGATGGACGAGCTATACAAGTAG', 'average_distance_score': 2.400642079220615, 'weakest_link_score': 2.400642079220615}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTGTCGAAGGGGGAGGAGGACAACATGGCCATAATAAAGGAGTTCATGAGGTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATAGAGGGGGAGGGGGAGGGGAGGCCCTACGAGGGGACCCAGACCGCCAAGCTAAAGGTGACCAAGGGGGGGCCCCTACCCTTCGCCTGGGACATACTATCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATACCCGACTACCTAAAGCTATCGTTCCCCGAGGGGTTCAAGTGGGAGAGGGTGATGAACTTCGAGGACGGGGGGGTGGTGACCGTGACCCAGGACTCGTCGCTACAGGACGGGGAGTTCATATACAAGGTGAAGCTAAGGGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAGACCATGGGGTGGGAGGCCTCGTCGGAGAGGATGTACCCCGAGGACGGGGCCCTAAAGGGGGAGATAAAGCAGAGGCTAAAGCTAAAGGACGGGGGGCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTACCCGGGGCCTACAACGTGAACATAAAGCTAGACATAACCTCGCACAACGAGGACTACACCATAGTGGAGCAGTACGAGAGGGCCGAGGGGAGGCACTCGACCGGGGGGATGGACGAGCTATACAAGTAG', 'average_distance_score': 2.400642079220615, 'weakest_link_score': 2.400642079220615}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTGTCGAAGGGGGAGGAGGACAACATGGCCATAATAAAGGAGTTCATGAGGTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATAGAGGGGGAGGGGGAGGGGAGGCCCTACGAGGGGACCCAGACCGCCAAGCTAAAGGTGACCAAGGGGGGGCCCCTACCCTTCGCCTGGGACATACTATCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATACCCGACTACCTAAAGCTATCGTTCCCCGAGGGGTTCAAGTGGGAGAGGGTGATGAACTTCGAGGACGGGGGGGTGGTGACCGTGACCCAGGACTCGTCGCTACAGGACGGGGAGTTCATATACAAGGTGAAGCTAAGGGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAGACCATGGGGTGGGAGGCCTCGTCGGAGAGGATGTACCCCGAGGACGGGGCCCTAAAGGGGGAGATAAAGCAGAGGCTAAAGCTAAAGGACGGGGGGCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTACCCGGGGCCTACAACGTGAACATAAAGCTAGACATAACCTCGCACAACGAGGACTACACCATAGTGGAGCAGTACGAGAGGGCCGAGGGGAGGCACTCGACCGGGGGGATGGACGAGCTATACAAGTAG', 'average_distance_score': 2.400642079220615, 'weakest_link_score': 2.400642079220615}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTGTCGAAGGGGGAGGAGGACAACATGGCCATAATAAAGGAGTTCATGAGGTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATAGAGGGGGAGGGGGAGGGGAGGCCCTACGAGGGGACCCAGACCGCCAAGCTAAAGGTGACCAAGGGGGGGCCCCTACCCTTCGCCTGGGACATACTATCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATACCCGACTACCTAAAGCTATCGTTCCCCGAGGGGTTCAAGTGGGAGAGGGTGATGAACTTCGAGGACGGGGGGGTGGTGACCGTGACCCAGGACTCGTCGCTACAGGACGGGGAGTTCATATACAAGGTGAAGCTAAGGGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAGACCATGGGGTGGGAGGCCTCGTCGGAGAGGATGTACCCCGAGGACGGGGCCCTAAAGGGGGAGATAAAGCAGAGGCTAAAGCTAAAGGACGGGGGGCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTACCCGGGGCCTACAACGTGAACATAAAGCTAGACATAACCTCGCACAACGAGGACTACACCATAGTGGAGCAGTACGAGAGGGCCGAGGGGAGGCACTCGACCGGGGGGATGGACGAGCTATACAAGTAG', 'average_distance_score': 2.400642079220615, 'weakest_link_score': 2.400642079220615}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTGTCGAAGGGGGAGGAGGACAACATGGCCATAATAAAGGAGTTCATGAGGTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATAGAGGGGGAGGGGGAGGGGAGGCCCTACGAGGGGACCCAGACCGCCAAGCTAAAGGTGACCAAGGGGGGGCCCCTACCCTTCGCCTGGGACATACTATCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATACCCGACTACCTAAAGCTATCGTTCCCCGAGGGGTTCAAGTGGGAGAGGGTGATGAACTTCGAGGACGGGGGGGTGGTGACCGTGACCCAGGACTCGTCGCTACAGGACGGGGAGTTCATATACAAGGTGAAGCTAAGGGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAGACCATGGGGTGGGAGGCCTCGTCGGAGAGGATGTACCCCGAGGACGGGGCCCTAAAGGGGGAGATAAAGCAGAGGCTAAAGCTAAAGGACGGGGGGCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTACCCGGGGCCTACAACGTGAACATAAAGCTAGACATAACCTCGCACAACGAGGACTACACCATAGTGGAGCAGTACGAGAGGGCCGAGGGGAGGCACTCGACCGGGGGGATGGACGAGCTATACAAGTAG', 'average_distance_score': 2.400642079220615, 'weakest_link_score': 2.400642079220615}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry_debug\\CAI_zscore_bulk_aa_ratio_ecoli_opt_True_Wi2e\n",
      "Final output: {'user_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'optimization_method': 'zscore_bulk_aa_ratio', 'optimization_cub_index': 'CAI', 'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.037037037037037035, 'TGA': 0.24183006535947713, 'GCA': 0.6554010349288486, 'GCC': 0.6652652005174644, 'GCG': 1.0, 'GCT': 0.6266170763260025, 'TGC': 1.0, 'TGT': 0.7162162162162162, 'GAC': 0.8422464375523889, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.36087048383688997, 'TTC': 1.0, 'TTT': 0.6891233766233766, 'GGA': 0.12073112073112073, 'GGC': 1.0, 'GGG': 0.19111752445085778, 'GGT': 0.9866923200256533, 'CAC': 1.0, 'CAT': 0.7259593679458239, 'ATA': 0.03593829462620784, 'ATC': 1.0, 'ATT': 0.7563993897270725, 'AAA': 1.0, 'AAG': 0.2715144230769231, 'CTA': 0.0255237398225231, 'CTC': 0.1333821242338304, 'CTG': 1.0, 'CTT': 0.11572591711645777, 'TTA': 0.10557131095050773, 'TTG': 0.12322751806788033, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.4079954093343534, 'CCA': 0.2681349767629825, 'CCC': 0.08830066680137402, 'CCG': 1.0, 'CCT': 0.19337239846433624, 'CAA': 0.3511947863866763, 'CAG': 1.0, 'AGA': 0.018922018348623854, 'AGG': 0.011085626911314986, 'CGA': 0.03631498470948013, 'CGC': 0.7138761467889909, 'CGG': 0.05600152905198777, 'CGT': 1.0, 'AGC': 1.0, 'AGT': 0.3131715081575806, 'TCA': 0.3048149621965778, 'TCC': 0.7942697970553123, 'TCG': 0.40907282132908873, 'TCT': 0.8901711102268206, 'ACA': 0.14701775384001597, 'ACC': 1.0, 'ACG': 0.3457011769399561, 'ACT': 0.4374625972471574, 'GTA': 0.531166518254675, 'GTC': 0.5075690115761353, 'GTG': 0.9971059661620658, 'GTT': 1.0, 'TGG': 1.0, 'TAC': 1.0, 'TAT': 0.857994781960492}, 'cai_avg': 0.6260311851192147, 'cai_std': 0.09192830548402788, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.10739856801909307, 'TGA': 0.1694510739856802, 'GCA': 1.0, 'GCC': 0.5246426811237062, 'GCG': 0.9196648595367175, 'GCT': 0.9361754558896007, 'TGC': 1.0, 'TGT': 0.7495652173913042, 'GAC': 0.6699738903394257, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.3869690068323596, 'TTC': 0.82402496099844, 'TTT': 1.0, 'GGA': 0.8480035296712993, 'GGC': 1.0, 'GGG': 0.29340392675932053, 'GGT': 0.6170306640194132, 'CAC': 0.6778131634819533, 'CAT': 1.0, 'ATA': 0.11114994757077944, 'ATC': 0.8386927647675637, 'ATT': 1.0, 'AAA': 1.0, 'AAG': 0.3398391420911528, 'CTA': 0.14688262702462837, 'CTC': 0.3046372309740404, 'CTG': 0.6640781007321944, 'CTT': 1.0, 'TTA': 0.5893055247392945, 'TTG': 0.41202573774129136, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.8348958333333333, 'CCA': 0.4923917284432306, 'CCC': 0.07452204447912603, 'CCG': 1.0, 'CCT': 0.7553648068669527, 'CAA': 1.0, 'CAG': 0.8251624335499115, 'AGA': 0.72005772005772, 'AGG': 0.09620009620009619, 'CGA': 0.19672919672919673, 'CGC': 0.984126984126984, 'CGG': 0.2751322751322751, 'CGT': 1.0, 'AGC': 0.8834134615384617, 'AGT': 0.33774038461538464, 'TCA': 0.9487179487179488, 'TCC': 0.42588141025641024, 'TCG': 0.2560096153846154, 'TCT': 1.0, 'ACA': 1.0, 'ACC': 0.22142692218887092, 'ACG': 0.5329023320249365, 'ACT': 0.4301546986839067, 'GTA': 0.7337249143416544, 'GTC': 0.681106216348507, 'GTG': 0.6669114047968673, 'GTT': 1.0, 'TGG': 1.0, 'TAC': 0.7787396121883656, 'TAT': 1.0}, 'cai_avg': 0.7354163127574911, 'cai_std': 0.04172052669683352, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}]}, 'orf': [{'initial_sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_sequence': 'ATGGTGTCGAAGGGGGAGGAGGACAACATGGCCATAATAAAGGAGTTCATGAGGTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATAGAGGGGGAGGGGGAGGGGAGGCCCTACGAGGGGACCCAGACCGCCAAGCTAAAGGTGACCAAGGGGGGGCCCCTACCCTTCGCCTGGGACATACTATCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATACCCGACTACCTAAAGCTATCGTTCCCCGAGGGGTTCAAGTGGGAGAGGGTGATGAACTTCGAGGACGGGGGGGTGGTGACCGTGACCCAGGACTCGTCGCTACAGGACGGGGAGTTCATATACAAGGTGAAGCTAAGGGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAGACCATGGGGTGGGAGGCCTCGTCGGAGAGGATGTACCCCGAGGACGGGGCCCTAAAGGGGGAGATAAAGCAGAGGCTAAAGCTAAAGGACGGGGGGCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTACCCGGGGCCTACAACGTGAACATAAAGCTAGACATAACCTCGCACAACGAGGACTACACCATAGTGGAGCAGTACGAGAGGGCCGAGGGGAGGCACTCGACCGGGGGGATGGACGAGCTATACAAGTAG', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TAG', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTA', 'H': 'CAC', 'R': 'AGG', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'initial_sequence_optimization_score': 1.0913186611767312, 'final_sequence_optimization_score': 1.3677525746264352, 'run_time': 17.45299999997951, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTA', 'H': 'CAC', 'R': 'AGG', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 1.22898899368115}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TAG', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTA', 'H': 'CAC', 'R': 'AGG', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 1.3676783765799794}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TAG', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTA', 'H': 'CAC', 'R': 'AGG', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 1.3677525746264352}]}, {'initial_sequence': 'ATGGTGTCGAAAGGGGAAGAAGACAACATGGCCATCATCAAAGAGTTCATGAGATTTAAAGTCCATATGGAAGGATCTGTGAACGGACACGAGTTCGAGATTGAAGGTGAAGGGGAAGGTAGACCATACGAGGGTACGCAGACAGCCAAGCTCAAGGTAACGAAGGGAGGCCCTCTCCCGTTCGCCTGGGACATACTTAGTCCGCAATTTATGTACGGAAGTAAGGCCTACGTGAAGCATCCTGCAGATATCCCTGATTACTTGAAGTTGAGTTTCCCTGAAGGCTTTAAATGGGAAAGAGTCATGAACTTTGAGGATGGTGGCGTTGTTACAGTGACCCAAGACTCCTCTCTTCAGGATGGAGAGTTTATCTACAAGGTTAAACTCCGTGGAACGAACTTTCCCTCAGATGGGCCGGTGATGCAGAAAAAGACCATGGGGTGGGAGGCGTCGTCGGAGAGGATGTACCCTGAGGACGGTGCACTCAAAGGTGAGATAAAGCAGCGCCTGAAACTAAAAGACGGTGGCCACTATGACGCTGAGGTGAAAACCACATATAAGGCGAAAAAACCAGTGCAACTTCCTGGCGCGTACAACGTAAACATCAAGCTTGATATCACTAGCCACAACGAGGACTACACTATCGTAGAACAGTATGAACGCGCTGAAGGGCGACACTCCACGGGTGGTATGGACGAGCTGTATAAGTAA', 'final_sequence': 'ATGGTGTCGAAGGGGGAGGAGGACAACATGGCCATAATAAAGGAGTTCATGAGGTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATAGAGGGGGAGGGGGAGGGGAGGCCCTACGAGGGGACCCAGACCGCCAAGCTAAAGGTGACCAAGGGGGGGCCCCTACCCTTCGCCTGGGACATACTATCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATACCCGACTACCTAAAGCTATCGTTCCCCGAGGGGTTCAAGTGGGAGAGGGTGATGAACTTCGAGGACGGGGGGGTGGTGACCGTGACCCAGGACTCGTCGCTACAGGACGGGGAGTTCATATACAAGGTGAAGCTAAGGGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAGACCATGGGGTGGGAGGCCTCGTCGGAGAGGATGTACCCCGAGGACGGGGCCCTAAAGGGGGAGATAAAGCAGAGGCTAAAGCTAAAGGACGGGGGGCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTACCCGGGGCCTACAACGTGAACATAAAGCTAGACATAACCTCGCACAACGAGGACTACACCATAGTGGAGCAGTACGAGAGGGCCGAGGGGAGGCACTCGACCGGGGGGATGGACGAGCTATACAAGTAG', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TAG', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTA', 'H': 'CAC', 'R': 'AGG', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'initial_sequence_optimization_score': 1.077173848257075, 'final_sequence_optimization_score': 1.3677525746264352, 'run_time': 17.07799999997951, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTA', 'H': 'CAC', 'R': 'AGG', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 1.2962964282636458}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TAG', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTA', 'H': 'CAC', 'R': 'AGG', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 1.3676783765799794}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TAG', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTA', 'H': 'CAC', 'R': 'AGG', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 1.3677525746264352}]}, {'initial_sequence': 'ATGGTTTCGAAGGGTGAAGAGGATAACATGGCCATCATTAAAGAATTCATGAGGTTCAAAGTACATATGGAAGGCTCGGTGAATGGGCATGAATTTGAAATCGAGGGCGAAGGCGAGGGACGCCCGTATGAGGGGACACAGACCGCCAAACTAAAAGTCACTAAAGGCGGGCCGTTGCCCTTCGCTTGGGACATTTTGTCACCGCAATTTATGTACGGGTCAAAAGCTTATGTTAAGCATCCGGCTGATATACCCGACTACCTAAAGCTTTCCTTTCCCGAGGGTTTTAAGTGGGAGCGAGTGATGAATTTTGAAGACGGAGGTGTAGTCACCGTCACGCAAGACAGCTCTCTACAAGATGGGGAGTTTATCTACAAGGTCAAATTGCGTGGGACAAACTTCCCGAGTGATGGACCAGTTATGCAGAAAAAAACCATGGGTTGGGAGGCCTCTTCAGAACGTATGTATCCAGAGGACGGAGCTCTGAAGGGGGAAATAAAACAGCGGCTAAAGTTAAAGGACGGAGGCCACTATGACGCCGAAGTAAAGACTACTTATAAAGCAAAGAAGCCCGTACAGCTTCCGGGCGCTTATAACGTAAATATAAAATTGGACATAACGTCTCACAACGAAGATTACACTATTGTTGAGCAATACGAAAGAGCGGAGGGACGACATTCAACCGGAGGAATGGACGAACTCTATAAATGA', 'final_sequence': 'ATGGTGTCGAAGGGGGAGGAGGACAACATGGCCATAATAAAGGAGTTCATGAGGTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATAGAGGGGGAGGGGGAGGGGAGGCCCTACGAGGGGACCCAGACCGCCAAGCTAAAGGTGACCAAGGGGGGGCCCCTACCCTTCGCCTGGGACATACTATCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATACCCGACTACCTAAAGCTATCGTTCCCCGAGGGGTTCAAGTGGGAGAGGGTGATGAACTTCGAGGACGGGGGGGTGGTGACCGTGACCCAGGACTCGTCGCTACAGGACGGGGAGTTCATATACAAGGTGAAGCTAAGGGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAGACCATGGGGTGGGAGGCCTCGTCGGAGAGGATGTACCCCGAGGACGGGGCCCTAAAGGGGGAGATAAAGCAGAGGCTAAAGCTAAAGGACGGGGGGCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTACCCGGGGCCTACAACGTGAACATAAAGCTAGACATAACCTCGCACAACGAGGACTACACCATAGTGGAGCAGTACGAGAGGGCCGAGGGGAGGCACTCGACCGGGGGGATGGACGAGCTATACAAGTAG', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TAG', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTA', 'H': 'CAC', 'R': 'AGG', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'initial_sequence_optimization_score': 1.1155377281205383, 'final_sequence_optimization_score': 1.3677525746264352, 'run_time': 16.85999999998603, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTA', 'H': 'CAC', 'R': 'AGG', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 1.3083865483157135}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TAG', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTA', 'H': 'CAC', 'R': 'AGG', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 1.3676783765799794}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TAG', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTA', 'H': 'CAC', 'R': 'AGG', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 1.3677525746264352}]}, {'initial_sequence': 'ATGGTTAGCAAAGGGGAAGAGGACAATATGGCGATCATCAAAGAGTTCATGAGGTTTAAGGTTCATATGGAGGGATCGGTGAATGGTCATGAGTTCGAGATTGAAGGCGAGGGAGAAGGGAGGCCCTACGAGGGAACTCAGACAGCGAAACTCAAGGTGACAAAAGGGGGGCCGCTCCCCTTCGCATGGGATATCCTCAGTCCTCAGTTCATGTACGGAAGTAAGGCCTATGTTAAACATCCAGCGGATATACCGGATTACTTGAAGTTATCGTTCCCGGAGGGCTTCAAGTGGGAGAGGGTAATGAACTTCGAAGATGGCGGGGTGGTTACCGTAACCCAAGATAGCTCACTCCAAGACGGGGAGTTCATTTATAAAGTGAAACTTCGCGGTACAAATTTCCCGAGCGACGGTCCAGTGATGCAGAAAAAGACCATGGGTTGGGAGGCCTCTTCCGAACGGATGTATCCCGAGGACGGCGCACTCAAGGGGGAAATTAAACAAAGACTGAAGTTAAAAGACGGTGGTCATTACGATGCTGAAGTCAAGACCACATACAAAGCTAAAAAGCCGGTTCAGCTGCCTGGCGCCTACAATGTCAATATCAAACTAGATATTACTTCCCATAACGAAGATTACACCATCGTCGAACAATACGAGAGAGCGGAAGGGAGACATTCAACTGGCGGGATGGATGAACTATACAAATAG', 'final_sequence': 'ATGGTGTCGAAGGGGGAGGAGGACAACATGGCCATAATAAAGGAGTTCATGAGGTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATAGAGGGGGAGGGGGAGGGGAGGCCCTACGAGGGGACCCAGACCGCCAAGCTAAAGGTGACCAAGGGGGGGCCCCTACCCTTCGCCTGGGACATACTATCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATACCCGACTACCTAAAGCTATCGTTCCCCGAGGGGTTCAAGTGGGAGAGGGTGATGAACTTCGAGGACGGGGGGGTGGTGACCGTGACCCAGGACTCGTCGCTACAGGACGGGGAGTTCATATACAAGGTGAAGCTAAGGGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAGACCATGGGGTGGGAGGCCTCGTCGGAGAGGATGTACCCCGAGGACGGGGCCCTAAAGGGGGAGATAAAGCAGAGGCTAAAGCTAAAGGACGGGGGGCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTACCCGGGGCCTACAACGTGAACATAAAGCTAGACATAACCTCGCACAACGAGGACTACACCATAGTGGAGCAGTACGAGAGGGCCGAGGGGAGGCACTCGACCGGGGGGATGGACGAGCTATACAAGTAG', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TAG', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTA', 'H': 'CAC', 'R': 'AGG', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'initial_sequence_optimization_score': 1.1187572281885003, 'final_sequence_optimization_score': 1.3677525746264352, 'run_time': 17.219000000040978, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTA', 'H': 'CAC', 'R': 'AGG', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 1.3012743470112405}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TAG', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTA', 'H': 'CAC', 'R': 'AGG', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 1.3676783765799794}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TAG', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTA', 'H': 'CAC', 'R': 'AGG', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 1.3677525746264352}]}], 'evaluation': [{'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'cai_initial_score': 0.7891479965436429, 'cai_final_score': 0.307495649675879, 'dist_score': -5.239434626057029}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'cai_initial_score': 0.7608140098814632, 'cai_final_score': 0.34190993377740575, 'dist_score': -10.040718784498258}], 'final_sequence': 'ATGGTGTCGAAGGGGGAGGAGGACAACATGGCCATAATAAAGGAGTTCATGAGGTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATAGAGGGGGAGGGGGAGGGGAGGCCCTACGAGGGGACCCAGACCGCCAAGCTAAAGGTGACCAAGGGGGGGCCCCTACCCTTCGCCTGGGACATACTATCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATACCCGACTACCTAAAGCTATCGTTCCCCGAGGGGTTCAAGTGGGAGAGGGTGATGAACTTCGAGGACGGGGGGGTGGTGACCGTGACCCAGGACTCGTCGCTACAGGACGGGGAGTTCATATACAAGGTGAAGCTAAGGGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAGACCATGGGGTGGGAGGCCTCGTCGGAGAGGATGTACCCCGAGGACGGGGCCCTAAAGGGGGAGATAAAGCAGAGGCTAAAGCTAAAGGACGGGGGGCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTACCCGGGGCCTACAACGTGAACATAAAGCTAGACATAACCTCGCACAACGAGGACTACACCATAGTGGAGCAGTACGAGAGGGCCGAGGGGAGGCACTCGACCGGGGGGATGGACGAGCTATACAAGTAG', 'average_distance_score': 2.400642079220615, 'weakest_link_score': 2.400642079220615}, {'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'cai_initial_score': 0.7891479965436429, 'cai_final_score': 0.307495649675879, 'dist_score': -5.239434626057029}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'cai_initial_score': 0.7608140098814632, 'cai_final_score': 0.34190993377740575, 'dist_score': -10.040718784498258}], 'final_sequence': 'ATGGTGTCGAAGGGGGAGGAGGACAACATGGCCATAATAAAGGAGTTCATGAGGTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATAGAGGGGGAGGGGGAGGGGAGGCCCTACGAGGGGACCCAGACCGCCAAGCTAAAGGTGACCAAGGGGGGGCCCCTACCCTTCGCCTGGGACATACTATCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATACCCGACTACCTAAAGCTATCGTTCCCCGAGGGGTTCAAGTGGGAGAGGGTGATGAACTTCGAGGACGGGGGGGTGGTGACCGTGACCCAGGACTCGTCGCTACAGGACGGGGAGTTCATATACAAGGTGAAGCTAAGGGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAGACCATGGGGTGGGAGGCCTCGTCGGAGAGGATGTACCCCGAGGACGGGGCCCTAAAGGGGGAGATAAAGCAGAGGCTAAAGCTAAAGGACGGGGGGCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTACCCGGGGCCTACAACGTGAACATAAAGCTAGACATAACCTCGCACAACGAGGACTACACCATAGTGGAGCAGTACGAGAGGGCCGAGGGGAGGCACTCGACCGGGGGGATGGACGAGCTATACAAGTAG', 'average_distance_score': 2.400642079220615, 'weakest_link_score': 2.400642079220615}, {'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'cai_initial_score': 0.7891479965436429, 'cai_final_score': 0.307495649675879, 'dist_score': -5.239434626057029}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'cai_initial_score': 0.7608140098814632, 'cai_final_score': 0.34190993377740575, 'dist_score': -10.040718784498258}], 'final_sequence': 'ATGGTGTCGAAGGGGGAGGAGGACAACATGGCCATAATAAAGGAGTTCATGAGGTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATAGAGGGGGAGGGGGAGGGGAGGCCCTACGAGGGGACCCAGACCGCCAAGCTAAAGGTGACCAAGGGGGGGCCCCTACCCTTCGCCTGGGACATACTATCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATACCCGACTACCTAAAGCTATCGTTCCCCGAGGGGTTCAAGTGGGAGAGGGTGATGAACTTCGAGGACGGGGGGGTGGTGACCGTGACCCAGGACTCGTCGCTACAGGACGGGGAGTTCATATACAAGGTGAAGCTAAGGGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAGACCATGGGGTGGGAGGCCTCGTCGGAGAGGATGTACCCCGAGGACGGGGCCCTAAAGGGGGAGATAAAGCAGAGGCTAAAGCTAAAGGACGGGGGGCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTACCCGGGGCCTACAACGTGAACATAAAGCTAGACATAACCTCGCACAACGAGGACTACACCATAGTGGAGCAGTACGAGAGGGCCGAGGGGAGGCACTCGACCGGGGGGATGGACGAGCTATACAAGTAG', 'average_distance_score': 2.400642079220615, 'weakest_link_score': 2.400642079220615}, {'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'cai_initial_score': 0.7891479965436429, 'cai_final_score': 0.307495649675879, 'dist_score': -5.239434626057029}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'cai_initial_score': 0.7608140098814632, 'cai_final_score': 0.34190993377740575, 'dist_score': -10.040718784498258}], 'final_sequence': 'ATGGTGTCGAAGGGGGAGGAGGACAACATGGCCATAATAAAGGAGTTCATGAGGTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATAGAGGGGGAGGGGGAGGGGAGGCCCTACGAGGGGACCCAGACCGCCAAGCTAAAGGTGACCAAGGGGGGGCCCCTACCCTTCGCCTGGGACATACTATCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATACCCGACTACCTAAAGCTATCGTTCCCCGAGGGGTTCAAGTGGGAGAGGGTGATGAACTTCGAGGACGGGGGGGTGGTGACCGTGACCCAGGACTCGTCGCTACAGGACGGGGAGTTCATATACAAGGTGAAGCTAAGGGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAGACCATGGGGTGGGAGGCCTCGTCGGAGAGGATGTACCCCGAGGACGGGGCCCTAAAGGGGGAGATAAAGCAGAGGCTAAAGCTAAAGGACGGGGGGCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTACCCGGGGCCTACAACGTGAACATAAAGCTAGACATAACCTCGCACAACGAGGACTACACCATAGTGGAGCAGTACGAGAGGGCCGAGGGGAGGCACTCGACCGGGGGGATGGACGAGCTATACAAGTAG', 'average_distance_score': 2.400642079220615, 'weakest_link_score': 2.400642079220615}], 'final_evaluation': {'final_sequence': 'ATGGTGTCGAAGGGGGAGGAGGACAACATGGCCATAATAAAGGAGTTCATGAGGTTCAAGGTGCACATGGAGGGGTCGGTGAACGGGCACGAGTTCGAGATAGAGGGGGAGGGGGAGGGGAGGCCCTACGAGGGGACCCAGACCGCCAAGCTAAAGGTGACCAAGGGGGGGCCCCTACCCTTCGCCTGGGACATACTATCGCCCCAGTTCATGTACGGGTCGAAGGCCTACGTGAAGCACCCCGCCGACATACCCGACTACCTAAAGCTATCGTTCCCCGAGGGGTTCAAGTGGGAGAGGGTGATGAACTTCGAGGACGGGGGGGTGGTGACCGTGACCCAGGACTCGTCGCTACAGGACGGGGAGTTCATATACAAGGTGAAGCTAAGGGGGACCAACTTCCCCTCGGACGGGCCCGTGATGCAGAAGAAGACCATGGGGTGGGAGGCCTCGTCGGAGAGGATGTACCCCGAGGACGGGGCCCTAAAGGGGGAGATAAAGCAGAGGCTAAAGCTAAAGGACGGGGGGCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTACCCGGGGCCTACAACGTGAACATAAAGCTAGACATAACCTCGCACAACGAGGACTACACCATAGTGGAGCAGTACGAGAGGGCCGAGGGGAGGCACTCGACCGGGGGGATGGACGAGCTATACAAGTAG', 'average_distance_score': 2.400642079220615, 'weakest_link_score': 2.400642079220615}, 'zip_output_file_path': 'results\\\\mcherry_debug\\\\CAI_zscore_bulk_aa_ratio_ecoli_opt_True_Wi2e\\\\communique_results.zip'}\n",
      "##########################\n",
      "# USER INPUT INFORMATION #\n",
      "##########################\n",
      "Organism is optimized\n",
      "Information about Bacillus subtilis:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 98 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3550\n",
      "Calculate CAI weights from a reference set of 532 highly expressed genes from estimated expression dictionary.\n",
      "name=Bacillus subtilis, cai_std=0.04172052669683352, cai_avg=0.7354163127574911\n",
      "Organism is de-optimized\n",
      "Information about Escherichia coli:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 90 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3895\n",
      "Calculate CAI weights from a reference set of 584 highly expressed genes from estimated expression dictionary.\n",
      "name=Escherichia coli, cai_std=0.09192830548402788, cai_avg=0.6260311851192147\n",
      "Bacillus subtilis has weight of 1.0\n",
      "Escherichia coli has weight of 1.0\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is: ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Total input processing time: 22.958054304122925\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "\n",
      "Optimization method used is: OptimizationMethod.zscore_bulk_aa_ratio\n",
      "Running ORF optimization for initial sequence: ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Running ORF optimization for initial sequence: ATGGTGAGTAAAGGTGAGGAGGATAACATGGCGATTATTAAAGAATTCATGCGATTCAAGGTTCATATGGAAGGCTCAGTTAACGGTCACGAATTTGAAATAGAAGGGGAAGGGGAAGGGCGTCCGTACGAGGGAACGCAGACCGCCAAGTTAAAGGTGACCAAAGGAGGTCCACTGCCATTTGCGTGGGACATACTCTCGCCCCAGTTTATGTACGGCTCCAAGGCTTATGTTAAACATCCAGCGGACATACCGGATTACTTGAAATTGAGTTTTCCAGAGGGTTTTAAGTGGGAGAGAGTAATGAATTTTGAAGACGGCGGGGTAGTCACAGTCACTCAGGACTCGAGCCTACAGGACGGCGAATTTATTTATAAAGTAAAACTACGGGGAACTAACTTCCCATCCGATGGGCCTGTAATGCAGAAAAAAACGATGGGGTGGGAAGCAAGTTCTGAACGTATGTACCCAGAAGACGGAGCTTTGAAAGGCGAGATTAAACAGCGTTTGAAGCTTAAAGACGGTGGACATTACGACGCAGAGGTAAAAACCACCTATAAAGCCAAGAAACCGGTGCAACTCCCAGGTGCCTACAACGTCAATATCAAACTGGATATCACGTCCCACAATGAGGATTATACAATTGTAGAGCAGTATGAGCGAGCTGAAGGACGACATTCTACTGGCGGCATGGACGAACTATATAAGTAG\n",
      "Running ORF optimization for initial sequence: ATGGTGAGCAAAGGTGAAGAAGACAACATGGCCATTATCAAAGAATTCATGAGATTCAAGGTGCACATGGAGGGCTCTGTTAATGGCCATGAGTTTGAAATAGAAGGTGAAGGAGAAGGCCGTCCCTATGAGGGTACTCAGACCGCCAAGTTAAAGGTAACAAAGGGCGGACCGTTACCTTTTGCGTGGGATATTCTATCCCCTCAGTTTATGTATGGTTCCAAAGCTTATGTGAAACACCCCGCAGACATCCCCGATTACTTAAAATTATCTTTTCCCGAAGGCTTCAAATGGGAGCGGGTCATGAACTTCGAAGATGGCGGCGTCGTTACGGTCACTCAGGACAGCAGCCTACAAGATGGAGAATTTATCTACAAGGTAAAATTGCGCGGCACCAATTTTCCATCGGACGGCCCCGTCATGCAAAAGAAAACAATGGGATGGGAGGCTTCTAGTGAACGGATGTATCCAGAAGATGGGGCGCTAAAGGGCGAGATTAAGCAACGACTAAAGCTTAAGGACGGAGGACACTACGATGCGGAAGTTAAAACCACCTATAAAGCCAAGAAGCCGGTCCAACTTCCTGGAGCCTACAACGTGAATATAAAACTTGATATAACAAGTCATAATGAGGATTATACAATTGTGGAGCAGTATGAGAGGGCGGAGGGTAGGCATAGTACTGGTGGGATGGATGAACTGTACAAATAG\n",
      "Running ORF optimization for initial sequence: ATGGTAAGTAAAGGAGAAGAGGACAACATGGCAATTATCAAAGAGTTCATGAGATTTAAAGTGCATATGGAGGGCTCGGTCAACGGGCATGAGTTTGAAATAGAAGGAGAGGGCGAAGGACGCCCGTATGAGGGCACGCAGACGGCCAAGCTTAAAGTGACGAAGGGAGGTCCACTACCGTTTGCCTGGGATATACTTTCTCCCCAGTTCATGTATGGATCGAAGGCTTACGTAAAACACCCTGCGGATATCCCCGACTACCTCAAGCTGAGTTTCCCCGAAGGATTCAAATGGGAACGCGTTATGAACTTTGAGGATGGGGGGGTAGTAACAGTGACACAAGATTCCTCGTTACAGGACGGTGAGTTCATCTACAAGGTGAAACTCAGGGGAACAAATTTTCCATCTGATGGCCCCGTTATGCAAAAGAAGACCATGGGATGGGAGGCGTCCTCCGAAAGGATGTACCCTGAAGATGGTGCTCTCAAAGGAGAAATTAAACAGCGTCTAAAGTTGAAAGACGGCGGGCATTATGATGCGGAAGTAAAGACAACTTATAAAGCCAAAAAGCCGGTACAGCTGCCCGGCGCTTACAATGTCAACATTAAGTTGGATATAACATCCCATAATGAAGACTACACGATAGTCGAACAGTACGAAAGGGCCGAAGGTAGGCATTCAACGGGTGGGATGGATGAGCTTTACAAATAA\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGAGAAGAAGATAATATGGCAATTATTAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGAGTTGTTACAGTTACACAAGATTCATCACTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCATCATCAGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGACATTATGATGCAGAAGTTAAAACAACATATAAAGCAAAAAAACCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGAATGGATGAACTTTATAAATAA', 'average_distance_score': 4.297672133245705, 'weakest_link_score': 4.297672133245705}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGAGAAGAAGATAATATGGCAATTATTAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGAGTTGTTACAGTTACACAAGATTCATCACTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCATCATCAGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGACATTATGATGCAGAAGTTAAAACAACATATAAAGCAAAAAAACCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGAATGGATGAACTTTATAAATAA', 'average_distance_score': 4.297672133245705, 'weakest_link_score': 4.297672133245705}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGAGAAGAAGATAATATGGCAATTATTAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGAGTTGTTACAGTTACACAAGATTCATCACTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCATCATCAGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGACATTATGATGCAGAAGTTAAAACAACATATAAAGCAAAAAAACCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGAATGGATGAACTTTATAAATAA', 'average_distance_score': 4.297672133245705, 'weakest_link_score': 4.297672133245705}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGAGAAGAAGATAATATGGCAATTATTAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGAGTTGTTACAGTTACACAAGATTCATCACTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCATCATCAGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGACATTATGATGCAGAAGTTAAAACAACATATAAAGCAAAAAAACCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGAATGGATGAACTTTATAAATAA', 'average_distance_score': 4.297672133245705, 'weakest_link_score': 4.297672133245705}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGAGAAGAAGATAATATGGCAATTATTAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGAGTTGTTACAGTTACACAAGATTCATCACTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCATCATCAGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGACATTATGATGCAGAAGTTAAAACAACATATAAAGCAAAAAAACCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGAATGGATGAACTTTATAAATAA', 'average_distance_score': 4.297672133245705, 'weakest_link_score': 4.297672133245705}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry_debug\\CAI_zscore_bulk_aa_ratio_ecoli_opt_False_bsu6\n",
      "Final output: {'user_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'optimization_method': 'zscore_bulk_aa_ratio', 'optimization_cub_index': 'CAI', 'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.10739856801909307, 'TGA': 0.1694510739856802, 'GCA': 1.0, 'GCC': 0.5246426811237062, 'GCG': 0.9196648595367175, 'GCT': 0.9361754558896007, 'TGC': 1.0, 'TGT': 0.7495652173913042, 'GAC': 0.6699738903394257, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.3869690068323596, 'TTC': 0.82402496099844, 'TTT': 1.0, 'GGA': 0.8480035296712993, 'GGC': 1.0, 'GGG': 0.29340392675932053, 'GGT': 0.6170306640194132, 'CAC': 0.6778131634819533, 'CAT': 1.0, 'ATA': 0.11114994757077944, 'ATC': 0.8386927647675637, 'ATT': 1.0, 'AAA': 1.0, 'AAG': 0.3398391420911528, 'CTA': 0.14688262702462837, 'CTC': 0.3046372309740404, 'CTG': 0.6640781007321944, 'CTT': 1.0, 'TTA': 0.5893055247392945, 'TTG': 0.41202573774129136, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.8348958333333333, 'CCA': 0.4923917284432306, 'CCC': 0.07452204447912603, 'CCG': 1.0, 'CCT': 0.7553648068669527, 'CAA': 1.0, 'CAG': 0.8251624335499115, 'AGA': 0.72005772005772, 'AGG': 0.09620009620009619, 'CGA': 0.19672919672919673, 'CGC': 0.984126984126984, 'CGG': 0.2751322751322751, 'CGT': 1.0, 'AGC': 0.8834134615384617, 'AGT': 0.33774038461538464, 'TCA': 0.9487179487179488, 'TCC': 0.42588141025641024, 'TCG': 0.2560096153846154, 'TCT': 1.0, 'ACA': 1.0, 'ACC': 0.22142692218887092, 'ACG': 0.5329023320249365, 'ACT': 0.4301546986839067, 'GTA': 0.7337249143416544, 'GTC': 0.681106216348507, 'GTG': 0.6669114047968673, 'GTT': 1.0, 'TGG': 1.0, 'TAC': 0.7787396121883656, 'TAT': 1.0}, 'cai_avg': 0.7354163127574911, 'cai_std': 0.04172052669683352, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}, {'name': 'Escherichia coli', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.037037037037037035, 'TGA': 0.24183006535947713, 'GCA': 0.6554010349288486, 'GCC': 0.6652652005174644, 'GCG': 1.0, 'GCT': 0.6266170763260025, 'TGC': 1.0, 'TGT': 0.7162162162162162, 'GAC': 0.8422464375523889, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.36087048383688997, 'TTC': 1.0, 'TTT': 0.6891233766233766, 'GGA': 0.12073112073112073, 'GGC': 1.0, 'GGG': 0.19111752445085778, 'GGT': 0.9866923200256533, 'CAC': 1.0, 'CAT': 0.7259593679458239, 'ATA': 0.03593829462620784, 'ATC': 1.0, 'ATT': 0.7563993897270725, 'AAA': 1.0, 'AAG': 0.2715144230769231, 'CTA': 0.0255237398225231, 'CTC': 0.1333821242338304, 'CTG': 1.0, 'CTT': 0.11572591711645777, 'TTA': 0.10557131095050773, 'TTG': 0.12322751806788033, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.4079954093343534, 'CCA': 0.2681349767629825, 'CCC': 0.08830066680137402, 'CCG': 1.0, 'CCT': 0.19337239846433624, 'CAA': 0.3511947863866763, 'CAG': 1.0, 'AGA': 0.018922018348623854, 'AGG': 0.011085626911314986, 'CGA': 0.03631498470948013, 'CGC': 0.7138761467889909, 'CGG': 0.05600152905198777, 'CGT': 1.0, 'AGC': 1.0, 'AGT': 0.3131715081575806, 'TCA': 0.3048149621965778, 'TCC': 0.7942697970553123, 'TCG': 0.40907282132908873, 'TCT': 0.8901711102268206, 'ACA': 0.14701775384001597, 'ACC': 1.0, 'ACG': 0.3457011769399561, 'ACT': 0.4374625972471574, 'GTA': 0.531166518254675, 'GTC': 0.5075690115761353, 'GTG': 0.9971059661620658, 'GTT': 1.0, 'TGG': 1.0, 'TAC': 1.0, 'TAT': 0.857994781960492}, 'cai_avg': 0.6260311851192147, 'cai_std': 0.09192830548402788, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}]}, 'orf': [{'initial_sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_sequence': 'ATGGTTTCAAAAGGAGAAGAAGATAATATGGCAATTATTAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGAGTTGTTACAGTTACACAAGATTCATCACTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCATCATCAGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGACATTATGATGCAGAAGTTAAAACAACATATAAAGCAAAAAAACCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGAATGGATGAACTTTATAAATAA', 'iterations_count': 2, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'initial_sequence_optimization_score': 0.9163226430323611, 'final_sequence_optimization_score': 1.3505332122051765, 'run_time': 14.422000000020489, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 1.4142135623730951}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 1.3505332122051765}]}, {'initial_sequence': 'ATGGTGAGTAAAGGTGAGGAGGATAACATGGCGATTATTAAAGAATTCATGCGATTCAAGGTTCATATGGAAGGCTCAGTTAACGGTCACGAATTTGAAATAGAAGGGGAAGGGGAAGGGCGTCCGTACGAGGGAACGCAGACCGCCAAGTTAAAGGTGACCAAAGGAGGTCCACTGCCATTTGCGTGGGACATACTCTCGCCCCAGTTTATGTACGGCTCCAAGGCTTATGTTAAACATCCAGCGGACATACCGGATTACTTGAAATTGAGTTTTCCAGAGGGTTTTAAGTGGGAGAGAGTAATGAATTTTGAAGACGGCGGGGTAGTCACAGTCACTCAGGACTCGAGCCTACAGGACGGCGAATTTATTTATAAAGTAAAACTACGGGGAACTAACTTCCCATCCGATGGGCCTGTAATGCAGAAAAAAACGATGGGGTGGGAAGCAAGTTCTGAACGTATGTACCCAGAAGACGGAGCTTTGAAAGGCGAGATTAAACAGCGTTTGAAGCTTAAAGACGGTGGACATTACGACGCAGAGGTAAAAACCACCTATAAAGCCAAGAAACCGGTGCAACTCCCAGGTGCCTACAACGTCAATATCAAACTGGATATCACGTCCCACAATGAGGATTATACAATTGTAGAGCAGTATGAGCGAGCTGAAGGACGACATTCTACTGGCGGCATGGACGAACTATATAAGTAG', 'final_sequence': 'ATGGTTTCAAAAGGAGAAGAAGATAATATGGCAATTATTAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGAGTTGTTACAGTTACACAAGATTCATCACTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCATCATCAGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGACATTATGATGCAGAAGTTAAAACAACATATAAAGCAAAAAAACCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGAATGGATGAACTTTATAAATAA', 'iterations_count': 2, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'initial_sequence_optimization_score': 0.9093192033228018, 'final_sequence_optimization_score': 1.3505332122051765, 'run_time': 13.327999999979511, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 1.3102159732205356}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 1.3505332122051765}]}, {'initial_sequence': 'ATGGTGAGCAAAGGTGAAGAAGACAACATGGCCATTATCAAAGAATTCATGAGATTCAAGGTGCACATGGAGGGCTCTGTTAATGGCCATGAGTTTGAAATAGAAGGTGAAGGAGAAGGCCGTCCCTATGAGGGTACTCAGACCGCCAAGTTAAAGGTAACAAAGGGCGGACCGTTACCTTTTGCGTGGGATATTCTATCCCCTCAGTTTATGTATGGTTCCAAAGCTTATGTGAAACACCCCGCAGACATCCCCGATTACTTAAAATTATCTTTTCCCGAAGGCTTCAAATGGGAGCGGGTCATGAACTTCGAAGATGGCGGCGTCGTTACGGTCACTCAGGACAGCAGCCTACAAGATGGAGAATTTATCTACAAGGTAAAATTGCGCGGCACCAATTTTCCATCGGACGGCCCCGTCATGCAAAAGAAAACAATGGGATGGGAGGCTTCTAGTGAACGGATGTATCCAGAAGATGGGGCGCTAAAGGGCGAGATTAAGCAACGACTAAAGCTTAAGGACGGAGGACACTACGATGCGGAAGTTAAAACCACCTATAAAGCCAAGAAGCCGGTCCAACTTCCTGGAGCCTACAACGTGAATATAAAACTTGATATAACAAGTCATAATGAGGATTATACAATTGTGGAGCAGTATGAGAGGGCGGAGGGTAGGCATAGTACTGGTGGGATGGATGAACTGTACAAATAG', 'final_sequence': 'ATGGTTTCAAAAGGAGAAGAAGATAATATGGCAATTATTAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGAGTTGTTACAGTTACACAAGATTCATCACTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCATCATCAGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGACATTATGATGCAGAAGTTAAAACAACATATAAAGCAAAAAAACCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGAATGGATGAACTTTATAAATAA', 'iterations_count': 2, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'initial_sequence_optimization_score': 0.9202819737124921, 'final_sequence_optimization_score': 1.3505332122051765, 'run_time': 13.73499999998603, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 1.2953896209876032}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 1.3505332122051765}]}, {'initial_sequence': 'ATGGTAAGTAAAGGAGAAGAGGACAACATGGCAATTATCAAAGAGTTCATGAGATTTAAAGTGCATATGGAGGGCTCGGTCAACGGGCATGAGTTTGAAATAGAAGGAGAGGGCGAAGGACGCCCGTATGAGGGCACGCAGACGGCCAAGCTTAAAGTGACGAAGGGAGGTCCACTACCGTTTGCCTGGGATATACTTTCTCCCCAGTTCATGTATGGATCGAAGGCTTACGTAAAACACCCTGCGGATATCCCCGACTACCTCAAGCTGAGTTTCCCCGAAGGATTCAAATGGGAACGCGTTATGAACTTTGAGGATGGGGGGGTAGTAACAGTGACACAAGATTCCTCGTTACAGGACGGTGAGTTCATCTACAAGGTGAAACTCAGGGGAACAAATTTTCCATCTGATGGCCCCGTTATGCAAAAGAAGACCATGGGATGGGAGGCGTCCTCCGAAAGGATGTACCCTGAAGATGGTGCTCTCAAAGGAGAAATTAAACAGCGTCTAAAGTTGAAAGACGGCGGGCATTATGATGCGGAAGTAAAGACAACTTATAAAGCCAAAAAGCCGGTACAGCTGCCCGGCGCTTACAATGTCAACATTAAGTTGGATATAACATCCCATAATGAAGACTACACGATAGTCGAACAGTACGAAAGGGCCGAAGGTAGGCATTCAACGGGTGGGATGGATGAGCTTTACAAATAA', 'final_sequence': 'ATGGTTTCAAAAGGAGAAGAAGATAATATGGCAATTATTAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGAGTTGTTACAGTTACACAAGATTCATCACTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCATCATCAGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGACATTATGATGCAGAAGTTAAAACAACATATAAAGCAAAAAAACCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGAATGGATGAACTTTATAAATAA', 'iterations_count': 2, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'initial_sequence_optimization_score': 0.9103781957013569, 'final_sequence_optimization_score': 1.3505332122051765, 'run_time': 15.577999999979511, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 1.3004279472331128}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 1.3505332122051765}]}], 'evaluation': [{'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'cai_initial_score': 0.7608140098814632, 'cai_final_score': 0.9477930887720342, 'dist_score': 4.481704659417979}, {'name': 'Escherichia coli', 'is_wanted': False, 'cai_initial_score': 0.7891479965436429, 'cai_final_score': 0.4109880780934002, 'dist_score': -4.11363960707343}], 'final_sequence': 'ATGGTTTCAAAAGGAGAAGAAGATAATATGGCAATTATTAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGAGTTGTTACAGTTACACAAGATTCATCACTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCATCATCAGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGACATTATGATGCAGAAGTTAAAACAACATATAAAGCAAAAAAACCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGAATGGATGAACTTTATAAATAA', 'average_distance_score': 4.297672133245705, 'weakest_link_score': 4.297672133245705}, {'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'cai_initial_score': 0.7608140098814632, 'cai_final_score': 0.9477930887720342, 'dist_score': 4.481704659417979}, {'name': 'Escherichia coli', 'is_wanted': False, 'cai_initial_score': 0.7891479965436429, 'cai_final_score': 0.4109880780934002, 'dist_score': -4.11363960707343}], 'final_sequence': 'ATGGTTTCAAAAGGAGAAGAAGATAATATGGCAATTATTAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGAGTTGTTACAGTTACACAAGATTCATCACTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCATCATCAGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGACATTATGATGCAGAAGTTAAAACAACATATAAAGCAAAAAAACCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGAATGGATGAACTTTATAAATAA', 'average_distance_score': 4.297672133245705, 'weakest_link_score': 4.297672133245705}, {'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'cai_initial_score': 0.7608140098814632, 'cai_final_score': 0.9477930887720342, 'dist_score': 4.481704659417979}, {'name': 'Escherichia coli', 'is_wanted': False, 'cai_initial_score': 0.7891479965436429, 'cai_final_score': 0.4109880780934002, 'dist_score': -4.11363960707343}], 'final_sequence': 'ATGGTTTCAAAAGGAGAAGAAGATAATATGGCAATTATTAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGAGTTGTTACAGTTACACAAGATTCATCACTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCATCATCAGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGACATTATGATGCAGAAGTTAAAACAACATATAAAGCAAAAAAACCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGAATGGATGAACTTTATAAATAA', 'average_distance_score': 4.297672133245705, 'weakest_link_score': 4.297672133245705}, {'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'cai_initial_score': 0.7608140098814632, 'cai_final_score': 0.9477930887720342, 'dist_score': 4.481704659417979}, {'name': 'Escherichia coli', 'is_wanted': False, 'cai_initial_score': 0.7891479965436429, 'cai_final_score': 0.4109880780934002, 'dist_score': -4.11363960707343}], 'final_sequence': 'ATGGTTTCAAAAGGAGAAGAAGATAATATGGCAATTATTAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGAGTTGTTACAGTTACACAAGATTCATCACTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCATCATCAGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGACATTATGATGCAGAAGTTAAAACAACATATAAAGCAAAAAAACCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGAATGGATGAACTTTATAAATAA', 'average_distance_score': 4.297672133245705, 'weakest_link_score': 4.297672133245705}], 'final_evaluation': {'final_sequence': 'ATGGTTTCAAAAGGAGAAGAAGATAATATGGCAATTATTAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAATGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAATTTTGAAGATGGAGGAGTTGTTACAGTTACACAAGATTCATCACTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAATTTTCCTTCAGATGGACCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCATCATCAGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGACATTATGATGCAGAAGTTAAAACAACATATAAAGCAAAAAAACCTGTTCAACTTCCTGGAGCATATAATGTTAATATTAAACTTGATATTACATCACATAATGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGAATGGATGAACTTTATAAATAA', 'average_distance_score': 4.297672133245705, 'weakest_link_score': 4.297672133245705}, 'zip_output_file_path': 'results\\\\mcherry_debug\\\\CAI_zscore_bulk_aa_ratio_ecoli_opt_False_bsu6\\\\communique_results.zip'}\n",
      "##########################\n",
      "# USER INPUT INFORMATION #\n",
      "##########################\n",
      "Organism is optimized\n",
      "Information about Escherichia coli:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Escherichia coli. Calculating tAI profile.\n",
      "name=Escherichia coli, tai_std=0.023481599705588187, tai_avg=0.2504950429983414\n",
      "Organism is de-optimized\n",
      "Information about Bacillus subtilis:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Bacillus subtilis. Calculating tAI profile.\n",
      "name=Bacillus subtilis, tai_std=0.027038419737753252, tai_avg=0.3292474875006557\n",
      "Escherichia coli has weight of 1.0\n",
      "Bacillus subtilis has weight of 1.0\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is: ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Total input processing time: 21.117204427719116\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "\n",
      "Optimization method used is: OptimizationMethod.zscore_bulk_aa_ratio\n",
      "Running ORF optimization for initial sequence: ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Running ORF optimization for initial sequence: ATGGTGTCCAAAGGAGAGGAAGATAATATGGCTATAATTAAAGAGTTTATGCGGTTTAAAGTGCATATGGAGGGGTCCGTAAATGGGCATGAGTTTGAAATTGAAGGAGAAGGTGAGGGGCGTCCTTACGAGGGAACACAGACTGCCAAGTTGAAAGTCACAAAGGGTGGTCCCTTGCCTTTCGCCTGGGATATCCTGTCTCCACAGTTTATGTATGGGTCTAAAGCATATGTAAAACATCCTGCGGACATACCTGACTATCTTAAGTTAAGTTTTCCTGAGGGATTCAAATGGGAGCGTGTAATGAACTTCGAAGACGGAGGTGTCGTAACTGTCACGCAGGACTCCAGCCTTCAGGACGGTGAGTTCATATACAAAGTAAAATTGAGAGGAACGAACTTCCCCTCTGACGGCCCCGTTATGCAAAAAAAGACAATGGGATGGGAGGCCTCATCCGAGAGGATGTACCCCGAGGACGGGGCGCTCAAGGGTGAGATCAAACAACGGCTGAAGTTGAAGGACGGAGGGCATTATGATGCAGAAGTGAAGACGACTTATAAAGCCAAAAAGCCTGTCCAACTCCCCGGAGCTTACAATGTAAACATCAAGCTGGATATCACTTCTCATAATGAGGACTATACCATAGTAGAACAGTATGAGAGAGCAGAGGGGCGCCATAGTACCGGCGGTATGGACGAGCTTTACAAGTGA\n",
      "Running ORF optimization for initial sequence: ATGGTTTCGAAAGGTGAGGAAGATAACATGGCTATAATTAAAGAATTTATGAGATTTAAAGTACATATGGAGGGATCAGTTAACGGCCATGAGTTCGAGATAGAGGGCGAAGGAGAGGGGCGACCGTACGAGGGTACACAAACCGCCAAACTGAAAGTCACTAAGGGAGGCCCCCTGCCATTTGCTTGGGATATACTTAGTCCCCAGTTTATGTATGGCAGCAAAGCGTATGTGAAACACCCTGCGGACATACCCGATTACCTAAAACTCAGCTTTCCGGAAGGATTCAAATGGGAGCGTGTAATGAACTTTGAAGACGGTGGGGTGGTTACAGTTACACAAGACTCGAGTCTGCAGGATGGCGAATTCATTTATAAGGTTAAACTAAGAGGGACCAACTTCCCTAGTGATGGCCCTGTCATGCAAAAAAAGACCATGGGATGGGAAGCCTCATCGGAGCGTATGTATCCAGAAGACGGAGCTCTAAAGGGAGAGATAAAACAGCGATTGAAGCTAAAGGATGGTGGGCACTATGATGCCGAAGTGAAAACTACGTATAAAGCGAAAAAACCCGTGCAGCTCCCGGGTGCTTATAATGTGAATATAAAATTAGATATAACTAGTCACAACGAGGATTATACGATCGTTGAACAGTATGAGCGAGCAGAAGGTCGCCACTCTACAGGGGGTATGGACGAACTATATAAGTAG\n",
      "Running ORF optimization for initial sequence: ATGGTCTCGAAGGGTGAGGAGGATAATATGGCCATAATTAAAGAATTTATGAGGTTTAAAGTTCACATGGAGGGATCGGTCAACGGCCACGAATTTGAGATAGAGGGGGAGGGGGAAGGACGCCCGTATGAAGGTACGCAAACAGCGAAGCTCAAAGTTACAAAAGGGGGACCGTTACCTTTCGCCTGGGATATATTGTCACCACAGTTCATGTATGGGAGCAAAGCATATGTAAAACATCCAGCTGATATACCAGATTACCTGAAACTTAGTTTTCCCGAAGGCTTTAAGTGGGAAAGAGTAATGAACTTCGAGGACGGTGGGGTAGTAACAGTGACCCAAGATTCCTCACTGCAGGATGGCGAGTTCATTTATAAAGTTAAGCTACGTGGAACTAACTTTCCCTCAGACGGGCCCGTAATGCAGAAGAAAACCATGGGGTGGGAGGCTTCGAGCGAAAGGATGTATCCCGAGGATGGTGCGCTAAAGGGCGAAATCAAGCAAAGACTTAAGCTCAAAGATGGCGGCCACTATGACGCGGAGGTCAAAACAACATACAAGGCCAAAAAGCCAGTTCAGTTACCTGGCGCGTACAATGTAAATATCAAACTCGACATTACGAGTCACAATGAGGACTATACGATCGTCGAGCAGTATGAAAGGGCTGAAGGACGCCATTCCACGGGTGGTATGGATGAACTCTACAAATAG\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCTAAGGGGGAGGAGGATAATATGGCTATAATAAAGGAGTTTATGCGATTTAAGGTTCATATGGAGGGGTCTGTTAATGGGCATGAGTTTGAGATAGAGGGGGAGGGGGAGGGGCGACCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGGCCGCTGCCGTTTGCTTGGGATATACTGTCTCCGCAGTTTATGTATGGGTCTAAGGCTTATGTTAAGCATCCGGCTGATATACCGGATTATCTGAAGCTGTCTTTTCCGGAGGGGTTTAAGTGGGAGCGAGTTATGAATTTTGAGGATGGGGGGGTTGTTACTGTTACTCAGGATTCTTCTCTGCAGGATGGGGAGTTTATATATAAGGTTAAGCTGCGAGGGACTAATTTTCCGTCTGATGGGCCGGTTATGCAGAAGAAGACTATGGGGTGGGAGGCTTCTTCTGAGCGAATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATAAAGCAGCGACTGAAGCTGAAGGATGGGGGGCATTATGATGCTGAGGTTAAGACTACTTATAAGGCTAAGAAGCCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATAAAGCTGGATATAACTTCTCATAATGAGGATTATACTATAGTTGAGCAGTATGAGCGAGCTGAGGGGCGACATTCTACTGGGGGGATGGATGAGCTGTATAAGTAA', 'average_distance_score': 0.6490571983680749, 'weakest_link_score': 0.6490571983680749}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCTAAGGGGGAGGAGGATAATATGGCTATAATAAAGGAGTTTATGCGATTTAAGGTTCATATGGAGGGGTCTGTTAATGGGCATGAGTTTGAGATAGAGGGGGAGGGGGAGGGGCGACCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGGCCGCTGCCGTTTGCTTGGGATATACTGTCTCCGCAGTTTATGTATGGGTCTAAGGCTTATGTTAAGCATCCGGCTGATATACCGGATTATCTGAAGCTGTCTTTTCCGGAGGGGTTTAAGTGGGAGCGAGTTATGAATTTTGAGGATGGGGGGGTTGTTACTGTTACTCAGGATTCTTCTCTGCAGGATGGGGAGTTTATATATAAGGTTAAGCTGCGAGGGACTAATTTTCCGTCTGATGGGCCGGTTATGCAGAAGAAGACTATGGGGTGGGAGGCTTCTTCTGAGCGAATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATAAAGCAGCGACTGAAGCTGAAGGATGGGGGGCATTATGATGCTGAGGTTAAGACTACTTATAAGGCTAAGAAGCCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATAAAGCTGGATATAACTTCTCATAATGAGGATTATACTATAGTTGAGCAGTATGAGCGAGCTGAGGGGCGACATTCTACTGGGGGGATGGATGAGCTGTATAAGTGA', 'average_distance_score': 0.6490571983680749, 'weakest_link_score': 0.6490571983680749}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCTAAGGGGGAGGAGGATAATATGGCTATAATAAAGGAGTTTATGCGATTTAAGGTTCATATGGAGGGGTCTGTTAATGGGCATGAGTTTGAGATAGAGGGGGAGGGGGAGGGGCGACCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGGCCGCTGCCGTTTGCTTGGGATATACTGTCTCCGCAGTTTATGTATGGGTCTAAGGCTTATGTTAAGCATCCGGCTGATATACCGGATTATCTGAAGCTGTCTTTTCCGGAGGGGTTTAAGTGGGAGCGAGTTATGAATTTTGAGGATGGGGGGGTTGTTACTGTTACTCAGGATTCTTCTCTGCAGGATGGGGAGTTTATATATAAGGTTAAGCTGCGAGGGACTAATTTTCCGTCTGATGGGCCGGTTATGCAGAAGAAGACTATGGGGTGGGAGGCTTCTTCTGAGCGAATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATAAAGCAGCGACTGAAGCTGAAGGATGGGGGGCATTATGATGCTGAGGTTAAGACTACTTATAAGGCTAAGAAGCCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATAAAGCTGGATATAACTTCTCATAATGAGGATTATACTATAGTTGAGCAGTATGAGCGAGCTGAGGGGCGACATTCTACTGGGGGGATGGATGAGCTGTATAAGTAG', 'average_distance_score': 0.6490571983680749, 'weakest_link_score': 0.6490571983680749}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCTAAGGGGGAGGAGGATAATATGGCTATAATAAAGGAGTTTATGCGATTTAAGGTTCATATGGAGGGGTCTGTTAATGGGCATGAGTTTGAGATAGAGGGGGAGGGGGAGGGGCGACCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGGCCGCTGCCGTTTGCTTGGGATATACTGTCTCCGCAGTTTATGTATGGGTCTAAGGCTTATGTTAAGCATCCGGCTGATATACCGGATTATCTGAAGCTGTCTTTTCCGGAGGGGTTTAAGTGGGAGCGAGTTATGAATTTTGAGGATGGGGGGGTTGTTACTGTTACTCAGGATTCTTCTCTGCAGGATGGGGAGTTTATATATAAGGTTAAGCTGCGAGGGACTAATTTTCCGTCTGATGGGCCGGTTATGCAGAAGAAGACTATGGGGTGGGAGGCTTCTTCTGAGCGAATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATAAAGCAGCGACTGAAGCTGAAGGATGGGGGGCATTATGATGCTGAGGTTAAGACTACTTATAAGGCTAAGAAGCCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATAAAGCTGGATATAACTTCTCATAATGAGGATTATACTATAGTTGAGCAGTATGAGCGAGCTGAGGGGCGACATTCTACTGGGGGGATGGATGAGCTGTATAAGTAG', 'average_distance_score': 0.6490571983680749, 'weakest_link_score': 0.6490571983680749}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTTTCTAAGGGGGAGGAGGATAATATGGCTATAATAAAGGAGTTTATGCGATTTAAGGTTCATATGGAGGGGTCTGTTAATGGGCATGAGTTTGAGATAGAGGGGGAGGGGGAGGGGCGACCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGGCCGCTGCCGTTTGCTTGGGATATACTGTCTCCGCAGTTTATGTATGGGTCTAAGGCTTATGTTAAGCATCCGGCTGATATACCGGATTATCTGAAGCTGTCTTTTCCGGAGGGGTTTAAGTGGGAGCGAGTTATGAATTTTGAGGATGGGGGGGTTGTTACTGTTACTCAGGATTCTTCTCTGCAGGATGGGGAGTTTATATATAAGGTTAAGCTGCGAGGGACTAATTTTCCGTCTGATGGGCCGGTTATGCAGAAGAAGACTATGGGGTGGGAGGCTTCTTCTGAGCGAATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATAAAGCAGCGACTGAAGCTGAAGGATGGGGGGCATTATGATGCTGAGGTTAAGACTACTTATAAGGCTAAGAAGCCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATAAAGCTGGATATAACTTCTCATAATGAGGATTATACTATAGTTGAGCAGTATGAGCGAGCTGAGGGGCGACATTCTACTGGGGGGATGGATGAGCTGTATAAGTAA', 'average_distance_score': 0.6490571983680749, 'weakest_link_score': 0.6490571983680749}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry_debug\\tAI_zscore_bulk_aa_ratio_ecoli_opt_True_VqXq\n",
      "Final output: {'user_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'optimization_method': 'zscore_bulk_aa_ratio', 'optimization_cub_index': 'tAI', 'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': None, 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.75, 'AAC': 0.5, 'AAG': 0.23999999999999996, 'AAT': 0.29500000000000004, 'ACA': 0.1525, 'ACC': 0.25, 'ACG': 0.29, 'ACT': 0.14750000000000002, 'AGA': 0.13874999999999998, 'AGC': 0.125, 'AGG': 0.16499999999999998, 'AGT': 0.07375000000000001, 'ATA': 0.10999999999999999, 'ATC': 0.375, 'ATG': 1.0, 'ATT': 0.22125000000000003, 'CAA': 0.27749999999999997, 'CAC': 0.125, 'CAG': 0.32999999999999996, 'CAT': 0.07375000000000001, 'CCA': 0.13874999999999998, 'CCC': 0.125, 'CCG': 0.16499999999999998, 'CCT': 0.07375000000000001, 'CGA': 0.013799999999999993, 'CGC': 0.36, 'CGG': 0.125, 'CGT': 0.5, 'CTA': 0.18, 'CTC': 0.125, 'CTG': 0.54, 'CTT': 0.07375000000000001, 'GAA': 0.5, 'GAC': 0.375, 'GAG': 0.15999999999999998, 'GAT': 0.22125000000000003, 'GCA': 0.375, 'GCC': 0.25, 'GCG': 0.11999999999999998, 'GCT': 0.14750000000000002, 'GGA': 0.13874999999999998, 'GGC': 0.5, 'GGG': 0.16499999999999998, 'GGT': 0.29500000000000004, 'GTA': 0.625, 'GTC': 0.25, 'GTG': 0.19999999999999996, 'GTT': 0.14750000000000002, 'TAC': 0.375, 'TAT': 0.22125000000000003, 'TCA': 0.13874999999999998, 'TCC': 0.25, 'TCG': 0.16499999999999998, 'TCT': 0.14750000000000002, 'TGC': 0.125, 'TGG': 0.125, 'TGT': 0.07375000000000001, 'TTA': 0.13874999999999998, 'TTC': 0.25, 'TTG': 0.16499999999999998, 'TTT': 0.14750000000000002}, 'tai_avg': 0.2504950429983414, 'tai_std': 0.023481599705588187}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': None, 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.6666666666666666, 'AAC': 0.6666666666666666, 'AAG': 0.2133333333333333, 'AAT': 0.39333333333333337, 'ACA': 0.6666666666666666, 'ACC': 0.16666666666666666, 'ACG': 0.2133333333333333, 'ACT': 0.09833333333333334, 'AGA': 0.18499999999999997, 'AGC': 0.3333333333333333, 'AGG': 0.21999999999999997, 'AGT': 0.19666666666666668, 'ATA': 0.10999999999999999, 'ATC': 0.5, 'ATG': 1.0, 'ATT': 0.29500000000000004, 'CAA': 0.6666666666666666, 'CAC': 0.3333333333333333, 'CAG': 0.2133333333333333, 'CAT': 0.19666666666666668, 'CCA': 0.5, 'CCG': 0.15999999999999998, 'CGA': 0.01839999999999999, 'CGC': 0.48, 'CGG': 0.16666666666666666, 'CGT': 0.6666666666666666, 'CTA': 0.3516666666666666, 'CTC': 0.16666666666666666, 'CTG': 0.2733333333333333, 'CTT': 0.09833333333333334, 'GAA': 1.0, 'GAC': 0.6666666666666666, 'GAG': 0.31999999999999995, 'GAT': 0.39333333333333337, 'GCA': 0.8333333333333334, 'GCC': 0.16666666666666666, 'GCG': 0.2666666666666666, 'GCT': 0.09833333333333334, 'GGA': 0.5, 'GGC': 0.6666666666666666, 'GGG': 0.15999999999999998, 'GGT': 0.39333333333333337, 'GTA': 0.6666666666666666, 'GTC': 0.16666666666666666, 'GTG': 0.2133333333333333, 'GTT': 0.09833333333333334, 'TAC': 0.3333333333333333, 'TAT': 0.19666666666666668, 'TCA': 0.3333333333333333, 'TCC': 0.16666666666666666, 'TCG': 0.10666666666666665, 'TCT': 0.09833333333333334, 'TGC': 0.16666666666666666, 'TGG': 0.16666666666666666, 'TGT': 0.09833333333333334, 'TTA': 0.5183333333333333, 'TTC': 0.5, 'TTG': 0.32666666666666666, 'TTT': 0.29500000000000004}, 'tai_avg': 0.3292474875006557, 'tai_std': 0.027038419737753252}]}, 'orf': [{'initial_sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_sequence': 'ATGGTTTCTAAGGGGGAGGAGGATAATATGGCTATAATAAAGGAGTTTATGCGATTTAAGGTTCATATGGAGGGGTCTGTTAATGGGCATGAGTTTGAGATAGAGGGGGAGGGGGAGGGGCGACCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGGCCGCTGCCGTTTGCTTGGGATATACTGTCTCCGCAGTTTATGTATGGGTCTAAGGCTTATGTTAAGCATCCGGCTGATATACCGGATTATCTGAAGCTGTCTTTTCCGGAGGGGTTTAAGTGGGAGCGAGTTATGAATTTTGAGGATGGGGGGGTTGTTACTGTTACTCAGGATTCTTCTCTGCAGGATGGGGAGTTTATATATAAGGTTAAGCTGCGAGGGACTAATTTTCCGTCTGATGGGCCGGTTATGCAGAAGAAGACTATGGGGTGGGAGGCTTCTTCTGAGCGAATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATAAAGCAGCGACTGAAGCTGAAGGATGGGGGGCATTATGATGCTGAGGTTAAGACTACTTATAAGGCTAAGAAGCCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATAAAGCTGGATATAACTTCTCATAATGAGGATTATACTATAGTTGAGCAGTATGAGCGAGCTGAGGGGCGACATTCTACTGGGGGGATGGATGAGCTGTATAAGTAA', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCT', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAT', 'R': 'CGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'initial_sequence_optimization_score': 1.099065776777178, 'final_sequence_optimization_score': 1.255449786604244, 'run_time': 17.328000000095926, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAT', 'R': 'CGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'sequence_score': 1.1394096619456868}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCT', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAT', 'R': 'CGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'sequence_score': 1.2556268798430237}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCT', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAT', 'R': 'CGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'sequence_score': 1.255449786604244}]}, {'initial_sequence': 'ATGGTGTCCAAAGGAGAGGAAGATAATATGGCTATAATTAAAGAGTTTATGCGGTTTAAAGTGCATATGGAGGGGTCCGTAAATGGGCATGAGTTTGAAATTGAAGGAGAAGGTGAGGGGCGTCCTTACGAGGGAACACAGACTGCCAAGTTGAAAGTCACAAAGGGTGGTCCCTTGCCTTTCGCCTGGGATATCCTGTCTCCACAGTTTATGTATGGGTCTAAAGCATATGTAAAACATCCTGCGGACATACCTGACTATCTTAAGTTAAGTTTTCCTGAGGGATTCAAATGGGAGCGTGTAATGAACTTCGAAGACGGAGGTGTCGTAACTGTCACGCAGGACTCCAGCCTTCAGGACGGTGAGTTCATATACAAAGTAAAATTGAGAGGAACGAACTTCCCCTCTGACGGCCCCGTTATGCAAAAAAAGACAATGGGATGGGAGGCCTCATCCGAGAGGATGTACCCCGAGGACGGGGCGCTCAAGGGTGAGATCAAACAACGGCTGAAGTTGAAGGACGGAGGGCATTATGATGCAGAAGTGAAGACGACTTATAAAGCCAAAAAGCCTGTCCAACTCCCCGGAGCTTACAATGTAAACATCAAGCTGGATATCACTTCTCATAATGAGGACTATACCATAGTAGAACAGTATGAGAGAGCAGAGGGGCGCCATAGTACCGGCGGTATGGACGAGCTTTACAAGTGA', 'final_sequence': 'ATGGTTTCTAAGGGGGAGGAGGATAATATGGCTATAATAAAGGAGTTTATGCGATTTAAGGTTCATATGGAGGGGTCTGTTAATGGGCATGAGTTTGAGATAGAGGGGGAGGGGGAGGGGCGACCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGGCCGCTGCCGTTTGCTTGGGATATACTGTCTCCGCAGTTTATGTATGGGTCTAAGGCTTATGTTAAGCATCCGGCTGATATACCGGATTATCTGAAGCTGTCTTTTCCGGAGGGGTTTAAGTGGGAGCGAGTTATGAATTTTGAGGATGGGGGGGTTGTTACTGTTACTCAGGATTCTTCTCTGCAGGATGGGGAGTTTATATATAAGGTTAAGCTGCGAGGGACTAATTTTCCGTCTGATGGGCCGGTTATGCAGAAGAAGACTATGGGGTGGGAGGCTTCTTCTGAGCGAATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATAAAGCAGCGACTGAAGCTGAAGGATGGGGGGCATTATGATGCTGAGGTTAAGACTACTTATAAGGCTAAGAAGCCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATAAAGCTGGATATAACTTCTCATAATGAGGATTATACTATAGTTGAGCAGTATGAGCGAGCTGAGGGGCGACATTCTACTGGGGGGATGGATGAGCTGTATAAGTGA', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCT', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAT', 'R': 'CGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'initial_sequence_optimization_score': 0.9273369029892886, 'final_sequence_optimization_score': 1.255449786604244, 'run_time': 18.530999999959022, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAT', 'R': 'CGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'sequence_score': 1.1680196024699185}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCT', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAT', 'R': 'CGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'sequence_score': 1.2556268798430237}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCT', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAT', 'R': 'CGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'sequence_score': 1.255449786604244}]}, {'initial_sequence': 'ATGGTTTCGAAAGGTGAGGAAGATAACATGGCTATAATTAAAGAATTTATGAGATTTAAAGTACATATGGAGGGATCAGTTAACGGCCATGAGTTCGAGATAGAGGGCGAAGGAGAGGGGCGACCGTACGAGGGTACACAAACCGCCAAACTGAAAGTCACTAAGGGAGGCCCCCTGCCATTTGCTTGGGATATACTTAGTCCCCAGTTTATGTATGGCAGCAAAGCGTATGTGAAACACCCTGCGGACATACCCGATTACCTAAAACTCAGCTTTCCGGAAGGATTCAAATGGGAGCGTGTAATGAACTTTGAAGACGGTGGGGTGGTTACAGTTACACAAGACTCGAGTCTGCAGGATGGCGAATTCATTTATAAGGTTAAACTAAGAGGGACCAACTTCCCTAGTGATGGCCCTGTCATGCAAAAAAAGACCATGGGATGGGAAGCCTCATCGGAGCGTATGTATCCAGAAGACGGAGCTCTAAAGGGAGAGATAAAACAGCGATTGAAGCTAAAGGATGGTGGGCACTATGATGCCGAAGTGAAAACTACGTATAAAGCGAAAAAACCCGTGCAGCTCCCGGGTGCTTATAATGTGAATATAAAATTAGATATAACTAGTCACAACGAGGATTATACGATCGTTGAACAGTATGAGCGAGCAGAAGGTCGCCACTCTACAGGGGGTATGGACGAACTATATAAGTAG', 'final_sequence': 'ATGGTTTCTAAGGGGGAGGAGGATAATATGGCTATAATAAAGGAGTTTATGCGATTTAAGGTTCATATGGAGGGGTCTGTTAATGGGCATGAGTTTGAGATAGAGGGGGAGGGGGAGGGGCGACCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGGCCGCTGCCGTTTGCTTGGGATATACTGTCTCCGCAGTTTATGTATGGGTCTAAGGCTTATGTTAAGCATCCGGCTGATATACCGGATTATCTGAAGCTGTCTTTTCCGGAGGGGTTTAAGTGGGAGCGAGTTATGAATTTTGAGGATGGGGGGGTTGTTACTGTTACTCAGGATTCTTCTCTGCAGGATGGGGAGTTTATATATAAGGTTAAGCTGCGAGGGACTAATTTTCCGTCTGATGGGCCGGTTATGCAGAAGAAGACTATGGGGTGGGAGGCTTCTTCTGAGCGAATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATAAAGCAGCGACTGAAGCTGAAGGATGGGGGGCATTATGATGCTGAGGTTAAGACTACTTATAAGGCTAAGAAGCCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATAAAGCTGGATATAACTTCTCATAATGAGGATTATACTATAGTTGAGCAGTATGAGCGAGCTGAGGGGCGACATTCTACTGGGGGGATGGATGAGCTGTATAAGTAG', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCT', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAT', 'R': 'CGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'initial_sequence_optimization_score': 0.9360780309231128, 'final_sequence_optimization_score': 1.255449786604244, 'run_time': 19.10999999998603, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAT', 'R': 'CGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'sequence_score': 1.1783294842917285}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCT', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAT', 'R': 'CGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'sequence_score': 1.2556268798430237}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCT', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAT', 'R': 'CGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'sequence_score': 1.255449786604244}]}, {'initial_sequence': 'ATGGTCTCGAAGGGTGAGGAGGATAATATGGCCATAATTAAAGAATTTATGAGGTTTAAAGTTCACATGGAGGGATCGGTCAACGGCCACGAATTTGAGATAGAGGGGGAGGGGGAAGGACGCCCGTATGAAGGTACGCAAACAGCGAAGCTCAAAGTTACAAAAGGGGGACCGTTACCTTTCGCCTGGGATATATTGTCACCACAGTTCATGTATGGGAGCAAAGCATATGTAAAACATCCAGCTGATATACCAGATTACCTGAAACTTAGTTTTCCCGAAGGCTTTAAGTGGGAAAGAGTAATGAACTTCGAGGACGGTGGGGTAGTAACAGTGACCCAAGATTCCTCACTGCAGGATGGCGAGTTCATTTATAAAGTTAAGCTACGTGGAACTAACTTTCCCTCAGACGGGCCCGTAATGCAGAAGAAAACCATGGGGTGGGAGGCTTCGAGCGAAAGGATGTATCCCGAGGATGGTGCGCTAAAGGGCGAAATCAAGCAAAGACTTAAGCTCAAAGATGGCGGCCACTATGACGCGGAGGTCAAAACAACATACAAGGCCAAAAAGCCAGTTCAGTTACCTGGCGCGTACAATGTAAATATCAAACTCGACATTACGAGTCACAATGAGGACTATACGATCGTCGAGCAGTATGAAAGGGCTGAAGGACGCCATTCCACGGGTGGTATGGATGAACTCTACAAATAG', 'final_sequence': 'ATGGTTTCTAAGGGGGAGGAGGATAATATGGCTATAATAAAGGAGTTTATGCGATTTAAGGTTCATATGGAGGGGTCTGTTAATGGGCATGAGTTTGAGATAGAGGGGGAGGGGGAGGGGCGACCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGGCCGCTGCCGTTTGCTTGGGATATACTGTCTCCGCAGTTTATGTATGGGTCTAAGGCTTATGTTAAGCATCCGGCTGATATACCGGATTATCTGAAGCTGTCTTTTCCGGAGGGGTTTAAGTGGGAGCGAGTTATGAATTTTGAGGATGGGGGGGTTGTTACTGTTACTCAGGATTCTTCTCTGCAGGATGGGGAGTTTATATATAAGGTTAAGCTGCGAGGGACTAATTTTCCGTCTGATGGGCCGGTTATGCAGAAGAAGACTATGGGGTGGGAGGCTTCTTCTGAGCGAATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATAAAGCAGCGACTGAAGCTGAAGGATGGGGGGCATTATGATGCTGAGGTTAAGACTACTTATAAGGCTAAGAAGCCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATAAAGCTGGATATAACTTCTCATAATGAGGATTATACTATAGTTGAGCAGTATGAGCGAGCTGAGGGGCGACATTCTACTGGGGGGATGGATGAGCTGTATAAGTAG', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCT', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAT', 'R': 'CGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'initial_sequence_optimization_score': 0.9167352395097994, 'final_sequence_optimization_score': 1.255449786604244, 'run_time': 19.20299999997951, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAT', 'R': 'CGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'sequence_score': 1.1634346520800827}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCT', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAT', 'R': 'CGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'sequence_score': 1.2556268798430237}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCT', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAT', 'R': 'CGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'sequence_score': 1.255449786604244}]}], 'evaluation': [{'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'tai_initial_score': 0.2799603826276079, 'tai_final_score': 0.16935251405440976, 'dist_score': -4.710406018329131}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'tai_initial_score': 0.33499296920046817, 'tai_final_score': 0.17253207221507372, 'dist_score': -6.008520415065281}], 'final_sequence': 'ATGGTTTCTAAGGGGGAGGAGGATAATATGGCTATAATAAAGGAGTTTATGCGATTTAAGGTTCATATGGAGGGGTCTGTTAATGGGCATGAGTTTGAGATAGAGGGGGAGGGGGAGGGGCGACCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGGCCGCTGCCGTTTGCTTGGGATATACTGTCTCCGCAGTTTATGTATGGGTCTAAGGCTTATGTTAAGCATCCGGCTGATATACCGGATTATCTGAAGCTGTCTTTTCCGGAGGGGTTTAAGTGGGAGCGAGTTATGAATTTTGAGGATGGGGGGGTTGTTACTGTTACTCAGGATTCTTCTCTGCAGGATGGGGAGTTTATATATAAGGTTAAGCTGCGAGGGACTAATTTTCCGTCTGATGGGCCGGTTATGCAGAAGAAGACTATGGGGTGGGAGGCTTCTTCTGAGCGAATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATAAAGCAGCGACTGAAGCTGAAGGATGGGGGGCATTATGATGCTGAGGTTAAGACTACTTATAAGGCTAAGAAGCCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATAAAGCTGGATATAACTTCTCATAATGAGGATTATACTATAGTTGAGCAGTATGAGCGAGCTGAGGGGCGACATTCTACTGGGGGGATGGATGAGCTGTATAAGTAA', 'average_distance_score': 0.6490571983680749, 'weakest_link_score': 0.6490571983680749}, {'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'tai_initial_score': 0.2799603826276079, 'tai_final_score': 0.16935251405440976, 'dist_score': -4.710406018329131}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'tai_initial_score': 0.33499296920046817, 'tai_final_score': 0.17253207221507372, 'dist_score': -6.008520415065281}], 'final_sequence': 'ATGGTTTCTAAGGGGGAGGAGGATAATATGGCTATAATAAAGGAGTTTATGCGATTTAAGGTTCATATGGAGGGGTCTGTTAATGGGCATGAGTTTGAGATAGAGGGGGAGGGGGAGGGGCGACCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGGCCGCTGCCGTTTGCTTGGGATATACTGTCTCCGCAGTTTATGTATGGGTCTAAGGCTTATGTTAAGCATCCGGCTGATATACCGGATTATCTGAAGCTGTCTTTTCCGGAGGGGTTTAAGTGGGAGCGAGTTATGAATTTTGAGGATGGGGGGGTTGTTACTGTTACTCAGGATTCTTCTCTGCAGGATGGGGAGTTTATATATAAGGTTAAGCTGCGAGGGACTAATTTTCCGTCTGATGGGCCGGTTATGCAGAAGAAGACTATGGGGTGGGAGGCTTCTTCTGAGCGAATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATAAAGCAGCGACTGAAGCTGAAGGATGGGGGGCATTATGATGCTGAGGTTAAGACTACTTATAAGGCTAAGAAGCCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATAAAGCTGGATATAACTTCTCATAATGAGGATTATACTATAGTTGAGCAGTATGAGCGAGCTGAGGGGCGACATTCTACTGGGGGGATGGATGAGCTGTATAAGTGA', 'average_distance_score': 0.6490571983680749, 'weakest_link_score': 0.6490571983680749}, {'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'tai_initial_score': 0.2799603826276079, 'tai_final_score': 0.16935251405440976, 'dist_score': -4.710406018329131}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'tai_initial_score': 0.33499296920046817, 'tai_final_score': 0.17253207221507372, 'dist_score': -6.008520415065281}], 'final_sequence': 'ATGGTTTCTAAGGGGGAGGAGGATAATATGGCTATAATAAAGGAGTTTATGCGATTTAAGGTTCATATGGAGGGGTCTGTTAATGGGCATGAGTTTGAGATAGAGGGGGAGGGGGAGGGGCGACCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGGCCGCTGCCGTTTGCTTGGGATATACTGTCTCCGCAGTTTATGTATGGGTCTAAGGCTTATGTTAAGCATCCGGCTGATATACCGGATTATCTGAAGCTGTCTTTTCCGGAGGGGTTTAAGTGGGAGCGAGTTATGAATTTTGAGGATGGGGGGGTTGTTACTGTTACTCAGGATTCTTCTCTGCAGGATGGGGAGTTTATATATAAGGTTAAGCTGCGAGGGACTAATTTTCCGTCTGATGGGCCGGTTATGCAGAAGAAGACTATGGGGTGGGAGGCTTCTTCTGAGCGAATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATAAAGCAGCGACTGAAGCTGAAGGATGGGGGGCATTATGATGCTGAGGTTAAGACTACTTATAAGGCTAAGAAGCCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATAAAGCTGGATATAACTTCTCATAATGAGGATTATACTATAGTTGAGCAGTATGAGCGAGCTGAGGGGCGACATTCTACTGGGGGGATGGATGAGCTGTATAAGTAG', 'average_distance_score': 0.6490571983680749, 'weakest_link_score': 0.6490571983680749}, {'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'tai_initial_score': 0.2799603826276079, 'tai_final_score': 0.16935251405440976, 'dist_score': -4.710406018329131}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'tai_initial_score': 0.33499296920046817, 'tai_final_score': 0.17253207221507372, 'dist_score': -6.008520415065281}], 'final_sequence': 'ATGGTTTCTAAGGGGGAGGAGGATAATATGGCTATAATAAAGGAGTTTATGCGATTTAAGGTTCATATGGAGGGGTCTGTTAATGGGCATGAGTTTGAGATAGAGGGGGAGGGGGAGGGGCGACCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGGCCGCTGCCGTTTGCTTGGGATATACTGTCTCCGCAGTTTATGTATGGGTCTAAGGCTTATGTTAAGCATCCGGCTGATATACCGGATTATCTGAAGCTGTCTTTTCCGGAGGGGTTTAAGTGGGAGCGAGTTATGAATTTTGAGGATGGGGGGGTTGTTACTGTTACTCAGGATTCTTCTCTGCAGGATGGGGAGTTTATATATAAGGTTAAGCTGCGAGGGACTAATTTTCCGTCTGATGGGCCGGTTATGCAGAAGAAGACTATGGGGTGGGAGGCTTCTTCTGAGCGAATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATAAAGCAGCGACTGAAGCTGAAGGATGGGGGGCATTATGATGCTGAGGTTAAGACTACTTATAAGGCTAAGAAGCCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATAAAGCTGGATATAACTTCTCATAATGAGGATTATACTATAGTTGAGCAGTATGAGCGAGCTGAGGGGCGACATTCTACTGGGGGGATGGATGAGCTGTATAAGTAG', 'average_distance_score': 0.6490571983680749, 'weakest_link_score': 0.6490571983680749}], 'final_evaluation': {'final_sequence': 'ATGGTTTCTAAGGGGGAGGAGGATAATATGGCTATAATAAAGGAGTTTATGCGATTTAAGGTTCATATGGAGGGGTCTGTTAATGGGCATGAGTTTGAGATAGAGGGGGAGGGGGAGGGGCGACCGTATGAGGGGACTCAGACTGCTAAGCTGAAGGTTACTAAGGGGGGGCCGCTGCCGTTTGCTTGGGATATACTGTCTCCGCAGTTTATGTATGGGTCTAAGGCTTATGTTAAGCATCCGGCTGATATACCGGATTATCTGAAGCTGTCTTTTCCGGAGGGGTTTAAGTGGGAGCGAGTTATGAATTTTGAGGATGGGGGGGTTGTTACTGTTACTCAGGATTCTTCTCTGCAGGATGGGGAGTTTATATATAAGGTTAAGCTGCGAGGGACTAATTTTCCGTCTGATGGGCCGGTTATGCAGAAGAAGACTATGGGGTGGGAGGCTTCTTCTGAGCGAATGTATCCGGAGGATGGGGCTCTGAAGGGGGAGATAAAGCAGCGACTGAAGCTGAAGGATGGGGGGCATTATGATGCTGAGGTTAAGACTACTTATAAGGCTAAGAAGCCGGTTCAGCTGCCGGGGGCTTATAATGTTAATATAAAGCTGGATATAACTTCTCATAATGAGGATTATACTATAGTTGAGCAGTATGAGCGAGCTGAGGGGCGACATTCTACTGGGGGGATGGATGAGCTGTATAAGTAA', 'average_distance_score': 0.6490571983680749, 'weakest_link_score': 0.6490571983680749}, 'zip_output_file_path': 'results\\\\mcherry_debug\\\\tAI_zscore_bulk_aa_ratio_ecoli_opt_True_VqXq\\\\communique_results.zip'}\n",
      "##########################\n",
      "# USER INPUT INFORMATION #\n",
      "##########################\n",
      "Organism is optimized\n",
      "Information about Bacillus subtilis:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Bacillus subtilis. Calculating tAI profile.\n",
      "name=Bacillus subtilis, tai_std=0.027038419737753252, tai_avg=0.3292474875006557\n",
      "Organism is de-optimized\n",
      "Information about Escherichia coli:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Escherichia coli. Calculating tAI profile.\n",
      "Encountered unknown error when running modules.\n",
      "Final output: {'error_message': 'Traceback (most recent call last):\\n  File \"C:\\\\Users\\\\Kama\\\\AppData\\\\Local\\\\Programs\\\\Python\\\\Python37-32\\\\lib\\\\urllib\\\\request.py\", line 1317, in do_open\\n    encode_chunked=req.has_header(\\'Transfer-encoding\\'))\\n  File \"C:\\\\Users\\\\Kama\\\\AppData\\\\Local\\\\Programs\\\\Python\\\\Python37-32\\\\lib\\\\http\\\\client.py\", line 1229, in request\\n    self._send_request(method, url, body, headers, encode_chunked)\\n  File \"C:\\\\Users\\\\Kama\\\\AppData\\\\Local\\\\Programs\\\\Python\\\\Python37-32\\\\lib\\\\http\\\\client.py\", line 1275, in _send_request\\n    self.endheaders(body, encode_chunked=encode_chunked)\\n  File \"C:\\\\Users\\\\Kama\\\\AppData\\\\Local\\\\Programs\\\\Python\\\\Python37-32\\\\lib\\\\http\\\\client.py\", line 1224, in endheaders\\n    self._send_output(message_body, encode_chunked=encode_chunked)\\n  File \"C:\\\\Users\\\\Kama\\\\AppData\\\\Local\\\\Programs\\\\Python\\\\Python37-32\\\\lib\\\\http\\\\client.py\", line 1016, in _send_output\\n    self.send(msg)\\n  File \"C:\\\\Users\\\\Kama\\\\AppData\\\\Local\\\\Programs\\\\Python\\\\Python37-32\\\\lib\\\\http\\\\client.py\", line 956, in send\\n    self.connect()\\n  File \"C:\\\\Users\\\\Kama\\\\AppData\\\\Local\\\\Programs\\\\Python\\\\Python37-32\\\\lib\\\\http\\\\client.py\", line 928, in connect\\n    (self.host,self.port), self.timeout, self.source_address)\\n  File \"C:\\\\Users\\\\Kama\\\\AppData\\\\Local\\\\Programs\\\\Python\\\\Python37-32\\\\lib\\\\socket.py\", line 727, in create_connection\\n    raise err\\n  File \"C:\\\\Users\\\\Kama\\\\AppData\\\\Local\\\\Programs\\\\Python\\\\Python37-32\\\\lib\\\\socket.py\", line 716, in create_connection\\n    sock.connect(sa)\\nTimeoutError: [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond\\n\\nDuring handling of the above exception, another exception occurred:\\n\\nTraceback (most recent call last):\\n  File \"C:\\\\projects\\\\Igem_TAU_2021_moran\\\\modules\\\\user_IO\\\\user_input.py\", line 98, in _parse_organisms_list\\n    optimization_cub_index=optimization_cub_index)\\n  File \"C:\\\\projects\\\\Igem_TAU_2021_moran\\\\modules\\\\user_IO\\\\user_input.py\", line 282, in _parse_single_organism_input\\n    tai = calculate_tai_weights(organism_name)\\n  File \"C:\\\\projects\\\\Igem_TAU_2021_moran\\\\modules\\\\user_IO\\\\input_functions.py\", line 237, in calculate_tai_weights\\n    return cb.scores.TrnaAdaptationIndex(url=organism_name_to_url_mapping[organism_name], prokaryote=True)\\n  File \"C:\\\\projects\\\\Igem_TAU_2021_moran\\\\venv\\\\lib\\\\site-packages\\\\codonbias\\\\scores.py\", line 575, in __init__\\n    tGCN = fetch_GCN_from_GtRNAdb(url=url, domain=domain, genome=genome_id)\\n  File \"C:\\\\projects\\\\Igem_TAU_2021_moran\\\\venv\\\\lib\\\\site-packages\\\\codonbias\\\\utils.py\", line 143, in fetch_GCN_from_GtRNAdb\\n    tables = pd.read_html(url)\\n  File \"C:\\\\projects\\\\Igem_TAU_2021_moran\\\\venv\\\\lib\\\\site-packages\\\\pandas\\\\util\\\\_decorators.py\", line 311, in wrapper\\n    return func(*args, **kwargs)\\n  File \"C:\\\\projects\\\\Igem_TAU_2021_moran\\\\venv\\\\lib\\\\site-packages\\\\pandas\\\\io\\\\html.py\", line 1113, in read_html\\n    displayed_only=displayed_only,\\n  File \"C:\\\\projects\\\\Igem_TAU_2021_moran\\\\venv\\\\lib\\\\site-packages\\\\pandas\\\\io\\\\html.py\", line 906, in _parse\\n    tables = p.parse_tables()\\n  File \"C:\\\\projects\\\\Igem_TAU_2021_moran\\\\venv\\\\lib\\\\site-packages\\\\pandas\\\\io\\\\html.py\", line 222, in parse_tables\\n    tables = self._parse_tables(self._build_doc(), self.match, self.attrs)\\n  File \"C:\\\\projects\\\\Igem_TAU_2021_moran\\\\venv\\\\lib\\\\site-packages\\\\pandas\\\\io\\\\html.py\", line 745, in _build_doc\\n    raise e\\n  File \"C:\\\\projects\\\\Igem_TAU_2021_moran\\\\venv\\\\lib\\\\site-packages\\\\pandas\\\\io\\\\html.py\", line 726, in _build_doc\\n    with urlopen(self.io) as f:\\n  File \"C:\\\\projects\\\\Igem_TAU_2021_moran\\\\venv\\\\lib\\\\site-packages\\\\pandas\\\\io\\\\common.py\", line 212, in urlopen\\n    return urllib.request.urlopen(*args, **kwargs)\\n  File \"C:\\\\Users\\\\Kama\\\\AppData\\\\Local\\\\Programs\\\\Python\\\\Python37-32\\\\lib\\\\urllib\\\\request.py\", line 222, in urlopen\\n    return opener.open(url, data, timeout)\\n  File \"C:\\\\Users\\\\Kama\\\\AppData\\\\Local\\\\Programs\\\\Python\\\\Python37-32\\\\lib\\\\urllib\\\\request.py\", line 525, in open\\n    response = self._open(req, data)\\n  File \"C:\\\\Users\\\\Kama\\\\AppData\\\\Local\\\\Programs\\\\Python\\\\Python37-32\\\\lib\\\\urllib\\\\request.py\", line 543, in _open\\n    \\'_open\\', req)\\n  File \"C:\\\\Users\\\\Kama\\\\AppData\\\\Local\\\\Programs\\\\Python\\\\Python37-32\\\\lib\\\\urllib\\\\request.py\", line 503, in _call_chain\\n    result = func(*args)\\n  File \"C:\\\\Users\\\\Kama\\\\AppData\\\\Local\\\\Programs\\\\Python\\\\Python37-32\\\\lib\\\\urllib\\\\request.py\", line 1345, in http_open\\n    return self.do_open(http.client.HTTPConnection, req)\\n  File \"C:\\\\Users\\\\Kama\\\\AppData\\\\Local\\\\Programs\\\\Python\\\\Python37-32\\\\lib\\\\urllib\\\\request.py\", line 1319, in do_open\\n    raise URLError(err)\\nurllib.error.URLError: <urlopen error [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond>\\n\\nDuring handling of the above exception, another exception occurred:\\n\\nTraceback (most recent call last):\\n  File \"C:\\\\projects\\\\Igem_TAU_2021_moran\\\\modules\\\\main.py\", line 36, in run_modules\\n    user_input = user_IO.UserInputModule.run_module(user_inp_raw=user_input_dict, run_summary=run_summary)\\n  File \"C:\\\\projects\\\\Igem_TAU_2021_moran\\\\modules\\\\user_IO\\\\user_input.py\", line 22, in run_module\\n    return cls._parse_input(user_input=user_inp_raw, run_summary=run_summary)\\n  File \"C:\\\\projects\\\\Igem_TAU_2021_moran\\\\modules\\\\user_IO\\\\user_input.py\", line 55, in _parse_input\\n    optimization_cub_index=optimization_cub_index)\\n  File \"C:\\\\projects\\\\Igem_TAU_2021_moran\\\\modules\\\\user_IO\\\\user_input.py\", line 100, in _parse_organisms_list\\n    raise ValueError(f\"Error in organism input: {organism_key}, re-check your input\")\\nValueError: Error in organism input: Escherichia-coli, re-check your input\\n'}\n",
      "##########################\n",
      "# USER INPUT INFORMATION #\n",
      "##########################\n",
      "Organism is optimized\n",
      "Information about Escherichia coli:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 90 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3895\n",
      "Calculate CAI weights from a reference set of 584 highly expressed genes from estimated expression dictionary.\n",
      "name=Escherichia coli, cai_std=0.09192830548402788, cai_avg=0.6260311851192147\n",
      "Organism is de-optimized\n",
      "Information about Bacillus subtilis:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 98 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3550\n",
      "Calculate CAI weights from a reference set of 532 highly expressed genes from estimated expression dictionary.\n",
      "name=Bacillus subtilis, cai_std=0.04172052669683352, cai_avg=0.7354163127574911\n",
      "Escherichia coli has weight of 1.0\n",
      "Bacillus subtilis has weight of 1.0\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is: ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Total input processing time: 19.563666820526123\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "\n",
      "Optimization method used is: OptimizationMethod.zscore_bulk_aa_diff\n",
      "Running ORF optimization for initial sequence: ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Running ORF optimization for initial sequence: ATGGTCAGTAAGGGCGAAGAAGATAATATGGCCATTATAAAGGAGTTTATGCGTTTCAAAGTGCACATGGAAGGGAGCGTTAACGGTCACGAGTTTGAAATCGAGGGTGAGGGAGAAGGGCGCCCGTACGAGGGAACTCAAACTGCGAAGCTTAAGGTGACGAAGGGCGGACCTTTACCCTTCGCATGGGACATACTCTCACCGCAGTTTATGTATGGTTCGAAAGCATATGTGAAGCACCCCGCAGATATACCGGACTACCTCAAATTGTCCTTTCCCGAGGGATTCAAGTGGGAACGAGTGATGAACTTCGAAGATGGTGGAGTGGTCACAGTAACTCAAGACTCCTCCCTCCAAGACGGTGAATTCATCTATAAAGTGAAGCTGAGAGGAACTAATTTTCCATCAGATGGCCCTGTGATGCAGAAGAAAACTATGGGATGGGAAGCATCGTCCGAGCGTATGTACCCGGAGGACGGGGCCTTAAAAGGAGAAATTAAACAGAGGCTGAAACTAAAGGATGGGGGGCATTACGACGCCGAAGTCAAAACCACATACAAAGCGAAAAAACCCGTCCAGTTACCGGGGGCATACAACGTAAACATTAAACTCGATATTACTAGCCACAACGAGGATTATACCATAGTCGAGCAATATGAGCGCGCAGAAGGGAGGCATTCTACTGGTGGCATGGACGAGCTATATAAATAG\n",
      "Running ORF optimization for initial sequence: ATGGTATCGAAGGGGGAAGAGGATAATATGGCCATAATTAAGGAATTCATGCGATTCAAAGTCCATATGGAGGGCTCTGTGAACGGGCACGAATTCGAAATCGAAGGCGAGGGAGAAGGACGCCCGTATGAAGGCACGCAGACGGCGAAACTGAAGGTAACAAAGGGAGGGCCCTTGCCGTTTGCGTGGGACATTCTCTCCCCCCAGTTTATGTATGGATCCAAAGCTTATGTAAAACATCCTGCGGACATACCGGATTATCTCAAGTTATCATTCCCAGAGGGGTTCAAGTGGGAACGTGTAATGAATTTCGAGGACGGTGGCGTGGTAACCGTTACGCAAGACAGCTCCCTACAAGATGGTGAGTTCATTTACAAAGTAAAATTGCGGGGCACCAATTTCCCATCTGACGGTCCCGTAATGCAGAAAAAAACAATGGGCTGGGAAGCATCCTCAGAGCGTATGTATCCGGAGGATGGCGCATTAAAAGGAGAAATTAAGCAGAGGCTCAAATTGAAAGACGGGGGTCACTATGACGCCGAAGTAAAAACTACGTATAAAGCTAAAAAGCCGGTGCAGCTACCCGGGGCCTATAATGTCAACATAAAACTGGATATCACAAGCCACAACGAGGATTATACCATAGTTGAGCAATACGAACGGGCAGAAGGCAGGCACAGTACGGGAGGTATGGATGAGCTTTATAAGTAA\n",
      "Running ORF optimization for initial sequence: ATGGTGAGTAAGGGGGAAGAAGACAACATGGCCATAATCAAGGAGTTTATGAGATTCAAAGTACATATGGAAGGCAGCGTTAATGGCCATGAGTTCGAGATAGAAGGGGAAGGTGAAGGGCGGCCATATGAAGGCACACAGACCGCCAAACTGAAGGTGACGAAAGGAGGCCCTCTACCGTTCGCTTGGGACATCCTCTCGCCACAGTTCATGTACGGAAGCAAGGCTTACGTGAAACACCCTGCTGACATCCCTGACTACCTAAAGCTCTCCTTTCCGGAAGGATTCAAGTGGGAGCGGGTCATGAACTTTGAGGACGGAGGAGTTGTTACTGTTACGCAAGACTCTAGCTTGCAGGACGGAGAGTTCATATATAAAGTGAAATTACGGGGAACTAATTTTCCTAGCGACGGTCCAGTTATGCAGAAGAAAACGATGGGATGGGAGGCCTCCAGTGAACGTATGTATCCTGAGGACGGAGCACTCAAGGGAGAGATCAAGCAGCGTTTGAAACTCAAAGATGGTGGTCACTATGATGCGGAGGTGAAGACTACGTACAAGGCGAAAAAGCCTGTGCAACTTCCTGGCGCCTATAACGTGAACATCAAACTGGACATTACGTCCCATAATGAAGACTACACGATTGTAGAGCAGTACGAACGTGCAGAAGGCAGGCACTCCACTGGGGGCATGGACGAACTCTACAAGTAA\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTGTCGAAGGGTGAGGAGGACAACATGGCCATAATAAAGGAGTTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATAGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGTCCCCTGCCCTTCGCCTGGGACATACTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATACCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGTGTGGTGACCGTGACCCAGGACTCGTCGCTGCAGGACGGTGAGTTCATATACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAGACCATGGGTTGGGAGGCCTCGTCGGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATAAAGCAGCGTCTGAAGCTGAAGGACGGTGGTCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATAAAGCTGGACATAACCTCGCACAACGAGGACTACACCATAGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGTATGGACGAGCTGTACAAGTGA', 'average_distance_score': 2.4750766201733048, 'weakest_link_score': 2.4750766201733048}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTGTCGAAGGGTGAGGAGGACAACATGGCCATAATAAAGGAGTTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATAGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGTCCCCTGCCCTTCGCCTGGGACATACTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATACCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGTGTGGTGACCGTGACCCAGGACTCGTCGCTGCAGGACGGTGAGTTCATATACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAGACCATGGGTTGGGAGGCCTCGTCGGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATAAAGCAGCGTCTGAAGCTGAAGGACGGTGGTCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATAAAGCTGGACATAACCTCGCACAACGAGGACTACACCATAGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGTATGGACGAGCTGTACAAGTGA', 'average_distance_score': 2.4750766201733048, 'weakest_link_score': 2.4750766201733048}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTGTCGAAGGGTGAGGAGGACAACATGGCCATAATAAAGGAGTTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATAGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGTCCCCTGCCCTTCGCCTGGGACATACTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATACCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGTGTGGTGACCGTGACCCAGGACTCGTCGCTGCAGGACGGTGAGTTCATATACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAGACCATGGGTTGGGAGGCCTCGTCGGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATAAAGCAGCGTCTGAAGCTGAAGGACGGTGGTCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATAAAGCTGGACATAACCTCGCACAACGAGGACTACACCATAGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGTATGGACGAGCTGTACAAGTGA', 'average_distance_score': 2.4750766201733048, 'weakest_link_score': 2.4750766201733048}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTGTCGAAGGGTGAGGAGGACAACATGGCCATAATAAAGGAGTTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATAGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGTCCCCTGCCCTTCGCCTGGGACATACTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATACCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGTGTGGTGACCGTGACCCAGGACTCGTCGCTGCAGGACGGTGAGTTCATATACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAGACCATGGGTTGGGAGGCCTCGTCGGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATAAAGCAGCGTCTGAAGCTGAAGGACGGTGGTCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATAAAGCTGGACATAACCTCGCACAACGAGGACTACACCATAGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGTATGGACGAGCTGTACAAGTGA', 'average_distance_score': 2.4750766201733048, 'weakest_link_score': 2.4750766201733048}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTGTCGAAGGGTGAGGAGGACAACATGGCCATAATAAAGGAGTTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATAGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGTCCCCTGCCCTTCGCCTGGGACATACTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATACCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGTGTGGTGACCGTGACCCAGGACTCGTCGCTGCAGGACGGTGAGTTCATATACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAGACCATGGGTTGGGAGGCCTCGTCGGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATAAAGCAGCGTCTGAAGCTGAAGGACGGTGGTCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATAAAGCTGGACATAACCTCGCACAACGAGGACTACACCATAGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGTATGGACGAGCTGTACAAGTGA', 'average_distance_score': 2.4750766201733048, 'weakest_link_score': 2.4750766201733048}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry_debug\\CAI_zscore_bulk_aa_diff_ecoli_opt_True_sSCa\n",
      "Final output: {'user_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'optimization_method': 'zscore_bulk_aa_diff', 'optimization_cub_index': 'CAI', 'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.037037037037037035, 'TGA': 0.24183006535947713, 'GCA': 0.6554010349288486, 'GCC': 0.6652652005174644, 'GCG': 1.0, 'GCT': 0.6266170763260025, 'TGC': 1.0, 'TGT': 0.7162162162162162, 'GAC': 0.8422464375523889, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.36087048383688997, 'TTC': 1.0, 'TTT': 0.6891233766233766, 'GGA': 0.12073112073112073, 'GGC': 1.0, 'GGG': 0.19111752445085778, 'GGT': 0.9866923200256533, 'CAC': 1.0, 'CAT': 0.7259593679458239, 'ATA': 0.03593829462620784, 'ATC': 1.0, 'ATT': 0.7563993897270725, 'AAA': 1.0, 'AAG': 0.2715144230769231, 'CTA': 0.0255237398225231, 'CTC': 0.1333821242338304, 'CTG': 1.0, 'CTT': 0.11572591711645777, 'TTA': 0.10557131095050773, 'TTG': 0.12322751806788033, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.4079954093343534, 'CCA': 0.2681349767629825, 'CCC': 0.08830066680137402, 'CCG': 1.0, 'CCT': 0.19337239846433624, 'CAA': 0.3511947863866763, 'CAG': 1.0, 'AGA': 0.018922018348623854, 'AGG': 0.011085626911314986, 'CGA': 0.03631498470948013, 'CGC': 0.7138761467889909, 'CGG': 0.05600152905198777, 'CGT': 1.0, 'AGC': 1.0, 'AGT': 0.3131715081575806, 'TCA': 0.3048149621965778, 'TCC': 0.7942697970553123, 'TCG': 0.40907282132908873, 'TCT': 0.8901711102268206, 'ACA': 0.14701775384001597, 'ACC': 1.0, 'ACG': 0.3457011769399561, 'ACT': 0.4374625972471574, 'GTA': 0.531166518254675, 'GTC': 0.5075690115761353, 'GTG': 0.9971059661620658, 'GTT': 1.0, 'TGG': 1.0, 'TAC': 1.0, 'TAT': 0.857994781960492}, 'cai_avg': 0.6260311851192147, 'cai_std': 0.09192830548402788, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.10739856801909307, 'TGA': 0.1694510739856802, 'GCA': 1.0, 'GCC': 0.5246426811237062, 'GCG': 0.9196648595367175, 'GCT': 0.9361754558896007, 'TGC': 1.0, 'TGT': 0.7495652173913042, 'GAC': 0.6699738903394257, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.3869690068323596, 'TTC': 0.82402496099844, 'TTT': 1.0, 'GGA': 0.8480035296712993, 'GGC': 1.0, 'GGG': 0.29340392675932053, 'GGT': 0.6170306640194132, 'CAC': 0.6778131634819533, 'CAT': 1.0, 'ATA': 0.11114994757077944, 'ATC': 0.8386927647675637, 'ATT': 1.0, 'AAA': 1.0, 'AAG': 0.3398391420911528, 'CTA': 0.14688262702462837, 'CTC': 0.3046372309740404, 'CTG': 0.6640781007321944, 'CTT': 1.0, 'TTA': 0.5893055247392945, 'TTG': 0.41202573774129136, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.8348958333333333, 'CCA': 0.4923917284432306, 'CCC': 0.07452204447912603, 'CCG': 1.0, 'CCT': 0.7553648068669527, 'CAA': 1.0, 'CAG': 0.8251624335499115, 'AGA': 0.72005772005772, 'AGG': 0.09620009620009619, 'CGA': 0.19672919672919673, 'CGC': 0.984126984126984, 'CGG': 0.2751322751322751, 'CGT': 1.0, 'AGC': 0.8834134615384617, 'AGT': 0.33774038461538464, 'TCA': 0.9487179487179488, 'TCC': 0.42588141025641024, 'TCG': 0.2560096153846154, 'TCT': 1.0, 'ACA': 1.0, 'ACC': 0.22142692218887092, 'ACG': 0.5329023320249365, 'ACT': 0.4301546986839067, 'GTA': 0.7337249143416544, 'GTC': 0.681106216348507, 'GTG': 0.6669114047968673, 'GTT': 1.0, 'TGG': 1.0, 'TAC': 0.7787396121883656, 'TAT': 1.0}, 'cai_avg': 0.7354163127574911, 'cai_std': 0.04172052669683352, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}]}, 'orf': [{'initial_sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_sequence': 'ATGGTGTCGAAGGGTGAGGAGGACAACATGGCCATAATAAAGGAGTTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATAGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGTCCCCTGCCCTTCGCCTGGGACATACTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATACCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGTGTGGTGACCGTGACCCAGGACTCGTCGCTGCAGGACGGTGAGTTCATATACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAGACCATGGGTTGGGAGGCCTCGTCGGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATAAAGCAGCGTCTGAAGCTGAAGGACGGTGGTCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATAAAGCTGGACATAACCTCGCACAACGAGGACTACACCATAGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGTATGGACGAGCTGTACAAGTGA', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'initial_sequence_optimization_score': 0.5828168782323468, 'final_sequence_optimization_score': 3.0578934984056514, 'run_time': 19.405999999959022, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'AGG', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 3.045400268179898}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 3.0578934984056514}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 3.0578934984056514}]}, {'initial_sequence': 'ATGGTCAGTAAGGGCGAAGAAGATAATATGGCCATTATAAAGGAGTTTATGCGTTTCAAAGTGCACATGGAAGGGAGCGTTAACGGTCACGAGTTTGAAATCGAGGGTGAGGGAGAAGGGCGCCCGTACGAGGGAACTCAAACTGCGAAGCTTAAGGTGACGAAGGGCGGACCTTTACCCTTCGCATGGGACATACTCTCACCGCAGTTTATGTATGGTTCGAAAGCATATGTGAAGCACCCCGCAGATATACCGGACTACCTCAAATTGTCCTTTCCCGAGGGATTCAAGTGGGAACGAGTGATGAACTTCGAAGATGGTGGAGTGGTCACAGTAACTCAAGACTCCTCCCTCCAAGACGGTGAATTCATCTATAAAGTGAAGCTGAGAGGAACTAATTTTCCATCAGATGGCCCTGTGATGCAGAAGAAAACTATGGGATGGGAAGCATCGTCCGAGCGTATGTACCCGGAGGACGGGGCCTTAAAAGGAGAAATTAAACAGAGGCTGAAACTAAAGGATGGGGGGCATTACGACGCCGAAGTCAAAACCACATACAAAGCGAAAAAACCCGTCCAGTTACCGGGGGCATACAACGTAAACATTAAACTCGATATTACTAGCCACAACGAGGATTATACCATAGTCGAGCAATATGAGCGCGCAGAAGGGAGGCATTCTACTGGTGGCATGGACGAGCTATATAAATAG', 'final_sequence': 'ATGGTGTCGAAGGGTGAGGAGGACAACATGGCCATAATAAAGGAGTTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATAGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGTCCCCTGCCCTTCGCCTGGGACATACTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATACCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGTGTGGTGACCGTGACCCAGGACTCGTCGCTGCAGGACGGTGAGTTCATATACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAGACCATGGGTTGGGAGGCCTCGTCGGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATAAAGCAGCGTCTGAAGCTGAAGGACGGTGGTCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATAAAGCTGGACATAACCTCGCACAACGAGGACTACACCATAGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGTATGGACGAGCTGTACAAGTGA', 'iterations_count': 5, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'initial_sequence_optimization_score': 0.6460779210273655, 'final_sequence_optimization_score': 3.0578934984056514, 'run_time': 31.344000000040978, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTA', 'H': 'CAC', 'R': 'AGG', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 2.9891767389230144}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'AGG', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 3.013554455347072}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'AGG', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 3.045400268179898}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 3.0578934984056514}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 3.0578934984056514}]}, {'initial_sequence': 'ATGGTATCGAAGGGGGAAGAGGATAATATGGCCATAATTAAGGAATTCATGCGATTCAAAGTCCATATGGAGGGCTCTGTGAACGGGCACGAATTCGAAATCGAAGGCGAGGGAGAAGGACGCCCGTATGAAGGCACGCAGACGGCGAAACTGAAGGTAACAAAGGGAGGGCCCTTGCCGTTTGCGTGGGACATTCTCTCCCCCCAGTTTATGTATGGATCCAAAGCTTATGTAAAACATCCTGCGGACATACCGGATTATCTCAAGTTATCATTCCCAGAGGGGTTCAAGTGGGAACGTGTAATGAATTTCGAGGACGGTGGCGTGGTAACCGTTACGCAAGACAGCTCCCTACAAGATGGTGAGTTCATTTACAAAGTAAAATTGCGGGGCACCAATTTCCCATCTGACGGTCCCGTAATGCAGAAAAAAACAATGGGCTGGGAAGCATCCTCAGAGCGTATGTATCCGGAGGATGGCGCATTAAAAGGAGAAATTAAGCAGAGGCTCAAATTGAAAGACGGGGGTCACTATGACGCCGAAGTAAAAACTACGTATAAAGCTAAAAAGCCGGTGCAGCTACCCGGGGCCTATAATGTCAACATAAAACTGGATATCACAAGCCACAACGAGGATTATACCATAGTTGAGCAATACGAACGGGCAGAAGGCAGGCACAGTACGGGAGGTATGGATGAGCTTTATAAGTAA', 'final_sequence': 'ATGGTGTCGAAGGGTGAGGAGGACAACATGGCCATAATAAAGGAGTTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATAGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGTCCCCTGCCCTTCGCCTGGGACATACTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATACCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGTGTGGTGACCGTGACCCAGGACTCGTCGCTGCAGGACGGTGAGTTCATATACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAGACCATGGGTTGGGAGGCCTCGTCGGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATAAAGCAGCGTCTGAAGCTGAAGGACGGTGGTCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATAAAGCTGGACATAACCTCGCACAACGAGGACTACACCATAGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGTATGGACGAGCTGTACAAGTGA', 'iterations_count': 5, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'initial_sequence_optimization_score': 0.5268456727356423, 'final_sequence_optimization_score': 3.0578934984056514, 'run_time': 31.06200000003446, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTA', 'H': 'CAC', 'R': 'AGG', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 2.9891767389230144}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'AGG', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 3.013554455347072}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'AGG', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 3.045400268179898}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 3.0578934984056514}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 3.0578934984056514}]}, {'initial_sequence': 'ATGGTGAGTAAGGGGGAAGAAGACAACATGGCCATAATCAAGGAGTTTATGAGATTCAAAGTACATATGGAAGGCAGCGTTAATGGCCATGAGTTCGAGATAGAAGGGGAAGGTGAAGGGCGGCCATATGAAGGCACACAGACCGCCAAACTGAAGGTGACGAAAGGAGGCCCTCTACCGTTCGCTTGGGACATCCTCTCGCCACAGTTCATGTACGGAAGCAAGGCTTACGTGAAACACCCTGCTGACATCCCTGACTACCTAAAGCTCTCCTTTCCGGAAGGATTCAAGTGGGAGCGGGTCATGAACTTTGAGGACGGAGGAGTTGTTACTGTTACGCAAGACTCTAGCTTGCAGGACGGAGAGTTCATATATAAAGTGAAATTACGGGGAACTAATTTTCCTAGCGACGGTCCAGTTATGCAGAAGAAAACGATGGGATGGGAGGCCTCCAGTGAACGTATGTATCCTGAGGACGGAGCACTCAAGGGAGAGATCAAGCAGCGTTTGAAACTCAAAGATGGTGGTCACTATGATGCGGAGGTGAAGACTACGTACAAGGCGAAAAAGCCTGTGCAACTTCCTGGCGCCTATAACGTGAACATCAAACTGGACATTACGTCCCATAATGAAGACTACACGATTGTAGAGCAGTACGAACGTGCAGAAGGCAGGCACTCCACTGGGGGCATGGACGAACTCTACAAGTAA', 'final_sequence': 'ATGGTGTCGAAGGGTGAGGAGGACAACATGGCCATAATAAAGGAGTTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATAGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGTCCCCTGCCCTTCGCCTGGGACATACTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATACCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGTGTGGTGACCGTGACCCAGGACTCGTCGCTGCAGGACGGTGAGTTCATATACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAGACCATGGGTTGGGAGGCCTCGTCGGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATAAAGCAGCGTCTGAAGCTGAAGGACGGTGGTCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATAAAGCTGGACATAACCTCGCACAACGAGGACTACACCATAGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGTATGGACGAGCTGTACAAGTGA', 'iterations_count': 5, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'initial_sequence_optimization_score': 0.3515399015442491, 'final_sequence_optimization_score': 3.0578934984056514, 'run_time': 31.015999999945052, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTA', 'H': 'CAC', 'R': 'AGG', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 2.9891767389230144}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'AGG', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 3.013554455347072}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'AGG', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 3.045400268179898}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 3.0578934984056514}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCC', 'K': 'AAG', '_': 'TGA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGT', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTG', 'E': 'GAG', 'Y': 'TAC'}, 'sequence_score': 3.0578934984056514}]}], 'evaluation': [{'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'cai_initial_score': 0.7891479965436429, 'cai_final_score': 0.5411573485546257, 'dist_score': -2.6976527706377054}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'cai_initial_score': 0.7608140098814632, 'cai_final_score': 0.4417435150279882, 'dist_score': -7.647806010984315}], 'final_sequence': 'ATGGTGTCGAAGGGTGAGGAGGACAACATGGCCATAATAAAGGAGTTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATAGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGTCCCCTGCCCTTCGCCTGGGACATACTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATACCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGTGTGGTGACCGTGACCCAGGACTCGTCGCTGCAGGACGGTGAGTTCATATACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAGACCATGGGTTGGGAGGCCTCGTCGGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATAAAGCAGCGTCTGAAGCTGAAGGACGGTGGTCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATAAAGCTGGACATAACCTCGCACAACGAGGACTACACCATAGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGTATGGACGAGCTGTACAAGTGA', 'average_distance_score': 2.4750766201733048, 'weakest_link_score': 2.4750766201733048}, {'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'cai_initial_score': 0.7891479965436429, 'cai_final_score': 0.5411573485546257, 'dist_score': -2.6976527706377054}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'cai_initial_score': 0.7608140098814632, 'cai_final_score': 0.4417435150279882, 'dist_score': -7.647806010984315}], 'final_sequence': 'ATGGTGTCGAAGGGTGAGGAGGACAACATGGCCATAATAAAGGAGTTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATAGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGTCCCCTGCCCTTCGCCTGGGACATACTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATACCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGTGTGGTGACCGTGACCCAGGACTCGTCGCTGCAGGACGGTGAGTTCATATACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAGACCATGGGTTGGGAGGCCTCGTCGGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATAAAGCAGCGTCTGAAGCTGAAGGACGGTGGTCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATAAAGCTGGACATAACCTCGCACAACGAGGACTACACCATAGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGTATGGACGAGCTGTACAAGTGA', 'average_distance_score': 2.4750766201733048, 'weakest_link_score': 2.4750766201733048}, {'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'cai_initial_score': 0.7891479965436429, 'cai_final_score': 0.5411573485546257, 'dist_score': -2.6976527706377054}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'cai_initial_score': 0.7608140098814632, 'cai_final_score': 0.4417435150279882, 'dist_score': -7.647806010984315}], 'final_sequence': 'ATGGTGTCGAAGGGTGAGGAGGACAACATGGCCATAATAAAGGAGTTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATAGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGTCCCCTGCCCTTCGCCTGGGACATACTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATACCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGTGTGGTGACCGTGACCCAGGACTCGTCGCTGCAGGACGGTGAGTTCATATACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAGACCATGGGTTGGGAGGCCTCGTCGGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATAAAGCAGCGTCTGAAGCTGAAGGACGGTGGTCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATAAAGCTGGACATAACCTCGCACAACGAGGACTACACCATAGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGTATGGACGAGCTGTACAAGTGA', 'average_distance_score': 2.4750766201733048, 'weakest_link_score': 2.4750766201733048}, {'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'cai_initial_score': 0.7891479965436429, 'cai_final_score': 0.5411573485546257, 'dist_score': -2.6976527706377054}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'cai_initial_score': 0.7608140098814632, 'cai_final_score': 0.4417435150279882, 'dist_score': -7.647806010984315}], 'final_sequence': 'ATGGTGTCGAAGGGTGAGGAGGACAACATGGCCATAATAAAGGAGTTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATAGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGTCCCCTGCCCTTCGCCTGGGACATACTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATACCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGTGTGGTGACCGTGACCCAGGACTCGTCGCTGCAGGACGGTGAGTTCATATACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAGACCATGGGTTGGGAGGCCTCGTCGGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATAAAGCAGCGTCTGAAGCTGAAGGACGGTGGTCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATAAAGCTGGACATAACCTCGCACAACGAGGACTACACCATAGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGTATGGACGAGCTGTACAAGTGA', 'average_distance_score': 2.4750766201733048, 'weakest_link_score': 2.4750766201733048}], 'final_evaluation': {'final_sequence': 'ATGGTGTCGAAGGGTGAGGAGGACAACATGGCCATAATAAAGGAGTTCATGCGTTTCAAGGTGCACATGGAGGGTTCGGTGAACGGTCACGAGTTCGAGATAGAGGGTGAGGGTGAGGGTCGTCCCTACGAGGGTACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGTCCCCTGCCCTTCGCCTGGGACATACTGTCGCCCCAGTTCATGTACGGTTCGAAGGCCTACGTGAAGCACCCCGCCGACATACCCGACTACCTGAAGCTGTCGTTCCCCGAGGGTTTCAAGTGGGAGCGTGTGATGAACTTCGAGGACGGTGGTGTGGTGACCGTGACCCAGGACTCGTCGCTGCAGGACGGTGAGTTCATATACAAGGTGAAGCTGCGTGGTACCAACTTCCCCTCGGACGGTCCCGTGATGCAGAAGAAGACCATGGGTTGGGAGGCCTCGTCGGAGCGTATGTACCCCGAGGACGGTGCCCTGAAGGGTGAGATAAAGCAGCGTCTGAAGCTGAAGGACGGTGGTCACTACGACGCCGAGGTGAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTGCCCGGTGCCTACAACGTGAACATAAAGCTGGACATAACCTCGCACAACGAGGACTACACCATAGTGGAGCAGTACGAGCGTGCCGAGGGTCGTCACTCGACCGGTGGTATGGACGAGCTGTACAAGTGA', 'average_distance_score': 2.4750766201733048, 'weakest_link_score': 2.4750766201733048}, 'zip_output_file_path': 'results\\\\mcherry_debug\\\\CAI_zscore_bulk_aa_diff_ecoli_opt_True_sSCa\\\\communique_results.zip'}\n",
      "##########################\n",
      "# USER INPUT INFORMATION #\n",
      "##########################\n",
      "Organism is optimized\n",
      "Information about Bacillus subtilis:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 98 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3550\n",
      "Calculate CAI weights from a reference set of 532 highly expressed genes from estimated expression dictionary.\n",
      "name=Bacillus subtilis, cai_std=0.04172052669683352, cai_avg=0.7354163127574911\n",
      "Organism is de-optimized\n",
      "Information about Escherichia coli:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "Found 90 ribosomal proteins in input genome.\n",
      "CAI will be calculated from a reference set of estimated expression dictionary.\n",
      "Expression levels were found for 3895\n",
      "Calculate CAI weights from a reference set of 584 highly expressed genes from estimated expression dictionary.\n",
      "name=Escherichia coli, cai_std=0.09192830548402788, cai_avg=0.6260311851192147\n",
      "Bacillus subtilis has weight of 1.0\n",
      "Escherichia coli has weight of 1.0\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is: ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Total input processing time: 20.947362899780273\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "\n",
      "Optimization method used is: OptimizationMethod.zscore_bulk_aa_diff\n",
      "Running ORF optimization for initial sequence: ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Running ORF optimization for initial sequence: ATGGTATCCAAAGGCGAAGAGGACAATATGGCCATTATTAAAGAGTTCATGAGATTCAAGGTTCATATGGAGGGTTCGGTTAACGGGCATGAGTTCGAAATAGAGGGCGAAGGAGAAGGCCGTCCCTATGAAGGCACGCAAACCGCTAAGCTTAAGGTCACTAAGGGGGGACCCTTGCCATTTGCGTGGGACATCTTATCACCCCAATTCATGTACGGTTCTAAAGCCTACGTCAAGCATCCCGCAGACATCCCGGATTATTTGAAGCTGTCGTTCCCAGAAGGATTTAAGTGGGAGCGTGTGATGAATTTCGAAGATGGAGGAGTGGTTACGGTGACACAGGACTCCTCCTTACAAGACGGAGAGTTCATATATAAAGTTAAGCTTCGAGGTACGAACTTTCCAAGTGATGGACCTGTAATGCAAAAAAAAACGATGGGGTGGGAAGCTAGTTCTGAACGAATGTACCCGGAAGATGGCGCACTTAAGGGTGAGATCAAACAGCGCCTGAAGCTCAAAGATGGTGGTCATTATGATGCGGAAGTTAAAACAACCTATAAAGCCAAAAAGCCAGTTCAACTCCCGGGTGCTTACAATGTAAACATAAAACTAGATATTACTTCGCATAATGAGGACTACACCATTGTCGAACAATACGAGCGAGCTGAGGGCCGTCATTCAACCGGTGGTATGGACGAACTATATAAGTGA\n",
      "Running ORF optimization for initial sequence: ATGGTAAGCAAGGGCGAGGAAGACAACATGGCGATCATAAAAGAATTTATGCGCTTCAAGGTACACATGGAGGGTTCGGTCAATGGACATGAGTTTGAAATAGAAGGTGAAGGAGAAGGGAGACCTTATGAGGGGACTCAGACAGCCAAATTGAAAGTGACGAAGGGTGGTCCGCTACCTTTTGCGTGGGATATACTGAGCCCGCAGTTCATGTATGGTAGCAAAGCCTATGTGAAGCATCCAGCAGACATTCCGGACTATTTAAAACTATCATTCCCTGAAGGTTTCAAATGGGAACGTGTTATGAACTTCGAGGATGGGGGAGTTGTAACCGTGACTCAAGATAGCTCCTTACAGGACGGCGAGTTTATCTATAAGGTAAAGCTCAGGGGTACGAATTTTCCGTCGGATGGCCCAGTAATGCAAAAGAAGACGATGGGTTGGGAAGCGTCATCAGAGAGAATGTACCCTGAAGATGGTGCATTAAAGGGAGAGATCAAGCAACGTTTAAAATTAAAAGACGGGGGACATTACGATGCTGAAGTGAAGACAACTTATAAGGCTAAAAAGCCGGTGCAATTACCAGGCGCGTATAATGTAAACATAAAGCTGGACATTACTTCGCACAATGAAGATTATACTATAGTTGAACAGTACGAAAGGGCCGAAGGGCGCCATTCTACGGGTGGGATGGATGAGTTGTACAAATGA\n",
      "Running ORF optimization for initial sequence: ATGGTCAGCAAGGGAGAGGAGGATAACATGGCGATTATAAAAGAATTTATGAGGTTTAAGGTTCATATGGAGGGCTCCGTGAACGGACATGAATTTGAAATAGAGGGAGAGGGCGAAGGCCGACCGTACGAAGGGACACAAACCGCAAAACTCAAAGTTACCAAGGGTGGTCCACTCCCATTCGCTTGGGACATACTAAGCCCTCAGTTCATGTACGGAAGCAAAGCTTACGTGAAGCACCCGGCCGATATACCCGACTACCTTAAGTTGTCCTTTCCCGAAGGATTTAAGTGGGAACGGGTTATGAATTTTGAGGATGGCGGAGTGGTCACGGTCACTCAGGATTCCTCGCTCCAGGATGGCGAATTCATCTATAAGGTTAAGCTTCGAGGTACTAATTTCCCCTCGGATGGACCTGTGATGCAGAAGAAGACGATGGGCTGGGAAGCATCATCCGAGCGCATGTATCCTGAAGACGGCGCACTAAAGGGCGAAATCAAGCAAAGGCTTAAGTTGAAGGACGGCGGCCATTACGATGCTGAAGTAAAAACGACGTACAAAGCAAAGAAACCAGTTCAGTTACCTGGAGCTTATAATGTGAATATAAAACTGGACATAACCTCTCACAACGAGGATTATACGATTGTCGAACAGTATGAACGTGCAGAGGGTAGACACTCCACAGGCGGGATGGACGAATTATATAAATGA\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGAGAAGAAGATAACATGGCAATTATTAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGAGTTGTTACAGTTACACAAGATTCATCACTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCATCATCAGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGACATTATGATGCAGAAGTTAAAACAACATATAAAGCAAAAAAACCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGAATGGATGAACTTTATAAATAA', 'average_distance_score': 4.298394830741209, 'weakest_link_score': 4.298394830741209}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGAGAAGAAGATAACATGGCAATTATTAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGAGTTGTTACAGTTACACAAGATTCATCACTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCATCATCAGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGACATTATGATGCAGAAGTTAAAACAACATATAAAGCAAAAAAACCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGAATGGATGAACTTTATAAATAA', 'average_distance_score': 4.298394830741209, 'weakest_link_score': 4.298394830741209}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGAGAAGAAGATAACATGGCAATTATTAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGAGTTGTTACAGTTACACAAGATTCATCACTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCATCATCAGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGACATTATGATGCAGAAGTTAAAACAACATATAAAGCAAAAAAACCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGAATGGATGAACTTTATAAATAA', 'average_distance_score': 4.298394830741209, 'weakest_link_score': 4.298394830741209}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGAGAAGAAGATAACATGGCAATTATTAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGAGTTGTTACAGTTACACAAGATTCATCACTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCATCATCAGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGACATTATGATGCAGAAGTTAAAACAACATATAAAGCAAAAAAACCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGAATGGATGAACTTTATAAATAA', 'average_distance_score': 4.298394830741209, 'weakest_link_score': 4.298394830741209}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTTTCAAAAGGAGAAGAAGATAACATGGCAATTATTAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGAGTTGTTACAGTTACACAAGATTCATCACTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCATCATCAGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGACATTATGATGCAGAAGTTAAAACAACATATAAAGCAAAAAAACCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGAATGGATGAACTTTATAAATAA', 'average_distance_score': 4.298394830741209, 'weakest_link_score': 4.298394830741209}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry_debug\\CAI_zscore_bulk_aa_diff_ecoli_opt_False_ANxb\n",
      "Final output: {'user_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'optimization_method': 'zscore_bulk_aa_diff', 'optimization_cub_index': 'CAI', 'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.10739856801909307, 'TGA': 0.1694510739856802, 'GCA': 1.0, 'GCC': 0.5246426811237062, 'GCG': 0.9196648595367175, 'GCT': 0.9361754558896007, 'TGC': 1.0, 'TGT': 0.7495652173913042, 'GAC': 0.6699738903394257, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.3869690068323596, 'TTC': 0.82402496099844, 'TTT': 1.0, 'GGA': 0.8480035296712993, 'GGC': 1.0, 'GGG': 0.29340392675932053, 'GGT': 0.6170306640194132, 'CAC': 0.6778131634819533, 'CAT': 1.0, 'ATA': 0.11114994757077944, 'ATC': 0.8386927647675637, 'ATT': 1.0, 'AAA': 1.0, 'AAG': 0.3398391420911528, 'CTA': 0.14688262702462837, 'CTC': 0.3046372309740404, 'CTG': 0.6640781007321944, 'CTT': 1.0, 'TTA': 0.5893055247392945, 'TTG': 0.41202573774129136, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.8348958333333333, 'CCA': 0.4923917284432306, 'CCC': 0.07452204447912603, 'CCG': 1.0, 'CCT': 0.7553648068669527, 'CAA': 1.0, 'CAG': 0.8251624335499115, 'AGA': 0.72005772005772, 'AGG': 0.09620009620009619, 'CGA': 0.19672919672919673, 'CGC': 0.984126984126984, 'CGG': 0.2751322751322751, 'CGT': 1.0, 'AGC': 0.8834134615384617, 'AGT': 0.33774038461538464, 'TCA': 0.9487179487179488, 'TCC': 0.42588141025641024, 'TCG': 0.2560096153846154, 'TCT': 1.0, 'ACA': 1.0, 'ACC': 0.22142692218887092, 'ACG': 0.5329023320249365, 'ACT': 0.4301546986839067, 'GTA': 0.7337249143416544, 'GTC': 0.681106216348507, 'GTG': 0.6669114047968673, 'GTT': 1.0, 'TGG': 1.0, 'TAC': 0.7787396121883656, 'TAT': 1.0}, 'cai_avg': 0.7354163127574911, 'cai_std': 0.04172052669683352, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}, {'name': 'Escherichia coli', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': {'TAA': 1.0, 'TAG': 0.037037037037037035, 'TGA': 0.24183006535947713, 'GCA': 0.6554010349288486, 'GCC': 0.6652652005174644, 'GCG': 1.0, 'GCT': 0.6266170763260025, 'TGC': 1.0, 'TGT': 0.7162162162162162, 'GAC': 0.8422464375523889, 'GAT': 1.0, 'GAA': 1.0, 'GAG': 0.36087048383688997, 'TTC': 1.0, 'TTT': 0.6891233766233766, 'GGA': 0.12073112073112073, 'GGC': 1.0, 'GGG': 0.19111752445085778, 'GGT': 0.9866923200256533, 'CAC': 1.0, 'CAT': 0.7259593679458239, 'ATA': 0.03593829462620784, 'ATC': 1.0, 'ATT': 0.7563993897270725, 'AAA': 1.0, 'AAG': 0.2715144230769231, 'CTA': 0.0255237398225231, 'CTC': 0.1333821242338304, 'CTG': 1.0, 'CTT': 0.11572591711645777, 'TTA': 0.10557131095050773, 'TTG': 0.12322751806788033, 'ATG': 1.0, 'AAC': 1.0, 'AAT': 0.4079954093343534, 'CCA': 0.2681349767629825, 'CCC': 0.08830066680137402, 'CCG': 1.0, 'CCT': 0.19337239846433624, 'CAA': 0.3511947863866763, 'CAG': 1.0, 'AGA': 0.018922018348623854, 'AGG': 0.011085626911314986, 'CGA': 0.03631498470948013, 'CGC': 0.7138761467889909, 'CGG': 0.05600152905198777, 'CGT': 1.0, 'AGC': 1.0, 'AGT': 0.3131715081575806, 'TCA': 0.3048149621965778, 'TCC': 0.7942697970553123, 'TCG': 0.40907282132908873, 'TCT': 0.8901711102268206, 'ACA': 0.14701775384001597, 'ACC': 1.0, 'ACG': 0.3457011769399561, 'ACT': 0.4374625972471574, 'GTA': 0.531166518254675, 'GTC': 0.5075690115761353, 'GTG': 0.9971059661620658, 'GTT': 1.0, 'TGG': 1.0, 'TAC': 1.0, 'TAT': 0.857994781960492}, 'cai_avg': 0.6260311851192147, 'cai_std': 0.09192830548402788, 'tai_weights': None, 'tai_avg': None, 'tai_std': None}]}, 'orf': [{'initial_sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_sequence': 'ATGGTTTCAAAAGGAGAAGAAGATAACATGGCAATTATTAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGAGTTGTTACAGTTACACAAGATTCATCACTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCATCATCAGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGACATTATGATGCAGAAGTTAAAACAACATATAAAGCAAAAAAACCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGAATGGATGAACTTTATAAATAA', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'initial_sequence_optimization_score': -0.5828168782323468, 'final_sequence_optimization_score': 3.7155779525088617, 'run_time': 18.375, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 3.7148552550133576}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 3.7155779525088617}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 3.7155779525088617}]}, {'initial_sequence': 'ATGGTATCCAAAGGCGAAGAGGACAATATGGCCATTATTAAAGAGTTCATGAGATTCAAGGTTCATATGGAGGGTTCGGTTAACGGGCATGAGTTCGAAATAGAGGGCGAAGGAGAAGGCCGTCCCTATGAAGGCACGCAAACCGCTAAGCTTAAGGTCACTAAGGGGGGACCCTTGCCATTTGCGTGGGACATCTTATCACCCCAATTCATGTACGGTTCTAAAGCCTACGTCAAGCATCCCGCAGACATCCCGGATTATTTGAAGCTGTCGTTCCCAGAAGGATTTAAGTGGGAGCGTGTGATGAATTTCGAAGATGGAGGAGTGGTTACGGTGACACAGGACTCCTCCTTACAAGACGGAGAGTTCATATATAAAGTTAAGCTTCGAGGTACGAACTTTCCAAGTGATGGACCTGTAATGCAAAAAAAAACGATGGGGTGGGAAGCTAGTTCTGAACGAATGTACCCGGAAGATGGCGCACTTAAGGGTGAGATCAAACAGCGCCTGAAGCTCAAAGATGGTGGTCATTATGATGCGGAAGTTAAAACAACCTATAAAGCCAAAAAGCCAGTTCAACTCCCGGGTGCTTACAATGTAAACATAAAACTAGATATTACTTCGCATAATGAGGACTACACCATTGTCGAACAATACGAGCGAGCTGAGGGCCGTCATTCAACCGGTGGTATGGACGAACTATATAAGTGA', 'final_sequence': 'ATGGTTTCAAAAGGAGAAGAAGATAACATGGCAATTATTAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGAGTTGTTACAGTTACACAAGATTCATCACTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCATCATCAGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGACATTATGATGCAGAAGTTAAAACAACATATAAAGCAAAAAAACCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGAATGGATGAACTTTATAAATAA', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'initial_sequence_optimization_score': -0.487092339477761, 'final_sequence_optimization_score': 3.7155779525088617, 'run_time': 18.891000000061467, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 3.7148552550133576}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 3.7155779525088617}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 3.7155779525088617}]}, {'initial_sequence': 'ATGGTAAGCAAGGGCGAGGAAGACAACATGGCGATCATAAAAGAATTTATGCGCTTCAAGGTACACATGGAGGGTTCGGTCAATGGACATGAGTTTGAAATAGAAGGTGAAGGAGAAGGGAGACCTTATGAGGGGACTCAGACAGCCAAATTGAAAGTGACGAAGGGTGGTCCGCTACCTTTTGCGTGGGATATACTGAGCCCGCAGTTCATGTATGGTAGCAAAGCCTATGTGAAGCATCCAGCAGACATTCCGGACTATTTAAAACTATCATTCCCTGAAGGTTTCAAATGGGAACGTGTTATGAACTTCGAGGATGGGGGAGTTGTAACCGTGACTCAAGATAGCTCCTTACAGGACGGCGAGTTTATCTATAAGGTAAAGCTCAGGGGTACGAATTTTCCGTCGGATGGCCCAGTAATGCAAAAGAAGACGATGGGTTGGGAAGCGTCATCAGAGAGAATGTACCCTGAAGATGGTGCATTAAAGGGAGAGATCAAGCAACGTTTAAAATTAAAAGACGGGGGACATTACGATGCTGAAGTGAAGACAACTTATAAGGCTAAAAAGCCGGTGCAATTACCAGGCGCGTATAATGTAAACATAAAGCTGGACATTACTTCGCACAATGAAGATTATACTATAGTTGAACAGTACGAAAGGGCCGAAGGGCGCCATTCTACGGGTGGGATGGATGAGTTGTACAAATGA', 'final_sequence': 'ATGGTTTCAAAAGGAGAAGAAGATAACATGGCAATTATTAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGAGTTGTTACAGTTACACAAGATTCATCACTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCATCATCAGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGACATTATGATGCAGAAGTTAAAACAACATATAAAGCAAAAAAACCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGAATGGATGAACTTTATAAATAA', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'initial_sequence_optimization_score': -0.2663248547247289, 'final_sequence_optimization_score': 3.7155779525088617, 'run_time': 19.70299999997951, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 3.7148552550133576}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 3.7155779525088617}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 3.7155779525088617}]}, {'initial_sequence': 'ATGGTCAGCAAGGGAGAGGAGGATAACATGGCGATTATAAAAGAATTTATGAGGTTTAAGGTTCATATGGAGGGCTCCGTGAACGGACATGAATTTGAAATAGAGGGAGAGGGCGAAGGCCGACCGTACGAAGGGACACAAACCGCAAAACTCAAAGTTACCAAGGGTGGTCCACTCCCATTCGCTTGGGACATACTAAGCCCTCAGTTCATGTACGGAAGCAAAGCTTACGTGAAGCACCCGGCCGATATACCCGACTACCTTAAGTTGTCCTTTCCCGAAGGATTTAAGTGGGAACGGGTTATGAATTTTGAGGATGGCGGAGTGGTCACGGTCACTCAGGATTCCTCGCTCCAGGATGGCGAATTCATCTATAAGGTTAAGCTTCGAGGTACTAATTTCCCCTCGGATGGACCTGTGATGCAGAAGAAGACGATGGGCTGGGAAGCATCATCCGAGCGCATGTATCCTGAAGACGGCGCACTAAAGGGCGAAATCAAGCAAAGGCTTAAGTTGAAGGACGGCGGCCATTACGATGCTGAAGTAAAAACGACGTACAAAGCAAAGAAACCAGTTCAGTTACCTGGAGCTTATAATGTGAATATAAAACTGGACATAACCTCTCACAACGAGGATTATACGATTGTCGAACAGTATGAACGTGCAGAGGGTAGACACTCCACAGGCGGGATGGACGAATTATATAAATGA', 'final_sequence': 'ATGGTTTCAAAAGGAGAAGAAGATAACATGGCAATTATTAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGAGTTGTTACAGTTACACAAGATTCATCACTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCATCATCAGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGACATTATGATGCAGAAGTTAAAACAACATATAAAGCAAAAAAACCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGAATGGATGAACTTTATAAATAA', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'initial_sequence_optimization_score': -0.4585566670312776, 'final_sequence_optimization_score': 3.7155779525088617, 'run_time': 18.5, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAT', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 3.7148552550133576}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 3.7155779525088617}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCA', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTT', 'A': 'GCA', 'G': 'GGA', 'I': 'ATT', 'L': 'CTT', 'H': 'CAT', 'R': 'AGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAA', 'Y': 'TAT'}, 'sequence_score': 3.7155779525088617}]}], 'evaluation': [{'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'cai_initial_score': 0.7608140098814632, 'cai_final_score': 0.9531528007844668, 'dist_score': 4.610171686006104}, {'name': 'Escherichia coli', 'is_wanted': False, 'cai_initial_score': 0.7891479965436429, 'cai_final_score': 0.4226649614459395, 'dist_score': -3.9866179754763142}], 'final_sequence': 'ATGGTTTCAAAAGGAGAAGAAGATAACATGGCAATTATTAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGAGTTGTTACAGTTACACAAGATTCATCACTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCATCATCAGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGACATTATGATGCAGAAGTTAAAACAACATATAAAGCAAAAAAACCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGAATGGATGAACTTTATAAATAA', 'average_distance_score': 4.298394830741209, 'weakest_link_score': 4.298394830741209}, {'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'cai_initial_score': 0.7608140098814632, 'cai_final_score': 0.9531528007844668, 'dist_score': 4.610171686006104}, {'name': 'Escherichia coli', 'is_wanted': False, 'cai_initial_score': 0.7891479965436429, 'cai_final_score': 0.4226649614459395, 'dist_score': -3.9866179754763142}], 'final_sequence': 'ATGGTTTCAAAAGGAGAAGAAGATAACATGGCAATTATTAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGAGTTGTTACAGTTACACAAGATTCATCACTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCATCATCAGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGACATTATGATGCAGAAGTTAAAACAACATATAAAGCAAAAAAACCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGAATGGATGAACTTTATAAATAA', 'average_distance_score': 4.298394830741209, 'weakest_link_score': 4.298394830741209}, {'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'cai_initial_score': 0.7608140098814632, 'cai_final_score': 0.9531528007844668, 'dist_score': 4.610171686006104}, {'name': 'Escherichia coli', 'is_wanted': False, 'cai_initial_score': 0.7891479965436429, 'cai_final_score': 0.4226649614459395, 'dist_score': -3.9866179754763142}], 'final_sequence': 'ATGGTTTCAAAAGGAGAAGAAGATAACATGGCAATTATTAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGAGTTGTTACAGTTACACAAGATTCATCACTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCATCATCAGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGACATTATGATGCAGAAGTTAAAACAACATATAAAGCAAAAAAACCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGAATGGATGAACTTTATAAATAA', 'average_distance_score': 4.298394830741209, 'weakest_link_score': 4.298394830741209}, {'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'cai_initial_score': 0.7608140098814632, 'cai_final_score': 0.9531528007844668, 'dist_score': 4.610171686006104}, {'name': 'Escherichia coli', 'is_wanted': False, 'cai_initial_score': 0.7891479965436429, 'cai_final_score': 0.4226649614459395, 'dist_score': -3.9866179754763142}], 'final_sequence': 'ATGGTTTCAAAAGGAGAAGAAGATAACATGGCAATTATTAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGAGTTGTTACAGTTACACAAGATTCATCACTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCATCATCAGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGACATTATGATGCAGAAGTTAAAACAACATATAAAGCAAAAAAACCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGAATGGATGAACTTTATAAATAA', 'average_distance_score': 4.298394830741209, 'weakest_link_score': 4.298394830741209}], 'final_evaluation': {'final_sequence': 'ATGGTTTCAAAAGGAGAAGAAGATAACATGGCAATTATTAAAGAATTTATGAGATTTAAAGTTCATATGGAAGGATCAGTTAACGGACATGAATTTGAAATTGAAGGAGAAGGAGAAGGAAGACCTTATGAAGGAACACAAACAGCAAAACTTAAAGTTACAAAAGGAGGACCTCTTCCTTTTGCATGGGATATTCTTTCACCTCAATTTATGTATGGATCAAAAGCATATGTTAAACATCCTGCAGATATTCCTGATTATCTTAAACTTTCATTTCCTGAAGGATTTAAATGGGAAAGAGTTATGAACTTTGAAGATGGAGGAGTTGTTACAGTTACACAAGATTCATCACTTCAAGATGGAGAATTTATTTATAAAGTTAAACTTAGAGGAACAAACTTTCCTTCAGATGGACCTGTTATGCAAAAAAAAACAATGGGATGGGAAGCATCATCAGAAAGAATGTATCCTGAAGATGGAGCACTTAAAGGAGAAATTAAACAAAGACTTAAACTTAAAGATGGAGGACATTATGATGCAGAAGTTAAAACAACATATAAAGCAAAAAAACCTGTTCAACTTCCTGGAGCATATAACGTTAACATTAAACTTGATATTACATCACATAACGAAGATTATACAATTGTTGAACAATATGAAAGAGCAGAAGGAAGACATTCAACAGGAGGAATGGATGAACTTTATAAATAA', 'average_distance_score': 4.298394830741209, 'weakest_link_score': 4.298394830741209}, 'zip_output_file_path': 'results\\\\mcherry_debug\\\\CAI_zscore_bulk_aa_diff_ecoli_opt_False_ANxb\\\\communique_results.zip'}\n",
      "##########################\n",
      "# USER INPUT INFORMATION #\n",
      "##########################\n",
      "Organism is optimized\n",
      "Information about Escherichia coli:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Escherichia coli. Calculating tAI profile.\n",
      "name=Escherichia coli, tai_std=0.023481599705588187, tai_avg=0.2504950429983414\n",
      "Organism is de-optimized\n",
      "Information about Bacillus subtilis:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Bacillus subtilis. Calculating tAI profile.\n",
      "name=Bacillus subtilis, tai_std=0.027038419737753252, tai_avg=0.3292474875006557\n",
      "Escherichia coli has weight of 1.0\n",
      "Bacillus subtilis has weight of 1.0\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is: ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Total input processing time: 20.276329278945923\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "\n",
      "Optimization method used is: OptimizationMethod.zscore_bulk_aa_diff\n",
      "Running ORF optimization for initial sequence: ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Running ORF optimization for initial sequence: ATGGTCTCCAAGGGCGAAGAAGATAACATGGCGATCATCAAAGAATTTATGCGGTTTAAGGTCCATATGGAAGGTTCTGTGAACGGGCATGAATTCGAAATAGAAGGGGAGGGCGAAGGACGGCCCTATGAAGGGACCCAAACCGCGAAACTGAAGGTGACTAAGGGGGGCCCACTTCCATTCGCATGGGACATTTTGTCACCCCAGTTCATGTATGGAAGTAAAGCTTACGTAAAGCACCCGGCTGATATACCAGACTACCTTAAACTAAGCTTTCCCGAGGGTTTTAAGTGGGAGCGTGTAATGAATTTCGAAGATGGAGGGGTAGTAACGGTCACTCAGGACTCTTCGCTGCAGGACGGAGAATTCATATATAAAGTTAAATTACGCGGTACCAACTTCCCCTCTGATGGTCCCGTTATGCAAAAGAAGACCATGGGATGGGAAGCGTCTAGCGAGCGCATGTACCCCGAGGATGGAGCACTAAAGGGCGAGATAAAGCAAAGGCTAAAGCTAAAAGACGGCGGGCATTACGATGCAGAGGTGAAAACCACCTACAAAGCGAAAAAGCCGGTGCAACTGCCGGGTGCGTACAACGTGAATATCAAGTTGGATATAACCAGTCATAATGAAGACTATACGATCGTGGAGCAGTACGAGCGGGCTGAAGGTAGGCATTCTACAGGCGGGATGGACGAACTTTATAAGTAA\n",
      "Running ORF optimization for initial sequence: ATGGTTTCCAAAGGCGAGGAAGATAATATGGCTATAATAAAAGAATTTATGCGTTTCAAGGTCCACATGGAAGGATCTGTTAACGGACATGAATTTGAGATAGAAGGCGAAGGGGAGGGACGGCCCTATGAAGGCACCCAAACAGCAAAATTAAAAGTTACGAAAGGAGGTCCCCTGCCATTTGCGTGGGATATACTCTCTCCACAATTTATGTACGGTTCTAAGGCTTATGTTAAACATCCTGCGGATATCCCCGACTACCTGAAGTTATCGTTCCCGGAAGGATTCAAGTGGGAGAGGGTTATGAACTTCGAGGATGGTGGGGTGGTCACGGTGACTCAAGACTCGTCCTTGCAAGACGGAGAATTCATCTACAAAGTAAAATTGCGCGGGACAAATTTCCCGTCCGATGGACCCGTCATGCAAAAAAAGACTATGGGTTGGGAGGCGAGCAGTGAACGAATGTACCCTGAGGATGGAGCACTCAAGGGCGAAATAAAGCAACGGTTAAAGCTCAAAGATGGCGGTCACTACGATGCTGAGGTGAAGACGACCTACAAGGCGAAAAAACCAGTGCAGTTGCCGGGGGCATATAATGTAAACATTAAGCTTGACATCACTTCGCACAACGAAGATTATACGATAGTTGAACAATATGAGCGAGCCGAGGGGCGGCACAGCACGGGAGGCATGGATGAGCTATACAAATAA\n",
      "Running ORF optimization for initial sequence: ATGGTATCAAAGGGCGAAGAAGACAATATGGCGATCATCAAGGAATTTATGCGGTTCAAGGTCCACATGGAAGGCTCTGTTAATGGGCACGAGTTTGAAATCGAGGGGGAGGGAGAAGGAAGACCGTATGAAGGGACCCAAACTGCTAAACTCAAAGTCACAAAGGGTGGACCACTGCCATTCGCGTGGGACATCCTATCTCCACAGTTCATGTACGGTTCGAAAGCGTATGTAAAACATCCAGCGGATATACCCGACTACTTAAAACTATCTTTTCCCGAGGGATTCAAGTGGGAGAGGGTTATGAACTTCGAAGATGGTGGCGTTGTGACAGTCACGCAGGATTCTAGCTTACAAGACGGCGAATTCATATACAAAGTCAAGCTCCGTGGAACAAACTTTCCTTCCGATGGTCCAGTAATGCAAAAGAAAACGATGGGATGGGAAGCCAGTTCAGAGCGTATGTATCCCGAAGATGGTGCTCTTAAGGGTGAGATTAAACAGCGCTTAAAACTAAAAGATGGAGGACATTACGATGCCGAGGTTAAGACAACGTACAAAGCAAAGAAGCCTGTGCAGCTCCCGGGCGCATACAATGTAAACATTAAACTAGATATTACCTCCCATAATGAGGACTACACGATCGTAGAGCAATATGAGCGTGCTGAGGGAAGACATTCTACCGGGGGAATGGATGAGCTCTATAAATGA\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTCTCCAAAGGGGAAGAAGACAACATGGCCATAATAAAAGAATTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATAGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGGCCGCTGCCGTTCGCCTGGGACATACTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATACCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGGGTCGTCACCGTCACCCAGGACTCCTCCCTGCAGGACGGGGAATTCATATACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAAACCATGGGGTGGGAAGCCTCCTCCGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATAAAACAGCGTCTGAAACTGAAAGACGGGGGGCACTACGACGCCGAAGTCAAAACCACCTACAAAGCCAAAAAACCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATAAAACTGGACATAACCTCCCACAACGAAGACTACACCATAGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGGATGGACGAACTGTACAAATAA', 'average_distance_score': 0.9949357248987034, 'weakest_link_score': 0.9949357248987034}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTCTCCAAAGGGGAAGAAGACAACATGGCCATAATAAAAGAATTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATAGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGGCCGCTGCCGTTCGCCTGGGACATACTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATACCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGGGTCGTCACCGTCACCCAGGACTCCTCCCTGCAGGACGGGGAATTCATATACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAAACCATGGGGTGGGAAGCCTCCTCCGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATAAAACAGCGTCTGAAACTGAAAGACGGGGGGCACTACGACGCCGAAGTCAAAACCACCTACAAAGCCAAAAAACCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATAAAACTGGACATAACCTCCCACAACGAAGACTACACCATAGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGGATGGACGAACTGTACAAATAA', 'average_distance_score': 0.9949357248987034, 'weakest_link_score': 0.9949357248987034}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTCTCCAAAGGGGAAGAAGACAACATGGCCATAATAAAAGAATTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATAGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGGCCGCTGCCGTTCGCCTGGGACATACTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATACCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGGGTCGTCACCGTCACCCAGGACTCCTCCCTGCAGGACGGGGAATTCATATACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAAACCATGGGGTGGGAAGCCTCCTCCGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATAAAACAGCGTCTGAAACTGAAAGACGGGGGGCACTACGACGCCGAAGTCAAAACCACCTACAAAGCCAAAAAACCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATAAAACTGGACATAACCTCCCACAACGAAGACTACACCATAGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGGATGGACGAACTGTACAAATAA', 'average_distance_score': 0.9949357248987034, 'weakest_link_score': 0.9949357248987034}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTCTCCAAAGGGGAAGAAGACAACATGGCCATAATAAAAGAATTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATAGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGGCCGCTGCCGTTCGCCTGGGACATACTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATACCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGGGTCGTCACCGTCACCCAGGACTCCTCCCTGCAGGACGGGGAATTCATATACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAAACCATGGGGTGGGAAGCCTCCTCCGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATAAAACAGCGTCTGAAACTGAAAGACGGGGGGCACTACGACGCCGAAGTCAAAACCACCTACAAAGCCAAAAAACCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATAAAACTGGACATAACCTCCCACAACGAAGACTACACCATAGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGGATGGACGAACTGTACAAATGA', 'average_distance_score': 0.9949357248987034, 'weakest_link_score': 0.9949357248987034}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTCTCCAAAGGGGAAGAAGACAACATGGCCATAATAAAAGAATTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATAGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGGCCGCTGCCGTTCGCCTGGGACATACTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATACCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGGGTCGTCACCGTCACCCAGGACTCCTCCCTGCAGGACGGGGAATTCATATACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAAACCATGGGGTGGGAAGCCTCCTCCGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATAAAACAGCGTCTGAAACTGAAAGACGGGGGGCACTACGACGCCGAAGTCAAAACCACCTACAAAGCCAAAAAACCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATAAAACTGGACATAACCTCCCACAACGAAGACTACACCATAGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGGATGGACGAACTGTACAAATAA', 'average_distance_score': 0.9949357248987034, 'weakest_link_score': 0.9949357248987034}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry_debug\\tAI_zscore_bulk_aa_diff_ecoli_opt_True_GLpr\n",
      "Final output: {'user_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'optimization_method': 'zscore_bulk_aa_diff', 'optimization_cub_index': 'tAI', 'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': None, 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.75, 'AAC': 0.5, 'AAG': 0.23999999999999996, 'AAT': 0.29500000000000004, 'ACA': 0.1525, 'ACC': 0.25, 'ACG': 0.29, 'ACT': 0.14750000000000002, 'AGA': 0.13874999999999998, 'AGC': 0.125, 'AGG': 0.16499999999999998, 'AGT': 0.07375000000000001, 'ATA': 0.10999999999999999, 'ATC': 0.375, 'ATG': 1.0, 'ATT': 0.22125000000000003, 'CAA': 0.27749999999999997, 'CAC': 0.125, 'CAG': 0.32999999999999996, 'CAT': 0.07375000000000001, 'CCA': 0.13874999999999998, 'CCC': 0.125, 'CCG': 0.16499999999999998, 'CCT': 0.07375000000000001, 'CGA': 0.013799999999999993, 'CGC': 0.36, 'CGG': 0.125, 'CGT': 0.5, 'CTA': 0.18, 'CTC': 0.125, 'CTG': 0.54, 'CTT': 0.07375000000000001, 'GAA': 0.5, 'GAC': 0.375, 'GAG': 0.15999999999999998, 'GAT': 0.22125000000000003, 'GCA': 0.375, 'GCC': 0.25, 'GCG': 0.11999999999999998, 'GCT': 0.14750000000000002, 'GGA': 0.13874999999999998, 'GGC': 0.5, 'GGG': 0.16499999999999998, 'GGT': 0.29500000000000004, 'GTA': 0.625, 'GTC': 0.25, 'GTG': 0.19999999999999996, 'GTT': 0.14750000000000002, 'TAC': 0.375, 'TAT': 0.22125000000000003, 'TCA': 0.13874999999999998, 'TCC': 0.25, 'TCG': 0.16499999999999998, 'TCT': 0.14750000000000002, 'TGC': 0.125, 'TGG': 0.125, 'TGT': 0.07375000000000001, 'TTA': 0.13874999999999998, 'TTC': 0.25, 'TTG': 0.16499999999999998, 'TTT': 0.14750000000000002}, 'tai_avg': 0.2504950429983414, 'tai_std': 0.023481599705588187}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': None, 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.6666666666666666, 'AAC': 0.6666666666666666, 'AAG': 0.2133333333333333, 'AAT': 0.39333333333333337, 'ACA': 0.6666666666666666, 'ACC': 0.16666666666666666, 'ACG': 0.2133333333333333, 'ACT': 0.09833333333333334, 'AGA': 0.18499999999999997, 'AGC': 0.3333333333333333, 'AGG': 0.21999999999999997, 'AGT': 0.19666666666666668, 'ATA': 0.10999999999999999, 'ATC': 0.5, 'ATG': 1.0, 'ATT': 0.29500000000000004, 'CAA': 0.6666666666666666, 'CAC': 0.3333333333333333, 'CAG': 0.2133333333333333, 'CAT': 0.19666666666666668, 'CCA': 0.5, 'CCG': 0.15999999999999998, 'CGA': 0.01839999999999999, 'CGC': 0.48, 'CGG': 0.16666666666666666, 'CGT': 0.6666666666666666, 'CTA': 0.3516666666666666, 'CTC': 0.16666666666666666, 'CTG': 0.2733333333333333, 'CTT': 0.09833333333333334, 'GAA': 1.0, 'GAC': 0.6666666666666666, 'GAG': 0.31999999999999995, 'GAT': 0.39333333333333337, 'GCA': 0.8333333333333334, 'GCC': 0.16666666666666666, 'GCG': 0.2666666666666666, 'GCT': 0.09833333333333334, 'GGA': 0.5, 'GGC': 0.6666666666666666, 'GGG': 0.15999999999999998, 'GGT': 0.39333333333333337, 'GTA': 0.6666666666666666, 'GTC': 0.16666666666666666, 'GTG': 0.2133333333333333, 'GTT': 0.09833333333333334, 'TAC': 0.3333333333333333, 'TAT': 0.19666666666666668, 'TCA': 0.3333333333333333, 'TCC': 0.16666666666666666, 'TCG': 0.10666666666666665, 'TCT': 0.09833333333333334, 'TGC': 0.16666666666666666, 'TGG': 0.16666666666666666, 'TGT': 0.09833333333333334, 'TTA': 0.5183333333333333, 'TTC': 0.5, 'TTG': 0.32666666666666666, 'TTT': 0.29500000000000004}, 'tai_avg': 0.3292474875006557, 'tai_std': 0.027038419737753252}]}, 'orf': [{'initial_sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_sequence': 'ATGGTCTCCAAAGGGGAAGAAGACAACATGGCCATAATAAAAGAATTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATAGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGGCCGCTGCCGTTCGCCTGGGACATACTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATACCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGGGTCGTCACCGTCACCCAGGACTCCTCCCTGCAGGACGGGGAATTCATATACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAAACCATGGGGTGGGAAGCCTCCTCCGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATAAAACAGCGTCTGAAACTGAAAGACGGGGGGCACTACGACGCCGAAGTCAAAACCACCTACAAAGCCAAAAAACCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATAAAACTGGACATAACCTCCCACAACGAAGACTACACCATAGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGGATGGACGAACTGTACAAATAA', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCC', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCG', 'K': 'AAA', '_': 'TAA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTC', 'E': 'GAA', 'Y': 'TAC'}, 'initial_sequence_optimization_score': 0.5211667536834308, 'final_sequence_optimization_score': 1.5161024785821342, 'run_time': 19.79700000002049, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAT', 'R': 'CGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'sequence_score': 1.1780142992504663}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCC', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCG', 'K': 'AAA', '_': 'TAA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTC', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 1.5161024785821342}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCC', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCG', 'K': 'AAA', '_': 'TAA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTC', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 1.5161024785821342}]}, {'initial_sequence': 'ATGGTCTCCAAGGGCGAAGAAGATAACATGGCGATCATCAAAGAATTTATGCGGTTTAAGGTCCATATGGAAGGTTCTGTGAACGGGCATGAATTCGAAATAGAAGGGGAGGGCGAAGGACGGCCCTATGAAGGGACCCAAACCGCGAAACTGAAGGTGACTAAGGGGGGCCCACTTCCATTCGCATGGGACATTTTGTCACCCCAGTTCATGTATGGAAGTAAAGCTTACGTAAAGCACCCGGCTGATATACCAGACTACCTTAAACTAAGCTTTCCCGAGGGTTTTAAGTGGGAGCGTGTAATGAATTTCGAAGATGGAGGGGTAGTAACGGTCACTCAGGACTCTTCGCTGCAGGACGGAGAATTCATATATAAAGTTAAATTACGCGGTACCAACTTCCCCTCTGATGGTCCCGTTATGCAAAAGAAGACCATGGGATGGGAAGCGTCTAGCGAGCGCATGTACCCCGAGGATGGAGCACTAAAGGGCGAGATAAAGCAAAGGCTAAAGCTAAAAGACGGCGGGCATTACGATGCAGAGGTGAAAACCACCTACAAAGCGAAAAAGCCGGTGCAACTGCCGGGTGCGTACAACGTGAATATCAAGTTGGATATAACCAGTCATAATGAAGACTATACGATCGTGGAGCAGTACGAGCGGGCTGAAGGTAGGCATTCTACAGGCGGGATGGACGAACTTTATAAGTAA', 'final_sequence': 'ATGGTCTCCAAAGGGGAAGAAGACAACATGGCCATAATAAAAGAATTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATAGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGGCCGCTGCCGTTCGCCTGGGACATACTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATACCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGGGTCGTCACCGTCACCCAGGACTCCTCCCTGCAGGACGGGGAATTCATATACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAAACCATGGGGTGGGAAGCCTCCTCCGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATAAAACAGCGTCTGAAACTGAAAGACGGGGGGCACTACGACGCCGAAGTCAAAACCACCTACAAAGCCAAAAAACCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATAAAACTGGACATAACCTCCCACAACGAAGACTACACCATAGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGGATGGACGAACTGTACAAATAA', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCC', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCG', 'K': 'AAA', '_': 'TAA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTC', 'E': 'GAA', 'Y': 'TAC'}, 'initial_sequence_optimization_score': -0.16387024449958065, 'final_sequence_optimization_score': 1.5161024785821342, 'run_time': 17.859000000054948, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAT', 'R': 'CGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'sequence_score': 1.1780142992504663}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCC', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCG', 'K': 'AAA', '_': 'TAA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTC', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 1.5161024785821342}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCC', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCG', 'K': 'AAA', '_': 'TAA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTC', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 1.5161024785821342}]}, {'initial_sequence': 'ATGGTTTCCAAAGGCGAGGAAGATAATATGGCTATAATAAAAGAATTTATGCGTTTCAAGGTCCACATGGAAGGATCTGTTAACGGACATGAATTTGAGATAGAAGGCGAAGGGGAGGGACGGCCCTATGAAGGCACCCAAACAGCAAAATTAAAAGTTACGAAAGGAGGTCCCCTGCCATTTGCGTGGGATATACTCTCTCCACAATTTATGTACGGTTCTAAGGCTTATGTTAAACATCCTGCGGATATCCCCGACTACCTGAAGTTATCGTTCCCGGAAGGATTCAAGTGGGAGAGGGTTATGAACTTCGAGGATGGTGGGGTGGTCACGGTGACTCAAGACTCGTCCTTGCAAGACGGAGAATTCATCTACAAAGTAAAATTGCGCGGGACAAATTTCCCGTCCGATGGACCCGTCATGCAAAAAAAGACTATGGGTTGGGAGGCGAGCAGTGAACGAATGTACCCTGAGGATGGAGCACTCAAGGGCGAAATAAAGCAACGGTTAAAGCTCAAAGATGGCGGTCACTACGATGCTGAGGTGAAGACGACCTACAAGGCGAAAAAACCAGTGCAGTTGCCGGGGGCATATAATGTAAACATTAAGCTTGACATCACTTCGCACAACGAAGATTATACGATAGTTGAACAATATGAGCGAGCCGAGGGGCGGCACAGCACGGGAGGCATGGATGAGCTATACAAATAA', 'final_sequence': 'ATGGTCTCCAAAGGGGAAGAAGACAACATGGCCATAATAAAAGAATTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATAGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGGCCGCTGCCGTTCGCCTGGGACATACTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATACCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGGGTCGTCACCGTCACCCAGGACTCCTCCCTGCAGGACGGGGAATTCATATACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAAACCATGGGGTGGGAAGCCTCCTCCGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATAAAACAGCGTCTGAAACTGAAAGACGGGGGGCACTACGACGCCGAAGTCAAAACCACCTACAAAGCCAAAAAACCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATAAAACTGGACATAACCTCCCACAACGAAGACTACACCATAGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGGATGGACGAACTGTACAAATAA', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCC', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCG', 'K': 'AAA', '_': 'TAA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTC', 'E': 'GAA', 'Y': 'TAC'}, 'initial_sequence_optimization_score': -0.3005264579774237, 'final_sequence_optimization_score': 1.5161024785821342, 'run_time': 18.5, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAT', 'R': 'CGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'sequence_score': 1.1780142992504663}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCC', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCG', 'K': 'AAA', '_': 'TAA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTC', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 1.5161024785821342}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCC', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCG', 'K': 'AAA', '_': 'TAA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTC', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 1.5161024785821342}]}, {'initial_sequence': 'ATGGTATCAAAGGGCGAAGAAGACAATATGGCGATCATCAAGGAATTTATGCGGTTCAAGGTCCACATGGAAGGCTCTGTTAATGGGCACGAGTTTGAAATCGAGGGGGAGGGAGAAGGAAGACCGTATGAAGGGACCCAAACTGCTAAACTCAAAGTCACAAAGGGTGGACCACTGCCATTCGCGTGGGACATCCTATCTCCACAGTTCATGTACGGTTCGAAAGCGTATGTAAAACATCCAGCGGATATACCCGACTACTTAAAACTATCTTTTCCCGAGGGATTCAAGTGGGAGAGGGTTATGAACTTCGAAGATGGTGGCGTTGTGACAGTCACGCAGGATTCTAGCTTACAAGACGGCGAATTCATATACAAAGTCAAGCTCCGTGGAACAAACTTTCCTTCCGATGGTCCAGTAATGCAAAAGAAAACGATGGGATGGGAAGCCAGTTCAGAGCGTATGTATCCCGAAGATGGTGCTCTTAAGGGTGAGATTAAACAGCGCTTAAAACTAAAAGATGGAGGACATTACGATGCCGAGGTTAAGACAACGTACAAAGCAAAGAAGCCTGTGCAGCTCCCGGGCGCATACAATGTAAACATTAAACTAGATATTACCTCCCATAATGAGGACTACACGATCGTAGAGCAATATGAGCGTGCTGAGGGAAGACATTCTACCGGGGGAATGGATGAGCTCTATAAATGA', 'final_sequence': 'ATGGTCTCCAAAGGGGAAGAAGACAACATGGCCATAATAAAAGAATTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATAGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGGCCGCTGCCGTTCGCCTGGGACATACTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATACCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGGGTCGTCACCGTCACCCAGGACTCCTCCCTGCAGGACGGGGAATTCATATACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAAACCATGGGGTGGGAAGCCTCCTCCGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATAAAACAGCGTCTGAAACTGAAAGACGGGGGGCACTACGACGCCGAAGTCAAAACCACCTACAAAGCCAAAAAACCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATAAAACTGGACATAACCTCCCACAACGAAGACTACACCATAGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGGATGGACGAACTGTACAAATGA', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCC', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCG', 'K': 'AAA', '_': 'TAA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTC', 'E': 'GAA', 'Y': 'TAC'}, 'initial_sequence_optimization_score': -0.49067235917325247, 'final_sequence_optimization_score': 1.5161024785821342, 'run_time': 19.0, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAT', 'S': 'TCG', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAT', 'P': 'CCG', 'K': 'AAG', '_': 'TAA', 'T': 'ACT', 'F': 'TTT', 'A': 'GCT', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAT', 'R': 'CGA', 'W': 'TGG', 'V': 'GTT', 'E': 'GAG', 'Y': 'TAT'}, 'sequence_score': 1.1780142992504663}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCC', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCG', 'K': 'AAA', '_': 'TAA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTC', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 1.5161024785821342}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'TCC', 'Q': 'CAG', 'M': 'ATG', 'N': 'AAC', 'P': 'CCG', 'K': 'AAA', '_': 'TAA', 'T': 'ACC', 'F': 'TTC', 'A': 'GCC', 'G': 'GGG', 'I': 'ATA', 'L': 'CTG', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTC', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 1.5161024785821342}]}], 'evaluation': [{'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'tai_initial_score': 0.2799603826276079, 'tai_final_score': 0.3103044359154907, 'dist_score': 1.2922481291026109}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'tai_initial_score': 0.33499296920046817, 'tai_final_score': 0.31613033703667703, 'dist_score': -0.6976233206947959}], 'final_sequence': 'ATGGTCTCCAAAGGGGAAGAAGACAACATGGCCATAATAAAAGAATTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATAGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGGCCGCTGCCGTTCGCCTGGGACATACTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATACCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGGGTCGTCACCGTCACCCAGGACTCCTCCCTGCAGGACGGGGAATTCATATACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAAACCATGGGGTGGGAAGCCTCCTCCGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATAAAACAGCGTCTGAAACTGAAAGACGGGGGGCACTACGACGCCGAAGTCAAAACCACCTACAAAGCCAAAAAACCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATAAAACTGGACATAACCTCCCACAACGAAGACTACACCATAGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGGATGGACGAACTGTACAAATAA', 'average_distance_score': 0.9949357248987034, 'weakest_link_score': 0.9949357248987034}, {'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'tai_initial_score': 0.2799603826276079, 'tai_final_score': 0.3103044359154907, 'dist_score': 1.2922481291026109}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'tai_initial_score': 0.33499296920046817, 'tai_final_score': 0.31613033703667703, 'dist_score': -0.6976233206947959}], 'final_sequence': 'ATGGTCTCCAAAGGGGAAGAAGACAACATGGCCATAATAAAAGAATTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATAGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGGCCGCTGCCGTTCGCCTGGGACATACTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATACCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGGGTCGTCACCGTCACCCAGGACTCCTCCCTGCAGGACGGGGAATTCATATACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAAACCATGGGGTGGGAAGCCTCCTCCGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATAAAACAGCGTCTGAAACTGAAAGACGGGGGGCACTACGACGCCGAAGTCAAAACCACCTACAAAGCCAAAAAACCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATAAAACTGGACATAACCTCCCACAACGAAGACTACACCATAGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGGATGGACGAACTGTACAAATAA', 'average_distance_score': 0.9949357248987034, 'weakest_link_score': 0.9949357248987034}, {'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'tai_initial_score': 0.2799603826276079, 'tai_final_score': 0.3103044359154907, 'dist_score': 1.2922481291026109}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'tai_initial_score': 0.33499296920046817, 'tai_final_score': 0.31613033703667703, 'dist_score': -0.6976233206947959}], 'final_sequence': 'ATGGTCTCCAAAGGGGAAGAAGACAACATGGCCATAATAAAAGAATTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATAGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGGCCGCTGCCGTTCGCCTGGGACATACTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATACCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGGGTCGTCACCGTCACCCAGGACTCCTCCCTGCAGGACGGGGAATTCATATACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAAACCATGGGGTGGGAAGCCTCCTCCGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATAAAACAGCGTCTGAAACTGAAAGACGGGGGGCACTACGACGCCGAAGTCAAAACCACCTACAAAGCCAAAAAACCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATAAAACTGGACATAACCTCCCACAACGAAGACTACACCATAGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGGATGGACGAACTGTACAAATAA', 'average_distance_score': 0.9949357248987034, 'weakest_link_score': 0.9949357248987034}, {'organisms': [{'name': 'Escherichia coli', 'is_wanted': True, 'tai_initial_score': 0.2799603826276079, 'tai_final_score': 0.3103044359154907, 'dist_score': 1.2922481291026109}, {'name': 'Bacillus subtilis', 'is_wanted': False, 'tai_initial_score': 0.33499296920046817, 'tai_final_score': 0.31613033703667703, 'dist_score': -0.6976233206947959}], 'final_sequence': 'ATGGTCTCCAAAGGGGAAGAAGACAACATGGCCATAATAAAAGAATTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATAGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGGCCGCTGCCGTTCGCCTGGGACATACTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATACCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGGGTCGTCACCGTCACCCAGGACTCCTCCCTGCAGGACGGGGAATTCATATACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAAACCATGGGGTGGGAAGCCTCCTCCGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATAAAACAGCGTCTGAAACTGAAAGACGGGGGGCACTACGACGCCGAAGTCAAAACCACCTACAAAGCCAAAAAACCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATAAAACTGGACATAACCTCCCACAACGAAGACTACACCATAGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGGATGGACGAACTGTACAAATGA', 'average_distance_score': 0.9949357248987034, 'weakest_link_score': 0.9949357248987034}], 'final_evaluation': {'final_sequence': 'ATGGTCTCCAAAGGGGAAGAAGACAACATGGCCATAATAAAAGAATTCATGCGTTTCAAAGTCCACATGGAAGGGTCCGTCAACGGGCACGAATTCGAAATAGAAGGGGAAGGGGAAGGGCGTCCGTACGAAGGGACCCAGACCGCCAAACTGAAAGTCACCAAAGGGGGGCCGCTGCCGTTCGCCTGGGACATACTGTCCCCGCAGTTCATGTACGGGTCCAAAGCCTACGTCAAACACCCGGCCGACATACCGGACTACCTGAAACTGTCCTTCCCGGAAGGGTTCAAATGGGAACGTGTCATGAACTTCGAAGACGGGGGGGTCGTCACCGTCACCCAGGACTCCTCCCTGCAGGACGGGGAATTCATATACAAAGTCAAACTGCGTGGGACCAACTTCCCGTCCGACGGGCCGGTCATGCAGAAAAAAACCATGGGGTGGGAAGCCTCCTCCGAACGTATGTACCCGGAAGACGGGGCCCTGAAAGGGGAAATAAAACAGCGTCTGAAACTGAAAGACGGGGGGCACTACGACGCCGAAGTCAAAACCACCTACAAAGCCAAAAAACCGGTCCAGCTGCCGGGGGCCTACAACGTCAACATAAAACTGGACATAACCTCCCACAACGAAGACTACACCATAGTCGAACAGTACGAACGTGCCGAAGGGCGTCACTCCACCGGGGGGATGGACGAACTGTACAAATAA', 'average_distance_score': 0.9949357248987034, 'weakest_link_score': 0.9949357248987034}, 'zip_output_file_path': 'results\\\\mcherry_debug\\\\tAI_zscore_bulk_aa_diff_ecoli_opt_True_GLpr\\\\communique_results.zip'}\n",
      "##########################\n",
      "# USER INPUT INFORMATION #\n",
      "##########################\n",
      "Organism is optimized\n",
      "Information about Bacillus subtilis:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Bacillus subtilis. Calculating tAI profile.\n",
      "name=Bacillus subtilis, tai_std=0.027038419737753252, tai_avg=0.3292474875006557\n",
      "Organism is de-optimized\n",
      "Information about Escherichia coli:\n",
      "Extracting expression levels from: protein_abundance file.\n",
      "tGCN values were found for Escherichia coli. Calculating tAI profile.\n",
      "name=Escherichia coli, tai_std=0.023481599705588187, tai_avg=0.2504950429983414\n",
      "Bacillus subtilis has weight of 1.0\n",
      "Escherichia coli has weight of 1.0\n",
      "Sequence to be optimized given in the following file: C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\mCherry_original.fasta\n",
      "Open reading frame sequence for optimization is: ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Total input processing time: 20.629909992218018\n",
      "##########################\n",
      "# ORF #\n",
      "##########################\n",
      "\n",
      "Optimization method used is: OptimizationMethod.zscore_bulk_aa_diff\n",
      "Running ORF optimization for initial sequence: ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA\n",
      "Running ORF optimization for initial sequence: ATGGTATCAAAGGGCGAAGAAGATAACATGGCTATAATAAAGGAATTTATGAGGTTCAAGGTCCATATGGAGGGCTCAGTCAACGGACACGAGTTTGAAATAGAAGGTGAAGGCGAAGGAAGGCCCTACGAGGGAACTCAAACGGCCAAACTGAAAGTAACGAAAGGCGGTCCTCTCCCCTTCGCCTGGGACATTCTCAGCCCCCAGTTTATGTACGGATCCAAAGCATATGTGAAGCACCCAGCTGATATTCCAGATTACCTGAAGCTTTCATTCCCTGAGGGTTTTAAATGGGAAAGAGTTATGAACTTCGAAGACGGCGGGGTAGTCACAGTAACCCAGGATTCTAGCTTACAAGATGGTGAGTTTATATACAAGGTTAAGCTCCGAGGCACAAATTTTCCTTCGGATGGACCAGTTATGCAAAAAAAGACTATGGGCTGGGAAGCTAGCAGTGAGCGCATGTACCCTGAAGATGGCGCACTTAAAGGGGAAATTAAACAACGATTGAAGTTGAAGGATGGAGGGCATTACGACGCGGAAGTTAAGACAACGTATAAGGCCAAGAAGCCCGTTCAGCTACCGGGCGCCTACAATGTAAATATCAAGCTCGACATAACTTCGCATAACGAGGATTACACCATCGTCGAACAATACGAACGCGCAGAGGGGCGTCATTCTACCGGCGGAATGGATGAACTTTACAAATAG\n",
      "Running ORF optimization for initial sequence: ATGGTCAGTAAAGGAGAGGAAGATAACATGGCTATAATCAAAGAGTTCATGCGCTTTAAAGTTCATATGGAGGGTTCGGTCAACGGTCATGAGTTTGAGATAGAGGGAGAGGGCGAGGGGAGGCCCTATGAGGGAACACAAACCGCGAAGCTTAAGGTTACCAAGGGCGGACCCCTTCCATTTGCTTGGGATATACTCAGCCCCCAATTTATGTATGGGTCGAAAGCATACGTGAAACATCCTGCCGATATCCCTGATTATCTGAAATTAAGTTTCCCTGAGGGTTTTAAGTGGGAGCGCGTAATGAATTTCGAAGATGGTGGTGTAGTCACGGTCACTCAAGATAGTAGTCTCCAAGACGGCGAATTTATTTATAAGGTCAAGCTACGGGGGACCAATTTCCCTAGTGATGGACCCGTGATGCAGAAGAAGACGATGGGATGGGAAGCGTCAAGTGAGCGGATGTATCCAGAAGACGGTGCTCTTAAAGGAGAAATCAAACAAAGATTGAAGTTAAAAGACGGAGGGCACTACGACGCGGAGGTAAAGACGACATACAAAGCCAAAAAGCCGGTGCAGCTCCCAGGGGCATACAACGTGAATATAAAGCTTGACATCACATCGCACAACGAAGATTACACGATTGTGGAACAATATGAGCGCGCTGAAGGCCGGCATTCAACAGGCGGTATGGACGAGTTATACAAGTGA\n",
      "Running ORF optimization for initial sequence: ATGGTATCGAAAGGCGAAGAGGATAATATGGCCATAATAAAAGAATTCATGCGTTTCAAGGTGCATATGGAGGGATCTGTAAACGGGCATGAGTTTGAGATTGAGGGCGAAGGAGAAGGCCGGCCGTACGAAGGCACGCAAACCGCAAAACTAAAGGTAACTAAAGGGGGCCCTCTTCCATTTGCCTGGGATATATTAAGTCCTCAGTTCATGTATGGGAGTAAAGCATACGTGAAGCACCCGGCCGATATTCCTGACTACCTGAAATTATCATTCCCTGAAGGCTTTAAGTGGGAGCGTGTTATGAACTTCGAAGATGGGGGTGTAGTTACCGTTACCCAAGACTCGTCCTTGCAAGACGGGGAGTTTATCTATAAGGTCAAATTAAGAGGTACGAACTTTCCGAGTGACGGTCCTGTAATGCAGAAAAAGACAATGGGTTGGGAGGCCTCCTCAGAGAGAATGTATCCAGAGGACGGGGCGCTAAAGGGCGAAATAAAGCAGCGTCTAAAGCTGAAAGACGGTGGACATTACGACGCTGAAGTAAAGACGACATATAAAGCAAAAAAACCTGTGCAACTTCCCGGGGCATATAATGTAAATATCAAATTAGACATCACATCCCACAACGAGGATTATACAATCGTGGAACAATATGAACGTGCAGAGGGACGGCATTCTACTGGGGGAATGGATGAATTGTATAAGTAG\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTAAGCAAAGGAGAAGAAGACAACATGGCAATCATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGACCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGAGTAGTAACAGTAACACAAGACAGCAGCTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAAACAATGGGATGGGAAGCAAGCAGCGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGACACTACGACGCAGAAGTAAAAACAACATACAAAGCAAAAAAACCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGAATGGACGAATTATACAAATAA', 'average_distance_score': 4.259816678961814, 'weakest_link_score': 4.259816678961814}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTAAGCAAAGGAGAAGAAGACAACATGGCAATCATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGACCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGAGTAGTAACAGTAACACAAGACAGCAGCTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAAACAATGGGATGGGAAGCAAGCAGCGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGACACTACGACGCAGAAGTAAAAACAACATACAAAGCAAAAAAACCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGAATGGACGAATTATACAAATAG', 'average_distance_score': 4.259816678961814, 'weakest_link_score': 4.259816678961814}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTAAGCAAAGGAGAAGAAGACAACATGGCAATCATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGACCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGAGTAGTAACAGTAACACAAGACAGCAGCTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAAACAATGGGATGGGAAGCAAGCAGCGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGACACTACGACGCAGAAGTAAAAACAACATACAAAGCAAAAAAACCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGAATGGACGAATTATACAAATGA', 'average_distance_score': 4.259816678961814, 'weakest_link_score': 4.259816678961814}\n",
      "Analyzing evaluation result: {'final_sequence': 'ATGGTAAGCAAAGGAGAAGAAGACAACATGGCAATCATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGACCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGAGTAGTAACAGTAACACAAGACAGCAGCTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAAACAATGGGATGGGAAGCAAGCAGCGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGACACTACGACGCAGAAGTAAAAACAACATACAAAGCAAAAAAACCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGAATGGACGAATTATACAAATAG', 'average_distance_score': 4.259816678961814, 'weakest_link_score': 4.259816678961814}\n",
      "Final evaluation result: {'final_sequence': 'ATGGTAAGCAAAGGAGAAGAAGACAACATGGCAATCATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGACCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGAGTAGTAACAGTAACACAAGACAGCAGCTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAAACAATGGGATGGGAAGCAAGCAGCGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGACACTACGACGCAGAAGTAAAAACAACATACAAAGCAAAAAAACCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGAATGGACGAATTATACAAATAA', 'average_distance_score': 4.259816678961814, 'weakest_link_score': 4.259816678961814}\n",
      "###########################\n",
      "# USER OUTPUT INFORMATION #\n",
      "###########################\n",
      "Output zip file directory path: results\\mcherry_debug\\tAI_zscore_bulk_aa_diff_ecoli_opt_False_ZpVj\n",
      "Final output: {'user_input': {'sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'tuning_parameter': 0.5, 'optimization_method': 'zscore_bulk_aa_diff', 'optimization_cub_index': 'tAI', 'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'optimization_priority': 1.0, 'cai_weights': None, 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.6666666666666666, 'AAC': 0.6666666666666666, 'AAG': 0.2133333333333333, 'AAT': 0.39333333333333337, 'ACA': 0.6666666666666666, 'ACC': 0.16666666666666666, 'ACG': 0.2133333333333333, 'ACT': 0.09833333333333334, 'AGA': 0.18499999999999997, 'AGC': 0.3333333333333333, 'AGG': 0.21999999999999997, 'AGT': 0.19666666666666668, 'ATA': 0.10999999999999999, 'ATC': 0.5, 'ATG': 1.0, 'ATT': 0.29500000000000004, 'CAA': 0.6666666666666666, 'CAC': 0.3333333333333333, 'CAG': 0.2133333333333333, 'CAT': 0.19666666666666668, 'CCA': 0.5, 'CCG': 0.15999999999999998, 'CGA': 0.01839999999999999, 'CGC': 0.48, 'CGG': 0.16666666666666666, 'CGT': 0.6666666666666666, 'CTA': 0.3516666666666666, 'CTC': 0.16666666666666666, 'CTG': 0.2733333333333333, 'CTT': 0.09833333333333334, 'GAA': 1.0, 'GAC': 0.6666666666666666, 'GAG': 0.31999999999999995, 'GAT': 0.39333333333333337, 'GCA': 0.8333333333333334, 'GCC': 0.16666666666666666, 'GCG': 0.2666666666666666, 'GCT': 0.09833333333333334, 'GGA': 0.5, 'GGC': 0.6666666666666666, 'GGG': 0.15999999999999998, 'GGT': 0.39333333333333337, 'GTA': 0.6666666666666666, 'GTC': 0.16666666666666666, 'GTG': 0.2133333333333333, 'GTT': 0.09833333333333334, 'TAC': 0.3333333333333333, 'TAT': 0.19666666666666668, 'TCA': 0.3333333333333333, 'TCC': 0.16666666666666666, 'TCG': 0.10666666666666665, 'TCT': 0.09833333333333334, 'TGC': 0.16666666666666666, 'TGG': 0.16666666666666666, 'TGT': 0.09833333333333334, 'TTA': 0.5183333333333333, 'TTC': 0.5, 'TTG': 0.32666666666666666, 'TTT': 0.29500000000000004}, 'tai_avg': 0.3292474875006557, 'tai_std': 0.027038419737753252}, {'name': 'Escherichia coli', 'is_wanted': False, 'optimization_priority': 1.0, 'cai_weights': None, 'cai_avg': None, 'cai_std': None, 'tai_weights': {'AAA': 0.75, 'AAC': 0.5, 'AAG': 0.23999999999999996, 'AAT': 0.29500000000000004, 'ACA': 0.1525, 'ACC': 0.25, 'ACG': 0.29, 'ACT': 0.14750000000000002, 'AGA': 0.13874999999999998, 'AGC': 0.125, 'AGG': 0.16499999999999998, 'AGT': 0.07375000000000001, 'ATA': 0.10999999999999999, 'ATC': 0.375, 'ATG': 1.0, 'ATT': 0.22125000000000003, 'CAA': 0.27749999999999997, 'CAC': 0.125, 'CAG': 0.32999999999999996, 'CAT': 0.07375000000000001, 'CCA': 0.13874999999999998, 'CCC': 0.125, 'CCG': 0.16499999999999998, 'CCT': 0.07375000000000001, 'CGA': 0.013799999999999993, 'CGC': 0.36, 'CGG': 0.125, 'CGT': 0.5, 'CTA': 0.18, 'CTC': 0.125, 'CTG': 0.54, 'CTT': 0.07375000000000001, 'GAA': 0.5, 'GAC': 0.375, 'GAG': 0.15999999999999998, 'GAT': 0.22125000000000003, 'GCA': 0.375, 'GCC': 0.25, 'GCG': 0.11999999999999998, 'GCT': 0.14750000000000002, 'GGA': 0.13874999999999998, 'GGC': 0.5, 'GGG': 0.16499999999999998, 'GGT': 0.29500000000000004, 'GTA': 0.625, 'GTC': 0.25, 'GTG': 0.19999999999999996, 'GTT': 0.14750000000000002, 'TAC': 0.375, 'TAT': 0.22125000000000003, 'TCA': 0.13874999999999998, 'TCC': 0.25, 'TCG': 0.16499999999999998, 'TCT': 0.14750000000000002, 'TGC': 0.125, 'TGG': 0.125, 'TGT': 0.07375000000000001, 'TTA': 0.13874999999999998, 'TTC': 0.25, 'TTG': 0.16499999999999998, 'TTT': 0.14750000000000002}, 'tai_avg': 0.2504950429983414, 'tai_std': 0.023481599705588187}]}, 'orf': [{'initial_sequence': 'ATGGTTTCCAAGGGCGAGGAGGATAACATGGCTATCATTAAAGAGTTCATGCGCTTCAAAGTTCACATGGAGGGTTCTGTTAACGGTCACGAGTTCGAGATCGAAGGCGAAGGCGAGGGCCGTCCGTATGAAGGCACCCAGACCGCCAAACTGAAAGTGACTAAAGGCGGCCCGCTGCCTTTTGCGTGGGACATCCTGAGCCCGCAATTTATGTACGGTTCTAAAGCGTATGTTAAACACCCAGCGGATATCCCGGACTATCTGAAGCTGTCTTTTCCGGAAGGTTTCAAGTGGGAACGCGTAATGAATTTTGAAGATGGTGGTGTCGTGACCGTCACTCAGGACTCCTCCCTGCAGGATGGCGAGTTCATCTATAAAGTTAAACTGCGTGGTACTAATTTTCCATCTGATGGCCCGGTGATGCAGAAAAAGACGATGGGTTGGGAGGCGTCTAGCGAACGCATGTATCCGGAAGATGGTGCGCTGAAAGGCGAAATTAAACAGCGCCTGAAACTGAAAGATGGCGGCCATTATGACGCTGAAGTGAAAACCACGTACAAAGCCAAGAAACCTGTGCAGCTGCCTGGCGCGTACAATGTGAATATTAAACTGGACATCACCTCTCATAATGAAGATTATACGATCGTAGAGCAATATGAGCGCGCGGAGGGTCGTCATTCTACCGGTGGCATGGATGAGCTGTACAAATAA', 'final_sequence': 'ATGGTAAGCAAAGGAGAAGAAGACAACATGGCAATCATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGACCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGAGTAGTAACAGTAACACAAGACAGCAGCTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAAACAATGGGATGGGAAGCAAGCAGCGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGACACTACGACGCAGAAGTAAAAACAACATACAAAGCAAAAAAACCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGAATGGACGAATTATACAAATAA', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCA', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'initial_sequence_optimization_score': -0.5211667536834308, 'final_sequence_optimization_score': 3.7386499252783834, 'run_time': 18.359000000054948, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 3.730263970744237}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCA', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 3.7386499252783834}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCA', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 3.7386499252783834}]}, {'initial_sequence': 'ATGGTATCAAAGGGCGAAGAAGATAACATGGCTATAATAAAGGAATTTATGAGGTTCAAGGTCCATATGGAGGGCTCAGTCAACGGACACGAGTTTGAAATAGAAGGTGAAGGCGAAGGAAGGCCCTACGAGGGAACTCAAACGGCCAAACTGAAAGTAACGAAAGGCGGTCCTCTCCCCTTCGCCTGGGACATTCTCAGCCCCCAGTTTATGTACGGATCCAAAGCATATGTGAAGCACCCAGCTGATATTCCAGATTACCTGAAGCTTTCATTCCCTGAGGGTTTTAAATGGGAAAGAGTTATGAACTTCGAAGACGGCGGGGTAGTCACAGTAACCCAGGATTCTAGCTTACAAGATGGTGAGTTTATATACAAGGTTAAGCTCCGAGGCACAAATTTTCCTTCGGATGGACCAGTTATGCAAAAAAAGACTATGGGCTGGGAAGCTAGCAGTGAGCGCATGTACCCTGAAGATGGCGCACTTAAAGGGGAAATTAAACAACGATTGAAGTTGAAGGATGGAGGGCATTACGACGCGGAAGTTAAGACAACGTATAAGGCCAAGAAGCCCGTTCAGCTACCGGGCGCCTACAATGTAAATATCAAGCTCGACATAACTTCGCATAACGAGGATTACACCATCGTCGAACAATACGAACGCGCAGAGGGGCGTCATTCTACCGGCGGAATGGATGAACTTTACAAATAG', 'final_sequence': 'ATGGTAAGCAAAGGAGAAGAAGACAACATGGCAATCATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGACCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGAGTAGTAACAGTAACACAAGACAGCAGCTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAAACAATGGGATGGGAAGCAAGCAGCGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGACACTACGACGCAGAAGTAAAAACAACATACAAAGCAAAAAAACCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGAATGGACGAATTATACAAATAG', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCA', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'initial_sequence_optimization_score': 0.3251098935596737, 'final_sequence_optimization_score': 3.7386499252783834, 'run_time': 19.468999999924563, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 3.730263970744237}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCA', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 3.7386499252783834}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCA', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 3.7386499252783834}]}, {'initial_sequence': 'ATGGTCAGTAAAGGAGAGGAAGATAACATGGCTATAATCAAAGAGTTCATGCGCTTTAAAGTTCATATGGAGGGTTCGGTCAACGGTCATGAGTTTGAGATAGAGGGAGAGGGCGAGGGGAGGCCCTATGAGGGAACACAAACCGCGAAGCTTAAGGTTACCAAGGGCGGACCCCTTCCATTTGCTTGGGATATACTCAGCCCCCAATTTATGTATGGGTCGAAAGCATACGTGAAACATCCTGCCGATATCCCTGATTATCTGAAATTAAGTTTCCCTGAGGGTTTTAAGTGGGAGCGCGTAATGAATTTCGAAGATGGTGGTGTAGTCACGGTCACTCAAGATAGTAGTCTCCAAGACGGCGAATTTATTTATAAGGTCAAGCTACGGGGGACCAATTTCCCTAGTGATGGACCCGTGATGCAGAAGAAGACGATGGGATGGGAAGCGTCAAGTGAGCGGATGTATCCAGAAGACGGTGCTCTTAAAGGAGAAATCAAACAAAGATTGAAGTTAAAAGACGGAGGGCACTACGACGCGGAGGTAAAGACGACATACAAAGCCAAAAAGCCGGTGCAGCTCCCAGGGGCATACAACGTGAATATAAAGCTTGACATCACATCGCACAACGAAGATTACACGATTGTGGAACAATATGAGCGCGCTGAAGGCCGGCATTCAACAGGCGGTATGGACGAGTTATACAAGTGA', 'final_sequence': 'ATGGTAAGCAAAGGAGAAGAAGACAACATGGCAATCATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGACCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGAGTAGTAACAGTAACACAAGACAGCAGCTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAAACAATGGGATGGGAAGCAAGCAGCGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGACACTACGACGCAGAAGTAAAAACAACATACAAAGCAAAAAAACCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGAATGGACGAATTATACAAATGA', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCA', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'initial_sequence_optimization_score': 0.5691069105856675, 'final_sequence_optimization_score': 3.7386499252783834, 'run_time': 19.04700000002049, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 3.730263970744237}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCA', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 3.7386499252783834}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCA', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 3.7386499252783834}]}, {'initial_sequence': 'ATGGTATCGAAAGGCGAAGAGGATAATATGGCCATAATAAAAGAATTCATGCGTTTCAAGGTGCATATGGAGGGATCTGTAAACGGGCATGAGTTTGAGATTGAGGGCGAAGGAGAAGGCCGGCCGTACGAAGGCACGCAAACCGCAAAACTAAAGGTAACTAAAGGGGGCCCTCTTCCATTTGCCTGGGATATATTAAGTCCTCAGTTCATGTATGGGAGTAAAGCATACGTGAAGCACCCGGCCGATATTCCTGACTACCTGAAATTATCATTCCCTGAAGGCTTTAAGTGGGAGCGTGTTATGAACTTCGAAGATGGGGGTGTAGTTACCGTTACCCAAGACTCGTCCTTGCAAGACGGGGAGTTTATCTATAAGGTCAAATTAAGAGGTACGAACTTTCCGAGTGACGGTCCTGTAATGCAGAAAAAGACAATGGGTTGGGAGGCCTCCTCAGAGAGAATGTATCCAGAGGACGGGGCGCTAAAGGGCGAAATAAAGCAGCGTCTAAAGCTGAAAGACGGTGGACATTACGACGCTGAAGTAAAGACGACATATAAAGCAAAAAAACCTGTGCAACTTCCCGGGGCATATAATGTAAATATCAAATTAGACATCACATCCCACAACGAGGATTATACAATCGTGGAACAATATGAACGTGCAGAGGGACGGCATTCTACTGGGGGAATGGATGAATTGTATAAGTAG', 'final_sequence': 'ATGGTAAGCAAAGGAGAAGAAGACAACATGGCAATCATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGACCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGAGTAGTAACAGTAACACAAGACAGCAGCTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAAACAATGGGATGGGAAGCAAGCAGCGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGACACTACGACGCAGAAGTAAAAACAACATACAAAGCAAAAAAACCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGAATGGACGAATTATACAAATAG', 'iterations_count': 3, 'aa_to_optimal_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCA', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'initial_sequence_optimization_score': 0.47588461707629015, 'final_sequence_optimization_score': 3.7386499252783834, 'run_time': 18.32799999997951, 'iterations_summary': [{'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCT', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 3.730263970744237}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCA', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 3.7386499252783834}, {'aa_to_selected_codon': {'C': 'TGC', 'D': 'GAC', 'S': 'AGC', 'Q': 'CAA', 'M': 'ATG', 'N': 'AAC', 'P': 'CCA', 'K': 'AAA', '_': 'TAA', 'T': 'ACA', 'F': 'TTC', 'A': 'GCA', 'G': 'GGA', 'I': 'ATC', 'L': 'TTA', 'H': 'CAC', 'R': 'CGT', 'W': 'TGG', 'V': 'GTA', 'E': 'GAA', 'Y': 'TAC'}, 'sequence_score': 3.7386499252783834}]}], 'evaluation': [{'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'tai_initial_score': 0.33499296920046817, 'tai_final_score': 0.5858466922697777, 'dist_score': 9.277676931653186}, {'name': 'Escherichia coli', 'is_wanted': False, 'tai_initial_score': 0.2799603826276079, 'tai_final_score': 0.29776045838531895, 'dist_score': 0.7580435737295591}], 'final_sequence': 'ATGGTAAGCAAAGGAGAAGAAGACAACATGGCAATCATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGACCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGAGTAGTAACAGTAACACAAGACAGCAGCTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAAACAATGGGATGGGAAGCAAGCAGCGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGACACTACGACGCAGAAGTAAAAACAACATACAAAGCAAAAAAACCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGAATGGACGAATTATACAAATAA', 'average_distance_score': 4.259816678961814, 'weakest_link_score': 4.259816678961814}, {'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'tai_initial_score': 0.33499296920046817, 'tai_final_score': 0.5858466922697777, 'dist_score': 9.277676931653186}, {'name': 'Escherichia coli', 'is_wanted': False, 'tai_initial_score': 0.2799603826276079, 'tai_final_score': 0.29776045838531895, 'dist_score': 0.7580435737295591}], 'final_sequence': 'ATGGTAAGCAAAGGAGAAGAAGACAACATGGCAATCATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGACCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGAGTAGTAACAGTAACACAAGACAGCAGCTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAAACAATGGGATGGGAAGCAAGCAGCGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGACACTACGACGCAGAAGTAAAAACAACATACAAAGCAAAAAAACCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGAATGGACGAATTATACAAATAG', 'average_distance_score': 4.259816678961814, 'weakest_link_score': 4.259816678961814}, {'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'tai_initial_score': 0.33499296920046817, 'tai_final_score': 0.5858466922697777, 'dist_score': 9.277676931653186}, {'name': 'Escherichia coli', 'is_wanted': False, 'tai_initial_score': 0.2799603826276079, 'tai_final_score': 0.29776045838531895, 'dist_score': 0.7580435737295591}], 'final_sequence': 'ATGGTAAGCAAAGGAGAAGAAGACAACATGGCAATCATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGACCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGAGTAGTAACAGTAACACAAGACAGCAGCTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAAACAATGGGATGGGAAGCAAGCAGCGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGACACTACGACGCAGAAGTAAAAACAACATACAAAGCAAAAAAACCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGAATGGACGAATTATACAAATGA', 'average_distance_score': 4.259816678961814, 'weakest_link_score': 4.259816678961814}, {'organisms': [{'name': 'Bacillus subtilis', 'is_wanted': True, 'tai_initial_score': 0.33499296920046817, 'tai_final_score': 0.5858466922697777, 'dist_score': 9.277676931653186}, {'name': 'Escherichia coli', 'is_wanted': False, 'tai_initial_score': 0.2799603826276079, 'tai_final_score': 0.29776045838531895, 'dist_score': 0.7580435737295591}], 'final_sequence': 'ATGGTAAGCAAAGGAGAAGAAGACAACATGGCAATCATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGACCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGAGTAGTAACAGTAACACAAGACAGCAGCTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAAACAATGGGATGGGAAGCAAGCAGCGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGACACTACGACGCAGAAGTAAAAACAACATACAAAGCAAAAAAACCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGAATGGACGAATTATACAAATAG', 'average_distance_score': 4.259816678961814, 'weakest_link_score': 4.259816678961814}], 'final_evaluation': {'final_sequence': 'ATGGTAAGCAAAGGAGAAGAAGACAACATGGCAATCATCAAAGAATTCATGCGTTTCAAAGTACACATGGAAGGAAGCGTAAACGGACACGAATTCGAAATCGAAGGAGAAGGAGAAGGACGTCCATACGAAGGAACACAAACAGCAAAATTAAAAGTAACAAAAGGAGGACCATTACCATTCGCATGGGACATCTTAAGCCCACAATTCATGTACGGAAGCAAAGCATACGTAAAACACCCAGCAGACATCCCAGACTACTTAAAATTAAGCTTCCCAGAAGGATTCAAATGGGAACGTGTAATGAACTTCGAAGACGGAGGAGTAGTAACAGTAACACAAGACAGCAGCTTACAAGACGGAGAATTCATCTACAAAGTAAAATTACGTGGAACAAACTTCCCAAGCGACGGACCAGTAATGCAAAAAAAAACAATGGGATGGGAAGCAAGCAGCGAACGTATGTACCCAGAAGACGGAGCATTAAAAGGAGAAATCAAACAACGTTTAAAATTAAAAGACGGAGGACACTACGACGCAGAAGTAAAAACAACATACAAAGCAAAAAAACCAGTACAATTACCAGGAGCATACAACGTAAACATCAAATTAGACATCACAAGCCACAACGAAGACTACACAATCGTAGAACAATACGAACGTGCAGAAGGACGTCACAGCACAGGAGGAATGGACGAATTATACAAATAA', 'average_distance_score': 4.259816678961814, 'weakest_link_score': 4.259816678961814}, 'zip_output_file_path': 'results\\\\mcherry_debug\\\\tAI_zscore_bulk_aa_diff_ecoli_opt_False_ZpVj\\\\communique_results.zip'}\n"
     ]
    }
   ],
   "source": [
    "run_all_methods(orf_sequence_file=macherry_file_path, output_path=\"mcherry\", should_use_protein_abundance=True, should_use_mrna_levels=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "list indices must be integers or slices, not str",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Temp\\ipykernel_20504\\3592340000.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      2\u001b[0m \u001b[0mmcherry_variants_file\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mos\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmcherry_root_dir\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\"mcherry_variants.fasta\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[0mgenerate_sequences_fasta_file\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmcherry_root_dir\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 4\u001b[1;33m \u001b[0mgenerate_summary\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mresults_directory\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mmcherry_root_dir\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32mC:\\projects\\Igem_TAU_2021_moran\\analysis\\orf_model_analysis\\generate_summary_file.py\u001b[0m in \u001b[0;36mgenerate_summary\u001b[1;34m(results_directory)\u001b[0m\n\u001b[0;32m     90\u001b[0m                                         ordered_organisms=ordered_organisms)\n\u001b[0;32m     91\u001b[0m                 \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 92\u001b[1;33m                     \u001b[0mordered_organisms\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mupdate_from_summary\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfile_path\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mfile_path\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mworksheet\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mworksheet\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     93\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     94\u001b[0m     \u001b[0mworkbook\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msave\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mos\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mresults_directory\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\"summary.xlsx\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mC:\\projects\\Igem_TAU_2021_moran\\analysis\\orf_model_analysis\\generate_summary_file.py\u001b[0m in \u001b[0;36mupdate_from_summary\u001b[1;34m(file_path, worksheet, ordered_organisms)\u001b[0m\n\u001b[0;32m    251\u001b[0m     \u001b[0msummary_row\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[0moptimization_method\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0moptimization_cub_index\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    252\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 253\u001b[1;33m     \u001b[0morganisms\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0msummary\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m\"evaluation\"\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m\"organisms\"\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    254\u001b[0m     \u001b[1;31m# organisms.sort(key=lambda x: x.get(\"is_wanted\"))\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    255\u001b[0m     \u001b[1;32mfor\u001b[0m \u001b[0morganism_name\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mordered_organisms\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mTypeError\u001b[0m: list indices must be integers or slices, not str"
     ]
    }
   ],
   "source": [
    "mcherry_root_dir = r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\orf_model_analysis\\results\\mcherry\"\n",
    "mcherry_variants_file = os.path.join(mcherry_root_dir, \"mcherry_variants.fasta\")\n",
    "generate_sequences_fasta_file(mcherry_root_dir)\n",
    "generate_summary(results_directory=mcherry_root_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tuning parameter = 1 (optimizing only wanted host and neglecting unwanted host)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run_all_methods(orf_sequence_file=macherry_file_path, output_path=\"mcherry_greater_alpha\", tuning_param=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Using protein abundance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " run_all_methods(orf_sequence_file=macherry_file_path, output_path=\"mcherry_greater_alpha\", tuning_param=1, should_use_protein_abundance=True, should_use_mrna_levels=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mcherry_greater_alpha_root_dir = r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\orf_model_analysis\\results\\mcherry_greater_alpha\"\n",
    "mcherry_greater_alpha_variants_file = os.path.join(mcherry_greater_alpha_root_dir, \"mcherry_variants.fasta\")\n",
    "generate_sequences_fasta_file(mcherry_greater_alpha_root_dir)\n",
    "generate_summary(results_directory=mcherry_greater_alpha_root_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tuning parameter = 0 (de-optimizing uwanted host and neglecting wanted host)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run_all_methods(orf_sequence_file=macherry_file_path, output_path=\"mcherry_lower_alpha\", tuning_param=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Using protein abundance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "run_all_methods(orf_sequence_file=macherry_file_path, output_path=\"mcherry_lower_alpha\", tuning_param=0, should_use_protein_abundance=True, should_use_mrna_levels=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mcherry_lower_alpha_root_dir = r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\orf_model_analysis\\results\\mcherry_lower_alpha\"\n",
    "mcherry_lower_alpha_variants_file = os.path.join(mcherry_lower_alpha_root_dir, \"mcherry_variants.fasta\")\n",
    "generate_sequences_fasta_file(mcherry_lower_alpha_root_dir)\n",
    "generate_summary(results_directory=mcherry_lower_alpha_root_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Using ribosomal genes as highly expressed genes (instead of mRNA levels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run_all_methods(orf_sequence_file=macherry_file_path, output_path=\"ribosomal\", should_use_mrna_levels=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mcherry_ribosomal_root_dir = r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\orf_model_analysis\\results\\mrna\\ribosomal\"\n",
    "mcherry_ribosomal_variants_file = os.path.join(mcherry_ribosomal_root_dir, \"mcherry_variants.fasta\")\n",
    "generate_sequences_fasta_file(mcherry_ribosomal_root_dir)\n",
    "generate_summary(results_directory=mcherry_ribosomal_root_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CUB and codon frequencies of Bacillus and E.coli"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bacillus_endogenous_genes_fasta_file_path = r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\Bacillus-subtilis.fasta\"\n",
    "ecoli_endogenous_genes_fasta_file_path =  r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\Escherichia-coli.fasta\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(bacillus_endogenous_genes_fasta_file_path, \"r\") as bacillus_endogenous_genes_fasta_file:\n",
    "    bacillus_endogenous_genes_dict = SeqIO.to_dict(SeqIO.parse(bacillus_endogenous_genes_fasta_file, \"fasta\"))\n",
    "\n",
    "bacillus_codon_frequencies = calculate_codon_frequencies([str(gene.seq) for gene in bacillus_endogenous_genes_dict.values()])\n",
    "bacillus_codon_frequencies = bacillus_codon_frequencies.sort_values(\"codon\")\n",
    "\n",
    "bacillus_codon_frequencies_lamda = calculate_codon_frequencies_lamda([str(gene.seq) for gene in bacillus_endogenous_genes_dict.values()])\n",
    "bacillus_codon_frequencies_lamda = bacillus_codon_frequencies_lamda.sort_values(\"codon\")\n",
    "\n",
    "with open(ecoli_endogenous_genes_fasta_file_path, \"r\") as ecoli_endogenous_genes_fasta_file:\n",
    "    ecoli_endogenous_genes_dict = SeqIO.to_dict(SeqIO.parse(ecoli_endogenous_genes_fasta_file, \"fasta\"))\n",
    "\n",
    "ecoli_codon_frequencies = calculate_codon_frequencies([str(gene.seq) for gene in ecoli_endogenous_genes_dict.values()])\n",
    "ecoli_codon_frequencies = ecoli_codon_frequencies.sort_values(\"codon\")\n",
    "\n",
    "ecoli_codon_frequencies_lamda = calculate_codon_frequencies_lamda([str(gene.seq) for gene in ecoli_endogenous_genes_dict.values()])\n",
    "ecoli_codon_frequencies_lamda = ecoli_codon_frequencies_lamda.sort_values(\"codon\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bacillus_codon_frequencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ecoli_codon_frequencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ecoli_highly_expressed_genes = get_highly_expressed_genes(endogenous_genes_fasta_file_path=ecoli_endogenous_genes_fasta_file_path,\n",
    "                                                          expression_levels_file_path=r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\PA_ecoli.json\",\n",
    "                                                          expression_type=\"protein_abundance\")\n",
    "# ecoli_highly_expressed_genes = get_highly_expressed_genes(endogenous_genes_fasta_file_path=ecoli_endogenous_genes_fasta_file_path,\n",
    "#                                                           expression_levels_file_path=r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\ecoli_mrna_level.csv\",\n",
    "#                                                           expression_type=\"mrna_levels\")\n",
    "ecoli_highly_expressed_genes_codon_frequencies = calculate_codon_frequencies([str(gene[\"cds\"].seq) for gene in ecoli_highly_expressed_genes.values()])\n",
    "ecoli_highly_expressed_genes_codon_frequencies = ecoli_highly_expressed_genes_codon_frequencies.sort_values(\"codon\")\n",
    "\n",
    "ecoli_highly_expressed_genes_codon_frequencies_lamda = calculate_codon_frequencies_lamda([str(gene[\"cds\"].seq) for gene in ecoli_highly_expressed_genes.values()])\n",
    "ecoli_highly_expressed_genes_codon_frequencies_lamda = ecoli_highly_expressed_genes_codon_frequencies_lamda.sort_values(\"codon\")\n",
    "\n",
    "# bacillus_highly_expressed_genes = get_highly_expressed_genes(endogenous_genes_fasta_file_path=bacillus_endogenous_genes_fasta_file_path,\n",
    "#                                                           expression_levels_file_path=r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\bacillus_mrna_level.csv\",\n",
    "#                                                             expression_type=\"mrna_levels\")\n",
    "bacillus_highly_expressed_genes = get_highly_expressed_genes(endogenous_genes_fasta_file_path=bacillus_endogenous_genes_fasta_file_path,\n",
    "                                                             expression_levels_file_path=r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\PA_bacillus.json\",\n",
    "                                                            expression_type=\"protein_abundance\")\n",
    "bacillus_highly_expressed_genes_codon_frequencies = calculate_codon_frequencies([str(gene[\"cds\"].seq) for gene in bacillus_highly_expressed_genes.values()])\n",
    "bacillus_highly_expressed_genes_codon_frequencies = bacillus_highly_expressed_genes_codon_frequencies.sort_values(\"codon\")\n",
    "\n",
    "bacillus_highly_expressed_genes_codon_frequencies_lamda = calculate_codon_frequencies_lamda([str(gene[\"cds\"].seq) for gene in bacillus_highly_expressed_genes.values()])\n",
    "bacillus_highly_expressed_genes_codon_frequencies_lamda = bacillus_highly_expressed_genes_codon_frequencies_lamda.sort_values(\"codon\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run only once - for generating both tai and cai weights dict:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from modules.user_IO.user_input import UserInputModule\n",
    "from modules.run_summary import RunSummary\n",
    "\n",
    "from generate_summary_file import generate_summary\n",
    "from input_testing_data.generate_input_testing_data_for_modules import generate_testing_data_for_ecoli_and_bacillus\n",
    "from modules.main import run_modules\n",
    "from modules import shared_functions_and_vars\n",
    "\n",
    "default_user_inp_raw = generate_testing_data_for_ecoli_and_bacillus(\n",
    "    optimization_method=\"single_codon_diff\",\n",
    "    optimization_cub_index=\"max_CAI_tAI\",\n",
    "    clusters_count=1,\n",
    "    tuning_param=0.5,\n",
    "    is_ecoli_optimized=True,\n",
    "    sequence=original_mcherry_seq,\n",
    "    output_path=\"results\",\n",
    "    should_use_mrna_levels=False,\n",
    "    should_use_protein_abundance=True,\n",
    ")\n",
    "\n",
    "UserInputModule.run_module(default_user_inp_raw, RunSummary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ecoli_cai_weights, ecoli_tai_weights = get_cub_weights(r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\Escherichia-coli_True_parsed.json\")\n",
    "bacillus_cai_weights, bacillus_tai_weights = get_cub_weights(r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\Bacillus-subtilis_False_parsed.json\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import codonbias as cb\n",
    "\n",
    "# # cai = cb.scores.CodonAdaptationIndex(ignore_stop=False, ref_seq=[str(gene[\"cds\"].seq) for gene in ecoli_highly_expressed_genes.values()], k_mer=1)\n",
    "# ecoli_tai_weights = cb.scores.TrnaAdaptationIndex(url = \"http://gtrnadb.ucsc.edu/genomes/bacteria/Esch_coli_K_12_MG1655/\").weights\n",
    "# bacillus_tai_weights = cb.scores.TrnaAdaptationIndex(url = \"http://gtrnadb.ucsc.edu/genomes/bacteria/Baci_subt_subtilis_168/\").weights"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "codons = ecoli_cai_weights.keys()\n",
    "weights_df_dict = {\n",
    "    \"codon\": codons, \n",
    "    \"amino_acid\": [shared_functions_and_vars.nt_to_aa[k] for k in codons],\n",
    "    \"ecoli_cai_weights\": [ecoli_cai_weights[k] for k in codons],\n",
    "    \"ecoli_tai_weights\": [ecoli_tai_weights.get(k, 0) for k in codons],\n",
    "    \"bacillus_cai_weights\": [bacillus_cai_weights[k] for k in codons],\n",
    "    \"bacillus_tai_weights\": [bacillus_tai_weights.get(k, 0) for k in codons],\n",
    "}\n",
    "weights_dataframe = pd.DataFrame(weights_df_dict)\n",
    "weights_dataframe.sort_values(\"amino_acid\")\n",
    "weights_dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(nrows=1,ncols=2, figsize=(20,5))\n",
    "scatter = weights_dataframe.plot.scatter(x=\"ecoli_cai_weights\", y=\"bacillus_cai_weights\", title=\"E.coli CAI weights vs. Bacillus CAI weights\", ax=axes[0])\n",
    "r, p = stats.spearmanr(weights_dataframe[\"ecoli_cai_weights\"], weights_dataframe[\"bacillus_cai_weights\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = weights_dataframe.plot.scatter(x=\"ecoli_tai_weights\", y=\"bacillus_tai_weights\", title=\"E.coli tAI weights vs. Bacillus tAI weights\", ax=axes[1])\n",
    "r, p = stats.spearmanr(weights_dataframe[\"ecoli_tai_weights\"], weights_dataframe[\"bacillus_tai_weights\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_weights_codon_frequencies_df = pd.merge(weights_dataframe, ecoli_codon_frequencies, on='codon', how='inner')\n",
    "merged_weights_codon_frequencies_df = pd.merge(merged_weights_codon_frequencies_df, bacillus_codon_frequencies, on='codon', how='inner', suffixes=(\"_ecoli\", \"_bacillus\"))\n",
    "\n",
    "fig, axes = plt.subplots(nrows=2,ncols=2, figsize=(20,15))\n",
    "scatter = merged_weights_codon_frequencies_df.plot.scatter(x=\"ecoli_cai_weights\", y=\"frequency_ecoli\", title=\"E.coli CAI weights vs. E.coli codon frequencies\", ax=axes[0,0])\n",
    "r, p = stats.spearmanr(merged_weights_codon_frequencies_df[\"ecoli_cai_weights\"], merged_weights_codon_frequencies_df[\"frequency_ecoli\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_weights_codon_frequencies_df.plot.scatter(x=\"bacillus_cai_weights\", y=\"frequency_bacillus\", title=\"Bacillus CAI weights vs. Bacillus codon frequencies\", ax=axes[0,1])\n",
    "r, p = stats.spearmanr(merged_weights_codon_frequencies_df[\"ecoli_cai_weights\"], merged_weights_codon_frequencies_df[\"frequency_bacillus\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_weights_codon_frequencies_df.plot.scatter(x=\"ecoli_tai_weights\", y=\"frequency_ecoli\", title=\"E.coli tAI weights vs. E.coli codon frequencies\", ax=axes[1,0])\n",
    "r, p = stats.spearmanr(merged_weights_codon_frequencies_df[\"ecoli_tai_weights\"], merged_weights_codon_frequencies_df[\"frequency_ecoli\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_weights_codon_frequencies_df.plot.scatter(x=\"bacillus_tai_weights\", y=\"frequency_bacillus\", title=\"Bacillus tAI weights vs. Bacillus codon frequencies\", ax=axes[1,1])\n",
    "r, p = stats.spearmanr(merged_weights_codon_frequencies_df[\"ecoli_tai_weights\"], merged_weights_codon_frequencies_df[\"frequency_bacillus\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_weights_codon_frequencies_df = pd.merge(weights_dataframe, ecoli_highly_expressed_genes_codon_frequencies, on='codon', how='inner')\n",
    "merged_weights_codon_frequencies_df = pd.merge(merged_weights_codon_frequencies_df, bacillus_highly_expressed_genes_codon_frequencies, on='codon', how='inner', suffixes=(\"_ecoli\", \"_bacillus\"))\n",
    "\n",
    "fig, axes = plt.subplots(nrows=2,ncols=2, figsize=(20,15))\n",
    "scatter = merged_weights_codon_frequencies_df.plot.scatter(x=\"ecoli_cai_weights\", y=\"frequency_ecoli\", title=\"E.coli CAI weights vs. E.coli codon frequencies in highly expressed genes\", ax=axes[0,0])\n",
    "r, p = stats.spearmanr(merged_weights_codon_frequencies_df[\"ecoli_cai_weights\"], merged_weights_codon_frequencies_df[\"frequency_ecoli\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_weights_codon_frequencies_df.plot.scatter(x=\"bacillus_cai_weights\", y=\"frequency_bacillus\", title=\"Bacillus CAI weights vs. Bacillus codon frequencies in highly expressed genes\", ax=axes[0,1])\n",
    "r, p = stats.spearmanr(merged_weights_codon_frequencies_df[\"ecoli_cai_weights\"], merged_weights_codon_frequencies_df[\"frequency_bacillus\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_weights_codon_frequencies_df.plot.scatter(x=\"ecoli_tai_weights\", y=\"frequency_ecoli\", title=\"E.coli tAI weights vs. E.coli codon frequencies in highly expressed genes\", ax=axes[1,0])\n",
    "r, p = stats.spearmanr(merged_weights_codon_frequencies_df[\"ecoli_tai_weights\"], merged_weights_codon_frequencies_df[\"frequency_ecoli\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_weights_codon_frequencies_df.plot.scatter(x=\"bacillus_tai_weights\", y=\"frequency_bacillus\", title=\"Bacillus tAI weights vs. Bacillus codon frequencies in highly expressed genes\", ax=axes[1,1])\n",
    "r, p = stats.spearmanr(merged_weights_codon_frequencies_df[\"ecoli_tai_weights\"], merged_weights_codon_frequencies_df[\"frequency_bacillus\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ecoli_cai_scores, ecoli_tai_scores = get_cub_scores(r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\Escherichia-coli_True_parsed.json\")\n",
    "bacillus_cai_scores, bacillus_tai_scores = get_cub_scores(r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\Bacillus-subtilis_False_parsed.json\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_scores(estimated_expression_dict, cai_scores_dict, tai_scores_dict, expression_type, selected_genes=None):\n",
    "    all_cds = cai_scores_dict.keys()\n",
    "    selected_genes = selected_genes or all_cds\n",
    "    filtered_cds = []\n",
    "    expression_levels = []\n",
    "    \n",
    "    expression_levels_key = \"mrna_level\" if expression_type == \"mrna_levels\" else \"pa\"\n",
    "    for k in all_cds:\n",
    "        if not any(k.startswith(x) for x in selected_genes):\n",
    "            continue\n",
    "        short_name = k.split(\" \")[0]\n",
    "        if short_name in estimated_expression_dict:\n",
    "            expression_levels.append(estimated_expression_dict[short_name][expression_levels_key])\n",
    "            filtered_cds.append(k)\n",
    "\n",
    "    scores_expression_levels_df_dict = {\n",
    "        \"gene\": filtered_cds, \n",
    "        \"cai_scores\": [cai_scores_dict[k] for k in filtered_cds],\n",
    "        \"tai_scores\": [tai_scores_dict[k] for k in filtered_cds],\n",
    "        expression_levels_key: expression_levels,\n",
    "    }\n",
    "    return pd.DataFrame(scores_expression_levels_df_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_type = \"mrna_levels\"\n",
    "ecoli_estimated_expression = get_estimated_expression(endogenous_genes_fasta_file_path=ecoli_endogenous_genes_fasta_file_path,\n",
    "                                                      expression_levels_file_path=r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\ecoli_mrna_level.csv\",\n",
    "                                                      expression_type=expression_type)\n",
    "ecoli_scores_dataframe =  get_scores(estimated_expression_dict=ecoli_estimated_expression, cai_scores_dict=ecoli_cai_scores, tai_scores_dict=ecoli_tai_scores, expression_type=expression_type)\n",
    "print(f\"Ecoli mrna levels stats:\\n records:{len(ecoli_scores_dataframe['mrna_level'])}\\n min: {ecoli_scores_dataframe['mrna_level'].min()} \\n max: {ecoli_scores_dataframe['mrna_level'].max()}\\n mean: {ecoli_scores_dataframe['mrna_level'].mean()}\\n std: {ecoli_scores_dataframe['mrna_level'].std()}\")  \n",
    "ecoli_scores_dataframe[\"mrna_level\"].sort_values(ascending=False).reset_index(drop=True).plot()\n",
    "\n",
    "bacillus_estimated_expression = get_estimated_expression(endogenous_genes_fasta_file_path=bacillus_endogenous_genes_fasta_file_path,\n",
    "                                                         expression_levels_file_path=r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\bacillus_mrna_level.csv\",\n",
    "                                                          expression_type=expression_type)\n",
    "bacillus_scores_dataframe =  get_scores(estimated_expression_dict=bacillus_estimated_expression, cai_scores_dict=bacillus_cai_scores, tai_scores_dict=bacillus_tai_scores, expression_type=expression_type)\n",
    "print(f\"Bacillus mrna levels stats:\\n records:{len(bacillus_scores_dataframe['mrna_level'])}\\n min: {bacillus_scores_dataframe['mrna_level'].min()} \\n max: {bacillus_scores_dataframe['mrna_level'].max()}\\n mean: {bacillus_scores_dataframe['mrna_level'].mean()}\\n std: {bacillus_scores_dataframe['mrna_level'].std()}\")  \n",
    "bacillus_scores_dataframe[\"mrna_level\"].sort_values(ascending=False).reset_index(drop=True).plot()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(nrows=2,ncols=2, figsize=(20,15))\n",
    "\n",
    "scatter = ecoli_scores_dataframe.plot.scatter(x=\"cai_scores\", y=\"mrna_level\", title=\"E.coli mRNA levels vs. E.coli CAI scores\", ax=axes[0,0])\n",
    "r, p = stats.spearmanr(ecoli_scores_dataframe[\"cai_scores\"], ecoli_scores_dataframe[\"mrna_level\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = ecoli_scores_dataframe.plot.scatter(x=\"tai_scores\", y=\"mrna_level\", title=\"E.coli mRNA levels vs. E.coli tAI scores\", ax=axes[0,1])\n",
    "r, p = stats.spearmanr(ecoli_scores_dataframe[\"tai_scores\"], ecoli_scores_dataframe[\"mrna_level\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = bacillus_scores_dataframe.plot.scatter(x=\"cai_scores\", y=\"mrna_level\", title=\"Baciilus mRNA levels vs. Baciilus CAI scores\", ax=axes[1,0])\n",
    "r, p = stats.spearmanr(bacillus_scores_dataframe[\"cai_scores\"], bacillus_scores_dataframe[\"mrna_level\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = bacillus_scores_dataframe.plot.scatter(x=\"tai_scores\", y=\"mrna_level\", title=\"Baciilus mRNA levels vs. Baciilus tAI scores\", ax=axes[1,1])\n",
    "r, p = stats.spearmanr(bacillus_scores_dataframe[\"tai_scores\"], bacillus_scores_dataframe[\"mrna_level\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_type = \"mrna_levels\"\n",
    "ecoli_estimated_expression = get_estimated_expression(endogenous_genes_fasta_file_path=ecoli_endogenous_genes_fasta_file_path,\n",
    "                                                      expression_levels_file_path=r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\ecoli_mrna_level.csv\",\n",
    "                                                      expression_type=expression_type)\n",
    "ecoli_highly_expressed_scores_dataframe =  get_scores(estimated_expression_dict=ecoli_estimated_expression, cai_scores_dict=ecoli_cai_scores, tai_scores_dict=ecoli_tai_scores, selected_genes=ecoli_highly_expressed_genes.keys(), expression_type=expression_type)\n",
    "print(f\"Ecoli mrna levels stats:\\n records:{len(ecoli_highly_expressed_scores_dataframe['mrna_level'])}\\n min: {ecoli_highly_expressed_scores_dataframe['mrna_level'].min()} \\n max: {ecoli_highly_expressed_scores_dataframe['mrna_level'].max()}\\n mean: {ecoli_highly_expressed_scores_dataframe['mrna_level'].mean()}\\n std: {ecoli_highly_expressed_scores_dataframe['mrna_level'].std()}\")  \n",
    "ecoli_highly_expressed_scores_dataframe[\"mrna_level\"].sort_values(ascending=False).reset_index(drop=True).plot()\n",
    "bacillus_estimated_expression = get_estimated_expression(endogenous_genes_fasta_file_path=bacillus_endogenous_genes_fasta_file_path,\n",
    "                                                      expression_levels_file_path=r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\bacillus_mrna_level.csv\",\n",
    "                                                        expression_type=expression_type)\n",
    "bacillus_highly_expressed_scores_dataframe =  get_scores(estimated_expression_dict=bacillus_estimated_expression, cai_scores_dict=bacillus_cai_scores, tai_scores_dict=bacillus_tai_scores, selected_genes=bacillus_highly_expressed_genes.keys(), expression_type=expression_type)\n",
    "print(f\"Bacillus mrna levels stats:\\n records:{len(bacillus_highly_expressed_scores_dataframe['mrna_level'])}\\n min: {bacillus_highly_expressed_scores_dataframe['mrna_level'].min()} \\n max: {bacillus_highly_expressed_scores_dataframe['mrna_level'].max()}\\n mean: {bacillus_highly_expressed_scores_dataframe['mrna_level'].mean()}\\n std: {bacillus_highly_expressed_scores_dataframe['mrna_level'].std()}\")  \n",
    "bacillus_highly_expressed_scores_dataframe[\"mrna_level\"].sort_values(ascending=False).reset_index(drop=True).plot()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(nrows=2,ncols=2, figsize=(20,15))\n",
    "\n",
    "scatter = ecoli_highly_expressed_scores_dataframe.plot.scatter(x=\"cai_scores\", y=\"mrna_level\", title=\"E.coli mRNA levels (Highly expressed) vs. E.coli CAI scores\", ax=axes[0,0])\n",
    "r, p = stats.spearmanr(ecoli_highly_expressed_scores_dataframe[\"cai_scores\"], ecoli_highly_expressed_scores_dataframe[\"mrna_level\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = ecoli_highly_expressed_scores_dataframe.plot.scatter(x=\"tai_scores\", y=\"mrna_level\", title=\"E.coli mRNA levels (Highly expressed) vs. E.coli tAI scores\", ax=axes[0,1])\n",
    "r, p = stats.spearmanr(ecoli_highly_expressed_scores_dataframe[\"tai_scores\"], ecoli_highly_expressed_scores_dataframe[\"mrna_level\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = bacillus_highly_expressed_scores_dataframe.plot.scatter(x=\"cai_scores\", y=\"mrna_level\", title=\"Baciilus mRNA levels (Highly expressed) vs. Baciilus CAI scores\", ax=axes[1,0])\n",
    "r, p = stats.spearmanr(bacillus_highly_expressed_scores_dataframe[\"cai_scores\"], bacillus_highly_expressed_scores_dataframe[\"mrna_level\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = bacillus_highly_expressed_scores_dataframe.plot.scatter(x=\"tai_scores\", y=\"mrna_level\", title=\"Baciilus mRNA levels (Highly expressed) vs. Baciilus tAI scores\", ax=axes[1,1])\n",
    "r, p = stats.spearmanr(bacillus_highly_expressed_scores_dataframe[\"tai_scores\"], bacillus_highly_expressed_scores_dataframe[\"mrna_level\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_type = \"protein_abundance\"\n",
    "ecoli_estimated_expression_pa = get_estimated_expression(endogenous_genes_fasta_file_path=ecoli_endogenous_genes_fasta_file_path,\n",
    "                                                         expression_levels_file_path=r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\PA_ecoli.json\",\n",
    "                                                         expression_type=expression_type)\n",
    "ecoli_scores_dataframe_pa =  get_scores(estimated_expression_dict=ecoli_estimated_expression_pa, cai_scores_dict=ecoli_cai_scores, tai_scores_dict=ecoli_tai_scores, expression_type=expression_type)\n",
    "print(f\"Ecoli protein abundance stats:\\n records:{len(ecoli_scores_dataframe_pa['pa'])}\\n min: {ecoli_scores_dataframe_pa['pa'].min()} \\n max: {ecoli_scores_dataframe_pa['pa'].max()}\\n mean: {ecoli_scores_dataframe_pa['pa'].mean()}\\n std: {ecoli_scores_dataframe_pa['pa'].std()}\")  \n",
    "ecoli_scores_dataframe_pa[\"pa\"].sort_values(ascending=False).reset_index(drop=True).plot()\n",
    "\n",
    "bacillus_estimated_expression_pa = get_estimated_expression(endogenous_genes_fasta_file_path=bacillus_endogenous_genes_fasta_file_path,\n",
    "                                                         expression_levels_file_path=r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\PA_bacillus.json\",\n",
    "                                                          expression_type=expression_type)\n",
    "bacillus_scores_dataframe_pa =  get_scores(estimated_expression_dict=bacillus_estimated_expression_pa, cai_scores_dict=bacillus_cai_scores, tai_scores_dict=bacillus_tai_scores, expression_type=expression_type)\n",
    "print(f\"Bacillus protein abundance stats:\\n records:{len(bacillus_scores_dataframe_pa['pa'])}\\n min: {bacillus_scores_dataframe_pa['pa'].min()} \\n max: {bacillus_scores_dataframe_pa['pa'].max()}\\n mean: {bacillus_scores_dataframe_pa['pa'].mean()}\\n std: {bacillus_scores_dataframe_pa['pa'].std()}\")  \n",
    "bacillus_scores_dataframe_pa[\"pa\"].sort_values(ascending=False).reset_index(drop=True).plot()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(nrows=2,ncols=2, figsize=(20,15))\n",
    "\n",
    "scatter = ecoli_scores_dataframe_pa.plot.scatter(x=\"cai_scores\", y=\"pa\", title=\"E.coli protein abundance vs. E.coli CAI scores\", ax=axes[0,0])\n",
    "r, p = stats.spearmanr(ecoli_scores_dataframe_pa[\"cai_scores\"], ecoli_scores_dataframe_pa[\"pa\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = ecoli_scores_dataframe_pa.plot.scatter(x=\"tai_scores\", y=\"pa\", title=\"E.coli protein abundance vs. E.coli tAI scores\", ax=axes[0,1])\n",
    "r, p = stats.spearmanr(ecoli_scores_dataframe_pa[\"tai_scores\"], ecoli_scores_dataframe_pa[\"pa\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = bacillus_scores_dataframe_pa.plot.scatter(x=\"cai_scores\", y=\"pa\", title=\"Baciilus protein abundance vs. Baciilus CAI scores\", ax=axes[1,0])\n",
    "r, p = stats.spearmanr(bacillus_scores_dataframe_pa[\"cai_scores\"], bacillus_scores_dataframe_pa[\"pa\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = bacillus_scores_dataframe_pa.plot.scatter(x=\"tai_scores\", y=\"pa\", title=\"Baciilus protein abundance vs. Baciilus tAI scores\", ax=axes[1,1])\n",
    "r, p = stats.spearmanr(bacillus_scores_dataframe_pa[\"tai_scores\"], bacillus_scores_dataframe_pa[\"pa\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_type = \"protein_abundance\"\n",
    "ecoli_estimated_expression_pa = get_estimated_expression(endogenous_genes_fasta_file_path=ecoli_endogenous_genes_fasta_file_path,\n",
    "                                                         expression_levels_file_path=r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\PA_ecoli.json\",\n",
    "                                                         expression_type=expression_type)\n",
    "ecoli_scores_dataframe_pa_highly_expressed =  get_scores(estimated_expression_dict=ecoli_estimated_expression_pa, cai_scores_dict=ecoli_cai_scores, tai_scores_dict=ecoli_tai_scores, expression_type=expression_type, selected_genes=ecoli_highly_expressed_genes.keys())\n",
    "print(f\"Ecoli protein abundance stats:\\n records:{len(ecoli_scores_dataframe_pa_highly_expressed['pa'])}\\n min: {ecoli_scores_dataframe_pa_highly_expressed['pa'].min()} \\n max: {ecoli_scores_dataframe_pa_highly_expressed['pa'].max()}\\n mean: {ecoli_scores_dataframe_pa_highly_expressed['pa'].mean()}\\n std: {ecoli_scores_dataframe_pa_highly_expressed['pa'].std()}\")  \n",
    "ecoli_scores_dataframe_pa_highly_expressed[\"pa\"].sort_values(ascending=False).reset_index(drop=True).plot()\n",
    "\n",
    "bacillus_estimated_expression_pa = get_estimated_expression(endogenous_genes_fasta_file_path=bacillus_endogenous_genes_fasta_file_path,\n",
    "                                                         expression_levels_file_path=r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\PA_bacillus.json\",\n",
    "                                                          expression_type=expression_type)\n",
    "bacillus_scores_dataframe_pa_highly_expressed =  get_scores(estimated_expression_dict=bacillus_estimated_expression_pa, cai_scores_dict=bacillus_cai_scores, tai_scores_dict=bacillus_tai_scores, expression_type=expression_type, selected_genes=bacillus_highly_expressed_genes.keys())\n",
    "print(f\"Bacillus protein abundance stats:\\n records:{len(bacillus_scores_dataframe_pa_highly_expressed['pa'])}\\n min: {bacillus_scores_dataframe_pa_highly_expressed['pa'].min()} \\n max: {bacillus_scores_dataframe_pa_highly_expressed['pa'].max()}\\n mean: {bacillus_scores_dataframe_pa_highly_expressed['pa'].mean()}\\n std: {bacillus_scores_dataframe_pa_highly_expressed['pa'].std()}\")  \n",
    "bacillus_scores_dataframe_pa_highly_expressed[\"pa\"].sort_values(ascending=False).reset_index(drop=True).plot()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(nrows=2,ncols=2, figsize=(20,15))\n",
    "\n",
    "scatter = ecoli_scores_dataframe_pa_highly_expressed.plot.scatter(x=\"cai_scores\", y=\"pa\", title=\"E.coli protein abundance vs. E.coli CAI scores - Highly expressed\", ax=axes[0,0])\n",
    "r, p = stats.spearmanr(ecoli_scores_dataframe_pa_highly_expressed[\"cai_scores\"], ecoli_scores_dataframe_pa_highly_expressed[\"pa\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = ecoli_scores_dataframe_pa_highly_expressed.plot.scatter(x=\"tai_scores\", y=\"pa\", title=\"E.coli protein abundance vs. E.coli tAI scores - Highly expressed\", ax=axes[0,1])\n",
    "r, p = stats.spearmanr(ecoli_scores_dataframe_pa_highly_expressed[\"tai_scores\"], ecoli_scores_dataframe_pa_highly_expressed[\"pa\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = bacillus_scores_dataframe_pa_highly_expressed.plot.scatter(x=\"cai_scores\", y=\"pa\", title=\"Baciilus protein abundance vs. Baciilus CAI scores - Highly expressed \", ax=axes[1,0])\n",
    "r, p = stats.spearmanr(bacillus_scores_dataframe_pa_highly_expressed[\"cai_scores\"], bacillus_scores_dataframe_pa_highly_expressed[\"pa\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = bacillus_scores_dataframe_pa.plot.scatter(x=\"tai_scores\", y=\"pa\", title=\"Baciilus protein abundance vs. Baciilus tAI scores - Highly expressed\", ax=axes[1,1])\n",
    "r, p = stats.spearmanr(bacillus_scores_dataframe_pa_highly_expressed[\"tai_scores\"], bacillus_scores_dataframe_pa_highly_expressed[\"pa\"])\n",
    "scatter.annotate('r = {:.4f}, p={}'.format(r, p), xy=(0.7, 0.9), xycoords='axes fraction')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Verification of variants translation to correct amino acid sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Validating original mcherry sequence:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reference aa sequence is taken from: https://www.addgene.org/browse/sequence_vdb/6496/\n",
    "reference_mcherry_protein_seq  = \"MVSKGEEDNMAIIKEFMRFKVHMEGSVNGHEFEIEGEGEGRPYEGTQTAKLKVTKGGPLPFAWDILSPQFMYGSKAYVKHPADIPDYLKLSFPEGFKWERVMNFEDGGVVTVTQDSSLQDGEFIYKVKLRGTNFPSDGPVMQKKTMGWEASSERMYPEDGALKGEIKQRLKLKDGGHYDAEVKTTYKAKKPVQLPGAYNVNIKLDITSHNEDYTIVEQYERAEGRHSTGGMDELYK*\"\n",
    "formatted_reference_mcherry_protein_seq = reference_mcherry_protein_seq.replace(\"*\", \"_\")\n",
    "\n",
    "translated_original_mcherry_seq = shared_functions_and_vars.translate(original_mcherry_seq)\n",
    "\n",
    "assert translated_original_mcherry_seq == formatted_reference_mcherry_protein_seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Validating generated mcherry variants:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(mcherry_variants_file, \"r\") as mcherry_variants_fasta_file:\n",
    "    mcherry_variants_dict = SeqIO.to_dict(SeqIO.parse(mcherry_variants_fasta_file, \"fasta\"))\n",
    "\n",
    "for variant_name, variant_seq in mcherry_variants_dict.items():\n",
    "    translated_seq = shared_functions_and_vars.translate(variant_seq.seq)\n",
    "    if translated_seq != formatted_reference_mcherry_protein_seq:\n",
    "        raise ValueError(f\"Variant {variant_name} is not coding to correct sequence. \\nReference sequence: {formatted_reference_mcherry_protein_seq} \\nVariant sequence: {translated_seq}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Measurements for assessing variants scores "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Eucledean distance of codon frequencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "original_mcherry_codon_frequencies = calculate_codon_frequencies((str(original_mcherry_seq),))\n",
    "original_mcherry_codon_frequencies = original_mcherry_codon_frequencies.sort_values(\"codon\")\n",
    "\n",
    "original_mcherry_codon_frequencies_lamda = calculate_codon_frequencies_lamda((str(original_mcherry_seq),))\n",
    "original_mcherry_codon_frequencies_lamda = original_mcherry_codon_frequencies_lamda.sort_values(\"codon\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "original_bacillus_dist = calculate_distance_between_frequencies(original_mcherry_codon_frequencies[\"frequency\"].reset_index(drop=True), bacillus_codon_frequencies[\"frequency\"].reset_index(drop=True))\n",
    "original_ecoli_dist = calculate_distance_between_frequencies(original_mcherry_codon_frequencies[\"frequency\"].reset_index(drop=True), ecoli_codon_frequencies[\"frequency\"].reset_index(drop=True))\n",
    "\n",
    "original_bacillus_dist_lamda = calculate_distance_between_frequencies(original_mcherry_codon_frequencies_lamda[\"frequency\"].reset_index(drop=True), bacillus_codon_frequencies_lamda[\"frequency\"].reset_index(drop=True))\n",
    "original_ecoli_dist_lamda = calculate_distance_between_frequencies(original_mcherry_codon_frequencies_lamda[\"frequency\"].reset_index(drop=True), ecoli_codon_frequencies_lamda[\"frequency\"].reset_index(drop=True))\n",
    "\n",
    "original_bacillus_dist_highly_expressed = calculate_distance_between_frequencies(original_mcherry_codon_frequencies[\"frequency\"].reset_index(drop=True), bacillus_highly_expressed_genes_codon_frequencies[\"frequency\"].reset_index(drop=True))\n",
    "original_ecoli_dist_highly_expressed = calculate_distance_between_frequencies(original_mcherry_codon_frequencies[\"frequency\"].reset_index(drop=True), ecoli_highly_expressed_genes_codon_frequencies[\"frequency\"].reset_index(drop=True))\n",
    "\n",
    "original_bacillus_dist_highly_expressed_lamda = calculate_distance_between_frequencies(original_mcherry_codon_frequencies_lamda[\"frequency\"].reset_index(drop=True), bacillus_highly_expressed_genes_codon_frequencies_lamda[\"frequency\"].reset_index(drop=True))\n",
    "original_ecoli_dist_highly_expressed_lamda = calculate_distance_between_frequencies(original_mcherry_codon_frequencies_lamda[\"frequency\"].reset_index(drop=True), ecoli_highly_expressed_genes_codon_frequencies_lamda[\"frequency\"].reset_index(drop=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_dict = get_variants_df_dict(mcherry_variants_file)\n",
    "\n",
    "ecoli_opt_df_dict = {key: value for key, value in df_dict.items() if not key.endswith(\"for_optimized_bacillus\")}\n",
    "ecoli_opt_variants_dataframe = pd.DataFrame(ecoli_opt_df_dict)\n",
    "\n",
    "bacillus_opt_df_dict = {key: value for key, value in df_dict.items() if not key.endswith(\"for_optimized_ecoli\")}\n",
    "bacillus_opt_variants_dataframe = pd.DataFrame(bacillus_opt_df_dict)\n",
    "\n",
    "df_dict_highly_expressed = get_variants_df_dict_highly_expressed(mcherry_variants_file)\n",
    "\n",
    "ecoli_opt_df_dict_highly_expressed = {key: value for key, value in df_dict_highly_expressed.items() if not key.endswith(\"for_optimized_bacillus\")}\n",
    "ecoli_opt_variants_dataframe_highly_expressed = pd.DataFrame(ecoli_opt_df_dict_highly_expressed)\n",
    "\n",
    "bacillus_opt_df_dict_highly_expressed = {key: value for key, value in df_dict_highly_expressed.items() if not key.endswith(\"for_optimized_ecoli\")}\n",
    "bacillus_opt_variants_dataframe_highly_expressed = pd.DataFrame(bacillus_opt_df_dict_highly_expressed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bacillus_A = bacillus_codon_frequencies[bacillus_codon_frequencies[\"amino_acid\"] == \"A\"].sort_values(\"codon\")\n",
    "x = pd.DataFrame(get_codon_frequencies_df_dict(mcherry_variants_file))\n",
    "x_A = x[x[\"amino_acid\"] == \"A\"].sort_values(\"codon\")\n",
    "original_A = original_mcherry_codon_frequencies[original_mcherry_codon_frequencies[\"amino_acid\"] == \"A\"].sort_values(\"codon\")\n",
    "\n",
    "\n",
    "original_dist = calculate_distance_between_frequencies(original_A[\"frequency\"].reset_index(drop=True), bacillus_A[\"frequency\"].reset_index(drop=True))\n",
    "variant_dist = calculate_distance_between_frequencies(x_A[\"frequency\"].reset_index(drop=True), bacillus_A[\"frequency\"].reset_index(drop=True))\n",
    "bacillus_A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "original_A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x_A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"original seq dist: {original_dist}, variant dist: {variant_dist}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### CUB scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_dict = get_cub_score_df_dict(mcherry_root_dir)\n",
    "cub_ecoli_opt_df_dict = {key.replace(\"_for_optimized_ecoli\", \"\"): value for key, value in df_dict.items() if not (key.endswith(\"_for_optimized_bacillus\") or key.startswith(\"evaluation_score\"))}\n",
    "cub_ecoli_opt_variants_dataframe = pd.DataFrame(cub_ecoli_opt_df_dict)\n",
    "\n",
    "evaluation_ecoli_opt_df_dict = {key: value for key, value in df_dict.items() if key in [\"variant\", \"evaluation_score_for_optimized_ecoli\"]}\n",
    "evaluation_ecoli_opt_variants_dataframe = pd.DataFrame(evaluation_ecoli_opt_df_dict)\n",
    "\n",
    "cub_bacillus_opt_df_dict = {key.replace(\"_for_optimized_bacillus\", \"\"): value for key, value in df_dict.items() if not (key.endswith(\"for_optimized_ecoli\") or key.startswith(\"evaluation_score\"))}\n",
    "cub_bacillus_opt_variants_dataframe = pd.DataFrame(cub_bacillus_opt_df_dict)\n",
    "\n",
    "evaluation_bacillus_opt_df_dict = {key: value for key, value in df_dict.items() if key in [\"variant\", \"evaluation_score_for_optimized_bacillus\"]}\n",
    "evaluation_bacillus_opt_variants_dataframe = pd.DataFrame(evaluation_bacillus_opt_df_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### tuning parameter = 1 - Eucledean distance of codon frequencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "greater_alpha_df_dict = get_variants_df_dict(mcherry_greater_alpha_variants_file)\n",
    "    \n",
    "greater_alpha_ecoli_opt_df_dict = {key: value for key, value in greater_alpha_df_dict.items() if not key.endswith(\"for_optimized_bacillus\")}\n",
    "greater_alpha_ecoli_opt_variants_dataframe = pd.DataFrame(greater_alpha_ecoli_opt_df_dict)\n",
    "\n",
    "greater_alpha_bacillus_opt_df_dict = {key: value for key, value in greater_alpha_df_dict.items() if not key.endswith(\"for_optimized_ecoli\")}\n",
    "greater_alpha_bacillus_opt_variants_dataframe = pd.DataFrame(greater_alpha_bacillus_opt_df_dict)\n",
    "\n",
    "greater_alpha_df_dict_highly_expressed = get_variants_df_dict_highly_expressed(mcherry_greater_alpha_variants_file)\n",
    "\n",
    "greater_alpha_ecoli_opt_df_dict_highly_expressed = {key: value for key, value in greater_alpha_df_dict_highly_expressed.items() if not key.endswith(\"for_optimized_bacillus\")}\n",
    "greater_alpha_ecoli_opt_variants_dataframe_highly_expressed = pd.DataFrame(greater_alpha_ecoli_opt_df_dict_highly_expressed)\n",
    "\n",
    "greater_alpha_bacillus_opt_df_dict_highly_expressed = {key: value for key, value in greater_alpha_df_dict_highly_expressed.items() if not key.endswith(\"for_optimized_ecoli\")}\n",
    "greater_alpha_bacillus_opt_variants_dataframe_highly_expressed = pd.DataFrame(greater_alpha_bacillus_opt_df_dict_highly_expressed)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### tuning parameter = 1 - CUB scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "greater_alpha_df_dict = get_cub_score_df_dict(mcherry_greater_alpha_root_dir)\n",
    "cub_ecoli_opt_greater_alpha_df_dict = {key.replace(\"_for_optimized_ecoli\", \"\"): value for key, value in greater_alpha_df_dict.items() if not (key.endswith(\"_for_optimized_bacillus\") or key.startswith(\"evaluation_score\"))}\n",
    "cub_ecoli_opt_greater_alpha_variants_dataframe = pd.DataFrame(cub_ecoli_opt_greater_alpha_df_dict)\n",
    "\n",
    "evaluation_ecoli_opt_greate_alpha_df_dict = {key: value for key, value in greater_alpha_df_dict.items() if key in [\"variant\", \"evaluation_score_for_optimized_ecoli\"]}\n",
    "evaluation_ecoli_opt_greater_alpha_variants_dataframe = pd.DataFrame(evaluation_ecoli_opt_greate_alpha_df_dict)\n",
    "\n",
    "cub_bacillus_opt_greater_alpha_df_dict = {key.replace(\"_for_optimized_bacillus\", \"\"): value for key, value in greater_alpha_df_dict.items() if not (key.endswith(\"for_optimized_ecoli\") or key.startswith(\"evaluation_score\"))}\n",
    "cub_bacillus_opt_greater_alpha_variants_dataframe = pd.DataFrame(cub_bacillus_opt_greater_alpha_df_dict)\n",
    "\n",
    "evaluation_bacillus_opt_greater_alpha_df_dict = {key: value for key, value in greater_alpha_df_dict.items() if key in [\"variant\", \"evaluation_score_for_optimized_bacillus\"]}\n",
    "evaluation_bacillus_opt_greater_alpha_variants_dataframe = pd.DataFrame(evaluation_bacillus_opt_greater_alpha_df_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### tuning parameter = 0 - Eucledean distance of codon frequencies & CUB scores "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lower_alpha_df_dict = get_variants_df_dict(mcherry_lower_alpha_variants_file)\n",
    "    \n",
    "lower_alpha_ecoli_opt_df_dict = {key: value for key, value in lower_alpha_df_dict.items() if not key.endswith(\"for_optimized_bacillus\")}\n",
    "lower_alpha_ecoli_opt_variants_dataframe = pd.DataFrame(lower_alpha_ecoli_opt_df_dict)\n",
    "\n",
    "lower_alpha_bacillus_opt_df_dict = {key: value for key, value in lower_alpha_df_dict.items() if not key.endswith(\"for_optimized_ecoli\")}\n",
    "lower_alpha_bacillus_opt_variants_dataframe = pd.DataFrame(lower_alpha_bacillus_opt_df_dict)\n",
    "\n",
    "lower_alpha_df_dict_highly_expressed = get_variants_df_dict_highly_expressed(mcherry_lower_alpha_variants_file)\n",
    "\n",
    "lower_alpha_ecoli_opt_df_dict_highly_expressed = {key: value for key, value in lower_alpha_df_dict_highly_expressed.items() if not key.endswith(\"for_optimized_bacillus\")}\n",
    "lower_alpha_ecoli_opt_variants_dataframe_highly_expressed = pd.DataFrame(lower_alpha_ecoli_opt_df_dict_highly_expressed)\n",
    "\n",
    "lower_alpha_bacillus_opt_df_dict_highly_expressed = {key: value for key, value in lower_alpha_df_dict_highly_expressed.items() if not key.endswith(\"for_optimized_ecoli\")}\n",
    "lower_alpha_bacillus_opt_variants_dataframe_highly_expressed = pd.DataFrame(lower_alpha_bacillus_opt_df_dict_highly_expressed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lower_alpha_df_dict = get_cub_score_df_dict(mcherry_lower_alpha_root_dir)\n",
    "cub_ecoli_opt_lower_alpha_df_dict = {key.replace(\"_for_optimized_ecoli\", \"\"): value for key, value in lower_alpha_df_dict.items() if not (key.endswith(\"_for_optimized_bacillus\") or key.startswith(\"evaluation_score\"))}\n",
    "cub_ecoli_opt_lower_alpha_variants_dataframe = pd.DataFrame(cub_ecoli_opt_lower_alpha_df_dict)\n",
    "\n",
    "evaluation_ecoli_opt_lower_alpha_df_dict = {key: value for key, value in lower_alpha_df_dict.items() if key in [\"variant\", \"evaluation_score_for_optimized_ecoli\"]}\n",
    "evaluation_ecoli_opt_lower_alpha_variants_dataframe = pd.DataFrame(evaluation_ecoli_opt_lower_alpha_df_dict)\n",
    "\n",
    "cub_bacillus_opt_lower_alpha_df_dict = {key.replace(\"_for_optimized_bacillus\", \"\"): value for key, value in lower_alpha_df_dict.items() if not (key.endswith(\"for_optimized_ecoli\") or key.startswith(\"evaluation_score\"))}\n",
    "cub_bacillus_opt_lower_alpha_variants_dataframe = pd.DataFrame(cub_bacillus_opt_lower_alpha_df_dict)\n",
    "\n",
    "evaluation_bacillus_opt_lower_alpha_df_dict = {key: value for key, value in lower_alpha_df_dict.items() if key in [\"variant\", \"evaluation_score_for_optimized_bacillus\"]}\n",
    "evaluation_bacillus_opt_lower_alpha_variants_dataframe = pd.DataFrame(evaluation_bacillus_opt_lower_alpha_df_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Comparative analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(nrows=1,ncols=2, figsize=(27,10))\n",
    "\n",
    "fig.suptitle(\"Variants Evaluation Scores\")\n",
    "evaluation_ecoli_opt_variants_dataframe.plot(\"variant\", kind=\"barh\", ax=axes[0], title=\"E.coli optimized\")\n",
    "evaluation_bacillus_opt_variants_dataframe.plot(\"variant\", kind=\"barh\", ax=axes[1], title=\"Bacillus optimized\", color=\"#dc4d01\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df_ecoli = pd.merge(evaluation_ecoli_opt_variants_dataframe, ecoli_opt_variants_dataframe, on='variant', how='inner')\n",
    "merged_df_ecoli['ecoli_freq_dist_diff'] =  merged_df_ecoli['ecoli_dist_for_optimized_ecoli'] - merged_df_ecoli['ecoli_dist_for_original_mcherry']\n",
    "merged_df_ecoli['bacillus_freq_dist_diff'] = merged_df_ecoli['bacillus_dist_for_optimized_ecoli'] - merged_df_ecoli['bacillus_dist_for_original_mcherry']\n",
    "\n",
    "merged_df_bacillus = pd.merge(evaluation_bacillus_opt_variants_dataframe, bacillus_opt_variants_dataframe, on='variant', how='inner')\n",
    "merged_df_bacillus['ecoli_freq_dist_diff'] =  merged_df_bacillus['ecoli_dist_for_optimized_bacillus'] - merged_df_bacillus['ecoli_dist_for_original_mcherry']\n",
    "merged_df_bacillus['bacillus_freq_dist_diff'] = merged_df_bacillus['bacillus_dist_for_optimized_bacillus'] - merged_df_bacillus['bacillus_dist_for_original_mcherry'] \n",
    "\n",
    "\n",
    "fig, axes = plt.subplots(nrows=2,ncols=2, figsize=(20,15))\n",
    "fig.suptitle(\"Ecaluation Score vs. Codon Frequencies distance\")\n",
    "\n",
    "scatter = merged_df_ecoli.plot.scatter(x=\"evaluation_score_for_optimized_ecoli\", y=\"ecoli_freq_dist_diff\", ax=axes[0,0], title=\"E.coli optimized - Evaluation score vs.\\nE.coli codon frequencies diff (modified seq dist-original seq dist)\")\n",
    "r, p = stats.pearsonr(merged_df_ecoli[\"evaluation_score_for_optimized_ecoli\"], merged_df_ecoli[\"ecoli_freq_dist_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_ecoli.plot.scatter(x=\"evaluation_score_for_optimized_ecoli\", y=\"bacillus_freq_dist_diff\", ax=axes[0,1], title=\"E.coli optimized - Evaluation score vs.\\nBacillus codon frequencies diff (modified seq dist-original seq dist)\")\n",
    "r, p = stats.pearsonr(merged_df_ecoli[\"evaluation_score_for_optimized_ecoli\"], merged_df_ecoli[\"bacillus_freq_dist_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_bacillus.plot.scatter(x=\"evaluation_score_for_optimized_bacillus\", y=\"ecoli_freq_dist_diff\", ax=axes[1,0], title=\"Bacillus optimized - Evaluation score vs.\\nE.coli codon frequencies diff (modified seq dist-original seq dist)\")\n",
    "r, p = stats.pearsonr(merged_df_bacillus[\"evaluation_score_for_optimized_bacillus\"], merged_df_bacillus[\"ecoli_freq_dist_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_bacillus.plot.scatter(x=\"evaluation_score_for_optimized_bacillus\", y=\"bacillus_freq_dist_diff\", ax=axes[1,1], title=\"Bacillus optimized - Evaluation score vs.\\nBacillus codon frequencies diff (modified seq dist-original seq dist)\")\n",
    "r, p = stats.pearsonr(merged_df_bacillus[\"evaluation_score_for_optimized_bacillus\"], merged_df_bacillus[\"bacillus_freq_dist_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df_ecoli = pd.merge(evaluation_ecoli_opt_greater_alpha_variants_dataframe, greater_alpha_ecoli_opt_variants_dataframe, on='variant', how='inner')\n",
    "merged_df_ecoli['ecoli_freq_dist_diff'] =  merged_df_ecoli['ecoli_dist_for_optimized_ecoli'] - merged_df_ecoli['ecoli_dist_for_original_mcherry']\n",
    "merged_df_ecoli['bacillus_freq_dist_diff'] = merged_df_ecoli['bacillus_dist_for_optimized_ecoli'] - merged_df_ecoli['bacillus_dist_for_original_mcherry']\n",
    "\n",
    "merged_df_bacillus = pd.merge(evaluation_bacillus_opt_greater_alpha_variants_dataframe, greater_alpha_bacillus_opt_variants_dataframe, on='variant', how='inner')\n",
    "merged_df_bacillus['ecoli_freq_dist_diff'] =  merged_df_bacillus['ecoli_dist_for_optimized_bacillus'] - merged_df_bacillus['ecoli_dist_for_original_mcherry']\n",
    "merged_df_bacillus['bacillus_freq_dist_diff'] = merged_df_bacillus['bacillus_dist_for_optimized_bacillus'] - merged_df_bacillus['bacillus_dist_for_original_mcherry'] \n",
    "\n",
    "\n",
    "fig, axes = plt.subplots(nrows=2,ncols=2, figsize=(20,13))\n",
    "fig.suptitle(\"Ecaluation Score vs. Codon Frequencies distance (alpha=1)\")\n",
    "\n",
    "scatter = merged_df_ecoli.plot.scatter(x=\"evaluation_score_for_optimized_ecoli\", y=\"ecoli_freq_dist_diff\", ax=axes[0,0], title=\"E.coli optimized - Evaluation score vs.\\nE.coli codon frequencies diff (modified seq dist-original seq dist)\")\n",
    "r, p = stats.pearsonr(merged_df_ecoli[\"evaluation_score_for_optimized_ecoli\"], merged_df_ecoli[\"ecoli_freq_dist_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_ecoli.plot.scatter(x=\"evaluation_score_for_optimized_ecoli\", y=\"bacillus_freq_dist_diff\", ax=axes[0,1], title=\"E.coli optimized - Evaluation score vs.\\nBacillus codon frequencies diff (modified seq dist-original seq dist)\")\n",
    "r, p = stats.pearsonr(merged_df_ecoli[\"evaluation_score_for_optimized_ecoli\"], merged_df_ecoli[\"bacillus_freq_dist_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_bacillus.plot.scatter(x=\"evaluation_score_for_optimized_bacillus\", y=\"ecoli_freq_dist_diff\", ax=axes[1,0], title=\"Bacillus optimized - Evaluation score vs.\\nE.coli codon frequencies diff (modified seq dist-original seq dist)\")\n",
    "r, p = stats.pearsonr(merged_df_bacillus[\"evaluation_score_for_optimized_bacillus\"], merged_df_bacillus[\"ecoli_freq_dist_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_bacillus.plot.scatter(x=\"evaluation_score_for_optimized_bacillus\", y=\"bacillus_freq_dist_diff\", ax=axes[1,1], title=\"Bacillus optimized - Evaluation score vs.\\nBacillus codon frequencies diff (modified seq dist-original seq dist)\")\n",
    "r, p = stats.pearsonr(merged_df_bacillus[\"evaluation_score_for_optimized_bacillus\"], merged_df_bacillus[\"bacillus_freq_dist_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df_ecoli = pd.merge(evaluation_ecoli_opt_lower_alpha_variants_dataframe, lower_alpha_ecoli_opt_variants_dataframe, on='variant', how='inner')\n",
    "merged_df_ecoli['ecoli_freq_dist_diff'] =  merged_df_ecoli['ecoli_dist_for_optimized_ecoli'] - merged_df_ecoli['ecoli_dist_for_original_mcherry']\n",
    "merged_df_ecoli['bacillus_freq_dist_diff'] = merged_df_ecoli['bacillus_dist_for_optimized_ecoli'] - merged_df_ecoli['bacillus_dist_for_original_mcherry']\n",
    "\n",
    "merged_df_bacillus = pd.merge(evaluation_bacillus_opt_lower_alpha_variants_dataframe, lower_alpha_bacillus_opt_variants_dataframe, on='variant', how='inner')\n",
    "merged_df_bacillus['ecoli_freq_dist_diff'] =  merged_df_bacillus['ecoli_dist_for_optimized_bacillus'] - merged_df_bacillus['ecoli_dist_for_original_mcherry']\n",
    "merged_df_bacillus['bacillus_freq_dist_diff'] = merged_df_bacillus['bacillus_dist_for_optimized_bacillus'] - merged_df_bacillus['bacillus_dist_for_original_mcherry'] \n",
    "\n",
    "\n",
    "fig, axes = plt.subplots(nrows=2,ncols=2, figsize=(20,13))\n",
    "fig.suptitle(\"Ecaluation Score vs. Codon Frequencies distance (alpha=0)\")\n",
    "\n",
    "scatter = merged_df_ecoli.plot.scatter(x=\"evaluation_score_for_optimized_ecoli\", y=\"ecoli_freq_dist_diff\", ax=axes[0,0], title=\"Bacillus deoptimized - Evaluation score vs.\\nE.coli codon frequencies diff (modified seq dist-original seq dist)\")\n",
    "r, p = stats.pearsonr(merged_df_ecoli[\"evaluation_score_for_optimized_ecoli\"], merged_df_ecoli[\"ecoli_freq_dist_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_ecoli.plot.scatter(x=\"evaluation_score_for_optimized_ecoli\", y=\"bacillus_freq_dist_diff\", ax=axes[0,1], title=\"Bacillus deoptimized - Evaluation score vs.\\nBacillus codon frequencies diff (modified seq dist-original seq dist)\")\n",
    "r, p = stats.pearsonr(merged_df_ecoli[\"evaluation_score_for_optimized_ecoli\"], merged_df_ecoli[\"bacillus_freq_dist_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_bacillus.plot.scatter(x=\"evaluation_score_for_optimized_bacillus\", y=\"ecoli_freq_dist_diff\", ax=axes[1,0], title=\"E.coli deoptimized - Evaluation score vs.\\nE.coli codon frequencies diff (modified seq dist-original seq dist)\")\n",
    "r, p = stats.pearsonr(merged_df_bacillus[\"evaluation_score_for_optimized_bacillus\"], merged_df_bacillus[\"ecoli_freq_dist_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_bacillus.plot.scatter(x=\"evaluation_score_for_optimized_bacillus\", y=\"bacillus_freq_dist_diff\", ax=axes[1,1], title=\"E.coli deoptimized - Evaluation score vs.\\nBacillus codon frequencies diff (modified seq dist-original seq dist)\")\n",
    "r, p = stats.pearsonr(merged_df_bacillus[\"evaluation_score_for_optimized_bacillus\"], merged_df_bacillus[\"bacillus_freq_dist_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df_ecoli = pd.merge(evaluation_ecoli_opt_variants_dataframe, cub_ecoli_opt_variants_dataframe, on='variant', how='inner')\n",
    "merged_df_ecoli['ecoli_score_diff'] = merged_df_ecoli['ecoli_final_score'] - merged_df_ecoli['ecoli_initial_score']\n",
    "merged_df_ecoli['bacillus_score_diff'] = merged_df_ecoli['bacillus_final_score'] - merged_df_ecoli['bacillus_initial_score']\n",
    "\n",
    "merged_df_bacillus = pd.merge(evaluation_bacillus_opt_variants_dataframe, cub_bacillus_opt_variants_dataframe, on='variant', how='inner')\n",
    "merged_df_bacillus['ecoli_score_diff'] = merged_df_bacillus['ecoli_final_score'] - merged_df_bacillus['ecoli_initial_score']\n",
    "merged_df_bacillus['bacillus_score_diff'] = merged_df_bacillus['bacillus_final_score'] - merged_df_bacillus['bacillus_initial_score']\n",
    "\n",
    "fig, axes = plt.subplots(nrows=2,ncols=2, figsize=(20,15))\n",
    "\n",
    "scatter = merged_df_ecoli.plot.scatter(x=\"evaluation_score_for_optimized_ecoli\", y=\"ecoli_score_diff\", ax=axes[0,0], title=\"E.coli optimized - Evaluation score vs.\\nE.coli CUB score diff (modified seq score-original seq score)\")\n",
    "r, p = stats.spearmanr(merged_df_ecoli[\"evaluation_score_for_optimized_ecoli\"], merged_df_ecoli[\"ecoli_score_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_ecoli.plot.scatter(x=\"evaluation_score_for_optimized_ecoli\", y=\"bacillus_score_diff\", ax=axes[0,1], title=\"E.coli optimized - Evaluation score vs.\\nBacillus CUB score diff (modified seq score-original seq score)\")\n",
    "r, p = stats.spearmanr(merged_df_ecoli[\"evaluation_score_for_optimized_ecoli\"], merged_df_ecoli[\"bacillus_score_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_bacillus.plot.scatter(x=\"evaluation_score_for_optimized_bacillus\", y=\"ecoli_score_diff\", ax=axes[1,0], title=\"Bacillus optimized - Evaluation score vs.\\nE.coli CUB score diff (modified seq score-original seq score)\")\n",
    "r, p = stats.spearmanr(merged_df_bacillus[\"evaluation_score_for_optimized_bacillus\"], merged_df_bacillus[\"ecoli_score_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_bacillus.plot.scatter(x=\"evaluation_score_for_optimized_bacillus\", y=\"bacillus_score_diff\", ax=axes[1,1], title=\"Bacillus optimized - Evaluation score vs.\\nBacillus CUB score diff (modified seq score-original seq score)\")\n",
    "r, p = stats.spearmanr(merged_df_bacillus[\"evaluation_score_for_optimized_bacillus\"], merged_df_bacillus[\"bacillus_score_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df_ecoli_greater_alpha = pd.merge(evaluation_ecoli_opt_greater_alpha_variants_dataframe, cub_ecoli_opt_greater_alpha_variants_dataframe, on='variant', how='inner')\n",
    "merged_df_ecoli_greater_alpha['ecoli_score_diff'] = merged_df_ecoli_greater_alpha['ecoli_final_score'] - merged_df_ecoli_greater_alpha['ecoli_initial_score']\n",
    "merged_df_ecoli_greater_alpha['bacillus_score_diff'] = merged_df_ecoli_greater_alpha['bacillus_final_score'] - merged_df_ecoli_greater_alpha['bacillus_initial_score']\n",
    "\n",
    "merged_df_bacillus_greater_alpha = pd.merge(evaluation_bacillus_opt_greater_alpha_variants_dataframe, cub_bacillus_opt_greater_alpha_variants_dataframe, on='variant', how='inner')\n",
    "merged_df_bacillus_greater_alpha['ecoli_score_diff'] = merged_df_bacillus_greater_alpha['ecoli_final_score'] - merged_df_bacillus_greater_alpha['ecoli_initial_score']\n",
    "merged_df_bacillus_greater_alpha['bacillus_score_diff'] = merged_df_bacillus_greater_alpha['bacillus_final_score'] - merged_df_bacillus_greater_alpha['bacillus_initial_score']\n",
    "\n",
    "fig, axes = plt.subplots(nrows=2,ncols=2, figsize=(20,15))\n",
    "\n",
    "scatter = merged_df_ecoli_greater_alpha.plot.scatter(x=\"evaluation_score_for_optimized_ecoli\", y=\"ecoli_score_diff\", ax=axes[0,0], title=\"E.coli optimized (alpha=1)- Evaluation score vs.\\nE.coli CUB score diff (modified seq score-original seq score\")\n",
    "r, p = stats.spearmanr(merged_df_ecoli_greater_alpha[\"evaluation_score_for_optimized_ecoli\"], merged_df_ecoli_greater_alpha[\"ecoli_score_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_ecoli_greater_alpha.plot.scatter(x=\"evaluation_score_for_optimized_ecoli\", y=\"bacillus_score_diff\", ax=axes[0,1], title=\"E.coli optimized (alpha=1)- Evaluation score vs.\\nBacillus CUB score diff (modified seq score-original seq score\")\n",
    "r, p = stats.spearmanr(merged_df_ecoli_greater_alpha[\"evaluation_score_for_optimized_ecoli\"], merged_df_ecoli_greater_alpha[\"bacillus_score_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_bacillus_greater_alpha.plot.scatter(x=\"evaluation_score_for_optimized_bacillus\", y=\"ecoli_score_diff\", ax=axes[1,0], title=\"Bacillus optimized (alpha=1)- Evaluation score vs.\\nE.coli CUB score diff (modified seq score-original seq score\")\n",
    "r, p = stats.spearmanr(merged_df_bacillus_greater_alpha[\"evaluation_score_for_optimized_bacillus\"], merged_df_bacillus_greater_alpha[\"ecoli_score_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_bacillus_greater_alpha.plot.scatter(x=\"evaluation_score_for_optimized_bacillus\", y=\"bacillus_score_diff\", ax=axes[1,1], title=\"Bacillus optimized (alpha=1)- Evaluation score vs.\\nBacillus CUB score diff (modified seq score-original seq score\")\n",
    "r, p = stats.spearmanr(merged_df_bacillus_greater_alpha[\"evaluation_score_for_optimized_bacillus\"], merged_df_bacillus_greater_alpha[\"bacillus_score_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df_ecoli_lower_alpha = pd.merge(evaluation_ecoli_opt_lower_alpha_variants_dataframe, cub_ecoli_opt_lower_alpha_variants_dataframe, on='variant', how='inner')\n",
    "merged_df_ecoli_lower_alpha['ecoli_score_diff'] = merged_df_ecoli_lower_alpha['ecoli_final_score'] - merged_df_ecoli_lower_alpha['ecoli_initial_score']\n",
    "merged_df_ecoli_lower_alpha['bacillus_score_diff'] = merged_df_ecoli_lower_alpha['bacillus_final_score'] - merged_df_ecoli_lower_alpha['bacillus_initial_score']\n",
    "\n",
    "merged_df_bacillus_lower_alpha = pd.merge(evaluation_bacillus_opt_lower_alpha_variants_dataframe, cub_bacillus_opt_lower_alpha_variants_dataframe, on='variant', how='inner')\n",
    "merged_df_bacillus_lower_alpha['ecoli_score_diff'] = merged_df_bacillus_lower_alpha['ecoli_final_score'] - merged_df_bacillus_lower_alpha['ecoli_initial_score']\n",
    "merged_df_bacillus_lower_alpha['bacillus_score_diff'] = merged_df_bacillus_lower_alpha['bacillus_final_score'] - merged_df_bacillus_lower_alpha['bacillus_initial_score']\n",
    "\n",
    "fig, axes = plt.subplots(nrows=2,ncols=2, figsize=(20,15))\n",
    "\n",
    "scatter = merged_df_ecoli_lower_alpha.plot.scatter(x=\"evaluation_score_for_optimized_ecoli\", y=\"ecoli_score_diff\", ax=axes[0,0], title=\"Bacillus deoptimized (alpha=0)- Evaluation score vs.\\nE.coli CUB score diff (modified seq score-original seq score\")\n",
    "r, p = stats.spearmanr(merged_df_ecoli_lower_alpha[\"evaluation_score_for_optimized_ecoli\"], merged_df_ecoli_lower_alpha[\"ecoli_score_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_ecoli_lower_alpha.plot.scatter(x=\"evaluation_score_for_optimized_ecoli\", y=\"bacillus_score_diff\", ax=axes[0,1], title=\"Bacillus deoptimized (alpha=0)- Evaluation score vs.\\nBacillus CUB score diff (modified seq score-original seq score\")\n",
    "r, p = stats.spearmanr(merged_df_ecoli_lower_alpha[\"evaluation_score_for_optimized_ecoli\"], merged_df_ecoli_lower_alpha[\"bacillus_score_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_bacillus_lower_alpha.plot.scatter(x=\"evaluation_score_for_optimized_bacillus\", y=\"ecoli_score_diff\", ax=axes[1,0], title=\"E.coli deoptimized (alpha=0)- Evaluation score vs.\\nE.coli CUB score diff (modified seq score-original seq score\")\n",
    "r, p = stats.spearmanr(merged_df_bacillus_lower_alpha[\"evaluation_score_for_optimized_bacillus\"], merged_df_bacillus_lower_alpha[\"ecoli_score_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')\n",
    "\n",
    "scatter = merged_df_bacillus_lower_alpha.plot.scatter(x=\"evaluation_score_for_optimized_bacillus\", y=\"bacillus_score_diff\", ax=axes[1,1], title=\"E.coli deoptimized (alpha=0)- Evaluation score vs.\\nBacillus CUB score diff (modified seq score-original seq score\")\n",
    "r, p = stats.spearmanr(merged_df_bacillus_lower_alpha[\"evaluation_score_for_optimized_bacillus\"], merged_df_bacillus_lower_alpha[\"bacillus_score_diff\"])\n",
    "scatter.annotate('r = {:.3f}'.format(r), xy=(0.7, 0.9), xycoords='axes fraction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "my_colors = ['#3b5b92', '#3b5b92', '#dc4d01', '#dc4d01'] \n",
    "my_colors_extended = ['cornflowerblue', 'cornflowerblue', 'navy', 'navy', 'orange', 'orange', 'orangered', 'orangered'] \n",
    "for q in range(cub_ecoli_opt_greater_alpha_variants_dataframe.shape[0]):\n",
    "\n",
    "    fig, axes = plt.subplots(nrows=1,ncols=3, figsize=(20,5))\n",
    "    fig.suptitle(greater_alpha_ecoli_opt_variants_dataframe.iloc[q,0] + \" - codon frequency distance\")\n",
    "    lower_alpha_ecoli_opt_variants_dataframe.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[0], stacked=True, color=my_colors_extended)\n",
    "    axes[0].set_title(\"alpha = 0\")\n",
    "    ecoli_opt_variants_dataframe.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[1], stacked=True, color=my_colors_extended)\n",
    "    axes[1].set_title(\"alpha = 0.5\")\n",
    "    greater_alpha_ecoli_opt_variants_dataframe.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[2], stacked=True, color=my_colors_extended)\n",
    "    axes[2].set_title(\"alpha = 1\")\n",
    "    \n",
    "    fig, axes = plt.subplots(nrows=1,ncols=3, figsize=(20,5))\n",
    "    fig.suptitle(greater_alpha_ecoli_opt_variants_dataframe_highly_expressed.iloc[q,0] + \" - codon frequency distance (highly expressed)\")\n",
    "    lower_alpha_ecoli_opt_variants_dataframe_highly_expressed.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[0], stacked=True, color=my_colors_extended)\n",
    "    axes[0].set_title(\"alpha = 0\")\n",
    "    ecoli_opt_variants_dataframe_highly_expressed.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[1], stacked=True, color=my_colors_extended)\n",
    "    axes[1].set_title(\"alpha = 0.5\")\n",
    "    greater_alpha_ecoli_opt_variants_dataframe_highly_expressed.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[2], stacked=True, color=my_colors_extended)\n",
    "    axes[2].set_title(\"alpha = 1\")\n",
    "\n",
    "\n",
    "    fig, axes = plt.subplots(nrows=1,ncols=3, figsize=(20,5))\n",
    "    fig.suptitle(cub_ecoli_opt_greater_alpha_variants_dataframe.iloc[q,0] + \" - CUB scores\")\n",
    "    cub_ecoli_opt_lower_alpha_variants_dataframe.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[0], stacked=True, color=my_colors)\n",
    "    axes[0].set_title(\"alpha = 0\")\n",
    "    cub_ecoli_opt_variants_dataframe.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[1], stacked=True, color=my_colors)\n",
    "    axes[1].set_title(\"alpha = 0.5\")\n",
    "    cub_ecoli_opt_greater_alpha_variants_dataframe.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[2], stacked=True, color=my_colors)\n",
    "    axes[2].set_title(\"alpha = 1\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Optimizing for Bacillus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "my_colors = ['#3b5b92', '#3b5b92', '#dc4d01', '#dc4d01'] \n",
    "my_colors_extended = ['cornflowerblue', 'cornflowerblue', 'navy', 'navy', 'orange', 'orange', 'orangered', 'orangered'] \n",
    "for q in range(cub_bacillus_opt_greater_alpha_variants_dataframe.shape[0]):\n",
    "\n",
    "    fig, axes = plt.subplots(nrows=1,ncols=3, figsize=(20,5))\n",
    "    fig.suptitle(greater_alpha_bacillus_opt_variants_dataframe.iloc[q,0] + \" - codon frequency distance\")\n",
    "    lower_alpha_bacillus_opt_variants_dataframe.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[0], stacked=True, color=my_colors_extended)\n",
    "    axes[0].set_title(\"alpha = 0\")\n",
    "    bacillus_opt_variants_dataframe.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[1], stacked=True, color=my_colors_extended)\n",
    "    axes[1].set_title(\"alpha = 0.5\")\n",
    "    greater_alpha_bacillus_opt_variants_dataframe.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[2], stacked=True, color=my_colors_extended)\n",
    "    axes[2].set_title(\"alpha = 1\")\n",
    "    \n",
    "    fig, axes = plt.subplots(nrows=1,ncols=3, figsize=(20,5))\n",
    "    fig.suptitle(greater_alpha_bacillus_opt_variants_dataframe_highly_expressed.iloc[q,0] + \" - codon frequency distance (highly expressed)\")\n",
    "    lower_alpha_bacillus_opt_variants_dataframe_highly_expressed.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[0], stacked=True, color=my_colors_extended)\n",
    "    axes[0].set_title(\"alpha = 0\")\n",
    "    bacillus_opt_variants_dataframe_highly_expressed.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[1], stacked=True, color=my_colors_extended)\n",
    "    axes[1].set_title(\"alpha = 0.5\")\n",
    "    greater_alpha_bacillus_opt_variants_dataframe_highly_expressed.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[2], stacked=True, color=my_colors_extended)\n",
    "    axes[2].set_title(\"alpha = 1\")\n",
    "\n",
    "\n",
    "    fig, axes = plt.subplots(nrows=1,ncols=3, figsize=(20,5))\n",
    "    fig.suptitle(cub_bacillus_opt_greater_alpha_variants_dataframe.iloc[q,0] + \" - CUB scores\")\n",
    "    cub_bacillus_opt_lower_alpha_variants_dataframe.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[0], stacked=True, color=my_colors)\n",
    "    axes[0].set_title(\"alpha = 0\")\n",
    "    cub_bacillus_opt_variants_dataframe.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[1], stacked=True, color=my_colors)\n",
    "    axes[1].set_title(\"alpha = 0.5\")\n",
    "    cub_bacillus_opt_greater_alpha_variants_dataframe.iloc[q,1:].plot(x=\"variant\", kind=\"bar\", ax=axes[2], stacked=True, color=my_colors)\n",
    "    axes[2].set_title(\"alpha = 1\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Debugging negative non normalized score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(r\"C:\\projects\\Igem_TAU_2021_moran\\analysis\\example_data\\Bacillus-subtilis.fasta\", \"r\") as fasta_handle:\n",
    "        genome_dict = SeqIO.to_dict(SeqIO.parse(fasta_handle, \"fasta\"))\n",
    "gene_name = \"rpmJ|ribosomal\"\n",
    "\n",
    "gene_sequence = genome_dict[gene_name]\n",
    "gene_sequence = str(gene_sequence.seq)\n",
    "\n",
    "results = run_single_method_for_orf_sequence(\n",
    "    optimization_method=\"single_codon_diff\",\n",
    "    optimization_cub_index=\"CAI\",\n",
    "    is_ecoli_optimized=False,\n",
    "    output_path=\"endogenous\",\n",
    "    orf_sequence=gene_sequence,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results_df_dict = {\"selected_codon\": [], \"amino_acid\": [], \"bacillus_weight\": [], \"ecoli_weight\": [], \"loss\": []}\n",
    "for amino_acid in results[\"orf_debug\"].keys():\n",
    "    results_df_dict[\"amino_acid\"].append(amino_acid)\n",
    "    selected_codon = results[\"orf\"][\"aa_to_optimal_codon\"][amino_acid]\n",
    "    results_df_dict[\"selected_codon\"].append(selected_codon)\n",
    "    results_df_dict[\"loss\"].append( results[\"orf_debug\"][amino_acid][selected_codon])\n",
    "    results_df_dict[\"bacillus_weight\"].append(results[\"user_input\"][\"organisms\"][0][\"cai_weights\"][selected_codon])\n",
    "    results_df_dict[\"ecoli_weight\"].append(results[\"user_input\"][\"organisms\"][1][\"cai_weights\"][selected_codon])\n",
    "results_df = pd.DataFrame(results_df_dict)\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "initial_seq = results[\"user_input\"][\"sequence\"]\n",
    "\n",
    "codons_stats = defaultdict(int)\n",
    "\n",
    "for i in range(0, len(initial_seq), 3):\n",
    "    codon = initial_seq[i:i+3]\n",
    "    codons_stats[codon] += 1\n",
    "initial_seq_df_dict = {\"codon\": [], \"count\": [], \"amino_acid\": [], \"bacillus_weight\": [], \"ecoli_weight\": [], \"loss\": []}\n",
    "\n",
    "for codon, amino_acid in shared_functions_and_vars.nt_to_aa.items():\n",
    "    initial_seq_df_dict[\"codon\"].append(codon)\n",
    "    initial_seq_df_dict[\"amino_acid\"].append(amino_acid)\n",
    "    # count = codons_stats.get(codon) or 0\n",
    "    initial_seq_df_dict[\"count\"].append(codons_stats[codon])\n",
    "    initial_seq_df_dict[\"bacillus_weight\"].append(results[\"user_input\"][\"organisms\"][0][\"cai_weights\"][codon])\n",
    "    initial_seq_df_dict[\"ecoli_weight\"].append(results[\"user_input\"][\"organisms\"][1][\"cai_weights\"][codon])\n",
    "    initial_seq_df_dict[\"loss\"].append(results[\"orf_debug\"][amino_acid][codon])\n",
    "initial_seq_df = pd.DataFrame(initial_seq_df_dict)\n",
    "initial_seq_df.sort_values(\"count\")\n",
    "initial_seq_df.nlargest(20, \"count\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_results_df = pd.merge(initial_seq_df, results_df, on='amino_acid', how='inner', suffixes=(\"\", \"_selected_codon\"))\n",
    "merged_results_df.nlargest(20, \"count\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO - parse the .out file instead. \n",
    "\n",
    "# e_coli_dict = {\"name\": \"Escherichia-coli\", \n",
    "#                \"AGC\": 0, \"GGC\": 2, \"CGC\": 0, \"TGC\": 3,\n",
    "#                \"ACC\": 0, \"GCC\": 4, \"CCC\": 1, \"TCC\": 1,\n",
    "#                \"AGG\": 0, \"GGG\": 1, \"CGG\": 1, \"TGG\": 1, \n",
    "#                \"AGT\": 0, \"GGT\": 2, \"CGT\": 2, \"TGT\": 1,\n",
    "#                \"AAC\": 0, }\n",
    "bacillus_dict = {\"name\": \"Bacillus-subtilis\", \"val\": 2}\n",
    "json_dict = [e_coli_dict, bacillus_dict]\n",
    "\n",
    "with open(\"test.json\", \"w\") as json_file:\n",
    "    json.dump(json_dict, json_file)\n",
    "\n",
    "df = pd.read_json(\"test.json\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "venv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
